{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "22cdfd58",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a99ab4f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "A = np.matrix([[1,2],[3,4]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "de82f5dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "A"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "169159b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "B = np.matrix('1,2;3,4')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "3b3505d1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "matrix([[1, 2],\n",
       "        [3, 4]])"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "B"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "83fe033b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "matrix([[ 7, 10],\n",
       "        [15, 22]])"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "A*B"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "2a0bf51a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "matrix([[-2. ,  1. ],\n",
       "        [ 1.5, -0.5]])"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "A.I"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "0ad4a3ce",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "matrix([[1, 3],\n",
       "        [2, 4]])"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "A.T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "d5d42edb",
   "metadata": {},
   "outputs": [],
   "source": [
    "C = np.asarray(A)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "46c3d24d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[1, 2],\n",
       "       [3, 4]])"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "C"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "aee4c6f1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on function isreal in module numpy:\n",
      "\n",
      "isreal(x)\n",
      "    Returns a bool array, where True if input element is real.\n",
      "    \n",
      "    If element has complex type with zero complex part, the return value\n",
      "    for that element is True.\n",
      "    \n",
      "    Parameters\n",
      "    ----------\n",
      "    x : array_like\n",
      "        Input array.\n",
      "    \n",
      "    Returns\n",
      "    -------\n",
      "    out : ndarray, bool\n",
      "        Boolean array of same shape as `x`.\n",
      "    \n",
      "    See Also\n",
      "    --------\n",
      "    iscomplex\n",
      "    isrealobj : Return True if x is not a complex type.\n",
      "    \n",
      "    Examples\n",
      "    --------\n",
      "    >>> np.isreal([1+1j, 1+0j, 4.5, 3, 2, 2j])\n",
      "    array([False,  True,  True,  True,  True, False])\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(np.isreal)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "aed86b13",
   "metadata": {},
   "outputs": [],
   "source": [
    "C = np.array([[1+1j, 1+0j], [4.5, 3j]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "9db303a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "C = np.asmatrix(C)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "5bb751ee",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "matrix([[1. +1.j, 1. +0.j],\n",
       "        [4.5+0.j, 0. +3.j]])"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "C"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "008429fd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "matrix([[1. -1.j, 4.5-0.j],\n",
       "        [1. -0.j, 0. -3.j]])"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "C.H"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "4a50c21f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scipy as sp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e2afa160",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on package scipy:\n",
      "\n",
      "NAME\n",
      "    scipy\n",
      "\n",
      "DESCRIPTION\n",
      "    SciPy: A scientific computing package for Python\n",
      "    ================================================\n",
      "    \n",
      "    Documentation is available in the docstrings and\n",
      "    online at https://docs.scipy.org.\n",
      "    \n",
      "    Contents\n",
      "    --------\n",
      "    SciPy imports all the functions from the NumPy namespace, and in\n",
      "    addition provides:\n",
      "    \n",
      "    Subpackages\n",
      "    -----------\n",
      "    Using any of these subpackages requires an explicit import. For example,\n",
      "    ``import scipy.cluster``.\n",
      "    \n",
      "    ::\n",
      "    \n",
      "     cluster                      --- Vector Quantization / Kmeans\n",
      "     fft                          --- Discrete Fourier transforms\n",
      "     fftpack                      --- Legacy discrete Fourier transforms\n",
      "     integrate                    --- Integration routines\n",
      "     interpolate                  --- Interpolation Tools\n",
      "     io                           --- Data input and output\n",
      "     linalg                       --- Linear algebra routines\n",
      "     linalg.blas                  --- Wrappers to BLAS library\n",
      "     linalg.lapack                --- Wrappers to LAPACK library\n",
      "     misc                         --- Various utilities that don't have\n",
      "                                      another home.\n",
      "     ndimage                      --- N-D image package\n",
      "     odr                          --- Orthogonal Distance Regression\n",
      "     optimize                     --- Optimization Tools\n",
      "     signal                       --- Signal Processing Tools\n",
      "     signal.windows               --- Window functions\n",
      "     sparse                       --- Sparse Matrices\n",
      "     sparse.linalg                --- Sparse Linear Algebra\n",
      "     sparse.linalg.dsolve         --- Linear Solvers\n",
      "     sparse.linalg.dsolve.umfpack --- :Interface to the UMFPACK library:\n",
      "                                      Conjugate Gradient Method (LOBPCG)\n",
      "     sparse.linalg.eigen          --- Sparse Eigenvalue Solvers\n",
      "     sparse.linalg.eigen.lobpcg   --- Locally Optimal Block Preconditioned\n",
      "                                      Conjugate Gradient Method (LOBPCG)\n",
      "     spatial                      --- Spatial data structures and algorithms\n",
      "     special                      --- Special functions\n",
      "     stats                        --- Statistical Functions\n",
      "    \n",
      "    Utility tools\n",
      "    -------------\n",
      "    ::\n",
      "    \n",
      "     test              --- Run scipy unittests\n",
      "     show_config       --- Show scipy build configuration\n",
      "     show_numpy_config --- Show numpy build configuration\n",
      "     __version__       --- SciPy version string\n",
      "     __numpy_version__ --- Numpy version string\n",
      "\n",
      "PACKAGE CONTENTS\n",
      "    __config__\n",
      "    _build_utils (package)\n",
      "    _distributor_init\n",
      "    _lib (package)\n",
      "    cluster (package)\n",
      "    conftest\n",
      "    constants (package)\n",
      "    fft (package)\n",
      "    fftpack (package)\n",
      "    integrate (package)\n",
      "    interpolate (package)\n",
      "    io (package)\n",
      "    linalg (package)\n",
      "    misc (package)\n",
      "    ndimage (package)\n",
      "    odr (package)\n",
      "    optimize (package)\n",
      "    setup\n",
      "    signal (package)\n",
      "    sparse (package)\n",
      "    spatial (package)\n",
      "    special (package)\n",
      "    stats (package)\n",
      "    version\n",
      "\n",
      "CLASSES\n",
      "    builtins.tuple(builtins.object)\n",
      "        scipy._lib._ccallback.LowLevelCallable\n",
      "    \n",
      "    class LowLevelCallable(builtins.tuple)\n",
      "     |  LowLevelCallable(function, user_data=None, signature=None)\n",
      "     |  \n",
      "     |  Low-level callback function.\n",
      "     |  \n",
      "     |  Parameters\n",
      "     |  ----------\n",
      "     |  function : {PyCapsule, ctypes function pointer, cffi function pointer}\n",
      "     |      Low-level callback function.\n",
      "     |  user_data : {PyCapsule, ctypes void pointer, cffi void pointer}\n",
      "     |      User data to pass on to the callback function.\n",
      "     |  signature : str, optional\n",
      "     |      Signature of the function. If omitted, determined from *function*,\n",
      "     |      if possible.\n",
      "     |  \n",
      "     |  Attributes\n",
      "     |  ----------\n",
      "     |  function\n",
      "     |      Callback function given.\n",
      "     |  user_data\n",
      "     |      User data given.\n",
      "     |  signature\n",
      "     |      Signature of the function.\n",
      "     |  \n",
      "     |  Methods\n",
      "     |  -------\n",
      "     |  from_cython\n",
      "     |      Class method for constructing callables from Cython C-exported\n",
      "     |      functions.\n",
      "     |  \n",
      "     |  Notes\n",
      "     |  -----\n",
      "     |  The argument ``function`` can be one of:\n",
      "     |  \n",
      "     |  - PyCapsule, whose name contains the C function signature\n",
      "     |  - ctypes function pointer\n",
      "     |  - cffi function pointer\n",
      "     |  \n",
      "     |  The signature of the low-level callback must match one of those expected\n",
      "     |  by the routine it is passed to.\n",
      "     |  \n",
      "     |  If constructing low-level functions from a PyCapsule, the name of the\n",
      "     |  capsule must be the corresponding signature, in the format::\n",
      "     |  \n",
      "     |      return_type (arg1_type, arg2_type, ...)\n",
      "     |  \n",
      "     |  For example::\n",
      "     |  \n",
      "     |      \"void (double)\"\n",
      "     |      \"double (double, int *, void *)\"\n",
      "     |  \n",
      "     |  The context of a PyCapsule passed in as ``function`` is used as ``user_data``,\n",
      "     |  if an explicit value for ``user_data`` was not given.\n",
      "     |  \n",
      "     |  Method resolution order:\n",
      "     |      LowLevelCallable\n",
      "     |      builtins.tuple\n",
      "     |      builtins.object\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __getitem__(self, idx)\n",
      "     |      Return self[key].\n",
      "     |  \n",
      "     |  __repr__(self)\n",
      "     |      Return repr(self).\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Class methods defined here:\n",
      "     |  \n",
      "     |  from_cython(module, name, user_data=None, signature=None) from builtins.type\n",
      "     |      Create a low-level callback function from an exported Cython function.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      module : module\n",
      "     |          Cython module where the exported function resides\n",
      "     |      name : str\n",
      "     |          Name of the exported function\n",
      "     |      user_data : {PyCapsule, ctypes void pointer, cffi void pointer}, optional\n",
      "     |          User data to pass on to the callback function.\n",
      "     |      signature : str, optional\n",
      "     |          Signature of the function. If omitted, determined from *function*.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Static methods defined here:\n",
      "     |  \n",
      "     |  __new__(cls, function, user_data=None, signature=None)\n",
      "     |      Create and return a new object.  See help(type) for accurate signature.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Readonly properties defined here:\n",
      "     |  \n",
      "     |  function\n",
      "     |  \n",
      "     |  signature\n",
      "     |  \n",
      "     |  user_data\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from builtins.tuple:\n",
      "     |  \n",
      "     |  __add__(self, value, /)\n",
      "     |      Return self+value.\n",
      "     |  \n",
      "     |  __contains__(self, key, /)\n",
      "     |      Return key in self.\n",
      "     |  \n",
      "     |  __eq__(self, value, /)\n",
      "     |      Return self==value.\n",
      "     |  \n",
      "     |  __ge__(self, value, /)\n",
      "     |      Return self>=value.\n",
      "     |  \n",
      "     |  __getattribute__(self, name, /)\n",
      "     |      Return getattr(self, name).\n",
      "     |  \n",
      "     |  __getnewargs__(self, /)\n",
      "     |  \n",
      "     |  __gt__(self, value, /)\n",
      "     |      Return self>value.\n",
      "     |  \n",
      "     |  __hash__(self, /)\n",
      "     |      Return hash(self).\n",
      "     |  \n",
      "     |  __iter__(self, /)\n",
      "     |      Implement iter(self).\n",
      "     |  \n",
      "     |  __le__(self, value, /)\n",
      "     |      Return self<=value.\n",
      "     |  \n",
      "     |  __len__(self, /)\n",
      "     |      Return len(self).\n",
      "     |  \n",
      "     |  __lt__(self, value, /)\n",
      "     |      Return self<value.\n",
      "     |  \n",
      "     |  __mul__(self, value, /)\n",
      "     |      Return self*value.\n",
      "     |  \n",
      "     |  __ne__(self, value, /)\n",
      "     |      Return self!=value.\n",
      "     |  \n",
      "     |  __rmul__(self, value, /)\n",
      "     |      Return value*self.\n",
      "     |  \n",
      "     |  count(self, value, /)\n",
      "     |      Return number of occurrences of value.\n",
      "     |  \n",
      "     |  index(self, value, start=0, stop=9223372036854775807, /)\n",
      "     |      Return first index of value.\n",
      "     |      \n",
      "     |      Raises ValueError if the value is not present.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Class methods inherited from builtins.tuple:\n",
      "     |  \n",
      "     |  __class_getitem__(...) from builtins.type\n",
      "     |      See PEP 585\n",
      "\n",
      "FUNCTIONS\n",
      "    show_config = show()\n",
      "        Show libraries in the system on which NumPy was built.\n",
      "        \n",
      "        Print information about various resources (libraries, library\n",
      "        directories, include directories, etc.) in the system on which\n",
      "        NumPy was built.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        get_include : Returns the directory containing NumPy C\n",
      "                      header files.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        Classes specifying the information to be printed are defined\n",
      "        in the `numpy.distutils.system_info` module.\n",
      "        \n",
      "        Information may include:\n",
      "        \n",
      "        * ``language``: language used to write the libraries (mostly\n",
      "          C or f77)\n",
      "        * ``libraries``: names of libraries found in the system\n",
      "        * ``library_dirs``: directories containing the libraries\n",
      "        * ``include_dirs``: directories containing library header files\n",
      "        * ``src_dirs``: directories containing library source files\n",
      "        * ``define_macros``: preprocessor macros used by\n",
      "          ``distutils.setup``\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> np.show_config()\n",
      "        blas_opt_info:\n",
      "            language = c\n",
      "            define_macros = [('HAVE_CBLAS', None)]\n",
      "            libraries = ['openblas', 'openblas']\n",
      "            library_dirs = ['/usr/local/lib']\n",
      "\n",
      "DATA\n",
      "    __numpy_version__ = '1.21.5'\n",
      "    test = <scipy._lib._testutils.PytestTester object>\n",
      "\n",
      "VERSION\n",
      "    1.9.3\n",
      "\n",
      "FILE\n",
      "    c:\\users\\bsrvprasad\\anaconda3\\lib\\site-packages\\scipy\\__init__.py\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(sp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "aee8eb5f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scipy.constants as sc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "6d92b176",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on package scipy.constants in scipy:\n",
      "\n",
      "NAME\n",
      "    scipy.constants\n",
      "\n",
      "DESCRIPTION\n",
      "    ==================================\n",
      "    Constants (:mod:`scipy.constants`)\n",
      "    ==================================\n",
      "    \n",
      "    .. currentmodule:: scipy.constants\n",
      "    \n",
      "    Physical and mathematical constants and units.\n",
      "    \n",
      "    \n",
      "    Mathematical constants\n",
      "    ======================\n",
      "    \n",
      "    ================  =================================================================\n",
      "    ``pi``            Pi\n",
      "    ``golden``        Golden ratio\n",
      "    ``golden_ratio``  Golden ratio\n",
      "    ================  =================================================================\n",
      "    \n",
      "    \n",
      "    Physical constants\n",
      "    ==================\n",
      "    \n",
      "    ===========================  =================================================================\n",
      "    ``c``                        speed of light in vacuum\n",
      "    ``speed_of_light``           speed of light in vacuum\n",
      "    ``mu_0``                     the magnetic constant :math:`\\mu_0`\n",
      "    ``epsilon_0``                the electric constant (vacuum permittivity), :math:`\\epsilon_0`\n",
      "    ``h``                        the Planck constant :math:`h`\n",
      "    ``Planck``                   the Planck constant :math:`h`\n",
      "    ``hbar``                     :math:`\\hbar = h/(2\\pi)`\n",
      "    ``G``                        Newtonian constant of gravitation\n",
      "    ``gravitational_constant``   Newtonian constant of gravitation\n",
      "    ``g``                        standard acceleration of gravity\n",
      "    ``e``                        elementary charge\n",
      "    ``elementary_charge``        elementary charge\n",
      "    ``R``                        molar gas constant\n",
      "    ``gas_constant``             molar gas constant\n",
      "    ``alpha``                    fine-structure constant\n",
      "    ``fine_structure``           fine-structure constant\n",
      "    ``N_A``                      Avogadro constant\n",
      "    ``Avogadro``                 Avogadro constant\n",
      "    ``k``                        Boltzmann constant\n",
      "    ``Boltzmann``                Boltzmann constant\n",
      "    ``sigma``                    Stefan-Boltzmann constant :math:`\\sigma`\n",
      "    ``Stefan_Boltzmann``         Stefan-Boltzmann constant :math:`\\sigma`\n",
      "    ``Wien``                     Wien displacement law constant\n",
      "    ``Rydberg``                  Rydberg constant\n",
      "    ``m_e``                      electron mass\n",
      "    ``electron_mass``            electron mass\n",
      "    ``m_p``                      proton mass\n",
      "    ``proton_mass``              proton mass\n",
      "    ``m_n``                      neutron mass\n",
      "    ``neutron_mass``             neutron mass\n",
      "    ===========================  =================================================================\n",
      "    \n",
      "    \n",
      "    Constants database\n",
      "    ------------------\n",
      "    \n",
      "    In addition to the above variables, :mod:`scipy.constants` also contains the\n",
      "    2018 CODATA recommended values [CODATA2018]_ database containing more physical\n",
      "    constants.\n",
      "    \n",
      "    .. autosummary::\n",
      "       :toctree: generated/\n",
      "    \n",
      "       value      -- Value in physical_constants indexed by key\n",
      "       unit       -- Unit in physical_constants indexed by key\n",
      "       precision  -- Relative precision in physical_constants indexed by key\n",
      "       find       -- Return list of physical_constant keys with a given string\n",
      "       ConstantWarning -- Constant sought not in newest CODATA data set\n",
      "    \n",
      "    .. data:: physical_constants\n",
      "    \n",
      "       Dictionary of physical constants, of the format\n",
      "       ``physical_constants[name] = (value, unit, uncertainty)``.\n",
      "    \n",
      "    Available constants:\n",
      "    \n",
      "    ======================================================================  ====\n",
      "    ``alpha particle mass``                                                 6.6446573357e-27 kg\n",
      "    ``alpha particle mass energy equivalent``                               5.9719201914e-10 J\n",
      "    ``alpha particle mass energy equivalent in MeV``                        3727.3794066 MeV\n",
      "    ``alpha particle mass in u``                                            4.001506179127 u\n",
      "    ``alpha particle molar mass``                                           0.0040015061777 kg mol^-1\n",
      "    ``alpha particle relative atomic mass``                                 4.001506179127 \n",
      "    ``alpha particle-electron mass ratio``                                  7294.29954142 \n",
      "    ``alpha particle-proton mass ratio``                                    3.97259969009 \n",
      "    ``Angstrom star``                                                       1.00001495e-10 m\n",
      "    ``atomic mass constant``                                                1.6605390666e-27 kg\n",
      "    ``atomic mass constant energy equivalent``                              1.4924180856e-10 J\n",
      "    ``atomic mass constant energy equivalent in MeV``                       931.49410242 MeV\n",
      "    ``atomic mass unit-electron volt relationship``                         931494102.42 eV\n",
      "    ``atomic mass unit-hartree relationship``                               34231776.874 E_h\n",
      "    ``atomic mass unit-hertz relationship``                                 2.25234271871e+23 Hz\n",
      "    ``atomic mass unit-inverse meter relationship``                         751300661040000.0 m^-1\n",
      "    ``atomic mass unit-joule relationship``                                 1.4924180856e-10 J\n",
      "    ``atomic mass unit-kelvin relationship``                                10809540191600.0 K\n",
      "    ``atomic mass unit-kilogram relationship``                              1.6605390666e-27 kg\n",
      "    ``atomic unit of 1st hyperpolarizability``                              3.2063613061e-53 C^3 m^3 J^-2\n",
      "    ``atomic unit of 2nd hyperpolarizability``                              6.2353799905e-65 C^4 m^4 J^-3\n",
      "    ``atomic unit of action``                                               1.054571817e-34 J s\n",
      "    ``atomic unit of charge``                                               1.602176634e-19 C\n",
      "    ``atomic unit of charge density``                                       1081202384570.0 C m^-3\n",
      "    ``atomic unit of current``                                              0.00662361823751 A\n",
      "    ``atomic unit of electric dipole mom.``                                 8.4783536255e-30 C m\n",
      "    ``atomic unit of electric field``                                       514220674763.0 V m^-1\n",
      "    ``atomic unit of electric field gradient``                              9.7173624292e+21 V m^-2\n",
      "    ``atomic unit of electric polarizability``                              1.64877727436e-41 C^2 m^2 J^-1\n",
      "    ``atomic unit of electric potential``                                   27.211386245988 V\n",
      "    ``atomic unit of electric quadrupole mom.``                             4.4865515246e-40 C m^2\n",
      "    ``atomic unit of energy``                                               4.3597447222071e-18 J\n",
      "    ``atomic unit of force``                                                8.2387234983e-08 N\n",
      "    ``atomic unit of length``                                               5.29177210903e-11 m\n",
      "    ``atomic unit of mag. dipole mom.``                                     1.85480201566e-23 J T^-1\n",
      "    ``atomic unit of mag. flux density``                                    235051.756758 T\n",
      "    ``atomic unit of magnetizability``                                      7.8910366008e-29 J T^-2\n",
      "    ``atomic unit of mass``                                                 9.1093837015e-31 kg\n",
      "    ``atomic unit of momentum``                                             1.9928519141e-24 kg m s^-1\n",
      "    ``atomic unit of permittivity``                                         1.11265005545e-10 F m^-1\n",
      "    ``atomic unit of time``                                                 2.4188843265857e-17 s\n",
      "    ``atomic unit of velocity``                                             2187691.26364 m s^-1\n",
      "    ``Avogadro constant``                                                   6.02214076e+23 mol^-1\n",
      "    ``Bohr magneton``                                                       9.2740100783e-24 J T^-1\n",
      "    ``Bohr magneton in eV/T``                                               5.788381806e-05 eV T^-1\n",
      "    ``Bohr magneton in Hz/T``                                               13996244936.1 Hz T^-1\n",
      "    ``Bohr magneton in inverse meter per tesla``                            46.686447783 m^-1 T^-1\n",
      "    ``Bohr magneton in K/T``                                                0.67171381563 K T^-1\n",
      "    ``Bohr radius``                                                         5.29177210903e-11 m\n",
      "    ``Boltzmann constant``                                                  1.380649e-23 J K^-1\n",
      "    ``Boltzmann constant in eV/K``                                          8.617333262e-05 eV K^-1\n",
      "    ``Boltzmann constant in Hz/K``                                          20836619120.0 Hz K^-1\n",
      "    ``Boltzmann constant in inverse meter per kelvin``                      69.50348004 m^-1 K^-1\n",
      "    ``classical electron radius``                                           2.8179403262e-15 m\n",
      "    ``Compton wavelength``                                                  2.42631023867e-12 m\n",
      "    ``conductance quantum``                                                 7.748091729e-05 S\n",
      "    ``conventional value of ampere-90``                                     1.00000008887 A\n",
      "    ``conventional value of coulomb-90``                                    1.00000008887 C\n",
      "    ``conventional value of farad-90``                                      0.9999999822 F\n",
      "    ``conventional value of henry-90``                                      1.00000001779 H\n",
      "    ``conventional value of Josephson constant``                            483597900000000.0 Hz V^-1\n",
      "    ``conventional value of ohm-90``                                        1.00000001779 ohm\n",
      "    ``conventional value of volt-90``                                       1.00000010666 V\n",
      "    ``conventional value of von Klitzing constant``                         25812.807 ohm\n",
      "    ``conventional value of watt-90``                                       1.00000019553 W\n",
      "    ``Cu x unit``                                                           1.00207697e-13 m\n",
      "    ``deuteron g factor``                                                   0.8574382338 \n",
      "    ``deuteron mag. mom.``                                                  4.330735094e-27 J T^-1\n",
      "    ``deuteron mag. mom. to Bohr magneton ratio``                           0.000466975457 \n",
      "    ``deuteron mag. mom. to nuclear magneton ratio``                        0.8574382338 \n",
      "    ``deuteron mass``                                                       3.3435837724e-27 kg\n",
      "    ``deuteron mass energy equivalent``                                     3.00506323102e-10 J\n",
      "    ``deuteron mass energy equivalent in MeV``                              1875.61294257 MeV\n",
      "    ``deuteron mass in u``                                                  2.013553212745 u\n",
      "    ``deuteron molar mass``                                                 0.00201355321205 kg mol^-1\n",
      "    ``deuteron relative atomic mass``                                       2.013553212745 \n",
      "    ``deuteron rms charge radius``                                          2.12799e-15 m\n",
      "    ``deuteron-electron mag. mom. ratio``                                   -0.0004664345551 \n",
      "    ``deuteron-electron mass ratio``                                        3670.48296788 \n",
      "    ``deuteron-neutron mag. mom. ratio``                                    -0.44820653 \n",
      "    ``deuteron-proton mag. mom. ratio``                                     0.30701220939 \n",
      "    ``deuteron-proton mass ratio``                                          1.99900750139 \n",
      "    ``electron charge to mass quotient``                                    -175882001076.0 C kg^-1\n",
      "    ``electron g factor``                                                   -2.00231930436256 \n",
      "    ``electron gyromag. ratio``                                             176085963023.0 s^-1 T^-1\n",
      "    ``electron gyromag. ratio in MHz/T``                                    28024.9514242 MHz T^-1\n",
      "    ``electron mag. mom.``                                                  -9.2847647043e-24 J T^-1\n",
      "    ``electron mag. mom. anomaly``                                          0.00115965218128 \n",
      "    ``electron mag. mom. to Bohr magneton ratio``                           -1.00115965218128 \n",
      "    ``electron mag. mom. to nuclear magneton ratio``                        -1838.28197188 \n",
      "    ``electron mass``                                                       9.1093837015e-31 kg\n",
      "    ``electron mass energy equivalent``                                     8.1871057769e-14 J\n",
      "    ``electron mass energy equivalent in MeV``                              0.51099895 MeV\n",
      "    ``electron mass in u``                                                  0.000548579909065 u\n",
      "    ``electron molar mass``                                                 5.4857990888e-07 kg mol^-1\n",
      "    ``electron relative atomic mass``                                       0.000548579909065 \n",
      "    ``electron to alpha particle mass ratio``                               0.0001370933554787 \n",
      "    ``electron to shielded helion mag. mom. ratio``                         864.058257 \n",
      "    ``electron to shielded proton mag. mom. ratio``                         -658.2275971 \n",
      "    ``electron volt``                                                       1.602176634e-19 J\n",
      "    ``electron volt-atomic mass unit relationship``                         1.07354410233e-09 u\n",
      "    ``electron volt-hartree relationship``                                  0.036749322175655 E_h\n",
      "    ``electron volt-hertz relationship``                                    241798924200000.0 Hz\n",
      "    ``electron volt-inverse meter relationship``                            806554.3937 m^-1\n",
      "    ``electron volt-joule relationship``                                    1.602176634e-19 J\n",
      "    ``electron volt-kelvin relationship``                                   11604.51812 K\n",
      "    ``electron volt-kilogram relationship``                                 1.782661921e-36 kg\n",
      "    ``electron-deuteron mag. mom. ratio``                                   -2143.9234915 \n",
      "    ``electron-deuteron mass ratio``                                        0.0002724437107462 \n",
      "    ``electron-helion mass ratio``                                          0.0001819543074573 \n",
      "    ``electron-muon mag. mom. ratio``                                       206.7669883 \n",
      "    ``electron-muon mass ratio``                                            0.00483633169 \n",
      "    ``electron-neutron mag. mom. ratio``                                    960.9205 \n",
      "    ``electron-neutron mass ratio``                                         0.00054386734424 \n",
      "    ``electron-proton mag. mom. ratio``                                     -658.21068789 \n",
      "    ``electron-proton mass ratio``                                          0.000544617021487 \n",
      "    ``electron-tau mass ratio``                                             0.000287585 \n",
      "    ``electron-triton mass ratio``                                          0.0001819200062251 \n",
      "    ``elementary charge``                                                   1.602176634e-19 C\n",
      "    ``elementary charge over h-bar``                                        1519267447000000.0 A J^-1\n",
      "    ``Faraday constant``                                                    96485.33212 C mol^-1\n",
      "    ``Fermi coupling constant``                                             1.1663787e-05 GeV^-2\n",
      "    ``fine-structure constant``                                             0.0072973525693 \n",
      "    ``first radiation constant``                                            3.741771852e-16 W m^2\n",
      "    ``first radiation constant for spectral radiance``                      1.191042972e-16 W m^2 sr^-1\n",
      "    ``Hartree energy``                                                      4.3597447222071e-18 J\n",
      "    ``Hartree energy in eV``                                                27.211386245988 eV\n",
      "    ``hartree-atomic mass unit relationship``                               2.92126232205e-08 u\n",
      "    ``hartree-electron volt relationship``                                  27.211386245988 eV\n",
      "    ``hartree-hertz relationship``                                          6579683920502000.0 Hz\n",
      "    ``hartree-inverse meter relationship``                                  21947463.13632 m^-1\n",
      "    ``hartree-joule relationship``                                          4.3597447222071e-18 J\n",
      "    ``hartree-kelvin relationship``                                         315775.02480407 K\n",
      "    ``hartree-kilogram relationship``                                       4.8508702095432e-35 kg\n",
      "    ``helion g factor``                                                     -4.255250615 \n",
      "    ``helion mag. mom.``                                                    -1.074617532e-26 J T^-1\n",
      "    ``helion mag. mom. to Bohr magneton ratio``                             -0.001158740958 \n",
      "    ``helion mag. mom. to nuclear magneton ratio``                          -2.127625307 \n",
      "    ``helion mass``                                                         5.0064127796e-27 kg\n",
      "    ``helion mass energy equivalent``                                       4.4995394125e-10 J\n",
      "    ``helion mass energy equivalent in MeV``                                2808.39160743 MeV\n",
      "    ``helion mass in u``                                                    3.014932247175 u\n",
      "    ``helion molar mass``                                                   0.00301493224613 kg mol^-1\n",
      "    ``helion relative atomic mass``                                         3.014932247175 \n",
      "    ``helion shielding shift``                                              5.996743e-05 \n",
      "    ``helion-electron mass ratio``                                          5495.88528007 \n",
      "    ``helion-proton mass ratio``                                            2.99315267167 \n",
      "    ``hertz-atomic mass unit relationship``                                 4.4398216652e-24 u\n",
      "    ``hertz-electron volt relationship``                                    4.135667696e-15 eV\n",
      "    ``hertz-hartree relationship``                                          1.519829846057e-16 E_h\n",
      "    ``hertz-inverse meter relationship``                                    3.3356409519815204e-09 m^-1\n",
      "    ``hertz-joule relationship``                                            6.62607015e-34 J\n",
      "    ``hertz-kelvin relationship``                                           4.799243073e-11 K\n",
      "    ``hertz-kilogram relationship``                                         7.372497323e-51 kg\n",
      "    ``hyperfine transition frequency of Cs-133``                            9192631770.0 Hz\n",
      "    ``inverse fine-structure constant``                                     137.035999084 \n",
      "    ``inverse meter-atomic mass unit relationship``                         1.3310250501e-15 u\n",
      "    ``inverse meter-electron volt relationship``                            1.239841984e-06 eV\n",
      "    ``inverse meter-hartree relationship``                                  4.556335252912e-08 E_h\n",
      "    ``inverse meter-hertz relationship``                                    299792458.0 Hz\n",
      "    ``inverse meter-joule relationship``                                    1.986445857e-25 J\n",
      "    ``inverse meter-kelvin relationship``                                   0.01438776877 K\n",
      "    ``inverse meter-kilogram relationship``                                 2.210219094e-42 kg\n",
      "    ``inverse of conductance quantum``                                      12906.40372 ohm\n",
      "    ``Josephson constant``                                                  483597848400000.0 Hz V^-1\n",
      "    ``joule-atomic mass unit relationship``                                 6700535256.5 u\n",
      "    ``joule-electron volt relationship``                                    6.241509074e+18 eV\n",
      "    ``joule-hartree relationship``                                          2.2937122783963e+17 E_h\n",
      "    ``joule-hertz relationship``                                            1.509190179e+33 Hz\n",
      "    ``joule-inverse meter relationship``                                    5.034116567e+24 m^-1\n",
      "    ``joule-kelvin relationship``                                           7.242970516e+22 K\n",
      "    ``joule-kilogram relationship``                                         1.1126500560536185e-17 kg\n",
      "    ``kelvin-atomic mass unit relationship``                                9.2510873014e-14 u\n",
      "    ``kelvin-electron volt relationship``                                   8.617333262e-05 eV\n",
      "    ``kelvin-hartree relationship``                                         3.1668115634556e-06 E_h\n",
      "    ``kelvin-hertz relationship``                                           20836619120.0 Hz\n",
      "    ``kelvin-inverse meter relationship``                                   69.50348004 m^-1\n",
      "    ``kelvin-joule relationship``                                           1.380649e-23 J\n",
      "    ``kelvin-kilogram relationship``                                        1.536179187e-40 kg\n",
      "    ``kilogram-atomic mass unit relationship``                              6.0221407621e+26 u\n",
      "    ``kilogram-electron volt relationship``                                 5.609588603e+35 eV\n",
      "    ``kilogram-hartree relationship``                                       2.0614857887409e+34 E_h\n",
      "    ``kilogram-hertz relationship``                                         1.356392489e+50 Hz\n",
      "    ``kilogram-inverse meter relationship``                                 4.524438335e+41 m^-1\n",
      "    ``kilogram-joule relationship``                                         8.987551787368176e+16 J\n",
      "    ``kilogram-kelvin relationship``                                        6.50965726e+39 K\n",
      "    ``lattice parameter of silicon``                                        5.431020511e-10 m\n",
      "    ``lattice spacing of ideal Si (220)``                                   1.920155716e-10 m\n",
      "    ``Loschmidt constant (273.15 K, 100 kPa)``                              2.651645804e+25 m^-3\n",
      "    ``Loschmidt constant (273.15 K, 101.325 kPa)``                          2.686780111e+25 m^-3\n",
      "    ``luminous efficacy``                                                   683.0 lm W^-1\n",
      "    ``mag. flux quantum``                                                   2.067833848e-15 Wb\n",
      "    ``Mo x unit``                                                           1.00209952e-13 m\n",
      "    ``molar gas constant``                                                  8.314462618 J mol^-1 K^-1\n",
      "    ``molar mass constant``                                                 0.00099999999965 kg mol^-1\n",
      "    ``molar mass of carbon-12``                                             0.0119999999958 kg mol^-1\n",
      "    ``molar Planck constant``                                               3.990312712e-10 J Hz^-1 mol^-1\n",
      "    ``molar volume of ideal gas (273.15 K, 100 kPa)``                       0.02271095464 m^3 mol^-1\n",
      "    ``molar volume of ideal gas (273.15 K, 101.325 kPa)``                   0.02241396954 m^3 mol^-1\n",
      "    ``molar volume of silicon``                                             1.205883199e-05 m^3 mol^-1\n",
      "    ``muon Compton wavelength``                                             1.17344411e-14 m\n",
      "    ``muon g factor``                                                       -2.0023318418 \n",
      "    ``muon mag. mom.``                                                      -4.4904483e-26 J T^-1\n",
      "    ``muon mag. mom. anomaly``                                              0.00116592089 \n",
      "    ``muon mag. mom. to Bohr magneton ratio``                               -0.00484197047 \n",
      "    ``muon mag. mom. to nuclear magneton ratio``                            -8.89059703 \n",
      "    ``muon mass``                                                           1.883531627e-28 kg\n",
      "    ``muon mass energy equivalent``                                         1.692833804e-11 J\n",
      "    ``muon mass energy equivalent in MeV``                                  105.6583755 MeV\n",
      "    ``muon mass in u``                                                      0.1134289259 u\n",
      "    ``muon molar mass``                                                     0.0001134289259 kg mol^-1\n",
      "    ``muon-electron mass ratio``                                            206.768283 \n",
      "    ``muon-neutron mass ratio``                                             0.112454517 \n",
      "    ``muon-proton mag. mom. ratio``                                         -3.183345142 \n",
      "    ``muon-proton mass ratio``                                              0.1126095264 \n",
      "    ``muon-tau mass ratio``                                                 0.0594635 \n",
      "    ``natural unit of action``                                              1.054571817e-34 J s\n",
      "    ``natural unit of action in eV s``                                      6.582119569e-16 eV s\n",
      "    ``natural unit of energy``                                              8.1871057769e-14 J\n",
      "    ``natural unit of energy in MeV``                                       0.51099895 MeV\n",
      "    ``natural unit of length``                                              3.8615926796e-13 m\n",
      "    ``natural unit of mass``                                                9.1093837015e-31 kg\n",
      "    ``natural unit of momentum``                                            2.730924488e-22 kg m s^-1\n",
      "    ``natural unit of momentum in MeV/c``                                   0.5109989461 MeV/c\n",
      "    ``natural unit of time``                                                1.28808866819e-21 s\n",
      "    ``natural unit of velocity``                                            299792458.0 m s^-1\n",
      "    ``neutron Compton wavelength``                                          1.31959090581e-15 m\n",
      "    ``neutron g factor``                                                    -3.82608545 \n",
      "    ``neutron gyromag. ratio``                                              183247171.0 s^-1 T^-1\n",
      "    ``neutron gyromag. ratio in MHz/T``                                     29.1646931 MHz T^-1\n",
      "    ``neutron mag. mom.``                                                   -9.6623651e-27 J T^-1\n",
      "    ``neutron mag. mom. to Bohr magneton ratio``                            -0.00104187563 \n",
      "    ``neutron mag. mom. to nuclear magneton ratio``                         -1.91304273 \n",
      "    ``neutron mass``                                                        1.67492749804e-27 kg\n",
      "    ``neutron mass energy equivalent``                                      1.50534976287e-10 J\n",
      "    ``neutron mass energy equivalent in MeV``                               939.56542052 MeV\n",
      "    ``neutron mass in u``                                                   1.00866491595 u\n",
      "    ``neutron molar mass``                                                  0.0010086649156 kg mol^-1\n",
      "    ``neutron relative atomic mass``                                        1.00866491595 \n",
      "    ``neutron to shielded proton mag. mom. ratio``                          -0.68499694 \n",
      "    ``neutron-electron mag. mom. ratio``                                    0.00104066882 \n",
      "    ``neutron-electron mass ratio``                                         1838.68366173 \n",
      "    ``neutron-muon mass ratio``                                             8.89248406 \n",
      "    ``neutron-proton mag. mom. ratio``                                      -0.68497934 \n",
      "    ``neutron-proton mass difference``                                      2.30557435e-30 kg\n",
      "    ``neutron-proton mass difference energy equivalent``                    2.07214689e-13 J\n",
      "    ``neutron-proton mass difference energy equivalent in MeV``             1.29333236 MeV\n",
      "    ``neutron-proton mass difference in u``                                 0.00138844933 u\n",
      "    ``neutron-proton mass ratio``                                           1.00137841931 \n",
      "    ``neutron-tau mass ratio``                                              0.528779 \n",
      "    ``Newtonian constant of gravitation``                                   6.6743e-11 m^3 kg^-1 s^-2\n",
      "    ``Newtonian constant of gravitation over h-bar c``                      6.70883e-39 (GeV/c^2)^-2\n",
      "    ``nuclear magneton``                                                    5.0507837461e-27 J T^-1\n",
      "    ``nuclear magneton in eV/T``                                            3.15245125844e-08 eV T^-1\n",
      "    ``nuclear magneton in inverse meter per tesla``                         0.0254262341353 m^-1 T^-1\n",
      "    ``nuclear magneton in K/T``                                             0.00036582677756 K T^-1\n",
      "    ``nuclear magneton in MHz/T``                                           7.6225932291 MHz T^-1\n",
      "    ``Planck constant``                                                     6.62607015e-34 J Hz^-1\n",
      "    ``Planck constant in eV/Hz``                                            4.135667696e-15 eV Hz^-1\n",
      "    ``Planck length``                                                       1.616255e-35 m\n",
      "    ``Planck mass``                                                         2.176434e-08 kg\n",
      "    ``Planck mass energy equivalent in GeV``                                1.22089e+19 GeV\n",
      "    ``Planck temperature``                                                  1.416784e+32 K\n",
      "    ``Planck time``                                                         5.391247e-44 s\n",
      "    ``proton charge to mass quotient``                                      95788331.56 C kg^-1\n",
      "    ``proton Compton wavelength``                                           1.32140985539e-15 m\n",
      "    ``proton g factor``                                                     5.5856946893 \n",
      "    ``proton gyromag. ratio``                                               267522187.44 s^-1 T^-1\n",
      "    ``proton gyromag. ratio in MHz/T``                                      42.577478518 MHz T^-1\n",
      "    ``proton mag. mom.``                                                    1.41060679736e-26 J T^-1\n",
      "    ``proton mag. mom. to Bohr magneton ratio``                             0.0015210322023 \n",
      "    ``proton mag. mom. to nuclear magneton ratio``                          2.79284734463 \n",
      "    ``proton mag. shielding correction``                                    2.5689e-05 \n",
      "    ``proton mass``                                                         1.67262192369e-27 kg\n",
      "    ``proton mass energy equivalent``                                       1.50327761598e-10 J\n",
      "    ``proton mass energy equivalent in MeV``                                938.27208816 MeV\n",
      "    ``proton mass in u``                                                    1.007276466621 u\n",
      "    ``proton molar mass``                                                   0.00100727646627 kg mol^-1\n",
      "    ``proton relative atomic mass``                                         1.007276466621 \n",
      "    ``proton rms charge radius``                                            8.414e-16 m\n",
      "    ``proton-electron mass ratio``                                          1836.15267343 \n",
      "    ``proton-muon mass ratio``                                              8.88024337 \n",
      "    ``proton-neutron mag. mom. ratio``                                      -1.45989805 \n",
      "    ``proton-neutron mass ratio``                                           0.99862347812 \n",
      "    ``proton-tau mass ratio``                                               0.528051 \n",
      "    ``quantum of circulation``                                              0.00036369475516 m^2 s^-1\n",
      "    ``quantum of circulation times 2``                                      0.00072738951032 m^2 s^-1\n",
      "    ``reduced Compton wavelength``                                          3.8615926796e-13 m\n",
      "    ``reduced muon Compton wavelength``                                     1.867594306e-15 m\n",
      "    ``reduced neutron Compton wavelength``                                  2.1001941552e-16 m\n",
      "    ``reduced Planck constant``                                             1.054571817e-34 J s\n",
      "    ``reduced Planck constant in eV s``                                     6.582119569e-16 eV s\n",
      "    ``reduced Planck constant times c in MeV fm``                           197.3269804 MeV fm\n",
      "    ``reduced proton Compton wavelength``                                   2.10308910336e-16 m\n",
      "    ``reduced tau Compton wavelength``                                      1.110538e-16 m\n",
      "    ``Rydberg constant``                                                    10973731.56816 m^-1\n",
      "    ``Rydberg constant times c in Hz``                                      3289841960250800.0 Hz\n",
      "    ``Rydberg constant times hc in eV``                                     13.605693122994 eV\n",
      "    ``Rydberg constant times hc in J``                                      2.1798723611035e-18 J\n",
      "    ``Sackur-Tetrode constant (1 K, 100 kPa)``                              -1.15170753706 \n",
      "    ``Sackur-Tetrode constant (1 K, 101.325 kPa)``                          -1.16487052358 \n",
      "    ``second radiation constant``                                           0.01438776877 m K\n",
      "    ``shielded helion gyromag. ratio``                                      203789456.9 s^-1 T^-1\n",
      "    ``shielded helion gyromag. ratio in MHz/T``                             32.43409942 MHz T^-1\n",
      "    ``shielded helion mag. mom.``                                           -1.07455309e-26 J T^-1\n",
      "    ``shielded helion mag. mom. to Bohr magneton ratio``                    -0.001158671471 \n",
      "    ``shielded helion mag. mom. to nuclear magneton ratio``                 -2.127497719 \n",
      "    ``shielded helion to proton mag. mom. ratio``                           -0.7617665618 \n",
      "    ``shielded helion to shielded proton mag. mom. ratio``                  -0.7617861313 \n",
      "    ``shielded proton gyromag. ratio``                                      267515315.1 s^-1 T^-1\n",
      "    ``shielded proton gyromag. ratio in MHz/T``                             42.57638474 MHz T^-1\n",
      "    ``shielded proton mag. mom.``                                           1.41057056e-26 J T^-1\n",
      "    ``shielded proton mag. mom. to Bohr magneton ratio``                    0.001520993128 \n",
      "    ``shielded proton mag. mom. to nuclear magneton ratio``                 2.792775599 \n",
      "    ``shielding difference of d and p in HD``                               2.02e-08 \n",
      "    ``shielding difference of t and p in HT``                               2.414e-08 \n",
      "    ``speed of light in vacuum``                                            299792458.0 m s^-1\n",
      "    ``standard acceleration of gravity``                                    9.80665 m s^-2\n",
      "    ``standard atmosphere``                                                 101325.0 Pa\n",
      "    ``standard-state pressure``                                             100000.0 Pa\n",
      "    ``Stefan-Boltzmann constant``                                           5.670374419e-08 W m^-2 K^-4\n",
      "    ``tau Compton wavelength``                                              6.97771e-16 m\n",
      "    ``tau energy equivalent``                                               1776.86 MeV\n",
      "    ``tau mass``                                                            3.16754e-27 kg\n",
      "    ``tau mass energy equivalent``                                          2.84684e-10 J\n",
      "    ``tau mass in u``                                                       1.90754 u\n",
      "    ``tau molar mass``                                                      0.00190754 kg mol^-1\n",
      "    ``tau-electron mass ratio``                                             3477.23 \n",
      "    ``tau-muon mass ratio``                                                 16.817 \n",
      "    ``tau-neutron mass ratio``                                              1.89115 \n",
      "    ``tau-proton mass ratio``                                               1.89376 \n",
      "    ``Thomson cross section``                                               6.6524587321e-29 m^2\n",
      "    ``triton g factor``                                                     5.957924931 \n",
      "    ``triton mag. mom.``                                                    1.5046095202e-26 J T^-1\n",
      "    ``triton mag. mom. to Bohr magneton ratio``                             0.0016223936651 \n",
      "    ``triton mag. mom. to nuclear magneton ratio``                          2.9789624656 \n",
      "    ``triton mass``                                                         5.0073567446e-27 kg\n",
      "    ``triton mass energy equivalent``                                       4.500387806e-10 J\n",
      "    ``triton mass energy equivalent in MeV``                                2808.92113298 MeV\n",
      "    ``triton mass in u``                                                    3.01550071621 u\n",
      "    ``triton molar mass``                                                   0.00301550071517 kg mol^-1\n",
      "    ``triton relative atomic mass``                                         3.01550071621 \n",
      "    ``triton to proton mag. mom. ratio``                                    1.0666399191 \n",
      "    ``triton-electron mass ratio``                                          5496.92153573 \n",
      "    ``triton-proton mass ratio``                                            2.99371703414 \n",
      "    ``unified atomic mass unit``                                            1.6605390666e-27 kg\n",
      "    ``vacuum electric permittivity``                                        8.8541878128e-12 F m^-1\n",
      "    ``vacuum mag. permeability``                                            1.25663706212e-06 N A^-2\n",
      "    ``von Klitzing constant``                                               25812.80745 ohm\n",
      "    ``W to Z mass ratio``                                                   0.88153 \n",
      "    ``weak mixing angle``                                                   0.2229 \n",
      "    ``Wien frequency displacement law constant``                            58789257570.0 Hz K^-1\n",
      "    ``Wien wavelength displacement law constant``                           0.002897771955 m K\n",
      "    ======================================================================  ====\n",
      "    \n",
      "    \n",
      "    Units\n",
      "    =====\n",
      "    \n",
      "    SI prefixes\n",
      "    -----------\n",
      "    \n",
      "    ============  =================================================================\n",
      "    ``yotta``     :math:`10^{24}`\n",
      "    ``zetta``     :math:`10^{21}`\n",
      "    ``exa``       :math:`10^{18}`\n",
      "    ``peta``      :math:`10^{15}`\n",
      "    ``tera``      :math:`10^{12}`\n",
      "    ``giga``      :math:`10^{9}`\n",
      "    ``mega``      :math:`10^{6}`\n",
      "    ``kilo``      :math:`10^{3}`\n",
      "    ``hecto``     :math:`10^{2}`\n",
      "    ``deka``      :math:`10^{1}`\n",
      "    ``deci``      :math:`10^{-1}`\n",
      "    ``centi``     :math:`10^{-2}`\n",
      "    ``milli``     :math:`10^{-3}`\n",
      "    ``micro``     :math:`10^{-6}`\n",
      "    ``nano``      :math:`10^{-9}`\n",
      "    ``pico``      :math:`10^{-12}`\n",
      "    ``femto``     :math:`10^{-15}`\n",
      "    ``atto``      :math:`10^{-18}`\n",
      "    ``zepto``     :math:`10^{-21}`\n",
      "    ============  =================================================================\n",
      "    \n",
      "    Binary prefixes\n",
      "    ---------------\n",
      "    \n",
      "    ============  =================================================================\n",
      "    ``kibi``      :math:`2^{10}`\n",
      "    ``mebi``      :math:`2^{20}`\n",
      "    ``gibi``      :math:`2^{30}`\n",
      "    ``tebi``      :math:`2^{40}`\n",
      "    ``pebi``      :math:`2^{50}`\n",
      "    ``exbi``      :math:`2^{60}`\n",
      "    ``zebi``      :math:`2^{70}`\n",
      "    ``yobi``      :math:`2^{80}`\n",
      "    ============  =================================================================\n",
      "    \n",
      "    Mass\n",
      "    ----\n",
      "    \n",
      "    =================  ============================================================\n",
      "    ``gram``           :math:`10^{-3}` kg\n",
      "    ``metric_ton``     :math:`10^{3}` kg\n",
      "    ``grain``          one grain in kg\n",
      "    ``lb``             one pound (avoirdupous) in kg\n",
      "    ``pound``          one pound (avoirdupous) in kg\n",
      "    ``blob``           one inch version of a slug in kg (added in 1.0.0)\n",
      "    ``slinch``         one inch version of a slug in kg (added in 1.0.0)\n",
      "    ``slug``           one slug in kg (added in 1.0.0)\n",
      "    ``oz``             one ounce in kg\n",
      "    ``ounce``          one ounce in kg\n",
      "    ``stone``          one stone in kg\n",
      "    ``grain``          one grain in kg\n",
      "    ``long_ton``       one long ton in kg\n",
      "    ``short_ton``      one short ton in kg\n",
      "    ``troy_ounce``     one Troy ounce in kg\n",
      "    ``troy_pound``     one Troy pound in kg\n",
      "    ``carat``          one carat in kg\n",
      "    ``m_u``            atomic mass constant (in kg)\n",
      "    ``u``              atomic mass constant (in kg)\n",
      "    ``atomic_mass``    atomic mass constant (in kg)\n",
      "    =================  ============================================================\n",
      "    \n",
      "    Angle\n",
      "    -----\n",
      "    \n",
      "    =================  ============================================================\n",
      "    ``degree``         degree in radians\n",
      "    ``arcmin``         arc minute in radians\n",
      "    ``arcminute``      arc minute in radians\n",
      "    ``arcsec``         arc second in radians\n",
      "    ``arcsecond``      arc second in radians\n",
      "    =================  ============================================================\n",
      "    \n",
      "    \n",
      "    Time\n",
      "    ----\n",
      "    \n",
      "    =================  ============================================================\n",
      "    ``minute``         one minute in seconds\n",
      "    ``hour``           one hour in seconds\n",
      "    ``day``            one day in seconds\n",
      "    ``week``           one week in seconds\n",
      "    ``year``           one year (365 days) in seconds\n",
      "    ``Julian_year``    one Julian year (365.25 days) in seconds\n",
      "    =================  ============================================================\n",
      "    \n",
      "    \n",
      "    Length\n",
      "    ------\n",
      "    \n",
      "    =====================  ============================================================\n",
      "    ``inch``               one inch in meters\n",
      "    ``foot``               one foot in meters\n",
      "    ``yard``               one yard in meters\n",
      "    ``mile``               one mile in meters\n",
      "    ``mil``                one mil in meters\n",
      "    ``pt``                 one point in meters\n",
      "    ``point``              one point in meters\n",
      "    ``survey_foot``        one survey foot in meters\n",
      "    ``survey_mile``        one survey mile in meters\n",
      "    ``nautical_mile``      one nautical mile in meters\n",
      "    ``fermi``              one Fermi in meters\n",
      "    ``angstrom``           one Angstrom in meters\n",
      "    ``micron``             one micron in meters\n",
      "    ``au``                 one astronomical unit in meters\n",
      "    ``astronomical_unit``  one astronomical unit in meters\n",
      "    ``light_year``         one light year in meters\n",
      "    ``parsec``             one parsec in meters\n",
      "    =====================  ============================================================\n",
      "    \n",
      "    Pressure\n",
      "    --------\n",
      "    \n",
      "    =================  ============================================================\n",
      "    ``atm``            standard atmosphere in pascals\n",
      "    ``atmosphere``     standard atmosphere in pascals\n",
      "    ``bar``            one bar in pascals\n",
      "    ``torr``           one torr (mmHg) in pascals\n",
      "    ``mmHg``           one torr (mmHg) in pascals\n",
      "    ``psi``            one psi in pascals\n",
      "    =================  ============================================================\n",
      "    \n",
      "    Area\n",
      "    ----\n",
      "    \n",
      "    =================  ============================================================\n",
      "    ``hectare``        one hectare in square meters\n",
      "    ``acre``           one acre in square meters\n",
      "    =================  ============================================================\n",
      "    \n",
      "    \n",
      "    Volume\n",
      "    ------\n",
      "    \n",
      "    ===================    ========================================================\n",
      "    ``liter``              one liter in cubic meters\n",
      "    ``litre``              one liter in cubic meters\n",
      "    ``gallon``             one gallon (US) in cubic meters\n",
      "    ``gallon_US``          one gallon (US) in cubic meters\n",
      "    ``gallon_imp``         one gallon (UK) in cubic meters\n",
      "    ``fluid_ounce``        one fluid ounce (US) in cubic meters\n",
      "    ``fluid_ounce_US``     one fluid ounce (US) in cubic meters\n",
      "    ``fluid_ounce_imp``    one fluid ounce (UK) in cubic meters\n",
      "    ``bbl``                one barrel in cubic meters\n",
      "    ``barrel``             one barrel in cubic meters\n",
      "    ===================    ========================================================\n",
      "    \n",
      "    Speed\n",
      "    -----\n",
      "    \n",
      "    ==================    ==========================================================\n",
      "    ``kmh``               kilometers per hour in meters per second\n",
      "    ``mph``               miles per hour in meters per second\n",
      "    ``mach``              one Mach (approx., at 15 C, 1 atm) in meters per second\n",
      "    ``speed_of_sound``    one Mach (approx., at 15 C, 1 atm) in meters per second\n",
      "    ``knot``              one knot in meters per second\n",
      "    ==================    ==========================================================\n",
      "    \n",
      "    \n",
      "    Temperature\n",
      "    -----------\n",
      "    \n",
      "    =====================  =======================================================\n",
      "    ``zero_Celsius``       zero of Celsius scale in Kelvin\n",
      "    ``degree_Fahrenheit``  one Fahrenheit (only differences) in Kelvins\n",
      "    =====================  =======================================================\n",
      "    \n",
      "    .. autosummary::\n",
      "       :toctree: generated/\n",
      "    \n",
      "       convert_temperature\n",
      "    \n",
      "    Energy\n",
      "    ------\n",
      "    \n",
      "    ====================  =======================================================\n",
      "    ``eV``                one electron volt in Joules\n",
      "    ``electron_volt``     one electron volt in Joules\n",
      "    ``calorie``           one calorie (thermochemical) in Joules\n",
      "    ``calorie_th``        one calorie (thermochemical) in Joules\n",
      "    ``calorie_IT``        one calorie (International Steam Table calorie, 1956) in Joules\n",
      "    ``erg``               one erg in Joules\n",
      "    ``Btu``               one British thermal unit (International Steam Table) in Joules\n",
      "    ``Btu_IT``            one British thermal unit (International Steam Table) in Joules\n",
      "    ``Btu_th``            one British thermal unit (thermochemical) in Joules\n",
      "    ``ton_TNT``           one ton of TNT in Joules\n",
      "    ====================  =======================================================\n",
      "    \n",
      "    Power\n",
      "    -----\n",
      "    \n",
      "    ====================  =======================================================\n",
      "    ``hp``                one horsepower in watts\n",
      "    ``horsepower``        one horsepower in watts\n",
      "    ====================  =======================================================\n",
      "    \n",
      "    Force\n",
      "    -----\n",
      "    \n",
      "    ====================  =======================================================\n",
      "    ``dyn``               one dyne in newtons\n",
      "    ``dyne``              one dyne in newtons\n",
      "    ``lbf``               one pound force in newtons\n",
      "    ``pound_force``       one pound force in newtons\n",
      "    ``kgf``               one kilogram force in newtons\n",
      "    ``kilogram_force``    one kilogram force in newtons\n",
      "    ====================  =======================================================\n",
      "    \n",
      "    Optics\n",
      "    ------\n",
      "    \n",
      "    .. autosummary::\n",
      "       :toctree: generated/\n",
      "    \n",
      "       lambda2nu\n",
      "       nu2lambda\n",
      "    \n",
      "    References\n",
      "    ==========\n",
      "    \n",
      "    .. [CODATA2018] CODATA Recommended Values of the Fundamental\n",
      "       Physical Constants 2018.\n",
      "    \n",
      "       https://physics.nist.gov/cuu/Constants/\n",
      "\n",
      "PACKAGE CONTENTS\n",
      "    codata\n",
      "    constants\n",
      "    setup\n",
      "    tests (package)\n",
      "\n",
      "CLASSES\n",
      "    builtins.DeprecationWarning(builtins.Warning)\n",
      "        scipy.constants.codata.ConstantWarning\n",
      "    \n",
      "    class ConstantWarning(builtins.DeprecationWarning)\n",
      "     |  Accessing a constant no longer in current CODATA data set\n",
      "     |  \n",
      "     |  Method resolution order:\n",
      "     |      ConstantWarning\n",
      "     |      builtins.DeprecationWarning\n",
      "     |      builtins.Warning\n",
      "     |      builtins.Exception\n",
      "     |      builtins.BaseException\n",
      "     |      builtins.object\n",
      "     |  \n",
      "     |  Data descriptors defined here:\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from builtins.DeprecationWarning:\n",
      "     |  \n",
      "     |  __init__(self, /, *args, **kwargs)\n",
      "     |      Initialize self.  See help(type(self)) for accurate signature.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Static methods inherited from builtins.DeprecationWarning:\n",
      "     |  \n",
      "     |  __new__(*args, **kwargs) from builtins.type\n",
      "     |      Create and return a new object.  See help(type) for accurate signature.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from builtins.BaseException:\n",
      "     |  \n",
      "     |  __delattr__(self, name, /)\n",
      "     |      Implement delattr(self, name).\n",
      "     |  \n",
      "     |  __getattribute__(self, name, /)\n",
      "     |      Return getattr(self, name).\n",
      "     |  \n",
      "     |  __reduce__(...)\n",
      "     |      Helper for pickle.\n",
      "     |  \n",
      "     |  __repr__(self, /)\n",
      "     |      Return repr(self).\n",
      "     |  \n",
      "     |  __setattr__(self, name, value, /)\n",
      "     |      Implement setattr(self, name, value).\n",
      "     |  \n",
      "     |  __setstate__(...)\n",
      "     |  \n",
      "     |  __str__(self, /)\n",
      "     |      Return str(self).\n",
      "     |  \n",
      "     |  with_traceback(...)\n",
      "     |      Exception.with_traceback(tb) --\n",
      "     |      set self.__traceback__ to tb and return self.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from builtins.BaseException:\n",
      "     |  \n",
      "     |  __cause__\n",
      "     |      exception cause\n",
      "     |  \n",
      "     |  __context__\n",
      "     |      exception context\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |  \n",
      "     |  __suppress_context__\n",
      "     |  \n",
      "     |  __traceback__\n",
      "     |  \n",
      "     |  args\n",
      "\n",
      "FUNCTIONS\n",
      "    convert_temperature(val, old_scale, new_scale)\n",
      "        Convert from a temperature scale to another one among Celsius, Kelvin,\n",
      "        Fahrenheit, and Rankine scales.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        val : array_like\n",
      "            Value(s) of the temperature(s) to be converted expressed in the\n",
      "            original scale.\n",
      "        \n",
      "        old_scale: str\n",
      "            Specifies as a string the original scale from which the temperature\n",
      "            value(s) will be converted. Supported scales are Celsius ('Celsius',\n",
      "            'celsius', 'C' or 'c'), Kelvin ('Kelvin', 'kelvin', 'K', 'k'),\n",
      "            Fahrenheit ('Fahrenheit', 'fahrenheit', 'F' or 'f'), and Rankine\n",
      "            ('Rankine', 'rankine', 'R', 'r').\n",
      "        \n",
      "        new_scale: str\n",
      "            Specifies as a string the new scale to which the temperature\n",
      "            value(s) will be converted. Supported scales are Celsius ('Celsius',\n",
      "            'celsius', 'C' or 'c'), Kelvin ('Kelvin', 'kelvin', 'K', 'k'),\n",
      "            Fahrenheit ('Fahrenheit', 'fahrenheit', 'F' or 'f'), and Rankine\n",
      "            ('Rankine', 'rankine', 'R', 'r').\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        res : float or array of floats\n",
      "            Value(s) of the converted temperature(s) expressed in the new scale.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        .. versionadded:: 0.18.0\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy.constants import convert_temperature\n",
      "        >>> convert_temperature(np.array([-40, 40]), 'Celsius', 'Kelvin')\n",
      "        array([ 233.15,  313.15])\n",
      "    \n",
      "    find(sub=None, disp=False)\n",
      "        Return list of physical_constant keys containing a given string.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        sub : str, unicode\n",
      "            Sub-string to search keys for. By default, return all keys.\n",
      "        disp : bool\n",
      "            If True, print the keys that are found and return None.\n",
      "            Otherwise, return the list of keys without printing anything.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        keys : list or None\n",
      "            If `disp` is False, the list of keys is returned.\n",
      "            Otherwise, None is returned.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy.constants import find, physical_constants\n",
      "        \n",
      "        Which keys in the ``physical_constants`` dictionary contain 'boltzmann'?\n",
      "        \n",
      "        >>> find('boltzmann')\n",
      "        ['Boltzmann constant',\n",
      "         'Boltzmann constant in Hz/K',\n",
      "         'Boltzmann constant in eV/K',\n",
      "         'Boltzmann constant in inverse meter per kelvin',\n",
      "         'Stefan-Boltzmann constant']\n",
      "        \n",
      "        Get the constant called 'Boltzmann constant in Hz/K':\n",
      "        \n",
      "        >>> physical_constants['Boltzmann constant in Hz/K']\n",
      "        (20836619120.0, 'Hz K^-1', 0.0)\n",
      "        \n",
      "        Find constants with 'radius' in the key:\n",
      "        \n",
      "        >>> find('radius')\n",
      "        ['Bohr radius',\n",
      "         'classical electron radius',\n",
      "         'deuteron rms charge radius',\n",
      "         'proton rms charge radius']\n",
      "        >>> physical_constants['classical electron radius']\n",
      "        (2.8179403262e-15, 'm', 1.3e-24)\n",
      "    \n",
      "    lambda2nu(lambda_)\n",
      "        Convert wavelength to optical frequency\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        lambda_ : array_like\n",
      "            Wavelength(s) to be converted.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        nu : float or array of floats\n",
      "            Equivalent optical frequency.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        Computes ``nu = c / lambda`` where c = 299792458.0, i.e., the\n",
      "        (vacuum) speed of light in meters/second.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy.constants import lambda2nu, speed_of_light\n",
      "        >>> lambda2nu(np.array((1, speed_of_light)))\n",
      "        array([  2.99792458e+08,   1.00000000e+00])\n",
      "    \n",
      "    nu2lambda(nu)\n",
      "        Convert optical frequency to wavelength.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        nu : array_like\n",
      "            Optical frequency to be converted.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        lambda : float or array of floats\n",
      "            Equivalent wavelength(s).\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        Computes ``lambda = c / nu`` where c = 299792458.0, i.e., the\n",
      "        (vacuum) speed of light in meters/second.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy.constants import nu2lambda, speed_of_light\n",
      "        >>> nu2lambda(np.array((1, speed_of_light)))\n",
      "        array([  2.99792458e+08,   1.00000000e+00])\n",
      "    \n",
      "    precision(key)\n",
      "        Relative precision in physical_constants indexed by key\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        key : Python string or unicode\n",
      "            Key in dictionary `physical_constants`\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        prec : float\n",
      "            Relative precision in `physical_constants` corresponding to `key`\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy import constants\n",
      "        >>> constants.precision(u'proton mass')\n",
      "        5.1e-37\n",
      "    \n",
      "    unit(key)\n",
      "        Unit in physical_constants indexed by key\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        key : Python string or unicode\n",
      "            Key in dictionary `physical_constants`\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        unit : Python string\n",
      "            Unit in `physical_constants` corresponding to `key`\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy import constants\n",
      "        >>> constants.unit(u'proton mass')\n",
      "        'kg'\n",
      "    \n",
      "    value(key)\n",
      "        Value in physical_constants indexed by key\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        key : Python string or unicode\n",
      "            Key in dictionary `physical_constants`\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        value : float\n",
      "            Value in `physical_constants` corresponding to `key`\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy import constants\n",
      "        >>> constants.value(u'elementary charge')\n",
      "        1.602176634e-19\n",
      "\n",
      "DATA\n",
      "    Avogadro = 6.02214076e+23\n",
      "    Boltzmann = 1.380649e-23\n",
      "    Btu = 1055.05585262\n",
      "    Btu_IT = 1055.05585262\n",
      "    Btu_th = 1054.3502644888888\n",
      "    G = 6.6743e-11\n",
      "    Julian_year = 31557600.0\n",
      "    N_A = 6.02214076e+23\n",
      "    Planck = 6.62607015e-34\n",
      "    R = 8.314462618\n",
      "    Rydberg = 10973731.56816\n",
      "    Stefan_Boltzmann = 5.670374419e-08\n",
      "    Wien = 0.002897771955\n",
      "    __all__ = ['Avogadro', 'Boltzmann', 'Btu', 'Btu_IT', 'Btu_th', 'Consta...\n",
      "    acre = 4046.8564223999992\n",
      "    alpha = 0.0072973525693\n",
      "    angstrom = 1e-10\n",
      "    arcmin = 0.0002908882086657216\n",
      "    arcminute = 0.0002908882086657216\n",
      "    arcsec = 4.84813681109536e-06\n",
      "    arcsecond = 4.84813681109536e-06\n",
      "    astronomical_unit = 149597870700.0\n",
      "    atm = 101325.0\n",
      "    atmosphere = 101325.0\n",
      "    atomic_mass = 1.6605390666e-27\n",
      "    atto = 1e-18\n",
      "    au = 149597870700.0\n",
      "    bar = 100000.0\n",
      "    barrel = 0.15898729492799998\n",
      "    bbl = 0.15898729492799998\n",
      "    blob = 175.12683524647636\n",
      "    c = 299792458.0\n",
      "    calorie = 4.184\n",
      "    calorie_IT = 4.1868\n",
      "    calorie_th = 4.184\n",
      "    carat = 0.0002\n",
      "    centi = 0.01\n",
      "    day = 86400.0\n",
      "    deci = 0.1\n",
      "    degree = 0.017453292519943295\n",
      "    degree_Fahrenheit = 0.5555555555555556\n",
      "    deka = 10.0\n",
      "    dyn = 1e-05\n",
      "    dyne = 1e-05\n",
      "    e = 1.602176634e-19\n",
      "    eV = 1.602176634e-19\n",
      "    electron_mass = 9.1093837015e-31\n",
      "    electron_volt = 1.602176634e-19\n",
      "    elementary_charge = 1.602176634e-19\n",
      "    epsilon_0 = 8.8541878128e-12\n",
      "    erg = 1e-07\n",
      "    exa = 1e+18\n",
      "    exbi = 1152921504606846976\n",
      "    femto = 1e-15\n",
      "    fermi = 1e-15\n",
      "    fine_structure = 0.0072973525693\n",
      "    fluid_ounce = 2.9573529562499998e-05\n",
      "    fluid_ounce_US = 2.9573529562499998e-05\n",
      "    fluid_ounce_imp = 2.84130625e-05\n",
      "    foot = 0.30479999999999996\n",
      "    g = 9.80665\n",
      "    gallon = 0.0037854117839999997\n",
      "    gallon_US = 0.0037854117839999997\n",
      "    gallon_imp = 0.00454609\n",
      "    gas_constant = 8.314462618\n",
      "    gibi = 1073741824\n",
      "    giga = 1000000000.0\n",
      "    golden = 1.618033988749895\n",
      "    golden_ratio = 1.618033988749895\n",
      "    grain = 6.479891e-05\n",
      "    gram = 0.001\n",
      "    gravitational_constant = 6.6743e-11\n",
      "    h = 6.62607015e-34\n",
      "    hbar = 1.0545718176461565e-34\n",
      "    hectare = 10000.0\n",
      "    hecto = 100.0\n",
      "    horsepower = 745.6998715822701\n",
      "    hour = 3600.0\n",
      "    hp = 745.6998715822701\n",
      "    inch = 0.0254\n",
      "    k = 1.380649e-23\n",
      "    kgf = 9.80665\n",
      "    kibi = 1024\n",
      "    kilo = 1000.0\n",
      "    kilogram_force = 9.80665\n",
      "    kmh = 0.2777777777777778\n",
      "    knot = 0.5144444444444445\n",
      "    lb = 0.45359236999999997\n",
      "    lbf = 4.4482216152605\n",
      "    light_year = 9460730472580800.0\n",
      "    liter = 0.001\n",
      "    litre = 0.001\n",
      "    long_ton = 1016.0469088\n",
      "    m_e = 9.1093837015e-31\n",
      "    m_n = 1.67492749804e-27\n",
      "    m_p = 1.67262192369e-27\n",
      "    m_u = 1.6605390666e-27\n",
      "    mach = 340.5\n",
      "    mebi = 1048576\n",
      "    mega = 1000000.0\n",
      "    metric_ton = 1000.0\n",
      "    micro = 1e-06\n",
      "    micron = 1e-06\n",
      "    mil = 2.5399999999999997e-05\n",
      "    mile = 1609.3439999999998\n",
      "    milli = 0.001\n",
      "    minute = 60.0\n",
      "    mmHg = 133.32236842105263\n",
      "    mph = 0.44703999999999994\n",
      "    mu_0 = 1.25663706212e-06\n",
      "    nano = 1e-09\n",
      "    nautical_mile = 1852.0\n",
      "    neutron_mass = 1.67492749804e-27\n",
      "    ounce = 0.028349523124999998\n",
      "    oz = 0.028349523124999998\n",
      "    parsec = 3.085677581491367e+16\n",
      "    pebi = 1125899906842624\n",
      "    peta = 1000000000000000.0\n",
      "    physical_constants = {'Angstrom star': (1.00001495e-10, 'm', 9e-17), '...\n",
      "    pi = 3.141592653589793\n",
      "    pico = 1e-12\n",
      "    point = 0.00035277777777777776\n",
      "    pound = 0.45359236999999997\n",
      "    pound_force = 4.4482216152605\n",
      "    proton_mass = 1.67262192369e-27\n",
      "    psi = 6894.757293168361\n",
      "    pt = 0.00035277777777777776\n",
      "    short_ton = 907.1847399999999\n",
      "    sigma = 5.670374419e-08\n",
      "    slinch = 175.12683524647636\n",
      "    slug = 14.593902937206364\n",
      "    speed_of_light = 299792458.0\n",
      "    speed_of_sound = 340.5\n",
      "    stone = 6.3502931799999995\n",
      "    survey_foot = 0.3048006096012192\n",
      "    survey_mile = 1609.3472186944373\n",
      "    tebi = 1099511627776\n",
      "    tera = 1000000000000.0\n",
      "    ton_TNT = 4184000000.0\n",
      "    torr = 133.32236842105263\n",
      "    troy_ounce = 0.031103476799999998\n",
      "    troy_pound = 0.37324172159999996\n",
      "    u = 1.6605390666e-27\n",
      "    week = 604800.0\n",
      "    yard = 0.9143999999999999\n",
      "    year = 31536000.0\n",
      "    yobi = 1208925819614629174706176\n",
      "    yotta = 1e+24\n",
      "    zebi = 1180591620717411303424\n",
      "    zepto = 1e-21\n",
      "    zero_Celsius = 273.15\n",
      "    zetta = 1e+21\n",
      "\n",
      "FILE\n",
      "    c:\\programdata\\anaconda3\\lib\\site-packages\\scipy\\constants\\__init__.py\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(sc)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "5318e71d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scipy.interpolate as sci"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "34db2bd9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on package scipy.interpolate in scipy:\n",
      "\n",
      "NAME\n",
      "    scipy.interpolate\n",
      "\n",
      "DESCRIPTION\n",
      "    ========================================\n",
      "    Interpolation (:mod:`scipy.interpolate`)\n",
      "    ========================================\n",
      "    \n",
      "    .. currentmodule:: scipy.interpolate\n",
      "    \n",
      "    Sub-package for objects used in interpolation.\n",
      "    \n",
      "    As listed below, this sub-package contains spline functions and classes,\n",
      "    1-D and multidimensional (univariate and multivariate)\n",
      "    interpolation classes, Lagrange and Taylor polynomial interpolators, and\n",
      "    wrappers for `FITPACK <http://www.netlib.org/dierckx/>`__\n",
      "    and DFITPACK functions.\n",
      "    \n",
      "    Univariate interpolation\n",
      "    ========================\n",
      "    \n",
      "    .. autosummary::\n",
      "       :toctree: generated/\n",
      "    \n",
      "       interp1d\n",
      "       BarycentricInterpolator\n",
      "       KroghInterpolator\n",
      "       barycentric_interpolate\n",
      "       krogh_interpolate\n",
      "       pchip_interpolate\n",
      "       CubicHermiteSpline\n",
      "       PchipInterpolator\n",
      "       Akima1DInterpolator\n",
      "       CubicSpline\n",
      "       PPoly\n",
      "       BPoly\n",
      "    \n",
      "    \n",
      "    Multivariate interpolation\n",
      "    ==========================\n",
      "    \n",
      "    Unstructured data:\n",
      "    \n",
      "    .. autosummary::\n",
      "       :toctree: generated/\n",
      "    \n",
      "       griddata\n",
      "       LinearNDInterpolator\n",
      "       NearestNDInterpolator\n",
      "       CloughTocher2DInterpolator\n",
      "       RBFInterpolator\n",
      "       Rbf\n",
      "       interp2d\n",
      "    \n",
      "    For data on a grid:\n",
      "    \n",
      "    .. autosummary::\n",
      "       :toctree: generated/\n",
      "    \n",
      "       interpn\n",
      "       RegularGridInterpolator\n",
      "       RectBivariateSpline\n",
      "    \n",
      "    .. seealso::\n",
      "    \n",
      "        `scipy.ndimage.map_coordinates`\n",
      "    \n",
      "    Tensor product polynomials:\n",
      "    \n",
      "    .. autosummary::\n",
      "       :toctree: generated/\n",
      "    \n",
      "       NdPPoly\n",
      "    \n",
      "    \n",
      "    1-D Splines\n",
      "    ===========\n",
      "    \n",
      "    .. autosummary::\n",
      "       :toctree: generated/\n",
      "    \n",
      "       BSpline\n",
      "       make_interp_spline\n",
      "       make_lsq_spline\n",
      "    \n",
      "    Functional interface to FITPACK routines:\n",
      "    \n",
      "    .. autosummary::\n",
      "       :toctree: generated/\n",
      "    \n",
      "       splrep\n",
      "       splprep\n",
      "       splev\n",
      "       splint\n",
      "       sproot\n",
      "       spalde\n",
      "       splder\n",
      "       splantider\n",
      "       insert\n",
      "    \n",
      "    Object-oriented FITPACK interface:\n",
      "    \n",
      "    .. autosummary::\n",
      "       :toctree: generated/\n",
      "    \n",
      "       UnivariateSpline\n",
      "       InterpolatedUnivariateSpline\n",
      "       LSQUnivariateSpline\n",
      "    \n",
      "    \n",
      "    \n",
      "    2-D Splines\n",
      "    ===========\n",
      "    \n",
      "    For data on a grid:\n",
      "    \n",
      "    .. autosummary::\n",
      "       :toctree: generated/\n",
      "    \n",
      "       RectBivariateSpline\n",
      "       RectSphereBivariateSpline\n",
      "    \n",
      "    For unstructured data:\n",
      "    \n",
      "    .. autosummary::\n",
      "       :toctree: generated/\n",
      "    \n",
      "       BivariateSpline\n",
      "       SmoothBivariateSpline\n",
      "       SmoothSphereBivariateSpline\n",
      "       LSQBivariateSpline\n",
      "       LSQSphereBivariateSpline\n",
      "    \n",
      "    Low-level interface to FITPACK functions:\n",
      "    \n",
      "    .. autosummary::\n",
      "       :toctree: generated/\n",
      "    \n",
      "       bisplrep\n",
      "       bisplev\n",
      "    \n",
      "    Additional tools\n",
      "    ================\n",
      "    \n",
      "    .. autosummary::\n",
      "       :toctree: generated/\n",
      "    \n",
      "       lagrange\n",
      "       approximate_taylor_polynomial\n",
      "       pade\n",
      "    \n",
      "    .. seealso::\n",
      "    \n",
      "       `scipy.ndimage.map_coordinates`,\n",
      "       `scipy.ndimage.spline_filter`,\n",
      "       `scipy.signal.resample`,\n",
      "       `scipy.signal.bspline`,\n",
      "       `scipy.signal.gauss_spline`,\n",
      "       `scipy.signal.qspline1d`,\n",
      "       `scipy.signal.cspline1d`,\n",
      "       `scipy.signal.qspline1d_eval`,\n",
      "       `scipy.signal.cspline1d_eval`,\n",
      "       `scipy.signal.qspline2d`,\n",
      "       `scipy.signal.cspline2d`.\n",
      "    \n",
      "    ``pchip`` is an alias of `PchipInterpolator` for backward compatibility\n",
      "    (should not be used in new code).\n",
      "\n",
      "PACKAGE CONTENTS\n",
      "    _bspl\n",
      "    _bsplines\n",
      "    _cubic\n",
      "    _fitpack\n",
      "    _fitpack_impl\n",
      "    _pade\n",
      "    _ppoly\n",
      "    _rbfinterp\n",
      "    _rbfinterp_pythran\n",
      "    dfitpack\n",
      "    fitpack\n",
      "    fitpack2\n",
      "    interpnd\n",
      "    interpnd_info\n",
      "    interpolate\n",
      "    ndgriddata\n",
      "    polyint\n",
      "    rbf\n",
      "    setup\n",
      "    tests (package)\n",
      "\n",
      "CLASSES\n",
      "    builtins.object\n",
      "        scipy.interpolate._bsplines.BSpline\n",
      "        scipy.interpolate._rbfinterp.RBFInterpolator\n",
      "        scipy.interpolate.fitpack2.UnivariateSpline\n",
      "            scipy.interpolate.fitpack2.InterpolatedUnivariateSpline\n",
      "            scipy.interpolate.fitpack2.LSQUnivariateSpline\n",
      "        scipy.interpolate.interpolate.NdPPoly\n",
      "        scipy.interpolate.interpolate.RegularGridInterpolator\n",
      "        scipy.interpolate.interpolate.interp2d\n",
      "        scipy.interpolate.rbf.Rbf\n",
      "    scipy.interpolate.fitpack2.SphereBivariateSpline(scipy.interpolate.fitpack2._BivariateSplineBase)\n",
      "        scipy.interpolate.fitpack2.LSQSphereBivariateSpline\n",
      "        scipy.interpolate.fitpack2.RectSphereBivariateSpline\n",
      "        scipy.interpolate.fitpack2.SmoothSphereBivariateSpline\n",
      "    scipy.interpolate.fitpack2._BivariateSplineBase(builtins.object)\n",
      "        scipy.interpolate.fitpack2.BivariateSpline\n",
      "            scipy.interpolate.fitpack2.LSQBivariateSpline\n",
      "            scipy.interpolate.fitpack2.RectBivariateSpline\n",
      "            scipy.interpolate.fitpack2.SmoothBivariateSpline\n",
      "    scipy.interpolate.interpnd.NDInterpolatorBase(builtins.object)\n",
      "        scipy.interpolate.interpnd.CloughTocher2DInterpolator\n",
      "        scipy.interpolate.interpnd.LinearNDInterpolator\n",
      "        scipy.interpolate.ndgriddata.NearestNDInterpolator\n",
      "    scipy.interpolate.interpolate._PPolyBase(builtins.object)\n",
      "        scipy.interpolate.interpolate.BPoly\n",
      "        scipy.interpolate.interpolate.PPoly\n",
      "            scipy.interpolate._cubic.CubicHermiteSpline\n",
      "                scipy.interpolate._cubic.Akima1DInterpolator\n",
      "                scipy.interpolate._cubic.CubicSpline\n",
      "                scipy.interpolate._cubic.PchipInterpolator\n",
      "    scipy.interpolate.polyint._Interpolator1D(builtins.object)\n",
      "        scipy.interpolate.interpolate.interp1d\n",
      "        scipy.interpolate.polyint.BarycentricInterpolator\n",
      "    scipy.interpolate.polyint._Interpolator1DWithDerivatives(scipy.interpolate.polyint._Interpolator1D)\n",
      "        scipy.interpolate.polyint.KroghInterpolator\n",
      "    \n",
      "    class Akima1DInterpolator(CubicHermiteSpline)\n",
      "     |  Akima1DInterpolator(x, y, axis=0)\n",
      "     |  \n",
      "     |  Akima interpolator\n",
      "     |  \n",
      "     |  Fit piecewise cubic polynomials, given vectors x and y. The interpolation\n",
      "     |  method by Akima uses a continuously differentiable sub-spline built from\n",
      "     |  piecewise cubic polynomials. The resultant curve passes through the given\n",
      "     |  data points and will appear smooth and natural.\n",
      "     |  \n",
      "     |  Parameters\n",
      "     |  ----------\n",
      "     |  x : ndarray, shape (m, )\n",
      "     |      1-D array of monotonically increasing real values.\n",
      "     |  y : ndarray, shape (m, ...)\n",
      "     |      N-D array of real values. The length of ``y`` along the first axis\n",
      "     |      must be equal to the length of ``x``.\n",
      "     |  axis : int, optional\n",
      "     |      Specifies the axis of ``y`` along which to interpolate. Interpolation\n",
      "     |      defaults to the first axis of ``y``.\n",
      "     |  \n",
      "     |  Methods\n",
      "     |  -------\n",
      "     |  __call__\n",
      "     |  derivative\n",
      "     |  antiderivative\n",
      "     |  roots\n",
      "     |  \n",
      "     |  See Also\n",
      "     |  --------\n",
      "     |  PchipInterpolator : PCHIP 1-D monotonic cubic interpolator.\n",
      "     |  CubicSpline : Cubic spline data interpolator.\n",
      "     |  PPoly : Piecewise polynomial in terms of coefficients and breakpoints\n",
      "     |  \n",
      "     |  Notes\n",
      "     |  -----\n",
      "     |  .. versionadded:: 0.14\n",
      "     |  \n",
      "     |  Use only for precise data, as the fitted curve passes through the given\n",
      "     |  points exactly. This routine is useful for plotting a pleasingly smooth\n",
      "     |  curve through a few given points for purposes of plotting.\n",
      "     |  \n",
      "     |  References\n",
      "     |  ----------\n",
      "     |  [1] A new method of interpolation and smooth curve fitting based\n",
      "     |      on local procedures. Hiroshi Akima, J. ACM, October 1970, 17(4),\n",
      "     |      589-602.\n",
      "     |  \n",
      "     |  Method resolution order:\n",
      "     |      Akima1DInterpolator\n",
      "     |      CubicHermiteSpline\n",
      "     |      scipy.interpolate.interpolate.PPoly\n",
      "     |      scipy.interpolate.interpolate._PPolyBase\n",
      "     |      builtins.object\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __init__(self, x, y, axis=0)\n",
      "     |      Initialize self.  See help(type(self)) for accurate signature.\n",
      "     |  \n",
      "     |  extend(self, c, x, right=True)\n",
      "     |      Add additional breakpoints and coefficients to the polynomial.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      c : ndarray, size (k, m, ...)\n",
      "     |          Additional coefficients for polynomials in intervals. Note that\n",
      "     |          the first additional interval will be formed using one of the\n",
      "     |          ``self.x`` end points.\n",
      "     |      x : ndarray, size (m,)\n",
      "     |          Additional breakpoints. Must be sorted in the same order as\n",
      "     |          ``self.x`` and either to the right or to the left of the current\n",
      "     |          breakpoints.\n",
      "     |      right\n",
      "     |          Deprecated argument. Has no effect.\n",
      "     |      \n",
      "     |          .. deprecated:: 0.19\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Class methods defined here:\n",
      "     |  \n",
      "     |  from_bernstein_basis(bp, extrapolate=None) from builtins.type\n",
      "     |      Construct a piecewise polynomial in the power basis\n",
      "     |      from a polynomial in Bernstein basis.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      bp : BPoly\n",
      "     |          A Bernstein basis polynomial, as created by BPoly\n",
      "     |      extrapolate : bool or 'periodic', optional\n",
      "     |          If bool, determines whether to extrapolate to out-of-bounds points\n",
      "     |          based on first and last intervals, or to return NaNs.\n",
      "     |          If 'periodic', periodic extrapolation is used. Default is True.\n",
      "     |  \n",
      "     |  from_spline(tck, extrapolate=None) from builtins.type\n",
      "     |      Construct a piecewise polynomial from a spline\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      tck\n",
      "     |          A spline, as returned by `splrep` or a BSpline object.\n",
      "     |      extrapolate : bool or 'periodic', optional\n",
      "     |          If bool, determines whether to extrapolate to out-of-bounds points\n",
      "     |          based on first and last intervals, or to return NaNs.\n",
      "     |          If 'periodic', periodic extrapolation is used. Default is True.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from scipy.interpolate.interpolate.PPoly:\n",
      "     |  \n",
      "     |  antiderivative(self, nu=1)\n",
      "     |      Construct a new piecewise polynomial representing the antiderivative.\n",
      "     |      \n",
      "     |      Antiderivative is also the indefinite integral of the function,\n",
      "     |      and derivative is its inverse operation.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      nu : int, optional\n",
      "     |          Order of antiderivative to evaluate. Default is 1, i.e., compute\n",
      "     |          the first integral. If negative, the derivative is returned.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      pp : PPoly\n",
      "     |          Piecewise polynomial of order k2 = k + n representing\n",
      "     |          the antiderivative of this polynomial.\n",
      "     |      \n",
      "     |      Notes\n",
      "     |      -----\n",
      "     |      The antiderivative returned by this function is continuous and\n",
      "     |      continuously differentiable to order n-1, up to floating point\n",
      "     |      rounding error.\n",
      "     |      \n",
      "     |      If antiderivative is computed and ``self.extrapolate='periodic'``,\n",
      "     |      it will be set to False for the returned instance. This is done because\n",
      "     |      the antiderivative is no longer periodic and its correct evaluation\n",
      "     |      outside of the initially given x interval is difficult.\n",
      "     |  \n",
      "     |  derivative(self, nu=1)\n",
      "     |      Construct a new piecewise polynomial representing the derivative.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      nu : int, optional\n",
      "     |          Order of derivative to evaluate. Default is 1, i.e., compute the\n",
      "     |          first derivative. If negative, the antiderivative is returned.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      pp : PPoly\n",
      "     |          Piecewise polynomial of order k2 = k - n representing the derivative\n",
      "     |          of this polynomial.\n",
      "     |      \n",
      "     |      Notes\n",
      "     |      -----\n",
      "     |      Derivatives are evaluated piecewise for each polynomial\n",
      "     |      segment, even if the polynomial is not differentiable at the\n",
      "     |      breakpoints. The polynomial intervals are considered half-open,\n",
      "     |      ``[a, b)``, except for the last interval which is closed\n",
      "     |      ``[a, b]``.\n",
      "     |  \n",
      "     |  integrate(self, a, b, extrapolate=None)\n",
      "     |      Compute a definite integral over a piecewise polynomial.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      a : float\n",
      "     |          Lower integration bound\n",
      "     |      b : float\n",
      "     |          Upper integration bound\n",
      "     |      extrapolate : {bool, 'periodic', None}, optional\n",
      "     |          If bool, determines whether to extrapolate to out-of-bounds points\n",
      "     |          based on first and last intervals, or to return NaNs.\n",
      "     |          If 'periodic', periodic extrapolation is used.\n",
      "     |          If None (default), use `self.extrapolate`.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      ig : array_like\n",
      "     |          Definite integral of the piecewise polynomial over [a, b]\n",
      "     |  \n",
      "     |  roots(self, discontinuity=True, extrapolate=None)\n",
      "     |      Find real roots of the the piecewise polynomial.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      discontinuity : bool, optional\n",
      "     |          Whether to report sign changes across discontinuities at\n",
      "     |          breakpoints as roots.\n",
      "     |      extrapolate : {bool, 'periodic', None}, optional\n",
      "     |          If bool, determines whether to return roots from the polynomial\n",
      "     |          extrapolated based on first and last intervals, 'periodic' works\n",
      "     |          the same as False. If None (default), use `self.extrapolate`.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      roots : ndarray\n",
      "     |          Roots of the polynomial(s).\n",
      "     |      \n",
      "     |          If the PPoly object describes multiple polynomials, the\n",
      "     |          return value is an object array whose each element is an\n",
      "     |          ndarray containing the roots.\n",
      "     |      \n",
      "     |      See Also\n",
      "     |      --------\n",
      "     |      PPoly.solve\n",
      "     |  \n",
      "     |  solve(self, y=0.0, discontinuity=True, extrapolate=None)\n",
      "     |      Find real solutions of the the equation ``pp(x) == y``.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      y : float, optional\n",
      "     |          Right-hand side. Default is zero.\n",
      "     |      discontinuity : bool, optional\n",
      "     |          Whether to report sign changes across discontinuities at\n",
      "     |          breakpoints as roots.\n",
      "     |      extrapolate : {bool, 'periodic', None}, optional\n",
      "     |          If bool, determines whether to return roots from the polynomial\n",
      "     |          extrapolated based on first and last intervals, 'periodic' works\n",
      "     |          the same as False. If None (default), use `self.extrapolate`.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      roots : ndarray\n",
      "     |          Roots of the polynomial(s).\n",
      "     |      \n",
      "     |          If the PPoly object describes multiple polynomials, the\n",
      "     |          return value is an object array whose each element is an\n",
      "     |          ndarray containing the roots.\n",
      "     |      \n",
      "     |      Notes\n",
      "     |      -----\n",
      "     |      This routine works only on real-valued polynomials.\n",
      "     |      \n",
      "     |      If the piecewise polynomial contains sections that are\n",
      "     |      identically zero, the root list will contain the start point\n",
      "     |      of the corresponding interval, followed by a ``nan`` value.\n",
      "     |      \n",
      "     |      If the polynomial is discontinuous across a breakpoint, and\n",
      "     |      there is a sign change across the breakpoint, this is reported\n",
      "     |      if the `discont` parameter is True.\n",
      "     |      \n",
      "     |      Examples\n",
      "     |      --------\n",
      "     |      \n",
      "     |      Finding roots of ``[x**2 - 1, (x - 1)**2]`` defined on intervals\n",
      "     |      ``[-2, 1], [1, 2]``:\n",
      "     |      \n",
      "     |      >>> from scipy.interpolate import PPoly\n",
      "     |      >>> pp = PPoly(np.array([[1, -4, 3], [1, 0, 0]]).T, [-2, 1, 2])\n",
      "     |      >>> pp.solve()\n",
      "     |      array([-1.,  1.])\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from scipy.interpolate.interpolate.PPoly:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |      dictionary for instance variables (if defined)\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from scipy.interpolate.interpolate._PPolyBase:\n",
      "     |  \n",
      "     |  __call__(self, x, nu=0, extrapolate=None)\n",
      "     |      Evaluate the piecewise polynomial or its derivative.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      x : array_like\n",
      "     |          Points to evaluate the interpolant at.\n",
      "     |      nu : int, optional\n",
      "     |          Order of derivative to evaluate. Must be non-negative.\n",
      "     |      extrapolate : {bool, 'periodic', None}, optional\n",
      "     |          If bool, determines whether to extrapolate to out-of-bounds points\n",
      "     |          based on first and last intervals, or to return NaNs.\n",
      "     |          If 'periodic', periodic extrapolation is used.\n",
      "     |          If None (default), use `self.extrapolate`.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      y : array_like\n",
      "     |          Interpolated values. Shape is determined by replacing\n",
      "     |          the interpolation axis in the original array with the shape of x.\n",
      "     |      \n",
      "     |      Notes\n",
      "     |      -----\n",
      "     |      Derivatives are evaluated piecewise for each polynomial\n",
      "     |      segment, even if the polynomial is not differentiable at the\n",
      "     |      breakpoints. The polynomial intervals are considered half-open,\n",
      "     |      ``[a, b)``, except for the last interval which is closed\n",
      "     |      ``[a, b]``.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Class methods inherited from scipy.interpolate.interpolate._PPolyBase:\n",
      "     |  \n",
      "     |  construct_fast(c, x, extrapolate=None, axis=0) from builtins.type\n",
      "     |      Construct the piecewise polynomial without making checks.\n",
      "     |      \n",
      "     |      Takes the same parameters as the constructor. Input arguments\n",
      "     |      ``c`` and ``x`` must be arrays of the correct shape and type. The\n",
      "     |      ``c`` array can only be of dtypes float and complex, and ``x``\n",
      "     |      array must have dtype float.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from scipy.interpolate.interpolate._PPolyBase:\n",
      "     |  \n",
      "     |  axis\n",
      "     |  \n",
      "     |  c\n",
      "     |  \n",
      "     |  extrapolate\n",
      "     |  \n",
      "     |  x\n",
      "    \n",
      "    class BPoly(_PPolyBase)\n",
      "     |  BPoly(c, x, extrapolate=None, axis=0)\n",
      "     |  \n",
      "     |  Piecewise polynomial in terms of coefficients and breakpoints.\n",
      "     |  \n",
      "     |  The polynomial between ``x[i]`` and ``x[i + 1]`` is written in the\n",
      "     |  Bernstein polynomial basis::\n",
      "     |  \n",
      "     |      S = sum(c[a, i] * b(a, k; x) for a in range(k+1)),\n",
      "     |  \n",
      "     |  where ``k`` is the degree of the polynomial, and::\n",
      "     |  \n",
      "     |      b(a, k; x) = binom(k, a) * t**a * (1 - t)**(k - a),\n",
      "     |  \n",
      "     |  with ``t = (x - x[i]) / (x[i+1] - x[i])`` and ``binom`` is the binomial\n",
      "     |  coefficient.\n",
      "     |  \n",
      "     |  Parameters\n",
      "     |  ----------\n",
      "     |  c : ndarray, shape (k, m, ...)\n",
      "     |      Polynomial coefficients, order `k` and `m` intervals\n",
      "     |  x : ndarray, shape (m+1,)\n",
      "     |      Polynomial breakpoints. Must be sorted in either increasing or\n",
      "     |      decreasing order.\n",
      "     |  extrapolate : bool, optional\n",
      "     |      If bool, determines whether to extrapolate to out-of-bounds points\n",
      "     |      based on first and last intervals, or to return NaNs. If 'periodic',\n",
      "     |      periodic extrapolation is used. Default is True.\n",
      "     |  axis : int, optional\n",
      "     |      Interpolation axis. Default is zero.\n",
      "     |  \n",
      "     |  Attributes\n",
      "     |  ----------\n",
      "     |  x : ndarray\n",
      "     |      Breakpoints.\n",
      "     |  c : ndarray\n",
      "     |      Coefficients of the polynomials. They are reshaped\n",
      "     |      to a 3-D array with the last dimension representing\n",
      "     |      the trailing dimensions of the original coefficient array.\n",
      "     |  axis : int\n",
      "     |      Interpolation axis.\n",
      "     |  \n",
      "     |  Methods\n",
      "     |  -------\n",
      "     |  __call__\n",
      "     |  extend\n",
      "     |  derivative\n",
      "     |  antiderivative\n",
      "     |  integrate\n",
      "     |  construct_fast\n",
      "     |  from_power_basis\n",
      "     |  from_derivatives\n",
      "     |  \n",
      "     |  See also\n",
      "     |  --------\n",
      "     |  PPoly : piecewise polynomials in the power basis\n",
      "     |  \n",
      "     |  Notes\n",
      "     |  -----\n",
      "     |  Properties of Bernstein polynomials are well documented in the literature,\n",
      "     |  see for example [1]_ [2]_ [3]_.\n",
      "     |  \n",
      "     |  References\n",
      "     |  ----------\n",
      "     |  .. [1] https://en.wikipedia.org/wiki/Bernstein_polynomial\n",
      "     |  \n",
      "     |  .. [2] Kenneth I. Joy, Bernstein polynomials,\n",
      "     |     http://www.idav.ucdavis.edu/education/CAGDNotes/Bernstein-Polynomials.pdf\n",
      "     |  \n",
      "     |  .. [3] E. H. Doha, A. H. Bhrawy, and M. A. Saker, Boundary Value Problems,\n",
      "     |         vol 2011, article ID 829546, :doi:`10.1155/2011/829543`.\n",
      "     |  \n",
      "     |  Examples\n",
      "     |  --------\n",
      "     |  >>> from scipy.interpolate import BPoly\n",
      "     |  >>> x = [0, 1]\n",
      "     |  >>> c = [[1], [2], [3]]\n",
      "     |  >>> bp = BPoly(c, x)\n",
      "     |  \n",
      "     |  This creates a 2nd order polynomial\n",
      "     |  \n",
      "     |  .. math::\n",
      "     |  \n",
      "     |      B(x) = 1 \\times b_{0, 2}(x) + 2 \\times b_{1, 2}(x) + 3 \\times b_{2, 2}(x) \\\\\n",
      "     |           = 1 \\times (1-x)^2 + 2 \\times 2 x (1 - x) + 3 \\times x^2\n",
      "     |  \n",
      "     |  Method resolution order:\n",
      "     |      BPoly\n",
      "     |      _PPolyBase\n",
      "     |      builtins.object\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  antiderivative(self, nu=1)\n",
      "     |      Construct a new piecewise polynomial representing the antiderivative.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      nu : int, optional\n",
      "     |          Order of antiderivative to evaluate. Default is 1, i.e., compute\n",
      "     |          the first integral. If negative, the derivative is returned.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      bp : BPoly\n",
      "     |          Piecewise polynomial of order k + nu representing the\n",
      "     |          antiderivative of this polynomial.\n",
      "     |      \n",
      "     |      Notes\n",
      "     |      -----\n",
      "     |      If antiderivative is computed and ``self.extrapolate='periodic'``,\n",
      "     |      it will be set to False for the returned instance. This is done because\n",
      "     |      the antiderivative is no longer periodic and its correct evaluation\n",
      "     |      outside of the initially given x interval is difficult.\n",
      "     |  \n",
      "     |  derivative(self, nu=1)\n",
      "     |      Construct a new piecewise polynomial representing the derivative.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      nu : int, optional\n",
      "     |          Order of derivative to evaluate. Default is 1, i.e., compute the\n",
      "     |          first derivative. If negative, the antiderivative is returned.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      bp : BPoly\n",
      "     |          Piecewise polynomial of order k - nu representing the derivative of\n",
      "     |          this polynomial.\n",
      "     |  \n",
      "     |  extend(self, c, x, right=None)\n",
      "     |      Add additional breakpoints and coefficients to the polynomial.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      c : ndarray, size (k, m, ...)\n",
      "     |          Additional coefficients for polynomials in intervals. Note that\n",
      "     |          the first additional interval will be formed using one of the\n",
      "     |          ``self.x`` end points.\n",
      "     |      x : ndarray, size (m,)\n",
      "     |          Additional breakpoints. Must be sorted in the same order as\n",
      "     |          ``self.x`` and either to the right or to the left of the current\n",
      "     |          breakpoints.\n",
      "     |      right\n",
      "     |          Deprecated argument. Has no effect.\n",
      "     |      \n",
      "     |          .. deprecated:: 0.19\n",
      "     |  \n",
      "     |  integrate(self, a, b, extrapolate=None)\n",
      "     |      Compute a definite integral over a piecewise polynomial.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      a : float\n",
      "     |          Lower integration bound\n",
      "     |      b : float\n",
      "     |          Upper integration bound\n",
      "     |      extrapolate : {bool, 'periodic', None}, optional\n",
      "     |          Whether to extrapolate to out-of-bounds points based on first\n",
      "     |          and last intervals, or to return NaNs. If 'periodic', periodic\n",
      "     |          extrapolation is used. If None (default), use `self.extrapolate`.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      array_like\n",
      "     |          Definite integral of the piecewise polynomial over [a, b]\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Class methods defined here:\n",
      "     |  \n",
      "     |  from_derivatives(xi, yi, orders=None, extrapolate=None) from builtins.type\n",
      "     |      Construct a piecewise polynomial in the Bernstein basis,\n",
      "     |      compatible with the specified values and derivatives at breakpoints.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      xi : array_like\n",
      "     |          sorted 1-D array of x-coordinates\n",
      "     |      yi : array_like or list of array_likes\n",
      "     |          ``yi[i][j]`` is the ``j``th derivative known at ``xi[i]``\n",
      "     |      orders : None or int or array_like of ints. Default: None.\n",
      "     |          Specifies the degree of local polynomials. If not None, some\n",
      "     |          derivatives are ignored.\n",
      "     |      extrapolate : bool or 'periodic', optional\n",
      "     |          If bool, determines whether to extrapolate to out-of-bounds points\n",
      "     |          based on first and last intervals, or to return NaNs.\n",
      "     |          If 'periodic', periodic extrapolation is used. Default is True.\n",
      "     |      \n",
      "     |      Notes\n",
      "     |      -----\n",
      "     |      If ``k`` derivatives are specified at a breakpoint ``x``, the\n",
      "     |      constructed polynomial is exactly ``k`` times continuously\n",
      "     |      differentiable at ``x``, unless the ``order`` is provided explicitly.\n",
      "     |      In the latter case, the smoothness of the polynomial at\n",
      "     |      the breakpoint is controlled by the ``order``.\n",
      "     |      \n",
      "     |      Deduces the number of derivatives to match at each end\n",
      "     |      from ``order`` and the number of derivatives available. If\n",
      "     |      possible it uses the same number of derivatives from\n",
      "     |      each end; if the number is odd it tries to take the\n",
      "     |      extra one from y2. In any case if not enough derivatives\n",
      "     |      are available at one end or another it draws enough to\n",
      "     |      make up the total from the other end.\n",
      "     |      \n",
      "     |      If the order is too high and not enough derivatives are available,\n",
      "     |      an exception is raised.\n",
      "     |      \n",
      "     |      Examples\n",
      "     |      --------\n",
      "     |      \n",
      "     |      >>> from scipy.interpolate import BPoly\n",
      "     |      >>> BPoly.from_derivatives([0, 1], [[1, 2], [3, 4]])\n",
      "     |      \n",
      "     |      Creates a polynomial `f(x)` of degree 3, defined on `[0, 1]`\n",
      "     |      such that `f(0) = 1, df/dx(0) = 2, f(1) = 3, df/dx(1) = 4`\n",
      "     |      \n",
      "     |      >>> BPoly.from_derivatives([0, 1, 2], [[0, 1], [0], [2]])\n",
      "     |      \n",
      "     |      Creates a piecewise polynomial `f(x)`, such that\n",
      "     |      `f(0) = f(1) = 0`, `f(2) = 2`, and `df/dx(0) = 1`.\n",
      "     |      Based on the number of derivatives provided, the order of the\n",
      "     |      local polynomials is 2 on `[0, 1]` and 1 on `[1, 2]`.\n",
      "     |      Notice that no restriction is imposed on the derivatives at\n",
      "     |      ``x = 1`` and ``x = 2``.\n",
      "     |      \n",
      "     |      Indeed, the explicit form of the polynomial is::\n",
      "     |      \n",
      "     |          f(x) = | x * (1 - x),  0 <= x < 1\n",
      "     |                 | 2 * (x - 1),  1 <= x <= 2\n",
      "     |      \n",
      "     |      So that f'(1-0) = -1 and f'(1+0) = 2\n",
      "     |  \n",
      "     |  from_power_basis(pp, extrapolate=None) from builtins.type\n",
      "     |      Construct a piecewise polynomial in Bernstein basis\n",
      "     |      from a power basis polynomial.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      pp : PPoly\n",
      "     |          A piecewise polynomial in the power basis\n",
      "     |      extrapolate : bool or 'periodic', optional\n",
      "     |          If bool, determines whether to extrapolate to out-of-bounds points\n",
      "     |          based on first and last intervals, or to return NaNs.\n",
      "     |          If 'periodic', periodic extrapolation is used. Default is True.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors defined here:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |      dictionary for instance variables (if defined)\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from _PPolyBase:\n",
      "     |  \n",
      "     |  __call__(self, x, nu=0, extrapolate=None)\n",
      "     |      Evaluate the piecewise polynomial or its derivative.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      x : array_like\n",
      "     |          Points to evaluate the interpolant at.\n",
      "     |      nu : int, optional\n",
      "     |          Order of derivative to evaluate. Must be non-negative.\n",
      "     |      extrapolate : {bool, 'periodic', None}, optional\n",
      "     |          If bool, determines whether to extrapolate to out-of-bounds points\n",
      "     |          based on first and last intervals, or to return NaNs.\n",
      "     |          If 'periodic', periodic extrapolation is used.\n",
      "     |          If None (default), use `self.extrapolate`.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      y : array_like\n",
      "     |          Interpolated values. Shape is determined by replacing\n",
      "     |          the interpolation axis in the original array with the shape of x.\n",
      "     |      \n",
      "     |      Notes\n",
      "     |      -----\n",
      "     |      Derivatives are evaluated piecewise for each polynomial\n",
      "     |      segment, even if the polynomial is not differentiable at the\n",
      "     |      breakpoints. The polynomial intervals are considered half-open,\n",
      "     |      ``[a, b)``, except for the last interval which is closed\n",
      "     |      ``[a, b]``.\n",
      "     |  \n",
      "     |  __init__(self, c, x, extrapolate=None, axis=0)\n",
      "     |      Initialize self.  See help(type(self)) for accurate signature.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Class methods inherited from _PPolyBase:\n",
      "     |  \n",
      "     |  construct_fast(c, x, extrapolate=None, axis=0) from builtins.type\n",
      "     |      Construct the piecewise polynomial without making checks.\n",
      "     |      \n",
      "     |      Takes the same parameters as the constructor. Input arguments\n",
      "     |      ``c`` and ``x`` must be arrays of the correct shape and type. The\n",
      "     |      ``c`` array can only be of dtypes float and complex, and ``x``\n",
      "     |      array must have dtype float.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from _PPolyBase:\n",
      "     |  \n",
      "     |  axis\n",
      "     |  \n",
      "     |  c\n",
      "     |  \n",
      "     |  extrapolate\n",
      "     |  \n",
      "     |  x\n",
      "    \n",
      "    class BSpline(builtins.object)\n",
      "     |  BSpline(t, c, k, extrapolate=True, axis=0)\n",
      "     |  \n",
      "     |  Univariate spline in the B-spline basis.\n",
      "     |  \n",
      "     |  .. math::\n",
      "     |  \n",
      "     |      S(x) = \\sum_{j=0}^{n-1} c_j  B_{j, k; t}(x)\n",
      "     |  \n",
      "     |  where :math:`B_{j, k; t}` are B-spline basis functions of degree `k`\n",
      "     |  and knots `t`.\n",
      "     |  \n",
      "     |  Parameters\n",
      "     |  ----------\n",
      "     |  t : ndarray, shape (n+k+1,)\n",
      "     |      knots\n",
      "     |  c : ndarray, shape (>=n, ...)\n",
      "     |      spline coefficients\n",
      "     |  k : int\n",
      "     |      B-spline degree\n",
      "     |  extrapolate : bool or 'periodic', optional\n",
      "     |      whether to extrapolate beyond the base interval, ``t[k] .. t[n]``,\n",
      "     |      or to return nans.\n",
      "     |      If True, extrapolates the first and last polynomial pieces of b-spline\n",
      "     |      functions active on the base interval.\n",
      "     |      If 'periodic', periodic extrapolation is used.\n",
      "     |      Default is True.\n",
      "     |  axis : int, optional\n",
      "     |      Interpolation axis. Default is zero.\n",
      "     |  \n",
      "     |  Attributes\n",
      "     |  ----------\n",
      "     |  t : ndarray\n",
      "     |      knot vector\n",
      "     |  c : ndarray\n",
      "     |      spline coefficients\n",
      "     |  k : int\n",
      "     |      spline degree\n",
      "     |  extrapolate : bool\n",
      "     |      If True, extrapolates the first and last polynomial pieces of b-spline\n",
      "     |      functions active on the base interval.\n",
      "     |  axis : int\n",
      "     |      Interpolation axis.\n",
      "     |  tck : tuple\n",
      "     |      A read-only equivalent of ``(self.t, self.c, self.k)``\n",
      "     |  \n",
      "     |  Methods\n",
      "     |  -------\n",
      "     |  __call__\n",
      "     |  basis_element\n",
      "     |  derivative\n",
      "     |  antiderivative\n",
      "     |  integrate\n",
      "     |  construct_fast\n",
      "     |  \n",
      "     |  Notes\n",
      "     |  -----\n",
      "     |  B-spline basis elements are defined via\n",
      "     |  \n",
      "     |  .. math::\n",
      "     |  \n",
      "     |      B_{i, 0}(x) = 1, \\textrm{if $t_i \\le x < t_{i+1}$, otherwise $0$,}\n",
      "     |  \n",
      "     |      B_{i, k}(x) = \\frac{x - t_i}{t_{i+k} - t_i} B_{i, k-1}(x)\n",
      "     |               + \\frac{t_{i+k+1} - x}{t_{i+k+1} - t_{i+1}} B_{i+1, k-1}(x)\n",
      "     |  \n",
      "     |  **Implementation details**\n",
      "     |  \n",
      "     |  - At least ``k+1`` coefficients are required for a spline of degree `k`,\n",
      "     |    so that ``n >= k+1``. Additional coefficients, ``c[j]`` with\n",
      "     |    ``j > n``, are ignored.\n",
      "     |  \n",
      "     |  - B-spline basis elements of degree `k` form a partition of unity on the\n",
      "     |    *base interval*, ``t[k] <= x <= t[n]``.\n",
      "     |  \n",
      "     |  \n",
      "     |  Examples\n",
      "     |  --------\n",
      "     |  \n",
      "     |  Translating the recursive definition of B-splines into Python code, we have:\n",
      "     |  \n",
      "     |  >>> def B(x, k, i, t):\n",
      "     |  ...    if k == 0:\n",
      "     |  ...       return 1.0 if t[i] <= x < t[i+1] else 0.0\n",
      "     |  ...    if t[i+k] == t[i]:\n",
      "     |  ...       c1 = 0.0\n",
      "     |  ...    else:\n",
      "     |  ...       c1 = (x - t[i])/(t[i+k] - t[i]) * B(x, k-1, i, t)\n",
      "     |  ...    if t[i+k+1] == t[i+1]:\n",
      "     |  ...       c2 = 0.0\n",
      "     |  ...    else:\n",
      "     |  ...       c2 = (t[i+k+1] - x)/(t[i+k+1] - t[i+1]) * B(x, k-1, i+1, t)\n",
      "     |  ...    return c1 + c2\n",
      "     |  \n",
      "     |  >>> def bspline(x, t, c, k):\n",
      "     |  ...    n = len(t) - k - 1\n",
      "     |  ...    assert (n >= k+1) and (len(c) >= n)\n",
      "     |  ...    return sum(c[i] * B(x, k, i, t) for i in range(n))\n",
      "     |  \n",
      "     |  Note that this is an inefficient (if straightforward) way to\n",
      "     |  evaluate B-splines --- this spline class does it in an equivalent,\n",
      "     |  but much more efficient way.\n",
      "     |  \n",
      "     |  Here we construct a quadratic spline function on the base interval\n",
      "     |  ``2 <= x <= 4`` and compare with the naive way of evaluating the spline:\n",
      "     |  \n",
      "     |  >>> from scipy.interpolate import BSpline\n",
      "     |  >>> k = 2\n",
      "     |  >>> t = [0, 1, 2, 3, 4, 5, 6]\n",
      "     |  >>> c = [-1, 2, 0, -1]\n",
      "     |  >>> spl = BSpline(t, c, k)\n",
      "     |  >>> spl(2.5)\n",
      "     |  array(1.375)\n",
      "     |  >>> bspline(2.5, t, c, k)\n",
      "     |  1.375\n",
      "     |  \n",
      "     |  Note that outside of the base interval results differ. This is because\n",
      "     |  `BSpline` extrapolates the first and last polynomial pieces of B-spline\n",
      "     |  functions active on the base interval.\n",
      "     |  \n",
      "     |  >>> import matplotlib.pyplot as plt\n",
      "     |  >>> fig, ax = plt.subplots()\n",
      "     |  >>> xx = np.linspace(1.5, 4.5, 50)\n",
      "     |  >>> ax.plot(xx, [bspline(x, t, c ,k) for x in xx], 'r-', lw=3, label='naive')\n",
      "     |  >>> ax.plot(xx, spl(xx), 'b-', lw=4, alpha=0.7, label='BSpline')\n",
      "     |  >>> ax.grid(True)\n",
      "     |  >>> ax.legend(loc='best')\n",
      "     |  >>> plt.show()\n",
      "     |  \n",
      "     |  \n",
      "     |  References\n",
      "     |  ----------\n",
      "     |  .. [1] Tom Lyche and Knut Morken, Spline methods,\n",
      "     |      http://www.uio.no/studier/emner/matnat/ifi/INF-MAT5340/v05/undervisningsmateriale/\n",
      "     |  .. [2] Carl de Boor, A practical guide to splines, Springer, 2001.\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __call__(self, x, nu=0, extrapolate=None)\n",
      "     |      Evaluate a spline function.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      x : array_like\n",
      "     |          points to evaluate the spline at.\n",
      "     |      nu: int, optional\n",
      "     |          derivative to evaluate (default is 0).\n",
      "     |      extrapolate : bool or 'periodic', optional\n",
      "     |          whether to extrapolate based on the first and last intervals\n",
      "     |          or return nans. If 'periodic', periodic extrapolation is used.\n",
      "     |          Default is `self.extrapolate`.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      y : array_like\n",
      "     |          Shape is determined by replacing the interpolation axis\n",
      "     |          in the coefficient array with the shape of `x`.\n",
      "     |  \n",
      "     |  __init__(self, t, c, k, extrapolate=True, axis=0)\n",
      "     |      Initialize self.  See help(type(self)) for accurate signature.\n",
      "     |  \n",
      "     |  antiderivative(self, nu=1)\n",
      "     |      Return a B-spline representing the antiderivative.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      nu : int, optional\n",
      "     |          Antiderivative order. Default is 1.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      b : BSpline object\n",
      "     |          A new instance representing the antiderivative.\n",
      "     |      \n",
      "     |      Notes\n",
      "     |      -----\n",
      "     |      If antiderivative is computed and ``self.extrapolate='periodic'``,\n",
      "     |      it will be set to False for the returned instance. This is done because\n",
      "     |      the antiderivative is no longer periodic and its correct evaluation\n",
      "     |      outside of the initially given x interval is difficult.\n",
      "     |      \n",
      "     |      See Also\n",
      "     |      --------\n",
      "     |      splder, splantider\n",
      "     |  \n",
      "     |  derivative(self, nu=1)\n",
      "     |      Return a B-spline representing the derivative.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      nu : int, optional\n",
      "     |          Derivative order.\n",
      "     |          Default is 1.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      b : BSpline object\n",
      "     |          A new instance representing the derivative.\n",
      "     |      \n",
      "     |      See Also\n",
      "     |      --------\n",
      "     |      splder, splantider\n",
      "     |  \n",
      "     |  integrate(self, a, b, extrapolate=None)\n",
      "     |      Compute a definite integral of the spline.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      a : float\n",
      "     |          Lower limit of integration.\n",
      "     |      b : float\n",
      "     |          Upper limit of integration.\n",
      "     |      extrapolate : bool or 'periodic', optional\n",
      "     |          whether to extrapolate beyond the base interval,\n",
      "     |          ``t[k] .. t[-k-1]``, or take the spline to be zero outside of the\n",
      "     |          base interval. If 'periodic', periodic extrapolation is used.\n",
      "     |          If None (default), use `self.extrapolate`.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      I : array_like\n",
      "     |          Definite integral of the spline over the interval ``[a, b]``.\n",
      "     |      \n",
      "     |      Examples\n",
      "     |      --------\n",
      "     |      Construct the linear spline ``x if x < 1 else 2 - x`` on the base\n",
      "     |      interval :math:`[0, 2]`, and integrate it\n",
      "     |      \n",
      "     |      >>> from scipy.interpolate import BSpline\n",
      "     |      >>> b = BSpline.basis_element([0, 1, 2])\n",
      "     |      >>> b.integrate(0, 1)\n",
      "     |      array(0.5)\n",
      "     |      \n",
      "     |      If the integration limits are outside of the base interval, the result\n",
      "     |      is controlled by the `extrapolate` parameter\n",
      "     |      \n",
      "     |      >>> b.integrate(-1, 1)\n",
      "     |      array(0.0)\n",
      "     |      >>> b.integrate(-1, 1, extrapolate=False)\n",
      "     |      array(0.5)\n",
      "     |      \n",
      "     |      >>> import matplotlib.pyplot as plt\n",
      "     |      >>> fig, ax = plt.subplots()\n",
      "     |      >>> ax.grid(True)\n",
      "     |      >>> ax.axvline(0, c='r', lw=5, alpha=0.5)  # base interval\n",
      "     |      >>> ax.axvline(2, c='r', lw=5, alpha=0.5)\n",
      "     |      >>> xx = [-1, 1, 2]\n",
      "     |      >>> ax.plot(xx, b(xx))\n",
      "     |      >>> plt.show()\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Class methods defined here:\n",
      "     |  \n",
      "     |  basis_element(t, extrapolate=True) from builtins.type\n",
      "     |      Return a B-spline basis element ``B(x | t[0], ..., t[k+1])``.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      t : ndarray, shape (k+1,)\n",
      "     |          internal knots\n",
      "     |      extrapolate : bool or 'periodic', optional\n",
      "     |          whether to extrapolate beyond the base interval, ``t[0] .. t[k+1]``,\n",
      "     |          or to return nans.\n",
      "     |          If 'periodic', periodic extrapolation is used.\n",
      "     |          Default is True.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      basis_element : callable\n",
      "     |          A callable representing a B-spline basis element for the knot\n",
      "     |          vector `t`.\n",
      "     |      \n",
      "     |      Notes\n",
      "     |      -----\n",
      "     |      The degree of the B-spline, `k`, is inferred from the length of `t` as\n",
      "     |      ``len(t)-2``. The knot vector is constructed by appending and prepending\n",
      "     |      ``k+1`` elements to internal knots `t`.\n",
      "     |      \n",
      "     |      Examples\n",
      "     |      --------\n",
      "     |      \n",
      "     |      Construct a cubic B-spline:\n",
      "     |      \n",
      "     |      >>> from scipy.interpolate import BSpline\n",
      "     |      >>> b = BSpline.basis_element([0, 1, 2, 3, 4])\n",
      "     |      >>> k = b.k\n",
      "     |      >>> b.t[k:-k]\n",
      "     |      array([ 0.,  1.,  2.,  3.,  4.])\n",
      "     |      >>> k\n",
      "     |      3\n",
      "     |      \n",
      "     |      Construct a quadratic B-spline on ``[0, 1, 1, 2]``, and compare\n",
      "     |      to its explicit form:\n",
      "     |      \n",
      "     |      >>> t = [-1, 0, 1, 1, 2]\n",
      "     |      >>> b = BSpline.basis_element(t[1:])\n",
      "     |      >>> def f(x):\n",
      "     |      ...     return np.where(x < 1, x*x, (2. - x)**2)\n",
      "     |      \n",
      "     |      >>> import matplotlib.pyplot as plt\n",
      "     |      >>> fig, ax = plt.subplots()\n",
      "     |      >>> x = np.linspace(0, 2, 51)\n",
      "     |      >>> ax.plot(x, b(x), 'g', lw=3)\n",
      "     |      >>> ax.plot(x, f(x), 'r', lw=8, alpha=0.4)\n",
      "     |      >>> ax.grid(True)\n",
      "     |      >>> plt.show()\n",
      "     |  \n",
      "     |  construct_fast(t, c, k, extrapolate=True, axis=0) from builtins.type\n",
      "     |      Construct a spline without making checks.\n",
      "     |      \n",
      "     |      Accepts same parameters as the regular constructor. Input arrays\n",
      "     |      `t` and `c` must of correct shape and dtype.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Readonly properties defined here:\n",
      "     |  \n",
      "     |  tck\n",
      "     |      Equivalent to ``(self.t, self.c, self.k)`` (read-only).\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors defined here:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |      dictionary for instance variables (if defined)\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "    \n",
      "    class BarycentricInterpolator(_Interpolator1D)\n",
      "     |  BarycentricInterpolator(xi, yi=None, axis=0)\n",
      "     |  \n",
      "     |  The interpolating polynomial for a set of points\n",
      "     |  \n",
      "     |  Constructs a polynomial that passes through a given set of points.\n",
      "     |  Allows evaluation of the polynomial, efficient changing of the y\n",
      "     |  values to be interpolated, and updating by adding more x values.\n",
      "     |  For reasons of numerical stability, this function does not compute\n",
      "     |  the coefficients of the polynomial.\n",
      "     |  \n",
      "     |  The values yi need to be provided before the function is\n",
      "     |  evaluated, but none of the preprocessing depends on them, so rapid\n",
      "     |  updates are possible.\n",
      "     |  \n",
      "     |  Parameters\n",
      "     |  ----------\n",
      "     |  xi : array_like\n",
      "     |      1-D array of x coordinates of the points the polynomial\n",
      "     |      should pass through\n",
      "     |  yi : array_like, optional\n",
      "     |      The y coordinates of the points the polynomial should pass through.\n",
      "     |      If None, the y values will be supplied later via the `set_y` method.\n",
      "     |  axis : int, optional\n",
      "     |      Axis in the yi array corresponding to the x-coordinate values.\n",
      "     |  \n",
      "     |  Notes\n",
      "     |  -----\n",
      "     |  This class uses a \"barycentric interpolation\" method that treats\n",
      "     |  the problem as a special case of rational function interpolation.\n",
      "     |  This algorithm is quite stable, numerically, but even in a world of\n",
      "     |  exact computation, unless the x coordinates are chosen very\n",
      "     |  carefully - Chebyshev zeros (e.g., cos(i*pi/n)) are a good choice -\n",
      "     |  polynomial interpolation itself is a very ill-conditioned process\n",
      "     |  due to the Runge phenomenon.\n",
      "     |  \n",
      "     |  Based on Berrut and Trefethen 2004, \"Barycentric Lagrange Interpolation\".\n",
      "     |  \n",
      "     |  Method resolution order:\n",
      "     |      BarycentricInterpolator\n",
      "     |      _Interpolator1D\n",
      "     |      builtins.object\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __call__(self, x)\n",
      "     |      Evaluate the interpolating polynomial at the points x\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      x : array_like\n",
      "     |          Points to evaluate the interpolant at.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      y : array_like\n",
      "     |          Interpolated values. Shape is determined by replacing\n",
      "     |          the interpolation axis in the original array with the shape of x.\n",
      "     |      \n",
      "     |      Notes\n",
      "     |      -----\n",
      "     |      Currently the code computes an outer product between x and the\n",
      "     |      weights, that is, it constructs an intermediate array of size\n",
      "     |      N by len(x), where N is the degree of the polynomial.\n",
      "     |  \n",
      "     |  __init__(self, xi, yi=None, axis=0)\n",
      "     |      Initialize self.  See help(type(self)) for accurate signature.\n",
      "     |  \n",
      "     |  add_xi(self, xi, yi=None)\n",
      "     |      Add more x values to the set to be interpolated\n",
      "     |      \n",
      "     |      The barycentric interpolation algorithm allows easy updating by\n",
      "     |      adding more points for the polynomial to pass through.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      xi : array_like\n",
      "     |          The x coordinates of the points that the polynomial should pass\n",
      "     |          through.\n",
      "     |      yi : array_like, optional\n",
      "     |          The y coordinates of the points the polynomial should pass through.\n",
      "     |          Should have shape ``(xi.size, R)``; if R > 1 then the polynomial is\n",
      "     |          vector-valued.\n",
      "     |          If `yi` is not given, the y values will be supplied later. `yi` should\n",
      "     |          be given if and only if the interpolator has y values specified.\n",
      "     |  \n",
      "     |  set_yi(self, yi, axis=None)\n",
      "     |      Update the y values to be interpolated\n",
      "     |      \n",
      "     |      The barycentric interpolation algorithm requires the calculation\n",
      "     |      of weights, but these depend only on the xi. The yi can be changed\n",
      "     |      at any time.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      yi : array_like\n",
      "     |          The y coordinates of the points the polynomial should pass through.\n",
      "     |          If None, the y values will be supplied later.\n",
      "     |      axis : int, optional\n",
      "     |          Axis in the yi array corresponding to the x-coordinate values.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors defined here:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |      dictionary for instance variables (if defined)\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from _Interpolator1D:\n",
      "     |  \n",
      "     |  dtype\n",
      "    \n",
      "    class BivariateSpline(_BivariateSplineBase)\n",
      "     |  Base class for bivariate splines.\n",
      "     |  \n",
      "     |  This describes a spline ``s(x, y)`` of degrees ``kx`` and ``ky`` on\n",
      "     |  the rectangle ``[xb, xe] * [yb, ye]`` calculated from a given set\n",
      "     |  of data points ``(x, y, z)``.\n",
      "     |  \n",
      "     |  This class is meant to be subclassed, not instantiated directly.\n",
      "     |  To construct these splines, call either `SmoothBivariateSpline` or\n",
      "     |  `LSQBivariateSpline` or `RectBivariateSpline`.\n",
      "     |  \n",
      "     |  See Also\n",
      "     |  --------\n",
      "     |  UnivariateSpline :\n",
      "     |      a smooth univariate spline to fit a given set of data points.\n",
      "     |  SmoothBivariateSpline :\n",
      "     |      a smoothing bivariate spline through the given points\n",
      "     |  LSQBivariateSpline :\n",
      "     |      a bivariate spline using weighted least-squares fitting\n",
      "     |  RectSphereBivariateSpline :\n",
      "     |      a bivariate spline over a rectangular mesh on a sphere\n",
      "     |  SmoothSphereBivariateSpline :\n",
      "     |      a smoothing bivariate spline in spherical coordinates\n",
      "     |  LSQSphereBivariateSpline :\n",
      "     |      a bivariate spline in spherical coordinates using weighted\n",
      "     |      least-squares fitting\n",
      "     |  RectBivariateSpline :\n",
      "     |      a bivariate spline over a rectangular mesh.\n",
      "     |  bisplrep :\n",
      "     |      a function to find a bivariate B-spline representation of a surface\n",
      "     |  bisplev :\n",
      "     |      a function to evaluate a bivariate B-spline and its derivatives\n",
      "     |  \n",
      "     |  Method resolution order:\n",
      "     |      BivariateSpline\n",
      "     |      _BivariateSplineBase\n",
      "     |      builtins.object\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  ev(self, xi, yi, dx=0, dy=0)\n",
      "     |      Evaluate the spline at points\n",
      "     |      \n",
      "     |      Returns the interpolated value at ``(xi[i], yi[i]),\n",
      "     |      i=0,...,len(xi)-1``.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      xi, yi : array_like\n",
      "     |          Input coordinates. Standard Numpy broadcasting is obeyed.\n",
      "     |      dx : int, optional\n",
      "     |          Order of x-derivative\n",
      "     |      \n",
      "     |          .. versionadded:: 0.14.0\n",
      "     |      dy : int, optional\n",
      "     |          Order of y-derivative\n",
      "     |      \n",
      "     |          .. versionadded:: 0.14.0\n",
      "     |  \n",
      "     |  integral(self, xa, xb, ya, yb)\n",
      "     |      Evaluate the integral of the spline over area [xa,xb] x [ya,yb].\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      xa, xb : float\n",
      "     |          The end-points of the x integration interval.\n",
      "     |      ya, yb : float\n",
      "     |          The end-points of the y integration interval.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      integ : float\n",
      "     |          The value of the resulting integral.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from _BivariateSplineBase:\n",
      "     |  \n",
      "     |  __call__(self, x, y, dx=0, dy=0, grid=True)\n",
      "     |      Evaluate the spline or its derivatives at given positions.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      x, y : array_like\n",
      "     |          Input coordinates.\n",
      "     |      \n",
      "     |          If `grid` is False, evaluate the spline at points ``(x[i],\n",
      "     |          y[i]), i=0, ..., len(x)-1``.  Standard Numpy broadcasting\n",
      "     |          is obeyed.\n",
      "     |      \n",
      "     |          If `grid` is True: evaluate spline at the grid points\n",
      "     |          defined by the coordinate arrays x, y. The arrays must be\n",
      "     |          sorted to increasing order.\n",
      "     |      \n",
      "     |          Note that the axis ordering is inverted relative to\n",
      "     |          the output of meshgrid.\n",
      "     |      dx : int\n",
      "     |          Order of x-derivative\n",
      "     |      \n",
      "     |          .. versionadded:: 0.14.0\n",
      "     |      dy : int\n",
      "     |          Order of y-derivative\n",
      "     |      \n",
      "     |          .. versionadded:: 0.14.0\n",
      "     |      grid : bool\n",
      "     |          Whether to evaluate the results on a grid spanned by the\n",
      "     |          input arrays, or at points specified by the input arrays.\n",
      "     |      \n",
      "     |          .. versionadded:: 0.14.0\n",
      "     |  \n",
      "     |  get_coeffs(self)\n",
      "     |      Return spline coefficients.\n",
      "     |  \n",
      "     |  get_knots(self)\n",
      "     |      Return a tuple (tx,ty) where tx,ty contain knots positions\n",
      "     |      of the spline with respect to x-, y-variable, respectively.\n",
      "     |      The position of interior and additional knots are given as\n",
      "     |      t[k+1:-k-1] and t[:k+1]=b, t[-k-1:]=e, respectively.\n",
      "     |  \n",
      "     |  get_residual(self)\n",
      "     |      Return weighted sum of squared residuals of the spline\n",
      "     |      approximation: sum ((w[i]*(z[i]-s(x[i],y[i])))**2,axis=0)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from _BivariateSplineBase:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |      dictionary for instance variables (if defined)\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "    \n",
      "    class CloughTocher2DInterpolator(NDInterpolatorBase)\n",
      "     |  CloughTocher2DInterpolator(points, values, fill_value=nan, tol=1e-06, maxiter=400, rescale=False)\n",
      "     |  \n",
      "     |  CloughTocher2DInterpolator(points, values, tol=1e-6).\n",
      "     |  \n",
      "     |  Piecewise cubic, C1 smooth, curvature-minimizing interpolant in 2D.\n",
      "     |  \n",
      "     |  .. versionadded:: 0.9\n",
      "     |  \n",
      "     |  Methods\n",
      "     |  -------\n",
      "     |  __call__\n",
      "     |  \n",
      "     |  Parameters\n",
      "     |  ----------\n",
      "     |  points : ndarray of floats, shape (npoints, ndims); or Delaunay\n",
      "     |      Data point coordinates, or a precomputed Delaunay triangulation.\n",
      "     |  values : ndarray of float or complex, shape (npoints, ...)\n",
      "     |      Data values.\n",
      "     |  fill_value : float, optional\n",
      "     |      Value used to fill in for requested points outside of the\n",
      "     |      convex hull of the input points.  If not provided, then\n",
      "     |      the default is ``nan``.\n",
      "     |  tol : float, optional\n",
      "     |      Absolute/relative tolerance for gradient estimation.\n",
      "     |  maxiter : int, optional\n",
      "     |      Maximum number of iterations in gradient estimation.\n",
      "     |  rescale : bool, optional\n",
      "     |      Rescale points to unit cube before performing interpolation.\n",
      "     |      This is useful if some of the input dimensions have\n",
      "     |      incommensurable units and differ by many orders of magnitude.\n",
      "     |  \n",
      "     |  Notes\n",
      "     |  -----\n",
      "     |  The interpolant is constructed by triangulating the input data\n",
      "     |  with Qhull [1]_, and constructing a piecewise cubic\n",
      "     |  interpolating Bezier polynomial on each triangle, using a\n",
      "     |  Clough-Tocher scheme [CT]_.  The interpolant is guaranteed to be\n",
      "     |  continuously differentiable.\n",
      "     |  \n",
      "     |  The gradients of the interpolant are chosen so that the curvature\n",
      "     |  of the interpolating surface is approximatively minimized. The\n",
      "     |  gradients necessary for this are estimated using the global\n",
      "     |  algorithm described in [Nielson83]_ and [Renka84]_.\n",
      "     |  \n",
      "     |  Examples\n",
      "     |  --------\n",
      "     |  We can interpolate values on a 2D plane:\n",
      "     |  \n",
      "     |  >>> from scipy.interpolate import CloughTocher2DInterpolator\n",
      "     |  >>> import matplotlib.pyplot as plt\n",
      "     |  >>> rng = np.random.default_rng()\n",
      "     |  >>> x = rng.random(10) - 0.5\n",
      "     |  >>> y = rng.random(10) - 0.5\n",
      "     |  >>> z = np.hypot(x, y)\n",
      "     |  >>> X = np.linspace(min(x), max(x))\n",
      "     |  >>> Y = np.linspace(min(y), max(y))\n",
      "     |  >>> X, Y = np.meshgrid(X, Y)  # 2D grid for interpolation\n",
      "     |  >>> interp = CloughTocher2DInterpolator(list(zip(x, y)), z)\n",
      "     |  >>> Z = interp(X, Y)\n",
      "     |  >>> plt.pcolormesh(X, Y, Z, shading='auto')\n",
      "     |  >>> plt.plot(x, y, \"ok\", label=\"input point\")\n",
      "     |  >>> plt.legend()\n",
      "     |  >>> plt.colorbar()\n",
      "     |  >>> plt.axis(\"equal\")\n",
      "     |  >>> plt.show()\n",
      "     |  \n",
      "     |  See also\n",
      "     |  --------\n",
      "     |  griddata :\n",
      "     |      Interpolate unstructured D-D data.\n",
      "     |  LinearNDInterpolator :\n",
      "     |      Piecewise linear interpolant in N > 1 dimensions.\n",
      "     |  NearestNDInterpolator :\n",
      "     |      Nearest-neighbor interpolation in N > 1 dimensions.\n",
      "     |  \n",
      "     |  References\n",
      "     |  ----------\n",
      "     |  .. [1] http://www.qhull.org/\n",
      "     |  \n",
      "     |  .. [CT] See, for example,\n",
      "     |     P. Alfeld,\n",
      "     |     ''A trivariate Clough-Tocher scheme for tetrahedral data''.\n",
      "     |     Computer Aided Geometric Design, 1, 169 (1984);\n",
      "     |     G. Farin,\n",
      "     |     ''Triangular Bernstein-Bezier patches''.\n",
      "     |     Computer Aided Geometric Design, 3, 83 (1986).\n",
      "     |  \n",
      "     |  .. [Nielson83] G. Nielson,\n",
      "     |     ''A method for interpolating scattered data based upon a minimum norm\n",
      "     |     network''.\n",
      "     |     Math. Comp., 40, 253 (1983).\n",
      "     |  \n",
      "     |  .. [Renka84] R. J. Renka and A. K. Cline.\n",
      "     |     ''A Triangle-based C1 interpolation method.'',\n",
      "     |     Rocky Mountain J. Math., 14, 223 (1984).\n",
      "     |  \n",
      "     |  Method resolution order:\n",
      "     |      CloughTocher2DInterpolator\n",
      "     |      NDInterpolatorBase\n",
      "     |      builtins.object\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __init__(self, points, values, fill_value=nan, tol=1e-06, maxiter=400, rescale=False)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from NDInterpolatorBase:\n",
      "     |  \n",
      "     |  __call__(self, *args)\n",
      "     |      interpolator(xi)\n",
      "     |      \n",
      "     |      Evaluate interpolator at given points.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      x1, x2, ... xn: array-like of float\n",
      "     |          Points where to interpolate data at.\n",
      "     |          x1, x2, ... xn can be array-like of float with broadcastable shape.\n",
      "     |          or x1 can be array-like of float with shape ``(..., ndim)``\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from NDInterpolatorBase:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |      dictionary for instance variables (if defined)\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "    \n",
      "    class CubicHermiteSpline(scipy.interpolate.interpolate.PPoly)\n",
      "     |  CubicHermiteSpline(x, y, dydx, axis=0, extrapolate=None)\n",
      "     |  \n",
      "     |  Piecewise-cubic interpolator matching values and first derivatives.\n",
      "     |  \n",
      "     |  The result is represented as a `PPoly` instance.\n",
      "     |  \n",
      "     |  Parameters\n",
      "     |  ----------\n",
      "     |  x : array_like, shape (n,)\n",
      "     |      1-D array containing values of the independent variable.\n",
      "     |      Values must be real, finite and in strictly increasing order.\n",
      "     |  y : array_like\n",
      "     |      Array containing values of the dependent variable. It can have\n",
      "     |      arbitrary number of dimensions, but the length along ``axis``\n",
      "     |      (see below) must match the length of ``x``. Values must be finite.\n",
      "     |  dydx : array_like\n",
      "     |      Array containing derivatives of the dependent variable. It can have\n",
      "     |      arbitrary number of dimensions, but the length along ``axis``\n",
      "     |      (see below) must match the length of ``x``. Values must be finite.\n",
      "     |  axis : int, optional\n",
      "     |      Axis along which `y` is assumed to be varying. Meaning that for\n",
      "     |      ``x[i]`` the corresponding values are ``np.take(y, i, axis=axis)``.\n",
      "     |      Default is 0.\n",
      "     |  extrapolate : {bool, 'periodic', None}, optional\n",
      "     |      If bool, determines whether to extrapolate to out-of-bounds points\n",
      "     |      based on first and last intervals, or to return NaNs. If 'periodic',\n",
      "     |      periodic extrapolation is used. If None (default), it is set to True.\n",
      "     |  \n",
      "     |  Attributes\n",
      "     |  ----------\n",
      "     |  x : ndarray, shape (n,)\n",
      "     |      Breakpoints. The same ``x`` which was passed to the constructor.\n",
      "     |  c : ndarray, shape (4, n-1, ...)\n",
      "     |      Coefficients of the polynomials on each segment. The trailing\n",
      "     |      dimensions match the dimensions of `y`, excluding ``axis``.\n",
      "     |      For example, if `y` is 1-D, then ``c[k, i]`` is a coefficient for\n",
      "     |      ``(x-x[i])**(3-k)`` on the segment between ``x[i]`` and ``x[i+1]``.\n",
      "     |  axis : int\n",
      "     |      Interpolation axis. The same axis which was passed to the\n",
      "     |      constructor.\n",
      "     |  \n",
      "     |  Methods\n",
      "     |  -------\n",
      "     |  __call__\n",
      "     |  derivative\n",
      "     |  antiderivative\n",
      "     |  integrate\n",
      "     |  roots\n",
      "     |  \n",
      "     |  See Also\n",
      "     |  --------\n",
      "     |  Akima1DInterpolator : Akima 1D interpolator.\n",
      "     |  PchipInterpolator : PCHIP 1-D monotonic cubic interpolator.\n",
      "     |  CubicSpline : Cubic spline data interpolator.\n",
      "     |  PPoly : Piecewise polynomial in terms of coefficients and breakpoints\n",
      "     |  \n",
      "     |  Notes\n",
      "     |  -----\n",
      "     |  If you want to create a higher-order spline matching higher-order\n",
      "     |  derivatives, use `BPoly.from_derivatives`.\n",
      "     |  \n",
      "     |  References\n",
      "     |  ----------\n",
      "     |  .. [1] `Cubic Hermite spline\n",
      "     |          <https://en.wikipedia.org/wiki/Cubic_Hermite_spline>`_\n",
      "     |          on Wikipedia.\n",
      "     |  \n",
      "     |  Method resolution order:\n",
      "     |      CubicHermiteSpline\n",
      "     |      scipy.interpolate.interpolate.PPoly\n",
      "     |      scipy.interpolate.interpolate._PPolyBase\n",
      "     |      builtins.object\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __init__(self, x, y, dydx, axis=0, extrapolate=None)\n",
      "     |      Initialize self.  See help(type(self)) for accurate signature.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from scipy.interpolate.interpolate.PPoly:\n",
      "     |  \n",
      "     |  antiderivative(self, nu=1)\n",
      "     |      Construct a new piecewise polynomial representing the antiderivative.\n",
      "     |      \n",
      "     |      Antiderivative is also the indefinite integral of the function,\n",
      "     |      and derivative is its inverse operation.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      nu : int, optional\n",
      "     |          Order of antiderivative to evaluate. Default is 1, i.e., compute\n",
      "     |          the first integral. If negative, the derivative is returned.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      pp : PPoly\n",
      "     |          Piecewise polynomial of order k2 = k + n representing\n",
      "     |          the antiderivative of this polynomial.\n",
      "     |      \n",
      "     |      Notes\n",
      "     |      -----\n",
      "     |      The antiderivative returned by this function is continuous and\n",
      "     |      continuously differentiable to order n-1, up to floating point\n",
      "     |      rounding error.\n",
      "     |      \n",
      "     |      If antiderivative is computed and ``self.extrapolate='periodic'``,\n",
      "     |      it will be set to False for the returned instance. This is done because\n",
      "     |      the antiderivative is no longer periodic and its correct evaluation\n",
      "     |      outside of the initially given x interval is difficult.\n",
      "     |  \n",
      "     |  derivative(self, nu=1)\n",
      "     |      Construct a new piecewise polynomial representing the derivative.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      nu : int, optional\n",
      "     |          Order of derivative to evaluate. Default is 1, i.e., compute the\n",
      "     |          first derivative. If negative, the antiderivative is returned.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      pp : PPoly\n",
      "     |          Piecewise polynomial of order k2 = k - n representing the derivative\n",
      "     |          of this polynomial.\n",
      "     |      \n",
      "     |      Notes\n",
      "     |      -----\n",
      "     |      Derivatives are evaluated piecewise for each polynomial\n",
      "     |      segment, even if the polynomial is not differentiable at the\n",
      "     |      breakpoints. The polynomial intervals are considered half-open,\n",
      "     |      ``[a, b)``, except for the last interval which is closed\n",
      "     |      ``[a, b]``.\n",
      "     |  \n",
      "     |  integrate(self, a, b, extrapolate=None)\n",
      "     |      Compute a definite integral over a piecewise polynomial.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      a : float\n",
      "     |          Lower integration bound\n",
      "     |      b : float\n",
      "     |          Upper integration bound\n",
      "     |      extrapolate : {bool, 'periodic', None}, optional\n",
      "     |          If bool, determines whether to extrapolate to out-of-bounds points\n",
      "     |          based on first and last intervals, or to return NaNs.\n",
      "     |          If 'periodic', periodic extrapolation is used.\n",
      "     |          If None (default), use `self.extrapolate`.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      ig : array_like\n",
      "     |          Definite integral of the piecewise polynomial over [a, b]\n",
      "     |  \n",
      "     |  roots(self, discontinuity=True, extrapolate=None)\n",
      "     |      Find real roots of the the piecewise polynomial.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      discontinuity : bool, optional\n",
      "     |          Whether to report sign changes across discontinuities at\n",
      "     |          breakpoints as roots.\n",
      "     |      extrapolate : {bool, 'periodic', None}, optional\n",
      "     |          If bool, determines whether to return roots from the polynomial\n",
      "     |          extrapolated based on first and last intervals, 'periodic' works\n",
      "     |          the same as False. If None (default), use `self.extrapolate`.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      roots : ndarray\n",
      "     |          Roots of the polynomial(s).\n",
      "     |      \n",
      "     |          If the PPoly object describes multiple polynomials, the\n",
      "     |          return value is an object array whose each element is an\n",
      "     |          ndarray containing the roots.\n",
      "     |      \n",
      "     |      See Also\n",
      "     |      --------\n",
      "     |      PPoly.solve\n",
      "     |  \n",
      "     |  solve(self, y=0.0, discontinuity=True, extrapolate=None)\n",
      "     |      Find real solutions of the the equation ``pp(x) == y``.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      y : float, optional\n",
      "     |          Right-hand side. Default is zero.\n",
      "     |      discontinuity : bool, optional\n",
      "     |          Whether to report sign changes across discontinuities at\n",
      "     |          breakpoints as roots.\n",
      "     |      extrapolate : {bool, 'periodic', None}, optional\n",
      "     |          If bool, determines whether to return roots from the polynomial\n",
      "     |          extrapolated based on first and last intervals, 'periodic' works\n",
      "     |          the same as False. If None (default), use `self.extrapolate`.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      roots : ndarray\n",
      "     |          Roots of the polynomial(s).\n",
      "     |      \n",
      "     |          If the PPoly object describes multiple polynomials, the\n",
      "     |          return value is an object array whose each element is an\n",
      "     |          ndarray containing the roots.\n",
      "     |      \n",
      "     |      Notes\n",
      "     |      -----\n",
      "     |      This routine works only on real-valued polynomials.\n",
      "     |      \n",
      "     |      If the piecewise polynomial contains sections that are\n",
      "     |      identically zero, the root list will contain the start point\n",
      "     |      of the corresponding interval, followed by a ``nan`` value.\n",
      "     |      \n",
      "     |      If the polynomial is discontinuous across a breakpoint, and\n",
      "     |      there is a sign change across the breakpoint, this is reported\n",
      "     |      if the `discont` parameter is True.\n",
      "     |      \n",
      "     |      Examples\n",
      "     |      --------\n",
      "     |      \n",
      "     |      Finding roots of ``[x**2 - 1, (x - 1)**2]`` defined on intervals\n",
      "     |      ``[-2, 1], [1, 2]``:\n",
      "     |      \n",
      "     |      >>> from scipy.interpolate import PPoly\n",
      "     |      >>> pp = PPoly(np.array([[1, -4, 3], [1, 0, 0]]).T, [-2, 1, 2])\n",
      "     |      >>> pp.solve()\n",
      "     |      array([-1.,  1.])\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Class methods inherited from scipy.interpolate.interpolate.PPoly:\n",
      "     |  \n",
      "     |  from_bernstein_basis(bp, extrapolate=None) from builtins.type\n",
      "     |      Construct a piecewise polynomial in the power basis\n",
      "     |      from a polynomial in Bernstein basis.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      bp : BPoly\n",
      "     |          A Bernstein basis polynomial, as created by BPoly\n",
      "     |      extrapolate : bool or 'periodic', optional\n",
      "     |          If bool, determines whether to extrapolate to out-of-bounds points\n",
      "     |          based on first and last intervals, or to return NaNs.\n",
      "     |          If 'periodic', periodic extrapolation is used. Default is True.\n",
      "     |  \n",
      "     |  from_spline(tck, extrapolate=None) from builtins.type\n",
      "     |      Construct a piecewise polynomial from a spline\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      tck\n",
      "     |          A spline, as returned by `splrep` or a BSpline object.\n",
      "     |      extrapolate : bool or 'periodic', optional\n",
      "     |          If bool, determines whether to extrapolate to out-of-bounds points\n",
      "     |          based on first and last intervals, or to return NaNs.\n",
      "     |          If 'periodic', periodic extrapolation is used. Default is True.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from scipy.interpolate.interpolate.PPoly:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |      dictionary for instance variables (if defined)\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from scipy.interpolate.interpolate._PPolyBase:\n",
      "     |  \n",
      "     |  __call__(self, x, nu=0, extrapolate=None)\n",
      "     |      Evaluate the piecewise polynomial or its derivative.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      x : array_like\n",
      "     |          Points to evaluate the interpolant at.\n",
      "     |      nu : int, optional\n",
      "     |          Order of derivative to evaluate. Must be non-negative.\n",
      "     |      extrapolate : {bool, 'periodic', None}, optional\n",
      "     |          If bool, determines whether to extrapolate to out-of-bounds points\n",
      "     |          based on first and last intervals, or to return NaNs.\n",
      "     |          If 'periodic', periodic extrapolation is used.\n",
      "     |          If None (default), use `self.extrapolate`.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      y : array_like\n",
      "     |          Interpolated values. Shape is determined by replacing\n",
      "     |          the interpolation axis in the original array with the shape of x.\n",
      "     |      \n",
      "     |      Notes\n",
      "     |      -----\n",
      "     |      Derivatives are evaluated piecewise for each polynomial\n",
      "     |      segment, even if the polynomial is not differentiable at the\n",
      "     |      breakpoints. The polynomial intervals are considered half-open,\n",
      "     |      ``[a, b)``, except for the last interval which is closed\n",
      "     |      ``[a, b]``.\n",
      "     |  \n",
      "     |  extend(self, c, x, right=None)\n",
      "     |      Add additional breakpoints and coefficients to the polynomial.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      c : ndarray, size (k, m, ...)\n",
      "     |          Additional coefficients for polynomials in intervals. Note that\n",
      "     |          the first additional interval will be formed using one of the\n",
      "     |          ``self.x`` end points.\n",
      "     |      x : ndarray, size (m,)\n",
      "     |          Additional breakpoints. Must be sorted in the same order as\n",
      "     |          ``self.x`` and either to the right or to the left of the current\n",
      "     |          breakpoints.\n",
      "     |      right\n",
      "     |          Deprecated argument. Has no effect.\n",
      "     |      \n",
      "     |          .. deprecated:: 0.19\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Class methods inherited from scipy.interpolate.interpolate._PPolyBase:\n",
      "     |  \n",
      "     |  construct_fast(c, x, extrapolate=None, axis=0) from builtins.type\n",
      "     |      Construct the piecewise polynomial without making checks.\n",
      "     |      \n",
      "     |      Takes the same parameters as the constructor. Input arguments\n",
      "     |      ``c`` and ``x`` must be arrays of the correct shape and type. The\n",
      "     |      ``c`` array can only be of dtypes float and complex, and ``x``\n",
      "     |      array must have dtype float.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from scipy.interpolate.interpolate._PPolyBase:\n",
      "     |  \n",
      "     |  axis\n",
      "     |  \n",
      "     |  c\n",
      "     |  \n",
      "     |  extrapolate\n",
      "     |  \n",
      "     |  x\n",
      "    \n",
      "    class CubicSpline(CubicHermiteSpline)\n",
      "     |  CubicSpline(x, y, axis=0, bc_type='not-a-knot', extrapolate=None)\n",
      "     |  \n",
      "     |  Cubic spline data interpolator.\n",
      "     |  \n",
      "     |  Interpolate data with a piecewise cubic polynomial which is twice\n",
      "     |  continuously differentiable [1]_. The result is represented as a `PPoly`\n",
      "     |  instance with breakpoints matching the given data.\n",
      "     |  \n",
      "     |  Parameters\n",
      "     |  ----------\n",
      "     |  x : array_like, shape (n,)\n",
      "     |      1-D array containing values of the independent variable.\n",
      "     |      Values must be real, finite and in strictly increasing order.\n",
      "     |  y : array_like\n",
      "     |      Array containing values of the dependent variable. It can have\n",
      "     |      arbitrary number of dimensions, but the length along ``axis``\n",
      "     |      (see below) must match the length of ``x``. Values must be finite.\n",
      "     |  axis : int, optional\n",
      "     |      Axis along which `y` is assumed to be varying. Meaning that for\n",
      "     |      ``x[i]`` the corresponding values are ``np.take(y, i, axis=axis)``.\n",
      "     |      Default is 0.\n",
      "     |  bc_type : string or 2-tuple, optional\n",
      "     |      Boundary condition type. Two additional equations, given by the\n",
      "     |      boundary conditions, are required to determine all coefficients of\n",
      "     |      polynomials on each segment [2]_.\n",
      "     |  \n",
      "     |      If `bc_type` is a string, then the specified condition will be applied\n",
      "     |      at both ends of a spline. Available conditions are:\n",
      "     |  \n",
      "     |      * 'not-a-knot' (default): The first and second segment at a curve end\n",
      "     |        are the same polynomial. It is a good default when there is no\n",
      "     |        information on boundary conditions.\n",
      "     |      * 'periodic': The interpolated functions is assumed to be periodic\n",
      "     |        of period ``x[-1] - x[0]``. The first and last value of `y` must be\n",
      "     |        identical: ``y[0] == y[-1]``. This boundary condition will result in\n",
      "     |        ``y'[0] == y'[-1]`` and ``y''[0] == y''[-1]``.\n",
      "     |      * 'clamped': The first derivative at curves ends are zero. Assuming\n",
      "     |        a 1D `y`, ``bc_type=((1, 0.0), (1, 0.0))`` is the same condition.\n",
      "     |      * 'natural': The second derivative at curve ends are zero. Assuming\n",
      "     |        a 1D `y`, ``bc_type=((2, 0.0), (2, 0.0))`` is the same condition.\n",
      "     |  \n",
      "     |      If `bc_type` is a 2-tuple, the first and the second value will be\n",
      "     |      applied at the curve start and end respectively. The tuple values can\n",
      "     |      be one of the previously mentioned strings (except 'periodic') or a\n",
      "     |      tuple `(order, deriv_values)` allowing to specify arbitrary\n",
      "     |      derivatives at curve ends:\n",
      "     |  \n",
      "     |      * `order`: the derivative order, 1 or 2.\n",
      "     |      * `deriv_value`: array_like containing derivative values, shape must\n",
      "     |        be the same as `y`, excluding ``axis`` dimension. For example, if\n",
      "     |        `y` is 1-D, then `deriv_value` must be a scalar. If `y` is 3-D with\n",
      "     |        the shape (n0, n1, n2) and axis=2, then `deriv_value` must be 2-D\n",
      "     |        and have the shape (n0, n1).\n",
      "     |  extrapolate : {bool, 'periodic', None}, optional\n",
      "     |      If bool, determines whether to extrapolate to out-of-bounds points\n",
      "     |      based on first and last intervals, or to return NaNs. If 'periodic',\n",
      "     |      periodic extrapolation is used. If None (default), ``extrapolate`` is\n",
      "     |      set to 'periodic' for ``bc_type='periodic'`` and to True otherwise.\n",
      "     |  \n",
      "     |  Attributes\n",
      "     |  ----------\n",
      "     |  x : ndarray, shape (n,)\n",
      "     |      Breakpoints. The same ``x`` which was passed to the constructor.\n",
      "     |  c : ndarray, shape (4, n-1, ...)\n",
      "     |      Coefficients of the polynomials on each segment. The trailing\n",
      "     |      dimensions match the dimensions of `y`, excluding ``axis``.\n",
      "     |      For example, if `y` is 1-d, then ``c[k, i]`` is a coefficient for\n",
      "     |      ``(x-x[i])**(3-k)`` on the segment between ``x[i]`` and ``x[i+1]``.\n",
      "     |  axis : int\n",
      "     |      Interpolation axis. The same axis which was passed to the\n",
      "     |      constructor.\n",
      "     |  \n",
      "     |  Methods\n",
      "     |  -------\n",
      "     |  __call__\n",
      "     |  derivative\n",
      "     |  antiderivative\n",
      "     |  integrate\n",
      "     |  roots\n",
      "     |  \n",
      "     |  See Also\n",
      "     |  --------\n",
      "     |  Akima1DInterpolator : Akima 1D interpolator.\n",
      "     |  PchipInterpolator : PCHIP 1-D monotonic cubic interpolator.\n",
      "     |  PPoly : Piecewise polynomial in terms of coefficients and breakpoints.\n",
      "     |  \n",
      "     |  Notes\n",
      "     |  -----\n",
      "     |  Parameters `bc_type` and ``interpolate`` work independently, i.e. the\n",
      "     |  former controls only construction of a spline, and the latter only\n",
      "     |  evaluation.\n",
      "     |  \n",
      "     |  When a boundary condition is 'not-a-knot' and n = 2, it is replaced by\n",
      "     |  a condition that the first derivative is equal to the linear interpolant\n",
      "     |  slope. When both boundary conditions are 'not-a-knot' and n = 3, the\n",
      "     |  solution is sought as a parabola passing through given points.\n",
      "     |  \n",
      "     |  When 'not-a-knot' boundary conditions is applied to both ends, the\n",
      "     |  resulting spline will be the same as returned by `splrep` (with ``s=0``)\n",
      "     |  and `InterpolatedUnivariateSpline`, but these two methods use a\n",
      "     |  representation in B-spline basis.\n",
      "     |  \n",
      "     |  .. versionadded:: 0.18.0\n",
      "     |  \n",
      "     |  Examples\n",
      "     |  --------\n",
      "     |  In this example the cubic spline is used to interpolate a sampled sinusoid.\n",
      "     |  You can see that the spline continuity property holds for the first and\n",
      "     |  second derivatives and violates only for the third derivative.\n",
      "     |  \n",
      "     |  >>> from scipy.interpolate import CubicSpline\n",
      "     |  >>> import matplotlib.pyplot as plt\n",
      "     |  >>> x = np.arange(10)\n",
      "     |  >>> y = np.sin(x)\n",
      "     |  >>> cs = CubicSpline(x, y)\n",
      "     |  >>> xs = np.arange(-0.5, 9.6, 0.1)\n",
      "     |  >>> fig, ax = plt.subplots(figsize=(6.5, 4))\n",
      "     |  >>> ax.plot(x, y, 'o', label='data')\n",
      "     |  >>> ax.plot(xs, np.sin(xs), label='true')\n",
      "     |  >>> ax.plot(xs, cs(xs), label=\"S\")\n",
      "     |  >>> ax.plot(xs, cs(xs, 1), label=\"S'\")\n",
      "     |  >>> ax.plot(xs, cs(xs, 2), label=\"S''\")\n",
      "     |  >>> ax.plot(xs, cs(xs, 3), label=\"S'''\")\n",
      "     |  >>> ax.set_xlim(-0.5, 9.5)\n",
      "     |  >>> ax.legend(loc='lower left', ncol=2)\n",
      "     |  >>> plt.show()\n",
      "     |  \n",
      "     |  In the second example, the unit circle is interpolated with a spline. A\n",
      "     |  periodic boundary condition is used. You can see that the first derivative\n",
      "     |  values, ds/dx=0, ds/dy=1 at the periodic point (1, 0) are correctly\n",
      "     |  computed. Note that a circle cannot be exactly represented by a cubic\n",
      "     |  spline. To increase precision, more breakpoints would be required.\n",
      "     |  \n",
      "     |  >>> theta = 2 * np.pi * np.linspace(0, 1, 5)\n",
      "     |  >>> y = np.c_[np.cos(theta), np.sin(theta)]\n",
      "     |  >>> cs = CubicSpline(theta, y, bc_type='periodic')\n",
      "     |  >>> print(\"ds/dx={:.1f} ds/dy={:.1f}\".format(cs(0, 1)[0], cs(0, 1)[1]))\n",
      "     |  ds/dx=0.0 ds/dy=1.0\n",
      "     |  >>> xs = 2 * np.pi * np.linspace(0, 1, 100)\n",
      "     |  >>> fig, ax = plt.subplots(figsize=(6.5, 4))\n",
      "     |  >>> ax.plot(y[:, 0], y[:, 1], 'o', label='data')\n",
      "     |  >>> ax.plot(np.cos(xs), np.sin(xs), label='true')\n",
      "     |  >>> ax.plot(cs(xs)[:, 0], cs(xs)[:, 1], label='spline')\n",
      "     |  >>> ax.axes.set_aspect('equal')\n",
      "     |  >>> ax.legend(loc='center')\n",
      "     |  >>> plt.show()\n",
      "     |  \n",
      "     |  The third example is the interpolation of a polynomial y = x**3 on the\n",
      "     |  interval 0 <= x<= 1. A cubic spline can represent this function exactly.\n",
      "     |  To achieve that we need to specify values and first derivatives at\n",
      "     |  endpoints of the interval. Note that y' = 3 * x**2 and thus y'(0) = 0 and\n",
      "     |  y'(1) = 3.\n",
      "     |  \n",
      "     |  >>> cs = CubicSpline([0, 1], [0, 1], bc_type=((1, 0), (1, 3)))\n",
      "     |  >>> x = np.linspace(0, 1)\n",
      "     |  >>> np.allclose(x**3, cs(x))\n",
      "     |  True\n",
      "     |  \n",
      "     |  References\n",
      "     |  ----------\n",
      "     |  .. [1] `Cubic Spline Interpolation\n",
      "     |          <https://en.wikiversity.org/wiki/Cubic_Spline_Interpolation>`_\n",
      "     |          on Wikiversity.\n",
      "     |  .. [2] Carl de Boor, \"A Practical Guide to Splines\", Springer-Verlag, 1978.\n",
      "     |  \n",
      "     |  Method resolution order:\n",
      "     |      CubicSpline\n",
      "     |      CubicHermiteSpline\n",
      "     |      scipy.interpolate.interpolate.PPoly\n",
      "     |      scipy.interpolate.interpolate._PPolyBase\n",
      "     |      builtins.object\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __init__(self, x, y, axis=0, bc_type='not-a-knot', extrapolate=None)\n",
      "     |      Initialize self.  See help(type(self)) for accurate signature.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from scipy.interpolate.interpolate.PPoly:\n",
      "     |  \n",
      "     |  antiderivative(self, nu=1)\n",
      "     |      Construct a new piecewise polynomial representing the antiderivative.\n",
      "     |      \n",
      "     |      Antiderivative is also the indefinite integral of the function,\n",
      "     |      and derivative is its inverse operation.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      nu : int, optional\n",
      "     |          Order of antiderivative to evaluate. Default is 1, i.e., compute\n",
      "     |          the first integral. If negative, the derivative is returned.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      pp : PPoly\n",
      "     |          Piecewise polynomial of order k2 = k + n representing\n",
      "     |          the antiderivative of this polynomial.\n",
      "     |      \n",
      "     |      Notes\n",
      "     |      -----\n",
      "     |      The antiderivative returned by this function is continuous and\n",
      "     |      continuously differentiable to order n-1, up to floating point\n",
      "     |      rounding error.\n",
      "     |      \n",
      "     |      If antiderivative is computed and ``self.extrapolate='periodic'``,\n",
      "     |      it will be set to False for the returned instance. This is done because\n",
      "     |      the antiderivative is no longer periodic and its correct evaluation\n",
      "     |      outside of the initially given x interval is difficult.\n",
      "     |  \n",
      "     |  derivative(self, nu=1)\n",
      "     |      Construct a new piecewise polynomial representing the derivative.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      nu : int, optional\n",
      "     |          Order of derivative to evaluate. Default is 1, i.e., compute the\n",
      "     |          first derivative. If negative, the antiderivative is returned.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      pp : PPoly\n",
      "     |          Piecewise polynomial of order k2 = k - n representing the derivative\n",
      "     |          of this polynomial.\n",
      "     |      \n",
      "     |      Notes\n",
      "     |      -----\n",
      "     |      Derivatives are evaluated piecewise for each polynomial\n",
      "     |      segment, even if the polynomial is not differentiable at the\n",
      "     |      breakpoints. The polynomial intervals are considered half-open,\n",
      "     |      ``[a, b)``, except for the last interval which is closed\n",
      "     |      ``[a, b]``.\n",
      "     |  \n",
      "     |  integrate(self, a, b, extrapolate=None)\n",
      "     |      Compute a definite integral over a piecewise polynomial.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      a : float\n",
      "     |          Lower integration bound\n",
      "     |      b : float\n",
      "     |          Upper integration bound\n",
      "     |      extrapolate : {bool, 'periodic', None}, optional\n",
      "     |          If bool, determines whether to extrapolate to out-of-bounds points\n",
      "     |          based on first and last intervals, or to return NaNs.\n",
      "     |          If 'periodic', periodic extrapolation is used.\n",
      "     |          If None (default), use `self.extrapolate`.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      ig : array_like\n",
      "     |          Definite integral of the piecewise polynomial over [a, b]\n",
      "     |  \n",
      "     |  roots(self, discontinuity=True, extrapolate=None)\n",
      "     |      Find real roots of the the piecewise polynomial.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      discontinuity : bool, optional\n",
      "     |          Whether to report sign changes across discontinuities at\n",
      "     |          breakpoints as roots.\n",
      "     |      extrapolate : {bool, 'periodic', None}, optional\n",
      "     |          If bool, determines whether to return roots from the polynomial\n",
      "     |          extrapolated based on first and last intervals, 'periodic' works\n",
      "     |          the same as False. If None (default), use `self.extrapolate`.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      roots : ndarray\n",
      "     |          Roots of the polynomial(s).\n",
      "     |      \n",
      "     |          If the PPoly object describes multiple polynomials, the\n",
      "     |          return value is an object array whose each element is an\n",
      "     |          ndarray containing the roots.\n",
      "     |      \n",
      "     |      See Also\n",
      "     |      --------\n",
      "     |      PPoly.solve\n",
      "     |  \n",
      "     |  solve(self, y=0.0, discontinuity=True, extrapolate=None)\n",
      "     |      Find real solutions of the the equation ``pp(x) == y``.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      y : float, optional\n",
      "     |          Right-hand side. Default is zero.\n",
      "     |      discontinuity : bool, optional\n",
      "     |          Whether to report sign changes across discontinuities at\n",
      "     |          breakpoints as roots.\n",
      "     |      extrapolate : {bool, 'periodic', None}, optional\n",
      "     |          If bool, determines whether to return roots from the polynomial\n",
      "     |          extrapolated based on first and last intervals, 'periodic' works\n",
      "     |          the same as False. If None (default), use `self.extrapolate`.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      roots : ndarray\n",
      "     |          Roots of the polynomial(s).\n",
      "     |      \n",
      "     |          If the PPoly object describes multiple polynomials, the\n",
      "     |          return value is an object array whose each element is an\n",
      "     |          ndarray containing the roots.\n",
      "     |      \n",
      "     |      Notes\n",
      "     |      -----\n",
      "     |      This routine works only on real-valued polynomials.\n",
      "     |      \n",
      "     |      If the piecewise polynomial contains sections that are\n",
      "     |      identically zero, the root list will contain the start point\n",
      "     |      of the corresponding interval, followed by a ``nan`` value.\n",
      "     |      \n",
      "     |      If the polynomial is discontinuous across a breakpoint, and\n",
      "     |      there is a sign change across the breakpoint, this is reported\n",
      "     |      if the `discont` parameter is True.\n",
      "     |      \n",
      "     |      Examples\n",
      "     |      --------\n",
      "     |      \n",
      "     |      Finding roots of ``[x**2 - 1, (x - 1)**2]`` defined on intervals\n",
      "     |      ``[-2, 1], [1, 2]``:\n",
      "     |      \n",
      "     |      >>> from scipy.interpolate import PPoly\n",
      "     |      >>> pp = PPoly(np.array([[1, -4, 3], [1, 0, 0]]).T, [-2, 1, 2])\n",
      "     |      >>> pp.solve()\n",
      "     |      array([-1.,  1.])\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Class methods inherited from scipy.interpolate.interpolate.PPoly:\n",
      "     |  \n",
      "     |  from_bernstein_basis(bp, extrapolate=None) from builtins.type\n",
      "     |      Construct a piecewise polynomial in the power basis\n",
      "     |      from a polynomial in Bernstein basis.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      bp : BPoly\n",
      "     |          A Bernstein basis polynomial, as created by BPoly\n",
      "     |      extrapolate : bool or 'periodic', optional\n",
      "     |          If bool, determines whether to extrapolate to out-of-bounds points\n",
      "     |          based on first and last intervals, or to return NaNs.\n",
      "     |          If 'periodic', periodic extrapolation is used. Default is True.\n",
      "     |  \n",
      "     |  from_spline(tck, extrapolate=None) from builtins.type\n",
      "     |      Construct a piecewise polynomial from a spline\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      tck\n",
      "     |          A spline, as returned by `splrep` or a BSpline object.\n",
      "     |      extrapolate : bool or 'periodic', optional\n",
      "     |          If bool, determines whether to extrapolate to out-of-bounds points\n",
      "     |          based on first and last intervals, or to return NaNs.\n",
      "     |          If 'periodic', periodic extrapolation is used. Default is True.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from scipy.interpolate.interpolate.PPoly:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |      dictionary for instance variables (if defined)\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from scipy.interpolate.interpolate._PPolyBase:\n",
      "     |  \n",
      "     |  __call__(self, x, nu=0, extrapolate=None)\n",
      "     |      Evaluate the piecewise polynomial or its derivative.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      x : array_like\n",
      "     |          Points to evaluate the interpolant at.\n",
      "     |      nu : int, optional\n",
      "     |          Order of derivative to evaluate. Must be non-negative.\n",
      "     |      extrapolate : {bool, 'periodic', None}, optional\n",
      "     |          If bool, determines whether to extrapolate to out-of-bounds points\n",
      "     |          based on first and last intervals, or to return NaNs.\n",
      "     |          If 'periodic', periodic extrapolation is used.\n",
      "     |          If None (default), use `self.extrapolate`.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      y : array_like\n",
      "     |          Interpolated values. Shape is determined by replacing\n",
      "     |          the interpolation axis in the original array with the shape of x.\n",
      "     |      \n",
      "     |      Notes\n",
      "     |      -----\n",
      "     |      Derivatives are evaluated piecewise for each polynomial\n",
      "     |      segment, even if the polynomial is not differentiable at the\n",
      "     |      breakpoints. The polynomial intervals are considered half-open,\n",
      "     |      ``[a, b)``, except for the last interval which is closed\n",
      "     |      ``[a, b]``.\n",
      "     |  \n",
      "     |  extend(self, c, x, right=None)\n",
      "     |      Add additional breakpoints and coefficients to the polynomial.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      c : ndarray, size (k, m, ...)\n",
      "     |          Additional coefficients for polynomials in intervals. Note that\n",
      "     |          the first additional interval will be formed using one of the\n",
      "     |          ``self.x`` end points.\n",
      "     |      x : ndarray, size (m,)\n",
      "     |          Additional breakpoints. Must be sorted in the same order as\n",
      "     |          ``self.x`` and either to the right or to the left of the current\n",
      "     |          breakpoints.\n",
      "     |      right\n",
      "     |          Deprecated argument. Has no effect.\n",
      "     |      \n",
      "     |          .. deprecated:: 0.19\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Class methods inherited from scipy.interpolate.interpolate._PPolyBase:\n",
      "     |  \n",
      "     |  construct_fast(c, x, extrapolate=None, axis=0) from builtins.type\n",
      "     |      Construct the piecewise polynomial without making checks.\n",
      "     |      \n",
      "     |      Takes the same parameters as the constructor. Input arguments\n",
      "     |      ``c`` and ``x`` must be arrays of the correct shape and type. The\n",
      "     |      ``c`` array can only be of dtypes float and complex, and ``x``\n",
      "     |      array must have dtype float.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from scipy.interpolate.interpolate._PPolyBase:\n",
      "     |  \n",
      "     |  axis\n",
      "     |  \n",
      "     |  c\n",
      "     |  \n",
      "     |  extrapolate\n",
      "     |  \n",
      "     |  x\n",
      "    \n",
      "    class InterpolatedUnivariateSpline(UnivariateSpline)\n",
      "     |  InterpolatedUnivariateSpline(x, y, w=None, bbox=[None, None], k=3, ext=0, check_finite=False)\n",
      "     |  \n",
      "     |  1-D interpolating spline for a given set of data points.\n",
      "     |  \n",
      "     |  Fits a spline y = spl(x) of degree `k` to the provided `x`, `y` data.\n",
      "     |  Spline function passes through all provided points. Equivalent to\n",
      "     |  `UnivariateSpline` with  s=0.\n",
      "     |  \n",
      "     |  Parameters\n",
      "     |  ----------\n",
      "     |  x : (N,) array_like\n",
      "     |      Input dimension of data points -- must be strictly increasing\n",
      "     |  y : (N,) array_like\n",
      "     |      input dimension of data points\n",
      "     |  w : (N,) array_like, optional\n",
      "     |      Weights for spline fitting.  Must be positive.  If None (default),\n",
      "     |      weights are all equal.\n",
      "     |  bbox : (2,) array_like, optional\n",
      "     |      2-sequence specifying the boundary of the approximation interval. If\n",
      "     |      None (default), ``bbox=[x[0], x[-1]]``.\n",
      "     |  k : int, optional\n",
      "     |      Degree of the smoothing spline.  Must be 1 <= `k` <= 5.\n",
      "     |  ext : int or str, optional\n",
      "     |      Controls the extrapolation mode for elements\n",
      "     |      not in the interval defined by the knot sequence.\n",
      "     |  \n",
      "     |      * if ext=0 or 'extrapolate', return the extrapolated value.\n",
      "     |      * if ext=1 or 'zeros', return 0\n",
      "     |      * if ext=2 or 'raise', raise a ValueError\n",
      "     |      * if ext=3 of 'const', return the boundary value.\n",
      "     |  \n",
      "     |      The default value is 0.\n",
      "     |  \n",
      "     |  check_finite : bool, optional\n",
      "     |      Whether to check that the input arrays contain only finite numbers.\n",
      "     |      Disabling may give a performance gain, but may result in problems\n",
      "     |      (crashes, non-termination or non-sensical results) if the inputs\n",
      "     |      do contain infinities or NaNs.\n",
      "     |      Default is False.\n",
      "     |  \n",
      "     |  See Also\n",
      "     |  --------\n",
      "     |  UnivariateSpline :\n",
      "     |      a smooth univariate spline to fit a given set of data points.\n",
      "     |  LSQUnivariateSpline :\n",
      "     |      a spline for which knots are user-selected\n",
      "     |  SmoothBivariateSpline :\n",
      "     |      a smoothing bivariate spline through the given points\n",
      "     |  LSQBivariateSpline :\n",
      "     |      a bivariate spline using weighted least-squares fitting\n",
      "     |  splrep :\n",
      "     |      a function to find the B-spline representation of a 1-D curve\n",
      "     |  splev :\n",
      "     |      a function to evaluate a B-spline or its derivatives\n",
      "     |  sproot :\n",
      "     |      a function to find the roots of a cubic B-spline\n",
      "     |  splint :\n",
      "     |      a function to evaluate the definite integral of a B-spline between two\n",
      "     |      given points\n",
      "     |  spalde :\n",
      "     |      a function to evaluate all derivatives of a B-spline\n",
      "     |  \n",
      "     |  Notes\n",
      "     |  -----\n",
      "     |  The number of data points must be larger than the spline degree `k`.\n",
      "     |  \n",
      "     |  Examples\n",
      "     |  --------\n",
      "     |  >>> import matplotlib.pyplot as plt\n",
      "     |  >>> from scipy.interpolate import InterpolatedUnivariateSpline\n",
      "     |  >>> rng = np.random.default_rng()\n",
      "     |  >>> x = np.linspace(-3, 3, 50)\n",
      "     |  >>> y = np.exp(-x**2) + 0.1 * rng.standard_normal(50)\n",
      "     |  >>> spl = InterpolatedUnivariateSpline(x, y)\n",
      "     |  >>> plt.plot(x, y, 'ro', ms=5)\n",
      "     |  >>> xs = np.linspace(-3, 3, 1000)\n",
      "     |  >>> plt.plot(xs, spl(xs), 'g', lw=3, alpha=0.7)\n",
      "     |  >>> plt.show()\n",
      "     |  \n",
      "     |  Notice that the ``spl(x)`` interpolates `y`:\n",
      "     |  \n",
      "     |  >>> spl.get_residual()\n",
      "     |  0.0\n",
      "     |  \n",
      "     |  Method resolution order:\n",
      "     |      InterpolatedUnivariateSpline\n",
      "     |      UnivariateSpline\n",
      "     |      builtins.object\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __init__(self, x, y, w=None, bbox=[None, None], k=3, ext=0, check_finite=False)\n",
      "     |      Initialize self.  See help(type(self)) for accurate signature.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from UnivariateSpline:\n",
      "     |  \n",
      "     |  __call__(self, x, nu=0, ext=None)\n",
      "     |      Evaluate spline (or its nu-th derivative) at positions x.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      x : array_like\n",
      "     |          A 1-D array of points at which to return the value of the smoothed\n",
      "     |          spline or its derivatives. Note: `x` can be unordered but the\n",
      "     |          evaluation is more efficient if `x` is (partially) ordered.\n",
      "     |      nu  : int\n",
      "     |          The order of derivative of the spline to compute.\n",
      "     |      ext : int\n",
      "     |          Controls the value returned for elements of `x` not in the\n",
      "     |          interval defined by the knot sequence.\n",
      "     |      \n",
      "     |          * if ext=0 or 'extrapolate', return the extrapolated value.\n",
      "     |          * if ext=1 or 'zeros', return 0\n",
      "     |          * if ext=2 or 'raise', raise a ValueError\n",
      "     |          * if ext=3 or 'const', return the boundary value.\n",
      "     |      \n",
      "     |          The default value is 0, passed from the initialization of\n",
      "     |          UnivariateSpline.\n",
      "     |  \n",
      "     |  antiderivative(self, n=1)\n",
      "     |      Construct a new spline representing the antiderivative of this spline.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      n : int, optional\n",
      "     |          Order of antiderivative to evaluate. Default: 1\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      spline : UnivariateSpline\n",
      "     |          Spline of order k2=k+n representing the antiderivative of this\n",
      "     |          spline.\n",
      "     |      \n",
      "     |      Notes\n",
      "     |      -----\n",
      "     |      \n",
      "     |      .. versionadded:: 0.13.0\n",
      "     |      \n",
      "     |      See Also\n",
      "     |      --------\n",
      "     |      splantider, derivative\n",
      "     |      \n",
      "     |      Examples\n",
      "     |      --------\n",
      "     |      >>> from scipy.interpolate import UnivariateSpline\n",
      "     |      >>> x = np.linspace(0, np.pi/2, 70)\n",
      "     |      >>> y = 1 / np.sqrt(1 - 0.8*np.sin(x)**2)\n",
      "     |      >>> spl = UnivariateSpline(x, y, s=0)\n",
      "     |      \n",
      "     |      The derivative is the inverse operation of the antiderivative,\n",
      "     |      although some floating point error accumulates:\n",
      "     |      \n",
      "     |      >>> spl(1.7), spl.antiderivative().derivative()(1.7)\n",
      "     |      (array(2.1565429877197317), array(2.1565429877201865))\n",
      "     |      \n",
      "     |      Antiderivative can be used to evaluate definite integrals:\n",
      "     |      \n",
      "     |      >>> ispl = spl.antiderivative()\n",
      "     |      >>> ispl(np.pi/2) - ispl(0)\n",
      "     |      2.2572053588768486\n",
      "     |      \n",
      "     |      This is indeed an approximation to the complete elliptic integral\n",
      "     |      :math:`K(m) = \\int_0^{\\pi/2} [1 - m\\sin^2 x]^{-1/2} dx`:\n",
      "     |      \n",
      "     |      >>> from scipy.special import ellipk\n",
      "     |      >>> ellipk(0.8)\n",
      "     |      2.2572053268208538\n",
      "     |  \n",
      "     |  derivative(self, n=1)\n",
      "     |      Construct a new spline representing the derivative of this spline.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      n : int, optional\n",
      "     |          Order of derivative to evaluate. Default: 1\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      spline : UnivariateSpline\n",
      "     |          Spline of order k2=k-n representing the derivative of this\n",
      "     |          spline.\n",
      "     |      \n",
      "     |      See Also\n",
      "     |      --------\n",
      "     |      splder, antiderivative\n",
      "     |      \n",
      "     |      Notes\n",
      "     |      -----\n",
      "     |      \n",
      "     |      .. versionadded:: 0.13.0\n",
      "     |      \n",
      "     |      Examples\n",
      "     |      --------\n",
      "     |      This can be used for finding maxima of a curve:\n",
      "     |      \n",
      "     |      >>> from scipy.interpolate import UnivariateSpline\n",
      "     |      >>> x = np.linspace(0, 10, 70)\n",
      "     |      >>> y = np.sin(x)\n",
      "     |      >>> spl = UnivariateSpline(x, y, k=4, s=0)\n",
      "     |      \n",
      "     |      Now, differentiate the spline and find the zeros of the\n",
      "     |      derivative. (NB: `sproot` only works for order 3 splines, so we\n",
      "     |      fit an order 4 spline):\n",
      "     |      \n",
      "     |      >>> spl.derivative().roots() / np.pi\n",
      "     |      array([ 0.50000001,  1.5       ,  2.49999998])\n",
      "     |      \n",
      "     |      This agrees well with roots :math:`\\pi/2 + n\\pi` of\n",
      "     |      :math:`\\cos(x) = \\sin'(x)`.\n",
      "     |  \n",
      "     |  derivatives(self, x)\n",
      "     |      Return all derivatives of the spline at the point x.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      x : float\n",
      "     |          The point to evaluate the derivatives at.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      der : ndarray, shape(k+1,)\n",
      "     |          Derivatives of the orders 0 to k.\n",
      "     |      \n",
      "     |      Examples\n",
      "     |      --------\n",
      "     |      >>> from scipy.interpolate import UnivariateSpline\n",
      "     |      >>> x = np.linspace(0, 3, 11)\n",
      "     |      >>> y = x**2\n",
      "     |      >>> spl = UnivariateSpline(x, y)\n",
      "     |      >>> spl.derivatives(1.5)\n",
      "     |      array([2.25, 3.0, 2.0, 0])\n",
      "     |  \n",
      "     |  get_coeffs(self)\n",
      "     |      Return spline coefficients.\n",
      "     |  \n",
      "     |  get_knots(self)\n",
      "     |      Return positions of interior knots of the spline.\n",
      "     |      \n",
      "     |      Internally, the knot vector contains ``2*k`` additional boundary knots.\n",
      "     |  \n",
      "     |  get_residual(self)\n",
      "     |      Return weighted sum of squared residuals of the spline approximation.\n",
      "     |      \n",
      "     |      This is equivalent to::\n",
      "     |      \n",
      "     |           sum((w[i] * (y[i]-spl(x[i])))**2, axis=0)\n",
      "     |  \n",
      "     |  integral(self, a, b)\n",
      "     |      Return definite integral of the spline between two given points.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      a : float\n",
      "     |          Lower limit of integration.\n",
      "     |      b : float\n",
      "     |          Upper limit of integration.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      integral : float\n",
      "     |          The value of the definite integral of the spline between limits.\n",
      "     |      \n",
      "     |      Examples\n",
      "     |      --------\n",
      "     |      >>> from scipy.interpolate import UnivariateSpline\n",
      "     |      >>> x = np.linspace(0, 3, 11)\n",
      "     |      >>> y = x**2\n",
      "     |      >>> spl = UnivariateSpline(x, y)\n",
      "     |      >>> spl.integral(0, 3)\n",
      "     |      9.0\n",
      "     |      \n",
      "     |      which agrees with :math:`\\int x^2 dx = x^3 / 3` between the limits\n",
      "     |      of 0 and 3.\n",
      "     |      \n",
      "     |      A caveat is that this routine assumes the spline to be zero outside of\n",
      "     |      the data limits:\n",
      "     |      \n",
      "     |      >>> spl.integral(-1, 4)\n",
      "     |      9.0\n",
      "     |      >>> spl.integral(-1, 0)\n",
      "     |      0.0\n",
      "     |  \n",
      "     |  roots(self)\n",
      "     |      Return the zeros of the spline.\n",
      "     |      \n",
      "     |      Restriction: only cubic splines are supported by fitpack.\n",
      "     |  \n",
      "     |  set_smoothing_factor(self, s)\n",
      "     |      Continue spline computation with the given smoothing\n",
      "     |      factor s and with the knots found at the last call.\n",
      "     |      \n",
      "     |      This routine modifies the spline in place.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Static methods inherited from UnivariateSpline:\n",
      "     |  \n",
      "     |  validate_input(x, y, w, bbox, k, s, ext, check_finite)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from UnivariateSpline:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |      dictionary for instance variables (if defined)\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "    \n",
      "    class KroghInterpolator(_Interpolator1DWithDerivatives)\n",
      "     |  KroghInterpolator(xi, yi, axis=0)\n",
      "     |  \n",
      "     |  Interpolating polynomial for a set of points.\n",
      "     |  \n",
      "     |  The polynomial passes through all the pairs (xi,yi). One may\n",
      "     |  additionally specify a number of derivatives at each point xi;\n",
      "     |  this is done by repeating the value xi and specifying the\n",
      "     |  derivatives as successive yi values.\n",
      "     |  \n",
      "     |  Allows evaluation of the polynomial and all its derivatives.\n",
      "     |  For reasons of numerical stability, this function does not compute\n",
      "     |  the coefficients of the polynomial, although they can be obtained\n",
      "     |  by evaluating all the derivatives.\n",
      "     |  \n",
      "     |  Parameters\n",
      "     |  ----------\n",
      "     |  xi : array_like, length N\n",
      "     |      Known x-coordinates. Must be sorted in increasing order.\n",
      "     |  yi : array_like\n",
      "     |      Known y-coordinates. When an xi occurs two or more times in\n",
      "     |      a row, the corresponding yi's represent derivative values.\n",
      "     |  axis : int, optional\n",
      "     |      Axis in the yi array corresponding to the x-coordinate values.\n",
      "     |  \n",
      "     |  Notes\n",
      "     |  -----\n",
      "     |  Be aware that the algorithms implemented here are not necessarily\n",
      "     |  the most numerically stable known. Moreover, even in a world of\n",
      "     |  exact computation, unless the x coordinates are chosen very\n",
      "     |  carefully - Chebyshev zeros (e.g., cos(i*pi/n)) are a good choice -\n",
      "     |  polynomial interpolation itself is a very ill-conditioned process\n",
      "     |  due to the Runge phenomenon. In general, even with well-chosen\n",
      "     |  x values, degrees higher than about thirty cause problems with\n",
      "     |  numerical instability in this code.\n",
      "     |  \n",
      "     |  Based on [1]_.\n",
      "     |  \n",
      "     |  References\n",
      "     |  ----------\n",
      "     |  .. [1] Krogh, \"Efficient Algorithms for Polynomial Interpolation\n",
      "     |      and Numerical Differentiation\", 1970.\n",
      "     |  \n",
      "     |  Examples\n",
      "     |  --------\n",
      "     |  To produce a polynomial that is zero at 0 and 1 and has\n",
      "     |  derivative 2 at 0, call\n",
      "     |  \n",
      "     |  >>> from scipy.interpolate import KroghInterpolator\n",
      "     |  >>> KroghInterpolator([0,0,1],[0,2,0])\n",
      "     |  \n",
      "     |  This constructs the quadratic 2*X**2-2*X. The derivative condition\n",
      "     |  is indicated by the repeated zero in the xi array; the corresponding\n",
      "     |  yi values are 0, the function value, and 2, the derivative value.\n",
      "     |  \n",
      "     |  For another example, given xi, yi, and a derivative ypi for each\n",
      "     |  point, appropriate arrays can be constructed as:\n",
      "     |  \n",
      "     |  >>> rng = np.random.default_rng()\n",
      "     |  >>> xi = np.linspace(0, 1, 5)\n",
      "     |  >>> yi, ypi = rng.random((2, 5))\n",
      "     |  >>> xi_k, yi_k = np.repeat(xi, 2), np.ravel(np.dstack((yi,ypi)))\n",
      "     |  >>> KroghInterpolator(xi_k, yi_k)\n",
      "     |  \n",
      "     |  To produce a vector-valued polynomial, supply a higher-dimensional\n",
      "     |  array for yi:\n",
      "     |  \n",
      "     |  >>> KroghInterpolator([0,1],[[2,3],[4,5]])\n",
      "     |  \n",
      "     |  This constructs a linear polynomial giving (2,3) at 0 and (4,5) at 1.\n",
      "     |  \n",
      "     |  Method resolution order:\n",
      "     |      KroghInterpolator\n",
      "     |      _Interpolator1DWithDerivatives\n",
      "     |      _Interpolator1D\n",
      "     |      builtins.object\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __init__(self, xi, yi, axis=0)\n",
      "     |      Initialize self.  See help(type(self)) for accurate signature.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from _Interpolator1DWithDerivatives:\n",
      "     |  \n",
      "     |  derivative(self, x, der=1)\n",
      "     |      Evaluate one derivative of the polynomial at the point x\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      x : array_like\n",
      "     |          Point or points at which to evaluate the derivatives\n",
      "     |      \n",
      "     |      der : integer, optional\n",
      "     |          Which derivative to extract. This number includes the\n",
      "     |          function value as 0th derivative.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      d : ndarray\n",
      "     |          Derivative interpolated at the x-points. Shape of d is\n",
      "     |          determined by replacing the interpolation axis in the\n",
      "     |          original array with the shape of x.\n",
      "     |      \n",
      "     |      Notes\n",
      "     |      -----\n",
      "     |      This is computed by evaluating all derivatives up to the desired\n",
      "     |      one (using self.derivatives()) and then discarding the rest.\n",
      "     |  \n",
      "     |  derivatives(self, x, der=None)\n",
      "     |      Evaluate many derivatives of the polynomial at the point x\n",
      "     |      \n",
      "     |      Produce an array of all derivative values at the point x.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      x : array_like\n",
      "     |          Point or points at which to evaluate the derivatives\n",
      "     |      der : int or None, optional\n",
      "     |          How many derivatives to extract; None for all potentially\n",
      "     |          nonzero derivatives (that is a number equal to the number\n",
      "     |          of points). This number includes the function value as 0th\n",
      "     |          derivative.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      d : ndarray\n",
      "     |          Array with derivatives; d[j] contains the jth derivative.\n",
      "     |          Shape of d[j] is determined by replacing the interpolation\n",
      "     |          axis in the original array with the shape of x.\n",
      "     |      \n",
      "     |      Examples\n",
      "     |      --------\n",
      "     |      >>> from scipy.interpolate import KroghInterpolator\n",
      "     |      >>> KroghInterpolator([0,0,0],[1,2,3]).derivatives(0)\n",
      "     |      array([1.0,2.0,3.0])\n",
      "     |      >>> KroghInterpolator([0,0,0],[1,2,3]).derivatives([0,0])\n",
      "     |      array([[1.0,1.0],\n",
      "     |             [2.0,2.0],\n",
      "     |             [3.0,3.0]])\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from _Interpolator1DWithDerivatives:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |      dictionary for instance variables (if defined)\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from _Interpolator1D:\n",
      "     |  \n",
      "     |  __call__(self, x)\n",
      "     |      Evaluate the interpolant\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      x : array_like\n",
      "     |          Points to evaluate the interpolant at.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      y : array_like\n",
      "     |          Interpolated values. Shape is determined by replacing\n",
      "     |          the interpolation axis in the original array with the shape of x.\n",
      "     |      \n",
      "     |      Notes\n",
      "     |      -----\n",
      "     |      Input values `x` must be convertible to `float` values like `int` \n",
      "     |      or `float`.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from _Interpolator1D:\n",
      "     |  \n",
      "     |  dtype\n",
      "    \n",
      "    class LSQBivariateSpline(BivariateSpline)\n",
      "     |  LSQBivariateSpline(x, y, z, tx, ty, w=None, bbox=[None, None, None, None], kx=3, ky=3, eps=None)\n",
      "     |  \n",
      "     |  Weighted least-squares bivariate spline approximation.\n",
      "     |  \n",
      "     |  Parameters\n",
      "     |  ----------\n",
      "     |  x, y, z : array_like\n",
      "     |      1-D sequences of data points (order is not important).\n",
      "     |  tx, ty : array_like\n",
      "     |      Strictly ordered 1-D sequences of knots coordinates.\n",
      "     |  w : array_like, optional\n",
      "     |      Positive 1-D array of weights, of the same length as `x`, `y` and `z`.\n",
      "     |  bbox : (4,) array_like, optional\n",
      "     |      Sequence of length 4 specifying the boundary of the rectangular\n",
      "     |      approximation domain.  By default,\n",
      "     |      ``bbox=[min(x,tx),max(x,tx), min(y,ty),max(y,ty)]``.\n",
      "     |  kx, ky : ints, optional\n",
      "     |      Degrees of the bivariate spline. Default is 3.\n",
      "     |  eps : float, optional\n",
      "     |      A threshold for determining the effective rank of an over-determined\n",
      "     |      linear system of equations. `eps` should have a value within the open\n",
      "     |      interval ``(0, 1)``, the default is 1e-16.\n",
      "     |  \n",
      "     |  See Also\n",
      "     |  --------\n",
      "     |  BivariateSpline :\n",
      "     |      a base class for bivariate splines.\n",
      "     |  UnivariateSpline :\n",
      "     |      a smooth univariate spline to fit a given set of data points.\n",
      "     |  SmoothBivariateSpline :\n",
      "     |      a smoothing bivariate spline through the given points\n",
      "     |  RectSphereBivariateSpline :\n",
      "     |      a bivariate spline over a rectangular mesh on a sphere\n",
      "     |  SmoothSphereBivariateSpline :\n",
      "     |      a smoothing bivariate spline in spherical coordinates\n",
      "     |  LSQSphereBivariateSpline :\n",
      "     |      a bivariate spline in spherical coordinates using weighted\n",
      "     |      least-squares fitting\n",
      "     |  RectBivariateSpline :\n",
      "     |      a bivariate spline over a rectangular mesh.\n",
      "     |  bisplrep :\n",
      "     |      a function to find a bivariate B-spline representation of a surface\n",
      "     |  bisplev :\n",
      "     |      a function to evaluate a bivariate B-spline and its derivatives\n",
      "     |  \n",
      "     |  Notes\n",
      "     |  -----\n",
      "     |  The length of `x`, `y` and `z` should be at least ``(kx+1) * (ky+1)``.\n",
      "     |  \n",
      "     |  Method resolution order:\n",
      "     |      LSQBivariateSpline\n",
      "     |      BivariateSpline\n",
      "     |      _BivariateSplineBase\n",
      "     |      builtins.object\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __init__(self, x, y, z, tx, ty, w=None, bbox=[None, None, None, None], kx=3, ky=3, eps=None)\n",
      "     |      Initialize self.  See help(type(self)) for accurate signature.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from BivariateSpline:\n",
      "     |  \n",
      "     |  ev(self, xi, yi, dx=0, dy=0)\n",
      "     |      Evaluate the spline at points\n",
      "     |      \n",
      "     |      Returns the interpolated value at ``(xi[i], yi[i]),\n",
      "     |      i=0,...,len(xi)-1``.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      xi, yi : array_like\n",
      "     |          Input coordinates. Standard Numpy broadcasting is obeyed.\n",
      "     |      dx : int, optional\n",
      "     |          Order of x-derivative\n",
      "     |      \n",
      "     |          .. versionadded:: 0.14.0\n",
      "     |      dy : int, optional\n",
      "     |          Order of y-derivative\n",
      "     |      \n",
      "     |          .. versionadded:: 0.14.0\n",
      "     |  \n",
      "     |  integral(self, xa, xb, ya, yb)\n",
      "     |      Evaluate the integral of the spline over area [xa,xb] x [ya,yb].\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      xa, xb : float\n",
      "     |          The end-points of the x integration interval.\n",
      "     |      ya, yb : float\n",
      "     |          The end-points of the y integration interval.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      integ : float\n",
      "     |          The value of the resulting integral.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from _BivariateSplineBase:\n",
      "     |  \n",
      "     |  __call__(self, x, y, dx=0, dy=0, grid=True)\n",
      "     |      Evaluate the spline or its derivatives at given positions.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      x, y : array_like\n",
      "     |          Input coordinates.\n",
      "     |      \n",
      "     |          If `grid` is False, evaluate the spline at points ``(x[i],\n",
      "     |          y[i]), i=0, ..., len(x)-1``.  Standard Numpy broadcasting\n",
      "     |          is obeyed.\n",
      "     |      \n",
      "     |          If `grid` is True: evaluate spline at the grid points\n",
      "     |          defined by the coordinate arrays x, y. The arrays must be\n",
      "     |          sorted to increasing order.\n",
      "     |      \n",
      "     |          Note that the axis ordering is inverted relative to\n",
      "     |          the output of meshgrid.\n",
      "     |      dx : int\n",
      "     |          Order of x-derivative\n",
      "     |      \n",
      "     |          .. versionadded:: 0.14.0\n",
      "     |      dy : int\n",
      "     |          Order of y-derivative\n",
      "     |      \n",
      "     |          .. versionadded:: 0.14.0\n",
      "     |      grid : bool\n",
      "     |          Whether to evaluate the results on a grid spanned by the\n",
      "     |          input arrays, or at points specified by the input arrays.\n",
      "     |      \n",
      "     |          .. versionadded:: 0.14.0\n",
      "     |  \n",
      "     |  get_coeffs(self)\n",
      "     |      Return spline coefficients.\n",
      "     |  \n",
      "     |  get_knots(self)\n",
      "     |      Return a tuple (tx,ty) where tx,ty contain knots positions\n",
      "     |      of the spline with respect to x-, y-variable, respectively.\n",
      "     |      The position of interior and additional knots are given as\n",
      "     |      t[k+1:-k-1] and t[:k+1]=b, t[-k-1:]=e, respectively.\n",
      "     |  \n",
      "     |  get_residual(self)\n",
      "     |      Return weighted sum of squared residuals of the spline\n",
      "     |      approximation: sum ((w[i]*(z[i]-s(x[i],y[i])))**2,axis=0)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from _BivariateSplineBase:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |      dictionary for instance variables (if defined)\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "    \n",
      "    class LSQSphereBivariateSpline(SphereBivariateSpline)\n",
      "     |  LSQSphereBivariateSpline(theta, phi, r, tt, tp, w=None, eps=1e-16)\n",
      "     |  \n",
      "     |  Weighted least-squares bivariate spline approximation in spherical\n",
      "     |  coordinates.\n",
      "     |  \n",
      "     |  Determines a smoothing bicubic spline according to a given\n",
      "     |  set of knots in the `theta` and `phi` directions.\n",
      "     |  \n",
      "     |  .. versionadded:: 0.11.0\n",
      "     |  \n",
      "     |  Parameters\n",
      "     |  ----------\n",
      "     |  theta, phi, r : array_like\n",
      "     |      1-D sequences of data points (order is not important). Coordinates\n",
      "     |      must be given in radians. Theta must lie within the interval\n",
      "     |      ``[0, pi]``, and phi must lie within the interval ``[0, 2pi]``.\n",
      "     |  tt, tp : array_like\n",
      "     |      Strictly ordered 1-D sequences of knots coordinates.\n",
      "     |      Coordinates must satisfy ``0 < tt[i] < pi``, ``0 < tp[i] < 2*pi``.\n",
      "     |  w : array_like, optional\n",
      "     |      Positive 1-D sequence of weights, of the same length as `theta`, `phi`\n",
      "     |      and `r`.\n",
      "     |  eps : float, optional\n",
      "     |      A threshold for determining the effective rank of an over-determined\n",
      "     |      linear system of equations. `eps` should have a value within the\n",
      "     |      open interval ``(0, 1)``, the default is 1e-16.\n",
      "     |  \n",
      "     |  See Also\n",
      "     |  --------\n",
      "     |  BivariateSpline :\n",
      "     |      a base class for bivariate splines.\n",
      "     |  UnivariateSpline :\n",
      "     |      a smooth univariate spline to fit a given set of data points.\n",
      "     |  SmoothBivariateSpline :\n",
      "     |      a smoothing bivariate spline through the given points\n",
      "     |  LSQBivariateSpline :\n",
      "     |      a bivariate spline using weighted least-squares fitting\n",
      "     |  RectSphereBivariateSpline :\n",
      "     |      a bivariate spline over a rectangular mesh on a sphere\n",
      "     |  SmoothSphereBivariateSpline :\n",
      "     |      a smoothing bivariate spline in spherical coordinates\n",
      "     |  RectBivariateSpline :\n",
      "     |      a bivariate spline over a rectangular mesh.\n",
      "     |  bisplrep :\n",
      "     |      a function to find a bivariate B-spline representation of a surface\n",
      "     |  bisplev :\n",
      "     |      a function to evaluate a bivariate B-spline and its derivatives\n",
      "     |  \n",
      "     |  Notes\n",
      "     |  -----\n",
      "     |  For more information, see the FITPACK_ site about this function.\n",
      "     |  \n",
      "     |  .. _FITPACK: http://www.netlib.org/dierckx/sphere.f\n",
      "     |  \n",
      "     |  Examples\n",
      "     |  --------\n",
      "     |  Suppose we have global data on a coarse grid (the input data does not\n",
      "     |  have to be on a grid):\n",
      "     |  \n",
      "     |  >>> from scipy.interpolate import LSQSphereBivariateSpline\n",
      "     |  >>> import matplotlib.pyplot as plt\n",
      "     |  \n",
      "     |  >>> theta = np.linspace(0, np.pi, num=7)\n",
      "     |  >>> phi = np.linspace(0, 2*np.pi, num=9)\n",
      "     |  >>> data = np.empty((theta.shape[0], phi.shape[0]))\n",
      "     |  >>> data[:,0], data[0,:], data[-1,:] = 0., 0., 0.\n",
      "     |  >>> data[1:-1,1], data[1:-1,-1] = 1., 1.\n",
      "     |  >>> data[1,1:-1], data[-2,1:-1] = 1., 1.\n",
      "     |  >>> data[2:-2,2], data[2:-2,-2] = 2., 2.\n",
      "     |  >>> data[2,2:-2], data[-3,2:-2] = 2., 2.\n",
      "     |  >>> data[3,3:-2] = 3.\n",
      "     |  >>> data = np.roll(data, 4, 1)\n",
      "     |  \n",
      "     |  We need to set up the interpolator object. Here, we must also specify the\n",
      "     |  coordinates of the knots to use.\n",
      "     |  \n",
      "     |  >>> lats, lons = np.meshgrid(theta, phi)\n",
      "     |  >>> knotst, knotsp = theta.copy(), phi.copy()\n",
      "     |  >>> knotst[0] += .0001\n",
      "     |  >>> knotst[-1] -= .0001\n",
      "     |  >>> knotsp[0] += .0001\n",
      "     |  >>> knotsp[-1] -= .0001\n",
      "     |  >>> lut = LSQSphereBivariateSpline(lats.ravel(), lons.ravel(),\n",
      "     |  ...                                data.T.ravel(), knotst, knotsp)\n",
      "     |  \n",
      "     |  As a first test, we'll see what the algorithm returns when run on the\n",
      "     |  input coordinates\n",
      "     |  \n",
      "     |  >>> data_orig = lut(theta, phi)\n",
      "     |  \n",
      "     |  Finally we interpolate the data to a finer grid\n",
      "     |  \n",
      "     |  >>> fine_lats = np.linspace(0., np.pi, 70)\n",
      "     |  >>> fine_lons = np.linspace(0., 2*np.pi, 90)\n",
      "     |  >>> data_lsq = lut(fine_lats, fine_lons)\n",
      "     |  \n",
      "     |  >>> fig = plt.figure()\n",
      "     |  >>> ax1 = fig.add_subplot(131)\n",
      "     |  >>> ax1.imshow(data, interpolation='nearest')\n",
      "     |  >>> ax2 = fig.add_subplot(132)\n",
      "     |  >>> ax2.imshow(data_orig, interpolation='nearest')\n",
      "     |  >>> ax3 = fig.add_subplot(133)\n",
      "     |  >>> ax3.imshow(data_lsq, interpolation='nearest')\n",
      "     |  >>> plt.show()\n",
      "     |  \n",
      "     |  Method resolution order:\n",
      "     |      LSQSphereBivariateSpline\n",
      "     |      SphereBivariateSpline\n",
      "     |      _BivariateSplineBase\n",
      "     |      builtins.object\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __call__(self, theta, phi, dtheta=0, dphi=0, grid=True)\n",
      "     |      Evaluate the spline or its derivatives at given positions.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      theta, phi : array_like\n",
      "     |          Input coordinates.\n",
      "     |      \n",
      "     |          If `grid` is False, evaluate the spline at points\n",
      "     |          ``(theta[i], phi[i]), i=0, ..., len(x)-1``.  Standard\n",
      "     |          Numpy broadcasting is obeyed.\n",
      "     |      \n",
      "     |          If `grid` is True: evaluate spline at the grid points\n",
      "     |          defined by the coordinate arrays theta, phi. The arrays\n",
      "     |          must be sorted to increasing order.\n",
      "     |      dtheta : int, optional\n",
      "     |          Order of theta-derivative\n",
      "     |      \n",
      "     |          .. versionadded:: 0.14.0\n",
      "     |      dphi : int\n",
      "     |          Order of phi-derivative\n",
      "     |      \n",
      "     |          .. versionadded:: 0.14.0\n",
      "     |      grid : bool\n",
      "     |          Whether to evaluate the results on a grid spanned by the\n",
      "     |          input arrays, or at points specified by the input arrays.\n",
      "     |      \n",
      "     |          .. versionadded:: 0.14.0\n",
      "     |  \n",
      "     |  __init__(self, theta, phi, r, tt, tp, w=None, eps=1e-16)\n",
      "     |      Initialize self.  See help(type(self)) for accurate signature.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from SphereBivariateSpline:\n",
      "     |  \n",
      "     |  ev(self, theta, phi, dtheta=0, dphi=0)\n",
      "     |      Evaluate the spline at points\n",
      "     |      \n",
      "     |      Returns the interpolated value at ``(theta[i], phi[i]),\n",
      "     |      i=0,...,len(theta)-1``.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      theta, phi : array_like\n",
      "     |          Input coordinates. Standard Numpy broadcasting is obeyed.\n",
      "     |      dtheta : int, optional\n",
      "     |          Order of theta-derivative\n",
      "     |      \n",
      "     |          .. versionadded:: 0.14.0\n",
      "     |      dphi : int, optional\n",
      "     |          Order of phi-derivative\n",
      "     |      \n",
      "     |          .. versionadded:: 0.14.0\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from _BivariateSplineBase:\n",
      "     |  \n",
      "     |  get_coeffs(self)\n",
      "     |      Return spline coefficients.\n",
      "     |  \n",
      "     |  get_knots(self)\n",
      "     |      Return a tuple (tx,ty) where tx,ty contain knots positions\n",
      "     |      of the spline with respect to x-, y-variable, respectively.\n",
      "     |      The position of interior and additional knots are given as\n",
      "     |      t[k+1:-k-1] and t[:k+1]=b, t[-k-1:]=e, respectively.\n",
      "     |  \n",
      "     |  get_residual(self)\n",
      "     |      Return weighted sum of squared residuals of the spline\n",
      "     |      approximation: sum ((w[i]*(z[i]-s(x[i],y[i])))**2,axis=0)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from _BivariateSplineBase:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |      dictionary for instance variables (if defined)\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "    \n",
      "    class LSQUnivariateSpline(UnivariateSpline)\n",
      "     |  LSQUnivariateSpline(x, y, t, w=None, bbox=[None, None], k=3, ext=0, check_finite=False)\n",
      "     |  \n",
      "     |  1-D spline with explicit internal knots.\n",
      "     |  \n",
      "     |  Fits a spline y = spl(x) of degree `k` to the provided `x`, `y` data.  `t`\n",
      "     |  specifies the internal knots of the spline\n",
      "     |  \n",
      "     |  Parameters\n",
      "     |  ----------\n",
      "     |  x : (N,) array_like\n",
      "     |      Input dimension of data points -- must be increasing\n",
      "     |  y : (N,) array_like\n",
      "     |      Input dimension of data points\n",
      "     |  t : (M,) array_like\n",
      "     |      interior knots of the spline.  Must be in ascending order and::\n",
      "     |  \n",
      "     |          bbox[0] < t[0] < ... < t[-1] < bbox[-1]\n",
      "     |  \n",
      "     |  w : (N,) array_like, optional\n",
      "     |      weights for spline fitting. Must be positive. If None (default),\n",
      "     |      weights are all equal.\n",
      "     |  bbox : (2,) array_like, optional\n",
      "     |      2-sequence specifying the boundary of the approximation interval. If\n",
      "     |      None (default), ``bbox = [x[0], x[-1]]``.\n",
      "     |  k : int, optional\n",
      "     |      Degree of the smoothing spline.  Must be 1 <= `k` <= 5.\n",
      "     |      Default is `k` = 3, a cubic spline.\n",
      "     |  ext : int or str, optional\n",
      "     |      Controls the extrapolation mode for elements\n",
      "     |      not in the interval defined by the knot sequence.\n",
      "     |  \n",
      "     |      * if ext=0 or 'extrapolate', return the extrapolated value.\n",
      "     |      * if ext=1 or 'zeros', return 0\n",
      "     |      * if ext=2 or 'raise', raise a ValueError\n",
      "     |      * if ext=3 of 'const', return the boundary value.\n",
      "     |  \n",
      "     |      The default value is 0.\n",
      "     |  \n",
      "     |  check_finite : bool, optional\n",
      "     |      Whether to check that the input arrays contain only finite numbers.\n",
      "     |      Disabling may give a performance gain, but may result in problems\n",
      "     |      (crashes, non-termination or non-sensical results) if the inputs\n",
      "     |      do contain infinities or NaNs.\n",
      "     |      Default is False.\n",
      "     |  \n",
      "     |  Raises\n",
      "     |  ------\n",
      "     |  ValueError\n",
      "     |      If the interior knots do not satisfy the Schoenberg-Whitney conditions\n",
      "     |  \n",
      "     |  See Also\n",
      "     |  --------\n",
      "     |  UnivariateSpline :\n",
      "     |      a smooth univariate spline to fit a given set of data points.\n",
      "     |  InterpolatedUnivariateSpline :\n",
      "     |      a interpolating univariate spline for a given set of data points.\n",
      "     |  splrep :\n",
      "     |      a function to find the B-spline representation of a 1-D curve\n",
      "     |  splev :\n",
      "     |      a function to evaluate a B-spline or its derivatives\n",
      "     |  sproot :\n",
      "     |      a function to find the roots of a cubic B-spline\n",
      "     |  splint :\n",
      "     |      a function to evaluate the definite integral of a B-spline between two\n",
      "     |      given points\n",
      "     |  spalde :\n",
      "     |      a function to evaluate all derivatives of a B-spline\n",
      "     |  \n",
      "     |  Notes\n",
      "     |  -----\n",
      "     |  The number of data points must be larger than the spline degree `k`.\n",
      "     |  \n",
      "     |  Knots `t` must satisfy the Schoenberg-Whitney conditions,\n",
      "     |  i.e., there must be a subset of data points ``x[j]`` such that\n",
      "     |  ``t[j] < x[j] < t[j+k+1]``, for ``j=0, 1,...,n-k-2``.\n",
      "     |  \n",
      "     |  Examples\n",
      "     |  --------\n",
      "     |  >>> from scipy.interpolate import LSQUnivariateSpline, UnivariateSpline\n",
      "     |  >>> import matplotlib.pyplot as plt\n",
      "     |  >>> rng = np.random.default_rng()\n",
      "     |  >>> x = np.linspace(-3, 3, 50)\n",
      "     |  >>> y = np.exp(-x**2) + 0.1 * rng.standard_normal(50)\n",
      "     |  \n",
      "     |  Fit a smoothing spline with a pre-defined internal knots:\n",
      "     |  \n",
      "     |  >>> t = [-1, 0, 1]\n",
      "     |  >>> spl = LSQUnivariateSpline(x, y, t)\n",
      "     |  \n",
      "     |  >>> xs = np.linspace(-3, 3, 1000)\n",
      "     |  >>> plt.plot(x, y, 'ro', ms=5)\n",
      "     |  >>> plt.plot(xs, spl(xs), 'g-', lw=3)\n",
      "     |  >>> plt.show()\n",
      "     |  \n",
      "     |  Check the knot vector:\n",
      "     |  \n",
      "     |  >>> spl.get_knots()\n",
      "     |  array([-3., -1., 0., 1., 3.])\n",
      "     |  \n",
      "     |  Constructing lsq spline using the knots from another spline:\n",
      "     |  \n",
      "     |  >>> x = np.arange(10)\n",
      "     |  >>> s = UnivariateSpline(x, x, s=0)\n",
      "     |  >>> s.get_knots()\n",
      "     |  array([ 0.,  2.,  3.,  4.,  5.,  6.,  7.,  9.])\n",
      "     |  >>> knt = s.get_knots()\n",
      "     |  >>> s1 = LSQUnivariateSpline(x, x, knt[1:-1])    # Chop 1st and last knot\n",
      "     |  >>> s1.get_knots()\n",
      "     |  array([ 0.,  2.,  3.,  4.,  5.,  6.,  7.,  9.])\n",
      "     |  \n",
      "     |  Method resolution order:\n",
      "     |      LSQUnivariateSpline\n",
      "     |      UnivariateSpline\n",
      "     |      builtins.object\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __init__(self, x, y, t, w=None, bbox=[None, None], k=3, ext=0, check_finite=False)\n",
      "     |      Initialize self.  See help(type(self)) for accurate signature.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from UnivariateSpline:\n",
      "     |  \n",
      "     |  __call__(self, x, nu=0, ext=None)\n",
      "     |      Evaluate spline (or its nu-th derivative) at positions x.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      x : array_like\n",
      "     |          A 1-D array of points at which to return the value of the smoothed\n",
      "     |          spline or its derivatives. Note: `x` can be unordered but the\n",
      "     |          evaluation is more efficient if `x` is (partially) ordered.\n",
      "     |      nu  : int\n",
      "     |          The order of derivative of the spline to compute.\n",
      "     |      ext : int\n",
      "     |          Controls the value returned for elements of `x` not in the\n",
      "     |          interval defined by the knot sequence.\n",
      "     |      \n",
      "     |          * if ext=0 or 'extrapolate', return the extrapolated value.\n",
      "     |          * if ext=1 or 'zeros', return 0\n",
      "     |          * if ext=2 or 'raise', raise a ValueError\n",
      "     |          * if ext=3 or 'const', return the boundary value.\n",
      "     |      \n",
      "     |          The default value is 0, passed from the initialization of\n",
      "     |          UnivariateSpline.\n",
      "     |  \n",
      "     |  antiderivative(self, n=1)\n",
      "     |      Construct a new spline representing the antiderivative of this spline.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      n : int, optional\n",
      "     |          Order of antiderivative to evaluate. Default: 1\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      spline : UnivariateSpline\n",
      "     |          Spline of order k2=k+n representing the antiderivative of this\n",
      "     |          spline.\n",
      "     |      \n",
      "     |      Notes\n",
      "     |      -----\n",
      "     |      \n",
      "     |      .. versionadded:: 0.13.0\n",
      "     |      \n",
      "     |      See Also\n",
      "     |      --------\n",
      "     |      splantider, derivative\n",
      "     |      \n",
      "     |      Examples\n",
      "     |      --------\n",
      "     |      >>> from scipy.interpolate import UnivariateSpline\n",
      "     |      >>> x = np.linspace(0, np.pi/2, 70)\n",
      "     |      >>> y = 1 / np.sqrt(1 - 0.8*np.sin(x)**2)\n",
      "     |      >>> spl = UnivariateSpline(x, y, s=0)\n",
      "     |      \n",
      "     |      The derivative is the inverse operation of the antiderivative,\n",
      "     |      although some floating point error accumulates:\n",
      "     |      \n",
      "     |      >>> spl(1.7), spl.antiderivative().derivative()(1.7)\n",
      "     |      (array(2.1565429877197317), array(2.1565429877201865))\n",
      "     |      \n",
      "     |      Antiderivative can be used to evaluate definite integrals:\n",
      "     |      \n",
      "     |      >>> ispl = spl.antiderivative()\n",
      "     |      >>> ispl(np.pi/2) - ispl(0)\n",
      "     |      2.2572053588768486\n",
      "     |      \n",
      "     |      This is indeed an approximation to the complete elliptic integral\n",
      "     |      :math:`K(m) = \\int_0^{\\pi/2} [1 - m\\sin^2 x]^{-1/2} dx`:\n",
      "     |      \n",
      "     |      >>> from scipy.special import ellipk\n",
      "     |      >>> ellipk(0.8)\n",
      "     |      2.2572053268208538\n",
      "     |  \n",
      "     |  derivative(self, n=1)\n",
      "     |      Construct a new spline representing the derivative of this spline.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      n : int, optional\n",
      "     |          Order of derivative to evaluate. Default: 1\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      spline : UnivariateSpline\n",
      "     |          Spline of order k2=k-n representing the derivative of this\n",
      "     |          spline.\n",
      "     |      \n",
      "     |      See Also\n",
      "     |      --------\n",
      "     |      splder, antiderivative\n",
      "     |      \n",
      "     |      Notes\n",
      "     |      -----\n",
      "     |      \n",
      "     |      .. versionadded:: 0.13.0\n",
      "     |      \n",
      "     |      Examples\n",
      "     |      --------\n",
      "     |      This can be used for finding maxima of a curve:\n",
      "     |      \n",
      "     |      >>> from scipy.interpolate import UnivariateSpline\n",
      "     |      >>> x = np.linspace(0, 10, 70)\n",
      "     |      >>> y = np.sin(x)\n",
      "     |      >>> spl = UnivariateSpline(x, y, k=4, s=0)\n",
      "     |      \n",
      "     |      Now, differentiate the spline and find the zeros of the\n",
      "     |      derivative. (NB: `sproot` only works for order 3 splines, so we\n",
      "     |      fit an order 4 spline):\n",
      "     |      \n",
      "     |      >>> spl.derivative().roots() / np.pi\n",
      "     |      array([ 0.50000001,  1.5       ,  2.49999998])\n",
      "     |      \n",
      "     |      This agrees well with roots :math:`\\pi/2 + n\\pi` of\n",
      "     |      :math:`\\cos(x) = \\sin'(x)`.\n",
      "     |  \n",
      "     |  derivatives(self, x)\n",
      "     |      Return all derivatives of the spline at the point x.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      x : float\n",
      "     |          The point to evaluate the derivatives at.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      der : ndarray, shape(k+1,)\n",
      "     |          Derivatives of the orders 0 to k.\n",
      "     |      \n",
      "     |      Examples\n",
      "     |      --------\n",
      "     |      >>> from scipy.interpolate import UnivariateSpline\n",
      "     |      >>> x = np.linspace(0, 3, 11)\n",
      "     |      >>> y = x**2\n",
      "     |      >>> spl = UnivariateSpline(x, y)\n",
      "     |      >>> spl.derivatives(1.5)\n",
      "     |      array([2.25, 3.0, 2.0, 0])\n",
      "     |  \n",
      "     |  get_coeffs(self)\n",
      "     |      Return spline coefficients.\n",
      "     |  \n",
      "     |  get_knots(self)\n",
      "     |      Return positions of interior knots of the spline.\n",
      "     |      \n",
      "     |      Internally, the knot vector contains ``2*k`` additional boundary knots.\n",
      "     |  \n",
      "     |  get_residual(self)\n",
      "     |      Return weighted sum of squared residuals of the spline approximation.\n",
      "     |      \n",
      "     |      This is equivalent to::\n",
      "     |      \n",
      "     |           sum((w[i] * (y[i]-spl(x[i])))**2, axis=0)\n",
      "     |  \n",
      "     |  integral(self, a, b)\n",
      "     |      Return definite integral of the spline between two given points.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      a : float\n",
      "     |          Lower limit of integration.\n",
      "     |      b : float\n",
      "     |          Upper limit of integration.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      integral : float\n",
      "     |          The value of the definite integral of the spline between limits.\n",
      "     |      \n",
      "     |      Examples\n",
      "     |      --------\n",
      "     |      >>> from scipy.interpolate import UnivariateSpline\n",
      "     |      >>> x = np.linspace(0, 3, 11)\n",
      "     |      >>> y = x**2\n",
      "     |      >>> spl = UnivariateSpline(x, y)\n",
      "     |      >>> spl.integral(0, 3)\n",
      "     |      9.0\n",
      "     |      \n",
      "     |      which agrees with :math:`\\int x^2 dx = x^3 / 3` between the limits\n",
      "     |      of 0 and 3.\n",
      "     |      \n",
      "     |      A caveat is that this routine assumes the spline to be zero outside of\n",
      "     |      the data limits:\n",
      "     |      \n",
      "     |      >>> spl.integral(-1, 4)\n",
      "     |      9.0\n",
      "     |      >>> spl.integral(-1, 0)\n",
      "     |      0.0\n",
      "     |  \n",
      "     |  roots(self)\n",
      "     |      Return the zeros of the spline.\n",
      "     |      \n",
      "     |      Restriction: only cubic splines are supported by fitpack.\n",
      "     |  \n",
      "     |  set_smoothing_factor(self, s)\n",
      "     |      Continue spline computation with the given smoothing\n",
      "     |      factor s and with the knots found at the last call.\n",
      "     |      \n",
      "     |      This routine modifies the spline in place.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Static methods inherited from UnivariateSpline:\n",
      "     |  \n",
      "     |  validate_input(x, y, w, bbox, k, s, ext, check_finite)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from UnivariateSpline:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |      dictionary for instance variables (if defined)\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "    \n",
      "    class LinearNDInterpolator(NDInterpolatorBase)\n",
      "     |  LinearNDInterpolator(points, values, fill_value=nan, rescale=False)\n",
      "     |  \n",
      "     |  LinearNDInterpolator(points, values, fill_value=np.nan, rescale=False)\n",
      "     |  \n",
      "     |  Piecewise linear interpolant in N > 1 dimensions.\n",
      "     |  \n",
      "     |  .. versionadded:: 0.9\n",
      "     |  \n",
      "     |  Methods\n",
      "     |  -------\n",
      "     |  __call__\n",
      "     |  \n",
      "     |  Parameters\n",
      "     |  ----------\n",
      "     |  points : ndarray of floats, shape (npoints, ndims); or Delaunay\n",
      "     |      Data point coordinates, or a precomputed Delaunay triangulation.\n",
      "     |  values : ndarray of float or complex, shape (npoints, ...)\n",
      "     |      Data values.\n",
      "     |  fill_value : float, optional\n",
      "     |      Value used to fill in for requested points outside of the\n",
      "     |      convex hull of the input points.  If not provided, then\n",
      "     |      the default is ``nan``.\n",
      "     |  rescale : bool, optional\n",
      "     |      Rescale points to unit cube before performing interpolation.\n",
      "     |      This is useful if some of the input dimensions have\n",
      "     |      incommensurable units and differ by many orders of magnitude.\n",
      "     |  \n",
      "     |  Notes\n",
      "     |  -----\n",
      "     |  The interpolant is constructed by triangulating the input data\n",
      "     |  with Qhull [1]_, and on each triangle performing linear\n",
      "     |  barycentric interpolation.\n",
      "     |  \n",
      "     |  Examples\n",
      "     |  --------\n",
      "     |  We can interpolate values on a 2D plane:\n",
      "     |  \n",
      "     |  >>> from scipy.interpolate import LinearNDInterpolator\n",
      "     |  >>> import matplotlib.pyplot as plt\n",
      "     |  >>> rng = np.random.default_rng()\n",
      "     |  >>> x = rng.random(10) - 0.5\n",
      "     |  >>> y = rng.random(10) - 0.5\n",
      "     |  >>> z = np.hypot(x, y)\n",
      "     |  >>> X = np.linspace(min(x), max(x))\n",
      "     |  >>> Y = np.linspace(min(y), max(y))\n",
      "     |  >>> X, Y = np.meshgrid(X, Y)  # 2D grid for interpolation\n",
      "     |  >>> interp = LinearNDInterpolator(list(zip(x, y)), z)\n",
      "     |  >>> Z = interp(X, Y)\n",
      "     |  >>> plt.pcolormesh(X, Y, Z, shading='auto')\n",
      "     |  >>> plt.plot(x, y, \"ok\", label=\"input point\")\n",
      "     |  >>> plt.legend()\n",
      "     |  >>> plt.colorbar()\n",
      "     |  >>> plt.axis(\"equal\")\n",
      "     |  >>> plt.show()\n",
      "     |  \n",
      "     |  See also\n",
      "     |  --------\n",
      "     |  griddata :\n",
      "     |      Interpolate unstructured D-D data.\n",
      "     |  NearestNDInterpolator :\n",
      "     |      Nearest-neighbor interpolation in N dimensions.\n",
      "     |  CloughTocher2DInterpolator :\n",
      "     |      Piecewise cubic, C1 smooth, curvature-minimizing interpolant in 2D.\n",
      "     |  \n",
      "     |  References\n",
      "     |  ----------\n",
      "     |  .. [1] http://www.qhull.org/\n",
      "     |  \n",
      "     |  Method resolution order:\n",
      "     |      LinearNDInterpolator\n",
      "     |      NDInterpolatorBase\n",
      "     |      builtins.object\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __init__(self, points, values, fill_value=nan, rescale=False)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from NDInterpolatorBase:\n",
      "     |  \n",
      "     |  __call__(self, *args)\n",
      "     |      interpolator(xi)\n",
      "     |      \n",
      "     |      Evaluate interpolator at given points.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      x1, x2, ... xn: array-like of float\n",
      "     |          Points where to interpolate data at.\n",
      "     |          x1, x2, ... xn can be array-like of float with broadcastable shape.\n",
      "     |          or x1 can be array-like of float with shape ``(..., ndim)``\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from NDInterpolatorBase:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |      dictionary for instance variables (if defined)\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "    \n",
      "    class NdPPoly(builtins.object)\n",
      "     |  NdPPoly(c, x, extrapolate=None)\n",
      "     |  \n",
      "     |  Piecewise tensor product polynomial\n",
      "     |  \n",
      "     |  The value at point ``xp = (x', y', z', ...)`` is evaluated by first\n",
      "     |  computing the interval indices `i` such that::\n",
      "     |  \n",
      "     |      x[0][i[0]] <= x' < x[0][i[0]+1]\n",
      "     |      x[1][i[1]] <= y' < x[1][i[1]+1]\n",
      "     |      ...\n",
      "     |  \n",
      "     |  and then computing::\n",
      "     |  \n",
      "     |      S = sum(c[k0-m0-1,...,kn-mn-1,i[0],...,i[n]]\n",
      "     |              * (xp[0] - x[0][i[0]])**m0\n",
      "     |              * ...\n",
      "     |              * (xp[n] - x[n][i[n]])**mn\n",
      "     |              for m0 in range(k[0]+1)\n",
      "     |              ...\n",
      "     |              for mn in range(k[n]+1))\n",
      "     |  \n",
      "     |  where ``k[j]`` is the degree of the polynomial in dimension j. This\n",
      "     |  representation is the piecewise multivariate power basis.\n",
      "     |  \n",
      "     |  Parameters\n",
      "     |  ----------\n",
      "     |  c : ndarray, shape (k0, ..., kn, m0, ..., mn, ...)\n",
      "     |      Polynomial coefficients, with polynomial order `kj` and\n",
      "     |      `mj+1` intervals for each dimension `j`.\n",
      "     |  x : ndim-tuple of ndarrays, shapes (mj+1,)\n",
      "     |      Polynomial breakpoints for each dimension. These must be\n",
      "     |      sorted in increasing order.\n",
      "     |  extrapolate : bool, optional\n",
      "     |      Whether to extrapolate to out-of-bounds points based on first\n",
      "     |      and last intervals, or to return NaNs. Default: True.\n",
      "     |  \n",
      "     |  Attributes\n",
      "     |  ----------\n",
      "     |  x : tuple of ndarrays\n",
      "     |      Breakpoints.\n",
      "     |  c : ndarray\n",
      "     |      Coefficients of the polynomials.\n",
      "     |  \n",
      "     |  Methods\n",
      "     |  -------\n",
      "     |  __call__\n",
      "     |  derivative\n",
      "     |  antiderivative\n",
      "     |  integrate\n",
      "     |  integrate_1d\n",
      "     |  construct_fast\n",
      "     |  \n",
      "     |  See also\n",
      "     |  --------\n",
      "     |  PPoly : piecewise polynomials in 1D\n",
      "     |  \n",
      "     |  Notes\n",
      "     |  -----\n",
      "     |  High-order polynomials in the power basis can be numerically\n",
      "     |  unstable.\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __call__(self, x, nu=None, extrapolate=None)\n",
      "     |      Evaluate the piecewise polynomial or its derivative\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      x : array-like\n",
      "     |          Points to evaluate the interpolant at.\n",
      "     |      nu : tuple, optional\n",
      "     |          Orders of derivatives to evaluate. Each must be non-negative.\n",
      "     |      extrapolate : bool, optional\n",
      "     |          Whether to extrapolate to out-of-bounds points based on first\n",
      "     |          and last intervals, or to return NaNs.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      y : array-like\n",
      "     |          Interpolated values. Shape is determined by replacing\n",
      "     |          the interpolation axis in the original array with the shape of x.\n",
      "     |      \n",
      "     |      Notes\n",
      "     |      -----\n",
      "     |      Derivatives are evaluated piecewise for each polynomial\n",
      "     |      segment, even if the polynomial is not differentiable at the\n",
      "     |      breakpoints. The polynomial intervals are considered half-open,\n",
      "     |      ``[a, b)``, except for the last interval which is closed\n",
      "     |      ``[a, b]``.\n",
      "     |  \n",
      "     |  __init__(self, c, x, extrapolate=None)\n",
      "     |      Initialize self.  See help(type(self)) for accurate signature.\n",
      "     |  \n",
      "     |  antiderivative(self, nu)\n",
      "     |      Construct a new piecewise polynomial representing the antiderivative.\n",
      "     |      \n",
      "     |      Antiderivative is also the indefinite integral of the function,\n",
      "     |      and derivative is its inverse operation.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      nu : ndim-tuple of int\n",
      "     |          Order of derivatives to evaluate for each dimension.\n",
      "     |          If negative, the derivative is returned.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      pp : PPoly\n",
      "     |          Piecewise polynomial of order k2 = k + n representing\n",
      "     |          the antiderivative of this polynomial.\n",
      "     |      \n",
      "     |      Notes\n",
      "     |      -----\n",
      "     |      The antiderivative returned by this function is continuous and\n",
      "     |      continuously differentiable to order n-1, up to floating point\n",
      "     |      rounding error.\n",
      "     |  \n",
      "     |  derivative(self, nu)\n",
      "     |      Construct a new piecewise polynomial representing the derivative.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      nu : ndim-tuple of int\n",
      "     |          Order of derivatives to evaluate for each dimension.\n",
      "     |          If negative, the antiderivative is returned.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      pp : NdPPoly\n",
      "     |          Piecewise polynomial of orders (k[0] - nu[0], ..., k[n] - nu[n])\n",
      "     |          representing the derivative of this polynomial.\n",
      "     |      \n",
      "     |      Notes\n",
      "     |      -----\n",
      "     |      Derivatives are evaluated piecewise for each polynomial\n",
      "     |      segment, even if the polynomial is not differentiable at the\n",
      "     |      breakpoints. The polynomial intervals in each dimension are\n",
      "     |      considered half-open, ``[a, b)``, except for the last interval\n",
      "     |      which is closed ``[a, b]``.\n",
      "     |  \n",
      "     |  integrate(self, ranges, extrapolate=None)\n",
      "     |      Compute a definite integral over a piecewise polynomial.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      ranges : ndim-tuple of 2-tuples float\n",
      "     |          Sequence of lower and upper bounds for each dimension,\n",
      "     |          ``[(a[0], b[0]), ..., (a[ndim-1], b[ndim-1])]``\n",
      "     |      extrapolate : bool, optional\n",
      "     |          Whether to extrapolate to out-of-bounds points based on first\n",
      "     |          and last intervals, or to return NaNs.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      ig : array_like\n",
      "     |          Definite integral of the piecewise polynomial over\n",
      "     |          [a[0], b[0]] x ... x [a[ndim-1], b[ndim-1]]\n",
      "     |  \n",
      "     |  integrate_1d(self, a, b, axis, extrapolate=None)\n",
      "     |      Compute NdPPoly representation for one dimensional definite integral\n",
      "     |      \n",
      "     |      The result is a piecewise polynomial representing the integral:\n",
      "     |      \n",
      "     |      .. math::\n",
      "     |      \n",
      "     |         p(y, z, ...) = \\int_a^b dx\\, p(x, y, z, ...)\n",
      "     |      \n",
      "     |      where the dimension integrated over is specified with the\n",
      "     |      `axis` parameter.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      a, b : float\n",
      "     |          Lower and upper bound for integration.\n",
      "     |      axis : int\n",
      "     |          Dimension over which to compute the 1-D integrals\n",
      "     |      extrapolate : bool, optional\n",
      "     |          Whether to extrapolate to out-of-bounds points based on first\n",
      "     |          and last intervals, or to return NaNs.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      ig : NdPPoly or array-like\n",
      "     |          Definite integral of the piecewise polynomial over [a, b].\n",
      "     |          If the polynomial was 1D, an array is returned,\n",
      "     |          otherwise, an NdPPoly object.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Class methods defined here:\n",
      "     |  \n",
      "     |  construct_fast(c, x, extrapolate=None) from builtins.type\n",
      "     |      Construct the piecewise polynomial without making checks.\n",
      "     |      \n",
      "     |      Takes the same parameters as the constructor. Input arguments\n",
      "     |      ``c`` and ``x`` must be arrays of the correct shape and type.  The\n",
      "     |      ``c`` array can only be of dtypes float and complex, and ``x``\n",
      "     |      array must have dtype float.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors defined here:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |      dictionary for instance variables (if defined)\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "    \n",
      "    class NearestNDInterpolator(scipy.interpolate.interpnd.NDInterpolatorBase)\n",
      "     |  NearestNDInterpolator(x, y, rescale=False, tree_options=None)\n",
      "     |  \n",
      "     |  NearestNDInterpolator(x, y).\n",
      "     |  \n",
      "     |  Nearest-neighbor interpolation in N > 1 dimensions.\n",
      "     |  \n",
      "     |  .. versionadded:: 0.9\n",
      "     |  \n",
      "     |  Methods\n",
      "     |  -------\n",
      "     |  __call__\n",
      "     |  \n",
      "     |  Parameters\n",
      "     |  ----------\n",
      "     |  x : (Npoints, Ndims) ndarray of floats\n",
      "     |      Data point coordinates.\n",
      "     |  y : (Npoints,) ndarray of float or complex\n",
      "     |      Data values.\n",
      "     |  rescale : boolean, optional\n",
      "     |      Rescale points to unit cube before performing interpolation.\n",
      "     |      This is useful if some of the input dimensions have\n",
      "     |      incommensurable units and differ by many orders of magnitude.\n",
      "     |  \n",
      "     |      .. versionadded:: 0.14.0\n",
      "     |  tree_options : dict, optional\n",
      "     |      Options passed to the underlying ``cKDTree``.\n",
      "     |  \n",
      "     |      .. versionadded:: 0.17.0\n",
      "     |  \n",
      "     |  \n",
      "     |  Notes\n",
      "     |  -----\n",
      "     |  Uses ``scipy.spatial.cKDTree``\n",
      "     |  \n",
      "     |  Examples\n",
      "     |  --------\n",
      "     |  We can interpolate values on a 2D plane:\n",
      "     |  \n",
      "     |  >>> from scipy.interpolate import NearestNDInterpolator\n",
      "     |  >>> import matplotlib.pyplot as plt\n",
      "     |  >>> rng = np.random.default_rng()\n",
      "     |  >>> x = rng.random(10) - 0.5\n",
      "     |  >>> y = rng.random(10) - 0.5\n",
      "     |  >>> z = np.hypot(x, y)\n",
      "     |  >>> X = np.linspace(min(x), max(x))\n",
      "     |  >>> Y = np.linspace(min(y), max(y))\n",
      "     |  >>> X, Y = np.meshgrid(X, Y)  # 2D grid for interpolation\n",
      "     |  >>> interp = NearestNDInterpolator(list(zip(x, y)), z)\n",
      "     |  >>> Z = interp(X, Y)\n",
      "     |  >>> plt.pcolormesh(X, Y, Z, shading='auto')\n",
      "     |  >>> plt.plot(x, y, \"ok\", label=\"input point\")\n",
      "     |  >>> plt.legend()\n",
      "     |  >>> plt.colorbar()\n",
      "     |  >>> plt.axis(\"equal\")\n",
      "     |  >>> plt.show()\n",
      "     |  \n",
      "     |  See also\n",
      "     |  --------\n",
      "     |  griddata :\n",
      "     |      Interpolate unstructured D-D data.\n",
      "     |  LinearNDInterpolator :\n",
      "     |      Piecewise linear interpolant in N dimensions.\n",
      "     |  CloughTocher2DInterpolator :\n",
      "     |      Piecewise cubic, C1 smooth, curvature-minimizing interpolant in 2D.\n",
      "     |  \n",
      "     |  Method resolution order:\n",
      "     |      NearestNDInterpolator\n",
      "     |      scipy.interpolate.interpnd.NDInterpolatorBase\n",
      "     |      builtins.object\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __call__(self, *args)\n",
      "     |      Evaluate interpolator at given points.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      x1, x2, ... xn: array-like of float\n",
      "     |          Points where to interpolate data at.\n",
      "     |          x1, x2, ... xn can be array-like of float with broadcastable shape.\n",
      "     |          or x1 can be array-like of float with shape ``(..., ndim)``\n",
      "     |  \n",
      "     |  __init__(self, x, y, rescale=False, tree_options=None)\n",
      "     |      Check shape of points and values arrays, and reshape values to\n",
      "     |      (npoints, nvalues).  Ensure the `points` and values arrays are\n",
      "     |      C-contiguous, and of correct type.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from scipy.interpolate.interpnd.NDInterpolatorBase:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |      dictionary for instance variables (if defined)\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "    \n",
      "    class PPoly(_PPolyBase)\n",
      "     |  PPoly(c, x, extrapolate=None, axis=0)\n",
      "     |  \n",
      "     |  Piecewise polynomial in terms of coefficients and breakpoints\n",
      "     |  \n",
      "     |  The polynomial between ``x[i]`` and ``x[i + 1]`` is written in the\n",
      "     |  local power basis::\n",
      "     |  \n",
      "     |      S = sum(c[m, i] * (xp - x[i])**(k-m) for m in range(k+1))\n",
      "     |  \n",
      "     |  where ``k`` is the degree of the polynomial.\n",
      "     |  \n",
      "     |  Parameters\n",
      "     |  ----------\n",
      "     |  c : ndarray, shape (k, m, ...)\n",
      "     |      Polynomial coefficients, order `k` and `m` intervals.\n",
      "     |  x : ndarray, shape (m+1,)\n",
      "     |      Polynomial breakpoints. Must be sorted in either increasing or\n",
      "     |      decreasing order.\n",
      "     |  extrapolate : bool or 'periodic', optional\n",
      "     |      If bool, determines whether to extrapolate to out-of-bounds points\n",
      "     |      based on first and last intervals, or to return NaNs. If 'periodic',\n",
      "     |      periodic extrapolation is used. Default is True.\n",
      "     |  axis : int, optional\n",
      "     |      Interpolation axis. Default is zero.\n",
      "     |  \n",
      "     |  Attributes\n",
      "     |  ----------\n",
      "     |  x : ndarray\n",
      "     |      Breakpoints.\n",
      "     |  c : ndarray\n",
      "     |      Coefficients of the polynomials. They are reshaped\n",
      "     |      to a 3-D array with the last dimension representing\n",
      "     |      the trailing dimensions of the original coefficient array.\n",
      "     |  axis : int\n",
      "     |      Interpolation axis.\n",
      "     |  \n",
      "     |  Methods\n",
      "     |  -------\n",
      "     |  __call__\n",
      "     |  derivative\n",
      "     |  antiderivative\n",
      "     |  integrate\n",
      "     |  solve\n",
      "     |  roots\n",
      "     |  extend\n",
      "     |  from_spline\n",
      "     |  from_bernstein_basis\n",
      "     |  construct_fast\n",
      "     |  \n",
      "     |  See also\n",
      "     |  --------\n",
      "     |  BPoly : piecewise polynomials in the Bernstein basis\n",
      "     |  \n",
      "     |  Notes\n",
      "     |  -----\n",
      "     |  High-order polynomials in the power basis can be numerically\n",
      "     |  unstable. Precision problems can start to appear for orders\n",
      "     |  larger than 20-30.\n",
      "     |  \n",
      "     |  Method resolution order:\n",
      "     |      PPoly\n",
      "     |      _PPolyBase\n",
      "     |      builtins.object\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  antiderivative(self, nu=1)\n",
      "     |      Construct a new piecewise polynomial representing the antiderivative.\n",
      "     |      \n",
      "     |      Antiderivative is also the indefinite integral of the function,\n",
      "     |      and derivative is its inverse operation.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      nu : int, optional\n",
      "     |          Order of antiderivative to evaluate. Default is 1, i.e., compute\n",
      "     |          the first integral. If negative, the derivative is returned.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      pp : PPoly\n",
      "     |          Piecewise polynomial of order k2 = k + n representing\n",
      "     |          the antiderivative of this polynomial.\n",
      "     |      \n",
      "     |      Notes\n",
      "     |      -----\n",
      "     |      The antiderivative returned by this function is continuous and\n",
      "     |      continuously differentiable to order n-1, up to floating point\n",
      "     |      rounding error.\n",
      "     |      \n",
      "     |      If antiderivative is computed and ``self.extrapolate='periodic'``,\n",
      "     |      it will be set to False for the returned instance. This is done because\n",
      "     |      the antiderivative is no longer periodic and its correct evaluation\n",
      "     |      outside of the initially given x interval is difficult.\n",
      "     |  \n",
      "     |  derivative(self, nu=1)\n",
      "     |      Construct a new piecewise polynomial representing the derivative.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      nu : int, optional\n",
      "     |          Order of derivative to evaluate. Default is 1, i.e., compute the\n",
      "     |          first derivative. If negative, the antiderivative is returned.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      pp : PPoly\n",
      "     |          Piecewise polynomial of order k2 = k - n representing the derivative\n",
      "     |          of this polynomial.\n",
      "     |      \n",
      "     |      Notes\n",
      "     |      -----\n",
      "     |      Derivatives are evaluated piecewise for each polynomial\n",
      "     |      segment, even if the polynomial is not differentiable at the\n",
      "     |      breakpoints. The polynomial intervals are considered half-open,\n",
      "     |      ``[a, b)``, except for the last interval which is closed\n",
      "     |      ``[a, b]``.\n",
      "     |  \n",
      "     |  integrate(self, a, b, extrapolate=None)\n",
      "     |      Compute a definite integral over a piecewise polynomial.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      a : float\n",
      "     |          Lower integration bound\n",
      "     |      b : float\n",
      "     |          Upper integration bound\n",
      "     |      extrapolate : {bool, 'periodic', None}, optional\n",
      "     |          If bool, determines whether to extrapolate to out-of-bounds points\n",
      "     |          based on first and last intervals, or to return NaNs.\n",
      "     |          If 'periodic', periodic extrapolation is used.\n",
      "     |          If None (default), use `self.extrapolate`.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      ig : array_like\n",
      "     |          Definite integral of the piecewise polynomial over [a, b]\n",
      "     |  \n",
      "     |  roots(self, discontinuity=True, extrapolate=None)\n",
      "     |      Find real roots of the the piecewise polynomial.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      discontinuity : bool, optional\n",
      "     |          Whether to report sign changes across discontinuities at\n",
      "     |          breakpoints as roots.\n",
      "     |      extrapolate : {bool, 'periodic', None}, optional\n",
      "     |          If bool, determines whether to return roots from the polynomial\n",
      "     |          extrapolated based on first and last intervals, 'periodic' works\n",
      "     |          the same as False. If None (default), use `self.extrapolate`.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      roots : ndarray\n",
      "     |          Roots of the polynomial(s).\n",
      "     |      \n",
      "     |          If the PPoly object describes multiple polynomials, the\n",
      "     |          return value is an object array whose each element is an\n",
      "     |          ndarray containing the roots.\n",
      "     |      \n",
      "     |      See Also\n",
      "     |      --------\n",
      "     |      PPoly.solve\n",
      "     |  \n",
      "     |  solve(self, y=0.0, discontinuity=True, extrapolate=None)\n",
      "     |      Find real solutions of the the equation ``pp(x) == y``.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      y : float, optional\n",
      "     |          Right-hand side. Default is zero.\n",
      "     |      discontinuity : bool, optional\n",
      "     |          Whether to report sign changes across discontinuities at\n",
      "     |          breakpoints as roots.\n",
      "     |      extrapolate : {bool, 'periodic', None}, optional\n",
      "     |          If bool, determines whether to return roots from the polynomial\n",
      "     |          extrapolated based on first and last intervals, 'periodic' works\n",
      "     |          the same as False. If None (default), use `self.extrapolate`.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      roots : ndarray\n",
      "     |          Roots of the polynomial(s).\n",
      "     |      \n",
      "     |          If the PPoly object describes multiple polynomials, the\n",
      "     |          return value is an object array whose each element is an\n",
      "     |          ndarray containing the roots.\n",
      "     |      \n",
      "     |      Notes\n",
      "     |      -----\n",
      "     |      This routine works only on real-valued polynomials.\n",
      "     |      \n",
      "     |      If the piecewise polynomial contains sections that are\n",
      "     |      identically zero, the root list will contain the start point\n",
      "     |      of the corresponding interval, followed by a ``nan`` value.\n",
      "     |      \n",
      "     |      If the polynomial is discontinuous across a breakpoint, and\n",
      "     |      there is a sign change across the breakpoint, this is reported\n",
      "     |      if the `discont` parameter is True.\n",
      "     |      \n",
      "     |      Examples\n",
      "     |      --------\n",
      "     |      \n",
      "     |      Finding roots of ``[x**2 - 1, (x - 1)**2]`` defined on intervals\n",
      "     |      ``[-2, 1], [1, 2]``:\n",
      "     |      \n",
      "     |      >>> from scipy.interpolate import PPoly\n",
      "     |      >>> pp = PPoly(np.array([[1, -4, 3], [1, 0, 0]]).T, [-2, 1, 2])\n",
      "     |      >>> pp.solve()\n",
      "     |      array([-1.,  1.])\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Class methods defined here:\n",
      "     |  \n",
      "     |  from_bernstein_basis(bp, extrapolate=None) from builtins.type\n",
      "     |      Construct a piecewise polynomial in the power basis\n",
      "     |      from a polynomial in Bernstein basis.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      bp : BPoly\n",
      "     |          A Bernstein basis polynomial, as created by BPoly\n",
      "     |      extrapolate : bool or 'periodic', optional\n",
      "     |          If bool, determines whether to extrapolate to out-of-bounds points\n",
      "     |          based on first and last intervals, or to return NaNs.\n",
      "     |          If 'periodic', periodic extrapolation is used. Default is True.\n",
      "     |  \n",
      "     |  from_spline(tck, extrapolate=None) from builtins.type\n",
      "     |      Construct a piecewise polynomial from a spline\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      tck\n",
      "     |          A spline, as returned by `splrep` or a BSpline object.\n",
      "     |      extrapolate : bool or 'periodic', optional\n",
      "     |          If bool, determines whether to extrapolate to out-of-bounds points\n",
      "     |          based on first and last intervals, or to return NaNs.\n",
      "     |          If 'periodic', periodic extrapolation is used. Default is True.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors defined here:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |      dictionary for instance variables (if defined)\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from _PPolyBase:\n",
      "     |  \n",
      "     |  __call__(self, x, nu=0, extrapolate=None)\n",
      "     |      Evaluate the piecewise polynomial or its derivative.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      x : array_like\n",
      "     |          Points to evaluate the interpolant at.\n",
      "     |      nu : int, optional\n",
      "     |          Order of derivative to evaluate. Must be non-negative.\n",
      "     |      extrapolate : {bool, 'periodic', None}, optional\n",
      "     |          If bool, determines whether to extrapolate to out-of-bounds points\n",
      "     |          based on first and last intervals, or to return NaNs.\n",
      "     |          If 'periodic', periodic extrapolation is used.\n",
      "     |          If None (default), use `self.extrapolate`.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      y : array_like\n",
      "     |          Interpolated values. Shape is determined by replacing\n",
      "     |          the interpolation axis in the original array with the shape of x.\n",
      "     |      \n",
      "     |      Notes\n",
      "     |      -----\n",
      "     |      Derivatives are evaluated piecewise for each polynomial\n",
      "     |      segment, even if the polynomial is not differentiable at the\n",
      "     |      breakpoints. The polynomial intervals are considered half-open,\n",
      "     |      ``[a, b)``, except for the last interval which is closed\n",
      "     |      ``[a, b]``.\n",
      "     |  \n",
      "     |  __init__(self, c, x, extrapolate=None, axis=0)\n",
      "     |      Initialize self.  See help(type(self)) for accurate signature.\n",
      "     |  \n",
      "     |  extend(self, c, x, right=None)\n",
      "     |      Add additional breakpoints and coefficients to the polynomial.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      c : ndarray, size (k, m, ...)\n",
      "     |          Additional coefficients for polynomials in intervals. Note that\n",
      "     |          the first additional interval will be formed using one of the\n",
      "     |          ``self.x`` end points.\n",
      "     |      x : ndarray, size (m,)\n",
      "     |          Additional breakpoints. Must be sorted in the same order as\n",
      "     |          ``self.x`` and either to the right or to the left of the current\n",
      "     |          breakpoints.\n",
      "     |      right\n",
      "     |          Deprecated argument. Has no effect.\n",
      "     |      \n",
      "     |          .. deprecated:: 0.19\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Class methods inherited from _PPolyBase:\n",
      "     |  \n",
      "     |  construct_fast(c, x, extrapolate=None, axis=0) from builtins.type\n",
      "     |      Construct the piecewise polynomial without making checks.\n",
      "     |      \n",
      "     |      Takes the same parameters as the constructor. Input arguments\n",
      "     |      ``c`` and ``x`` must be arrays of the correct shape and type. The\n",
      "     |      ``c`` array can only be of dtypes float and complex, and ``x``\n",
      "     |      array must have dtype float.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from _PPolyBase:\n",
      "     |  \n",
      "     |  axis\n",
      "     |  \n",
      "     |  c\n",
      "     |  \n",
      "     |  extrapolate\n",
      "     |  \n",
      "     |  x\n",
      "    \n",
      "    class PchipInterpolator(CubicHermiteSpline)\n",
      "     |  PchipInterpolator(x, y, axis=0, extrapolate=None)\n",
      "     |  \n",
      "     |  PCHIP 1-D monotonic cubic interpolation.\n",
      "     |  \n",
      "     |  ``x`` and ``y`` are arrays of values used to approximate some function f,\n",
      "     |  with ``y = f(x)``. The interpolant uses monotonic cubic splines\n",
      "     |  to find the value of new points. (PCHIP stands for Piecewise Cubic\n",
      "     |  Hermite Interpolating Polynomial).\n",
      "     |  \n",
      "     |  Parameters\n",
      "     |  ----------\n",
      "     |  x : ndarray\n",
      "     |      A 1-D array of monotonically increasing real values. ``x`` cannot\n",
      "     |      include duplicate values (otherwise f is overspecified)\n",
      "     |  y : ndarray\n",
      "     |      A 1-D array of real values. ``y``'s length along the interpolation\n",
      "     |      axis must be equal to the length of ``x``. If N-D array, use ``axis``\n",
      "     |      parameter to select correct axis.\n",
      "     |  axis : int, optional\n",
      "     |      Axis in the y array corresponding to the x-coordinate values.\n",
      "     |  extrapolate : bool, optional\n",
      "     |      Whether to extrapolate to out-of-bounds points based on first\n",
      "     |      and last intervals, or to return NaNs.\n",
      "     |  \n",
      "     |  Methods\n",
      "     |  -------\n",
      "     |  __call__\n",
      "     |  derivative\n",
      "     |  antiderivative\n",
      "     |  roots\n",
      "     |  \n",
      "     |  See Also\n",
      "     |  --------\n",
      "     |  CubicHermiteSpline : Piecewise-cubic interpolator.\n",
      "     |  Akima1DInterpolator : Akima 1D interpolator.\n",
      "     |  CubicSpline : Cubic spline data interpolator.\n",
      "     |  PPoly : Piecewise polynomial in terms of coefficients and breakpoints.\n",
      "     |  \n",
      "     |  Notes\n",
      "     |  -----\n",
      "     |  The interpolator preserves monotonicity in the interpolation data and does\n",
      "     |  not overshoot if the data is not smooth.\n",
      "     |  \n",
      "     |  The first derivatives are guaranteed to be continuous, but the second\n",
      "     |  derivatives may jump at :math:`x_k`.\n",
      "     |  \n",
      "     |  Determines the derivatives at the points :math:`x_k`, :math:`f'_k`,\n",
      "     |  by using PCHIP algorithm [1]_.\n",
      "     |  \n",
      "     |  Let :math:`h_k = x_{k+1} - x_k`, and  :math:`d_k = (y_{k+1} - y_k) / h_k`\n",
      "     |  are the slopes at internal points :math:`x_k`.\n",
      "     |  If the signs of :math:`d_k` and :math:`d_{k-1}` are different or either of\n",
      "     |  them equals zero, then :math:`f'_k = 0`. Otherwise, it is given by the\n",
      "     |  weighted harmonic mean\n",
      "     |  \n",
      "     |  .. math::\n",
      "     |  \n",
      "     |      \\frac{w_1 + w_2}{f'_k} = \\frac{w_1}{d_{k-1}} + \\frac{w_2}{d_k}\n",
      "     |  \n",
      "     |  where :math:`w_1 = 2 h_k + h_{k-1}` and :math:`w_2 = h_k + 2 h_{k-1}`.\n",
      "     |  \n",
      "     |  The end slopes are set using a one-sided scheme [2]_.\n",
      "     |  \n",
      "     |  \n",
      "     |  References\n",
      "     |  ----------\n",
      "     |  .. [1] F. N. Fritsch and J. Butland,\n",
      "     |         A method for constructing local\n",
      "     |         monotone piecewise cubic interpolants,\n",
      "     |         SIAM J. Sci. Comput., 5(2), 300-304 (1984).\n",
      "     |         :doi:`10.1137/0905021`.\n",
      "     |  .. [2] see, e.g., C. Moler, Numerical Computing with Matlab, 2004.\n",
      "     |         :doi:`10.1137/1.9780898717952`\n",
      "     |  \n",
      "     |  Method resolution order:\n",
      "     |      PchipInterpolator\n",
      "     |      CubicHermiteSpline\n",
      "     |      scipy.interpolate.interpolate.PPoly\n",
      "     |      scipy.interpolate.interpolate._PPolyBase\n",
      "     |      builtins.object\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __init__(self, x, y, axis=0, extrapolate=None)\n",
      "     |      Initialize self.  See help(type(self)) for accurate signature.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from scipy.interpolate.interpolate.PPoly:\n",
      "     |  \n",
      "     |  antiderivative(self, nu=1)\n",
      "     |      Construct a new piecewise polynomial representing the antiderivative.\n",
      "     |      \n",
      "     |      Antiderivative is also the indefinite integral of the function,\n",
      "     |      and derivative is its inverse operation.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      nu : int, optional\n",
      "     |          Order of antiderivative to evaluate. Default is 1, i.e., compute\n",
      "     |          the first integral. If negative, the derivative is returned.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      pp : PPoly\n",
      "     |          Piecewise polynomial of order k2 = k + n representing\n",
      "     |          the antiderivative of this polynomial.\n",
      "     |      \n",
      "     |      Notes\n",
      "     |      -----\n",
      "     |      The antiderivative returned by this function is continuous and\n",
      "     |      continuously differentiable to order n-1, up to floating point\n",
      "     |      rounding error.\n",
      "     |      \n",
      "     |      If antiderivative is computed and ``self.extrapolate='periodic'``,\n",
      "     |      it will be set to False for the returned instance. This is done because\n",
      "     |      the antiderivative is no longer periodic and its correct evaluation\n",
      "     |      outside of the initially given x interval is difficult.\n",
      "     |  \n",
      "     |  derivative(self, nu=1)\n",
      "     |      Construct a new piecewise polynomial representing the derivative.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      nu : int, optional\n",
      "     |          Order of derivative to evaluate. Default is 1, i.e., compute the\n",
      "     |          first derivative. If negative, the antiderivative is returned.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      pp : PPoly\n",
      "     |          Piecewise polynomial of order k2 = k - n representing the derivative\n",
      "     |          of this polynomial.\n",
      "     |      \n",
      "     |      Notes\n",
      "     |      -----\n",
      "     |      Derivatives are evaluated piecewise for each polynomial\n",
      "     |      segment, even if the polynomial is not differentiable at the\n",
      "     |      breakpoints. The polynomial intervals are considered half-open,\n",
      "     |      ``[a, b)``, except for the last interval which is closed\n",
      "     |      ``[a, b]``.\n",
      "     |  \n",
      "     |  integrate(self, a, b, extrapolate=None)\n",
      "     |      Compute a definite integral over a piecewise polynomial.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      a : float\n",
      "     |          Lower integration bound\n",
      "     |      b : float\n",
      "     |          Upper integration bound\n",
      "     |      extrapolate : {bool, 'periodic', None}, optional\n",
      "     |          If bool, determines whether to extrapolate to out-of-bounds points\n",
      "     |          based on first and last intervals, or to return NaNs.\n",
      "     |          If 'periodic', periodic extrapolation is used.\n",
      "     |          If None (default), use `self.extrapolate`.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      ig : array_like\n",
      "     |          Definite integral of the piecewise polynomial over [a, b]\n",
      "     |  \n",
      "     |  roots(self, discontinuity=True, extrapolate=None)\n",
      "     |      Find real roots of the the piecewise polynomial.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      discontinuity : bool, optional\n",
      "     |          Whether to report sign changes across discontinuities at\n",
      "     |          breakpoints as roots.\n",
      "     |      extrapolate : {bool, 'periodic', None}, optional\n",
      "     |          If bool, determines whether to return roots from the polynomial\n",
      "     |          extrapolated based on first and last intervals, 'periodic' works\n",
      "     |          the same as False. If None (default), use `self.extrapolate`.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      roots : ndarray\n",
      "     |          Roots of the polynomial(s).\n",
      "     |      \n",
      "     |          If the PPoly object describes multiple polynomials, the\n",
      "     |          return value is an object array whose each element is an\n",
      "     |          ndarray containing the roots.\n",
      "     |      \n",
      "     |      See Also\n",
      "     |      --------\n",
      "     |      PPoly.solve\n",
      "     |  \n",
      "     |  solve(self, y=0.0, discontinuity=True, extrapolate=None)\n",
      "     |      Find real solutions of the the equation ``pp(x) == y``.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      y : float, optional\n",
      "     |          Right-hand side. Default is zero.\n",
      "     |      discontinuity : bool, optional\n",
      "     |          Whether to report sign changes across discontinuities at\n",
      "     |          breakpoints as roots.\n",
      "     |      extrapolate : {bool, 'periodic', None}, optional\n",
      "     |          If bool, determines whether to return roots from the polynomial\n",
      "     |          extrapolated based on first and last intervals, 'periodic' works\n",
      "     |          the same as False. If None (default), use `self.extrapolate`.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      roots : ndarray\n",
      "     |          Roots of the polynomial(s).\n",
      "     |      \n",
      "     |          If the PPoly object describes multiple polynomials, the\n",
      "     |          return value is an object array whose each element is an\n",
      "     |          ndarray containing the roots.\n",
      "     |      \n",
      "     |      Notes\n",
      "     |      -----\n",
      "     |      This routine works only on real-valued polynomials.\n",
      "     |      \n",
      "     |      If the piecewise polynomial contains sections that are\n",
      "     |      identically zero, the root list will contain the start point\n",
      "     |      of the corresponding interval, followed by a ``nan`` value.\n",
      "     |      \n",
      "     |      If the polynomial is discontinuous across a breakpoint, and\n",
      "     |      there is a sign change across the breakpoint, this is reported\n",
      "     |      if the `discont` parameter is True.\n",
      "     |      \n",
      "     |      Examples\n",
      "     |      --------\n",
      "     |      \n",
      "     |      Finding roots of ``[x**2 - 1, (x - 1)**2]`` defined on intervals\n",
      "     |      ``[-2, 1], [1, 2]``:\n",
      "     |      \n",
      "     |      >>> from scipy.interpolate import PPoly\n",
      "     |      >>> pp = PPoly(np.array([[1, -4, 3], [1, 0, 0]]).T, [-2, 1, 2])\n",
      "     |      >>> pp.solve()\n",
      "     |      array([-1.,  1.])\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Class methods inherited from scipy.interpolate.interpolate.PPoly:\n",
      "     |  \n",
      "     |  from_bernstein_basis(bp, extrapolate=None) from builtins.type\n",
      "     |      Construct a piecewise polynomial in the power basis\n",
      "     |      from a polynomial in Bernstein basis.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      bp : BPoly\n",
      "     |          A Bernstein basis polynomial, as created by BPoly\n",
      "     |      extrapolate : bool or 'periodic', optional\n",
      "     |          If bool, determines whether to extrapolate to out-of-bounds points\n",
      "     |          based on first and last intervals, or to return NaNs.\n",
      "     |          If 'periodic', periodic extrapolation is used. Default is True.\n",
      "     |  \n",
      "     |  from_spline(tck, extrapolate=None) from builtins.type\n",
      "     |      Construct a piecewise polynomial from a spline\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      tck\n",
      "     |          A spline, as returned by `splrep` or a BSpline object.\n",
      "     |      extrapolate : bool or 'periodic', optional\n",
      "     |          If bool, determines whether to extrapolate to out-of-bounds points\n",
      "     |          based on first and last intervals, or to return NaNs.\n",
      "     |          If 'periodic', periodic extrapolation is used. Default is True.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from scipy.interpolate.interpolate.PPoly:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |      dictionary for instance variables (if defined)\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from scipy.interpolate.interpolate._PPolyBase:\n",
      "     |  \n",
      "     |  __call__(self, x, nu=0, extrapolate=None)\n",
      "     |      Evaluate the piecewise polynomial or its derivative.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      x : array_like\n",
      "     |          Points to evaluate the interpolant at.\n",
      "     |      nu : int, optional\n",
      "     |          Order of derivative to evaluate. Must be non-negative.\n",
      "     |      extrapolate : {bool, 'periodic', None}, optional\n",
      "     |          If bool, determines whether to extrapolate to out-of-bounds points\n",
      "     |          based on first and last intervals, or to return NaNs.\n",
      "     |          If 'periodic', periodic extrapolation is used.\n",
      "     |          If None (default), use `self.extrapolate`.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      y : array_like\n",
      "     |          Interpolated values. Shape is determined by replacing\n",
      "     |          the interpolation axis in the original array with the shape of x.\n",
      "     |      \n",
      "     |      Notes\n",
      "     |      -----\n",
      "     |      Derivatives are evaluated piecewise for each polynomial\n",
      "     |      segment, even if the polynomial is not differentiable at the\n",
      "     |      breakpoints. The polynomial intervals are considered half-open,\n",
      "     |      ``[a, b)``, except for the last interval which is closed\n",
      "     |      ``[a, b]``.\n",
      "     |  \n",
      "     |  extend(self, c, x, right=None)\n",
      "     |      Add additional breakpoints and coefficients to the polynomial.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      c : ndarray, size (k, m, ...)\n",
      "     |          Additional coefficients for polynomials in intervals. Note that\n",
      "     |          the first additional interval will be formed using one of the\n",
      "     |          ``self.x`` end points.\n",
      "     |      x : ndarray, size (m,)\n",
      "     |          Additional breakpoints. Must be sorted in the same order as\n",
      "     |          ``self.x`` and either to the right or to the left of the current\n",
      "     |          breakpoints.\n",
      "     |      right\n",
      "     |          Deprecated argument. Has no effect.\n",
      "     |      \n",
      "     |          .. deprecated:: 0.19\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Class methods inherited from scipy.interpolate.interpolate._PPolyBase:\n",
      "     |  \n",
      "     |  construct_fast(c, x, extrapolate=None, axis=0) from builtins.type\n",
      "     |      Construct the piecewise polynomial without making checks.\n",
      "     |      \n",
      "     |      Takes the same parameters as the constructor. Input arguments\n",
      "     |      ``c`` and ``x`` must be arrays of the correct shape and type. The\n",
      "     |      ``c`` array can only be of dtypes float and complex, and ``x``\n",
      "     |      array must have dtype float.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from scipy.interpolate.interpolate._PPolyBase:\n",
      "     |  \n",
      "     |  axis\n",
      "     |  \n",
      "     |  c\n",
      "     |  \n",
      "     |  extrapolate\n",
      "     |  \n",
      "     |  x\n",
      "    \n",
      "    class RBFInterpolator(builtins.object)\n",
      "     |  RBFInterpolator(y, d, neighbors=None, smoothing=0.0, kernel='thin_plate_spline', epsilon=None, degree=None)\n",
      "     |  \n",
      "     |  Radial basis function (RBF) interpolation in N dimensions.\n",
      "     |  \n",
      "     |  Parameters\n",
      "     |  ----------\n",
      "     |  y : (P, N) array_like\n",
      "     |      Data point coordinates.\n",
      "     |  d : (P, ...) array_like\n",
      "     |      Data values at `y`.\n",
      "     |  neighbors : int, optional\n",
      "     |      If specified, the value of the interpolant at each evaluation point\n",
      "     |      will be computed using only this many nearest data points. All the data\n",
      "     |      points are used by default.\n",
      "     |  smoothing : float or (P,) array_like, optional\n",
      "     |      Smoothing parameter. The interpolant perfectly fits the data when this\n",
      "     |      is set to 0. For large values, the interpolant approaches a least\n",
      "     |      squares fit of a polynomial with the specified degree. Default is 0.\n",
      "     |  kernel : str, optional\n",
      "     |      Type of RBF. This should be one of\n",
      "     |  \n",
      "     |          - 'linear'               : ``-r``\n",
      "     |          - 'thin_plate_spline'    : ``r**2 * log(r)``\n",
      "     |          - 'cubic'                : ``r**3``\n",
      "     |          - 'quintic'              : ``-r**5``\n",
      "     |          - 'multiquadric'         : ``-sqrt(1 + r**2)``\n",
      "     |          - 'inverse_multiquadric' : ``1/sqrt(1 + r**2)``\n",
      "     |          - 'inverse_quadratic'    : ``1/(1 + r**2)``\n",
      "     |          - 'gaussian'             : ``exp(-r**2)``\n",
      "     |  \n",
      "     |      Default is 'thin_plate_spline'.\n",
      "     |  epsilon : float, optional\n",
      "     |      Shape parameter that scales the input to the RBF. If `kernel` is\n",
      "     |      'linear', 'thin_plate_spline', 'cubic', or 'quintic', this defaults to\n",
      "     |      1 and can be ignored because it has the same effect as scaling the\n",
      "     |      smoothing parameter. Otherwise, this must be specified.\n",
      "     |  degree : int, optional\n",
      "     |      Degree of the added polynomial. For some RBFs the interpolant may not\n",
      "     |      be well-posed if the polynomial degree is too small. Those RBFs and\n",
      "     |      their corresponding minimum degrees are\n",
      "     |  \n",
      "     |          - 'multiquadric'      : 0\n",
      "     |          - 'linear'            : 0\n",
      "     |          - 'thin_plate_spline' : 1\n",
      "     |          - 'cubic'             : 1\n",
      "     |          - 'quintic'           : 2\n",
      "     |  \n",
      "     |      The default value is the minimum degree for `kernel` or 0 if there is\n",
      "     |      no minimum degree. Set this to -1 for no added polynomial.\n",
      "     |  \n",
      "     |  Notes\n",
      "     |  -----\n",
      "     |  An RBF is a scalar valued function in N-dimensional space whose value at\n",
      "     |  :math:`x` can be expressed in terms of :math:`r=||x - c||`, where :math:`c`\n",
      "     |  is the center of the RBF.\n",
      "     |  \n",
      "     |  An RBF interpolant for the vector of data values :math:`d`, which are from\n",
      "     |  locations :math:`y`, is a linear combination of RBFs centered at :math:`y`\n",
      "     |  plus a polynomial with a specified degree. The RBF interpolant is written\n",
      "     |  as\n",
      "     |  \n",
      "     |  .. math::\n",
      "     |      f(x) = K(x, y) a + P(x) b,\n",
      "     |  \n",
      "     |  where :math:`K(x, y)` is a matrix of RBFs with centers at :math:`y`\n",
      "     |  evaluated at the points :math:`x`, and :math:`P(x)` is a matrix of\n",
      "     |  monomials, which span polynomials with the specified degree, evaluated at\n",
      "     |  :math:`x`. The coefficients :math:`a` and :math:`b` are the solution to the\n",
      "     |  linear equations\n",
      "     |  \n",
      "     |  .. math::\n",
      "     |      (K(y, y) + \\lambda I) a + P(y) b = d\n",
      "     |  \n",
      "     |  and\n",
      "     |  \n",
      "     |  .. math::\n",
      "     |      P(y)^T a = 0,\n",
      "     |  \n",
      "     |  where :math:`\\lambda` is a non-negative smoothing parameter that controls\n",
      "     |  how well we want to fit the data. The data are fit exactly when the\n",
      "     |  smoothing parameter is 0.\n",
      "     |  \n",
      "     |  The above system is uniquely solvable if the following requirements are\n",
      "     |  met:\n",
      "     |  \n",
      "     |      - :math:`P(y)` must have full column rank. :math:`P(y)` always has full\n",
      "     |        column rank when `degree` is -1 or 0. When `degree` is 1,\n",
      "     |        :math:`P(y)` has full column rank if the data point locations are not\n",
      "     |        all collinear (N=2), coplanar (N=3), etc.\n",
      "     |      - If `kernel` is 'multiquadric', 'linear', 'thin_plate_spline',\n",
      "     |        'cubic', or 'quintic', then `degree` must not be lower than the\n",
      "     |        minimum value listed above.\n",
      "     |      - If `smoothing` is 0, then each data point location must be distinct.\n",
      "     |  \n",
      "     |  When using an RBF that is not scale invariant ('multiquadric',\n",
      "     |  'inverse_multiquadric', 'inverse_quadratic', or 'gaussian'), an appropriate\n",
      "     |  shape parameter must be chosen (e.g., through cross validation). Smaller\n",
      "     |  values for the shape parameter correspond to wider RBFs. The problem can\n",
      "     |  become ill-conditioned or singular when the shape parameter is too small.\n",
      "     |  \n",
      "     |  The memory required to solve for the RBF interpolation coefficients\n",
      "     |  increases quadratically with the number of data points, which can become\n",
      "     |  impractical when interpolating more than about a thousand data points.\n",
      "     |  To overcome memory limitations for large interpolation problems, the\n",
      "     |  `neighbors` argument can be specified to compute an RBF interpolant for\n",
      "     |  each evaluation point using only the nearest data points.\n",
      "     |  \n",
      "     |  .. versionadded:: 1.7.0\n",
      "     |  \n",
      "     |  See Also\n",
      "     |  --------\n",
      "     |  NearestNDInterpolator\n",
      "     |  LinearNDInterpolator\n",
      "     |  CloughTocher2DInterpolator\n",
      "     |  \n",
      "     |  References\n",
      "     |  ----------\n",
      "     |  .. [1] Fasshauer, G., 2007. Meshfree Approximation Methods with Matlab.\n",
      "     |      World Scientific Publishing Co.\n",
      "     |  \n",
      "     |  .. [2] http://amadeus.math.iit.edu/~fass/603_ch3.pdf\n",
      "     |  \n",
      "     |  .. [3] Wahba, G., 1990. Spline Models for Observational Data. SIAM.\n",
      "     |  \n",
      "     |  .. [4] http://pages.stat.wisc.edu/~wahba/stat860public/lect/lect8/lect8.pdf\n",
      "     |  \n",
      "     |  Examples\n",
      "     |  --------\n",
      "     |  Demonstrate interpolating scattered data to a grid in 2-D.\n",
      "     |  \n",
      "     |  >>> import matplotlib.pyplot as plt\n",
      "     |  >>> from scipy.interpolate import RBFInterpolator\n",
      "     |  >>> from scipy.stats.qmc import Halton\n",
      "     |  \n",
      "     |  >>> rng = np.random.default_rng()\n",
      "     |  >>> xobs = 2*Halton(2, seed=rng).random(100) - 1\n",
      "     |  >>> yobs = np.sum(xobs, axis=1)*np.exp(-6*np.sum(xobs**2, axis=1))\n",
      "     |  \n",
      "     |  >>> xgrid = np.mgrid[-1:1:50j, -1:1:50j]\n",
      "     |  >>> xflat = xgrid.reshape(2, -1).T\n",
      "     |  >>> yflat = RBFInterpolator(xobs, yobs)(xflat)\n",
      "     |  >>> ygrid = yflat.reshape(50, 50)\n",
      "     |  \n",
      "     |  >>> fig, ax = plt.subplots()\n",
      "     |  >>> ax.pcolormesh(*xgrid, ygrid, vmin=-0.25, vmax=0.25, shading='gouraud')\n",
      "     |  >>> p = ax.scatter(*xobs.T, c=yobs, s=50, ec='k', vmin=-0.25, vmax=0.25)\n",
      "     |  >>> fig.colorbar(p)\n",
      "     |  >>> plt.show()\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __call__(self, x)\n",
      "     |      Evaluate the interpolant at `x`.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      x : (Q, N) array_like\n",
      "     |          Evaluation point coordinates.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      (Q, ...) ndarray\n",
      "     |          Values of the interpolant at `x`.\n",
      "     |  \n",
      "     |  __init__(self, y, d, neighbors=None, smoothing=0.0, kernel='thin_plate_spline', epsilon=None, degree=None)\n",
      "     |      Initialize self.  See help(type(self)) for accurate signature.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors defined here:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |      dictionary for instance variables (if defined)\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "    \n",
      "    class Rbf(builtins.object)\n",
      "     |  Rbf(*args, **kwargs)\n",
      "     |  \n",
      "     |  Rbf(*args)\n",
      "     |  \n",
      "     |  A class for radial basis function interpolation of functions from\n",
      "     |  N-D scattered data to an M-D domain.\n",
      "     |  \n",
      "     |  .. note::\n",
      "     |      `Rbf` is legacy code, for new usage please use `RBFInterpolator`\n",
      "     |      instead.\n",
      "     |  \n",
      "     |  Parameters\n",
      "     |  ----------\n",
      "     |  *args : arrays\n",
      "     |      x, y, z, ..., d, where x, y, z, ... are the coordinates of the nodes\n",
      "     |      and d is the array of values at the nodes\n",
      "     |  function : str or callable, optional\n",
      "     |      The radial basis function, based on the radius, r, given by the norm\n",
      "     |      (default is Euclidean distance); the default is 'multiquadric'::\n",
      "     |  \n",
      "     |          'multiquadric': sqrt((r/self.epsilon)**2 + 1)\n",
      "     |          'inverse': 1.0/sqrt((r/self.epsilon)**2 + 1)\n",
      "     |          'gaussian': exp(-(r/self.epsilon)**2)\n",
      "     |          'linear': r\n",
      "     |          'cubic': r**3\n",
      "     |          'quintic': r**5\n",
      "     |          'thin_plate': r**2 * log(r)\n",
      "     |  \n",
      "     |      If callable, then it must take 2 arguments (self, r). The epsilon\n",
      "     |      parameter will be available as self.epsilon. Other keyword\n",
      "     |      arguments passed in will be available as well.\n",
      "     |  \n",
      "     |  epsilon : float, optional\n",
      "     |      Adjustable constant for gaussian or multiquadrics functions\n",
      "     |      - defaults to approximate average distance between nodes (which is\n",
      "     |      a good start).\n",
      "     |  smooth : float, optional\n",
      "     |      Values greater than zero increase the smoothness of the\n",
      "     |      approximation. 0 is for interpolation (default), the function will\n",
      "     |      always go through the nodal points in this case.\n",
      "     |  norm : str, callable, optional\n",
      "     |      A function that returns the 'distance' between two points, with\n",
      "     |      inputs as arrays of positions (x, y, z, ...), and an output as an\n",
      "     |      array of distance. E.g., the default: 'euclidean', such that the result\n",
      "     |      is a matrix of the distances from each point in ``x1`` to each point in\n",
      "     |      ``x2``. For more options, see documentation of\n",
      "     |      `scipy.spatial.distances.cdist`.\n",
      "     |  mode : str, optional\n",
      "     |      Mode of the interpolation, can be '1-D' (default) or 'N-D'. When it is\n",
      "     |      '1-D' the data `d` will be considered as 1-D and flattened\n",
      "     |      internally. When it is 'N-D' the data `d` is assumed to be an array of\n",
      "     |      shape (n_samples, m), where m is the dimension of the target domain.\n",
      "     |  \n",
      "     |  \n",
      "     |  Attributes\n",
      "     |  ----------\n",
      "     |  N : int\n",
      "     |      The number of data points (as determined by the input arrays).\n",
      "     |  di : ndarray\n",
      "     |      The 1-D array of data values at each of the data coordinates `xi`.\n",
      "     |  xi : ndarray\n",
      "     |      The 2-D array of data coordinates.\n",
      "     |  function : str or callable\n",
      "     |      The radial basis function. See description under Parameters.\n",
      "     |  epsilon : float\n",
      "     |      Parameter used by gaussian or multiquadrics functions. See Parameters.\n",
      "     |  smooth : float\n",
      "     |      Smoothing parameter. See description under Parameters.\n",
      "     |  norm : str or callable\n",
      "     |      The distance function. See description under Parameters.\n",
      "     |  mode : str\n",
      "     |      Mode of the interpolation. See description under Parameters.\n",
      "     |  nodes : ndarray\n",
      "     |      A 1-D array of node values for the interpolation.\n",
      "     |  A : internal property, do not use\n",
      "     |  \n",
      "     |  See Also\n",
      "     |  --------\n",
      "     |  RBFInterpolator\n",
      "     |  \n",
      "     |  Examples\n",
      "     |  --------\n",
      "     |  >>> from scipy.interpolate import Rbf\n",
      "     |  >>> rng = np.random.default_rng()\n",
      "     |  >>> x, y, z, d = rng.random((4, 50))\n",
      "     |  >>> rbfi = Rbf(x, y, z, d)  # radial basis function interpolator instance\n",
      "     |  >>> xi = yi = zi = np.linspace(0, 1, 20)\n",
      "     |  >>> di = rbfi(xi, yi, zi)   # interpolated values\n",
      "     |  >>> di.shape\n",
      "     |  (20,)\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __call__(self, *args)\n",
      "     |      Call self as a function.\n",
      "     |  \n",
      "     |  __init__(self, *args, **kwargs)\n",
      "     |      Initialize self.  See help(type(self)) for accurate signature.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Readonly properties defined here:\n",
      "     |  \n",
      "     |  A\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors defined here:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |      dictionary for instance variables (if defined)\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "    \n",
      "    class RectBivariateSpline(BivariateSpline)\n",
      "     |  RectBivariateSpline(x, y, z, bbox=[None, None, None, None], kx=3, ky=3, s=0)\n",
      "     |  \n",
      "     |  Bivariate spline approximation over a rectangular mesh.\n",
      "     |  \n",
      "     |  Can be used for both smoothing and interpolating data.\n",
      "     |  \n",
      "     |  Parameters\n",
      "     |  ----------\n",
      "     |  x,y : array_like\n",
      "     |      1-D arrays of coordinates in strictly ascending order.\n",
      "     |  z : array_like\n",
      "     |      2-D array of data with shape (x.size,y.size).\n",
      "     |  bbox : array_like, optional\n",
      "     |      Sequence of length 4 specifying the boundary of the rectangular\n",
      "     |      approximation domain.  By default,\n",
      "     |      ``bbox=[min(x), max(x), min(y), max(y)]``.\n",
      "     |  kx, ky : ints, optional\n",
      "     |      Degrees of the bivariate spline. Default is 3.\n",
      "     |  s : float, optional\n",
      "     |      Positive smoothing factor defined for estimation condition:\n",
      "     |      ``sum((z[i]-f(x[i], y[i]))**2, axis=0) <= s`` where f is a spline\n",
      "     |      function. Default is ``s=0``, which is for interpolation.\n",
      "     |  \n",
      "     |  See Also\n",
      "     |  --------\n",
      "     |  BivariateSpline :\n",
      "     |      a base class for bivariate splines.\n",
      "     |  UnivariateSpline :\n",
      "     |      a smooth univariate spline to fit a given set of data points.\n",
      "     |  SmoothBivariateSpline :\n",
      "     |      a smoothing bivariate spline through the given points\n",
      "     |  LSQBivariateSpline :\n",
      "     |      a bivariate spline using weighted least-squares fitting\n",
      "     |  RectSphereBivariateSpline :\n",
      "     |      a bivariate spline over a rectangular mesh on a sphere\n",
      "     |  SmoothSphereBivariateSpline :\n",
      "     |      a smoothing bivariate spline in spherical coordinates\n",
      "     |  LSQSphereBivariateSpline :\n",
      "     |      a bivariate spline in spherical coordinates using weighted\n",
      "     |      least-squares fitting\n",
      "     |  bisplrep :\n",
      "     |      a function to find a bivariate B-spline representation of a surface\n",
      "     |  bisplev :\n",
      "     |      a function to evaluate a bivariate B-spline and its derivatives\n",
      "     |  \n",
      "     |  Method resolution order:\n",
      "     |      RectBivariateSpline\n",
      "     |      BivariateSpline\n",
      "     |      _BivariateSplineBase\n",
      "     |      builtins.object\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __init__(self, x, y, z, bbox=[None, None, None, None], kx=3, ky=3, s=0)\n",
      "     |      Initialize self.  See help(type(self)) for accurate signature.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from BivariateSpline:\n",
      "     |  \n",
      "     |  ev(self, xi, yi, dx=0, dy=0)\n",
      "     |      Evaluate the spline at points\n",
      "     |      \n",
      "     |      Returns the interpolated value at ``(xi[i], yi[i]),\n",
      "     |      i=0,...,len(xi)-1``.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      xi, yi : array_like\n",
      "     |          Input coordinates. Standard Numpy broadcasting is obeyed.\n",
      "     |      dx : int, optional\n",
      "     |          Order of x-derivative\n",
      "     |      \n",
      "     |          .. versionadded:: 0.14.0\n",
      "     |      dy : int, optional\n",
      "     |          Order of y-derivative\n",
      "     |      \n",
      "     |          .. versionadded:: 0.14.0\n",
      "     |  \n",
      "     |  integral(self, xa, xb, ya, yb)\n",
      "     |      Evaluate the integral of the spline over area [xa,xb] x [ya,yb].\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      xa, xb : float\n",
      "     |          The end-points of the x integration interval.\n",
      "     |      ya, yb : float\n",
      "     |          The end-points of the y integration interval.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      integ : float\n",
      "     |          The value of the resulting integral.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from _BivariateSplineBase:\n",
      "     |  \n",
      "     |  __call__(self, x, y, dx=0, dy=0, grid=True)\n",
      "     |      Evaluate the spline or its derivatives at given positions.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      x, y : array_like\n",
      "     |          Input coordinates.\n",
      "     |      \n",
      "     |          If `grid` is False, evaluate the spline at points ``(x[i],\n",
      "     |          y[i]), i=0, ..., len(x)-1``.  Standard Numpy broadcasting\n",
      "     |          is obeyed.\n",
      "     |      \n",
      "     |          If `grid` is True: evaluate spline at the grid points\n",
      "     |          defined by the coordinate arrays x, y. The arrays must be\n",
      "     |          sorted to increasing order.\n",
      "     |      \n",
      "     |          Note that the axis ordering is inverted relative to\n",
      "     |          the output of meshgrid.\n",
      "     |      dx : int\n",
      "     |          Order of x-derivative\n",
      "     |      \n",
      "     |          .. versionadded:: 0.14.0\n",
      "     |      dy : int\n",
      "     |          Order of y-derivative\n",
      "     |      \n",
      "     |          .. versionadded:: 0.14.0\n",
      "     |      grid : bool\n",
      "     |          Whether to evaluate the results on a grid spanned by the\n",
      "     |          input arrays, or at points specified by the input arrays.\n",
      "     |      \n",
      "     |          .. versionadded:: 0.14.0\n",
      "     |  \n",
      "     |  get_coeffs(self)\n",
      "     |      Return spline coefficients.\n",
      "     |  \n",
      "     |  get_knots(self)\n",
      "     |      Return a tuple (tx,ty) where tx,ty contain knots positions\n",
      "     |      of the spline with respect to x-, y-variable, respectively.\n",
      "     |      The position of interior and additional knots are given as\n",
      "     |      t[k+1:-k-1] and t[:k+1]=b, t[-k-1:]=e, respectively.\n",
      "     |  \n",
      "     |  get_residual(self)\n",
      "     |      Return weighted sum of squared residuals of the spline\n",
      "     |      approximation: sum ((w[i]*(z[i]-s(x[i],y[i])))**2,axis=0)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from _BivariateSplineBase:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |      dictionary for instance variables (if defined)\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "    \n",
      "    class RectSphereBivariateSpline(SphereBivariateSpline)\n",
      "     |  RectSphereBivariateSpline(u, v, r, s=0.0, pole_continuity=False, pole_values=None, pole_exact=False, pole_flat=False)\n",
      "     |  \n",
      "     |  Bivariate spline approximation over a rectangular mesh on a sphere.\n",
      "     |  \n",
      "     |  Can be used for smoothing data.\n",
      "     |  \n",
      "     |  .. versionadded:: 0.11.0\n",
      "     |  \n",
      "     |  Parameters\n",
      "     |  ----------\n",
      "     |  u : array_like\n",
      "     |      1-D array of colatitude coordinates in strictly ascending order.\n",
      "     |      Coordinates must be given in radians and lie within the open interval\n",
      "     |      ``(0, pi)``.\n",
      "     |  v : array_like\n",
      "     |      1-D array of longitude coordinates in strictly ascending order.\n",
      "     |      Coordinates must be given in radians. First element (``v[0]``) must lie\n",
      "     |      within the interval ``[-pi, pi)``. Last element (``v[-1]``) must satisfy\n",
      "     |      ``v[-1] <= v[0] + 2*pi``.\n",
      "     |  r : array_like\n",
      "     |      2-D array of data with shape ``(u.size, v.size)``.\n",
      "     |  s : float, optional\n",
      "     |      Positive smoothing factor defined for estimation condition\n",
      "     |      (``s=0`` is for interpolation).\n",
      "     |  pole_continuity : bool or (bool, bool), optional\n",
      "     |      Order of continuity at the poles ``u=0`` (``pole_continuity[0]``) and\n",
      "     |      ``u=pi`` (``pole_continuity[1]``).  The order of continuity at the pole\n",
      "     |      will be 1 or 0 when this is True or False, respectively.\n",
      "     |      Defaults to False.\n",
      "     |  pole_values : float or (float, float), optional\n",
      "     |      Data values at the poles ``u=0`` and ``u=pi``.  Either the whole\n",
      "     |      parameter or each individual element can be None.  Defaults to None.\n",
      "     |  pole_exact : bool or (bool, bool), optional\n",
      "     |      Data value exactness at the poles ``u=0`` and ``u=pi``.  If True, the\n",
      "     |      value is considered to be the right function value, and it will be\n",
      "     |      fitted exactly. If False, the value will be considered to be a data\n",
      "     |      value just like the other data values.  Defaults to False.\n",
      "     |  pole_flat : bool or (bool, bool), optional\n",
      "     |      For the poles at ``u=0`` and ``u=pi``, specify whether or not the\n",
      "     |      approximation has vanishing derivatives.  Defaults to False.\n",
      "     |  \n",
      "     |  See Also\n",
      "     |  --------\n",
      "     |  BivariateSpline :\n",
      "     |      a base class for bivariate splines.\n",
      "     |  UnivariateSpline :\n",
      "     |      a smooth univariate spline to fit a given set of data points.\n",
      "     |  SmoothBivariateSpline :\n",
      "     |      a smoothing bivariate spline through the given points\n",
      "     |  LSQBivariateSpline :\n",
      "     |      a bivariate spline using weighted least-squares fitting\n",
      "     |  SmoothSphereBivariateSpline :\n",
      "     |      a smoothing bivariate spline in spherical coordinates\n",
      "     |  LSQSphereBivariateSpline :\n",
      "     |      a bivariate spline in spherical coordinates using weighted\n",
      "     |      least-squares fitting\n",
      "     |  RectBivariateSpline :\n",
      "     |      a bivariate spline over a rectangular mesh.\n",
      "     |  bisplrep :\n",
      "     |      a function to find a bivariate B-spline representation of a surface\n",
      "     |  bisplev :\n",
      "     |      a function to evaluate a bivariate B-spline and its derivatives\n",
      "     |  \n",
      "     |  Notes\n",
      "     |  -----\n",
      "     |  Currently, only the smoothing spline approximation (``iopt[0] = 0`` and\n",
      "     |  ``iopt[0] = 1`` in the FITPACK routine) is supported.  The exact\n",
      "     |  least-squares spline approximation is not implemented yet.\n",
      "     |  \n",
      "     |  When actually performing the interpolation, the requested `v` values must\n",
      "     |  lie within the same length 2pi interval that the original `v` values were\n",
      "     |  chosen from.\n",
      "     |  \n",
      "     |  For more information, see the FITPACK_ site about this function.\n",
      "     |  \n",
      "     |  .. _FITPACK: http://www.netlib.org/dierckx/spgrid.f\n",
      "     |  \n",
      "     |  Examples\n",
      "     |  --------\n",
      "     |  Suppose we have global data on a coarse grid\n",
      "     |  \n",
      "     |  >>> lats = np.linspace(10, 170, 9) * np.pi / 180.\n",
      "     |  >>> lons = np.linspace(0, 350, 18) * np.pi / 180.\n",
      "     |  >>> data = np.dot(np.atleast_2d(90. - np.linspace(-80., 80., 18)).T,\n",
      "     |  ...               np.atleast_2d(180. - np.abs(np.linspace(0., 350., 9)))).T\n",
      "     |  \n",
      "     |  We want to interpolate it to a global one-degree grid\n",
      "     |  \n",
      "     |  >>> new_lats = np.linspace(1, 180, 180) * np.pi / 180\n",
      "     |  >>> new_lons = np.linspace(1, 360, 360) * np.pi / 180\n",
      "     |  >>> new_lats, new_lons = np.meshgrid(new_lats, new_lons)\n",
      "     |  \n",
      "     |  We need to set up the interpolator object\n",
      "     |  \n",
      "     |  >>> from scipy.interpolate import RectSphereBivariateSpline\n",
      "     |  >>> lut = RectSphereBivariateSpline(lats, lons, data)\n",
      "     |  \n",
      "     |  Finally we interpolate the data.  The `RectSphereBivariateSpline` object\n",
      "     |  only takes 1-D arrays as input, therefore we need to do some reshaping.\n",
      "     |  \n",
      "     |  >>> data_interp = lut.ev(new_lats.ravel(),\n",
      "     |  ...                      new_lons.ravel()).reshape((360, 180)).T\n",
      "     |  \n",
      "     |  Looking at the original and the interpolated data, one can see that the\n",
      "     |  interpolant reproduces the original data very well:\n",
      "     |  \n",
      "     |  >>> import matplotlib.pyplot as plt\n",
      "     |  >>> fig = plt.figure()\n",
      "     |  >>> ax1 = fig.add_subplot(211)\n",
      "     |  >>> ax1.imshow(data, interpolation='nearest')\n",
      "     |  >>> ax2 = fig.add_subplot(212)\n",
      "     |  >>> ax2.imshow(data_interp, interpolation='nearest')\n",
      "     |  >>> plt.show()\n",
      "     |  \n",
      "     |  Choosing the optimal value of ``s`` can be a delicate task. Recommended\n",
      "     |  values for ``s`` depend on the accuracy of the data values.  If the user\n",
      "     |  has an idea of the statistical errors on the data, she can also find a\n",
      "     |  proper estimate for ``s``. By assuming that, if she specifies the\n",
      "     |  right ``s``, the interpolator will use a spline ``f(u,v)`` which exactly\n",
      "     |  reproduces the function underlying the data, she can evaluate\n",
      "     |  ``sum((r(i,j)-s(u(i),v(j)))**2)`` to find a good estimate for this ``s``.\n",
      "     |  For example, if she knows that the statistical errors on her\n",
      "     |  ``r(i,j)``-values are not greater than 0.1, she may expect that a good\n",
      "     |  ``s`` should have a value not larger than ``u.size * v.size * (0.1)**2``.\n",
      "     |  \n",
      "     |  If nothing is known about the statistical error in ``r(i,j)``, ``s`` must\n",
      "     |  be determined by trial and error.  The best is then to start with a very\n",
      "     |  large value of ``s`` (to determine the least-squares polynomial and the\n",
      "     |  corresponding upper bound ``fp0`` for ``s``) and then to progressively\n",
      "     |  decrease the value of ``s`` (say by a factor 10 in the beginning, i.e.\n",
      "     |  ``s = fp0 / 10, fp0 / 100, ...``  and more carefully as the approximation\n",
      "     |  shows more detail) to obtain closer fits.\n",
      "     |  \n",
      "     |  The interpolation results for different values of ``s`` give some insight\n",
      "     |  into this process:\n",
      "     |  \n",
      "     |  >>> fig2 = plt.figure()\n",
      "     |  >>> s = [3e9, 2e9, 1e9, 1e8]\n",
      "     |  >>> for ii in range(len(s)):\n",
      "     |  ...     lut = RectSphereBivariateSpline(lats, lons, data, s=s[ii])\n",
      "     |  ...     data_interp = lut.ev(new_lats.ravel(),\n",
      "     |  ...                          new_lons.ravel()).reshape((360, 180)).T\n",
      "     |  ...     ax = fig2.add_subplot(2, 2, ii+1)\n",
      "     |  ...     ax.imshow(data_interp, interpolation='nearest')\n",
      "     |  ...     ax.set_title(\"s = %g\" % s[ii])\n",
      "     |  >>> plt.show()\n",
      "     |  \n",
      "     |  Method resolution order:\n",
      "     |      RectSphereBivariateSpline\n",
      "     |      SphereBivariateSpline\n",
      "     |      _BivariateSplineBase\n",
      "     |      builtins.object\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __call__(self, theta, phi, dtheta=0, dphi=0, grid=True)\n",
      "     |      Evaluate the spline or its derivatives at given positions.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      theta, phi : array_like\n",
      "     |          Input coordinates.\n",
      "     |      \n",
      "     |          If `grid` is False, evaluate the spline at points\n",
      "     |          ``(theta[i], phi[i]), i=0, ..., len(x)-1``.  Standard\n",
      "     |          Numpy broadcasting is obeyed.\n",
      "     |      \n",
      "     |          If `grid` is True: evaluate spline at the grid points\n",
      "     |          defined by the coordinate arrays theta, phi. The arrays\n",
      "     |          must be sorted to increasing order.\n",
      "     |      dtheta : int, optional\n",
      "     |          Order of theta-derivative\n",
      "     |      \n",
      "     |          .. versionadded:: 0.14.0\n",
      "     |      dphi : int\n",
      "     |          Order of phi-derivative\n",
      "     |      \n",
      "     |          .. versionadded:: 0.14.0\n",
      "     |      grid : bool\n",
      "     |          Whether to evaluate the results on a grid spanned by the\n",
      "     |          input arrays, or at points specified by the input arrays.\n",
      "     |      \n",
      "     |          .. versionadded:: 0.14.0\n",
      "     |  \n",
      "     |  __init__(self, u, v, r, s=0.0, pole_continuity=False, pole_values=None, pole_exact=False, pole_flat=False)\n",
      "     |      Initialize self.  See help(type(self)) for accurate signature.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from SphereBivariateSpline:\n",
      "     |  \n",
      "     |  ev(self, theta, phi, dtheta=0, dphi=0)\n",
      "     |      Evaluate the spline at points\n",
      "     |      \n",
      "     |      Returns the interpolated value at ``(theta[i], phi[i]),\n",
      "     |      i=0,...,len(theta)-1``.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      theta, phi : array_like\n",
      "     |          Input coordinates. Standard Numpy broadcasting is obeyed.\n",
      "     |      dtheta : int, optional\n",
      "     |          Order of theta-derivative\n",
      "     |      \n",
      "     |          .. versionadded:: 0.14.0\n",
      "     |      dphi : int, optional\n",
      "     |          Order of phi-derivative\n",
      "     |      \n",
      "     |          .. versionadded:: 0.14.0\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from _BivariateSplineBase:\n",
      "     |  \n",
      "     |  get_coeffs(self)\n",
      "     |      Return spline coefficients.\n",
      "     |  \n",
      "     |  get_knots(self)\n",
      "     |      Return a tuple (tx,ty) where tx,ty contain knots positions\n",
      "     |      of the spline with respect to x-, y-variable, respectively.\n",
      "     |      The position of interior and additional knots are given as\n",
      "     |      t[k+1:-k-1] and t[:k+1]=b, t[-k-1:]=e, respectively.\n",
      "     |  \n",
      "     |  get_residual(self)\n",
      "     |      Return weighted sum of squared residuals of the spline\n",
      "     |      approximation: sum ((w[i]*(z[i]-s(x[i],y[i])))**2,axis=0)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from _BivariateSplineBase:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |      dictionary for instance variables (if defined)\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "    \n",
      "    class RegularGridInterpolator(builtins.object)\n",
      "     |  RegularGridInterpolator(points, values, method='linear', bounds_error=True, fill_value=nan)\n",
      "     |  \n",
      "     |  Interpolation on a regular grid in arbitrary dimensions\n",
      "     |  \n",
      "     |  The data must be defined on a regular grid; the grid spacing however may be\n",
      "     |  uneven. Linear and nearest-neighbor interpolation are supported. After\n",
      "     |  setting up the interpolator object, the interpolation method (*linear* or\n",
      "     |  *nearest*) may be chosen at each evaluation.\n",
      "     |  \n",
      "     |  Parameters\n",
      "     |  ----------\n",
      "     |  points : tuple of ndarray of float, with shapes (m1, ), ..., (mn, )\n",
      "     |      The points defining the regular grid in n dimensions.\n",
      "     |  \n",
      "     |  values : array_like, shape (m1, ..., mn, ...)\n",
      "     |      The data on the regular grid in n dimensions.\n",
      "     |  \n",
      "     |  method : str, optional\n",
      "     |      The method of interpolation to perform. Supported are \"linear\" and\n",
      "     |      \"nearest\". This parameter will become the default for the object's\n",
      "     |      ``__call__`` method. Default is \"linear\".\n",
      "     |  \n",
      "     |  bounds_error : bool, optional\n",
      "     |      If True, when interpolated values are requested outside of the\n",
      "     |      domain of the input data, a ValueError is raised.\n",
      "     |      If False, then `fill_value` is used.\n",
      "     |  \n",
      "     |  fill_value : number, optional\n",
      "     |      If provided, the value to use for points outside of the\n",
      "     |      interpolation domain. If None, values outside\n",
      "     |      the domain are extrapolated.\n",
      "     |  \n",
      "     |  Methods\n",
      "     |  -------\n",
      "     |  __call__\n",
      "     |  \n",
      "     |  Notes\n",
      "     |  -----\n",
      "     |  Contrary to LinearNDInterpolator and NearestNDInterpolator, this class\n",
      "     |  avoids expensive triangulation of the input data by taking advantage of the\n",
      "     |  regular grid structure.\n",
      "     |  \n",
      "     |  If any of `points` have a dimension of size 1, linear interpolation will\n",
      "     |  return an array of `nan` values. Nearest-neighbor interpolation will work\n",
      "     |  as usual in this case.\n",
      "     |  \n",
      "     |  .. versionadded:: 0.14\n",
      "     |  \n",
      "     |  Examples\n",
      "     |  --------\n",
      "     |  Evaluate a simple example function on the points of a 3-D grid:\n",
      "     |  \n",
      "     |  >>> from scipy.interpolate import RegularGridInterpolator\n",
      "     |  >>> def f(x, y, z):\n",
      "     |  ...     return 2 * x**3 + 3 * y**2 - z\n",
      "     |  >>> x = np.linspace(1, 4, 11)\n",
      "     |  >>> y = np.linspace(4, 7, 22)\n",
      "     |  >>> z = np.linspace(7, 9, 33)\n",
      "     |  >>> xg, yg ,zg = np.meshgrid(x, y, z, indexing='ij', sparse=True)\n",
      "     |  >>> data = f(xg, yg, zg)\n",
      "     |  \n",
      "     |  ``data`` is now a 3-D array with ``data[i,j,k] = f(x[i], y[j], z[k])``.\n",
      "     |  Next, define an interpolating function from this data:\n",
      "     |  \n",
      "     |  >>> my_interpolating_function = RegularGridInterpolator((x, y, z), data)\n",
      "     |  \n",
      "     |  Evaluate the interpolating function at the two points\n",
      "     |  ``(x,y,z) = (2.1, 6.2, 8.3)`` and ``(3.3, 5.2, 7.1)``:\n",
      "     |  \n",
      "     |  >>> pts = np.array([[2.1, 6.2, 8.3], [3.3, 5.2, 7.1]])\n",
      "     |  >>> my_interpolating_function(pts)\n",
      "     |  array([ 125.80469388,  146.30069388])\n",
      "     |  \n",
      "     |  which is indeed a close approximation to\n",
      "     |  ``[f(2.1, 6.2, 8.3), f(3.3, 5.2, 7.1)]``.\n",
      "     |  \n",
      "     |  See also\n",
      "     |  --------\n",
      "     |  NearestNDInterpolator : Nearest neighbor interpolation on unstructured\n",
      "     |                          data in N dimensions\n",
      "     |  \n",
      "     |  LinearNDInterpolator : Piecewise linear interpolant on unstructured data\n",
      "     |                         in N dimensions\n",
      "     |  \n",
      "     |  References\n",
      "     |  ----------\n",
      "     |  .. [1] Python package *regulargrid* by Johannes Buchner, see\n",
      "     |         https://pypi.python.org/pypi/regulargrid/\n",
      "     |  .. [2] Wikipedia, \"Trilinear interpolation\",\n",
      "     |         https://en.wikipedia.org/wiki/Trilinear_interpolation\n",
      "     |  .. [3] Weiser, Alan, and Sergio E. Zarantonello. \"A note on piecewise linear\n",
      "     |         and multilinear table interpolation in many dimensions.\" MATH.\n",
      "     |         COMPUT. 50.181 (1988): 189-196.\n",
      "     |         https://www.ams.org/journals/mcom/1988-50-181/S0025-5718-1988-0917826-0/S0025-5718-1988-0917826-0.pdf\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __call__(self, xi, method=None)\n",
      "     |      Interpolation at coordinates\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      xi : ndarray of shape (..., ndim)\n",
      "     |          The coordinates to sample the gridded data at\n",
      "     |      \n",
      "     |      method : str\n",
      "     |          The method of interpolation to perform. Supported are \"linear\" and\n",
      "     |          \"nearest\".\n",
      "     |  \n",
      "     |  __init__(self, points, values, method='linear', bounds_error=True, fill_value=nan)\n",
      "     |      Initialize self.  See help(type(self)) for accurate signature.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors defined here:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |      dictionary for instance variables (if defined)\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "    \n",
      "    class SmoothBivariateSpline(BivariateSpline)\n",
      "     |  SmoothBivariateSpline(x, y, z, w=None, bbox=[None, None, None, None], kx=3, ky=3, s=None, eps=1e-16)\n",
      "     |  \n",
      "     |  Smooth bivariate spline approximation.\n",
      "     |  \n",
      "     |  Parameters\n",
      "     |  ----------\n",
      "     |  x, y, z : array_like\n",
      "     |      1-D sequences of data points (order is not important).\n",
      "     |  w : array_like, optional\n",
      "     |      Positive 1-D sequence of weights, of same length as `x`, `y` and `z`.\n",
      "     |  bbox : array_like, optional\n",
      "     |      Sequence of length 4 specifying the boundary of the rectangular\n",
      "     |      approximation domain.  By default,\n",
      "     |      ``bbox=[min(x), max(x), min(y), max(y)]``.\n",
      "     |  kx, ky : ints, optional\n",
      "     |      Degrees of the bivariate spline. Default is 3.\n",
      "     |  s : float, optional\n",
      "     |      Positive smoothing factor defined for estimation condition:\n",
      "     |      ``sum((w[i]*(z[i]-s(x[i], y[i])))**2, axis=0) <= s``\n",
      "     |      Default ``s=len(w)`` which should be a good value if ``1/w[i]`` is an\n",
      "     |      estimate of the standard deviation of ``z[i]``.\n",
      "     |  eps : float, optional\n",
      "     |      A threshold for determining the effective rank of an over-determined\n",
      "     |      linear system of equations. `eps` should have a value within the open\n",
      "     |      interval ``(0, 1)``, the default is 1e-16.\n",
      "     |  \n",
      "     |  See Also\n",
      "     |  --------\n",
      "     |  BivariateSpline :\n",
      "     |      a base class for bivariate splines.\n",
      "     |  UnivariateSpline :\n",
      "     |      a smooth univariate spline to fit a given set of data points.\n",
      "     |  LSQBivariateSpline :\n",
      "     |      a bivariate spline using weighted least-squares fitting\n",
      "     |  RectSphereBivariateSpline :\n",
      "     |      a bivariate spline over a rectangular mesh on a sphere\n",
      "     |  SmoothSphereBivariateSpline :\n",
      "     |      a smoothing bivariate spline in spherical coordinates\n",
      "     |  LSQSphereBivariateSpline :\n",
      "     |      a bivariate spline in spherical coordinates using weighted\n",
      "     |      least-squares fitting\n",
      "     |  RectBivariateSpline :\n",
      "     |      a bivariate spline over a rectangular mesh\n",
      "     |  bisplrep :\n",
      "     |      a function to find a bivariate B-spline representation of a surface\n",
      "     |  bisplev :\n",
      "     |      a function to evaluate a bivariate B-spline and its derivatives\n",
      "     |  \n",
      "     |  Notes\n",
      "     |  -----\n",
      "     |  The length of `x`, `y` and `z` should be at least ``(kx+1) * (ky+1)``.\n",
      "     |  \n",
      "     |  Method resolution order:\n",
      "     |      SmoothBivariateSpline\n",
      "     |      BivariateSpline\n",
      "     |      _BivariateSplineBase\n",
      "     |      builtins.object\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __init__(self, x, y, z, w=None, bbox=[None, None, None, None], kx=3, ky=3, s=None, eps=1e-16)\n",
      "     |      Initialize self.  See help(type(self)) for accurate signature.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from BivariateSpline:\n",
      "     |  \n",
      "     |  ev(self, xi, yi, dx=0, dy=0)\n",
      "     |      Evaluate the spline at points\n",
      "     |      \n",
      "     |      Returns the interpolated value at ``(xi[i], yi[i]),\n",
      "     |      i=0,...,len(xi)-1``.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      xi, yi : array_like\n",
      "     |          Input coordinates. Standard Numpy broadcasting is obeyed.\n",
      "     |      dx : int, optional\n",
      "     |          Order of x-derivative\n",
      "     |      \n",
      "     |          .. versionadded:: 0.14.0\n",
      "     |      dy : int, optional\n",
      "     |          Order of y-derivative\n",
      "     |      \n",
      "     |          .. versionadded:: 0.14.0\n",
      "     |  \n",
      "     |  integral(self, xa, xb, ya, yb)\n",
      "     |      Evaluate the integral of the spline over area [xa,xb] x [ya,yb].\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      xa, xb : float\n",
      "     |          The end-points of the x integration interval.\n",
      "     |      ya, yb : float\n",
      "     |          The end-points of the y integration interval.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      integ : float\n",
      "     |          The value of the resulting integral.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from _BivariateSplineBase:\n",
      "     |  \n",
      "     |  __call__(self, x, y, dx=0, dy=0, grid=True)\n",
      "     |      Evaluate the spline or its derivatives at given positions.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      x, y : array_like\n",
      "     |          Input coordinates.\n",
      "     |      \n",
      "     |          If `grid` is False, evaluate the spline at points ``(x[i],\n",
      "     |          y[i]), i=0, ..., len(x)-1``.  Standard Numpy broadcasting\n",
      "     |          is obeyed.\n",
      "     |      \n",
      "     |          If `grid` is True: evaluate spline at the grid points\n",
      "     |          defined by the coordinate arrays x, y. The arrays must be\n",
      "     |          sorted to increasing order.\n",
      "     |      \n",
      "     |          Note that the axis ordering is inverted relative to\n",
      "     |          the output of meshgrid.\n",
      "     |      dx : int\n",
      "     |          Order of x-derivative\n",
      "     |      \n",
      "     |          .. versionadded:: 0.14.0\n",
      "     |      dy : int\n",
      "     |          Order of y-derivative\n",
      "     |      \n",
      "     |          .. versionadded:: 0.14.0\n",
      "     |      grid : bool\n",
      "     |          Whether to evaluate the results on a grid spanned by the\n",
      "     |          input arrays, or at points specified by the input arrays.\n",
      "     |      \n",
      "     |          .. versionadded:: 0.14.0\n",
      "     |  \n",
      "     |  get_coeffs(self)\n",
      "     |      Return spline coefficients.\n",
      "     |  \n",
      "     |  get_knots(self)\n",
      "     |      Return a tuple (tx,ty) where tx,ty contain knots positions\n",
      "     |      of the spline with respect to x-, y-variable, respectively.\n",
      "     |      The position of interior and additional knots are given as\n",
      "     |      t[k+1:-k-1] and t[:k+1]=b, t[-k-1:]=e, respectively.\n",
      "     |  \n",
      "     |  get_residual(self)\n",
      "     |      Return weighted sum of squared residuals of the spline\n",
      "     |      approximation: sum ((w[i]*(z[i]-s(x[i],y[i])))**2,axis=0)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from _BivariateSplineBase:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |      dictionary for instance variables (if defined)\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "    \n",
      "    class SmoothSphereBivariateSpline(SphereBivariateSpline)\n",
      "     |  SmoothSphereBivariateSpline(theta, phi, r, w=None, s=0.0, eps=1e-16)\n",
      "     |  \n",
      "     |  Smooth bivariate spline approximation in spherical coordinates.\n",
      "     |  \n",
      "     |  .. versionadded:: 0.11.0\n",
      "     |  \n",
      "     |  Parameters\n",
      "     |  ----------\n",
      "     |  theta, phi, r : array_like\n",
      "     |      1-D sequences of data points (order is not important). Coordinates\n",
      "     |      must be given in radians. Theta must lie within the interval\n",
      "     |      ``[0, pi]``, and phi must lie within the interval ``[0, 2pi]``.\n",
      "     |  w : array_like, optional\n",
      "     |      Positive 1-D sequence of weights.\n",
      "     |  s : float, optional\n",
      "     |      Positive smoothing factor defined for estimation condition:\n",
      "     |      ``sum((w(i)*(r(i) - s(theta(i), phi(i))))**2, axis=0) <= s``\n",
      "     |      Default ``s=len(w)`` which should be a good value if ``1/w[i]`` is an\n",
      "     |      estimate of the standard deviation of ``r[i]``.\n",
      "     |  eps : float, optional\n",
      "     |      A threshold for determining the effective rank of an over-determined\n",
      "     |      linear system of equations. `eps` should have a value within the open\n",
      "     |      interval ``(0, 1)``, the default is 1e-16.\n",
      "     |  \n",
      "     |  See Also\n",
      "     |  --------\n",
      "     |  BivariateSpline :\n",
      "     |      a base class for bivariate splines.\n",
      "     |  UnivariateSpline :\n",
      "     |      a smooth univariate spline to fit a given set of data points.\n",
      "     |  SmoothBivariateSpline :\n",
      "     |      a smoothing bivariate spline through the given points\n",
      "     |  LSQBivariateSpline :\n",
      "     |      a bivariate spline using weighted least-squares fitting\n",
      "     |  RectSphereBivariateSpline :\n",
      "     |      a bivariate spline over a rectangular mesh on a sphere\n",
      "     |  LSQSphereBivariateSpline :\n",
      "     |      a bivariate spline in spherical coordinates using weighted\n",
      "     |      least-squares fitting\n",
      "     |  RectBivariateSpline :\n",
      "     |      a bivariate spline over a rectangular mesh.\n",
      "     |  bisplrep :\n",
      "     |      a function to find a bivariate B-spline representation of a surface\n",
      "     |  bisplev :\n",
      "     |      a function to evaluate a bivariate B-spline and its derivatives\n",
      "     |  \n",
      "     |  Notes\n",
      "     |  -----\n",
      "     |  For more information, see the FITPACK_ site about this function.\n",
      "     |  \n",
      "     |  .. _FITPACK: http://www.netlib.org/dierckx/sphere.f\n",
      "     |  \n",
      "     |  Examples\n",
      "     |  --------\n",
      "     |  Suppose we have global data on a coarse grid (the input data does not\n",
      "     |  have to be on a grid):\n",
      "     |  \n",
      "     |  >>> theta = np.linspace(0., np.pi, 7)\n",
      "     |  >>> phi = np.linspace(0., 2*np.pi, 9)\n",
      "     |  >>> data = np.empty((theta.shape[0], phi.shape[0]))\n",
      "     |  >>> data[:,0], data[0,:], data[-1,:] = 0., 0., 0.\n",
      "     |  >>> data[1:-1,1], data[1:-1,-1] = 1., 1.\n",
      "     |  >>> data[1,1:-1], data[-2,1:-1] = 1., 1.\n",
      "     |  >>> data[2:-2,2], data[2:-2,-2] = 2., 2.\n",
      "     |  >>> data[2,2:-2], data[-3,2:-2] = 2., 2.\n",
      "     |  >>> data[3,3:-2] = 3.\n",
      "     |  >>> data = np.roll(data, 4, 1)\n",
      "     |  \n",
      "     |  We need to set up the interpolator object\n",
      "     |  \n",
      "     |  >>> lats, lons = np.meshgrid(theta, phi)\n",
      "     |  >>> from scipy.interpolate import SmoothSphereBivariateSpline\n",
      "     |  >>> lut = SmoothSphereBivariateSpline(lats.ravel(), lons.ravel(),\n",
      "     |  ...                                   data.T.ravel(), s=3.5)\n",
      "     |  \n",
      "     |  As a first test, we'll see what the algorithm returns when run on the\n",
      "     |  input coordinates\n",
      "     |  \n",
      "     |  >>> data_orig = lut(theta, phi)\n",
      "     |  \n",
      "     |  Finally we interpolate the data to a finer grid\n",
      "     |  \n",
      "     |  >>> fine_lats = np.linspace(0., np.pi, 70)\n",
      "     |  >>> fine_lons = np.linspace(0., 2 * np.pi, 90)\n",
      "     |  \n",
      "     |  >>> data_smth = lut(fine_lats, fine_lons)\n",
      "     |  \n",
      "     |  >>> import matplotlib.pyplot as plt\n",
      "     |  >>> fig = plt.figure()\n",
      "     |  >>> ax1 = fig.add_subplot(131)\n",
      "     |  >>> ax1.imshow(data, interpolation='nearest')\n",
      "     |  >>> ax2 = fig.add_subplot(132)\n",
      "     |  >>> ax2.imshow(data_orig, interpolation='nearest')\n",
      "     |  >>> ax3 = fig.add_subplot(133)\n",
      "     |  >>> ax3.imshow(data_smth, interpolation='nearest')\n",
      "     |  >>> plt.show()\n",
      "     |  \n",
      "     |  Method resolution order:\n",
      "     |      SmoothSphereBivariateSpline\n",
      "     |      SphereBivariateSpline\n",
      "     |      _BivariateSplineBase\n",
      "     |      builtins.object\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __call__(self, theta, phi, dtheta=0, dphi=0, grid=True)\n",
      "     |      Evaluate the spline or its derivatives at given positions.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      theta, phi : array_like\n",
      "     |          Input coordinates.\n",
      "     |      \n",
      "     |          If `grid` is False, evaluate the spline at points\n",
      "     |          ``(theta[i], phi[i]), i=0, ..., len(x)-1``.  Standard\n",
      "     |          Numpy broadcasting is obeyed.\n",
      "     |      \n",
      "     |          If `grid` is True: evaluate spline at the grid points\n",
      "     |          defined by the coordinate arrays theta, phi. The arrays\n",
      "     |          must be sorted to increasing order.\n",
      "     |      dtheta : int, optional\n",
      "     |          Order of theta-derivative\n",
      "     |      \n",
      "     |          .. versionadded:: 0.14.0\n",
      "     |      dphi : int\n",
      "     |          Order of phi-derivative\n",
      "     |      \n",
      "     |          .. versionadded:: 0.14.0\n",
      "     |      grid : bool\n",
      "     |          Whether to evaluate the results on a grid spanned by the\n",
      "     |          input arrays, or at points specified by the input arrays.\n",
      "     |      \n",
      "     |          .. versionadded:: 0.14.0\n",
      "     |  \n",
      "     |  __init__(self, theta, phi, r, w=None, s=0.0, eps=1e-16)\n",
      "     |      Initialize self.  See help(type(self)) for accurate signature.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from SphereBivariateSpline:\n",
      "     |  \n",
      "     |  ev(self, theta, phi, dtheta=0, dphi=0)\n",
      "     |      Evaluate the spline at points\n",
      "     |      \n",
      "     |      Returns the interpolated value at ``(theta[i], phi[i]),\n",
      "     |      i=0,...,len(theta)-1``.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      theta, phi : array_like\n",
      "     |          Input coordinates. Standard Numpy broadcasting is obeyed.\n",
      "     |      dtheta : int, optional\n",
      "     |          Order of theta-derivative\n",
      "     |      \n",
      "     |          .. versionadded:: 0.14.0\n",
      "     |      dphi : int, optional\n",
      "     |          Order of phi-derivative\n",
      "     |      \n",
      "     |          .. versionadded:: 0.14.0\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from _BivariateSplineBase:\n",
      "     |  \n",
      "     |  get_coeffs(self)\n",
      "     |      Return spline coefficients.\n",
      "     |  \n",
      "     |  get_knots(self)\n",
      "     |      Return a tuple (tx,ty) where tx,ty contain knots positions\n",
      "     |      of the spline with respect to x-, y-variable, respectively.\n",
      "     |      The position of interior and additional knots are given as\n",
      "     |      t[k+1:-k-1] and t[:k+1]=b, t[-k-1:]=e, respectively.\n",
      "     |  \n",
      "     |  get_residual(self)\n",
      "     |      Return weighted sum of squared residuals of the spline\n",
      "     |      approximation: sum ((w[i]*(z[i]-s(x[i],y[i])))**2,axis=0)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from _BivariateSplineBase:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |      dictionary for instance variables (if defined)\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "    \n",
      "    class UnivariateSpline(builtins.object)\n",
      "     |  UnivariateSpline(x, y, w=None, bbox=[None, None], k=3, s=None, ext=0, check_finite=False)\n",
      "     |  \n",
      "     |  1-D smoothing spline fit to a given set of data points.\n",
      "     |  \n",
      "     |  Fits a spline y = spl(x) of degree `k` to the provided `x`, `y` data.  `s`\n",
      "     |  specifies the number of knots by specifying a smoothing condition.\n",
      "     |  \n",
      "     |  Parameters\n",
      "     |  ----------\n",
      "     |  x : (N,) array_like\n",
      "     |      1-D array of independent input data. Must be increasing;\n",
      "     |      must be strictly increasing if `s` is 0.\n",
      "     |  y : (N,) array_like\n",
      "     |      1-D array of dependent input data, of the same length as `x`.\n",
      "     |  w : (N,) array_like, optional\n",
      "     |      Weights for spline fitting.  Must be positive.  If `w` is None,\n",
      "     |      weights are all equal. Default is None.\n",
      "     |  bbox : (2,) array_like, optional\n",
      "     |      2-sequence specifying the boundary of the approximation interval. If\n",
      "     |      `bbox` is None, ``bbox=[x[0], x[-1]]``. Default is None.\n",
      "     |  k : int, optional\n",
      "     |      Degree of the smoothing spline.  Must be 1 <= `k` <= 5.\n",
      "     |      ``k = 3`` is a cubic spline. Default is 3.\n",
      "     |  s : float or None, optional\n",
      "     |      Positive smoothing factor used to choose the number of knots.  Number\n",
      "     |      of knots will be increased until the smoothing condition is satisfied::\n",
      "     |  \n",
      "     |          sum((w[i] * (y[i]-spl(x[i])))**2, axis=0) <= s\n",
      "     |  \n",
      "     |      If `s` is None, ``s = len(w)`` which should be a good value if\n",
      "     |      ``1/w[i]`` is an estimate of the standard deviation of ``y[i]``.\n",
      "     |      If 0, spline will interpolate through all data points. Default is None.\n",
      "     |  ext : int or str, optional\n",
      "     |      Controls the extrapolation mode for elements\n",
      "     |      not in the interval defined by the knot sequence.\n",
      "     |  \n",
      "     |      * if ext=0 or 'extrapolate', return the extrapolated value.\n",
      "     |      * if ext=1 or 'zeros', return 0\n",
      "     |      * if ext=2 or 'raise', raise a ValueError\n",
      "     |      * if ext=3 of 'const', return the boundary value.\n",
      "     |  \n",
      "     |      Default is 0.\n",
      "     |  \n",
      "     |  check_finite : bool, optional\n",
      "     |      Whether to check that the input arrays contain only finite numbers.\n",
      "     |      Disabling may give a performance gain, but may result in problems\n",
      "     |      (crashes, non-termination or non-sensical results) if the inputs\n",
      "     |      do contain infinities or NaNs.\n",
      "     |      Default is False.\n",
      "     |  \n",
      "     |  See Also\n",
      "     |  --------\n",
      "     |  BivariateSpline :\n",
      "     |      a base class for bivariate splines.\n",
      "     |  SmoothBivariateSpline :\n",
      "     |      a smoothing bivariate spline through the given points\n",
      "     |  LSQBivariateSpline :\n",
      "     |      a bivariate spline using weighted least-squares fitting\n",
      "     |  RectSphereBivariateSpline :\n",
      "     |      a bivariate spline over a rectangular mesh on a sphere\n",
      "     |  SmoothSphereBivariateSpline :\n",
      "     |      a smoothing bivariate spline in spherical coordinates\n",
      "     |  LSQSphereBivariateSpline :\n",
      "     |      a bivariate spline in spherical coordinates using weighted\n",
      "     |      least-squares fitting\n",
      "     |  RectBivariateSpline :\n",
      "     |      a bivariate spline over a rectangular mesh\n",
      "     |  InterpolatedUnivariateSpline :\n",
      "     |      a interpolating univariate spline for a given set of data points.\n",
      "     |  bisplrep :\n",
      "     |      a function to find a bivariate B-spline representation of a surface\n",
      "     |  bisplev :\n",
      "     |      a function to evaluate a bivariate B-spline and its derivatives\n",
      "     |  splrep :\n",
      "     |      a function to find the B-spline representation of a 1-D curve\n",
      "     |  splev :\n",
      "     |      a function to evaluate a B-spline or its derivatives\n",
      "     |  sproot :\n",
      "     |      a function to find the roots of a cubic B-spline\n",
      "     |  splint :\n",
      "     |      a function to evaluate the definite integral of a B-spline between two\n",
      "     |      given points\n",
      "     |  spalde :\n",
      "     |      a function to evaluate all derivatives of a B-spline\n",
      "     |  \n",
      "     |  Notes\n",
      "     |  -----\n",
      "     |  The number of data points must be larger than the spline degree `k`.\n",
      "     |  \n",
      "     |  **NaN handling**: If the input arrays contain ``nan`` values, the result\n",
      "     |  is not useful, since the underlying spline fitting routines cannot deal\n",
      "     |  with ``nan``. A workaround is to use zero weights for not-a-number\n",
      "     |  data points:\n",
      "     |  \n",
      "     |  >>> from scipy.interpolate import UnivariateSpline\n",
      "     |  >>> x, y = np.array([1, 2, 3, 4]), np.array([1, np.nan, 3, 4])\n",
      "     |  >>> w = np.isnan(y)\n",
      "     |  >>> y[w] = 0.\n",
      "     |  >>> spl = UnivariateSpline(x, y, w=~w)\n",
      "     |  \n",
      "     |  Notice the need to replace a ``nan`` by a numerical value (precise value\n",
      "     |  does not matter as long as the corresponding weight is zero.)\n",
      "     |  \n",
      "     |  Examples\n",
      "     |  --------\n",
      "     |  >>> import matplotlib.pyplot as plt\n",
      "     |  >>> from scipy.interpolate import UnivariateSpline\n",
      "     |  >>> rng = np.random.default_rng()\n",
      "     |  >>> x = np.linspace(-3, 3, 50)\n",
      "     |  >>> y = np.exp(-x**2) + 0.1 * rng.standard_normal(50)\n",
      "     |  >>> plt.plot(x, y, 'ro', ms=5)\n",
      "     |  \n",
      "     |  Use the default value for the smoothing parameter:\n",
      "     |  \n",
      "     |  >>> spl = UnivariateSpline(x, y)\n",
      "     |  >>> xs = np.linspace(-3, 3, 1000)\n",
      "     |  >>> plt.plot(xs, spl(xs), 'g', lw=3)\n",
      "     |  \n",
      "     |  Manually change the amount of smoothing:\n",
      "     |  \n",
      "     |  >>> spl.set_smoothing_factor(0.5)\n",
      "     |  >>> plt.plot(xs, spl(xs), 'b', lw=3)\n",
      "     |  >>> plt.show()\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __call__(self, x, nu=0, ext=None)\n",
      "     |      Evaluate spline (or its nu-th derivative) at positions x.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      x : array_like\n",
      "     |          A 1-D array of points at which to return the value of the smoothed\n",
      "     |          spline or its derivatives. Note: `x` can be unordered but the\n",
      "     |          evaluation is more efficient if `x` is (partially) ordered.\n",
      "     |      nu  : int\n",
      "     |          The order of derivative of the spline to compute.\n",
      "     |      ext : int\n",
      "     |          Controls the value returned for elements of `x` not in the\n",
      "     |          interval defined by the knot sequence.\n",
      "     |      \n",
      "     |          * if ext=0 or 'extrapolate', return the extrapolated value.\n",
      "     |          * if ext=1 or 'zeros', return 0\n",
      "     |          * if ext=2 or 'raise', raise a ValueError\n",
      "     |          * if ext=3 or 'const', return the boundary value.\n",
      "     |      \n",
      "     |          The default value is 0, passed from the initialization of\n",
      "     |          UnivariateSpline.\n",
      "     |  \n",
      "     |  __init__(self, x, y, w=None, bbox=[None, None], k=3, s=None, ext=0, check_finite=False)\n",
      "     |      Initialize self.  See help(type(self)) for accurate signature.\n",
      "     |  \n",
      "     |  antiderivative(self, n=1)\n",
      "     |      Construct a new spline representing the antiderivative of this spline.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      n : int, optional\n",
      "     |          Order of antiderivative to evaluate. Default: 1\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      spline : UnivariateSpline\n",
      "     |          Spline of order k2=k+n representing the antiderivative of this\n",
      "     |          spline.\n",
      "     |      \n",
      "     |      Notes\n",
      "     |      -----\n",
      "     |      \n",
      "     |      .. versionadded:: 0.13.0\n",
      "     |      \n",
      "     |      See Also\n",
      "     |      --------\n",
      "     |      splantider, derivative\n",
      "     |      \n",
      "     |      Examples\n",
      "     |      --------\n",
      "     |      >>> from scipy.interpolate import UnivariateSpline\n",
      "     |      >>> x = np.linspace(0, np.pi/2, 70)\n",
      "     |      >>> y = 1 / np.sqrt(1 - 0.8*np.sin(x)**2)\n",
      "     |      >>> spl = UnivariateSpline(x, y, s=0)\n",
      "     |      \n",
      "     |      The derivative is the inverse operation of the antiderivative,\n",
      "     |      although some floating point error accumulates:\n",
      "     |      \n",
      "     |      >>> spl(1.7), spl.antiderivative().derivative()(1.7)\n",
      "     |      (array(2.1565429877197317), array(2.1565429877201865))\n",
      "     |      \n",
      "     |      Antiderivative can be used to evaluate definite integrals:\n",
      "     |      \n",
      "     |      >>> ispl = spl.antiderivative()\n",
      "     |      >>> ispl(np.pi/2) - ispl(0)\n",
      "     |      2.2572053588768486\n",
      "     |      \n",
      "     |      This is indeed an approximation to the complete elliptic integral\n",
      "     |      :math:`K(m) = \\int_0^{\\pi/2} [1 - m\\sin^2 x]^{-1/2} dx`:\n",
      "     |      \n",
      "     |      >>> from scipy.special import ellipk\n",
      "     |      >>> ellipk(0.8)\n",
      "     |      2.2572053268208538\n",
      "     |  \n",
      "     |  derivative(self, n=1)\n",
      "     |      Construct a new spline representing the derivative of this spline.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      n : int, optional\n",
      "     |          Order of derivative to evaluate. Default: 1\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      spline : UnivariateSpline\n",
      "     |          Spline of order k2=k-n representing the derivative of this\n",
      "     |          spline.\n",
      "     |      \n",
      "     |      See Also\n",
      "     |      --------\n",
      "     |      splder, antiderivative\n",
      "     |      \n",
      "     |      Notes\n",
      "     |      -----\n",
      "     |      \n",
      "     |      .. versionadded:: 0.13.0\n",
      "     |      \n",
      "     |      Examples\n",
      "     |      --------\n",
      "     |      This can be used for finding maxima of a curve:\n",
      "     |      \n",
      "     |      >>> from scipy.interpolate import UnivariateSpline\n",
      "     |      >>> x = np.linspace(0, 10, 70)\n",
      "     |      >>> y = np.sin(x)\n",
      "     |      >>> spl = UnivariateSpline(x, y, k=4, s=0)\n",
      "     |      \n",
      "     |      Now, differentiate the spline and find the zeros of the\n",
      "     |      derivative. (NB: `sproot` only works for order 3 splines, so we\n",
      "     |      fit an order 4 spline):\n",
      "     |      \n",
      "     |      >>> spl.derivative().roots() / np.pi\n",
      "     |      array([ 0.50000001,  1.5       ,  2.49999998])\n",
      "     |      \n",
      "     |      This agrees well with roots :math:`\\pi/2 + n\\pi` of\n",
      "     |      :math:`\\cos(x) = \\sin'(x)`.\n",
      "     |  \n",
      "     |  derivatives(self, x)\n",
      "     |      Return all derivatives of the spline at the point x.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      x : float\n",
      "     |          The point to evaluate the derivatives at.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      der : ndarray, shape(k+1,)\n",
      "     |          Derivatives of the orders 0 to k.\n",
      "     |      \n",
      "     |      Examples\n",
      "     |      --------\n",
      "     |      >>> from scipy.interpolate import UnivariateSpline\n",
      "     |      >>> x = np.linspace(0, 3, 11)\n",
      "     |      >>> y = x**2\n",
      "     |      >>> spl = UnivariateSpline(x, y)\n",
      "     |      >>> spl.derivatives(1.5)\n",
      "     |      array([2.25, 3.0, 2.0, 0])\n",
      "     |  \n",
      "     |  get_coeffs(self)\n",
      "     |      Return spline coefficients.\n",
      "     |  \n",
      "     |  get_knots(self)\n",
      "     |      Return positions of interior knots of the spline.\n",
      "     |      \n",
      "     |      Internally, the knot vector contains ``2*k`` additional boundary knots.\n",
      "     |  \n",
      "     |  get_residual(self)\n",
      "     |      Return weighted sum of squared residuals of the spline approximation.\n",
      "     |      \n",
      "     |      This is equivalent to::\n",
      "     |      \n",
      "     |           sum((w[i] * (y[i]-spl(x[i])))**2, axis=0)\n",
      "     |  \n",
      "     |  integral(self, a, b)\n",
      "     |      Return definite integral of the spline between two given points.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      a : float\n",
      "     |          Lower limit of integration.\n",
      "     |      b : float\n",
      "     |          Upper limit of integration.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      integral : float\n",
      "     |          The value of the definite integral of the spline between limits.\n",
      "     |      \n",
      "     |      Examples\n",
      "     |      --------\n",
      "     |      >>> from scipy.interpolate import UnivariateSpline\n",
      "     |      >>> x = np.linspace(0, 3, 11)\n",
      "     |      >>> y = x**2\n",
      "     |      >>> spl = UnivariateSpline(x, y)\n",
      "     |      >>> spl.integral(0, 3)\n",
      "     |      9.0\n",
      "     |      \n",
      "     |      which agrees with :math:`\\int x^2 dx = x^3 / 3` between the limits\n",
      "     |      of 0 and 3.\n",
      "     |      \n",
      "     |      A caveat is that this routine assumes the spline to be zero outside of\n",
      "     |      the data limits:\n",
      "     |      \n",
      "     |      >>> spl.integral(-1, 4)\n",
      "     |      9.0\n",
      "     |      >>> spl.integral(-1, 0)\n",
      "     |      0.0\n",
      "     |  \n",
      "     |  roots(self)\n",
      "     |      Return the zeros of the spline.\n",
      "     |      \n",
      "     |      Restriction: only cubic splines are supported by fitpack.\n",
      "     |  \n",
      "     |  set_smoothing_factor(self, s)\n",
      "     |      Continue spline computation with the given smoothing\n",
      "     |      factor s and with the knots found at the last call.\n",
      "     |      \n",
      "     |      This routine modifies the spline in place.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Static methods defined here:\n",
      "     |  \n",
      "     |  validate_input(x, y, w, bbox, k, s, ext, check_finite)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors defined here:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |      dictionary for instance variables (if defined)\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "    \n",
      "    class interp1d(scipy.interpolate.polyint._Interpolator1D)\n",
      "     |  interp1d(x, y, kind='linear', axis=-1, copy=True, bounds_error=None, fill_value=nan, assume_sorted=False)\n",
      "     |  \n",
      "     |  Interpolate a 1-D function.\n",
      "     |  \n",
      "     |  `x` and `y` are arrays of values used to approximate some function f:\n",
      "     |  ``y = f(x)``. This class returns a function whose call method uses\n",
      "     |  interpolation to find the value of new points.\n",
      "     |  \n",
      "     |  Parameters\n",
      "     |  ----------\n",
      "     |  x : (N,) array_like\n",
      "     |      A 1-D array of real values.\n",
      "     |  y : (...,N,...) array_like\n",
      "     |      A N-D array of real values. The length of `y` along the interpolation\n",
      "     |      axis must be equal to the length of `x`.\n",
      "     |  kind : str or int, optional\n",
      "     |      Specifies the kind of interpolation as a string or as an integer\n",
      "     |      specifying the order of the spline interpolator to use.\n",
      "     |      The string has to be one of 'linear', 'nearest', 'nearest-up', 'zero',\n",
      "     |      'slinear', 'quadratic', 'cubic', 'previous', or 'next'. 'zero',\n",
      "     |      'slinear', 'quadratic' and 'cubic' refer to a spline interpolation of\n",
      "     |      zeroth, first, second or third order; 'previous' and 'next' simply\n",
      "     |      return the previous or next value of the point; 'nearest-up' and\n",
      "     |      'nearest' differ when interpolating half-integers (e.g. 0.5, 1.5)\n",
      "     |      in that 'nearest-up' rounds up and 'nearest' rounds down. Default\n",
      "     |      is 'linear'.\n",
      "     |  axis : int, optional\n",
      "     |      Specifies the axis of `y` along which to interpolate.\n",
      "     |      Interpolation defaults to the last axis of `y`.\n",
      "     |  copy : bool, optional\n",
      "     |      If True, the class makes internal copies of x and y.\n",
      "     |      If False, references to `x` and `y` are used. The default is to copy.\n",
      "     |  bounds_error : bool, optional\n",
      "     |      If True, a ValueError is raised any time interpolation is attempted on\n",
      "     |      a value outside of the range of x (where extrapolation is\n",
      "     |      necessary). If False, out of bounds values are assigned `fill_value`.\n",
      "     |      By default, an error is raised unless ``fill_value=\"extrapolate\"``.\n",
      "     |  fill_value : array-like or (array-like, array_like) or \"extrapolate\", optional\n",
      "     |      - if a ndarray (or float), this value will be used to fill in for\n",
      "     |        requested points outside of the data range. If not provided, then\n",
      "     |        the default is NaN. The array-like must broadcast properly to the\n",
      "     |        dimensions of the non-interpolation axes.\n",
      "     |      - If a two-element tuple, then the first element is used as a\n",
      "     |        fill value for ``x_new < x[0]`` and the second element is used for\n",
      "     |        ``x_new > x[-1]``. Anything that is not a 2-element tuple (e.g.,\n",
      "     |        list or ndarray, regardless of shape) is taken to be a single\n",
      "     |        array-like argument meant to be used for both bounds as\n",
      "     |        ``below, above = fill_value, fill_value``.\n",
      "     |  \n",
      "     |        .. versionadded:: 0.17.0\n",
      "     |      - If \"extrapolate\", then points outside the data range will be\n",
      "     |        extrapolated.\n",
      "     |  \n",
      "     |        .. versionadded:: 0.17.0\n",
      "     |  assume_sorted : bool, optional\n",
      "     |      If False, values of `x` can be in any order and they are sorted first.\n",
      "     |      If True, `x` has to be an array of monotonically increasing values.\n",
      "     |  \n",
      "     |  Attributes\n",
      "     |  ----------\n",
      "     |  fill_value\n",
      "     |  \n",
      "     |  Methods\n",
      "     |  -------\n",
      "     |  __call__\n",
      "     |  \n",
      "     |  See Also\n",
      "     |  --------\n",
      "     |  splrep, splev\n",
      "     |      Spline interpolation/smoothing based on FITPACK.\n",
      "     |  UnivariateSpline : An object-oriented wrapper of the FITPACK routines.\n",
      "     |  interp2d : 2-D interpolation\n",
      "     |  \n",
      "     |  Notes\n",
      "     |  -----\n",
      "     |  Calling `interp1d` with NaNs present in input values results in\n",
      "     |  undefined behaviour.\n",
      "     |  \n",
      "     |  Input values `x` and `y` must be convertible to `float` values like\n",
      "     |  `int` or `float`.\n",
      "     |  \n",
      "     |  If the values in `x` are not unique, the resulting behavior is\n",
      "     |  undefined and specific to the choice of `kind`, i.e., changing\n",
      "     |  `kind` will change the behavior for duplicates.\n",
      "     |  \n",
      "     |  \n",
      "     |  Examples\n",
      "     |  --------\n",
      "     |  >>> import matplotlib.pyplot as plt\n",
      "     |  >>> from scipy import interpolate\n",
      "     |  >>> x = np.arange(0, 10)\n",
      "     |  >>> y = np.exp(-x/3.0)\n",
      "     |  >>> f = interpolate.interp1d(x, y)\n",
      "     |  \n",
      "     |  >>> xnew = np.arange(0, 9, 0.1)\n",
      "     |  >>> ynew = f(xnew)   # use interpolation function returned by `interp1d`\n",
      "     |  >>> plt.plot(x, y, 'o', xnew, ynew, '-')\n",
      "     |  >>> plt.show()\n",
      "     |  \n",
      "     |  Method resolution order:\n",
      "     |      interp1d\n",
      "     |      scipy.interpolate.polyint._Interpolator1D\n",
      "     |      builtins.object\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __init__(self, x, y, kind='linear', axis=-1, copy=True, bounds_error=None, fill_value=nan, assume_sorted=False)\n",
      "     |      Initialize a 1-D linear interpolation class.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors defined here:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |      dictionary for instance variables (if defined)\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "     |  \n",
      "     |  fill_value\n",
      "     |      The fill value.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from scipy.interpolate.polyint._Interpolator1D:\n",
      "     |  \n",
      "     |  __call__(self, x)\n",
      "     |      Evaluate the interpolant\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      x : array_like\n",
      "     |          Points to evaluate the interpolant at.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      y : array_like\n",
      "     |          Interpolated values. Shape is determined by replacing\n",
      "     |          the interpolation axis in the original array with the shape of x.\n",
      "     |      \n",
      "     |      Notes\n",
      "     |      -----\n",
      "     |      Input values `x` must be convertible to `float` values like `int` \n",
      "     |      or `float`.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from scipy.interpolate.polyint._Interpolator1D:\n",
      "     |  \n",
      "     |  dtype\n",
      "    \n",
      "    class interp2d(builtins.object)\n",
      "     |  interp2d(x, y, z, kind='linear', copy=True, bounds_error=False, fill_value=None)\n",
      "     |  \n",
      "     |  interp2d(x, y, z, kind='linear', copy=True, bounds_error=False,\n",
      "     |           fill_value=None)\n",
      "     |  \n",
      "     |  Interpolate over a 2-D grid.\n",
      "     |  \n",
      "     |  `x`, `y` and `z` are arrays of values used to approximate some function\n",
      "     |  f: ``z = f(x, y)`` which returns a scalar value `z`. This class returns a\n",
      "     |  function whose call method uses spline interpolation to find the value\n",
      "     |  of new points.\n",
      "     |  \n",
      "     |  If `x` and `y` represent a regular grid, consider using\n",
      "     |  `RectBivariateSpline`.\n",
      "     |  \n",
      "     |  If `z` is a vector value, consider using `interpn`.\n",
      "     |  \n",
      "     |  Note that calling `interp2d` with NaNs present in input values results in\n",
      "     |  undefined behaviour.\n",
      "     |  \n",
      "     |  Methods\n",
      "     |  -------\n",
      "     |  __call__\n",
      "     |  \n",
      "     |  Parameters\n",
      "     |  ----------\n",
      "     |  x, y : array_like\n",
      "     |      Arrays defining the data point coordinates.\n",
      "     |  \n",
      "     |      If the points lie on a regular grid, `x` can specify the column\n",
      "     |      coordinates and `y` the row coordinates, for example::\n",
      "     |  \n",
      "     |        >>> x = [0,1,2];  y = [0,3]; z = [[1,2,3], [4,5,6]]\n",
      "     |  \n",
      "     |      Otherwise, `x` and `y` must specify the full coordinates for each\n",
      "     |      point, for example::\n",
      "     |  \n",
      "     |        >>> x = [0,1,2,0,1,2];  y = [0,0,0,3,3,3]; z = [1,2,3,4,5,6]\n",
      "     |  \n",
      "     |      If `x` and `y` are multidimensional, they are flattened before use.\n",
      "     |  z : array_like\n",
      "     |      The values of the function to interpolate at the data points. If\n",
      "     |      `z` is a multidimensional array, it is flattened before use.  The\n",
      "     |      length of a flattened `z` array is either\n",
      "     |      len(`x`)*len(`y`) if `x` and `y` specify the column and row coordinates\n",
      "     |      or ``len(z) == len(x) == len(y)`` if `x` and `y` specify coordinates\n",
      "     |      for each point.\n",
      "     |  kind : {'linear', 'cubic', 'quintic'}, optional\n",
      "     |      The kind of spline interpolation to use. Default is 'linear'.\n",
      "     |  copy : bool, optional\n",
      "     |      If True, the class makes internal copies of x, y and z.\n",
      "     |      If False, references may be used. The default is to copy.\n",
      "     |  bounds_error : bool, optional\n",
      "     |      If True, when interpolated values are requested outside of the\n",
      "     |      domain of the input data (x,y), a ValueError is raised.\n",
      "     |      If False, then `fill_value` is used.\n",
      "     |  fill_value : number, optional\n",
      "     |      If provided, the value to use for points outside of the\n",
      "     |      interpolation domain. If omitted (None), values outside\n",
      "     |      the domain are extrapolated via nearest-neighbor extrapolation.\n",
      "     |  \n",
      "     |  See Also\n",
      "     |  --------\n",
      "     |  RectBivariateSpline :\n",
      "     |      Much faster 2-D interpolation if your input data is on a grid\n",
      "     |  bisplrep, bisplev :\n",
      "     |      Spline interpolation based on FITPACK\n",
      "     |  BivariateSpline : a more recent wrapper of the FITPACK routines\n",
      "     |  interp1d : 1-D version of this function\n",
      "     |  \n",
      "     |  Notes\n",
      "     |  -----\n",
      "     |  The minimum number of data points required along the interpolation\n",
      "     |  axis is ``(k+1)**2``, with k=1 for linear, k=3 for cubic and k=5 for\n",
      "     |  quintic interpolation.\n",
      "     |  \n",
      "     |  The interpolator is constructed by `bisplrep`, with a smoothing factor\n",
      "     |  of 0. If more control over smoothing is needed, `bisplrep` should be\n",
      "     |  used directly.\n",
      "     |  \n",
      "     |  Examples\n",
      "     |  --------\n",
      "     |  Construct a 2-D grid and interpolate on it:\n",
      "     |  \n",
      "     |  >>> from scipy import interpolate\n",
      "     |  >>> x = np.arange(-5.01, 5.01, 0.25)\n",
      "     |  >>> y = np.arange(-5.01, 5.01, 0.25)\n",
      "     |  >>> xx, yy = np.meshgrid(x, y)\n",
      "     |  >>> z = np.sin(xx**2+yy**2)\n",
      "     |  >>> f = interpolate.interp2d(x, y, z, kind='cubic')\n",
      "     |  \n",
      "     |  Now use the obtained interpolation function and plot the result:\n",
      "     |  \n",
      "     |  >>> import matplotlib.pyplot as plt\n",
      "     |  >>> xnew = np.arange(-5.01, 5.01, 1e-2)\n",
      "     |  >>> ynew = np.arange(-5.01, 5.01, 1e-2)\n",
      "     |  >>> znew = f(xnew, ynew)\n",
      "     |  >>> plt.plot(x, z[0, :], 'ro-', xnew, znew[0, :], 'b-')\n",
      "     |  >>> plt.show()\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __call__(self, x, y, dx=0, dy=0, assume_sorted=False)\n",
      "     |      Interpolate the function.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      x : 1-D array\n",
      "     |          x-coordinates of the mesh on which to interpolate.\n",
      "     |      y : 1-D array\n",
      "     |          y-coordinates of the mesh on which to interpolate.\n",
      "     |      dx : int >= 0, < kx\n",
      "     |          Order of partial derivatives in x.\n",
      "     |      dy : int >= 0, < ky\n",
      "     |          Order of partial derivatives in y.\n",
      "     |      assume_sorted : bool, optional\n",
      "     |          If False, values of `x` and `y` can be in any order and they are\n",
      "     |          sorted first.\n",
      "     |          If True, `x` and `y` have to be arrays of monotonically\n",
      "     |          increasing values.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      z : 2-D array with shape (len(y), len(x))\n",
      "     |          The interpolated values.\n",
      "     |  \n",
      "     |  __init__(self, x, y, z, kind='linear', copy=True, bounds_error=False, fill_value=None)\n",
      "     |      Initialize self.  See help(type(self)) for accurate signature.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors defined here:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |      dictionary for instance variables (if defined)\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "\n",
      "FUNCTIONS\n",
      "    approximate_taylor_polynomial(f, x, degree, scale, order=None)\n",
      "        Estimate the Taylor polynomial of f at x by polynomial fitting.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        f : callable\n",
      "            The function whose Taylor polynomial is sought. Should accept\n",
      "            a vector of `x` values.\n",
      "        x : scalar\n",
      "            The point at which the polynomial is to be evaluated.\n",
      "        degree : int\n",
      "            The degree of the Taylor polynomial\n",
      "        scale : scalar\n",
      "            The width of the interval to use to evaluate the Taylor polynomial.\n",
      "            Function values spread over a range this wide are used to fit the\n",
      "            polynomial. Must be chosen carefully.\n",
      "        order : int or None, optional\n",
      "            The order of the polynomial to be used in the fitting; `f` will be\n",
      "            evaluated ``order+1`` times. If None, use `degree`.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        p : poly1d instance\n",
      "            The Taylor polynomial (translated to the origin, so that\n",
      "            for example p(0)=f(x)).\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        The appropriate choice of \"scale\" is a trade-off; too large and the\n",
      "        function differs from its Taylor polynomial too much to get a good\n",
      "        answer, too small and round-off errors overwhelm the higher-order terms.\n",
      "        The algorithm used becomes numerically unstable around order 30 even\n",
      "        under ideal circumstances.\n",
      "        \n",
      "        Choosing order somewhat larger than degree may improve the higher-order\n",
      "        terms.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        We can calculate Taylor approximation polynomials of sin function with\n",
      "        various degrees:\n",
      "        \n",
      "        >>> import matplotlib.pyplot as plt\n",
      "        >>> from scipy.interpolate import approximate_taylor_polynomial\n",
      "        >>> x = np.linspace(-10.0, 10.0, num=100)\n",
      "        >>> plt.plot(x, np.sin(x), label=\"sin curve\")\n",
      "        >>> for degree in np.arange(1, 15, step=2):\n",
      "        ...     sin_taylor = approximate_taylor_polynomial(np.sin, 0, degree, 1,\n",
      "        ...                                                order=degree + 2)\n",
      "        ...     plt.plot(x, sin_taylor(x), label=f\"degree={degree}\")\n",
      "        >>> plt.legend(bbox_to_anchor=(1.05, 1), loc='upper left',\n",
      "        ...            borderaxespad=0.0, shadow=True)\n",
      "        >>> plt.tight_layout()\n",
      "        >>> plt.axis([-10, 10, -10, 10])\n",
      "        >>> plt.show()\n",
      "    \n",
      "    barycentric_interpolate(xi, yi, x, axis=0)\n",
      "        Convenience function for polynomial interpolation.\n",
      "        \n",
      "        Constructs a polynomial that passes through a given set of points,\n",
      "        then evaluates the polynomial. For reasons of numerical stability,\n",
      "        this function does not compute the coefficients of the polynomial.\n",
      "        \n",
      "        This function uses a \"barycentric interpolation\" method that treats\n",
      "        the problem as a special case of rational function interpolation.\n",
      "        This algorithm is quite stable, numerically, but even in a world of\n",
      "        exact computation, unless the `x` coordinates are chosen very\n",
      "        carefully - Chebyshev zeros (e.g., cos(i*pi/n)) are a good choice -\n",
      "        polynomial interpolation itself is a very ill-conditioned process\n",
      "        due to the Runge phenomenon.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        xi : array_like\n",
      "            1-D array of x coordinates of the points the polynomial should\n",
      "            pass through\n",
      "        yi : array_like\n",
      "            The y coordinates of the points the polynomial should pass through.\n",
      "        x : scalar or array_like\n",
      "            Points to evaluate the interpolator at.\n",
      "        axis : int, optional\n",
      "            Axis in the yi array corresponding to the x-coordinate values.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        y : scalar or array_like\n",
      "            Interpolated values. Shape is determined by replacing\n",
      "            the interpolation axis in the original array with the shape of x.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        BarycentricInterpolator : Bary centric interpolator\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        Construction of the interpolation weights is a relatively slow process.\n",
      "        If you want to call this many times with the same xi (but possibly\n",
      "        varying yi or x) you should use the class `BarycentricInterpolator`.\n",
      "        This is what this function uses internally.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        We can interpolate 2D observed data using barycentric interpolation:\n",
      "        \n",
      "        >>> import matplotlib.pyplot as plt\n",
      "        >>> from scipy.interpolate import barycentric_interpolate\n",
      "        >>> x_observed = np.linspace(0.0, 10.0, 11)\n",
      "        >>> y_observed = np.sin(x_observed)\n",
      "        >>> x = np.linspace(min(x_observed), max(x_observed), num=100)\n",
      "        >>> y = barycentric_interpolate(x_observed, y_observed, x)\n",
      "        >>> plt.plot(x_observed, y_observed, \"o\", label=\"observation\")\n",
      "        >>> plt.plot(x, y, label=\"barycentric interpolation\")\n",
      "        >>> plt.legend()\n",
      "        >>> plt.show()\n",
      "    \n",
      "    bisplev(x, y, tck, dx=0, dy=0)\n",
      "        Evaluate a bivariate B-spline and its derivatives.\n",
      "        \n",
      "        Return a rank-2 array of spline function values (or spline derivative\n",
      "        values) at points given by the cross-product of the rank-1 arrays `x` and\n",
      "        `y`.  In special cases, return an array or just a float if either `x` or\n",
      "        `y` or both are floats.  Based on BISPEV from FITPACK.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x, y : ndarray\n",
      "            Rank-1 arrays specifying the domain over which to evaluate the\n",
      "            spline or its derivative.\n",
      "        tck : tuple\n",
      "            A sequence of length 5 returned by `bisplrep` containing the knot\n",
      "            locations, the coefficients, and the degree of the spline:\n",
      "            [tx, ty, c, kx, ky].\n",
      "        dx, dy : int, optional\n",
      "            The orders of the partial derivatives in `x` and `y` respectively.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        vals : ndarray\n",
      "            The B-spline or its derivative evaluated over the set formed by\n",
      "            the cross-product of `x` and `y`.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        splprep, splrep, splint, sproot, splev\n",
      "        UnivariateSpline, BivariateSpline\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "            See `bisplrep` to generate the `tck` representation.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] Dierckx P. : An algorithm for surface fitting\n",
      "           with spline functions\n",
      "           Ima J. Numer. Anal. 1 (1981) 267-283.\n",
      "        .. [2] Dierckx P. : An algorithm for surface fitting\n",
      "           with spline functions\n",
      "           report tw50, Dept. Computer Science,K.U.Leuven, 1980.\n",
      "        .. [3] Dierckx P. : Curve and surface fitting with splines,\n",
      "           Monographs on Numerical Analysis, Oxford University Press, 1993.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        Examples are given :ref:`in the tutorial <tutorial-interpolate_2d_spline>`.\n",
      "    \n",
      "    bisplrep(x, y, z, w=None, xb=None, xe=None, yb=None, ye=None, kx=3, ky=3, task=0, s=None, eps=1e-16, tx=None, ty=None, full_output=0, nxest=None, nyest=None, quiet=1)\n",
      "        Find a bivariate B-spline representation of a surface.\n",
      "        \n",
      "        Given a set of data points (x[i], y[i], z[i]) representing a surface\n",
      "        z=f(x,y), compute a B-spline representation of the surface. Based on\n",
      "        the routine SURFIT from FITPACK.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x, y, z : ndarray\n",
      "            Rank-1 arrays of data points.\n",
      "        w : ndarray, optional\n",
      "            Rank-1 array of weights. By default ``w=np.ones(len(x))``.\n",
      "        xb, xe : float, optional\n",
      "            End points of approximation interval in `x`.\n",
      "            By default ``xb = x.min(), xe=x.max()``.\n",
      "        yb, ye : float, optional\n",
      "            End points of approximation interval in `y`.\n",
      "            By default ``yb=y.min(), ye = y.max()``.\n",
      "        kx, ky : int, optional\n",
      "            The degrees of the spline (1 <= kx, ky <= 5).\n",
      "            Third order (kx=ky=3) is recommended.\n",
      "        task : int, optional\n",
      "            If task=0, find knots in x and y and coefficients for a given\n",
      "            smoothing factor, s.\n",
      "            If task=1, find knots and coefficients for another value of the\n",
      "            smoothing factor, s.  bisplrep must have been previously called\n",
      "            with task=0 or task=1.\n",
      "            If task=-1, find coefficients for a given set of knots tx, ty.\n",
      "        s : float, optional\n",
      "            A non-negative smoothing factor. If weights correspond\n",
      "            to the inverse of the standard-deviation of the errors in z,\n",
      "            then a good s-value should be found in the range\n",
      "            ``(m-sqrt(2*m),m+sqrt(2*m))`` where m=len(x).\n",
      "        eps : float, optional\n",
      "            A threshold for determining the effective rank of an\n",
      "            over-determined linear system of equations (0 < eps < 1).\n",
      "            `eps` is not likely to need changing.\n",
      "        tx, ty : ndarray, optional\n",
      "            Rank-1 arrays of the knots of the spline for task=-1\n",
      "        full_output : int, optional\n",
      "            Non-zero to return optional outputs.\n",
      "        nxest, nyest : int, optional\n",
      "            Over-estimates of the total number of knots. If None then\n",
      "            ``nxest = max(kx+sqrt(m/2),2*kx+3)``,\n",
      "            ``nyest = max(ky+sqrt(m/2),2*ky+3)``.\n",
      "        quiet : int, optional\n",
      "            Non-zero to suppress printing of messages.\n",
      "            This parameter is deprecated; use standard Python warning filters\n",
      "            instead.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        tck : array_like\n",
      "            A list [tx, ty, c, kx, ky] containing the knots (tx, ty) and\n",
      "            coefficients (c) of the bivariate B-spline representation of the\n",
      "            surface along with the degree of the spline.\n",
      "        fp : ndarray\n",
      "            The weighted sum of squared residuals of the spline approximation.\n",
      "        ier : int\n",
      "            An integer flag about splrep success. Success is indicated if\n",
      "            ier<=0. If ier in [1,2,3] an error occurred but was not raised.\n",
      "            Otherwise an error is raised.\n",
      "        msg : str\n",
      "            A message corresponding to the integer flag, ier.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        splprep, splrep, splint, sproot, splev\n",
      "        UnivariateSpline, BivariateSpline\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        See `bisplev` to evaluate the value of the B-spline given its tck\n",
      "        representation.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] Dierckx P.:An algorithm for surface fitting with spline functions\n",
      "           Ima J. Numer. Anal. 1 (1981) 267-283.\n",
      "        .. [2] Dierckx P.:An algorithm for surface fitting with spline functions\n",
      "           report tw50, Dept. Computer Science,K.U.Leuven, 1980.\n",
      "        .. [3] Dierckx P.:Curve and surface fitting with splines, Monographs on\n",
      "           Numerical Analysis, Oxford University Press, 1993.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        Examples are given :ref:`in the tutorial <tutorial-interpolate_2d_spline>`.\n",
      "    \n",
      "    griddata(points, values, xi, method='linear', fill_value=nan, rescale=False)\n",
      "        Interpolate unstructured D-D data.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        points : 2-D ndarray of floats with shape (n, D), or length D tuple of 1-D ndarrays with shape (n,).\n",
      "            Data point coordinates.\n",
      "        values : ndarray of float or complex, shape (n,)\n",
      "            Data values.\n",
      "        xi : 2-D ndarray of floats with shape (m, D), or length D tuple of ndarrays broadcastable to the same shape.\n",
      "            Points at which to interpolate data.\n",
      "        method : {'linear', 'nearest', 'cubic'}, optional\n",
      "            Method of interpolation. One of\n",
      "        \n",
      "            ``nearest``\n",
      "              return the value at the data point closest to\n",
      "              the point of interpolation. See `NearestNDInterpolator` for\n",
      "              more details.\n",
      "        \n",
      "            ``linear``\n",
      "              tessellate the input point set to N-D\n",
      "              simplices, and interpolate linearly on each simplex. See\n",
      "              `LinearNDInterpolator` for more details.\n",
      "        \n",
      "            ``cubic`` (1-D)\n",
      "              return the value determined from a cubic\n",
      "              spline.\n",
      "        \n",
      "            ``cubic`` (2-D)\n",
      "              return the value determined from a\n",
      "              piecewise cubic, continuously differentiable (C1), and\n",
      "              approximately curvature-minimizing polynomial surface. See\n",
      "              `CloughTocher2DInterpolator` for more details.\n",
      "        fill_value : float, optional\n",
      "            Value used to fill in for requested points outside of the\n",
      "            convex hull of the input points. If not provided, then the\n",
      "            default is ``nan``. This option has no effect for the\n",
      "            'nearest' method.\n",
      "        rescale : bool, optional\n",
      "            Rescale points to unit cube before performing interpolation.\n",
      "            This is useful if some of the input dimensions have\n",
      "            incommensurable units and differ by many orders of magnitude.\n",
      "        \n",
      "            .. versionadded:: 0.14.0\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        ndarray\n",
      "            Array of interpolated values.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        \n",
      "        .. versionadded:: 0.9\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        \n",
      "        Suppose we want to interpolate the 2-D function\n",
      "        \n",
      "        >>> def func(x, y):\n",
      "        ...     return x*(1-x)*np.cos(4*np.pi*x) * np.sin(4*np.pi*y**2)**2\n",
      "        \n",
      "        on a grid in [0, 1]x[0, 1]\n",
      "        \n",
      "        >>> grid_x, grid_y = np.mgrid[0:1:100j, 0:1:200j]\n",
      "        \n",
      "        but we only know its values at 1000 data points:\n",
      "        \n",
      "        >>> rng = np.random.default_rng()\n",
      "        >>> points = rng.random((1000, 2))\n",
      "        >>> values = func(points[:,0], points[:,1])\n",
      "        \n",
      "        This can be done with `griddata` -- below we try out all of the\n",
      "        interpolation methods:\n",
      "        \n",
      "        >>> from scipy.interpolate import griddata\n",
      "        >>> grid_z0 = griddata(points, values, (grid_x, grid_y), method='nearest')\n",
      "        >>> grid_z1 = griddata(points, values, (grid_x, grid_y), method='linear')\n",
      "        >>> grid_z2 = griddata(points, values, (grid_x, grid_y), method='cubic')\n",
      "        \n",
      "        One can see that the exact result is reproduced by all of the\n",
      "        methods to some degree, but for this smooth function the piecewise\n",
      "        cubic interpolant gives the best results:\n",
      "        \n",
      "        >>> import matplotlib.pyplot as plt\n",
      "        >>> plt.subplot(221)\n",
      "        >>> plt.imshow(func(grid_x, grid_y).T, extent=(0,1,0,1), origin='lower')\n",
      "        >>> plt.plot(points[:,0], points[:,1], 'k.', ms=1)\n",
      "        >>> plt.title('Original')\n",
      "        >>> plt.subplot(222)\n",
      "        >>> plt.imshow(grid_z0.T, extent=(0,1,0,1), origin='lower')\n",
      "        >>> plt.title('Nearest')\n",
      "        >>> plt.subplot(223)\n",
      "        >>> plt.imshow(grid_z1.T, extent=(0,1,0,1), origin='lower')\n",
      "        >>> plt.title('Linear')\n",
      "        >>> plt.subplot(224)\n",
      "        >>> plt.imshow(grid_z2.T, extent=(0,1,0,1), origin='lower')\n",
      "        >>> plt.title('Cubic')\n",
      "        >>> plt.gcf().set_size_inches(6, 6)\n",
      "        >>> plt.show()\n",
      "        \n",
      "        See also\n",
      "        --------\n",
      "        LinearNDInterpolator :\n",
      "            Piecewise linear interpolant in N dimensions.\n",
      "        NearestNDInterpolator :\n",
      "            Nearest-neighbor interpolation in N dimensions.\n",
      "        CloughTocher2DInterpolator :\n",
      "            Piecewise cubic, C1 smooth, curvature-minimizing interpolant in 2D.\n",
      "    \n",
      "    insert(x, tck, m=1, per=0)\n",
      "        Insert knots into a B-spline.\n",
      "        \n",
      "        Given the knots and coefficients of a B-spline representation, create a\n",
      "        new B-spline with a knot inserted `m` times at point `x`.\n",
      "        This is a wrapper around the FORTRAN routine insert of FITPACK.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x (u) : array_like\n",
      "            A 1-D point at which to insert a new knot(s).  If `tck` was returned\n",
      "            from ``splprep``, then the parameter values, u should be given.\n",
      "        tck : a `BSpline` instance or a tuple\n",
      "            If tuple, then it is expected to be a tuple (t,c,k) containing\n",
      "            the vector of knots, the B-spline coefficients, and the degree of\n",
      "            the spline.\n",
      "        m : int, optional\n",
      "            The number of times to insert the given knot (its multiplicity).\n",
      "            Default is 1.\n",
      "        per : int, optional\n",
      "            If non-zero, the input spline is considered periodic.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        BSpline instance or a tuple\n",
      "            A new B-spline with knots t, coefficients c, and degree k.\n",
      "            ``t(k+1) <= x <= t(n-k)``, where k is the degree of the spline.\n",
      "            In case of a periodic spline (``per != 0``) there must be\n",
      "            either at least k interior knots t(j) satisfying ``t(k+1)<t(j)<=x``\n",
      "            or at least k interior knots t(j) satisfying ``x<=t(j)<t(n-k)``.\n",
      "            A tuple is returned iff the input argument `tck` is a tuple, otherwise\n",
      "            a BSpline object is constructed and returned.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        Based on algorithms from [1]_ and [2]_.\n",
      "        \n",
      "        Manipulating the tck-tuples directly is not recommended. In new code,\n",
      "        prefer using the `BSpline` objects.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] W. Boehm, \"Inserting new knots into b-spline curves.\",\n",
      "            Computer Aided Design, 12, p.199-201, 1980.\n",
      "        .. [2] P. Dierckx, \"Curve and surface fitting with splines, Monographs on\n",
      "            Numerical Analysis\", Oxford University Press, 1993.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        You can insert knots into a B-spline.\n",
      "        \n",
      "        >>> from scipy.interpolate import splrep, insert\n",
      "        >>> x = np.linspace(0, 10, 5)\n",
      "        >>> y = np.sin(x)\n",
      "        >>> tck = splrep(x, y)\n",
      "        >>> tck[0]\n",
      "        array([ 0.,  0.,  0.,  0.,  5., 10., 10., 10., 10.])\n",
      "        \n",
      "        A knot is inserted:\n",
      "        \n",
      "        >>> tck_inserted = insert(3, tck)\n",
      "        >>> tck_inserted[0]\n",
      "        array([ 0.,  0.,  0.,  0.,  3.,  5., 10., 10., 10., 10.])\n",
      "        \n",
      "        Some knots are inserted:\n",
      "        \n",
      "        >>> tck_inserted2 = insert(8, tck, m=3)\n",
      "        >>> tck_inserted2[0]\n",
      "        array([ 0.,  0.,  0.,  0.,  5.,  8.,  8.,  8., 10., 10., 10., 10.])\n",
      "    \n",
      "    interpn(points, values, xi, method='linear', bounds_error=True, fill_value=nan)\n",
      "        Multidimensional interpolation on regular grids.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        points : tuple of ndarray of float, with shapes (m1, ), ..., (mn, )\n",
      "            The points defining the regular grid in n dimensions.\n",
      "        \n",
      "        values : array_like, shape (m1, ..., mn, ...)\n",
      "            The data on the regular grid in n dimensions.\n",
      "        \n",
      "        xi : ndarray of shape (..., ndim)\n",
      "            The coordinates to sample the gridded data at\n",
      "        \n",
      "        method : str, optional\n",
      "            The method of interpolation to perform. Supported are \"linear\" and\n",
      "            \"nearest\", and \"splinef2d\". \"splinef2d\" is only supported for\n",
      "            2-dimensional data.\n",
      "        \n",
      "        bounds_error : bool, optional\n",
      "            If True, when interpolated values are requested outside of the\n",
      "            domain of the input data, a ValueError is raised.\n",
      "            If False, then `fill_value` is used.\n",
      "        \n",
      "        fill_value : number, optional\n",
      "            If provided, the value to use for points outside of the\n",
      "            interpolation domain. If None, values outside\n",
      "            the domain are extrapolated.  Extrapolation is not supported by method\n",
      "            \"splinef2d\".\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        values_x : ndarray, shape xi.shape[:-1] + values.shape[ndim:]\n",
      "            Interpolated values at input coordinates.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        \n",
      "        .. versionadded:: 0.14\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        Evaluate a simple example function on the points of a regular 3-D grid:\n",
      "        \n",
      "        >>> from scipy.interpolate import interpn\n",
      "        >>> def value_func_3d(x, y, z):\n",
      "        ...     return 2 * x + 3 * y - z\n",
      "        >>> x = np.linspace(0, 4, 5)\n",
      "        >>> y = np.linspace(0, 5, 6)\n",
      "        >>> z = np.linspace(0, 6, 7)\n",
      "        >>> points = (x, y, z)\n",
      "        >>> values = value_func_3d(*np.meshgrid(*points, indexing='ij'))\n",
      "        \n",
      "        Evaluate the interpolating function at a point\n",
      "        \n",
      "        >>> point = np.array([2.21, 3.12, 1.15])\n",
      "        >>> print(interpn(points, values, point))\n",
      "        [12.63]\n",
      "        \n",
      "        See also\n",
      "        --------\n",
      "        NearestNDInterpolator : Nearest neighbor interpolation on unstructured\n",
      "                                data in N dimensions\n",
      "        \n",
      "        LinearNDInterpolator : Piecewise linear interpolant on unstructured data\n",
      "                               in N dimensions\n",
      "        \n",
      "        RegularGridInterpolator : Linear and nearest-neighbor Interpolation on a\n",
      "                                  regular grid in arbitrary dimensions\n",
      "        \n",
      "        RectBivariateSpline : Bivariate spline approximation over a rectangular mesh\n",
      "    \n",
      "    krogh_interpolate(xi, yi, x, der=0, axis=0)\n",
      "        Convenience function for polynomial interpolation.\n",
      "        \n",
      "        See `KroghInterpolator` for more details.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        xi : array_like\n",
      "            Known x-coordinates.\n",
      "        yi : array_like\n",
      "            Known y-coordinates, of shape ``(xi.size, R)``. Interpreted as\n",
      "            vectors of length R, or scalars if R=1.\n",
      "        x : array_like\n",
      "            Point or points at which to evaluate the derivatives.\n",
      "        der : int or list, optional\n",
      "            How many derivatives to extract; None for all potentially\n",
      "            nonzero derivatives (that is a number equal to the number\n",
      "            of points), or a list of derivatives to extract. This number\n",
      "            includes the function value as 0th derivative.\n",
      "        axis : int, optional\n",
      "            Axis in the yi array corresponding to the x-coordinate values.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        d : ndarray\n",
      "            If the interpolator's values are R-D then the\n",
      "            returned array will be the number of derivatives by N by R.\n",
      "            If `x` is a scalar, the middle dimension will be dropped; if\n",
      "            the `yi` are scalars then the last dimension will be dropped.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        KroghInterpolator : Krogh interpolator\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        Construction of the interpolating polynomial is a relatively expensive\n",
      "        process. If you want to evaluate it repeatedly consider using the class\n",
      "        KroghInterpolator (which is what this function uses).\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        We can interpolate 2D observed data using krogh interpolation:\n",
      "        \n",
      "        >>> import matplotlib.pyplot as plt\n",
      "        >>> from scipy.interpolate import krogh_interpolate\n",
      "        >>> x_observed = np.linspace(0.0, 10.0, 11)\n",
      "        >>> y_observed = np.sin(x_observed)\n",
      "        >>> x = np.linspace(min(x_observed), max(x_observed), num=100)\n",
      "        >>> y = krogh_interpolate(x_observed, y_observed, x)\n",
      "        >>> plt.plot(x_observed, y_observed, \"o\", label=\"observation\")\n",
      "        >>> plt.plot(x, y, label=\"krogh interpolation\")\n",
      "        >>> plt.legend()\n",
      "        >>> plt.show()\n",
      "    \n",
      "    lagrange(x, w)\n",
      "        Return a Lagrange interpolating polynomial.\n",
      "        \n",
      "        Given two 1-D arrays `x` and `w,` returns the Lagrange interpolating\n",
      "        polynomial through the points ``(x, w)``.\n",
      "        \n",
      "        Warning: This implementation is numerically unstable. Do not expect to\n",
      "        be able to use more than about 20 points even if they are chosen optimally.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x : array_like\n",
      "            `x` represents the x-coordinates of a set of datapoints.\n",
      "        w : array_like\n",
      "            `w` represents the y-coordinates of a set of datapoints, i.e., f(`x`).\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        lagrange : `numpy.poly1d` instance\n",
      "            The Lagrange interpolating polynomial.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        Interpolate :math:`f(x) = x^3` by 3 points.\n",
      "        \n",
      "        >>> from scipy.interpolate import lagrange\n",
      "        >>> x = np.array([0, 1, 2])\n",
      "        >>> y = x**3\n",
      "        >>> poly = lagrange(x, y)\n",
      "        \n",
      "        Since there are only 3 points, Lagrange polynomial has degree 2. Explicitly,\n",
      "        it is given by\n",
      "        \n",
      "        .. math::\n",
      "        \n",
      "            \\begin{aligned}\n",
      "                L(x) &= 1\\times \\frac{x (x - 2)}{-1} + 8\\times \\frac{x (x-1)}{2} \\\\\n",
      "                     &= x (-2 + 3x)\n",
      "            \\end{aligned}\n",
      "        \n",
      "        >>> from numpy.polynomial.polynomial import Polynomial\n",
      "        >>> Polynomial(poly).coef\n",
      "        array([ 3., -2.,  0.])\n",
      "    \n",
      "    make_interp_spline(x, y, k=3, t=None, bc_type=None, axis=0, check_finite=True)\n",
      "        Compute the (coefficients of) interpolating B-spline.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x : array_like, shape (n,)\n",
      "            Abscissas.\n",
      "        y : array_like, shape (n, ...)\n",
      "            Ordinates.\n",
      "        k : int, optional\n",
      "            B-spline degree. Default is cubic, k=3.\n",
      "        t : array_like, shape (nt + k + 1,), optional.\n",
      "            Knots.\n",
      "            The number of knots needs to agree with the number of datapoints and\n",
      "            the number of derivatives at the edges. Specifically, ``nt - n`` must\n",
      "            equal ``len(deriv_l) + len(deriv_r)``.\n",
      "        bc_type : 2-tuple or None\n",
      "            Boundary conditions.\n",
      "            Default is None, which means choosing the boundary conditions\n",
      "            automatically. Otherwise, it must be a length-two tuple where the first\n",
      "            element sets the boundary conditions at ``x[0]`` and the second\n",
      "            element sets the boundary conditions at ``x[-1]``. Each of these must\n",
      "            be an iterable of pairs ``(order, value)`` which gives the values of\n",
      "            derivatives of specified orders at the given edge of the interpolation\n",
      "            interval.\n",
      "            Alternatively, the following string aliases are recognized:\n",
      "        \n",
      "            * ``\"clamped\"``: The first derivatives at the ends are zero. This is\n",
      "               equivalent to ``bc_type=([(1, 0.0)], [(1, 0.0)])``.\n",
      "            * ``\"natural\"``: The second derivatives at ends are zero. This is\n",
      "              equivalent to ``bc_type=([(2, 0.0)], [(2, 0.0)])``.\n",
      "            * ``\"not-a-knot\"`` (default): The first and second segments are the\n",
      "              same polynomial. This is equivalent to having ``bc_type=None``.\n",
      "            * ``\"periodic\"``: The values and the first ``k-1`` derivatives at the\n",
      "              ends are equivalent.\n",
      "        \n",
      "        axis : int, optional\n",
      "            Interpolation axis. Default is 0.\n",
      "        check_finite : bool, optional\n",
      "            Whether to check that the input arrays contain only finite numbers.\n",
      "            Disabling may give a performance gain, but may result in problems\n",
      "            (crashes, non-termination) if the inputs do contain infinities or NaNs.\n",
      "            Default is True.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        b : a BSpline object of the degree ``k`` and with knots ``t``.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        \n",
      "        Use cubic interpolation on Chebyshev nodes:\n",
      "        \n",
      "        >>> def cheb_nodes(N):\n",
      "        ...     jj = 2.*np.arange(N) + 1\n",
      "        ...     x = np.cos(np.pi * jj / 2 / N)[::-1]\n",
      "        ...     return x\n",
      "        \n",
      "        >>> x = cheb_nodes(20)\n",
      "        >>> y = np.sqrt(1 - x**2)\n",
      "        \n",
      "        >>> from scipy.interpolate import BSpline, make_interp_spline\n",
      "        >>> b = make_interp_spline(x, y)\n",
      "        >>> np.allclose(b(x), y)\n",
      "        True\n",
      "        \n",
      "        Note that the default is a cubic spline with a not-a-knot boundary condition\n",
      "        \n",
      "        >>> b.k\n",
      "        3\n",
      "        \n",
      "        Here we use a 'natural' spline, with zero 2nd derivatives at edges:\n",
      "        \n",
      "        >>> l, r = [(2, 0.0)], [(2, 0.0)]\n",
      "        >>> b_n = make_interp_spline(x, y, bc_type=(l, r))  # or, bc_type=\"natural\"\n",
      "        >>> np.allclose(b_n(x), y)\n",
      "        True\n",
      "        >>> x0, x1 = x[0], x[-1]\n",
      "        >>> np.allclose([b_n(x0, 2), b_n(x1, 2)], [0, 0])\n",
      "        True\n",
      "        \n",
      "        Interpolation of parametric curves is also supported. As an example, we\n",
      "        compute a discretization of a snail curve in polar coordinates\n",
      "        \n",
      "        >>> phi = np.linspace(0, 2.*np.pi, 40)\n",
      "        >>> r = 0.3 + np.cos(phi)\n",
      "        >>> x, y = r*np.cos(phi), r*np.sin(phi)  # convert to Cartesian coordinates\n",
      "        \n",
      "        Build an interpolating curve, parameterizing it by the angle\n",
      "        \n",
      "        >>> from scipy.interpolate import make_interp_spline\n",
      "        >>> spl = make_interp_spline(phi, np.c_[x, y])\n",
      "        \n",
      "        Evaluate the interpolant on a finer grid (note that we transpose the result\n",
      "        to unpack it into a pair of x- and y-arrays)\n",
      "        \n",
      "        >>> phi_new = np.linspace(0, 2.*np.pi, 100)\n",
      "        >>> x_new, y_new = spl(phi_new).T\n",
      "        \n",
      "        Plot the result\n",
      "        \n",
      "        >>> import matplotlib.pyplot as plt\n",
      "        >>> plt.plot(x, y, 'o')\n",
      "        >>> plt.plot(x_new, y_new, '-')\n",
      "        >>> plt.show()\n",
      "        \n",
      "        Build a B-spline curve with 2 dimensional y\n",
      "        \n",
      "        >>> x = np.linspace(0, 2*np.pi, 10)\n",
      "        >>> y = np.array([np.sin(x), np.cos(x)])\n",
      "        \n",
      "        Periodic condition is satisfied because y coordinates of points on the ends\n",
      "        are equivalent\n",
      "        \n",
      "        >>> ax = plt.axes(projection='3d')\n",
      "        >>> xx = np.linspace(0, 2*np.pi, 100)\n",
      "        >>> bspl = make_interp_spline(x, y, k=5, bc_type='periodic', axis=1)\n",
      "        >>> ax.plot3D(xx, *bspl(xx))\n",
      "        >>> ax.scatter3D(x, *y, color='red')\n",
      "        >>> plt.show()\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        BSpline : base class representing the B-spline objects\n",
      "        CubicSpline : a cubic spline in the polynomial basis\n",
      "        make_lsq_spline : a similar factory function for spline fitting\n",
      "        UnivariateSpline : a wrapper over FITPACK spline fitting routines\n",
      "        splrep : a wrapper over FITPACK spline fitting routines\n",
      "    \n",
      "    make_lsq_spline(x, y, t, k=3, w=None, axis=0, check_finite=True)\n",
      "        Compute the (coefficients of) an LSQ B-spline.\n",
      "        \n",
      "        The result is a linear combination\n",
      "        \n",
      "        .. math::\n",
      "        \n",
      "                S(x) = \\sum_j c_j B_j(x; t)\n",
      "        \n",
      "        of the B-spline basis elements, :math:`B_j(x; t)`, which minimizes\n",
      "        \n",
      "        .. math::\n",
      "        \n",
      "            \\sum_{j} \\left( w_j \\times (S(x_j) - y_j) \\right)^2\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x : array_like, shape (m,)\n",
      "            Abscissas.\n",
      "        y : array_like, shape (m, ...)\n",
      "            Ordinates.\n",
      "        t : array_like, shape (n + k + 1,).\n",
      "            Knots.\n",
      "            Knots and data points must satisfy Schoenberg-Whitney conditions.\n",
      "        k : int, optional\n",
      "            B-spline degree. Default is cubic, k=3.\n",
      "        w : array_like, shape (n,), optional\n",
      "            Weights for spline fitting. Must be positive. If ``None``,\n",
      "            then weights are all equal.\n",
      "            Default is ``None``.\n",
      "        axis : int, optional\n",
      "            Interpolation axis. Default is zero.\n",
      "        check_finite : bool, optional\n",
      "            Whether to check that the input arrays contain only finite numbers.\n",
      "            Disabling may give a performance gain, but may result in problems\n",
      "            (crashes, non-termination) if the inputs do contain infinities or NaNs.\n",
      "            Default is True.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        b : a BSpline object of the degree `k` with knots `t`.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        \n",
      "        The number of data points must be larger than the spline degree `k`.\n",
      "        \n",
      "        Knots `t` must satisfy the Schoenberg-Whitney conditions,\n",
      "        i.e., there must be a subset of data points ``x[j]`` such that\n",
      "        ``t[j] < x[j] < t[j+k+1]``, for ``j=0, 1,...,n-k-2``.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        Generate some noisy data:\n",
      "        \n",
      "        >>> rng = np.random.default_rng()\n",
      "        >>> x = np.linspace(-3, 3, 50)\n",
      "        >>> y = np.exp(-x**2) + 0.1 * rng.standard_normal(50)\n",
      "        \n",
      "        Now fit a smoothing cubic spline with a pre-defined internal knots.\n",
      "        Here we make the knot vector (k+1)-regular by adding boundary knots:\n",
      "        \n",
      "        >>> from scipy.interpolate import make_lsq_spline, BSpline\n",
      "        >>> t = [-1, 0, 1]\n",
      "        >>> k = 3\n",
      "        >>> t = np.r_[(x[0],)*(k+1),\n",
      "        ...           t,\n",
      "        ...           (x[-1],)*(k+1)]\n",
      "        >>> spl = make_lsq_spline(x, y, t, k)\n",
      "        \n",
      "        For comparison, we also construct an interpolating spline for the same\n",
      "        set of data:\n",
      "        \n",
      "        >>> from scipy.interpolate import make_interp_spline\n",
      "        >>> spl_i = make_interp_spline(x, y)\n",
      "        \n",
      "        Plot both:\n",
      "        \n",
      "        >>> import matplotlib.pyplot as plt\n",
      "        >>> xs = np.linspace(-3, 3, 100)\n",
      "        >>> plt.plot(x, y, 'ro', ms=5)\n",
      "        >>> plt.plot(xs, spl(xs), 'g-', lw=3, label='LSQ spline')\n",
      "        >>> plt.plot(xs, spl_i(xs), 'b-', lw=3, alpha=0.7, label='interp spline')\n",
      "        >>> plt.legend(loc='best')\n",
      "        >>> plt.show()\n",
      "        \n",
      "        **NaN handling**: If the input arrays contain ``nan`` values, the result is\n",
      "        not useful since the underlying spline fitting routines cannot deal with\n",
      "        ``nan``. A workaround is to use zero weights for not-a-number data points:\n",
      "        \n",
      "        >>> y[8] = np.nan\n",
      "        >>> w = np.isnan(y)\n",
      "        >>> y[w] = 0.\n",
      "        >>> tck = make_lsq_spline(x, y, t, w=~w)\n",
      "        \n",
      "        Notice the need to replace a ``nan`` by a numerical value (precise value\n",
      "        does not matter as long as the corresponding weight is zero.)\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        BSpline : base class representing the B-spline objects\n",
      "        make_interp_spline : a similar factory function for interpolating splines\n",
      "        LSQUnivariateSpline : a FITPACK-based spline fitting routine\n",
      "        splrep : a FITPACK-based fitting routine\n",
      "    \n",
      "    pade(an, m, n=None)\n",
      "        Return Pade approximation to a polynomial as the ratio of two polynomials.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        an : (N,) array_like\n",
      "            Taylor series coefficients.\n",
      "        m : int\n",
      "            The order of the returned approximating polynomial `q`.\n",
      "        n : int, optional\n",
      "            The order of the returned approximating polynomial `p`. By default,\n",
      "            the order is ``len(an)-m``.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        p, q : Polynomial class\n",
      "            The Pade approximation of the polynomial defined by `an` is\n",
      "            ``p(x)/q(x)``.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy.interpolate import pade\n",
      "        >>> e_exp = [1.0, 1.0, 1.0/2.0, 1.0/6.0, 1.0/24.0, 1.0/120.0]\n",
      "        >>> p, q = pade(e_exp, 2)\n",
      "        \n",
      "        >>> e_exp.reverse()\n",
      "        >>> e_poly = np.poly1d(e_exp)\n",
      "        \n",
      "        Compare ``e_poly(x)`` and the Pade approximation ``p(x)/q(x)``\n",
      "        \n",
      "        >>> e_poly(1)\n",
      "        2.7166666666666668\n",
      "        \n",
      "        >>> p(1)/q(1)\n",
      "        2.7179487179487181\n",
      "    \n",
      "    pchip_interpolate(xi, yi, x, der=0, axis=0)\n",
      "        Convenience function for pchip interpolation.\n",
      "        \n",
      "        xi and yi are arrays of values used to approximate some function f,\n",
      "        with ``yi = f(xi)``. The interpolant uses monotonic cubic splines\n",
      "        to find the value of new points x and the derivatives there.\n",
      "        \n",
      "        See `scipy.interpolate.PchipInterpolator` for details.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        xi : array_like\n",
      "            A sorted list of x-coordinates, of length N.\n",
      "        yi :  array_like\n",
      "            A 1-D array of real values. `yi`'s length along the interpolation\n",
      "            axis must be equal to the length of `xi`. If N-D array, use axis\n",
      "            parameter to select correct axis.\n",
      "        x : scalar or array_like\n",
      "            Of length M.\n",
      "        der : int or list, optional\n",
      "            Derivatives to extract. The 0th derivative can be included to\n",
      "            return the function value.\n",
      "        axis : int, optional\n",
      "            Axis in the yi array corresponding to the x-coordinate values.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        PchipInterpolator : PCHIP 1-D monotonic cubic interpolator.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        y : scalar or array_like\n",
      "            The result, of length R or length M or M by R,\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        We can interpolate 2D observed data using pchip interpolation:\n",
      "        \n",
      "        >>> import matplotlib.pyplot as plt\n",
      "        >>> from scipy.interpolate import pchip_interpolate\n",
      "        >>> x_observed = np.linspace(0.0, 10.0, 11)\n",
      "        >>> y_observed = np.sin(x_observed)\n",
      "        >>> x = np.linspace(min(x_observed), max(x_observed), num=100)\n",
      "        >>> y = pchip_interpolate(x_observed, y_observed, x)\n",
      "        >>> plt.plot(x_observed, y_observed, \"o\", label=\"observation\")\n",
      "        >>> plt.plot(x, y, label=\"pchip interpolation\")\n",
      "        >>> plt.legend()\n",
      "        >>> plt.show()\n",
      "    \n",
      "    spalde(x, tck)\n",
      "        Evaluate all derivatives of a B-spline.\n",
      "        \n",
      "        Given the knots and coefficients of a cubic B-spline compute all\n",
      "        derivatives up to order k at a point (or set of points).\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x : array_like\n",
      "            A point or a set of points at which to evaluate the derivatives.\n",
      "            Note that ``t(k) <= x <= t(n-k+1)`` must hold for each `x`.\n",
      "        tck : tuple\n",
      "            A tuple ``(t, c, k)``, containing the vector of knots, the B-spline\n",
      "            coefficients, and the degree of the spline (see `splev`).\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        results : {ndarray, list of ndarrays}\n",
      "            An array (or a list of arrays) containing all derivatives\n",
      "            up to order k inclusive for each point `x`.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        splprep, splrep, splint, sproot, splev, bisplrep, bisplev,\n",
      "        BSpline\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] C. de Boor: On calculating with b-splines, J. Approximation Theory\n",
      "           6 (1972) 50-62.\n",
      "        .. [2] M. G. Cox : The numerical evaluation of b-splines, J. Inst. Maths\n",
      "           applics 10 (1972) 134-149.\n",
      "        .. [3] P. Dierckx : Curve and surface fitting with splines, Monographs on\n",
      "           Numerical Analysis, Oxford University Press, 1993.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        Examples are given :ref:`in the tutorial <tutorial-interpolate_splXXX>`.\n",
      "    \n",
      "    splantider(tck, n=1)\n",
      "        Compute the spline for the antiderivative (integral) of a given spline.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        tck : BSpline instance or a tuple of (t, c, k)\n",
      "            Spline whose antiderivative to compute\n",
      "        n : int, optional\n",
      "            Order of antiderivative to evaluate. Default: 1\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        BSpline instance or a tuple of (t2, c2, k2)\n",
      "            Spline of order k2=k+n representing the antiderivative of the input\n",
      "            spline.\n",
      "            A tuple is returned iff the input argument `tck` is a tuple, otherwise\n",
      "            a BSpline object is constructed and returned.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        splder, splev, spalde\n",
      "        BSpline\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        The `splder` function is the inverse operation of this function.\n",
      "        Namely, ``splder(splantider(tck))`` is identical to `tck`, modulo\n",
      "        rounding error.\n",
      "        \n",
      "        .. versionadded:: 0.13.0\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy.interpolate import splrep, splder, splantider, splev\n",
      "        >>> x = np.linspace(0, np.pi/2, 70)\n",
      "        >>> y = 1 / np.sqrt(1 - 0.8*np.sin(x)**2)\n",
      "        >>> spl = splrep(x, y)\n",
      "        \n",
      "        The derivative is the inverse operation of the antiderivative,\n",
      "        although some floating point error accumulates:\n",
      "        \n",
      "        >>> splev(1.7, spl), splev(1.7, splder(splantider(spl)))\n",
      "        (array(2.1565429877197317), array(2.1565429877201865))\n",
      "        \n",
      "        Antiderivative can be used to evaluate definite integrals:\n",
      "        \n",
      "        >>> ispl = splantider(spl)\n",
      "        >>> splev(np.pi/2, ispl) - splev(0, ispl)\n",
      "        2.2572053588768486\n",
      "        \n",
      "        This is indeed an approximation to the complete elliptic integral\n",
      "        :math:`K(m) = \\int_0^{\\pi/2} [1 - m\\sin^2 x]^{-1/2} dx`:\n",
      "        \n",
      "        >>> from scipy.special import ellipk\n",
      "        >>> ellipk(0.8)\n",
      "        2.2572053268208538\n",
      "    \n",
      "    splder(tck, n=1)\n",
      "        Compute the spline representation of the derivative of a given spline\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        tck : BSpline instance or a tuple of (t, c, k)\n",
      "            Spline whose derivative to compute\n",
      "        n : int, optional\n",
      "            Order of derivative to evaluate. Default: 1\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        `BSpline` instance or tuple\n",
      "            Spline of order k2=k-n representing the derivative\n",
      "            of the input spline.\n",
      "            A tuple is returned iff the input argument `tck` is a tuple, otherwise\n",
      "            a BSpline object is constructed and returned.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        \n",
      "        .. versionadded:: 0.13.0\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        splantider, splev, spalde\n",
      "        BSpline\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        This can be used for finding maxima of a curve:\n",
      "        \n",
      "        >>> from scipy.interpolate import splrep, splder, sproot\n",
      "        >>> x = np.linspace(0, 10, 70)\n",
      "        >>> y = np.sin(x)\n",
      "        >>> spl = splrep(x, y, k=4)\n",
      "        \n",
      "        Now, differentiate the spline and find the zeros of the\n",
      "        derivative. (NB: `sproot` only works for order 3 splines, so we\n",
      "        fit an order 4 spline):\n",
      "        \n",
      "        >>> dspl = splder(spl)\n",
      "        >>> sproot(dspl) / np.pi\n",
      "        array([ 0.50000001,  1.5       ,  2.49999998])\n",
      "        \n",
      "        This agrees well with roots :math:`\\pi/2 + n\\pi` of\n",
      "        :math:`\\cos(x) = \\sin'(x)`.\n",
      "    \n",
      "    splev(x, tck, der=0, ext=0)\n",
      "        Evaluate a B-spline or its derivatives.\n",
      "        \n",
      "        Given the knots and coefficients of a B-spline representation, evaluate\n",
      "        the value of the smoothing polynomial and its derivatives. This is a\n",
      "        wrapper around the FORTRAN routines splev and splder of FITPACK.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x : array_like\n",
      "            An array of points at which to return the value of the smoothed\n",
      "            spline or its derivatives. If `tck` was returned from `splprep`,\n",
      "            then the parameter values, u should be given.\n",
      "        tck : 3-tuple or a BSpline object\n",
      "            If a tuple, then it should be a sequence of length 3 returned by\n",
      "            `splrep` or `splprep` containing the knots, coefficients, and degree\n",
      "            of the spline. (Also see Notes.)\n",
      "        der : int, optional\n",
      "            The order of derivative of the spline to compute (must be less than\n",
      "            or equal to k, the degree of the spline).\n",
      "        ext : int, optional\n",
      "            Controls the value returned for elements of ``x`` not in the\n",
      "            interval defined by the knot sequence.\n",
      "        \n",
      "            * if ext=0, return the extrapolated value.\n",
      "            * if ext=1, return 0\n",
      "            * if ext=2, raise a ValueError\n",
      "            * if ext=3, return the boundary value.\n",
      "        \n",
      "            The default value is 0.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        y : ndarray or list of ndarrays\n",
      "            An array of values representing the spline function evaluated at\n",
      "            the points in `x`.  If `tck` was returned from `splprep`, then this\n",
      "            is a list of arrays representing the curve in an N-D space.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        Manipulating the tck-tuples directly is not recommended. In new code,\n",
      "        prefer using `BSpline` objects.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        splprep, splrep, sproot, spalde, splint\n",
      "        bisplrep, bisplev\n",
      "        BSpline\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] C. de Boor, \"On calculating with b-splines\", J. Approximation\n",
      "            Theory, 6, p.50-62, 1972.\n",
      "        .. [2] M. G. Cox, \"The numerical evaluation of b-splines\", J. Inst. Maths\n",
      "            Applics, 10, p.134-149, 1972.\n",
      "        .. [3] P. Dierckx, \"Curve and surface fitting with splines\", Monographs\n",
      "            on Numerical Analysis, Oxford University Press, 1993.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        Examples are given :ref:`in the tutorial <tutorial-interpolate_splXXX>`.\n",
      "    \n",
      "    splint(a, b, tck, full_output=0)\n",
      "        Evaluate the definite integral of a B-spline between two given points.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        a, b : float\n",
      "            The end-points of the integration interval.\n",
      "        tck : tuple or a BSpline instance\n",
      "            If a tuple, then it should be a sequence of length 3, containing the\n",
      "            vector of knots, the B-spline coefficients, and the degree of the\n",
      "            spline (see `splev`).\n",
      "        full_output : int, optional\n",
      "            Non-zero to return optional output.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        integral : float\n",
      "            The resulting integral.\n",
      "        wrk : ndarray\n",
      "            An array containing the integrals of the normalized B-splines\n",
      "            defined on the set of knots.\n",
      "            (Only returned if `full_output` is non-zero)\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        `splint` silently assumes that the spline function is zero outside the data\n",
      "        interval (`a`, `b`).\n",
      "        \n",
      "        Manipulating the tck-tuples directly is not recommended. In new code,\n",
      "        prefer using the `BSpline` objects.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        splprep, splrep, sproot, spalde, splev\n",
      "        bisplrep, bisplev\n",
      "        BSpline\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] P.W. Gaffney, The calculation of indefinite integrals of b-splines\",\n",
      "            J. Inst. Maths Applics, 17, p.37-41, 1976.\n",
      "        .. [2] P. Dierckx, \"Curve and surface fitting with splines\", Monographs\n",
      "            on Numerical Analysis, Oxford University Press, 1993.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        Examples are given :ref:`in the tutorial <tutorial-interpolate_splXXX>`.\n",
      "    \n",
      "    splprep(x, w=None, u=None, ub=None, ue=None, k=3, task=0, s=None, t=None, full_output=0, nest=None, per=0, quiet=1)\n",
      "        Find the B-spline representation of an N-D curve.\n",
      "        \n",
      "        Given a list of N rank-1 arrays, `x`, which represent a curve in\n",
      "        N-D space parametrized by `u`, find a smooth approximating\n",
      "        spline curve g(`u`). Uses the FORTRAN routine parcur from FITPACK.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x : array_like\n",
      "            A list of sample vector arrays representing the curve.\n",
      "        w : array_like, optional\n",
      "            Strictly positive rank-1 array of weights the same length as `x[0]`.\n",
      "            The weights are used in computing the weighted least-squares spline\n",
      "            fit. If the errors in the `x` values have standard-deviation given by\n",
      "            the vector d, then `w` should be 1/d. Default is ``ones(len(x[0]))``.\n",
      "        u : array_like, optional\n",
      "            An array of parameter values. If not given, these values are\n",
      "            calculated automatically as ``M = len(x[0])``, where\n",
      "        \n",
      "                v[0] = 0\n",
      "        \n",
      "                v[i] = v[i-1] + distance(`x[i]`, `x[i-1]`)\n",
      "        \n",
      "                u[i] = v[i] / v[M-1]\n",
      "        \n",
      "        ub, ue : int, optional\n",
      "            The end-points of the parameters interval.  Defaults to\n",
      "            u[0] and u[-1].\n",
      "        k : int, optional\n",
      "            Degree of the spline. Cubic splines are recommended.\n",
      "            Even values of `k` should be avoided especially with a small s-value.\n",
      "            ``1 <= k <= 5``, default is 3.\n",
      "        task : int, optional\n",
      "            If task==0 (default), find t and c for a given smoothing factor, s.\n",
      "            If task==1, find t and c for another value of the smoothing factor, s.\n",
      "            There must have been a previous call with task=0 or task=1\n",
      "            for the same set of data.\n",
      "            If task=-1 find the weighted least square spline for a given set of\n",
      "            knots, t.\n",
      "        s : float, optional\n",
      "            A smoothing condition.  The amount of smoothness is determined by\n",
      "            satisfying the conditions: ``sum((w * (y - g))**2,axis=0) <= s``,\n",
      "            where g(x) is the smoothed interpolation of (x,y).  The user can\n",
      "            use `s` to control the trade-off between closeness and smoothness\n",
      "            of fit.  Larger `s` means more smoothing while smaller values of `s`\n",
      "            indicate less smoothing. Recommended values of `s` depend on the\n",
      "            weights, w.  If the weights represent the inverse of the\n",
      "            standard-deviation of y, then a good `s` value should be found in\n",
      "            the range ``(m-sqrt(2*m),m+sqrt(2*m))``, where m is the number of\n",
      "            data points in x, y, and w.\n",
      "        t : int, optional\n",
      "            The knots needed for task=-1.\n",
      "        full_output : int, optional\n",
      "            If non-zero, then return optional outputs.\n",
      "        nest : int, optional\n",
      "            An over-estimate of the total number of knots of the spline to\n",
      "            help in determining the storage space.  By default nest=m/2.\n",
      "            Always large enough is nest=m+k+1.\n",
      "        per : int, optional\n",
      "           If non-zero, data points are considered periodic with period\n",
      "           ``x[m-1] - x[0]`` and a smooth periodic spline approximation is\n",
      "           returned.  Values of ``y[m-1]`` and ``w[m-1]`` are not used.\n",
      "        quiet : int, optional\n",
      "             Non-zero to suppress messages.\n",
      "             This parameter is deprecated; use standard Python warning filters\n",
      "             instead.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        tck : tuple\n",
      "            (t,c,k) a tuple containing the vector of knots, the B-spline\n",
      "            coefficients, and the degree of the spline.\n",
      "        u : array\n",
      "            An array of the values of the parameter.\n",
      "        fp : float\n",
      "            The weighted sum of squared residuals of the spline approximation.\n",
      "        ier : int\n",
      "            An integer flag about splrep success.  Success is indicated\n",
      "            if ier<=0. If ier in [1,2,3] an error occurred but was not raised.\n",
      "            Otherwise an error is raised.\n",
      "        msg : str\n",
      "            A message corresponding to the integer flag, ier.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        splrep, splev, sproot, spalde, splint,\n",
      "        bisplrep, bisplev\n",
      "        UnivariateSpline, BivariateSpline\n",
      "        BSpline\n",
      "        make_interp_spline\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        See `splev` for evaluation of the spline and its derivatives.\n",
      "        The number of dimensions N must be smaller than 11.\n",
      "        \n",
      "        The number of coefficients in the `c` array is ``k+1`` less then the number\n",
      "        of knots, ``len(t)``. This is in contrast with `splrep`, which zero-pads\n",
      "        the array of coefficients to have the same length as the array of knots.\n",
      "        These additional coefficients are ignored by evaluation routines, `splev`\n",
      "        and `BSpline`.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] P. Dierckx, \"Algorithms for smoothing data with periodic and\n",
      "            parametric splines, Computer Graphics and Image Processing\",\n",
      "            20 (1982) 171-184.\n",
      "        .. [2] P. Dierckx, \"Algorithms for smoothing data with periodic and\n",
      "            parametric splines\", report tw55, Dept. Computer Science,\n",
      "            K.U.Leuven, 1981.\n",
      "        .. [3] P. Dierckx, \"Curve and surface fitting with splines\", Monographs on\n",
      "            Numerical Analysis, Oxford University Press, 1993.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        Generate a discretization of a limacon curve in the polar coordinates:\n",
      "        \n",
      "        >>> phi = np.linspace(0, 2.*np.pi, 40)\n",
      "        >>> r = 0.5 + np.cos(phi)         # polar coords\n",
      "        >>> x, y = r * np.cos(phi), r * np.sin(phi)    # convert to cartesian\n",
      "        \n",
      "        And interpolate:\n",
      "        \n",
      "        >>> from scipy.interpolate import splprep, splev\n",
      "        >>> tck, u = splprep([x, y], s=0)\n",
      "        >>> new_points = splev(u, tck)\n",
      "        \n",
      "        Notice that (i) we force interpolation by using `s=0`,\n",
      "        (ii) the parameterization, ``u``, is generated automatically.\n",
      "        Now plot the result:\n",
      "        \n",
      "        >>> import matplotlib.pyplot as plt\n",
      "        >>> fig, ax = plt.subplots()\n",
      "        >>> ax.plot(x, y, 'ro')\n",
      "        >>> ax.plot(new_points[0], new_points[1], 'r-')\n",
      "        >>> plt.show()\n",
      "    \n",
      "    splrep(x, y, w=None, xb=None, xe=None, k=3, task=0, s=None, t=None, full_output=0, per=0, quiet=1)\n",
      "        Find the B-spline representation of a 1-D curve.\n",
      "        \n",
      "        Given the set of data points ``(x[i], y[i])`` determine a smooth spline\n",
      "        approximation of degree k on the interval ``xb <= x <= xe``.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x, y : array_like\n",
      "            The data points defining a curve y = f(x).\n",
      "        w : array_like, optional\n",
      "            Strictly positive rank-1 array of weights the same length as x and y.\n",
      "            The weights are used in computing the weighted least-squares spline\n",
      "            fit. If the errors in the y values have standard-deviation given by the\n",
      "            vector d, then w should be 1/d. Default is ones(len(x)).\n",
      "        xb, xe : float, optional\n",
      "            The interval to fit.  If None, these default to x[0] and x[-1]\n",
      "            respectively.\n",
      "        k : int, optional\n",
      "            The degree of the spline fit. It is recommended to use cubic splines.\n",
      "            Even values of k should be avoided especially with small s values.\n",
      "            1 <= k <= 5\n",
      "        task : {1, 0, -1}, optional\n",
      "            If task==0 find t and c for a given smoothing factor, s.\n",
      "        \n",
      "            If task==1 find t and c for another value of the smoothing factor, s.\n",
      "            There must have been a previous call with task=0 or task=1 for the same\n",
      "            set of data (t will be stored an used internally)\n",
      "        \n",
      "            If task=-1 find the weighted least square spline for a given set of\n",
      "            knots, t. These should be interior knots as knots on the ends will be\n",
      "            added automatically.\n",
      "        s : float, optional\n",
      "            A smoothing condition. The amount of smoothness is determined by\n",
      "            satisfying the conditions: sum((w * (y - g))**2,axis=0) <= s where g(x)\n",
      "            is the smoothed interpolation of (x,y). The user can use s to control\n",
      "            the tradeoff between closeness and smoothness of fit. Larger s means\n",
      "            more smoothing while smaller values of s indicate less smoothing.\n",
      "            Recommended values of s depend on the weights, w. If the weights\n",
      "            represent the inverse of the standard-deviation of y, then a good s\n",
      "            value should be found in the range (m-sqrt(2*m),m+sqrt(2*m)) where m is\n",
      "            the number of datapoints in x, y, and w. default : s=m-sqrt(2*m) if\n",
      "            weights are supplied. s = 0.0 (interpolating) if no weights are\n",
      "            supplied.\n",
      "        t : array_like, optional\n",
      "            The knots needed for task=-1. If given then task is automatically set\n",
      "            to -1.\n",
      "        full_output : bool, optional\n",
      "            If non-zero, then return optional outputs.\n",
      "        per : bool, optional\n",
      "            If non-zero, data points are considered periodic with period x[m-1] -\n",
      "            x[0] and a smooth periodic spline approximation is returned. Values of\n",
      "            y[m-1] and w[m-1] are not used.\n",
      "        quiet : bool, optional\n",
      "            Non-zero to suppress messages.\n",
      "            This parameter is deprecated; use standard Python warning filters\n",
      "            instead.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        tck : tuple\n",
      "            A tuple (t,c,k) containing the vector of knots, the B-spline\n",
      "            coefficients, and the degree of the spline.\n",
      "        fp : array, optional\n",
      "            The weighted sum of squared residuals of the spline approximation.\n",
      "        ier : int, optional\n",
      "            An integer flag about splrep success. Success is indicated if ier<=0.\n",
      "            If ier in [1,2,3] an error occurred but was not raised. Otherwise an\n",
      "            error is raised.\n",
      "        msg : str, optional\n",
      "            A message corresponding to the integer flag, ier.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        UnivariateSpline, BivariateSpline\n",
      "        splprep, splev, sproot, spalde, splint\n",
      "        bisplrep, bisplev\n",
      "        BSpline\n",
      "        make_interp_spline\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        See `splev` for evaluation of the spline and its derivatives. Uses the\n",
      "        FORTRAN routine ``curfit`` from FITPACK.\n",
      "        \n",
      "        The user is responsible for assuring that the values of `x` are unique.\n",
      "        Otherwise, `splrep` will not return sensible results.\n",
      "        \n",
      "        If provided, knots `t` must satisfy the Schoenberg-Whitney conditions,\n",
      "        i.e., there must be a subset of data points ``x[j]`` such that\n",
      "        ``t[j] < x[j] < t[j+k+1]``, for ``j=0, 1,...,n-k-2``.\n",
      "        \n",
      "        This routine zero-pads the coefficients array ``c`` to have the same length\n",
      "        as the array of knots ``t`` (the trailing ``k + 1`` coefficients are ignored\n",
      "        by the evaluation routines, `splev` and `BSpline`.) This is in contrast with\n",
      "        `splprep`, which does not zero-pad the coefficients.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        Based on algorithms described in [1]_, [2]_, [3]_, and [4]_:\n",
      "        \n",
      "        .. [1] P. Dierckx, \"An algorithm for smoothing, differentiation and\n",
      "           integration of experimental data using spline functions\",\n",
      "           J.Comp.Appl.Maths 1 (1975) 165-184.\n",
      "        .. [2] P. Dierckx, \"A fast algorithm for smoothing data on a rectangular\n",
      "           grid while using spline functions\", SIAM J.Numer.Anal. 19 (1982)\n",
      "           1286-1304.\n",
      "        .. [3] P. Dierckx, \"An improved algorithm for curve fitting with spline\n",
      "           functions\", report tw54, Dept. Computer Science,K.U. Leuven, 1981.\n",
      "        .. [4] P. Dierckx, \"Curve and surface fitting with splines\", Monographs on\n",
      "           Numerical Analysis, Oxford University Press, 1993.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        You can interpolate 1-D points with a B-spline curve.\n",
      "        Further examples are given in\n",
      "        :ref:`in the tutorial <tutorial-interpolate_splXXX>`.\n",
      "        \n",
      "        >>> import matplotlib.pyplot as plt\n",
      "        >>> from scipy.interpolate import splev, splrep\n",
      "        >>> x = np.linspace(0, 10, 10)\n",
      "        >>> y = np.sin(x)\n",
      "        >>> spl = splrep(x, y)\n",
      "        >>> x2 = np.linspace(0, 10, 200)\n",
      "        >>> y2 = splev(x2, spl)\n",
      "        >>> plt.plot(x, y, 'o', x2, y2)\n",
      "        >>> plt.show()\n",
      "    \n",
      "    sproot(tck, mest=10)\n",
      "        Find the roots of a cubic B-spline.\n",
      "        \n",
      "        Given the knots (>=8) and coefficients of a cubic B-spline return the\n",
      "        roots of the spline.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        tck : tuple or a BSpline object\n",
      "            If a tuple, then it should be a sequence of length 3, containing the\n",
      "            vector of knots, the B-spline coefficients, and the degree of the\n",
      "            spline.\n",
      "            The number of knots must be >= 8, and the degree must be 3.\n",
      "            The knots must be a montonically increasing sequence.\n",
      "        mest : int, optional\n",
      "            An estimate of the number of zeros (Default is 10).\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        zeros : ndarray\n",
      "            An array giving the roots of the spline.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        Manipulating the tck-tuples directly is not recommended. In new code,\n",
      "        prefer using the `BSpline` objects.\n",
      "        \n",
      "        See also\n",
      "        --------\n",
      "        splprep, splrep, splint, spalde, splev\n",
      "        bisplrep, bisplev\n",
      "        BSpline\n",
      "        \n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] C. de Boor, \"On calculating with b-splines\", J. Approximation\n",
      "            Theory, 6, p.50-62, 1972.\n",
      "        .. [2] M. G. Cox, \"The numerical evaluation of b-splines\", J. Inst. Maths\n",
      "            Applics, 10, p.134-149, 1972.\n",
      "        .. [3] P. Dierckx, \"Curve and surface fitting with splines\", Monographs\n",
      "            on Numerical Analysis, Oxford University Press, 1993.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        Examples are given :ref:`in the tutorial <tutorial-interpolate_splXXX>`.\n",
      "\n",
      "DATA\n",
      "    __all__ = ['Akima1DInterpolator', 'BPoly', 'BSpline', 'BarycentricInte...\n",
      "\n",
      "FILE\n",
      "    c:\\programdata\\anaconda3\\lib\\site-packages\\scipy\\interpolate\\__init__.py\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(sci)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "b92b78dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scipy.stats as stat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "b04756cf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on package scipy.stats in scipy:\n",
      "\n",
      "NAME\n",
      "    scipy.stats - .. _statsrefmanual:\n",
      "\n",
      "DESCRIPTION\n",
      "    ==========================================\n",
      "    Statistical functions (:mod:`scipy.stats`)\n",
      "    ==========================================\n",
      "    \n",
      "    .. currentmodule:: scipy.stats\n",
      "    \n",
      "    This module contains a large number of probability distributions,\n",
      "    summary and frequency statistics, correlation functions and statistical\n",
      "    tests, masked statistics, kernel density estimation, quasi-Monte Carlo\n",
      "    functionality, and more.\n",
      "    \n",
      "    Statistics is a very large area, and there are topics that are out of scope\n",
      "    for SciPy and are covered by other packages. Some of the most important ones\n",
      "    are:\n",
      "    \n",
      "    - `statsmodels <https://www.statsmodels.org/stable/index.html>`__:\n",
      "      regression, linear models, time series analysis, extensions to topics\n",
      "      also covered by ``scipy.stats``.\n",
      "    - `Pandas <https://pandas.pydata.org/>`__: tabular data, time series\n",
      "      functionality, interfaces to other statistical languages.\n",
      "    - `PyMC3 <https://docs.pymc.io/>`__: Bayesian statistical\n",
      "      modeling, probabilistic machine learning.\n",
      "    - `scikit-learn <https://scikit-learn.org/>`__: classification, regression,\n",
      "      model selection.\n",
      "    - `Seaborn <https://seaborn.pydata.org/>`__: statistical data visualization.\n",
      "    - `rpy2 <https://rpy2.github.io/>`__: Python to R bridge.\n",
      "    \n",
      "    \n",
      "    Probability distributions\n",
      "    =========================\n",
      "    \n",
      "    Each univariate distribution is an instance of a subclass of `rv_continuous`\n",
      "    (`rv_discrete` for discrete distributions):\n",
      "    \n",
      "    .. autosummary::\n",
      "       :toctree: generated/\n",
      "    \n",
      "       rv_continuous\n",
      "       rv_discrete\n",
      "       rv_histogram\n",
      "    \n",
      "    Continuous distributions\n",
      "    ------------------------\n",
      "    \n",
      "    .. autosummary::\n",
      "       :toctree: generated/\n",
      "    \n",
      "       alpha             -- Alpha\n",
      "       anglit            -- Anglit\n",
      "       arcsine           -- Arcsine\n",
      "       argus             -- Argus\n",
      "       beta              -- Beta\n",
      "       betaprime         -- Beta Prime\n",
      "       bradford          -- Bradford\n",
      "       burr              -- Burr (Type III)\n",
      "       burr12            -- Burr (Type XII)\n",
      "       cauchy            -- Cauchy\n",
      "       chi               -- Chi\n",
      "       chi2              -- Chi-squared\n",
      "       cosine            -- Cosine\n",
      "       crystalball       -- Crystalball\n",
      "       dgamma            -- Double Gamma\n",
      "       dweibull          -- Double Weibull\n",
      "       erlang            -- Erlang\n",
      "       expon             -- Exponential\n",
      "       exponnorm         -- Exponentially Modified Normal\n",
      "       exponweib         -- Exponentiated Weibull\n",
      "       exponpow          -- Exponential Power\n",
      "       f                 -- F (Snecdor F)\n",
      "       fatiguelife       -- Fatigue Life (Birnbaum-Saunders)\n",
      "       fisk              -- Fisk\n",
      "       foldcauchy        -- Folded Cauchy\n",
      "       foldnorm          -- Folded Normal\n",
      "       genlogistic       -- Generalized Logistic\n",
      "       gennorm           -- Generalized normal\n",
      "       genpareto         -- Generalized Pareto\n",
      "       genexpon          -- Generalized Exponential\n",
      "       genextreme        -- Generalized Extreme Value\n",
      "       gausshyper        -- Gauss Hypergeometric\n",
      "       gamma             -- Gamma\n",
      "       gengamma          -- Generalized gamma\n",
      "       genhalflogistic   -- Generalized Half Logistic\n",
      "       genhyperbolic     -- Generalized Hyperbolic\n",
      "       geninvgauss       -- Generalized Inverse Gaussian\n",
      "       gilbrat           -- Gilbrat\n",
      "       gompertz          -- Gompertz (Truncated Gumbel)\n",
      "       gumbel_r          -- Right Sided Gumbel, Log-Weibull, Fisher-Tippett, Extreme Value Type I\n",
      "       gumbel_l          -- Left Sided Gumbel, etc.\n",
      "       halfcauchy        -- Half Cauchy\n",
      "       halflogistic      -- Half Logistic\n",
      "       halfnorm          -- Half Normal\n",
      "       halfgennorm       -- Generalized Half Normal\n",
      "       hypsecant         -- Hyperbolic Secant\n",
      "       invgamma          -- Inverse Gamma\n",
      "       invgauss          -- Inverse Gaussian\n",
      "       invweibull        -- Inverse Weibull\n",
      "       johnsonsb         -- Johnson SB\n",
      "       johnsonsu         -- Johnson SU\n",
      "       kappa4            -- Kappa 4 parameter\n",
      "       kappa3            -- Kappa 3 parameter\n",
      "       ksone             -- Distribution of Kolmogorov-Smirnov one-sided test statistic\n",
      "       kstwo             -- Distribution of Kolmogorov-Smirnov two-sided test statistic\n",
      "       kstwobign         -- Limiting Distribution of scaled Kolmogorov-Smirnov two-sided test statistic.\n",
      "       laplace           -- Laplace\n",
      "       laplace_asymmetric    -- Asymmetric Laplace\n",
      "       levy              -- Levy\n",
      "       levy_l\n",
      "       levy_stable\n",
      "       logistic          -- Logistic\n",
      "       loggamma          -- Log-Gamma\n",
      "       loglaplace        -- Log-Laplace (Log Double Exponential)\n",
      "       lognorm           -- Log-Normal\n",
      "       loguniform        -- Log-Uniform\n",
      "       lomax             -- Lomax (Pareto of the second kind)\n",
      "       maxwell           -- Maxwell\n",
      "       mielke            -- Mielke's Beta-Kappa\n",
      "       moyal             -- Moyal\n",
      "       nakagami          -- Nakagami\n",
      "       ncx2              -- Non-central chi-squared\n",
      "       ncf               -- Non-central F\n",
      "       nct               -- Non-central Student's T\n",
      "       norm              -- Normal (Gaussian)\n",
      "       norminvgauss      -- Normal Inverse Gaussian\n",
      "       pareto            -- Pareto\n",
      "       pearson3          -- Pearson type III\n",
      "       powerlaw          -- Power-function\n",
      "       powerlognorm      -- Power log normal\n",
      "       powernorm         -- Power normal\n",
      "       rdist             -- R-distribution\n",
      "       rayleigh          -- Rayleigh\n",
      "       rice              -- Rice\n",
      "       recipinvgauss     -- Reciprocal Inverse Gaussian\n",
      "       semicircular      -- Semicircular\n",
      "       skewcauchy        -- Skew Cauchy\n",
      "       skewnorm          -- Skew normal\n",
      "       studentized_range    -- Studentized Range\n",
      "       t                 -- Student's T\n",
      "       trapezoid         -- Trapezoidal\n",
      "       triang            -- Triangular\n",
      "       truncexpon        -- Truncated Exponential\n",
      "       truncnorm         -- Truncated Normal\n",
      "       tukeylambda       -- Tukey-Lambda\n",
      "       uniform           -- Uniform\n",
      "       vonmises          -- Von-Mises (Circular)\n",
      "       vonmises_line     -- Von-Mises (Line)\n",
      "       wald              -- Wald\n",
      "       weibull_min       -- Minimum Weibull (see Frechet)\n",
      "       weibull_max       -- Maximum Weibull (see Frechet)\n",
      "       wrapcauchy        -- Wrapped Cauchy\n",
      "    \n",
      "    Multivariate distributions\n",
      "    --------------------------\n",
      "    \n",
      "    .. autosummary::\n",
      "       :toctree: generated/\n",
      "    \n",
      "       multivariate_normal    -- Multivariate normal distribution\n",
      "       matrix_normal          -- Matrix normal distribution\n",
      "       dirichlet              -- Dirichlet\n",
      "       wishart                -- Wishart\n",
      "       invwishart             -- Inverse Wishart\n",
      "       multinomial            -- Multinomial distribution\n",
      "       special_ortho_group    -- SO(N) group\n",
      "       ortho_group            -- O(N) group\n",
      "       unitary_group          -- U(N) group\n",
      "       random_correlation     -- random correlation matrices\n",
      "       multivariate_t         -- Multivariate t-distribution\n",
      "       multivariate_hypergeom -- Multivariate hypergeometric distribution\n",
      "    \n",
      "    Discrete distributions\n",
      "    ----------------------\n",
      "    \n",
      "    .. autosummary::\n",
      "       :toctree: generated/\n",
      "    \n",
      "       bernoulli                -- Bernoulli\n",
      "       betabinom                -- Beta-Binomial\n",
      "       binom                    -- Binomial\n",
      "       boltzmann                -- Boltzmann (Truncated Discrete Exponential)\n",
      "       dlaplace                 -- Discrete Laplacian\n",
      "       geom                     -- Geometric\n",
      "       hypergeom                -- Hypergeometric\n",
      "       logser                   -- Logarithmic (Log-Series, Series)\n",
      "       nbinom                   -- Negative Binomial\n",
      "       nchypergeom_fisher       -- Fisher's Noncentral Hypergeometric\n",
      "       nchypergeom_wallenius    -- Wallenius's Noncentral Hypergeometric\n",
      "       nhypergeom               -- Negative Hypergeometric\n",
      "       planck                   -- Planck (Discrete Exponential)\n",
      "       poisson                  -- Poisson\n",
      "       randint                  -- Discrete Uniform\n",
      "       skellam                  -- Skellam\n",
      "       yulesimon                -- Yule-Simon\n",
      "       zipf                     -- Zipf (Zeta)\n",
      "       zipfian                  -- Zipfian\n",
      "    \n",
      "    An overview of statistical functions is given below.  Many of these functions\n",
      "    have a similar version in `scipy.stats.mstats` which work for masked arrays.\n",
      "    \n",
      "    Summary statistics\n",
      "    ==================\n",
      "    \n",
      "    .. autosummary::\n",
      "       :toctree: generated/\n",
      "    \n",
      "       describe          -- Descriptive statistics\n",
      "       gmean             -- Geometric mean\n",
      "       hmean             -- Harmonic mean\n",
      "       kurtosis          -- Fisher or Pearson kurtosis\n",
      "       mode              -- Modal value\n",
      "       moment            -- Central moment\n",
      "       skew              -- Skewness\n",
      "       kstat             --\n",
      "       kstatvar          --\n",
      "       tmean             -- Truncated arithmetic mean\n",
      "       tvar              -- Truncated variance\n",
      "       tmin              --\n",
      "       tmax              --\n",
      "       tstd              --\n",
      "       tsem              --\n",
      "       variation         -- Coefficient of variation\n",
      "       find_repeats\n",
      "       trim_mean\n",
      "       gstd              -- Geometric Standard Deviation\n",
      "       iqr\n",
      "       sem\n",
      "       bayes_mvs\n",
      "       mvsdist\n",
      "       entropy\n",
      "       differential_entropy\n",
      "       median_absolute_deviation\n",
      "       median_abs_deviation\n",
      "       bootstrap\n",
      "    \n",
      "    Frequency statistics\n",
      "    ====================\n",
      "    \n",
      "    .. autosummary::\n",
      "       :toctree: generated/\n",
      "    \n",
      "       cumfreq\n",
      "       itemfreq\n",
      "       percentileofscore\n",
      "       scoreatpercentile\n",
      "       relfreq\n",
      "    \n",
      "    .. autosummary::\n",
      "       :toctree: generated/\n",
      "    \n",
      "       binned_statistic     -- Compute a binned statistic for a set of data.\n",
      "       binned_statistic_2d  -- Compute a 2-D binned statistic for a set of data.\n",
      "       binned_statistic_dd  -- Compute a d-D binned statistic for a set of data.\n",
      "    \n",
      "    Correlation functions\n",
      "    =====================\n",
      "    \n",
      "    .. autosummary::\n",
      "       :toctree: generated/\n",
      "    \n",
      "       f_oneway\n",
      "       alexandergovern\n",
      "       pearsonr\n",
      "       spearmanr\n",
      "       pointbiserialr\n",
      "       kendalltau\n",
      "       weightedtau\n",
      "       somersd\n",
      "       linregress\n",
      "       siegelslopes\n",
      "       theilslopes\n",
      "       multiscale_graphcorr\n",
      "    \n",
      "    Statistical tests\n",
      "    =================\n",
      "    \n",
      "    .. autosummary::\n",
      "       :toctree: generated/\n",
      "    \n",
      "       ttest_1samp\n",
      "       ttest_ind\n",
      "       ttest_ind_from_stats\n",
      "       ttest_rel\n",
      "       chisquare\n",
      "       cramervonmises\n",
      "       cramervonmises_2samp\n",
      "       power_divergence\n",
      "       kstest\n",
      "       ks_1samp\n",
      "       ks_2samp\n",
      "       epps_singleton_2samp\n",
      "       mannwhitneyu\n",
      "       tiecorrect\n",
      "       rankdata\n",
      "       ranksums\n",
      "       wilcoxon\n",
      "       kruskal\n",
      "       friedmanchisquare\n",
      "       brunnermunzel\n",
      "       combine_pvalues\n",
      "       jarque_bera\n",
      "       page_trend_test\n",
      "    \n",
      "    .. autosummary::\n",
      "       :toctree: generated/\n",
      "    \n",
      "       ansari\n",
      "       bartlett\n",
      "       levene\n",
      "       shapiro\n",
      "       anderson\n",
      "       anderson_ksamp\n",
      "       binom_test\n",
      "       binomtest\n",
      "       fligner\n",
      "       median_test\n",
      "       mood\n",
      "       skewtest\n",
      "       kurtosistest\n",
      "       normaltest\n",
      "    \n",
      "    \n",
      "    Quasi-Monte Carlo\n",
      "    =================\n",
      "    \n",
      "    .. toctree::\n",
      "       :maxdepth: 4\n",
      "    \n",
      "       stats.qmc\n",
      "    \n",
      "    \n",
      "    Masked statistics functions\n",
      "    ===========================\n",
      "    \n",
      "    .. toctree::\n",
      "    \n",
      "       stats.mstats\n",
      "    \n",
      "    \n",
      "    Other statistical functionality\n",
      "    ===============================\n",
      "    \n",
      "    Transformations\n",
      "    ---------------\n",
      "    \n",
      "    .. autosummary::\n",
      "       :toctree: generated/\n",
      "    \n",
      "       boxcox\n",
      "       boxcox_normmax\n",
      "       boxcox_llf\n",
      "       yeojohnson\n",
      "       yeojohnson_normmax\n",
      "       yeojohnson_llf\n",
      "       obrientransform\n",
      "       sigmaclip\n",
      "       trimboth\n",
      "       trim1\n",
      "       zmap\n",
      "       zscore\n",
      "    \n",
      "    Statistical distances\n",
      "    ---------------------\n",
      "    \n",
      "    .. autosummary::\n",
      "       :toctree: generated/\n",
      "    \n",
      "       wasserstein_distance\n",
      "       energy_distance\n",
      "    \n",
      "    Random variate generation / CDF Inversion\n",
      "    =========================================\n",
      "    \n",
      "    .. autosummary::\n",
      "       :toctree: generated/\n",
      "    \n",
      "       rvs_ratio_uniforms\n",
      "       NumericalInverseHermite\n",
      "    \n",
      "    Circular statistical functions\n",
      "    ------------------------------\n",
      "    \n",
      "    .. autosummary::\n",
      "       :toctree: generated/\n",
      "    \n",
      "       circmean\n",
      "       circvar\n",
      "       circstd\n",
      "    \n",
      "    Contingency table functions\n",
      "    ---------------------------\n",
      "    \n",
      "    .. autosummary::\n",
      "       :toctree: generated/\n",
      "    \n",
      "       chi2_contingency\n",
      "       contingency.crosstab\n",
      "       contingency.expected_freq\n",
      "       contingency.margins\n",
      "       contingency.relative_risk\n",
      "       contingency.association\n",
      "       fisher_exact\n",
      "       barnard_exact\n",
      "       boschloo_exact\n",
      "    \n",
      "    Plot-tests\n",
      "    ----------\n",
      "    \n",
      "    .. autosummary::\n",
      "       :toctree: generated/\n",
      "    \n",
      "       ppcc_max\n",
      "       ppcc_plot\n",
      "       probplot\n",
      "       boxcox_normplot\n",
      "       yeojohnson_normplot\n",
      "    \n",
      "    Univariate and multivariate kernel density estimation\n",
      "    -----------------------------------------------------\n",
      "    \n",
      "    .. autosummary::\n",
      "       :toctree: generated/\n",
      "    \n",
      "       gaussian_kde\n",
      "    \n",
      "    Warnings used in :mod:`scipy.stats`\n",
      "    -----------------------------------\n",
      "    \n",
      "    .. autosummary::\n",
      "       :toctree: generated/\n",
      "    \n",
      "       F_onewayConstantInputWarning\n",
      "       F_onewayBadInputSizesWarning\n",
      "       PearsonRConstantInputWarning\n",
      "       PearsonRNearConstantInputWarning\n",
      "       SpearmanRConstantInputWarning\n",
      "\n",
      "PACKAGE CONTENTS\n",
      "    _binned_statistic\n",
      "    _binomtest\n",
      "    _boost (package)\n",
      "    _bootstrap\n",
      "    _common\n",
      "    _constants\n",
      "    _continuous_distns\n",
      "    _crosstab\n",
      "    _discrete_distns\n",
      "    _distn_infrastructure\n",
      "    _distr_params\n",
      "    _entropy\n",
      "    _generate_pyx\n",
      "    _hypotests\n",
      "    _ksstats\n",
      "    _mannwhitneyu\n",
      "    _multivariate\n",
      "    _page_trend_test\n",
      "    _qmc\n",
      "    _qmc_cy\n",
      "    _relative_risk\n",
      "    _result_classes\n",
      "    _rvs_sampling\n",
      "    _sobol\n",
      "    _stats\n",
      "    _stats_mstats_common\n",
      "    _tukeylambda_stats\n",
      "    _wilcoxon_data\n",
      "    biasedurn\n",
      "    contingency\n",
      "    distributions\n",
      "    kde\n",
      "    morestats\n",
      "    mstats\n",
      "    mstats_basic\n",
      "    mstats_extras\n",
      "    mvn\n",
      "    qmc\n",
      "    setup\n",
      "    statlib\n",
      "    stats\n",
      "    tests (package)\n",
      "\n",
      "CLASSES\n",
      "    builtins.RuntimeWarning(builtins.Warning)\n",
      "        scipy.stats.stats.F_onewayBadInputSizesWarning\n",
      "        scipy.stats.stats.F_onewayConstantInputWarning\n",
      "        scipy.stats.stats.PearsonRConstantInputWarning\n",
      "        scipy.stats.stats.PearsonRNearConstantInputWarning\n",
      "        scipy.stats.stats.SpearmanRConstantInputWarning\n",
      "    builtins.object\n",
      "        scipy.stats._rvs_sampling.NumericalInverseHermite\n",
      "        scipy.stats.kde.gaussian_kde\n",
      "    scipy.stats._distn_infrastructure.rv_generic(builtins.object)\n",
      "        scipy.stats._distn_infrastructure.rv_continuous\n",
      "            scipy.stats._continuous_distns.rv_histogram\n",
      "        scipy.stats._distn_infrastructure.rv_discrete\n",
      "    \n",
      "    class F_onewayBadInputSizesWarning(builtins.RuntimeWarning)\n",
      "     |  Warning generated by `f_oneway` when an input has length 0,\n",
      "     |  or if all the inputs have length 1.\n",
      "     |  \n",
      "     |  Method resolution order:\n",
      "     |      F_onewayBadInputSizesWarning\n",
      "     |      builtins.RuntimeWarning\n",
      "     |      builtins.Warning\n",
      "     |      builtins.Exception\n",
      "     |      builtins.BaseException\n",
      "     |      builtins.object\n",
      "     |  \n",
      "     |  Data descriptors defined here:\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from builtins.RuntimeWarning:\n",
      "     |  \n",
      "     |  __init__(self, /, *args, **kwargs)\n",
      "     |      Initialize self.  See help(type(self)) for accurate signature.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Static methods inherited from builtins.RuntimeWarning:\n",
      "     |  \n",
      "     |  __new__(*args, **kwargs) from builtins.type\n",
      "     |      Create and return a new object.  See help(type) for accurate signature.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from builtins.BaseException:\n",
      "     |  \n",
      "     |  __delattr__(self, name, /)\n",
      "     |      Implement delattr(self, name).\n",
      "     |  \n",
      "     |  __getattribute__(self, name, /)\n",
      "     |      Return getattr(self, name).\n",
      "     |  \n",
      "     |  __reduce__(...)\n",
      "     |      Helper for pickle.\n",
      "     |  \n",
      "     |  __repr__(self, /)\n",
      "     |      Return repr(self).\n",
      "     |  \n",
      "     |  __setattr__(self, name, value, /)\n",
      "     |      Implement setattr(self, name, value).\n",
      "     |  \n",
      "     |  __setstate__(...)\n",
      "     |  \n",
      "     |  __str__(self, /)\n",
      "     |      Return str(self).\n",
      "     |  \n",
      "     |  with_traceback(...)\n",
      "     |      Exception.with_traceback(tb) --\n",
      "     |      set self.__traceback__ to tb and return self.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from builtins.BaseException:\n",
      "     |  \n",
      "     |  __cause__\n",
      "     |      exception cause\n",
      "     |  \n",
      "     |  __context__\n",
      "     |      exception context\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |  \n",
      "     |  __suppress_context__\n",
      "     |  \n",
      "     |  __traceback__\n",
      "     |  \n",
      "     |  args\n",
      "    \n",
      "    class F_onewayConstantInputWarning(builtins.RuntimeWarning)\n",
      "     |  F_onewayConstantInputWarning(msg=None)\n",
      "     |  \n",
      "     |  Warning generated by `f_oneway` when an input is constant, e.g.\n",
      "     |  each of the samples provided is a constant array.\n",
      "     |  \n",
      "     |  Method resolution order:\n",
      "     |      F_onewayConstantInputWarning\n",
      "     |      builtins.RuntimeWarning\n",
      "     |      builtins.Warning\n",
      "     |      builtins.Exception\n",
      "     |      builtins.BaseException\n",
      "     |      builtins.object\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __init__(self, msg=None)\n",
      "     |      Initialize self.  See help(type(self)) for accurate signature.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors defined here:\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Static methods inherited from builtins.RuntimeWarning:\n",
      "     |  \n",
      "     |  __new__(*args, **kwargs) from builtins.type\n",
      "     |      Create and return a new object.  See help(type) for accurate signature.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from builtins.BaseException:\n",
      "     |  \n",
      "     |  __delattr__(self, name, /)\n",
      "     |      Implement delattr(self, name).\n",
      "     |  \n",
      "     |  __getattribute__(self, name, /)\n",
      "     |      Return getattr(self, name).\n",
      "     |  \n",
      "     |  __reduce__(...)\n",
      "     |      Helper for pickle.\n",
      "     |  \n",
      "     |  __repr__(self, /)\n",
      "     |      Return repr(self).\n",
      "     |  \n",
      "     |  __setattr__(self, name, value, /)\n",
      "     |      Implement setattr(self, name, value).\n",
      "     |  \n",
      "     |  __setstate__(...)\n",
      "     |  \n",
      "     |  __str__(self, /)\n",
      "     |      Return str(self).\n",
      "     |  \n",
      "     |  with_traceback(...)\n",
      "     |      Exception.with_traceback(tb) --\n",
      "     |      set self.__traceback__ to tb and return self.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from builtins.BaseException:\n",
      "     |  \n",
      "     |  __cause__\n",
      "     |      exception cause\n",
      "     |  \n",
      "     |  __context__\n",
      "     |      exception context\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |  \n",
      "     |  __suppress_context__\n",
      "     |  \n",
      "     |  __traceback__\n",
      "     |  \n",
      "     |  args\n",
      "    \n",
      "    class NumericalInverseHermite(builtins.object)\n",
      "     |  NumericalInverseHermite(dist, *, tol=1e-12, max_intervals=100000)\n",
      "     |  \n",
      "     |  A Hermite spline fast numerical inverse of a probability distribution.\n",
      "     |  \n",
      "     |  The initializer of `NumericalInverseHermite` accepts `dist`, an object\n",
      "     |  representing a continuous distribution, and provides an object with methods\n",
      "     |  that approximate `dist.ppf` and `dist.rvs`. For most distributions,\n",
      "     |  these methods are faster than those of `dist` itself.\n",
      "     |  \n",
      "     |  Parameters\n",
      "     |  ----------\n",
      "     |  dist : object\n",
      "     |      Object representing the distribution for which a fast numerical inverse\n",
      "     |      is desired; for instance, a frozen instance of a `scipy.stats`\n",
      "     |      continuous distribution. See Notes and Examples for details.\n",
      "     |  tol : float, optional\n",
      "     |      u-error tolerance (see Notes). The default is 1e-12.\n",
      "     |  max_intervals : int, optional\n",
      "     |      Maximum number of intervals in the cubic Hermite spline used to\n",
      "     |      approximate the percent point function. The default is 100000.\n",
      "     |  \n",
      "     |  Attributes\n",
      "     |  ----------\n",
      "     |  intervals : int\n",
      "     |      The number of intervals of the interpolant.\n",
      "     |  midpoint_error : float\n",
      "     |      The maximum u-error at an interpolant interval midpoint.\n",
      "     |  \n",
      "     |  Notes\n",
      "     |  -----\n",
      "     |  `NumericalInverseHermite` approximates the inverse of a continuous\n",
      "     |  statistical distribution's CDF with a cubic Hermite spline.\n",
      "     |  \n",
      "     |  As described in [1]_, it begins by evaluating the distribution's PDF and\n",
      "     |  CDF at a mesh of quantiles ``x`` within the distribution's support.\n",
      "     |  It uses the results to fit a cubic Hermite spline ``H`` such that\n",
      "     |  ``H(p) == x``, where ``p`` is the array of percentiles corresponding\n",
      "     |  with the quantiles ``x``. Therefore, the spline approximates the inverse\n",
      "     |  of the distribution's CDF to machine precision at the percentiles ``p``,\n",
      "     |  but typically, the spline will not be as accurate at the midpoints between\n",
      "     |  the percentile points::\n",
      "     |  \n",
      "     |      p_mid = (p[:-1] + p[1:])/2\n",
      "     |  \n",
      "     |  so the mesh of quantiles is refined as needed to reduce the maximum\n",
      "     |  \"u-error\"::\n",
      "     |  \n",
      "     |      u_error = np.max(np.abs(dist.cdf(H(p_mid)) - p_mid))\n",
      "     |  \n",
      "     |  below the specified tolerance `tol`. Refinement stops when the required\n",
      "     |  tolerance is achieved or when the number of mesh intervals after the next\n",
      "     |  refinement could exceed the maximum allowed number `max_intervals`.\n",
      "     |  \n",
      "     |  The object `dist` must have methods ``pdf``, ``cdf``, and ``ppf`` that\n",
      "     |  behave like those of a *frozen* instance of `scipy.stats.rv_continuous`.\n",
      "     |  Specifically, it must have methods ``pdf`` and ``cdf`` that accept exactly\n",
      "     |  one ndarray argument ``x`` and return the probability density function and\n",
      "     |  cumulative density function (respectively) at ``x``. The object must also\n",
      "     |  have a method ``ppf`` that accepts a float ``p`` and returns the percentile\n",
      "     |  point function at ``p``. The object may also have a method ``isf`` that\n",
      "     |  accepts a float ``p`` and returns the inverse survival function at ``p``;\n",
      "     |  if it does not, it will be assigned an attribute ``isf`` that calculates\n",
      "     |  the inverse survival function using ``ppf``. The ``ppf`` and\n",
      "     |  ``isf` methods will each be evaluated at a small positive float ``p``\n",
      "     |  (e.g. ``p = utol/10``), and the domain over which the approximate numerical\n",
      "     |  inverse is defined will be ``ppf(p)`` to ``isf(p)``. The approximation will\n",
      "     |  not be accurate in the extreme tails beyond this domain.\n",
      "     |  \n",
      "     |  References\n",
      "     |  ----------\n",
      "     |  .. [1] Hörmann, Wolfgang, and Josef Leydold. \"Continuous random variate\n",
      "     |         generation by fast numerical inversion.\" ACM Transactions on\n",
      "     |         Modeling and Computer Simulation (TOMACS) 13.4 (2003): 347-362.\n",
      "     |  \n",
      "     |  Examples\n",
      "     |  --------\n",
      "     |  For some distributions, ``dist.ppf`` and ``dist.rvs`` are quite slow.\n",
      "     |  For instance, consider `scipy.stats.genexpon`. We freeze the distribution\n",
      "     |  by passing all shape parameters into its initializer and time the resulting\n",
      "     |  object's ``ppf`` and ``rvs`` functions.\n",
      "     |  \n",
      "     |  >>> import numpy as np\n",
      "     |  >>> from scipy import stats\n",
      "     |  >>> from timeit import timeit\n",
      "     |  >>> time_once = lambda f: f\"{timeit(f, number=1)*1000:.6} ms\"\n",
      "     |  >>> dist = stats.genexpon(9, 16, 3)  # freeze the distribution\n",
      "     |  >>> p = np.linspace(0.01, 0.99, 99)  # percentiles from 1% to 99%\n",
      "     |  >>> time_once(lambda: dist.ppf(p))\n",
      "     |  '154.565 ms'  # may vary\n",
      "     |  \n",
      "     |  >>> time_once(lambda: dist.rvs(size=100))\n",
      "     |  '148.979 ms'  # may vary\n",
      "     |  \n",
      "     |  The `NumericalInverseHermite` has a method that approximates ``dist.ppf``.\n",
      "     |  \n",
      "     |  >>> from scipy.stats import NumericalInverseHermite\n",
      "     |  >>> fni = NumericalInverseHermite(dist)\n",
      "     |  >>> np.allclose(fni.ppf(p), dist.ppf(p))\n",
      "     |  True\n",
      "     |  \n",
      "     |  In some cases, it is faster to both generate the fast numerical inverse\n",
      "     |  and use it than to call ``dist.ppf``.\n",
      "     |  \n",
      "     |  >>> def time_me():\n",
      "     |  ...     fni = NumericalInverseHermite(dist)\n",
      "     |  ...     fni.ppf(p)\n",
      "     |  >>> time_once(time_me)\n",
      "     |  '11.9222 ms'  # may vary\n",
      "     |  \n",
      "     |  After generating the fast numerical inverse, subsequent calls to its\n",
      "     |  methods are much faster.\n",
      "     |  >>> time_once(lambda: fni.ppf(p))\n",
      "     |  '0.0819 ms'  # may vary\n",
      "     |  \n",
      "     |  The fast numerical inverse can also be used to generate random variates\n",
      "     |  using inverse transform sampling.\n",
      "     |  \n",
      "     |  >>> time_once(lambda: fni.rvs(size=100))\n",
      "     |  '0.0911 ms'  # may vary\n",
      "     |  \n",
      "     |  Depending on the implementation of the distribution's random sampling\n",
      "     |  method, the random variates generated may be nearly identical, given\n",
      "     |  the same random state.\n",
      "     |  \n",
      "     |  >>> # `seed` ensures identical random streams are used by each `rvs` method\n",
      "     |  >>> seed = 500072020\n",
      "     |  >>> rvs1 = dist.rvs(size=100, random_state=np.random.default_rng(seed))\n",
      "     |  >>> rvs2 = fni.rvs(size=100, random_state=np.random.default_rng(seed))\n",
      "     |  >>> np.allclose(rvs1, rvs2)\n",
      "     |  True\n",
      "     |  \n",
      "     |  To use `NumericalInverseHermite` with a custom distribution, users may\n",
      "     |  subclass  `scipy.stats.rv_continuous` and initialize a frozen instance or\n",
      "     |  create an object with equivalent ``pdf``, ``cdf``, and ``ppf`` methods.\n",
      "     |  For instance, the following object represents the standard normal\n",
      "     |  distribution. For simplicity, we use `scipy.special.ndtr` and\n",
      "     |  `scipy.special.ndtri` to compute the ``cdf`` and ``ppf``, respectively.\n",
      "     |  \n",
      "     |  >>> from scipy.special import ndtr, ndtri\n",
      "     |  >>>\n",
      "     |  >>> class MyNormal:\n",
      "     |  ...\n",
      "     |  ...     def pdf(self, x):\n",
      "     |  ...        return 1/np.sqrt(2*np.pi) * np.exp(-x**2 / 2)\n",
      "     |  ...\n",
      "     |  ...     def cdf(self, x):\n",
      "     |  ...        return ndtr(x)\n",
      "     |  ...\n",
      "     |  ...     def ppf(self, x):\n",
      "     |  ...        return ndtri(x)\n",
      "     |  ...\n",
      "     |  >>> dist1 = MyNormal()\n",
      "     |  >>> fni1 = NumericalInverseHermite(dist1)\n",
      "     |  >>>\n",
      "     |  >>> dist2 = stats.norm()\n",
      "     |  >>> fni2 = NumericalInverseHermite(dist2)\n",
      "     |  >>>\n",
      "     |  >>> print(fni1.rvs(random_state=seed), fni2.rvs(random_state=seed))\n",
      "     |  -1.9603810921759424 -1.9603810921747074\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __init__(self, dist, *, tol=1e-12, max_intervals=100000)\n",
      "     |      Initialize self.  See help(type(self)) for accurate signature.\n",
      "     |  \n",
      "     |  ppf(self, q)\n",
      "     |      Approximate percent point function (inverse `cdf`) of the given RV.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      q : array_like\n",
      "     |          lower tail probability.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      x : array_like\n",
      "     |          quantile corresponding to the lower tail probability `q`.\n",
      "     |  \n",
      "     |  qrvs(self, size=None, d=None, qmc_engine=None)\n",
      "     |      Quasi-random variates of the given RV.\n",
      "     |      \n",
      "     |      The `qmc_engine` is used to draw uniform quasi-random variates, and\n",
      "     |      these are converted to quasi-random variates of the given RV using\n",
      "     |      inverse transform sampling.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      size : int, tuple of ints, or None; optional\n",
      "     |          Defines shape of random variates array. Default is ``None``.\n",
      "     |      d : int or None, optional\n",
      "     |          Defines dimension of uniform quasi-random variates to be\n",
      "     |          transformed. Default is ``None``.\n",
      "     |      qmc_engine : scipy.stats.qmc.QMCEngine(d=1), optional\n",
      "     |          Defines the object to use for drawing\n",
      "     |          quasi-random variates. Default is ``None``, which uses\n",
      "     |          `scipy.stats.qmc.Halton(1)`.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      rvs : ndarray or scalar\n",
      "     |          Quasi-random variates. See Notes for shape information.\n",
      "     |      \n",
      "     |      Notes\n",
      "     |      -----\n",
      "     |      The shape of the output array depends on `size`, `d`, and `qmc_engine`.\n",
      "     |      The intent is for the interface to be natural, but the detailed rules\n",
      "     |      to achieve this are complicated.\n",
      "     |      \n",
      "     |      - If `qmc_engine` is ``None``, a `scipy.stats.qmc.Halton` instance is\n",
      "     |        created with dimension `d`. If `d` is not provided, ``d=1``.\n",
      "     |      - If `qmc_engine` is not ``None`` and `d` is ``None``, `d` is\n",
      "     |        determined from the dimension of the `qmc_engine`.\n",
      "     |      - If `qmc_engine` is not ``None`` and `d` is not ``None`` but the\n",
      "     |        dimensions are inconsistent, a ``ValueError`` is raised.\n",
      "     |      - After `d` is determined according to the rules above, the output\n",
      "     |        shape is ``tuple_shape + d_shape``, where:\n",
      "     |      \n",
      "     |            - ``tuple_shape = tuple()`` if `size` is ``None``,\n",
      "     |            - ``tuple_shape = (size,)`` if `size` is an ``int``,\n",
      "     |            - ``tuple_shape = size`` if `size` is a sequence,\n",
      "     |            - ``d_shape = tuple()`` if `d` is ``None`` or `d` is 1, and\n",
      "     |            - ``d_shape = (d,)`` if `d` is greater than 1.\n",
      "     |      \n",
      "     |      The elements of the returned array are part of a low-discrepancy\n",
      "     |      sequence. If `d` is 1, this means that none of the samples are truly\n",
      "     |      independent. If `d` > 1, each slice ``rvs[..., i]`` will be of a\n",
      "     |      quasi-independent sequence; see `scipy.stats.qmc.QMCEngine` for\n",
      "     |      details. Note that when `d` > 1, the samples returned are still those\n",
      "     |      of the provided univariate distribution, not a multivariate\n",
      "     |      generalization of that distribution.\n",
      "     |  \n",
      "     |  rvs(self, size=None, random_state=None)\n",
      "     |      Random variates of the given RV.\n",
      "     |      \n",
      "     |      The `random_state` is used to draw uniform pseudo-random variates, and\n",
      "     |      these are converted to pseudo-random variates of the given RV using\n",
      "     |      inverse transform sampling.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      size : int, tuple of ints, or None; optional\n",
      "     |          Defines shape of array of random variates. Default is ``None``.\n",
      "     |      random_state : {None, int, `numpy.random.Generator`,\n",
      "     |                      `numpy.random.RandomState`}, optional\n",
      "     |      \n",
      "     |          Defines the object to use for drawing pseudorandom variates.\n",
      "     |          If `random_state` is ``None`` the `np.random.RandomState`\n",
      "     |          singleton is used.\n",
      "     |          If `random_state` is an ``int``, a new ``RandomState`` instance is\n",
      "     |          used, seeded with `random_state`.\n",
      "     |          If `random_state` is already a ``RandomState`` or ``Generator``\n",
      "     |          instance, then that object is used.\n",
      "     |          Default is None.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      rvs : ndarray or scalar\n",
      "     |          Random variates of given `size`. If `size` is ``None``, a scalar\n",
      "     |          is returned.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors defined here:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |      dictionary for instance variables (if defined)\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "    \n",
      "    class PearsonRConstantInputWarning(builtins.RuntimeWarning)\n",
      "     |  PearsonRConstantInputWarning(msg=None)\n",
      "     |  \n",
      "     |  Warning generated by `pearsonr` when an input is constant.\n",
      "     |  \n",
      "     |  Method resolution order:\n",
      "     |      PearsonRConstantInputWarning\n",
      "     |      builtins.RuntimeWarning\n",
      "     |      builtins.Warning\n",
      "     |      builtins.Exception\n",
      "     |      builtins.BaseException\n",
      "     |      builtins.object\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __init__(self, msg=None)\n",
      "     |      Initialize self.  See help(type(self)) for accurate signature.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors defined here:\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Static methods inherited from builtins.RuntimeWarning:\n",
      "     |  \n",
      "     |  __new__(*args, **kwargs) from builtins.type\n",
      "     |      Create and return a new object.  See help(type) for accurate signature.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from builtins.BaseException:\n",
      "     |  \n",
      "     |  __delattr__(self, name, /)\n",
      "     |      Implement delattr(self, name).\n",
      "     |  \n",
      "     |  __getattribute__(self, name, /)\n",
      "     |      Return getattr(self, name).\n",
      "     |  \n",
      "     |  __reduce__(...)\n",
      "     |      Helper for pickle.\n",
      "     |  \n",
      "     |  __repr__(self, /)\n",
      "     |      Return repr(self).\n",
      "     |  \n",
      "     |  __setattr__(self, name, value, /)\n",
      "     |      Implement setattr(self, name, value).\n",
      "     |  \n",
      "     |  __setstate__(...)\n",
      "     |  \n",
      "     |  __str__(self, /)\n",
      "     |      Return str(self).\n",
      "     |  \n",
      "     |  with_traceback(...)\n",
      "     |      Exception.with_traceback(tb) --\n",
      "     |      set self.__traceback__ to tb and return self.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from builtins.BaseException:\n",
      "     |  \n",
      "     |  __cause__\n",
      "     |      exception cause\n",
      "     |  \n",
      "     |  __context__\n",
      "     |      exception context\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |  \n",
      "     |  __suppress_context__\n",
      "     |  \n",
      "     |  __traceback__\n",
      "     |  \n",
      "     |  args\n",
      "    \n",
      "    class PearsonRNearConstantInputWarning(builtins.RuntimeWarning)\n",
      "     |  PearsonRNearConstantInputWarning(msg=None)\n",
      "     |  \n",
      "     |  Warning generated by `pearsonr` when an input is nearly constant.\n",
      "     |  \n",
      "     |  Method resolution order:\n",
      "     |      PearsonRNearConstantInputWarning\n",
      "     |      builtins.RuntimeWarning\n",
      "     |      builtins.Warning\n",
      "     |      builtins.Exception\n",
      "     |      builtins.BaseException\n",
      "     |      builtins.object\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __init__(self, msg=None)\n",
      "     |      Initialize self.  See help(type(self)) for accurate signature.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors defined here:\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Static methods inherited from builtins.RuntimeWarning:\n",
      "     |  \n",
      "     |  __new__(*args, **kwargs) from builtins.type\n",
      "     |      Create and return a new object.  See help(type) for accurate signature.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from builtins.BaseException:\n",
      "     |  \n",
      "     |  __delattr__(self, name, /)\n",
      "     |      Implement delattr(self, name).\n",
      "     |  \n",
      "     |  __getattribute__(self, name, /)\n",
      "     |      Return getattr(self, name).\n",
      "     |  \n",
      "     |  __reduce__(...)\n",
      "     |      Helper for pickle.\n",
      "     |  \n",
      "     |  __repr__(self, /)\n",
      "     |      Return repr(self).\n",
      "     |  \n",
      "     |  __setattr__(self, name, value, /)\n",
      "     |      Implement setattr(self, name, value).\n",
      "     |  \n",
      "     |  __setstate__(...)\n",
      "     |  \n",
      "     |  __str__(self, /)\n",
      "     |      Return str(self).\n",
      "     |  \n",
      "     |  with_traceback(...)\n",
      "     |      Exception.with_traceback(tb) --\n",
      "     |      set self.__traceback__ to tb and return self.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from builtins.BaseException:\n",
      "     |  \n",
      "     |  __cause__\n",
      "     |      exception cause\n",
      "     |  \n",
      "     |  __context__\n",
      "     |      exception context\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |  \n",
      "     |  __suppress_context__\n",
      "     |  \n",
      "     |  __traceback__\n",
      "     |  \n",
      "     |  args\n",
      "    \n",
      "    class SpearmanRConstantInputWarning(builtins.RuntimeWarning)\n",
      "     |  SpearmanRConstantInputWarning(msg=None)\n",
      "     |  \n",
      "     |  Warning generated by `spearmanr` when an input is constant.\n",
      "     |  \n",
      "     |  Method resolution order:\n",
      "     |      SpearmanRConstantInputWarning\n",
      "     |      builtins.RuntimeWarning\n",
      "     |      builtins.Warning\n",
      "     |      builtins.Exception\n",
      "     |      builtins.BaseException\n",
      "     |      builtins.object\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __init__(self, msg=None)\n",
      "     |      Initialize self.  See help(type(self)) for accurate signature.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors defined here:\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Static methods inherited from builtins.RuntimeWarning:\n",
      "     |  \n",
      "     |  __new__(*args, **kwargs) from builtins.type\n",
      "     |      Create and return a new object.  See help(type) for accurate signature.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from builtins.BaseException:\n",
      "     |  \n",
      "     |  __delattr__(self, name, /)\n",
      "     |      Implement delattr(self, name).\n",
      "     |  \n",
      "     |  __getattribute__(self, name, /)\n",
      "     |      Return getattr(self, name).\n",
      "     |  \n",
      "     |  __reduce__(...)\n",
      "     |      Helper for pickle.\n",
      "     |  \n",
      "     |  __repr__(self, /)\n",
      "     |      Return repr(self).\n",
      "     |  \n",
      "     |  __setattr__(self, name, value, /)\n",
      "     |      Implement setattr(self, name, value).\n",
      "     |  \n",
      "     |  __setstate__(...)\n",
      "     |  \n",
      "     |  __str__(self, /)\n",
      "     |      Return str(self).\n",
      "     |  \n",
      "     |  with_traceback(...)\n",
      "     |      Exception.with_traceback(tb) --\n",
      "     |      set self.__traceback__ to tb and return self.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from builtins.BaseException:\n",
      "     |  \n",
      "     |  __cause__\n",
      "     |      exception cause\n",
      "     |  \n",
      "     |  __context__\n",
      "     |      exception context\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |  \n",
      "     |  __suppress_context__\n",
      "     |  \n",
      "     |  __traceback__\n",
      "     |  \n",
      "     |  args\n",
      "    \n",
      "    class gaussian_kde(builtins.object)\n",
      "     |  gaussian_kde(dataset, bw_method=None, weights=None)\n",
      "     |  \n",
      "     |  Representation of a kernel-density estimate using Gaussian kernels.\n",
      "     |  \n",
      "     |  Kernel density estimation is a way to estimate the probability density\n",
      "     |  function (PDF) of a random variable in a non-parametric way.\n",
      "     |  `gaussian_kde` works for both uni-variate and multi-variate data.   It\n",
      "     |  includes automatic bandwidth determination.  The estimation works best for\n",
      "     |  a unimodal distribution; bimodal or multi-modal distributions tend to be\n",
      "     |  oversmoothed.\n",
      "     |  \n",
      "     |  Parameters\n",
      "     |  ----------\n",
      "     |  dataset : array_like\n",
      "     |      Datapoints to estimate from. In case of univariate data this is a 1-D\n",
      "     |      array, otherwise a 2-D array with shape (# of dims, # of data).\n",
      "     |  bw_method : str, scalar or callable, optional\n",
      "     |      The method used to calculate the estimator bandwidth.  This can be\n",
      "     |      'scott', 'silverman', a scalar constant or a callable.  If a scalar,\n",
      "     |      this will be used directly as `kde.factor`.  If a callable, it should\n",
      "     |      take a `gaussian_kde` instance as only parameter and return a scalar.\n",
      "     |      If None (default), 'scott' is used.  See Notes for more details.\n",
      "     |  weights : array_like, optional\n",
      "     |      weights of datapoints. This must be the same shape as dataset.\n",
      "     |      If None (default), the samples are assumed to be equally weighted\n",
      "     |  \n",
      "     |  Attributes\n",
      "     |  ----------\n",
      "     |  dataset : ndarray\n",
      "     |      The dataset with which `gaussian_kde` was initialized.\n",
      "     |  d : int\n",
      "     |      Number of dimensions.\n",
      "     |  n : int\n",
      "     |      Number of datapoints.\n",
      "     |  neff : int\n",
      "     |      Effective number of datapoints.\n",
      "     |  \n",
      "     |      .. versionadded:: 1.2.0\n",
      "     |  factor : float\n",
      "     |      The bandwidth factor, obtained from `kde.covariance_factor`, with which\n",
      "     |      the covariance matrix is multiplied.\n",
      "     |  covariance : ndarray\n",
      "     |      The covariance matrix of `dataset`, scaled by the calculated bandwidth\n",
      "     |      (`kde.factor`).\n",
      "     |  inv_cov : ndarray\n",
      "     |      The inverse of `covariance`.\n",
      "     |  \n",
      "     |  Methods\n",
      "     |  -------\n",
      "     |  evaluate\n",
      "     |  __call__\n",
      "     |  integrate_gaussian\n",
      "     |  integrate_box_1d\n",
      "     |  integrate_box\n",
      "     |  integrate_kde\n",
      "     |  pdf\n",
      "     |  logpdf\n",
      "     |  resample\n",
      "     |  set_bandwidth\n",
      "     |  covariance_factor\n",
      "     |  \n",
      "     |  Notes\n",
      "     |  -----\n",
      "     |  Bandwidth selection strongly influences the estimate obtained from the KDE\n",
      "     |  (much more so than the actual shape of the kernel).  Bandwidth selection\n",
      "     |  can be done by a \"rule of thumb\", by cross-validation, by \"plug-in\n",
      "     |  methods\" or by other means; see [3]_, [4]_ for reviews.  `gaussian_kde`\n",
      "     |  uses a rule of thumb, the default is Scott's Rule.\n",
      "     |  \n",
      "     |  Scott's Rule [1]_, implemented as `scotts_factor`, is::\n",
      "     |  \n",
      "     |      n**(-1./(d+4)),\n",
      "     |  \n",
      "     |  with ``n`` the number of data points and ``d`` the number of dimensions.\n",
      "     |  In the case of unequally weighted points, `scotts_factor` becomes::\n",
      "     |  \n",
      "     |      neff**(-1./(d+4)),\n",
      "     |  \n",
      "     |  with ``neff`` the effective number of datapoints.\n",
      "     |  Silverman's Rule [2]_, implemented as `silverman_factor`, is::\n",
      "     |  \n",
      "     |      (n * (d + 2) / 4.)**(-1. / (d + 4)).\n",
      "     |  \n",
      "     |  or in the case of unequally weighted points::\n",
      "     |  \n",
      "     |      (neff * (d + 2) / 4.)**(-1. / (d + 4)).\n",
      "     |  \n",
      "     |  Good general descriptions of kernel density estimation can be found in [1]_\n",
      "     |  and [2]_, the mathematics for this multi-dimensional implementation can be\n",
      "     |  found in [1]_.\n",
      "     |  \n",
      "     |  With a set of weighted samples, the effective number of datapoints ``neff``\n",
      "     |  is defined by::\n",
      "     |  \n",
      "     |      neff = sum(weights)^2 / sum(weights^2)\n",
      "     |  \n",
      "     |  as detailed in [5]_.\n",
      "     |  \n",
      "     |  References\n",
      "     |  ----------\n",
      "     |  .. [1] D.W. Scott, \"Multivariate Density Estimation: Theory, Practice, and\n",
      "     |         Visualization\", John Wiley & Sons, New York, Chicester, 1992.\n",
      "     |  .. [2] B.W. Silverman, \"Density Estimation for Statistics and Data\n",
      "     |         Analysis\", Vol. 26, Monographs on Statistics and Applied Probability,\n",
      "     |         Chapman and Hall, London, 1986.\n",
      "     |  .. [3] B.A. Turlach, \"Bandwidth Selection in Kernel Density Estimation: A\n",
      "     |         Review\", CORE and Institut de Statistique, Vol. 19, pp. 1-33, 1993.\n",
      "     |  .. [4] D.M. Bashtannyk and R.J. Hyndman, \"Bandwidth selection for kernel\n",
      "     |         conditional density estimation\", Computational Statistics & Data\n",
      "     |         Analysis, Vol. 36, pp. 279-298, 2001.\n",
      "     |  .. [5] Gray P. G., 1969, Journal of the Royal Statistical Society.\n",
      "     |         Series A (General), 132, 272\n",
      "     |  \n",
      "     |  Examples\n",
      "     |  --------\n",
      "     |  Generate some random two-dimensional data:\n",
      "     |  \n",
      "     |  >>> from scipy import stats\n",
      "     |  >>> def measure(n):\n",
      "     |  ...     \"Measurement model, return two coupled measurements.\"\n",
      "     |  ...     m1 = np.random.normal(size=n)\n",
      "     |  ...     m2 = np.random.normal(scale=0.5, size=n)\n",
      "     |  ...     return m1+m2, m1-m2\n",
      "     |  \n",
      "     |  >>> m1, m2 = measure(2000)\n",
      "     |  >>> xmin = m1.min()\n",
      "     |  >>> xmax = m1.max()\n",
      "     |  >>> ymin = m2.min()\n",
      "     |  >>> ymax = m2.max()\n",
      "     |  \n",
      "     |  Perform a kernel density estimate on the data:\n",
      "     |  \n",
      "     |  >>> X, Y = np.mgrid[xmin:xmax:100j, ymin:ymax:100j]\n",
      "     |  >>> positions = np.vstack([X.ravel(), Y.ravel()])\n",
      "     |  >>> values = np.vstack([m1, m2])\n",
      "     |  >>> kernel = stats.gaussian_kde(values)\n",
      "     |  >>> Z = np.reshape(kernel(positions).T, X.shape)\n",
      "     |  \n",
      "     |  Plot the results:\n",
      "     |  \n",
      "     |  >>> import matplotlib.pyplot as plt\n",
      "     |  >>> fig, ax = plt.subplots()\n",
      "     |  >>> ax.imshow(np.rot90(Z), cmap=plt.cm.gist_earth_r,\n",
      "     |  ...           extent=[xmin, xmax, ymin, ymax])\n",
      "     |  >>> ax.plot(m1, m2, 'k.', markersize=2)\n",
      "     |  >>> ax.set_xlim([xmin, xmax])\n",
      "     |  >>> ax.set_ylim([ymin, ymax])\n",
      "     |  >>> plt.show()\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __call__ = evaluate(self, points)\n",
      "     |  \n",
      "     |  __init__(self, dataset, bw_method=None, weights=None)\n",
      "     |      Initialize self.  See help(type(self)) for accurate signature.\n",
      "     |  \n",
      "     |  covariance_factor = scotts_factor(self)\n",
      "     |  \n",
      "     |  evaluate(self, points)\n",
      "     |      Evaluate the estimated pdf on a set of points.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      points : (# of dimensions, # of points)-array\n",
      "     |          Alternatively, a (# of dimensions,) vector can be passed in and\n",
      "     |          treated as a single point.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      values : (# of points,)-array\n",
      "     |          The values at each point.\n",
      "     |      \n",
      "     |      Raises\n",
      "     |      ------\n",
      "     |      ValueError : if the dimensionality of the input points is different than\n",
      "     |                   the dimensionality of the KDE.\n",
      "     |  \n",
      "     |  integrate_box(self, low_bounds, high_bounds, maxpts=None)\n",
      "     |      Computes the integral of a pdf over a rectangular interval.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      low_bounds : array_like\n",
      "     |          A 1-D array containing the lower bounds of integration.\n",
      "     |      high_bounds : array_like\n",
      "     |          A 1-D array containing the upper bounds of integration.\n",
      "     |      maxpts : int, optional\n",
      "     |          The maximum number of points to use for integration.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      value : scalar\n",
      "     |          The result of the integral.\n",
      "     |  \n",
      "     |  integrate_box_1d(self, low, high)\n",
      "     |      Computes the integral of a 1D pdf between two bounds.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      low : scalar\n",
      "     |          Lower bound of integration.\n",
      "     |      high : scalar\n",
      "     |          Upper bound of integration.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      value : scalar\n",
      "     |          The result of the integral.\n",
      "     |      \n",
      "     |      Raises\n",
      "     |      ------\n",
      "     |      ValueError\n",
      "     |          If the KDE is over more than one dimension.\n",
      "     |  \n",
      "     |  integrate_gaussian(self, mean, cov)\n",
      "     |      Multiply estimated density by a multivariate Gaussian and integrate\n",
      "     |      over the whole space.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      mean : aray_like\n",
      "     |          A 1-D array, specifying the mean of the Gaussian.\n",
      "     |      cov : array_like\n",
      "     |          A 2-D array, specifying the covariance matrix of the Gaussian.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      result : scalar\n",
      "     |          The value of the integral.\n",
      "     |      \n",
      "     |      Raises\n",
      "     |      ------\n",
      "     |      ValueError\n",
      "     |          If the mean or covariance of the input Gaussian differs from\n",
      "     |          the KDE's dimensionality.\n",
      "     |  \n",
      "     |  integrate_kde(self, other)\n",
      "     |      Computes the integral of the product of this  kernel density estimate\n",
      "     |      with another.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      other : gaussian_kde instance\n",
      "     |          The other kde.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      value : scalar\n",
      "     |          The result of the integral.\n",
      "     |      \n",
      "     |      Raises\n",
      "     |      ------\n",
      "     |      ValueError\n",
      "     |          If the KDEs have different dimensionality.\n",
      "     |  \n",
      "     |  logpdf(self, x)\n",
      "     |      Evaluate the log of the estimated pdf on a provided set of points.\n",
      "     |  \n",
      "     |  pdf(self, x)\n",
      "     |      Evaluate the estimated pdf on a provided set of points.\n",
      "     |      \n",
      "     |      Notes\n",
      "     |      -----\n",
      "     |      This is an alias for `gaussian_kde.evaluate`.  See the ``evaluate``\n",
      "     |      docstring for more details.\n",
      "     |  \n",
      "     |  resample(self, size=None, seed=None)\n",
      "     |      Randomly sample a dataset from the estimated pdf.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      size : int, optional\n",
      "     |          The number of samples to draw.  If not provided, then the size is\n",
      "     |          the same as the effective number of samples in the underlying\n",
      "     |          dataset.\n",
      "     |      seed : {None, int, `numpy.random.Generator`,\n",
      "     |              `numpy.random.RandomState`}, optional\n",
      "     |      \n",
      "     |          If `seed` is None (or `np.random`), the `numpy.random.RandomState`\n",
      "     |          singleton is used.\n",
      "     |          If `seed` is an int, a new ``RandomState`` instance is used,\n",
      "     |          seeded with `seed`.\n",
      "     |          If `seed` is already a ``Generator`` or ``RandomState`` instance then\n",
      "     |          that instance is used.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      resample : (self.d, `size`) ndarray\n",
      "     |          The sampled dataset.\n",
      "     |  \n",
      "     |  scotts_factor(self)\n",
      "     |      Computes the coefficient (`kde.factor`) that\n",
      "     |      multiplies the data covariance matrix to obtain the kernel covariance\n",
      "     |      matrix. The default is `scotts_factor`.  A subclass can overwrite this\n",
      "     |      method to provide a different method, or set it through a call to\n",
      "     |      `kde.set_bandwidth`.\n",
      "     |  \n",
      "     |  set_bandwidth(self, bw_method=None)\n",
      "     |      Compute the estimator bandwidth with given method.\n",
      "     |      \n",
      "     |      The new bandwidth calculated after a call to `set_bandwidth` is used\n",
      "     |      for subsequent evaluations of the estimated density.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      bw_method : str, scalar or callable, optional\n",
      "     |          The method used to calculate the estimator bandwidth.  This can be\n",
      "     |          'scott', 'silverman', a scalar constant or a callable.  If a\n",
      "     |          scalar, this will be used directly as `kde.factor`.  If a callable,\n",
      "     |          it should take a `gaussian_kde` instance as only parameter and\n",
      "     |          return a scalar.  If None (default), nothing happens; the current\n",
      "     |          `kde.covariance_factor` method is kept.\n",
      "     |      \n",
      "     |      Notes\n",
      "     |      -----\n",
      "     |      .. versionadded:: 0.11\n",
      "     |      \n",
      "     |      Examples\n",
      "     |      --------\n",
      "     |      >>> import scipy.stats as stats\n",
      "     |      >>> x1 = np.array([-7, -5, 1, 4, 5.])\n",
      "     |      >>> kde = stats.gaussian_kde(x1)\n",
      "     |      >>> xs = np.linspace(-10, 10, num=50)\n",
      "     |      >>> y1 = kde(xs)\n",
      "     |      >>> kde.set_bandwidth(bw_method='silverman')\n",
      "     |      >>> y2 = kde(xs)\n",
      "     |      >>> kde.set_bandwidth(bw_method=kde.factor / 3.)\n",
      "     |      >>> y3 = kde(xs)\n",
      "     |      \n",
      "     |      >>> import matplotlib.pyplot as plt\n",
      "     |      >>> fig, ax = plt.subplots()\n",
      "     |      >>> ax.plot(x1, np.full(x1.shape, 1 / (4. * x1.size)), 'bo',\n",
      "     |      ...         label='Data points (rescaled)')\n",
      "     |      >>> ax.plot(xs, y1, label='Scott (default)')\n",
      "     |      >>> ax.plot(xs, y2, label='Silverman')\n",
      "     |      >>> ax.plot(xs, y3, label='Const (1/3 * Silverman)')\n",
      "     |      >>> ax.legend()\n",
      "     |      >>> plt.show()\n",
      "     |  \n",
      "     |  silverman_factor(self)\n",
      "     |      Compute the Silverman factor.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      s : float\n",
      "     |          The silverman factor.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Readonly properties defined here:\n",
      "     |  \n",
      "     |  neff\n",
      "     |  \n",
      "     |  weights\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors defined here:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |      dictionary for instance variables (if defined)\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "    \n",
      "    class rv_continuous(rv_generic)\n",
      "     |  rv_continuous(momtype=1, a=None, b=None, xtol=1e-14, badvalue=None, name=None, longname=None, shapes=None, extradoc=None, seed=None)\n",
      "     |  \n",
      "     |  A generic continuous random variable class meant for subclassing.\n",
      "     |  \n",
      "     |  `rv_continuous` is a base class to construct specific distribution classes\n",
      "     |  and instances for continuous random variables. It cannot be used\n",
      "     |  directly as a distribution.\n",
      "     |  \n",
      "     |  Parameters\n",
      "     |  ----------\n",
      "     |  momtype : int, optional\n",
      "     |      The type of generic moment calculation to use: 0 for pdf, 1 (default)\n",
      "     |      for ppf.\n",
      "     |  a : float, optional\n",
      "     |      Lower bound of the support of the distribution, default is minus\n",
      "     |      infinity.\n",
      "     |  b : float, optional\n",
      "     |      Upper bound of the support of the distribution, default is plus\n",
      "     |      infinity.\n",
      "     |  xtol : float, optional\n",
      "     |      The tolerance for fixed point calculation for generic ppf.\n",
      "     |  badvalue : float, optional\n",
      "     |      The value in a result arrays that indicates a value that for which\n",
      "     |      some argument restriction is violated, default is np.nan.\n",
      "     |  name : str, optional\n",
      "     |      The name of the instance. This string is used to construct the default\n",
      "     |      example for distributions.\n",
      "     |  longname : str, optional\n",
      "     |      This string is used as part of the first line of the docstring returned\n",
      "     |      when a subclass has no docstring of its own. Note: `longname` exists\n",
      "     |      for backwards compatibility, do not use for new subclasses.\n",
      "     |  shapes : str, optional\n",
      "     |      The shape of the distribution. For example ``\"m, n\"`` for a\n",
      "     |      distribution that takes two integers as the two shape arguments for all\n",
      "     |      its methods. If not provided, shape parameters will be inferred from\n",
      "     |      the signature of the private methods, ``_pdf`` and ``_cdf`` of the\n",
      "     |      instance.\n",
      "     |  extradoc :  str, optional, deprecated\n",
      "     |      This string is used as the last part of the docstring returned when a\n",
      "     |      subclass has no docstring of its own. Note: `extradoc` exists for\n",
      "     |      backwards compatibility, do not use for new subclasses.\n",
      "     |  seed : {None, int, `numpy.random.Generator`,\n",
      "     |          `numpy.random.RandomState`}, optional\n",
      "     |  \n",
      "     |      If `seed` is None (or `np.random`), the `numpy.random.RandomState`\n",
      "     |      singleton is used.\n",
      "     |      If `seed` is an int, a new ``RandomState`` instance is used,\n",
      "     |      seeded with `seed`.\n",
      "     |      If `seed` is already a ``Generator`` or ``RandomState`` instance then\n",
      "     |      that instance is used.\n",
      "     |  \n",
      "     |  Methods\n",
      "     |  -------\n",
      "     |  rvs\n",
      "     |  pdf\n",
      "     |  logpdf\n",
      "     |  cdf\n",
      "     |  logcdf\n",
      "     |  sf\n",
      "     |  logsf\n",
      "     |  ppf\n",
      "     |  isf\n",
      "     |  moment\n",
      "     |  stats\n",
      "     |  entropy\n",
      "     |  expect\n",
      "     |  median\n",
      "     |  mean\n",
      "     |  std\n",
      "     |  var\n",
      "     |  interval\n",
      "     |  __call__\n",
      "     |  fit\n",
      "     |  fit_loc_scale\n",
      "     |  nnlf\n",
      "     |  support\n",
      "     |  \n",
      "     |  Notes\n",
      "     |  -----\n",
      "     |  Public methods of an instance of a distribution class (e.g., ``pdf``,\n",
      "     |  ``cdf``) check their arguments and pass valid arguments to private,\n",
      "     |  computational methods (``_pdf``, ``_cdf``). For ``pdf(x)``, ``x`` is valid\n",
      "     |  if it is within the support of the distribution.\n",
      "     |  Whether a shape parameter is valid is decided by an ``_argcheck`` method\n",
      "     |  (which defaults to checking that its arguments are strictly positive.)\n",
      "     |  \n",
      "     |  **Subclassing**\n",
      "     |  \n",
      "     |  New random variables can be defined by subclassing the `rv_continuous` class\n",
      "     |  and re-defining at least the ``_pdf`` or the ``_cdf`` method (normalized\n",
      "     |  to location 0 and scale 1).\n",
      "     |  \n",
      "     |  If positive argument checking is not correct for your RV\n",
      "     |  then you will also need to re-define the ``_argcheck`` method.\n",
      "     |  \n",
      "     |  For most of the scipy.stats distributions, the support interval doesn't\n",
      "     |  depend on the shape parameters. ``x`` being in the support interval is\n",
      "     |  equivalent to ``self.a <= x <= self.b``.  If either of the endpoints of\n",
      "     |  the support do depend on the shape parameters, then\n",
      "     |  i) the distribution must implement the ``_get_support`` method; and\n",
      "     |  ii) those dependent endpoints must be omitted from the distribution's\n",
      "     |  call to the ``rv_continuous`` initializer.\n",
      "     |  \n",
      "     |  Correct, but potentially slow defaults exist for the remaining\n",
      "     |  methods but for speed and/or accuracy you can over-ride::\n",
      "     |  \n",
      "     |    _logpdf, _cdf, _logcdf, _ppf, _rvs, _isf, _sf, _logsf\n",
      "     |  \n",
      "     |  The default method ``_rvs`` relies on the inverse of the cdf, ``_ppf``,\n",
      "     |  applied to a uniform random variate. In order to generate random variates\n",
      "     |  efficiently, either the default ``_ppf`` needs to be overwritten (e.g.\n",
      "     |  if the inverse cdf can expressed in an explicit form) or a sampling\n",
      "     |  method needs to be implemented in a custom ``_rvs`` method.\n",
      "     |  \n",
      "     |  If possible, you should override ``_isf``, ``_sf`` or ``_logsf``.\n",
      "     |  The main reason would be to improve numerical accuracy: for example,\n",
      "     |  the survival function ``_sf`` is computed as ``1 - _cdf`` which can\n",
      "     |  result in loss of precision if ``_cdf(x)`` is close to one.\n",
      "     |  \n",
      "     |  **Methods that can be overwritten by subclasses**\n",
      "     |  ::\n",
      "     |  \n",
      "     |    _rvs\n",
      "     |    _pdf\n",
      "     |    _cdf\n",
      "     |    _sf\n",
      "     |    _ppf\n",
      "     |    _isf\n",
      "     |    _stats\n",
      "     |    _munp\n",
      "     |    _entropy\n",
      "     |    _argcheck\n",
      "     |    _get_support\n",
      "     |  \n",
      "     |  There are additional (internal and private) generic methods that can\n",
      "     |  be useful for cross-checking and for debugging, but might work in all\n",
      "     |  cases when directly called.\n",
      "     |  \n",
      "     |  A note on ``shapes``: subclasses need not specify them explicitly. In this\n",
      "     |  case, `shapes` will be automatically deduced from the signatures of the\n",
      "     |  overridden methods (`pdf`, `cdf` etc).\n",
      "     |  If, for some reason, you prefer to avoid relying on introspection, you can\n",
      "     |  specify ``shapes`` explicitly as an argument to the instance constructor.\n",
      "     |  \n",
      "     |  \n",
      "     |  **Frozen Distributions**\n",
      "     |  \n",
      "     |  Normally, you must provide shape parameters (and, optionally, location and\n",
      "     |  scale parameters to each call of a method of a distribution.\n",
      "     |  \n",
      "     |  Alternatively, the object may be called (as a function) to fix the shape,\n",
      "     |  location, and scale parameters returning a \"frozen\" continuous RV object:\n",
      "     |  \n",
      "     |  rv = generic(<shape(s)>, loc=0, scale=1)\n",
      "     |      `rv_frozen` object with the same methods but holding the given shape,\n",
      "     |      location, and scale fixed\n",
      "     |  \n",
      "     |  **Statistics**\n",
      "     |  \n",
      "     |  Statistics are computed using numerical integration by default.\n",
      "     |  For speed you can redefine this using ``_stats``:\n",
      "     |  \n",
      "     |   - take shape parameters and return mu, mu2, g1, g2\n",
      "     |   - If you can't compute one of these, return it as None\n",
      "     |   - Can also be defined with a keyword argument ``moments``, which is a\n",
      "     |     string composed of \"m\", \"v\", \"s\", and/or \"k\".\n",
      "     |     Only the components appearing in string should be computed and\n",
      "     |     returned in the order \"m\", \"v\", \"s\", or \"k\"  with missing values\n",
      "     |     returned as None.\n",
      "     |  \n",
      "     |  Alternatively, you can override ``_munp``, which takes ``n`` and shape\n",
      "     |  parameters and returns the n-th non-central moment of the distribution.\n",
      "     |  \n",
      "     |  Examples\n",
      "     |  --------\n",
      "     |  To create a new Gaussian distribution, we would do the following:\n",
      "     |  \n",
      "     |  >>> from scipy.stats import rv_continuous\n",
      "     |  >>> class gaussian_gen(rv_continuous):\n",
      "     |  ...     \"Gaussian distribution\"\n",
      "     |  ...     def _pdf(self, x):\n",
      "     |  ...         return np.exp(-x**2 / 2.) / np.sqrt(2.0 * np.pi)\n",
      "     |  >>> gaussian = gaussian_gen(name='gaussian')\n",
      "     |  \n",
      "     |  ``scipy.stats`` distributions are *instances*, so here we subclass\n",
      "     |  `rv_continuous` and create an instance. With this, we now have\n",
      "     |  a fully functional distribution with all relevant methods automagically\n",
      "     |  generated by the framework.\n",
      "     |  \n",
      "     |  Note that above we defined a standard normal distribution, with zero mean\n",
      "     |  and unit variance. Shifting and scaling of the distribution can be done\n",
      "     |  by using ``loc`` and ``scale`` parameters: ``gaussian.pdf(x, loc, scale)``\n",
      "     |  essentially computes ``y = (x - loc) / scale`` and\n",
      "     |  ``gaussian._pdf(y) / scale``.\n",
      "     |  \n",
      "     |  Method resolution order:\n",
      "     |      rv_continuous\n",
      "     |      rv_generic\n",
      "     |      builtins.object\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __getstate__(self)\n",
      "     |  \n",
      "     |  __init__(self, momtype=1, a=None, b=None, xtol=1e-14, badvalue=None, name=None, longname=None, shapes=None, extradoc=None, seed=None)\n",
      "     |      Initialize self.  See help(type(self)) for accurate signature.\n",
      "     |  \n",
      "     |  cdf(self, x, *args, **kwds)\n",
      "     |      Cumulative distribution function of the given RV.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      x : array_like\n",
      "     |          quantiles\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information)\n",
      "     |      loc : array_like, optional\n",
      "     |          location parameter (default=0)\n",
      "     |      scale : array_like, optional\n",
      "     |          scale parameter (default=1)\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      cdf : ndarray\n",
      "     |          Cumulative distribution function evaluated at `x`\n",
      "     |  \n",
      "     |  expect(self, func=None, args=(), loc=0, scale=1, lb=None, ub=None, conditional=False, **kwds)\n",
      "     |      Calculate expected value of a function with respect to the\n",
      "     |      distribution by numerical integration.\n",
      "     |      \n",
      "     |      The expected value of a function ``f(x)`` with respect to a\n",
      "     |      distribution ``dist`` is defined as::\n",
      "     |      \n",
      "     |                  ub\n",
      "     |          E[f(x)] = Integral(f(x) * dist.pdf(x)),\n",
      "     |                  lb\n",
      "     |      \n",
      "     |      where ``ub`` and ``lb`` are arguments and ``x`` has the ``dist.pdf(x)``\n",
      "     |      distribution. If the bounds ``lb`` and ``ub`` correspond to the\n",
      "     |      support of the distribution, e.g. ``[-inf, inf]`` in the default\n",
      "     |      case, then the integral is the unrestricted expectation of ``f(x)``.\n",
      "     |      Also, the function ``f(x)`` may be defined such that ``f(x)`` is ``0``\n",
      "     |      outside a finite interval in which case the expectation is\n",
      "     |      calculated within the finite range ``[lb, ub]``.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      func : callable, optional\n",
      "     |          Function for which integral is calculated. Takes only one argument.\n",
      "     |          The default is the identity mapping f(x) = x.\n",
      "     |      args : tuple, optional\n",
      "     |          Shape parameters of the distribution.\n",
      "     |      loc : float, optional\n",
      "     |          Location parameter (default=0).\n",
      "     |      scale : float, optional\n",
      "     |          Scale parameter (default=1).\n",
      "     |      lb, ub : scalar, optional\n",
      "     |          Lower and upper bound for integration. Default is set to the\n",
      "     |          support of the distribution.\n",
      "     |      conditional : bool, optional\n",
      "     |          If True, the integral is corrected by the conditional probability\n",
      "     |          of the integration interval.  The return value is the expectation\n",
      "     |          of the function, conditional on being in the given interval.\n",
      "     |          Default is False.\n",
      "     |      \n",
      "     |      Additional keyword arguments are passed to the integration routine.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      expect : float\n",
      "     |          The calculated expected value.\n",
      "     |      \n",
      "     |      Notes\n",
      "     |      -----\n",
      "     |      The integration behavior of this function is inherited from\n",
      "     |      `scipy.integrate.quad`. Neither this function nor\n",
      "     |      `scipy.integrate.quad` can verify whether the integral exists or is\n",
      "     |      finite. For example ``cauchy(0).mean()`` returns ``np.nan`` and\n",
      "     |      ``cauchy(0).expect()`` returns ``0.0``.\n",
      "     |      \n",
      "     |      The function is not vectorized.\n",
      "     |      \n",
      "     |      Examples\n",
      "     |      --------\n",
      "     |      \n",
      "     |      To understand the effect of the bounds of integration consider\n",
      "     |      \n",
      "     |      >>> from scipy.stats import expon\n",
      "     |      >>> expon(1).expect(lambda x: 1, lb=0.0, ub=2.0)\n",
      "     |      0.6321205588285578\n",
      "     |      \n",
      "     |      This is close to\n",
      "     |      \n",
      "     |      >>> expon(1).cdf(2.0) - expon(1).cdf(0.0)\n",
      "     |      0.6321205588285577\n",
      "     |      \n",
      "     |      If ``conditional=True``\n",
      "     |      \n",
      "     |      >>> expon(1).expect(lambda x: 1, lb=0.0, ub=2.0, conditional=True)\n",
      "     |      1.0000000000000002\n",
      "     |      \n",
      "     |      The slight deviation from 1 is due to numerical integration.\n",
      "     |  \n",
      "     |  fit(self, data, *args, **kwds)\n",
      "     |      Return estimates of shape (if applicable), location, and scale\n",
      "     |      parameters from data. The default estimation method is Maximum\n",
      "     |      Likelihood Estimation (MLE), but Method of Moments (MM)\n",
      "     |      is also available.\n",
      "     |      \n",
      "     |      Starting estimates for\n",
      "     |      the fit are given by input arguments; for any arguments not provided\n",
      "     |      with starting estimates, ``self._fitstart(data)`` is called to generate\n",
      "     |      such.\n",
      "     |      \n",
      "     |      One can hold some parameters fixed to specific values by passing in\n",
      "     |      keyword arguments ``f0``, ``f1``, ..., ``fn`` (for shape parameters)\n",
      "     |      and ``floc`` and ``fscale`` (for location and scale parameters,\n",
      "     |      respectively).\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      data : array_like\n",
      "     |          Data to use in estimating the distribution parameters.\n",
      "     |      arg1, arg2, arg3,... : floats, optional\n",
      "     |          Starting value(s) for any shape-characterizing arguments (those not\n",
      "     |          provided will be determined by a call to ``_fitstart(data)``).\n",
      "     |          No default value.\n",
      "     |      kwds : floats, optional\n",
      "     |          - `loc`: initial guess of the distribution's location parameter.\n",
      "     |          - `scale`: initial guess of the distribution's scale parameter.\n",
      "     |      \n",
      "     |          Special keyword arguments are recognized as holding certain\n",
      "     |          parameters fixed:\n",
      "     |      \n",
      "     |          - f0...fn : hold respective shape parameters fixed.\n",
      "     |            Alternatively, shape parameters to fix can be specified by name.\n",
      "     |            For example, if ``self.shapes == \"a, b\"``, ``fa`` and ``fix_a``\n",
      "     |            are equivalent to ``f0``, and ``fb`` and ``fix_b`` are\n",
      "     |            equivalent to ``f1``.\n",
      "     |      \n",
      "     |          - floc : hold location parameter fixed to specified value.\n",
      "     |      \n",
      "     |          - fscale : hold scale parameter fixed to specified value.\n",
      "     |      \n",
      "     |          - optimizer : The optimizer to use.\n",
      "     |            The optimizer must take ``func``,\n",
      "     |            and starting position as the first two arguments,\n",
      "     |            plus ``args`` (for extra arguments to pass to the\n",
      "     |            function to be optimized) and ``disp=0`` to suppress\n",
      "     |            output as keyword arguments.\n",
      "     |      \n",
      "     |          - method : The method to use. The default is \"MLE\" (Maximum\n",
      "     |            Likelihood Estimate); \"MM\" (Method of Moments)\n",
      "     |            is also available.\n",
      "     |      \n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      parameter_tuple : tuple of floats\n",
      "     |          Estimates for any shape parameters (if applicable),\n",
      "     |          followed by those for location and scale.\n",
      "     |          For most random variables, shape statistics\n",
      "     |          will be returned, but there are exceptions (e.g. ``norm``).\n",
      "     |      \n",
      "     |      Notes\n",
      "     |      -----\n",
      "     |      With ``method=\"MLE\"`` (default), the fit is computed by minimizing\n",
      "     |      the negative log-likelihood function. A large, finite penalty\n",
      "     |      (rather than infinite negative log-likelihood) is applied for\n",
      "     |      observations beyond the support of the distribution.\n",
      "     |      \n",
      "     |      With ``method=\"MM\"``, the fit is computed by minimizing the L2 norm\n",
      "     |      of the relative errors between the first *k* raw (about zero) data\n",
      "     |      moments and the corresponding distribution moments, where *k* is the\n",
      "     |      number of non-fixed parameters.\n",
      "     |      More precisely, the objective function is::\n",
      "     |      \n",
      "     |          (((data_moments - dist_moments)\n",
      "     |            / np.maximum(np.abs(data_moments), 1e-8))**2).sum()\n",
      "     |      \n",
      "     |      where the constant ``1e-8`` avoids division by zero in case of\n",
      "     |      vanishing data moments. Typically, this error norm can be reduced to\n",
      "     |      zero.\n",
      "     |      Note that the standard method of moments can produce parameters for\n",
      "     |      which some data are outside the support of the fitted distribution;\n",
      "     |      this implementation does nothing to prevent this.\n",
      "     |      \n",
      "     |      For either method,\n",
      "     |      the returned answer is not guaranteed to be globally optimal; it\n",
      "     |      may only be locally optimal, or the optimization may fail altogether.\n",
      "     |      If the data contain any of ``np.nan``, ``np.inf``, or ``-np.inf``,\n",
      "     |      the `fit` method will raise a ``RuntimeError``.\n",
      "     |      \n",
      "     |      Examples\n",
      "     |      --------\n",
      "     |      \n",
      "     |      Generate some data to fit: draw random variates from the `beta`\n",
      "     |      distribution\n",
      "     |      \n",
      "     |      >>> from scipy.stats import beta\n",
      "     |      >>> a, b = 1., 2.\n",
      "     |      >>> x = beta.rvs(a, b, size=1000)\n",
      "     |      \n",
      "     |      Now we can fit all four parameters (``a``, ``b``, ``loc``\n",
      "     |      and ``scale``):\n",
      "     |      \n",
      "     |      >>> a1, b1, loc1, scale1 = beta.fit(x)\n",
      "     |      \n",
      "     |      We can also use some prior knowledge about the dataset: let's keep\n",
      "     |      ``loc`` and ``scale`` fixed:\n",
      "     |      \n",
      "     |      >>> a1, b1, loc1, scale1 = beta.fit(x, floc=0, fscale=1)\n",
      "     |      >>> loc1, scale1\n",
      "     |      (0, 1)\n",
      "     |      \n",
      "     |      We can also keep shape parameters fixed by using ``f``-keywords. To\n",
      "     |      keep the zero-th shape parameter ``a`` equal 1, use ``f0=1`` or,\n",
      "     |      equivalently, ``fa=1``:\n",
      "     |      \n",
      "     |      >>> a1, b1, loc1, scale1 = beta.fit(x, fa=1, floc=0, fscale=1)\n",
      "     |      >>> a1\n",
      "     |      1\n",
      "     |      \n",
      "     |      Not all distributions return estimates for the shape parameters.\n",
      "     |      ``norm`` for example just returns estimates for location and scale:\n",
      "     |      \n",
      "     |      >>> from scipy.stats import norm\n",
      "     |      >>> x = norm.rvs(a, b, size=1000, random_state=123)\n",
      "     |      >>> loc1, scale1 = norm.fit(x)\n",
      "     |      >>> loc1, scale1\n",
      "     |      (0.92087172783841631, 2.0015750750324668)\n",
      "     |  \n",
      "     |  fit_loc_scale(self, data, *args)\n",
      "     |      Estimate loc and scale parameters from data using 1st and 2nd moments.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      data : array_like\n",
      "     |          Data to fit.\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information).\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      Lhat : float\n",
      "     |          Estimated location parameter for the data.\n",
      "     |      Shat : float\n",
      "     |          Estimated scale parameter for the data.\n",
      "     |  \n",
      "     |  isf(self, q, *args, **kwds)\n",
      "     |      Inverse survival function (inverse of `sf`) at q of the given RV.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      q : array_like\n",
      "     |          upper tail probability\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information)\n",
      "     |      loc : array_like, optional\n",
      "     |          location parameter (default=0)\n",
      "     |      scale : array_like, optional\n",
      "     |          scale parameter (default=1)\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      x : ndarray or scalar\n",
      "     |          Quantile corresponding to the upper tail probability q.\n",
      "     |  \n",
      "     |  logcdf(self, x, *args, **kwds)\n",
      "     |      Log of the cumulative distribution function at x of the given RV.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      x : array_like\n",
      "     |          quantiles\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information)\n",
      "     |      loc : array_like, optional\n",
      "     |          location parameter (default=0)\n",
      "     |      scale : array_like, optional\n",
      "     |          scale parameter (default=1)\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      logcdf : array_like\n",
      "     |          Log of the cumulative distribution function evaluated at x\n",
      "     |  \n",
      "     |  logpdf(self, x, *args, **kwds)\n",
      "     |      Log of the probability density function at x of the given RV.\n",
      "     |      \n",
      "     |      This uses a more numerically accurate calculation if available.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      x : array_like\n",
      "     |          quantiles\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information)\n",
      "     |      loc : array_like, optional\n",
      "     |          location parameter (default=0)\n",
      "     |      scale : array_like, optional\n",
      "     |          scale parameter (default=1)\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      logpdf : array_like\n",
      "     |          Log of the probability density function evaluated at x\n",
      "     |  \n",
      "     |  logsf(self, x, *args, **kwds)\n",
      "     |      Log of the survival function of the given RV.\n",
      "     |      \n",
      "     |      Returns the log of the \"survival function,\" defined as (1 - `cdf`),\n",
      "     |      evaluated at `x`.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      x : array_like\n",
      "     |          quantiles\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information)\n",
      "     |      loc : array_like, optional\n",
      "     |          location parameter (default=0)\n",
      "     |      scale : array_like, optional\n",
      "     |          scale parameter (default=1)\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      logsf : ndarray\n",
      "     |          Log of the survival function evaluated at `x`.\n",
      "     |  \n",
      "     |  nnlf(self, theta, x)\n",
      "     |      Negative loglikelihood function.\n",
      "     |      \n",
      "     |      Notes\n",
      "     |      -----\n",
      "     |      This is ``-sum(log pdf(x, theta), axis=0)`` where `theta` are the\n",
      "     |      parameters (including loc and scale).\n",
      "     |  \n",
      "     |  pdf(self, x, *args, **kwds)\n",
      "     |      Probability density function at x of the given RV.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      x : array_like\n",
      "     |          quantiles\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information)\n",
      "     |      loc : array_like, optional\n",
      "     |          location parameter (default=0)\n",
      "     |      scale : array_like, optional\n",
      "     |          scale parameter (default=1)\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      pdf : ndarray\n",
      "     |          Probability density function evaluated at x\n",
      "     |  \n",
      "     |  ppf(self, q, *args, **kwds)\n",
      "     |      Percent point function (inverse of `cdf`) at q of the given RV.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      q : array_like\n",
      "     |          lower tail probability\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information)\n",
      "     |      loc : array_like, optional\n",
      "     |          location parameter (default=0)\n",
      "     |      scale : array_like, optional\n",
      "     |          scale parameter (default=1)\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      x : array_like\n",
      "     |          quantile corresponding to the lower tail probability q.\n",
      "     |  \n",
      "     |  sf(self, x, *args, **kwds)\n",
      "     |      Survival function (1 - `cdf`) at x of the given RV.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      x : array_like\n",
      "     |          quantiles\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information)\n",
      "     |      loc : array_like, optional\n",
      "     |          location parameter (default=0)\n",
      "     |      scale : array_like, optional\n",
      "     |          scale parameter (default=1)\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      sf : array_like\n",
      "     |          Survival function evaluated at x\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from rv_generic:\n",
      "     |  \n",
      "     |  __call__(self, *args, **kwds)\n",
      "     |      Freeze the distribution for the given arguments.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution.  Should include all\n",
      "     |          the non-optional arguments, may include ``loc`` and ``scale``.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      rv_frozen : rv_frozen instance\n",
      "     |          The frozen distribution.\n",
      "     |  \n",
      "     |  __setstate__(self, state)\n",
      "     |  \n",
      "     |  entropy(self, *args, **kwds)\n",
      "     |      Differential entropy of the RV.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information).\n",
      "     |      loc : array_like, optional\n",
      "     |          Location parameter (default=0).\n",
      "     |      scale : array_like, optional  (continuous distributions only).\n",
      "     |          Scale parameter (default=1).\n",
      "     |      \n",
      "     |      Notes\n",
      "     |      -----\n",
      "     |      Entropy is defined base `e`:\n",
      "     |      \n",
      "     |      >>> drv = rv_discrete(values=((0, 1), (0.5, 0.5)))\n",
      "     |      >>> np.allclose(drv.entropy(), np.log(2.0))\n",
      "     |      True\n",
      "     |  \n",
      "     |  freeze(self, *args, **kwds)\n",
      "     |      Freeze the distribution for the given arguments.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution.  Should include all\n",
      "     |          the non-optional arguments, may include ``loc`` and ``scale``.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      rv_frozen : rv_frozen instance\n",
      "     |          The frozen distribution.\n",
      "     |  \n",
      "     |  interval(self, alpha, *args, **kwds)\n",
      "     |      Confidence interval with equal areas around the median.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      alpha : array_like of float\n",
      "     |          Probability that an rv will be drawn from the returned range.\n",
      "     |          Each value should be in the range [0, 1].\n",
      "     |      arg1, arg2, ... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information).\n",
      "     |      loc : array_like, optional\n",
      "     |          location parameter, Default is 0.\n",
      "     |      scale : array_like, optional\n",
      "     |          scale parameter, Default is 1.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      a, b : ndarray of float\n",
      "     |          end-points of range that contain ``100 * alpha %`` of the rv's\n",
      "     |          possible values.\n",
      "     |  \n",
      "     |  mean(self, *args, **kwds)\n",
      "     |      Mean of the distribution.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information)\n",
      "     |      loc : array_like, optional\n",
      "     |          location parameter (default=0)\n",
      "     |      scale : array_like, optional\n",
      "     |          scale parameter (default=1)\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      mean : float\n",
      "     |          the mean of the distribution\n",
      "     |  \n",
      "     |  median(self, *args, **kwds)\n",
      "     |      Median of the distribution.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information)\n",
      "     |      loc : array_like, optional\n",
      "     |          Location parameter, Default is 0.\n",
      "     |      scale : array_like, optional\n",
      "     |          Scale parameter, Default is 1.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      median : float\n",
      "     |          The median of the distribution.\n",
      "     |      \n",
      "     |      See Also\n",
      "     |      --------\n",
      "     |      rv_discrete.ppf\n",
      "     |          Inverse of the CDF\n",
      "     |  \n",
      "     |  moment(self, n, *args, **kwds)\n",
      "     |      n-th order non-central moment of distribution.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      n : int, n >= 1\n",
      "     |          Order of moment.\n",
      "     |      arg1, arg2, arg3,... : float\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information).\n",
      "     |      loc : array_like, optional\n",
      "     |          location parameter (default=0)\n",
      "     |      scale : array_like, optional\n",
      "     |          scale parameter (default=1)\n",
      "     |  \n",
      "     |  rvs(self, *args, **kwds)\n",
      "     |      Random variates of given type.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information).\n",
      "     |      loc : array_like, optional\n",
      "     |          Location parameter (default=0).\n",
      "     |      scale : array_like, optional\n",
      "     |          Scale parameter (default=1).\n",
      "     |      size : int or tuple of ints, optional\n",
      "     |          Defining number of random variates (default is 1).\n",
      "     |      random_state : {None, int, `numpy.random.Generator`,\n",
      "     |                      `numpy.random.RandomState`}, optional\n",
      "     |      \n",
      "     |          If `seed` is None (or `np.random`), the `numpy.random.RandomState`\n",
      "     |          singleton is used.\n",
      "     |          If `seed` is an int, a new ``RandomState`` instance is used,\n",
      "     |          seeded with `seed`.\n",
      "     |          If `seed` is already a ``Generator`` or ``RandomState`` instance\n",
      "     |          then that instance is used.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      rvs : ndarray or scalar\n",
      "     |          Random variates of given `size`.\n",
      "     |  \n",
      "     |  stats(self, *args, **kwds)\n",
      "     |      Some statistics of the given RV.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information)\n",
      "     |      loc : array_like, optional\n",
      "     |          location parameter (default=0)\n",
      "     |      scale : array_like, optional (continuous RVs only)\n",
      "     |          scale parameter (default=1)\n",
      "     |      moments : str, optional\n",
      "     |          composed of letters ['mvsk'] defining which moments to compute:\n",
      "     |          'm' = mean,\n",
      "     |          'v' = variance,\n",
      "     |          's' = (Fisher's) skew,\n",
      "     |          'k' = (Fisher's) kurtosis.\n",
      "     |          (default is 'mv')\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      stats : sequence\n",
      "     |          of requested moments.\n",
      "     |  \n",
      "     |  std(self, *args, **kwds)\n",
      "     |      Standard deviation of the distribution.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information)\n",
      "     |      loc : array_like, optional\n",
      "     |          location parameter (default=0)\n",
      "     |      scale : array_like, optional\n",
      "     |          scale parameter (default=1)\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      std : float\n",
      "     |          standard deviation of the distribution\n",
      "     |  \n",
      "     |  support(self, *args, **kwargs)\n",
      "     |      Support of the distribution.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      arg1, arg2, ... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information).\n",
      "     |      loc : array_like, optional\n",
      "     |          location parameter, Default is 0.\n",
      "     |      scale : array_like, optional\n",
      "     |          scale parameter, Default is 1.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      a, b : array_like\n",
      "     |          end-points of the distribution's support.\n",
      "     |  \n",
      "     |  var(self, *args, **kwds)\n",
      "     |      Variance of the distribution.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information)\n",
      "     |      loc : array_like, optional\n",
      "     |          location parameter (default=0)\n",
      "     |      scale : array_like, optional\n",
      "     |          scale parameter (default=1)\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      var : float\n",
      "     |          the variance of the distribution\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from rv_generic:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |      dictionary for instance variables (if defined)\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "     |  \n",
      "     |  random_state\n",
      "     |      Get or set the generator object for generating random variates.\n",
      "     |      \n",
      "     |      If `seed` is None (or `np.random`), the `numpy.random.RandomState`\n",
      "     |      singleton is used.\n",
      "     |      If `seed` is an int, a new ``RandomState`` instance is used,\n",
      "     |      seeded with `seed`.\n",
      "     |      If `seed` is already a ``Generator`` or ``RandomState`` instance then\n",
      "     |      that instance is used.\n",
      "    \n",
      "    class rv_discrete(rv_generic)\n",
      "     |  rv_discrete(a=0, b=inf, name=None, badvalue=None, moment_tol=1e-08, values=None, inc=1, longname=None, shapes=None, extradoc=None, seed=None)\n",
      "     |  \n",
      "     |  A generic discrete random variable class meant for subclassing.\n",
      "     |  \n",
      "     |  `rv_discrete` is a base class to construct specific distribution classes\n",
      "     |  and instances for discrete random variables. It can also be used\n",
      "     |  to construct an arbitrary distribution defined by a list of support\n",
      "     |  points and corresponding probabilities.\n",
      "     |  \n",
      "     |  Parameters\n",
      "     |  ----------\n",
      "     |  a : float, optional\n",
      "     |      Lower bound of the support of the distribution, default: 0\n",
      "     |  b : float, optional\n",
      "     |      Upper bound of the support of the distribution, default: plus infinity\n",
      "     |  moment_tol : float, optional\n",
      "     |      The tolerance for the generic calculation of moments.\n",
      "     |  values : tuple of two array_like, optional\n",
      "     |      ``(xk, pk)`` where ``xk`` are integers and ``pk`` are the non-zero\n",
      "     |      probabilities between 0 and 1 with ``sum(pk) = 1``. ``xk``\n",
      "     |      and ``pk`` must have the same shape.\n",
      "     |  inc : integer, optional\n",
      "     |      Increment for the support of the distribution.\n",
      "     |      Default is 1. (other values have not been tested)\n",
      "     |  badvalue : float, optional\n",
      "     |      The value in a result arrays that indicates a value that for which\n",
      "     |      some argument restriction is violated, default is np.nan.\n",
      "     |  name : str, optional\n",
      "     |      The name of the instance. This string is used to construct the default\n",
      "     |      example for distributions.\n",
      "     |  longname : str, optional\n",
      "     |      This string is used as part of the first line of the docstring returned\n",
      "     |      when a subclass has no docstring of its own. Note: `longname` exists\n",
      "     |      for backwards compatibility, do not use for new subclasses.\n",
      "     |  shapes : str, optional\n",
      "     |      The shape of the distribution. For example \"m, n\" for a distribution\n",
      "     |      that takes two integers as the two shape arguments for all its methods\n",
      "     |      If not provided, shape parameters will be inferred from\n",
      "     |      the signatures of the private methods, ``_pmf`` and ``_cdf`` of\n",
      "     |      the instance.\n",
      "     |  extradoc :  str, optional\n",
      "     |      This string is used as the last part of the docstring returned when a\n",
      "     |      subclass has no docstring of its own. Note: `extradoc` exists for\n",
      "     |      backwards compatibility, do not use for new subclasses.\n",
      "     |  seed : {None, int, `numpy.random.Generator`,\n",
      "     |          `numpy.random.RandomState`}, optional\n",
      "     |  \n",
      "     |      If `seed` is None (or `np.random`), the `numpy.random.RandomState`\n",
      "     |      singleton is used.\n",
      "     |      If `seed` is an int, a new ``RandomState`` instance is used,\n",
      "     |      seeded with `seed`.\n",
      "     |      If `seed` is already a ``Generator`` or ``RandomState`` instance then\n",
      "     |      that instance is used.\n",
      "     |  \n",
      "     |  Methods\n",
      "     |  -------\n",
      "     |  rvs\n",
      "     |  pmf\n",
      "     |  logpmf\n",
      "     |  cdf\n",
      "     |  logcdf\n",
      "     |  sf\n",
      "     |  logsf\n",
      "     |  ppf\n",
      "     |  isf\n",
      "     |  moment\n",
      "     |  stats\n",
      "     |  entropy\n",
      "     |  expect\n",
      "     |  median\n",
      "     |  mean\n",
      "     |  std\n",
      "     |  var\n",
      "     |  interval\n",
      "     |  __call__\n",
      "     |  support\n",
      "     |  \n",
      "     |  Notes\n",
      "     |  -----\n",
      "     |  This class is similar to `rv_continuous`. Whether a shape parameter is\n",
      "     |  valid is decided by an ``_argcheck`` method (which defaults to checking\n",
      "     |  that its arguments are strictly positive.)\n",
      "     |  The main differences are:\n",
      "     |  \n",
      "     |  - the support of the distribution is a set of integers\n",
      "     |  - instead of the probability density function, ``pdf`` (and the\n",
      "     |    corresponding private ``_pdf``), this class defines the\n",
      "     |    *probability mass function*, `pmf` (and the corresponding\n",
      "     |    private ``_pmf``.)\n",
      "     |  - scale parameter is not defined.\n",
      "     |  \n",
      "     |  To create a new discrete distribution, we would do the following:\n",
      "     |  \n",
      "     |  >>> from scipy.stats import rv_discrete\n",
      "     |  >>> class poisson_gen(rv_discrete):\n",
      "     |  ...     \"Poisson distribution\"\n",
      "     |  ...     def _pmf(self, k, mu):\n",
      "     |  ...         return exp(-mu) * mu**k / factorial(k)\n",
      "     |  \n",
      "     |  and create an instance::\n",
      "     |  \n",
      "     |  >>> poisson = poisson_gen(name=\"poisson\")\n",
      "     |  \n",
      "     |  Note that above we defined the Poisson distribution in the standard form.\n",
      "     |  Shifting the distribution can be done by providing the ``loc`` parameter\n",
      "     |  to the methods of the instance. For example, ``poisson.pmf(x, mu, loc)``\n",
      "     |  delegates the work to ``poisson._pmf(x-loc, mu)``.\n",
      "     |  \n",
      "     |  **Discrete distributions from a list of probabilities**\n",
      "     |  \n",
      "     |  Alternatively, you can construct an arbitrary discrete rv defined\n",
      "     |  on a finite set of values ``xk`` with ``Prob{X=xk} = pk`` by using the\n",
      "     |  ``values`` keyword argument to the `rv_discrete` constructor.\n",
      "     |  \n",
      "     |  Examples\n",
      "     |  --------\n",
      "     |  Custom made discrete distribution:\n",
      "     |  \n",
      "     |  >>> from scipy import stats\n",
      "     |  >>> xk = np.arange(7)\n",
      "     |  >>> pk = (0.1, 0.2, 0.3, 0.1, 0.1, 0.0, 0.2)\n",
      "     |  >>> custm = stats.rv_discrete(name='custm', values=(xk, pk))\n",
      "     |  >>>\n",
      "     |  >>> import matplotlib.pyplot as plt\n",
      "     |  >>> fig, ax = plt.subplots(1, 1)\n",
      "     |  >>> ax.plot(xk, custm.pmf(xk), 'ro', ms=12, mec='r')\n",
      "     |  >>> ax.vlines(xk, 0, custm.pmf(xk), colors='r', lw=4)\n",
      "     |  >>> plt.show()\n",
      "     |  \n",
      "     |  Random number generation:\n",
      "     |  \n",
      "     |  >>> R = custm.rvs(size=100)\n",
      "     |  \n",
      "     |  Method resolution order:\n",
      "     |      rv_discrete\n",
      "     |      rv_generic\n",
      "     |      builtins.object\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __getstate__(self)\n",
      "     |  \n",
      "     |  __init__(self, a=0, b=inf, name=None, badvalue=None, moment_tol=1e-08, values=None, inc=1, longname=None, shapes=None, extradoc=None, seed=None)\n",
      "     |      Initialize self.  See help(type(self)) for accurate signature.\n",
      "     |  \n",
      "     |  cdf(self, k, *args, **kwds)\n",
      "     |      Cumulative distribution function of the given RV.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      k : array_like, int\n",
      "     |          Quantiles.\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information).\n",
      "     |      loc : array_like, optional\n",
      "     |          Location parameter (default=0).\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      cdf : ndarray\n",
      "     |          Cumulative distribution function evaluated at `k`.\n",
      "     |  \n",
      "     |  expect(self, func=None, args=(), loc=0, lb=None, ub=None, conditional=False, maxcount=1000, tolerance=1e-10, chunksize=32)\n",
      "     |      Calculate expected value of a function with respect to the distribution\n",
      "     |      for discrete distribution by numerical summation.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      func : callable, optional\n",
      "     |          Function for which the expectation value is calculated.\n",
      "     |          Takes only one argument.\n",
      "     |          The default is the identity mapping f(k) = k.\n",
      "     |      args : tuple, optional\n",
      "     |          Shape parameters of the distribution.\n",
      "     |      loc : float, optional\n",
      "     |          Location parameter.\n",
      "     |          Default is 0.\n",
      "     |      lb, ub : int, optional\n",
      "     |          Lower and upper bound for the summation, default is set to the\n",
      "     |          support of the distribution, inclusive (``lb <= k <= ub``).\n",
      "     |      conditional : bool, optional\n",
      "     |          If true then the expectation is corrected by the conditional\n",
      "     |          probability of the summation interval. The return value is the\n",
      "     |          expectation of the function, `func`, conditional on being in\n",
      "     |          the given interval (k such that ``lb <= k <= ub``).\n",
      "     |          Default is False.\n",
      "     |      maxcount : int, optional\n",
      "     |          Maximal number of terms to evaluate (to avoid an endless loop for\n",
      "     |          an infinite sum). Default is 1000.\n",
      "     |      tolerance : float, optional\n",
      "     |          Absolute tolerance for the summation. Default is 1e-10.\n",
      "     |      chunksize : int, optional\n",
      "     |          Iterate over the support of a distributions in chunks of this size.\n",
      "     |          Default is 32.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      expect : float\n",
      "     |          Expected value.\n",
      "     |      \n",
      "     |      Notes\n",
      "     |      -----\n",
      "     |      For heavy-tailed distributions, the expected value may or\n",
      "     |      may not exist,\n",
      "     |      depending on the function, `func`. If it does exist, but the\n",
      "     |      sum converges\n",
      "     |      slowly, the accuracy of the result may be rather low. For instance, for\n",
      "     |      ``zipf(4)``, accuracy for mean, variance in example is only 1e-5.\n",
      "     |      increasing `maxcount` and/or `chunksize` may improve the result,\n",
      "     |      but may also make zipf very slow.\n",
      "     |      \n",
      "     |      The function is not vectorized.\n",
      "     |  \n",
      "     |  isf(self, q, *args, **kwds)\n",
      "     |      Inverse survival function (inverse of `sf`) at q of the given RV.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      q : array_like\n",
      "     |          Upper tail probability.\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information).\n",
      "     |      loc : array_like, optional\n",
      "     |          Location parameter (default=0).\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      k : ndarray or scalar\n",
      "     |          Quantile corresponding to the upper tail probability, q.\n",
      "     |  \n",
      "     |  logcdf(self, k, *args, **kwds)\n",
      "     |      Log of the cumulative distribution function at k of the given RV.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      k : array_like, int\n",
      "     |          Quantiles.\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information).\n",
      "     |      loc : array_like, optional\n",
      "     |          Location parameter (default=0).\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      logcdf : array_like\n",
      "     |          Log of the cumulative distribution function evaluated at k.\n",
      "     |  \n",
      "     |  logpmf(self, k, *args, **kwds)\n",
      "     |      Log of the probability mass function at k of the given RV.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      k : array_like\n",
      "     |          Quantiles.\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information).\n",
      "     |      loc : array_like, optional\n",
      "     |          Location parameter. Default is 0.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      logpmf : array_like\n",
      "     |          Log of the probability mass function evaluated at k.\n",
      "     |  \n",
      "     |  logsf(self, k, *args, **kwds)\n",
      "     |      Log of the survival function of the given RV.\n",
      "     |      \n",
      "     |      Returns the log of the \"survival function,\" defined as 1 - `cdf`,\n",
      "     |      evaluated at `k`.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      k : array_like\n",
      "     |          Quantiles.\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information).\n",
      "     |      loc : array_like, optional\n",
      "     |          Location parameter (default=0).\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      logsf : ndarray\n",
      "     |          Log of the survival function evaluated at `k`.\n",
      "     |  \n",
      "     |  pmf(self, k, *args, **kwds)\n",
      "     |      Probability mass function at k of the given RV.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      k : array_like\n",
      "     |          Quantiles.\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information)\n",
      "     |      loc : array_like, optional\n",
      "     |          Location parameter (default=0).\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      pmf : array_like\n",
      "     |          Probability mass function evaluated at k\n",
      "     |  \n",
      "     |  ppf(self, q, *args, **kwds)\n",
      "     |      Percent point function (inverse of `cdf`) at q of the given RV.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      q : array_like\n",
      "     |          Lower tail probability.\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information).\n",
      "     |      loc : array_like, optional\n",
      "     |          Location parameter (default=0).\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      k : array_like\n",
      "     |          Quantile corresponding to the lower tail probability, q.\n",
      "     |  \n",
      "     |  rvs(self, *args, **kwargs)\n",
      "     |      Random variates of given type.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information).\n",
      "     |      loc : array_like, optional\n",
      "     |          Location parameter (default=0).\n",
      "     |      size : int or tuple of ints, optional\n",
      "     |          Defining number of random variates (Default is 1). Note that `size`\n",
      "     |          has to be given as keyword, not as positional argument.\n",
      "     |      random_state : {None, int, `numpy.random.Generator`,\n",
      "     |                      `numpy.random.RandomState`}, optional\n",
      "     |      \n",
      "     |          If `seed` is None (or `np.random`), the `numpy.random.RandomState`\n",
      "     |          singleton is used.\n",
      "     |          If `seed` is an int, a new ``RandomState`` instance is used,\n",
      "     |          seeded with `seed`.\n",
      "     |          If `seed` is already a ``Generator`` or ``RandomState`` instance\n",
      "     |          then that instance is used.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      rvs : ndarray or scalar\n",
      "     |          Random variates of given `size`.\n",
      "     |  \n",
      "     |  sf(self, k, *args, **kwds)\n",
      "     |      Survival function (1 - `cdf`) at k of the given RV.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      k : array_like\n",
      "     |          Quantiles.\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information).\n",
      "     |      loc : array_like, optional\n",
      "     |          Location parameter (default=0).\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      sf : array_like\n",
      "     |          Survival function evaluated at k.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Static methods defined here:\n",
      "     |  \n",
      "     |  __new__(cls, a=0, b=inf, name=None, badvalue=None, moment_tol=1e-08, values=None, inc=1, longname=None, shapes=None, extradoc=None, seed=None)\n",
      "     |      Create and return a new object.  See help(type) for accurate signature.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from rv_generic:\n",
      "     |  \n",
      "     |  __call__(self, *args, **kwds)\n",
      "     |      Freeze the distribution for the given arguments.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution.  Should include all\n",
      "     |          the non-optional arguments, may include ``loc`` and ``scale``.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      rv_frozen : rv_frozen instance\n",
      "     |          The frozen distribution.\n",
      "     |  \n",
      "     |  __setstate__(self, state)\n",
      "     |  \n",
      "     |  entropy(self, *args, **kwds)\n",
      "     |      Differential entropy of the RV.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information).\n",
      "     |      loc : array_like, optional\n",
      "     |          Location parameter (default=0).\n",
      "     |      scale : array_like, optional  (continuous distributions only).\n",
      "     |          Scale parameter (default=1).\n",
      "     |      \n",
      "     |      Notes\n",
      "     |      -----\n",
      "     |      Entropy is defined base `e`:\n",
      "     |      \n",
      "     |      >>> drv = rv_discrete(values=((0, 1), (0.5, 0.5)))\n",
      "     |      >>> np.allclose(drv.entropy(), np.log(2.0))\n",
      "     |      True\n",
      "     |  \n",
      "     |  freeze(self, *args, **kwds)\n",
      "     |      Freeze the distribution for the given arguments.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution.  Should include all\n",
      "     |          the non-optional arguments, may include ``loc`` and ``scale``.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      rv_frozen : rv_frozen instance\n",
      "     |          The frozen distribution.\n",
      "     |  \n",
      "     |  interval(self, alpha, *args, **kwds)\n",
      "     |      Confidence interval with equal areas around the median.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      alpha : array_like of float\n",
      "     |          Probability that an rv will be drawn from the returned range.\n",
      "     |          Each value should be in the range [0, 1].\n",
      "     |      arg1, arg2, ... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information).\n",
      "     |      loc : array_like, optional\n",
      "     |          location parameter, Default is 0.\n",
      "     |      scale : array_like, optional\n",
      "     |          scale parameter, Default is 1.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      a, b : ndarray of float\n",
      "     |          end-points of range that contain ``100 * alpha %`` of the rv's\n",
      "     |          possible values.\n",
      "     |  \n",
      "     |  mean(self, *args, **kwds)\n",
      "     |      Mean of the distribution.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information)\n",
      "     |      loc : array_like, optional\n",
      "     |          location parameter (default=0)\n",
      "     |      scale : array_like, optional\n",
      "     |          scale parameter (default=1)\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      mean : float\n",
      "     |          the mean of the distribution\n",
      "     |  \n",
      "     |  median(self, *args, **kwds)\n",
      "     |      Median of the distribution.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information)\n",
      "     |      loc : array_like, optional\n",
      "     |          Location parameter, Default is 0.\n",
      "     |      scale : array_like, optional\n",
      "     |          Scale parameter, Default is 1.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      median : float\n",
      "     |          The median of the distribution.\n",
      "     |      \n",
      "     |      See Also\n",
      "     |      --------\n",
      "     |      rv_discrete.ppf\n",
      "     |          Inverse of the CDF\n",
      "     |  \n",
      "     |  moment(self, n, *args, **kwds)\n",
      "     |      n-th order non-central moment of distribution.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      n : int, n >= 1\n",
      "     |          Order of moment.\n",
      "     |      arg1, arg2, arg3,... : float\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information).\n",
      "     |      loc : array_like, optional\n",
      "     |          location parameter (default=0)\n",
      "     |      scale : array_like, optional\n",
      "     |          scale parameter (default=1)\n",
      "     |  \n",
      "     |  stats(self, *args, **kwds)\n",
      "     |      Some statistics of the given RV.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information)\n",
      "     |      loc : array_like, optional\n",
      "     |          location parameter (default=0)\n",
      "     |      scale : array_like, optional (continuous RVs only)\n",
      "     |          scale parameter (default=1)\n",
      "     |      moments : str, optional\n",
      "     |          composed of letters ['mvsk'] defining which moments to compute:\n",
      "     |          'm' = mean,\n",
      "     |          'v' = variance,\n",
      "     |          's' = (Fisher's) skew,\n",
      "     |          'k' = (Fisher's) kurtosis.\n",
      "     |          (default is 'mv')\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      stats : sequence\n",
      "     |          of requested moments.\n",
      "     |  \n",
      "     |  std(self, *args, **kwds)\n",
      "     |      Standard deviation of the distribution.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information)\n",
      "     |      loc : array_like, optional\n",
      "     |          location parameter (default=0)\n",
      "     |      scale : array_like, optional\n",
      "     |          scale parameter (default=1)\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      std : float\n",
      "     |          standard deviation of the distribution\n",
      "     |  \n",
      "     |  support(self, *args, **kwargs)\n",
      "     |      Support of the distribution.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      arg1, arg2, ... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information).\n",
      "     |      loc : array_like, optional\n",
      "     |          location parameter, Default is 0.\n",
      "     |      scale : array_like, optional\n",
      "     |          scale parameter, Default is 1.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      a, b : array_like\n",
      "     |          end-points of the distribution's support.\n",
      "     |  \n",
      "     |  var(self, *args, **kwds)\n",
      "     |      Variance of the distribution.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information)\n",
      "     |      loc : array_like, optional\n",
      "     |          location parameter (default=0)\n",
      "     |      scale : array_like, optional\n",
      "     |          scale parameter (default=1)\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      var : float\n",
      "     |          the variance of the distribution\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from rv_generic:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |      dictionary for instance variables (if defined)\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "     |  \n",
      "     |  random_state\n",
      "     |      Get or set the generator object for generating random variates.\n",
      "     |      \n",
      "     |      If `seed` is None (or `np.random`), the `numpy.random.RandomState`\n",
      "     |      singleton is used.\n",
      "     |      If `seed` is an int, a new ``RandomState`` instance is used,\n",
      "     |      seeded with `seed`.\n",
      "     |      If `seed` is already a ``Generator`` or ``RandomState`` instance then\n",
      "     |      that instance is used.\n",
      "    \n",
      "    class rv_histogram(scipy.stats._distn_infrastructure.rv_continuous)\n",
      "     |  rv_histogram(histogram, *args, **kwargs)\n",
      "     |  \n",
      "     |  Generates a distribution given by a histogram.\n",
      "     |  This is useful to generate a template distribution from a binned\n",
      "     |  datasample.\n",
      "     |  \n",
      "     |  As a subclass of the `rv_continuous` class, `rv_histogram` inherits from it\n",
      "     |  a collection of generic methods (see `rv_continuous` for the full list),\n",
      "     |  and implements them based on the properties of the provided binned\n",
      "     |  datasample.\n",
      "     |  \n",
      "     |  Parameters\n",
      "     |  ----------\n",
      "     |  histogram : tuple of array_like\n",
      "     |    Tuple containing two array_like objects\n",
      "     |    The first containing the content of n bins\n",
      "     |    The second containing the (n+1) bin boundaries\n",
      "     |    In particular the return value np.histogram is accepted\n",
      "     |  \n",
      "     |  Notes\n",
      "     |  -----\n",
      "     |  There are no additional shape parameters except for the loc and scale.\n",
      "     |  The pdf is defined as a stepwise function from the provided histogram\n",
      "     |  The cdf is a linear interpolation of the pdf.\n",
      "     |  \n",
      "     |  .. versionadded:: 0.19.0\n",
      "     |  \n",
      "     |  Examples\n",
      "     |  --------\n",
      "     |  \n",
      "     |  Create a scipy.stats distribution from a numpy histogram\n",
      "     |  \n",
      "     |  >>> import scipy.stats\n",
      "     |  >>> import numpy as np\n",
      "     |  >>> data = scipy.stats.norm.rvs(size=100000, loc=0, scale=1.5, random_state=123)\n",
      "     |  >>> hist = np.histogram(data, bins=100)\n",
      "     |  >>> hist_dist = scipy.stats.rv_histogram(hist)\n",
      "     |  \n",
      "     |  Behaves like an ordinary scipy rv_continuous distribution\n",
      "     |  \n",
      "     |  >>> hist_dist.pdf(1.0)\n",
      "     |  0.20538577847618705\n",
      "     |  >>> hist_dist.cdf(2.0)\n",
      "     |  0.90818568543056499\n",
      "     |  \n",
      "     |  PDF is zero above (below) the highest (lowest) bin of the histogram,\n",
      "     |  defined by the max (min) of the original dataset\n",
      "     |  \n",
      "     |  >>> hist_dist.pdf(np.max(data))\n",
      "     |  0.0\n",
      "     |  >>> hist_dist.cdf(np.max(data))\n",
      "     |  1.0\n",
      "     |  >>> hist_dist.pdf(np.min(data))\n",
      "     |  7.7591907244498314e-05\n",
      "     |  >>> hist_dist.cdf(np.min(data))\n",
      "     |  0.0\n",
      "     |  \n",
      "     |  PDF and CDF follow the histogram\n",
      "     |  \n",
      "     |  >>> import matplotlib.pyplot as plt\n",
      "     |  >>> X = np.linspace(-5.0, 5.0, 100)\n",
      "     |  >>> plt.title(\"PDF from Template\")\n",
      "     |  >>> plt.hist(data, density=True, bins=100)\n",
      "     |  >>> plt.plot(X, hist_dist.pdf(X), label='PDF')\n",
      "     |  >>> plt.plot(X, hist_dist.cdf(X), label='CDF')\n",
      "     |  >>> plt.show()\n",
      "     |  \n",
      "     |  Method resolution order:\n",
      "     |      rv_histogram\n",
      "     |      scipy.stats._distn_infrastructure.rv_continuous\n",
      "     |      scipy.stats._distn_infrastructure.rv_generic\n",
      "     |      builtins.object\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __init__(self, histogram, *args, **kwargs)\n",
      "     |      Create a new distribution using the given histogram\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      histogram : tuple of array_like\n",
      "     |        Tuple containing two array_like objects\n",
      "     |        The first containing the content of n bins\n",
      "     |        The second containing the (n+1) bin boundaries\n",
      "     |        In particular the return value np.histogram is accepted\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from scipy.stats._distn_infrastructure.rv_continuous:\n",
      "     |  \n",
      "     |  __getstate__(self)\n",
      "     |  \n",
      "     |  cdf(self, x, *args, **kwds)\n",
      "     |      Cumulative distribution function of the given RV.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      x : array_like\n",
      "     |          quantiles\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information)\n",
      "     |      loc : array_like, optional\n",
      "     |          location parameter (default=0)\n",
      "     |      scale : array_like, optional\n",
      "     |          scale parameter (default=1)\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      cdf : ndarray\n",
      "     |          Cumulative distribution function evaluated at `x`\n",
      "     |  \n",
      "     |  expect(self, func=None, args=(), loc=0, scale=1, lb=None, ub=None, conditional=False, **kwds)\n",
      "     |      Calculate expected value of a function with respect to the\n",
      "     |      distribution by numerical integration.\n",
      "     |      \n",
      "     |      The expected value of a function ``f(x)`` with respect to a\n",
      "     |      distribution ``dist`` is defined as::\n",
      "     |      \n",
      "     |                  ub\n",
      "     |          E[f(x)] = Integral(f(x) * dist.pdf(x)),\n",
      "     |                  lb\n",
      "     |      \n",
      "     |      where ``ub`` and ``lb`` are arguments and ``x`` has the ``dist.pdf(x)``\n",
      "     |      distribution. If the bounds ``lb`` and ``ub`` correspond to the\n",
      "     |      support of the distribution, e.g. ``[-inf, inf]`` in the default\n",
      "     |      case, then the integral is the unrestricted expectation of ``f(x)``.\n",
      "     |      Also, the function ``f(x)`` may be defined such that ``f(x)`` is ``0``\n",
      "     |      outside a finite interval in which case the expectation is\n",
      "     |      calculated within the finite range ``[lb, ub]``.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      func : callable, optional\n",
      "     |          Function for which integral is calculated. Takes only one argument.\n",
      "     |          The default is the identity mapping f(x) = x.\n",
      "     |      args : tuple, optional\n",
      "     |          Shape parameters of the distribution.\n",
      "     |      loc : float, optional\n",
      "     |          Location parameter (default=0).\n",
      "     |      scale : float, optional\n",
      "     |          Scale parameter (default=1).\n",
      "     |      lb, ub : scalar, optional\n",
      "     |          Lower and upper bound for integration. Default is set to the\n",
      "     |          support of the distribution.\n",
      "     |      conditional : bool, optional\n",
      "     |          If True, the integral is corrected by the conditional probability\n",
      "     |          of the integration interval.  The return value is the expectation\n",
      "     |          of the function, conditional on being in the given interval.\n",
      "     |          Default is False.\n",
      "     |      \n",
      "     |      Additional keyword arguments are passed to the integration routine.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      expect : float\n",
      "     |          The calculated expected value.\n",
      "     |      \n",
      "     |      Notes\n",
      "     |      -----\n",
      "     |      The integration behavior of this function is inherited from\n",
      "     |      `scipy.integrate.quad`. Neither this function nor\n",
      "     |      `scipy.integrate.quad` can verify whether the integral exists or is\n",
      "     |      finite. For example ``cauchy(0).mean()`` returns ``np.nan`` and\n",
      "     |      ``cauchy(0).expect()`` returns ``0.0``.\n",
      "     |      \n",
      "     |      The function is not vectorized.\n",
      "     |      \n",
      "     |      Examples\n",
      "     |      --------\n",
      "     |      \n",
      "     |      To understand the effect of the bounds of integration consider\n",
      "     |      \n",
      "     |      >>> from scipy.stats import expon\n",
      "     |      >>> expon(1).expect(lambda x: 1, lb=0.0, ub=2.0)\n",
      "     |      0.6321205588285578\n",
      "     |      \n",
      "     |      This is close to\n",
      "     |      \n",
      "     |      >>> expon(1).cdf(2.0) - expon(1).cdf(0.0)\n",
      "     |      0.6321205588285577\n",
      "     |      \n",
      "     |      If ``conditional=True``\n",
      "     |      \n",
      "     |      >>> expon(1).expect(lambda x: 1, lb=0.0, ub=2.0, conditional=True)\n",
      "     |      1.0000000000000002\n",
      "     |      \n",
      "     |      The slight deviation from 1 is due to numerical integration.\n",
      "     |  \n",
      "     |  fit(self, data, *args, **kwds)\n",
      "     |      Return estimates of shape (if applicable), location, and scale\n",
      "     |      parameters from data. The default estimation method is Maximum\n",
      "     |      Likelihood Estimation (MLE), but Method of Moments (MM)\n",
      "     |      is also available.\n",
      "     |      \n",
      "     |      Starting estimates for\n",
      "     |      the fit are given by input arguments; for any arguments not provided\n",
      "     |      with starting estimates, ``self._fitstart(data)`` is called to generate\n",
      "     |      such.\n",
      "     |      \n",
      "     |      One can hold some parameters fixed to specific values by passing in\n",
      "     |      keyword arguments ``f0``, ``f1``, ..., ``fn`` (for shape parameters)\n",
      "     |      and ``floc`` and ``fscale`` (for location and scale parameters,\n",
      "     |      respectively).\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      data : array_like\n",
      "     |          Data to use in estimating the distribution parameters.\n",
      "     |      arg1, arg2, arg3,... : floats, optional\n",
      "     |          Starting value(s) for any shape-characterizing arguments (those not\n",
      "     |          provided will be determined by a call to ``_fitstart(data)``).\n",
      "     |          No default value.\n",
      "     |      kwds : floats, optional\n",
      "     |          - `loc`: initial guess of the distribution's location parameter.\n",
      "     |          - `scale`: initial guess of the distribution's scale parameter.\n",
      "     |      \n",
      "     |          Special keyword arguments are recognized as holding certain\n",
      "     |          parameters fixed:\n",
      "     |      \n",
      "     |          - f0...fn : hold respective shape parameters fixed.\n",
      "     |            Alternatively, shape parameters to fix can be specified by name.\n",
      "     |            For example, if ``self.shapes == \"a, b\"``, ``fa`` and ``fix_a``\n",
      "     |            are equivalent to ``f0``, and ``fb`` and ``fix_b`` are\n",
      "     |            equivalent to ``f1``.\n",
      "     |      \n",
      "     |          - floc : hold location parameter fixed to specified value.\n",
      "     |      \n",
      "     |          - fscale : hold scale parameter fixed to specified value.\n",
      "     |      \n",
      "     |          - optimizer : The optimizer to use.\n",
      "     |            The optimizer must take ``func``,\n",
      "     |            and starting position as the first two arguments,\n",
      "     |            plus ``args`` (for extra arguments to pass to the\n",
      "     |            function to be optimized) and ``disp=0`` to suppress\n",
      "     |            output as keyword arguments.\n",
      "     |      \n",
      "     |          - method : The method to use. The default is \"MLE\" (Maximum\n",
      "     |            Likelihood Estimate); \"MM\" (Method of Moments)\n",
      "     |            is also available.\n",
      "     |      \n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      parameter_tuple : tuple of floats\n",
      "     |          Estimates for any shape parameters (if applicable),\n",
      "     |          followed by those for location and scale.\n",
      "     |          For most random variables, shape statistics\n",
      "     |          will be returned, but there are exceptions (e.g. ``norm``).\n",
      "     |      \n",
      "     |      Notes\n",
      "     |      -----\n",
      "     |      With ``method=\"MLE\"`` (default), the fit is computed by minimizing\n",
      "     |      the negative log-likelihood function. A large, finite penalty\n",
      "     |      (rather than infinite negative log-likelihood) is applied for\n",
      "     |      observations beyond the support of the distribution.\n",
      "     |      \n",
      "     |      With ``method=\"MM\"``, the fit is computed by minimizing the L2 norm\n",
      "     |      of the relative errors between the first *k* raw (about zero) data\n",
      "     |      moments and the corresponding distribution moments, where *k* is the\n",
      "     |      number of non-fixed parameters.\n",
      "     |      More precisely, the objective function is::\n",
      "     |      \n",
      "     |          (((data_moments - dist_moments)\n",
      "     |            / np.maximum(np.abs(data_moments), 1e-8))**2).sum()\n",
      "     |      \n",
      "     |      where the constant ``1e-8`` avoids division by zero in case of\n",
      "     |      vanishing data moments. Typically, this error norm can be reduced to\n",
      "     |      zero.\n",
      "     |      Note that the standard method of moments can produce parameters for\n",
      "     |      which some data are outside the support of the fitted distribution;\n",
      "     |      this implementation does nothing to prevent this.\n",
      "     |      \n",
      "     |      For either method,\n",
      "     |      the returned answer is not guaranteed to be globally optimal; it\n",
      "     |      may only be locally optimal, or the optimization may fail altogether.\n",
      "     |      If the data contain any of ``np.nan``, ``np.inf``, or ``-np.inf``,\n",
      "     |      the `fit` method will raise a ``RuntimeError``.\n",
      "     |      \n",
      "     |      Examples\n",
      "     |      --------\n",
      "     |      \n",
      "     |      Generate some data to fit: draw random variates from the `beta`\n",
      "     |      distribution\n",
      "     |      \n",
      "     |      >>> from scipy.stats import beta\n",
      "     |      >>> a, b = 1., 2.\n",
      "     |      >>> x = beta.rvs(a, b, size=1000)\n",
      "     |      \n",
      "     |      Now we can fit all four parameters (``a``, ``b``, ``loc``\n",
      "     |      and ``scale``):\n",
      "     |      \n",
      "     |      >>> a1, b1, loc1, scale1 = beta.fit(x)\n",
      "     |      \n",
      "     |      We can also use some prior knowledge about the dataset: let's keep\n",
      "     |      ``loc`` and ``scale`` fixed:\n",
      "     |      \n",
      "     |      >>> a1, b1, loc1, scale1 = beta.fit(x, floc=0, fscale=1)\n",
      "     |      >>> loc1, scale1\n",
      "     |      (0, 1)\n",
      "     |      \n",
      "     |      We can also keep shape parameters fixed by using ``f``-keywords. To\n",
      "     |      keep the zero-th shape parameter ``a`` equal 1, use ``f0=1`` or,\n",
      "     |      equivalently, ``fa=1``:\n",
      "     |      \n",
      "     |      >>> a1, b1, loc1, scale1 = beta.fit(x, fa=1, floc=0, fscale=1)\n",
      "     |      >>> a1\n",
      "     |      1\n",
      "     |      \n",
      "     |      Not all distributions return estimates for the shape parameters.\n",
      "     |      ``norm`` for example just returns estimates for location and scale:\n",
      "     |      \n",
      "     |      >>> from scipy.stats import norm\n",
      "     |      >>> x = norm.rvs(a, b, size=1000, random_state=123)\n",
      "     |      >>> loc1, scale1 = norm.fit(x)\n",
      "     |      >>> loc1, scale1\n",
      "     |      (0.92087172783841631, 2.0015750750324668)\n",
      "     |  \n",
      "     |  fit_loc_scale(self, data, *args)\n",
      "     |      Estimate loc and scale parameters from data using 1st and 2nd moments.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      data : array_like\n",
      "     |          Data to fit.\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information).\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      Lhat : float\n",
      "     |          Estimated location parameter for the data.\n",
      "     |      Shat : float\n",
      "     |          Estimated scale parameter for the data.\n",
      "     |  \n",
      "     |  isf(self, q, *args, **kwds)\n",
      "     |      Inverse survival function (inverse of `sf`) at q of the given RV.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      q : array_like\n",
      "     |          upper tail probability\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information)\n",
      "     |      loc : array_like, optional\n",
      "     |          location parameter (default=0)\n",
      "     |      scale : array_like, optional\n",
      "     |          scale parameter (default=1)\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      x : ndarray or scalar\n",
      "     |          Quantile corresponding to the upper tail probability q.\n",
      "     |  \n",
      "     |  logcdf(self, x, *args, **kwds)\n",
      "     |      Log of the cumulative distribution function at x of the given RV.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      x : array_like\n",
      "     |          quantiles\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information)\n",
      "     |      loc : array_like, optional\n",
      "     |          location parameter (default=0)\n",
      "     |      scale : array_like, optional\n",
      "     |          scale parameter (default=1)\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      logcdf : array_like\n",
      "     |          Log of the cumulative distribution function evaluated at x\n",
      "     |  \n",
      "     |  logpdf(self, x, *args, **kwds)\n",
      "     |      Log of the probability density function at x of the given RV.\n",
      "     |      \n",
      "     |      This uses a more numerically accurate calculation if available.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      x : array_like\n",
      "     |          quantiles\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information)\n",
      "     |      loc : array_like, optional\n",
      "     |          location parameter (default=0)\n",
      "     |      scale : array_like, optional\n",
      "     |          scale parameter (default=1)\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      logpdf : array_like\n",
      "     |          Log of the probability density function evaluated at x\n",
      "     |  \n",
      "     |  logsf(self, x, *args, **kwds)\n",
      "     |      Log of the survival function of the given RV.\n",
      "     |      \n",
      "     |      Returns the log of the \"survival function,\" defined as (1 - `cdf`),\n",
      "     |      evaluated at `x`.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      x : array_like\n",
      "     |          quantiles\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information)\n",
      "     |      loc : array_like, optional\n",
      "     |          location parameter (default=0)\n",
      "     |      scale : array_like, optional\n",
      "     |          scale parameter (default=1)\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      logsf : ndarray\n",
      "     |          Log of the survival function evaluated at `x`.\n",
      "     |  \n",
      "     |  nnlf(self, theta, x)\n",
      "     |      Negative loglikelihood function.\n",
      "     |      \n",
      "     |      Notes\n",
      "     |      -----\n",
      "     |      This is ``-sum(log pdf(x, theta), axis=0)`` where `theta` are the\n",
      "     |      parameters (including loc and scale).\n",
      "     |  \n",
      "     |  pdf(self, x, *args, **kwds)\n",
      "     |      Probability density function at x of the given RV.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      x : array_like\n",
      "     |          quantiles\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information)\n",
      "     |      loc : array_like, optional\n",
      "     |          location parameter (default=0)\n",
      "     |      scale : array_like, optional\n",
      "     |          scale parameter (default=1)\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      pdf : ndarray\n",
      "     |          Probability density function evaluated at x\n",
      "     |  \n",
      "     |  ppf(self, q, *args, **kwds)\n",
      "     |      Percent point function (inverse of `cdf`) at q of the given RV.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      q : array_like\n",
      "     |          lower tail probability\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information)\n",
      "     |      loc : array_like, optional\n",
      "     |          location parameter (default=0)\n",
      "     |      scale : array_like, optional\n",
      "     |          scale parameter (default=1)\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      x : array_like\n",
      "     |          quantile corresponding to the lower tail probability q.\n",
      "     |  \n",
      "     |  sf(self, x, *args, **kwds)\n",
      "     |      Survival function (1 - `cdf`) at x of the given RV.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      x : array_like\n",
      "     |          quantiles\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information)\n",
      "     |      loc : array_like, optional\n",
      "     |          location parameter (default=0)\n",
      "     |      scale : array_like, optional\n",
      "     |          scale parameter (default=1)\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      sf : array_like\n",
      "     |          Survival function evaluated at x\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from scipy.stats._distn_infrastructure.rv_generic:\n",
      "     |  \n",
      "     |  __call__(self, *args, **kwds)\n",
      "     |      Freeze the distribution for the given arguments.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution.  Should include all\n",
      "     |          the non-optional arguments, may include ``loc`` and ``scale``.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      rv_frozen : rv_frozen instance\n",
      "     |          The frozen distribution.\n",
      "     |  \n",
      "     |  __setstate__(self, state)\n",
      "     |  \n",
      "     |  entropy(self, *args, **kwds)\n",
      "     |      Differential entropy of the RV.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information).\n",
      "     |      loc : array_like, optional\n",
      "     |          Location parameter (default=0).\n",
      "     |      scale : array_like, optional  (continuous distributions only).\n",
      "     |          Scale parameter (default=1).\n",
      "     |      \n",
      "     |      Notes\n",
      "     |      -----\n",
      "     |      Entropy is defined base `e`:\n",
      "     |      \n",
      "     |      >>> drv = rv_discrete(values=((0, 1), (0.5, 0.5)))\n",
      "     |      >>> np.allclose(drv.entropy(), np.log(2.0))\n",
      "     |      True\n",
      "     |  \n",
      "     |  freeze(self, *args, **kwds)\n",
      "     |      Freeze the distribution for the given arguments.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution.  Should include all\n",
      "     |          the non-optional arguments, may include ``loc`` and ``scale``.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      rv_frozen : rv_frozen instance\n",
      "     |          The frozen distribution.\n",
      "     |  \n",
      "     |  interval(self, alpha, *args, **kwds)\n",
      "     |      Confidence interval with equal areas around the median.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      alpha : array_like of float\n",
      "     |          Probability that an rv will be drawn from the returned range.\n",
      "     |          Each value should be in the range [0, 1].\n",
      "     |      arg1, arg2, ... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information).\n",
      "     |      loc : array_like, optional\n",
      "     |          location parameter, Default is 0.\n",
      "     |      scale : array_like, optional\n",
      "     |          scale parameter, Default is 1.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      a, b : ndarray of float\n",
      "     |          end-points of range that contain ``100 * alpha %`` of the rv's\n",
      "     |          possible values.\n",
      "     |  \n",
      "     |  mean(self, *args, **kwds)\n",
      "     |      Mean of the distribution.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information)\n",
      "     |      loc : array_like, optional\n",
      "     |          location parameter (default=0)\n",
      "     |      scale : array_like, optional\n",
      "     |          scale parameter (default=1)\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      mean : float\n",
      "     |          the mean of the distribution\n",
      "     |  \n",
      "     |  median(self, *args, **kwds)\n",
      "     |      Median of the distribution.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information)\n",
      "     |      loc : array_like, optional\n",
      "     |          Location parameter, Default is 0.\n",
      "     |      scale : array_like, optional\n",
      "     |          Scale parameter, Default is 1.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      median : float\n",
      "     |          The median of the distribution.\n",
      "     |      \n",
      "     |      See Also\n",
      "     |      --------\n",
      "     |      rv_discrete.ppf\n",
      "     |          Inverse of the CDF\n",
      "     |  \n",
      "     |  moment(self, n, *args, **kwds)\n",
      "     |      n-th order non-central moment of distribution.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      n : int, n >= 1\n",
      "     |          Order of moment.\n",
      "     |      arg1, arg2, arg3,... : float\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information).\n",
      "     |      loc : array_like, optional\n",
      "     |          location parameter (default=0)\n",
      "     |      scale : array_like, optional\n",
      "     |          scale parameter (default=1)\n",
      "     |  \n",
      "     |  rvs(self, *args, **kwds)\n",
      "     |      Random variates of given type.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information).\n",
      "     |      loc : array_like, optional\n",
      "     |          Location parameter (default=0).\n",
      "     |      scale : array_like, optional\n",
      "     |          Scale parameter (default=1).\n",
      "     |      size : int or tuple of ints, optional\n",
      "     |          Defining number of random variates (default is 1).\n",
      "     |      random_state : {None, int, `numpy.random.Generator`,\n",
      "     |                      `numpy.random.RandomState`}, optional\n",
      "     |      \n",
      "     |          If `seed` is None (or `np.random`), the `numpy.random.RandomState`\n",
      "     |          singleton is used.\n",
      "     |          If `seed` is an int, a new ``RandomState`` instance is used,\n",
      "     |          seeded with `seed`.\n",
      "     |          If `seed` is already a ``Generator`` or ``RandomState`` instance\n",
      "     |          then that instance is used.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      rvs : ndarray or scalar\n",
      "     |          Random variates of given `size`.\n",
      "     |  \n",
      "     |  stats(self, *args, **kwds)\n",
      "     |      Some statistics of the given RV.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information)\n",
      "     |      loc : array_like, optional\n",
      "     |          location parameter (default=0)\n",
      "     |      scale : array_like, optional (continuous RVs only)\n",
      "     |          scale parameter (default=1)\n",
      "     |      moments : str, optional\n",
      "     |          composed of letters ['mvsk'] defining which moments to compute:\n",
      "     |          'm' = mean,\n",
      "     |          'v' = variance,\n",
      "     |          's' = (Fisher's) skew,\n",
      "     |          'k' = (Fisher's) kurtosis.\n",
      "     |          (default is 'mv')\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      stats : sequence\n",
      "     |          of requested moments.\n",
      "     |  \n",
      "     |  std(self, *args, **kwds)\n",
      "     |      Standard deviation of the distribution.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information)\n",
      "     |      loc : array_like, optional\n",
      "     |          location parameter (default=0)\n",
      "     |      scale : array_like, optional\n",
      "     |          scale parameter (default=1)\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      std : float\n",
      "     |          standard deviation of the distribution\n",
      "     |  \n",
      "     |  support(self, *args, **kwargs)\n",
      "     |      Support of the distribution.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      arg1, arg2, ... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information).\n",
      "     |      loc : array_like, optional\n",
      "     |          location parameter, Default is 0.\n",
      "     |      scale : array_like, optional\n",
      "     |          scale parameter, Default is 1.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      a, b : array_like\n",
      "     |          end-points of the distribution's support.\n",
      "     |  \n",
      "     |  var(self, *args, **kwds)\n",
      "     |      Variance of the distribution.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      arg1, arg2, arg3,... : array_like\n",
      "     |          The shape parameter(s) for the distribution (see docstring of the\n",
      "     |          instance object for more information)\n",
      "     |      loc : array_like, optional\n",
      "     |          location parameter (default=0)\n",
      "     |      scale : array_like, optional\n",
      "     |          scale parameter (default=1)\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      var : float\n",
      "     |          the variance of the distribution\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from scipy.stats._distn_infrastructure.rv_generic:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |      dictionary for instance variables (if defined)\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "     |  \n",
      "     |  random_state\n",
      "     |      Get or set the generator object for generating random variates.\n",
      "     |      \n",
      "     |      If `seed` is None (or `np.random`), the `numpy.random.RandomState`\n",
      "     |      singleton is used.\n",
      "     |      If `seed` is an int, a new ``RandomState`` instance is used,\n",
      "     |      seeded with `seed`.\n",
      "     |      If `seed` is already a ``Generator`` or ``RandomState`` instance then\n",
      "     |      that instance is used.\n",
      "\n",
      "FUNCTIONS\n",
      "    alexandergovern(*args, nan_policy='propagate')\n",
      "        Performs the Alexander Govern test.\n",
      "        \n",
      "        The Alexander-Govern approximation tests the equality of k independent\n",
      "        means in the face of heterogeneity of variance. The test is applied to\n",
      "        samples from two or more groups, possibly with differing sizes.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        sample1, sample2, ... : array_like\n",
      "            The sample measurements for each group.  There must be at least\n",
      "            two samples.\n",
      "        nan_policy : {'propagate', 'raise', 'omit'}, optional\n",
      "            Defines how to handle when input contains nan.\n",
      "            The following options are available (default is 'propagate'):\n",
      "        \n",
      "            * 'propagate': returns nan\n",
      "            * 'raise': throws an error\n",
      "            * 'omit': performs the calculations ignoring nan values\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        statistic : float\n",
      "            The computed A statistic of the test.\n",
      "        pvalue : float\n",
      "            The associated p-value from the chi-squared distribution.\n",
      "        \n",
      "        Warns\n",
      "        -----\n",
      "        AlexanderGovernConstantInputWarning\n",
      "            Raised if an input is a constant array.  The statistic is not defined\n",
      "            in this case, so ``np.nan`` is returned.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        f_oneway : one-way ANOVA\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        The use of this test relies on several assumptions.\n",
      "        \n",
      "        1. The samples are independent.\n",
      "        2. Each sample is from a normally distributed population.\n",
      "        3. Unlike `f_oneway`, this test does not assume on homoscedasticity,\n",
      "           instead relaxing the assumption of equal variances.\n",
      "        \n",
      "        Input samples must be finite, one dimensional, and with size greater than\n",
      "        one.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] Alexander, Ralph A., and Diane M. Govern. \"A New and Simpler\n",
      "               Approximation for ANOVA under Variance Heterogeneity.\" Journal\n",
      "               of Educational Statistics, vol. 19, no. 2, 1994, pp. 91-101.\n",
      "               JSTOR, www.jstor.org/stable/1165140. Accessed 12 Sept. 2020.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy.stats import alexandergovern\n",
      "        \n",
      "        Here are some data on annual percentage rate of interest charged on\n",
      "        new car loans at nine of the largest banks in four American cities\n",
      "        taken from the National Institute of Standards and Technology's\n",
      "        ANOVA dataset.\n",
      "        \n",
      "        We use `alexandergovern` to test the null hypothesis that all cities\n",
      "        have the same mean APR against the alternative that the cities do not\n",
      "        all have the same mean APR. We decide that a sigificance level of 5%\n",
      "        is required to reject the null hypothesis in favor of the alternative.\n",
      "        \n",
      "        >>> atlanta = [13.75, 13.75, 13.5, 13.5, 13.0, 13.0, 13.0, 12.75, 12.5]\n",
      "        >>> chicago = [14.25, 13.0, 12.75, 12.5, 12.5, 12.4, 12.3, 11.9, 11.9]\n",
      "        >>> houston = [14.0, 14.0, 13.51, 13.5, 13.5, 13.25, 13.0, 12.5, 12.5]\n",
      "        >>> memphis = [15.0, 14.0, 13.75, 13.59, 13.25, 12.97, 12.5, 12.25,\n",
      "        ...           11.89]\n",
      "        >>> alexandergovern(atlanta, chicago, houston, memphis)\n",
      "        AlexanderGovernResult(statistic=4.65087071883494,\n",
      "                              pvalue=0.19922132490385214)\n",
      "        \n",
      "        The p-value is 0.1992, indicating a nearly 20% chance of observing\n",
      "        such an extreme value of the test statistic under the null hypothesis.\n",
      "        This exceeds 5%, so we do not reject the null hypothesis in favor of\n",
      "        the alternative.\n",
      "    \n",
      "    anderson(x, dist='norm')\n",
      "        Anderson-Darling test for data coming from a particular distribution.\n",
      "        \n",
      "        The Anderson-Darling test tests the null hypothesis that a sample is\n",
      "        drawn from a population that follows a particular distribution.\n",
      "        For the Anderson-Darling test, the critical values depend on\n",
      "        which distribution is being tested against.  This function works\n",
      "        for normal, exponential, logistic, or Gumbel (Extreme Value\n",
      "        Type I) distributions.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x : array_like\n",
      "            Array of sample data.\n",
      "        dist : {'norm', 'expon', 'logistic', 'gumbel', 'gumbel_l', 'gumbel_r', 'extreme1'}, optional\n",
      "            The type of distribution to test against.  The default is 'norm'.\n",
      "            The names 'extreme1', 'gumbel_l' and 'gumbel' are synonyms for the\n",
      "            same distribution.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        statistic : float\n",
      "            The Anderson-Darling test statistic.\n",
      "        critical_values : list\n",
      "            The critical values for this distribution.\n",
      "        significance_level : list\n",
      "            The significance levels for the corresponding critical values\n",
      "            in percents.  The function returns critical values for a\n",
      "            differing set of significance levels depending on the\n",
      "            distribution that is being tested against.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        kstest : The Kolmogorov-Smirnov test for goodness-of-fit.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        Critical values provided are for the following significance levels:\n",
      "        \n",
      "        normal/exponential\n",
      "            15%, 10%, 5%, 2.5%, 1%\n",
      "        logistic\n",
      "            25%, 10%, 5%, 2.5%, 1%, 0.5%\n",
      "        Gumbel\n",
      "            25%, 10%, 5%, 2.5%, 1%\n",
      "        \n",
      "        If the returned statistic is larger than these critical values then\n",
      "        for the corresponding significance level, the null hypothesis that\n",
      "        the data come from the chosen distribution can be rejected.\n",
      "        The returned statistic is referred to as 'A2' in the references.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] https://www.itl.nist.gov/div898/handbook/prc/section2/prc213.htm\n",
      "        .. [2] Stephens, M. A. (1974). EDF Statistics for Goodness of Fit and\n",
      "               Some Comparisons, Journal of the American Statistical Association,\n",
      "               Vol. 69, pp. 730-737.\n",
      "        .. [3] Stephens, M. A. (1976). Asymptotic Results for Goodness-of-Fit\n",
      "               Statistics with Unknown Parameters, Annals of Statistics, Vol. 4,\n",
      "               pp. 357-369.\n",
      "        .. [4] Stephens, M. A. (1977). Goodness of Fit for the Extreme Value\n",
      "               Distribution, Biometrika, Vol. 64, pp. 583-588.\n",
      "        .. [5] Stephens, M. A. (1977). Goodness of Fit with Special Reference\n",
      "               to Tests for Exponentiality , Technical Report No. 262,\n",
      "               Department of Statistics, Stanford University, Stanford, CA.\n",
      "        .. [6] Stephens, M. A. (1979). Tests of Fit for the Logistic Distribution\n",
      "               Based on the Empirical Distribution Function, Biometrika, Vol. 66,\n",
      "               pp. 591-595.\n",
      "    \n",
      "    anderson_ksamp(samples, midrank=True)\n",
      "        The Anderson-Darling test for k-samples.\n",
      "        \n",
      "        The k-sample Anderson-Darling test is a modification of the\n",
      "        one-sample Anderson-Darling test. It tests the null hypothesis\n",
      "        that k-samples are drawn from the same population without having\n",
      "        to specify the distribution function of that population. The\n",
      "        critical values depend on the number of samples.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        samples : sequence of 1-D array_like\n",
      "            Array of sample data in arrays.\n",
      "        midrank : bool, optional\n",
      "            Type of Anderson-Darling test which is computed. Default\n",
      "            (True) is the midrank test applicable to continuous and\n",
      "            discrete populations. If False, the right side empirical\n",
      "            distribution is used.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        statistic : float\n",
      "            Normalized k-sample Anderson-Darling test statistic.\n",
      "        critical_values : array\n",
      "            The critical values for significance levels 25%, 10%, 5%, 2.5%, 1%,\n",
      "            0.5%, 0.1%.\n",
      "        significance_level : float\n",
      "            An approximate significance level at which the null hypothesis for the\n",
      "            provided samples can be rejected. The value is floored / capped at\n",
      "            0.1% / 25%.\n",
      "        \n",
      "        Raises\n",
      "        ------\n",
      "        ValueError\n",
      "            If less than 2 samples are provided, a sample is empty, or no\n",
      "            distinct observations are in the samples.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        ks_2samp : 2 sample Kolmogorov-Smirnov test\n",
      "        anderson : 1 sample Anderson-Darling test\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        [1]_ defines three versions of the k-sample Anderson-Darling test:\n",
      "        one for continuous distributions and two for discrete\n",
      "        distributions, in which ties between samples may occur. The\n",
      "        default of this routine is to compute the version based on the\n",
      "        midrank empirical distribution function. This test is applicable\n",
      "        to continuous and discrete data. If midrank is set to False, the\n",
      "        right side empirical distribution is used for a test for discrete\n",
      "        data. According to [1]_, the two discrete test statistics differ\n",
      "        only slightly if a few collisions due to round-off errors occur in\n",
      "        the test not adjusted for ties between samples.\n",
      "        \n",
      "        The critical values corresponding to the significance levels from 0.01\n",
      "        to 0.25 are taken from [1]_. p-values are floored / capped\n",
      "        at 0.1% / 25%. Since the range of critical values might be extended in\n",
      "        future releases, it is recommended not to test ``p == 0.25``, but rather\n",
      "        ``p >= 0.25`` (analogously for the lower bound).\n",
      "        \n",
      "        .. versionadded:: 0.14.0\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] Scholz, F. W and Stephens, M. A. (1987), K-Sample\n",
      "               Anderson-Darling Tests, Journal of the American Statistical\n",
      "               Association, Vol. 82, pp. 918-924.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy import stats\n",
      "        >>> rng = np.random.default_rng()\n",
      "        \n",
      "        The null hypothesis that the two random samples come from the same\n",
      "        distribution can be rejected at the 5% level because the returned\n",
      "        test value is greater than the critical value for 5% (1.961) but\n",
      "        not at the 2.5% level. The interpolation gives an approximate\n",
      "        significance level of 3.2%:\n",
      "        \n",
      "        >>> stats.anderson_ksamp([rng.normal(size=50),\n",
      "        ... rng.normal(loc=0.5, size=30)])\n",
      "        (1.974403288713695,\n",
      "          array([0.325, 1.226, 1.961, 2.718, 3.752, 4.592, 6.546]),\n",
      "          0.04991293614572478)\n",
      "        \n",
      "        \n",
      "        The null hypothesis cannot be rejected for three samples from an\n",
      "        identical distribution. The reported p-value (25%) has been capped and\n",
      "        may not be very accurate (since it corresponds to the value 0.449\n",
      "        whereas the statistic is -0.731):\n",
      "        \n",
      "        >>> stats.anderson_ksamp([rng.normal(size=50),\n",
      "        ... rng.normal(size=30), rng.normal(size=20)])\n",
      "        (-0.29103725200789504,\n",
      "          array([ 0.44925884,  1.3052767 ,  1.9434184 ,  2.57696569,  3.41634856,\n",
      "          4.07210043, 5.56419101]),\n",
      "          0.25)\n",
      "    \n",
      "    ansari(x, y, alternative='two-sided')\n",
      "        Perform the Ansari-Bradley test for equal scale parameters.\n",
      "        \n",
      "        The Ansari-Bradley test ([1]_, [2]_) is a non-parametric test\n",
      "        for the equality of the scale parameter of the distributions\n",
      "        from which two samples were drawn. The null hypothesis states that\n",
      "        the ratio of the scale of the distribution underlying `x` to the scale\n",
      "        of the distribution underlying `y` is 1.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x, y : array_like\n",
      "            Arrays of sample data.\n",
      "        alternative : {'two-sided', 'less', 'greater'}, optional\n",
      "            Defines the alternative hypothesis. Default is 'two-sided'.\n",
      "            The following options are available:\n",
      "        \n",
      "            * 'two-sided': the ratio of scales is not equal to 1.\n",
      "            * 'less': the ratio of scales is less than 1.\n",
      "            * 'greater': the ratio of scales is greater than 1.\n",
      "        \n",
      "            .. versionadded:: 1.7.0\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        statistic : float\n",
      "            The Ansari-Bradley test statistic.\n",
      "        pvalue : float\n",
      "            The p-value of the hypothesis test.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        fligner : A non-parametric test for the equality of k variances\n",
      "        mood : A non-parametric test for the equality of two scale parameters\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        The p-value given is exact when the sample sizes are both less than\n",
      "        55 and there are no ties, otherwise a normal approximation for the\n",
      "        p-value is used.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] Ansari, A. R. and Bradley, R. A. (1960) Rank-sum tests for\n",
      "               dispersions, Annals of Mathematical Statistics, 31, 1174-1189.\n",
      "        .. [2] Sprent, Peter and N.C. Smeeton.  Applied nonparametric\n",
      "               statistical methods.  3rd ed. Chapman and Hall/CRC. 2001.\n",
      "               Section 5.8.2.\n",
      "        .. [3] Nathaniel E. Helwig \"Nonparametric Dispersion and Equality\n",
      "               Tests\" at http://users.stat.umn.edu/~helwig/notes/npde-Notes.pdf\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy.stats import ansari\n",
      "        >>> rng = np.random.default_rng()\n",
      "        \n",
      "        For these examples, we'll create three random data sets.  The first\n",
      "        two, with sizes 35 and 25, are drawn from a normal distribution with\n",
      "        mean 0 and standard deviation 2.  The third data set has size 25 and\n",
      "        is drawn from a normal distribution with standard deviation 1.25.\n",
      "        \n",
      "        >>> x1 = rng.normal(loc=0, scale=2, size=35)\n",
      "        >>> x2 = rng.normal(loc=0, scale=2, size=25)\n",
      "        >>> x3 = rng.normal(loc=0, scale=1.25, size=25)\n",
      "        \n",
      "        First we apply `ansari` to `x1` and `x2`.  These samples are drawn\n",
      "        from the same distribution, so we expect the Ansari-Bradley test\n",
      "        should not lead us to conclude that the scales of the distributions\n",
      "        are different.\n",
      "        \n",
      "        >>> ansari(x1, x2)\n",
      "        AnsariResult(statistic=541.0, pvalue=0.9762532927399098)\n",
      "        \n",
      "        With a p-value close to 1, we cannot conclude that there is a\n",
      "        significant difference in the scales (as expected).\n",
      "        \n",
      "        Now apply the test to `x1` and `x3`:\n",
      "        \n",
      "        >>> ansari(x1, x3)\n",
      "        AnsariResult(statistic=425.0, pvalue=0.0003087020407974518)\n",
      "        \n",
      "        The probability of observing such an extreme value of the statistic\n",
      "        under the null hypothesis of equal scales is only 0.03087%. We take this\n",
      "        as evidence against the null hypothesis in favor of the alternative:\n",
      "        the scales of the distributions from which the samples were drawn\n",
      "        are not equal.\n",
      "        \n",
      "        We can use the `alternative` parameter to perform a one-tailed test.\n",
      "        In the above example, the scale of `x1` is greater than `x3` and so\n",
      "        the ratio of scales of `x1` and `x3` is greater than 1. This means\n",
      "        that the p-value when ``alternative='greater'`` should be near 0 and\n",
      "        hence we should be able to reject the null hypothesis:\n",
      "        \n",
      "        >>> ansari(x1, x3, alternative='greater')\n",
      "        AnsariResult(statistic=425.0, pvalue=0.0001543510203987259)\n",
      "        \n",
      "        As we can see, the p-value is indeed quite low. Use of\n",
      "        ``alternative='less'`` should thus yield a large p-value:\n",
      "        \n",
      "        >>> ansari(x1, x3, alternative='less')\n",
      "        AnsariResult(statistic=425.0, pvalue=0.9998643258449039)\n",
      "    \n",
      "    barnard_exact(table, alternative='two-sided', pooled=True, n=32)\n",
      "        Perform a Barnard exact test on a 2x2 contingency table.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        table : array_like of ints\n",
      "            A 2x2 contingency table.  Elements should be non-negative integers.\n",
      "        \n",
      "        alternative : {'two-sided', 'less', 'greater'}, optional\n",
      "            Defines the null and alternative hypotheses. Default is 'two-sided'.\n",
      "            Please see explanations in the Notes section below.\n",
      "        \n",
      "        pooled : bool, optional\n",
      "            Whether to compute score statistic with pooled variance (as in\n",
      "            Student's t-test, for example) or unpooled variance (as in Welch's\n",
      "            t-test). Default is ``True``.\n",
      "        \n",
      "        n : int, optional\n",
      "            Number of sampling points used in the construction of the sampling\n",
      "            method. Note that this argument will automatically be converted to\n",
      "            the next higher power of 2 since `scipy.stats.qmc.Sobol` is used to\n",
      "            select sample points. Default is 32. Must be positive. In most cases,\n",
      "            32 points is enough to reach good precision. More points comes at\n",
      "            performance cost.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        ber : BarnardExactResult\n",
      "            A result object with the following attributes.\n",
      "        \n",
      "            statistic : float\n",
      "                The Wald statistic with pooled or unpooled variance, depending\n",
      "                on the user choice of `pooled`.\n",
      "        \n",
      "            pvalue : float\n",
      "                P-value, the probability of obtaining a distribution at least as\n",
      "                extreme as the one that was actually observed, assuming that the\n",
      "                null hypothesis is true.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        chi2_contingency : Chi-square test of independence of variables in a\n",
      "            contingency table.\n",
      "        fisher_exact : Fisher exact test on a 2x2 contingency table.\n",
      "        boschloo_exact : Boschloo's exact test on a 2x2 contingency table,\n",
      "            which is an uniformly more powerful alternative to Fisher's exact test.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        Barnard's test is an exact test used in the analysis of contingency\n",
      "        tables. It examines the association of two categorical variables, and\n",
      "        is a more powerful alternative than Fisher's exact test\n",
      "        for 2x2 contingency tables.\n",
      "        \n",
      "        Let's define :math:`X_0` a 2x2 matrix representing the observed sample,\n",
      "        where each column stores the binomial experiment, as in the example\n",
      "        below. Let's also define :math:`p_1, p_2` the theoretical binomial\n",
      "        probabilities for  :math:`x_{11}` and :math:`x_{12}`. When using\n",
      "        Barnard exact test, we can assert three different null hypotheses :\n",
      "        \n",
      "        - :math:`H_0 : p_1 \\geq p_2` versus :math:`H_1 : p_1 < p_2`,\n",
      "          with `alternative` = \"less\"\n",
      "        \n",
      "        - :math:`H_0 : p_1 \\leq p_2` versus :math:`H_1 : p_1 > p_2`,\n",
      "          with `alternative` = \"greater\"\n",
      "        \n",
      "        - :math:`H_0 : p_1 = p_2` versus :math:`H_1 : p_1 \\neq p_2`,\n",
      "          with `alternative` = \"two-sided\" (default one)\n",
      "        \n",
      "        In order to compute Barnard's exact test, we are using the Wald\n",
      "        statistic [3]_ with pooled or unpooled variance.\n",
      "        Under the default assumption that both variances are equal\n",
      "        (``pooled = True``), the statistic is computed as:\n",
      "        \n",
      "        .. math::\n",
      "        \n",
      "            T(X) = \\frac{\n",
      "                \\hat{p}_1 - \\hat{p}_2\n",
      "            }{\n",
      "                \\sqrt{\n",
      "                    \\hat{p}(1 - \\hat{p})\n",
      "                    (\\frac{1}{c_1} +\n",
      "                    \\frac{1}{c_2})\n",
      "                }\n",
      "            }\n",
      "        \n",
      "        with :math:`\\hat{p}_1, \\hat{p}_2` and :math:`\\hat{p}` the estimator of\n",
      "        :math:`p_1, p_2` and :math:`p`, the latter being the combined probability,\n",
      "        given the assumption that :math:`p_1 = p_2`.\n",
      "        \n",
      "        If this assumption is invalid (``pooled = False``), the statistic is:\n",
      "        \n",
      "        .. math::\n",
      "        \n",
      "            T(X) = \\frac{\n",
      "                \\hat{p}_1 - \\hat{p}_2\n",
      "            }{\n",
      "                \\sqrt{\n",
      "                    \\frac{\\hat{p}_1 (1 - \\hat{p}_1)}{c_1} +\n",
      "                    \\frac{\\hat{p}_2 (1 - \\hat{p}_2)}{c_2}\n",
      "                }\n",
      "            }\n",
      "        \n",
      "        The p-value is then computed as:\n",
      "        \n",
      "        .. math::\n",
      "        \n",
      "            \\sum\n",
      "                \\binom{c_1}{x_{11}}\n",
      "                \\binom{c_2}{x_{12}}\n",
      "                \\pi^{x_{11} + x_{12}}\n",
      "                (1 - \\pi)^{t - x_{11} - x_{12}}\n",
      "        \n",
      "        where the sum is over all  2x2 contingency tables :math:`X` such that:\n",
      "        * :math:`T(X) \\leq T(X_0)` when `alternative` = \"less\",\n",
      "        * :math:`T(X) \\geq T(X_0)` when `alternative` = \"greater\", or\n",
      "        * :math:`T(X) \\geq |T(X_0)|` when `alternative` = \"two-sided\".\n",
      "        Above, :math:`c_1, c_2` are the sum of the columns 1 and 2,\n",
      "        and :math:`t` the total (sum of the 4 sample's element).\n",
      "        \n",
      "        The returned p-value is the maximum p-value taken over the nuisance\n",
      "        parameter :math:`\\pi`, where :math:`0 \\leq \\pi \\leq 1`.\n",
      "        \n",
      "        This function's complexity is :math:`O(n c_1 c_2)`, where `n` is the\n",
      "        number of sample points.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] Barnard, G. A. \"Significance Tests for 2x2 Tables\". *Biometrika*.\n",
      "               34.1/2 (1947): 123-138. :doi:`dpgkg3`\n",
      "        \n",
      "        .. [2] Mehta, Cyrus R., and Pralay Senchaudhuri. \"Conditional versus\n",
      "               unconditional exact tests for comparing two binomials.\"\n",
      "               *Cytel Software Corporation* 675 (2003): 1-5.\n",
      "        \n",
      "        .. [3] \"Wald Test\". *Wikipedia*. https://en.wikipedia.org/wiki/Wald_test\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        An example use of Barnard's test is presented in [2]_.\n",
      "        \n",
      "            Consider the following example of a vaccine efficacy study\n",
      "            (Chan, 1998). In a randomized clinical trial of 30 subjects, 15 were\n",
      "            inoculated with a recombinant DNA influenza vaccine and the 15 were\n",
      "            inoculated with a placebo. Twelve of the 15 subjects in the placebo\n",
      "            group (80%) eventually became infected with influenza whereas for the\n",
      "            vaccine group, only 7 of the 15 subjects (47%) became infected. The\n",
      "            data are tabulated as a 2 x 2 table::\n",
      "        \n",
      "                    Vaccine  Placebo\n",
      "                Yes     7        12\n",
      "                No      8        3\n",
      "        \n",
      "        When working with statistical hypothesis testing, we usually use a\n",
      "        threshold probability or significance level upon which we decide\n",
      "        to reject the null hypothesis :math:`H_0`. Suppose we choose the common\n",
      "        significance level of 5%.\n",
      "        \n",
      "        Our alternative hypothesis is that the vaccine will lower the chance of\n",
      "        becoming infected with the virus; that is, the probability :math:`p_1` of\n",
      "        catching the virus with the vaccine will be *less than* the probability\n",
      "        :math:`p_2` of catching the virus without the vaccine.  Therefore, we call\n",
      "        `barnard_exact` with the ``alternative=\"less\"`` option:\n",
      "        \n",
      "        >>> import scipy.stats as stats\n",
      "        >>> res = stats.barnard_exact([[7, 12], [8, 3]], alternative=\"less\")\n",
      "        >>> res.statistic\n",
      "        -1.894...\n",
      "        >>> res.pvalue\n",
      "        0.03407...\n",
      "        \n",
      "        Under the null hypothesis that the vaccine will not lower the chance of\n",
      "        becoming infected, the probability of obtaining test results at least as\n",
      "        extreme as the observed data is approximately 3.4%. Since this p-value is\n",
      "        less than our chosen significance level, we have evidence to reject\n",
      "        :math:`H_0` in favor of the alternative.\n",
      "        \n",
      "        Suppose we had used Fisher's exact test instead:\n",
      "        \n",
      "        >>> _, pvalue = stats.fisher_exact([[7, 12], [8, 3]], alternative=\"less\")\n",
      "        >>> pvalue\n",
      "        0.0640...\n",
      "        \n",
      "        With the same threshold significance of 5%, we would not have been able\n",
      "        to reject the null hypothesis in favor of the alternative. As stated in\n",
      "        [2]_, Barnard's test is uniformly more powerful than Fisher's exact test\n",
      "        because Barnard's test does not condition on any margin. Fisher's test\n",
      "        should only be used when both sets of marginals are fixed.\n",
      "    \n",
      "    bartlett(*args)\n",
      "        Perform Bartlett's test for equal variances.\n",
      "        \n",
      "        Bartlett's test tests the null hypothesis that all input samples\n",
      "        are from populations with equal variances.  For samples\n",
      "        from significantly non-normal populations, Levene's test\n",
      "        `levene` is more robust.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        sample1, sample2,... : array_like\n",
      "            arrays of sample data.  Only 1d arrays are accepted, they may have\n",
      "            different lengths.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        statistic : float\n",
      "            The test statistic.\n",
      "        pvalue : float\n",
      "            The p-value of the test.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        fligner : A non-parametric test for the equality of k variances\n",
      "        levene : A robust parametric test for equality of k variances\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        Conover et al. (1981) examine many of the existing parametric and\n",
      "        nonparametric tests by extensive simulations and they conclude that the\n",
      "        tests proposed by Fligner and Killeen (1976) and Levene (1960) appear to be\n",
      "        superior in terms of robustness of departures from normality and power\n",
      "        ([3]_).\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1]  https://www.itl.nist.gov/div898/handbook/eda/section3/eda357.htm\n",
      "        \n",
      "        .. [2]  Snedecor, George W. and Cochran, William G. (1989), Statistical\n",
      "                  Methods, Eighth Edition, Iowa State University Press.\n",
      "        \n",
      "        .. [3] Park, C. and Lindsay, B. G. (1999). Robust Scale Estimation and\n",
      "               Hypothesis Testing based on Quadratic Inference Function. Technical\n",
      "               Report #99-03, Center for Likelihood Studies, Pennsylvania State\n",
      "               University.\n",
      "        \n",
      "        .. [4] Bartlett, M. S. (1937). Properties of Sufficiency and Statistical\n",
      "               Tests. Proceedings of the Royal Society of London. Series A,\n",
      "               Mathematical and Physical Sciences, Vol. 160, No.901, pp. 268-282.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        Test whether or not the lists `a`, `b` and `c` come from populations\n",
      "        with equal variances.\n",
      "        \n",
      "        >>> from scipy.stats import bartlett\n",
      "        >>> a = [8.88, 9.12, 9.04, 8.98, 9.00, 9.08, 9.01, 8.85, 9.06, 8.99]\n",
      "        >>> b = [8.88, 8.95, 9.29, 9.44, 9.15, 9.58, 8.36, 9.18, 8.67, 9.05]\n",
      "        >>> c = [8.95, 9.12, 8.95, 8.85, 9.03, 8.84, 9.07, 8.98, 8.86, 8.98]\n",
      "        >>> stat, p = bartlett(a, b, c)\n",
      "        >>> p\n",
      "        1.1254782518834628e-05\n",
      "        \n",
      "        The very small p-value suggests that the populations do not have equal\n",
      "        variances.\n",
      "        \n",
      "        This is not surprising, given that the sample variance of `b` is much\n",
      "        larger than that of `a` and `c`:\n",
      "        \n",
      "        >>> [np.var(x, ddof=1) for x in [a, b, c]]\n",
      "        [0.007054444444444413, 0.13073888888888888, 0.008890000000000002]\n",
      "    \n",
      "    bayes_mvs(data, alpha=0.9)\n",
      "        Bayesian confidence intervals for the mean, var, and std.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        data : array_like\n",
      "            Input data, if multi-dimensional it is flattened to 1-D by `bayes_mvs`.\n",
      "            Requires 2 or more data points.\n",
      "        alpha : float, optional\n",
      "            Probability that the returned confidence interval contains\n",
      "            the true parameter.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        mean_cntr, var_cntr, std_cntr : tuple\n",
      "            The three results are for the mean, variance and standard deviation,\n",
      "            respectively.  Each result is a tuple of the form::\n",
      "        \n",
      "                (center, (lower, upper))\n",
      "        \n",
      "            with `center` the mean of the conditional pdf of the value given the\n",
      "            data, and `(lower, upper)` a confidence interval, centered on the\n",
      "            median, containing the estimate to a probability ``alpha``.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        mvsdist\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        Each tuple of mean, variance, and standard deviation estimates represent\n",
      "        the (center, (lower, upper)) with center the mean of the conditional pdf\n",
      "        of the value given the data and (lower, upper) is a confidence interval\n",
      "        centered on the median, containing the estimate to a probability\n",
      "        ``alpha``.\n",
      "        \n",
      "        Converts data to 1-D and assumes all data has the same mean and variance.\n",
      "        Uses Jeffrey's prior for variance and std.\n",
      "        \n",
      "        Equivalent to ``tuple((x.mean(), x.interval(alpha)) for x in mvsdist(dat))``\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        T.E. Oliphant, \"A Bayesian perspective on estimating mean, variance, and\n",
      "        standard-deviation from data\", https://scholarsarchive.byu.edu/facpub/278,\n",
      "        2006.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        First a basic example to demonstrate the outputs:\n",
      "        \n",
      "        >>> from scipy import stats\n",
      "        >>> data = [6, 9, 12, 7, 8, 8, 13]\n",
      "        >>> mean, var, std = stats.bayes_mvs(data)\n",
      "        >>> mean\n",
      "        Mean(statistic=9.0, minmax=(7.103650222612533, 10.896349777387467))\n",
      "        >>> var\n",
      "        Variance(statistic=10.0, minmax=(3.176724206..., 24.45910382...))\n",
      "        >>> std\n",
      "        Std_dev(statistic=2.9724954732045084, minmax=(1.7823367265645143, 4.945614605014631))\n",
      "        \n",
      "        Now we generate some normally distributed random data, and get estimates of\n",
      "        mean and standard deviation with 95% confidence intervals for those\n",
      "        estimates:\n",
      "        \n",
      "        >>> n_samples = 100000\n",
      "        >>> data = stats.norm.rvs(size=n_samples)\n",
      "        >>> res_mean, res_var, res_std = stats.bayes_mvs(data, alpha=0.95)\n",
      "        \n",
      "        >>> import matplotlib.pyplot as plt\n",
      "        >>> fig = plt.figure()\n",
      "        >>> ax = fig.add_subplot(111)\n",
      "        >>> ax.hist(data, bins=100, density=True, label='Histogram of data')\n",
      "        >>> ax.vlines(res_mean.statistic, 0, 0.5, colors='r', label='Estimated mean')\n",
      "        >>> ax.axvspan(res_mean.minmax[0],res_mean.minmax[1], facecolor='r',\n",
      "        ...            alpha=0.2, label=r'Estimated mean (95% limits)')\n",
      "        >>> ax.vlines(res_std.statistic, 0, 0.5, colors='g', label='Estimated scale')\n",
      "        >>> ax.axvspan(res_std.minmax[0],res_std.minmax[1], facecolor='g', alpha=0.2,\n",
      "        ...            label=r'Estimated scale (95% limits)')\n",
      "        \n",
      "        >>> ax.legend(fontsize=10)\n",
      "        >>> ax.set_xlim([-4, 4])\n",
      "        >>> ax.set_ylim([0, 0.5])\n",
      "        >>> plt.show()\n",
      "    \n",
      "    binned_statistic(x, values, statistic='mean', bins=10, range=None)\n",
      "        Compute a binned statistic for one or more sets of data.\n",
      "        \n",
      "        This is a generalization of a histogram function.  A histogram divides\n",
      "        the space into bins, and returns the count of the number of points in\n",
      "        each bin.  This function allows the computation of the sum, mean, median,\n",
      "        or other statistic of the values (or set of values) within each bin.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x : (N,) array_like\n",
      "            A sequence of values to be binned.\n",
      "        values : (N,) array_like or list of (N,) array_like\n",
      "            The data on which the statistic will be computed.  This must be\n",
      "            the same shape as `x`, or a set of sequences - each the same shape as\n",
      "            `x`.  If `values` is a set of sequences, the statistic will be computed\n",
      "            on each independently.\n",
      "        statistic : string or callable, optional\n",
      "            The statistic to compute (default is 'mean').\n",
      "            The following statistics are available:\n",
      "        \n",
      "              * 'mean' : compute the mean of values for points within each bin.\n",
      "                Empty bins will be represented by NaN.\n",
      "              * 'std' : compute the standard deviation within each bin. This\n",
      "                is implicitly calculated with ddof=0.\n",
      "              * 'median' : compute the median of values for points within each\n",
      "                bin. Empty bins will be represented by NaN.\n",
      "              * 'count' : compute the count of points within each bin.  This is\n",
      "                identical to an unweighted histogram.  `values` array is not\n",
      "                referenced.\n",
      "              * 'sum' : compute the sum of values for points within each bin.\n",
      "                This is identical to a weighted histogram.\n",
      "              * 'min' : compute the minimum of values for points within each bin.\n",
      "                Empty bins will be represented by NaN.\n",
      "              * 'max' : compute the maximum of values for point within each bin.\n",
      "                Empty bins will be represented by NaN.\n",
      "              * function : a user-defined function which takes a 1D array of\n",
      "                values, and outputs a single numerical statistic. This function\n",
      "                will be called on the values in each bin.  Empty bins will be\n",
      "                represented by function([]), or NaN if this returns an error.\n",
      "        \n",
      "        bins : int or sequence of scalars, optional\n",
      "            If `bins` is an int, it defines the number of equal-width bins in the\n",
      "            given range (10 by default).  If `bins` is a sequence, it defines the\n",
      "            bin edges, including the rightmost edge, allowing for non-uniform bin\n",
      "            widths.  Values in `x` that are smaller than lowest bin edge are\n",
      "            assigned to bin number 0, values beyond the highest bin are assigned to\n",
      "            ``bins[-1]``.  If the bin edges are specified, the number of bins will\n",
      "            be, (nx = len(bins)-1).\n",
      "        range : (float, float) or [(float, float)], optional\n",
      "            The lower and upper range of the bins.  If not provided, range\n",
      "            is simply ``(x.min(), x.max())``.  Values outside the range are\n",
      "            ignored.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        statistic : array\n",
      "            The values of the selected statistic in each bin.\n",
      "        bin_edges : array of dtype float\n",
      "            Return the bin edges ``(length(statistic)+1)``.\n",
      "        binnumber: 1-D ndarray of ints\n",
      "            Indices of the bins (corresponding to `bin_edges`) in which each value\n",
      "            of `x` belongs.  Same length as `values`.  A binnumber of `i` means the\n",
      "            corresponding value is between (bin_edges[i-1], bin_edges[i]).\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        numpy.digitize, numpy.histogram, binned_statistic_2d, binned_statistic_dd\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        All but the last (righthand-most) bin is half-open.  In other words, if\n",
      "        `bins` is ``[1, 2, 3, 4]``, then the first bin is ``[1, 2)`` (including 1,\n",
      "        but excluding 2) and the second ``[2, 3)``.  The last bin, however, is\n",
      "        ``[3, 4]``, which *includes* 4.\n",
      "        \n",
      "        .. versionadded:: 0.11.0\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy import stats\n",
      "        >>> import matplotlib.pyplot as plt\n",
      "        \n",
      "        First some basic examples:\n",
      "        \n",
      "        Create two evenly spaced bins in the range of the given sample, and sum the\n",
      "        corresponding values in each of those bins:\n",
      "        \n",
      "        >>> values = [1.0, 1.0, 2.0, 1.5, 3.0]\n",
      "        >>> stats.binned_statistic([1, 1, 2, 5, 7], values, 'sum', bins=2)\n",
      "        BinnedStatisticResult(statistic=array([4. , 4.5]),\n",
      "                bin_edges=array([1., 4., 7.]), binnumber=array([1, 1, 1, 2, 2]))\n",
      "        \n",
      "        Multiple arrays of values can also be passed.  The statistic is calculated\n",
      "        on each set independently:\n",
      "        \n",
      "        >>> values = [[1.0, 1.0, 2.0, 1.5, 3.0], [2.0, 2.0, 4.0, 3.0, 6.0]]\n",
      "        >>> stats.binned_statistic([1, 1, 2, 5, 7], values, 'sum', bins=2)\n",
      "        BinnedStatisticResult(statistic=array([[4. , 4.5],\n",
      "               [8. , 9. ]]), bin_edges=array([1., 4., 7.]),\n",
      "               binnumber=array([1, 1, 1, 2, 2]))\n",
      "        \n",
      "        >>> stats.binned_statistic([1, 2, 1, 2, 4], np.arange(5), statistic='mean',\n",
      "        ...                        bins=3)\n",
      "        BinnedStatisticResult(statistic=array([1., 2., 4.]),\n",
      "                bin_edges=array([1., 2., 3., 4.]),\n",
      "                binnumber=array([1, 2, 1, 2, 3]))\n",
      "        \n",
      "        As a second example, we now generate some random data of sailing boat speed\n",
      "        as a function of wind speed, and then determine how fast our boat is for\n",
      "        certain wind speeds:\n",
      "        \n",
      "        >>> rng = np.random.default_rng()\n",
      "        >>> windspeed = 8 * rng.random(500)\n",
      "        >>> boatspeed = .3 * windspeed**.5 + .2 * rng.random(500)\n",
      "        >>> bin_means, bin_edges, binnumber = stats.binned_statistic(windspeed,\n",
      "        ...                 boatspeed, statistic='median', bins=[1,2,3,4,5,6,7])\n",
      "        >>> plt.figure()\n",
      "        >>> plt.plot(windspeed, boatspeed, 'b.', label='raw data')\n",
      "        >>> plt.hlines(bin_means, bin_edges[:-1], bin_edges[1:], colors='g', lw=5,\n",
      "        ...            label='binned statistic of data')\n",
      "        >>> plt.legend()\n",
      "        \n",
      "        Now we can use ``binnumber`` to select all datapoints with a windspeed\n",
      "        below 1:\n",
      "        \n",
      "        >>> low_boatspeed = boatspeed[binnumber == 0]\n",
      "        \n",
      "        As a final example, we will use ``bin_edges`` and ``binnumber`` to make a\n",
      "        plot of a distribution that shows the mean and distribution around that\n",
      "        mean per bin, on top of a regular histogram and the probability\n",
      "        distribution function:\n",
      "        \n",
      "        >>> x = np.linspace(0, 5, num=500)\n",
      "        >>> x_pdf = stats.maxwell.pdf(x)\n",
      "        >>> samples = stats.maxwell.rvs(size=10000)\n",
      "        \n",
      "        >>> bin_means, bin_edges, binnumber = stats.binned_statistic(x, x_pdf,\n",
      "        ...         statistic='mean', bins=25)\n",
      "        >>> bin_width = (bin_edges[1] - bin_edges[0])\n",
      "        >>> bin_centers = bin_edges[1:] - bin_width/2\n",
      "        \n",
      "        >>> plt.figure()\n",
      "        >>> plt.hist(samples, bins=50, density=True, histtype='stepfilled',\n",
      "        ...          alpha=0.2, label='histogram of data')\n",
      "        >>> plt.plot(x, x_pdf, 'r-', label='analytical pdf')\n",
      "        >>> plt.hlines(bin_means, bin_edges[:-1], bin_edges[1:], colors='g', lw=2,\n",
      "        ...            label='binned statistic of data')\n",
      "        >>> plt.plot((binnumber - 0.5) * bin_width, x_pdf, 'g.', alpha=0.5)\n",
      "        >>> plt.legend(fontsize=10)\n",
      "        >>> plt.show()\n",
      "    \n",
      "    binned_statistic_2d(x, y, values, statistic='mean', bins=10, range=None, expand_binnumbers=False)\n",
      "        Compute a bidimensional binned statistic for one or more sets of data.\n",
      "        \n",
      "        This is a generalization of a histogram2d function.  A histogram divides\n",
      "        the space into bins, and returns the count of the number of points in\n",
      "        each bin.  This function allows the computation of the sum, mean, median,\n",
      "        or other statistic of the values (or set of values) within each bin.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x : (N,) array_like\n",
      "            A sequence of values to be binned along the first dimension.\n",
      "        y : (N,) array_like\n",
      "            A sequence of values to be binned along the second dimension.\n",
      "        values : (N,) array_like or list of (N,) array_like\n",
      "            The data on which the statistic will be computed.  This must be\n",
      "            the same shape as `x`, or a list of sequences - each with the same\n",
      "            shape as `x`.  If `values` is such a list, the statistic will be\n",
      "            computed on each independently.\n",
      "        statistic : string or callable, optional\n",
      "            The statistic to compute (default is 'mean').\n",
      "            The following statistics are available:\n",
      "        \n",
      "              * 'mean' : compute the mean of values for points within each bin.\n",
      "                Empty bins will be represented by NaN.\n",
      "              * 'std' : compute the standard deviation within each bin. This\n",
      "                is implicitly calculated with ddof=0.\n",
      "              * 'median' : compute the median of values for points within each\n",
      "                bin. Empty bins will be represented by NaN.\n",
      "              * 'count' : compute the count of points within each bin.  This is\n",
      "                identical to an unweighted histogram.  `values` array is not\n",
      "                referenced.\n",
      "              * 'sum' : compute the sum of values for points within each bin.\n",
      "                This is identical to a weighted histogram.\n",
      "              * 'min' : compute the minimum of values for points within each bin.\n",
      "                Empty bins will be represented by NaN.\n",
      "              * 'max' : compute the maximum of values for point within each bin.\n",
      "                Empty bins will be represented by NaN.\n",
      "              * function : a user-defined function which takes a 1D array of\n",
      "                values, and outputs a single numerical statistic. This function\n",
      "                will be called on the values in each bin.  Empty bins will be\n",
      "                represented by function([]), or NaN if this returns an error.\n",
      "        \n",
      "        bins : int or [int, int] or array_like or [array, array], optional\n",
      "            The bin specification:\n",
      "        \n",
      "              * the number of bins for the two dimensions (nx = ny = bins),\n",
      "              * the number of bins in each dimension (nx, ny = bins),\n",
      "              * the bin edges for the two dimensions (x_edge = y_edge = bins),\n",
      "              * the bin edges in each dimension (x_edge, y_edge = bins).\n",
      "        \n",
      "            If the bin edges are specified, the number of bins will be,\n",
      "            (nx = len(x_edge)-1, ny = len(y_edge)-1).\n",
      "        \n",
      "        range : (2,2) array_like, optional\n",
      "            The leftmost and rightmost edges of the bins along each dimension\n",
      "            (if not specified explicitly in the `bins` parameters):\n",
      "            [[xmin, xmax], [ymin, ymax]]. All values outside of this range will be\n",
      "            considered outliers and not tallied in the histogram.\n",
      "        expand_binnumbers : bool, optional\n",
      "            'False' (default): the returned `binnumber` is a shape (N,) array of\n",
      "            linearized bin indices.\n",
      "            'True': the returned `binnumber` is 'unraveled' into a shape (2,N)\n",
      "            ndarray, where each row gives the bin numbers in the corresponding\n",
      "            dimension.\n",
      "            See the `binnumber` returned value, and the `Examples` section.\n",
      "        \n",
      "            .. versionadded:: 0.17.0\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        statistic : (nx, ny) ndarray\n",
      "            The values of the selected statistic in each two-dimensional bin.\n",
      "        x_edge : (nx + 1) ndarray\n",
      "            The bin edges along the first dimension.\n",
      "        y_edge : (ny + 1) ndarray\n",
      "            The bin edges along the second dimension.\n",
      "        binnumber : (N,) array of ints or (2,N) ndarray of ints\n",
      "            This assigns to each element of `sample` an integer that represents the\n",
      "            bin in which this observation falls.  The representation depends on the\n",
      "            `expand_binnumbers` argument.  See `Notes` for details.\n",
      "        \n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        numpy.digitize, numpy.histogram2d, binned_statistic, binned_statistic_dd\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        Binedges:\n",
      "        All but the last (righthand-most) bin is half-open.  In other words, if\n",
      "        `bins` is ``[1, 2, 3, 4]``, then the first bin is ``[1, 2)`` (including 1,\n",
      "        but excluding 2) and the second ``[2, 3)``.  The last bin, however, is\n",
      "        ``[3, 4]``, which *includes* 4.\n",
      "        \n",
      "        `binnumber`:\n",
      "        This returned argument assigns to each element of `sample` an integer that\n",
      "        represents the bin in which it belongs.  The representation depends on the\n",
      "        `expand_binnumbers` argument. If 'False' (default): The returned\n",
      "        `binnumber` is a shape (N,) array of linearized indices mapping each\n",
      "        element of `sample` to its corresponding bin (using row-major ordering).\n",
      "        If 'True': The returned `binnumber` is a shape (2,N) ndarray where\n",
      "        each row indicates bin placements for each dimension respectively.  In each\n",
      "        dimension, a binnumber of `i` means the corresponding value is between\n",
      "        (D_edge[i-1], D_edge[i]), where 'D' is either 'x' or 'y'.\n",
      "        \n",
      "        .. versionadded:: 0.11.0\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy import stats\n",
      "        \n",
      "        Calculate the counts with explicit bin-edges:\n",
      "        \n",
      "        >>> x = [0.1, 0.1, 0.1, 0.6]\n",
      "        >>> y = [2.1, 2.6, 2.1, 2.1]\n",
      "        >>> binx = [0.0, 0.5, 1.0]\n",
      "        >>> biny = [2.0, 2.5, 3.0]\n",
      "        >>> ret = stats.binned_statistic_2d(x, y, None, 'count', bins=[binx, biny])\n",
      "        >>> ret.statistic\n",
      "        array([[2., 1.],\n",
      "               [1., 0.]])\n",
      "        \n",
      "        The bin in which each sample is placed is given by the `binnumber`\n",
      "        returned parameter.  By default, these are the linearized bin indices:\n",
      "        \n",
      "        >>> ret.binnumber\n",
      "        array([5, 6, 5, 9])\n",
      "        \n",
      "        The bin indices can also be expanded into separate entries for each\n",
      "        dimension using the `expand_binnumbers` parameter:\n",
      "        \n",
      "        >>> ret = stats.binned_statistic_2d(x, y, None, 'count', bins=[binx, biny],\n",
      "        ...                                 expand_binnumbers=True)\n",
      "        >>> ret.binnumber\n",
      "        array([[1, 1, 1, 2],\n",
      "               [1, 2, 1, 1]])\n",
      "        \n",
      "        Which shows that the first three elements belong in the xbin 1, and the\n",
      "        fourth into xbin 2; and so on for y.\n",
      "    \n",
      "    binned_statistic_dd(sample, values, statistic='mean', bins=10, range=None, expand_binnumbers=False, binned_statistic_result=None)\n",
      "        Compute a multidimensional binned statistic for a set of data.\n",
      "        \n",
      "        This is a generalization of a histogramdd function.  A histogram divides\n",
      "        the space into bins, and returns the count of the number of points in\n",
      "        each bin.  This function allows the computation of the sum, mean, median,\n",
      "        or other statistic of the values within each bin.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        sample : array_like\n",
      "            Data to histogram passed as a sequence of N arrays of length D, or\n",
      "            as an (N,D) array.\n",
      "        values : (N,) array_like or list of (N,) array_like\n",
      "            The data on which the statistic will be computed.  This must be\n",
      "            the same shape as `sample`, or a list of sequences - each with the\n",
      "            same shape as `sample`.  If `values` is such a list, the statistic\n",
      "            will be computed on each independently.\n",
      "        statistic : string or callable, optional\n",
      "            The statistic to compute (default is 'mean').\n",
      "            The following statistics are available:\n",
      "        \n",
      "              * 'mean' : compute the mean of values for points within each bin.\n",
      "                Empty bins will be represented by NaN.\n",
      "              * 'median' : compute the median of values for points within each\n",
      "                bin. Empty bins will be represented by NaN.\n",
      "              * 'count' : compute the count of points within each bin.  This is\n",
      "                identical to an unweighted histogram.  `values` array is not\n",
      "                referenced.\n",
      "              * 'sum' : compute the sum of values for points within each bin.\n",
      "                This is identical to a weighted histogram.\n",
      "              * 'std' : compute the standard deviation within each bin. This\n",
      "                is implicitly calculated with ddof=0. If the number of values\n",
      "                within a given bin is 0 or 1, the computed standard deviation value\n",
      "                will be 0 for the bin.\n",
      "              * 'min' : compute the minimum of values for points within each bin.\n",
      "                Empty bins will be represented by NaN.\n",
      "              * 'max' : compute the maximum of values for point within each bin.\n",
      "                Empty bins will be represented by NaN.\n",
      "              * function : a user-defined function which takes a 1D array of\n",
      "                values, and outputs a single numerical statistic. This function\n",
      "                will be called on the values in each bin.  Empty bins will be\n",
      "                represented by function([]), or NaN if this returns an error.\n",
      "        \n",
      "        bins : sequence or positive int, optional\n",
      "            The bin specification must be in one of the following forms:\n",
      "        \n",
      "              * A sequence of arrays describing the bin edges along each dimension.\n",
      "              * The number of bins for each dimension (nx, ny, ... = bins).\n",
      "              * The number of bins for all dimensions (nx = ny = ... = bins).\n",
      "        range : sequence, optional\n",
      "            A sequence of lower and upper bin edges to be used if the edges are\n",
      "            not given explicitly in `bins`. Defaults to the minimum and maximum\n",
      "            values along each dimension.\n",
      "        expand_binnumbers : bool, optional\n",
      "            'False' (default): the returned `binnumber` is a shape (N,) array of\n",
      "            linearized bin indices.\n",
      "            'True': the returned `binnumber` is 'unraveled' into a shape (D,N)\n",
      "            ndarray, where each row gives the bin numbers in the corresponding\n",
      "            dimension.\n",
      "            See the `binnumber` returned value, and the `Examples` section of\n",
      "            `binned_statistic_2d`.\n",
      "        binned_statistic_result : binnedStatisticddResult\n",
      "            Result of a previous call to the function in order to reuse bin edges\n",
      "            and bin numbers with new values and/or a different statistic.\n",
      "            To reuse bin numbers, `expand_binnumbers` must have been set to False\n",
      "            (the default)\n",
      "        \n",
      "            .. versionadded:: 0.17.0\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        statistic : ndarray, shape(nx1, nx2, nx3,...)\n",
      "            The values of the selected statistic in each two-dimensional bin.\n",
      "        bin_edges : list of ndarrays\n",
      "            A list of D arrays describing the (nxi + 1) bin edges for each\n",
      "            dimension.\n",
      "        binnumber : (N,) array of ints or (D,N) ndarray of ints\n",
      "            This assigns to each element of `sample` an integer that represents the\n",
      "            bin in which this observation falls.  The representation depends on the\n",
      "            `expand_binnumbers` argument.  See `Notes` for details.\n",
      "        \n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        numpy.digitize, numpy.histogramdd, binned_statistic, binned_statistic_2d\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        Binedges:\n",
      "        All but the last (righthand-most) bin is half-open in each dimension.  In\n",
      "        other words, if `bins` is ``[1, 2, 3, 4]``, then the first bin is\n",
      "        ``[1, 2)`` (including 1, but excluding 2) and the second ``[2, 3)``.  The\n",
      "        last bin, however, is ``[3, 4]``, which *includes* 4.\n",
      "        \n",
      "        `binnumber`:\n",
      "        This returned argument assigns to each element of `sample` an integer that\n",
      "        represents the bin in which it belongs.  The representation depends on the\n",
      "        `expand_binnumbers` argument. If 'False' (default): The returned\n",
      "        `binnumber` is a shape (N,) array of linearized indices mapping each\n",
      "        element of `sample` to its corresponding bin (using row-major ordering).\n",
      "        If 'True': The returned `binnumber` is a shape (D,N) ndarray where\n",
      "        each row indicates bin placements for each dimension respectively.  In each\n",
      "        dimension, a binnumber of `i` means the corresponding value is between\n",
      "        (bin_edges[D][i-1], bin_edges[D][i]), for each dimension 'D'.\n",
      "        \n",
      "        .. versionadded:: 0.11.0\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy import stats\n",
      "        >>> import matplotlib.pyplot as plt\n",
      "        >>> from mpl_toolkits.mplot3d import Axes3D\n",
      "        \n",
      "        Take an array of 600 (x, y) coordinates as an example.\n",
      "        `binned_statistic_dd` can handle arrays of higher dimension `D`. But a plot\n",
      "        of dimension `D+1` is required.\n",
      "        \n",
      "        >>> mu = np.array([0., 1.])\n",
      "        >>> sigma = np.array([[1., -0.5],[-0.5, 1.5]])\n",
      "        >>> multinormal = stats.multivariate_normal(mu, sigma)\n",
      "        >>> data = multinormal.rvs(size=600, random_state=235412)\n",
      "        >>> data.shape\n",
      "        (600, 2)\n",
      "        \n",
      "        Create bins and count how many arrays fall in each bin:\n",
      "        \n",
      "        >>> N = 60\n",
      "        >>> x = np.linspace(-3, 3, N)\n",
      "        >>> y = np.linspace(-3, 4, N)\n",
      "        >>> ret = stats.binned_statistic_dd(data, np.arange(600), bins=[x, y],\n",
      "        ...                                 statistic='count')\n",
      "        >>> bincounts = ret.statistic\n",
      "        \n",
      "        Set the volume and the location of bars:\n",
      "        \n",
      "        >>> dx = x[1] - x[0]\n",
      "        >>> dy = y[1] - y[0]\n",
      "        >>> x, y = np.meshgrid(x[:-1]+dx/2, y[:-1]+dy/2)\n",
      "        >>> z = 0\n",
      "        \n",
      "        >>> bincounts = bincounts.ravel()\n",
      "        >>> x = x.ravel()\n",
      "        >>> y = y.ravel()\n",
      "        \n",
      "        >>> fig = plt.figure()\n",
      "        >>> ax = fig.add_subplot(111, projection='3d')\n",
      "        >>> with np.errstate(divide='ignore'):   # silence random axes3d warning\n",
      "        ...     ax.bar3d(x, y, z, dx, dy, bincounts)\n",
      "        \n",
      "        Reuse bin numbers and bin edges with new values:\n",
      "        \n",
      "        >>> ret2 = stats.binned_statistic_dd(data, -np.arange(600),\n",
      "        ...                                  binned_statistic_result=ret,\n",
      "        ...                                  statistic='mean')\n",
      "    \n",
      "    binom_test(x, n=None, p=0.5, alternative='two-sided')\n",
      "        Perform a test that the probability of success is p.\n",
      "        \n",
      "        Note: `binom_test` is deprecated; it is recommended that `binomtest`\n",
      "        be used instead.\n",
      "        \n",
      "        This is an exact, two-sided test of the null hypothesis\n",
      "        that the probability of success in a Bernoulli experiment\n",
      "        is `p`.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x : int or array_like\n",
      "            The number of successes, or if x has length 2, it is the\n",
      "            number of successes and the number of failures.\n",
      "        n : int\n",
      "            The number of trials.  This is ignored if x gives both the\n",
      "            number of successes and failures.\n",
      "        p : float, optional\n",
      "            The hypothesized probability of success.  ``0 <= p <= 1``. The\n",
      "            default value is ``p = 0.5``.\n",
      "        alternative : {'two-sided', 'greater', 'less'}, optional\n",
      "            Indicates the alternative hypothesis. The default value is\n",
      "            'two-sided'.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        p-value : float\n",
      "            The p-value of the hypothesis test.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] https://en.wikipedia.org/wiki/Binomial_test\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy import stats\n",
      "        \n",
      "        A car manufacturer claims that no more than 10% of their cars are unsafe.\n",
      "        15 cars are inspected for safety, 3 were found to be unsafe. Test the\n",
      "        manufacturer's claim:\n",
      "        \n",
      "        >>> stats.binom_test(3, n=15, p=0.1, alternative='greater')\n",
      "        0.18406106910639114\n",
      "        \n",
      "        The null hypothesis cannot be rejected at the 5% level of significance\n",
      "        because the returned p-value is greater than the critical value of 5%.\n",
      "    \n",
      "    binomtest(k, n, p=0.5, alternative='two-sided')\n",
      "        Perform a test that the probability of success is p.\n",
      "        \n",
      "        The binomial test [1]_ is a test of the null hypothesis that the\n",
      "        probability of success in a Bernoulli experiment is `p`.\n",
      "        \n",
      "        Details of the test can be found in many texts on statistics, such\n",
      "        as section 24.5 of [2]_.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        k : int\n",
      "            The number of successes.\n",
      "        n : int\n",
      "            The number of trials.\n",
      "        p : float, optional\n",
      "            The hypothesized probability of success, i.e. the expected\n",
      "            proportion of successes.  The value must be in the interval\n",
      "            ``0 <= p <= 1``. The default value is ``p = 0.5``.\n",
      "        alternative : {'two-sided', 'greater', 'less'}, optional\n",
      "            Indicates the alternative hypothesis. The default value is\n",
      "            'two-sided'.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        result : `~scipy.stats._result_classes.BinomTestResult` instance\n",
      "            The return value is an object with the following attributes:\n",
      "        \n",
      "            k : int\n",
      "                The number of successes (copied from `binomtest` input).\n",
      "            n : int\n",
      "                The number of trials (copied from `binomtest` input).\n",
      "            alternative : str\n",
      "                Indicates the alternative hypothesis specified in the input\n",
      "                to `binomtest`.  It will be one of ``'two-sided'``, ``'greater'``,\n",
      "                or ``'less'``.\n",
      "            pvalue : float\n",
      "                The p-value of the hypothesis test.\n",
      "            proportion_estimate : float\n",
      "                The estimate of the proportion of successes.\n",
      "        \n",
      "            The object has the following methods:\n",
      "        \n",
      "            proportion_ci(confidence_level=0.95, method='exact') :\n",
      "                Compute the confidence interval for ``proportion_estimate``.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        .. versionadded:: 1.7.0\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] Binomial test, https://en.wikipedia.org/wiki/Binomial_test\n",
      "        .. [2] Jerrold H. Zar, Biostatistical Analysis (fifth edition),\n",
      "               Prentice Hall, Upper Saddle River, New Jersey USA (2010)\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy.stats import binomtest\n",
      "        \n",
      "        A car manufacturer claims that no more than 10% of their cars are unsafe.\n",
      "        15 cars are inspected for safety, 3 were found to be unsafe. Test the\n",
      "        manufacturer's claim:\n",
      "        \n",
      "        >>> result = binomtest(3, n=15, p=0.1, alternative='greater')\n",
      "        >>> result.pvalue\n",
      "        0.18406106910639114\n",
      "        \n",
      "        The null hypothesis cannot be rejected at the 5% level of significance\n",
      "        because the returned p-value is greater than the critical value of 5%.\n",
      "        \n",
      "        The estimated proportion is simply ``3/15``:\n",
      "        \n",
      "        >>> result.proportion_estimate\n",
      "        0.2\n",
      "        \n",
      "        We can use the `proportion_ci()` method of the result to compute the\n",
      "        confidence interval of the estimate:\n",
      "        \n",
      "        >>> result.proportion_ci(confidence_level=0.95)\n",
      "        ConfidenceInterval(low=0.05684686759024681, high=1.0)\n",
      "    \n",
      "    bootstrap(data, statistic, *, vectorized=True, paired=False, axis=0, confidence_level=0.95, n_resamples=9999, batch=None, method='BCa', random_state=None)\n",
      "        Compute a two-sided bootstrap confidence interval of a statistic.\n",
      "        \n",
      "        When `method` is ``'percentile'``, a bootstrap confidence interval is\n",
      "        computed according to the following procedure.\n",
      "        \n",
      "        1. Resample the data: for each sample in `data` and for each of\n",
      "           `n_resamples`, take a random sample of the original sample\n",
      "           (with replacement) of the same size as the original sample.\n",
      "        \n",
      "        2. Compute the bootstrap distribution of the statistic: for each set of\n",
      "           resamples, compute the test statistic.\n",
      "        \n",
      "        3. Determine the confidence interval: find the interval of the bootstrap\n",
      "           distribution that is\n",
      "        \n",
      "           - symmetric about the median and\n",
      "           - contains `confidence_level` of the resampled statistic values.\n",
      "        \n",
      "        While the ``'percentile'`` method is the most intuitive, it is rarely\n",
      "        used in practice. Two more common methods are available, ``'basic'``\n",
      "        ('reverse percentile') and ``'BCa'`` ('bias-corrected and accelerated');\n",
      "        they differ in how step 3 is performed.\n",
      "        \n",
      "        If the samples in `data` are  taken at random from their respective\n",
      "        distributions :math:`n` times, the confidence interval returned by\n",
      "        `bootstrap` will contain the true value of the statistic for those\n",
      "        distributions approximately `confidence_level`:math:`\\, \\times \\, n` times.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        data : sequence of array-like\n",
      "             Each element of data is a sample from an underlying distribution.\n",
      "        statistic : callable\n",
      "            Statistic for which the confidence interval is to be calculated.\n",
      "            `statistic` must be a callable that accepts ``len(data)`` samples\n",
      "            as separate arguments and returns the resulting statistic.\n",
      "            If `vectorized` is set ``True``,\n",
      "            `statistic` must also accept a keyword argument `axis` and be\n",
      "            vectorized to compute the statistic along the provided `axis`.\n",
      "        vectorized : bool, default: ``True``\n",
      "            If `vectorized` is set ``False``, `statistic` will not be passed\n",
      "            keyword argument `axis`, and is assumed to calculate the statistic\n",
      "            only for 1D samples.\n",
      "        paired : bool, default: ``False``\n",
      "            Whether the statistic treats corresponding elements of the samples\n",
      "            in `data` as paired.\n",
      "        axis : int, default: ``0``\n",
      "            The axis of the samples in `data` along which the `statistic` is\n",
      "            calculated.\n",
      "        confidence_level : float, default: ``0.95``\n",
      "            The confidence level of the confidence interval.\n",
      "        n_resamples : int, default: ``9999``\n",
      "            The number of resamples performed to form the bootstrap distribution\n",
      "            of the statistic.\n",
      "        batch : int, optional\n",
      "            The number of resamples to process in each vectorized call to\n",
      "            `statistic`. Memory usage is O(`batch`*``n``), where ``n`` is the\n",
      "            sample size. Default is ``None``, in which case ``batch = n_resamples``\n",
      "            (or ``batch = max(n_resamples, n)`` for ``method='BCa'``).\n",
      "        method : {'percentile', 'basic', 'bca'}, default: ``'BCa'``\n",
      "            Whether to return the 'percentile' bootstrap confidence interval\n",
      "            (``'percentile'``), the 'reverse' or the bias-corrected and accelerated\n",
      "            bootstrap confidence interval (``'BCa'``).\n",
      "            Note that only ``'percentile'`` and ``'basic'`` support multi-sample\n",
      "            statistics at this time.\n",
      "        random_state : {None, int, `numpy.random.Generator`,\n",
      "                        `numpy.random.RandomState`}, optional\n",
      "        \n",
      "            If `seed` is ``None`` (or `np.random`), the `numpy.random.RandomState`\n",
      "            singleton is used.\n",
      "            If `seed` is an int, a new ``RandomState`` instance is used,\n",
      "            seeded with `seed`.\n",
      "            If `seed` is already a ``Generator`` or ``RandomState`` instance then\n",
      "            that instance is used.\n",
      "        \n",
      "            Pseudorandom number generator state used to generate resamples.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        res : BootstrapResult\n",
      "            An object with attributes:\n",
      "        \n",
      "            confidence_interval : ConfidenceInterval\n",
      "                The bootstrap confidence interval as an instance of\n",
      "                `collections.namedtuple` with attributes `low` and `high`.\n",
      "            standard_error : float or ndarray\n",
      "                The bootstrap standard error, that is, the sample standard\n",
      "                deviation of the bootstrap distribution\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] B. Efron and R. J. Tibshirani, An Introduction to the Bootstrap,\n",
      "           Chapman & Hall/CRC, Boca Raton, FL, USA (1993)\n",
      "        .. [2] Nathaniel E. Helwig, \"Bootstrap Confidence Intervals\",\n",
      "           http://users.stat.umn.edu/~helwig/notes/bootci-Notes.pdf\n",
      "        .. [3] Bootstrapping (statistics), Wikipedia,\n",
      "           https://en.wikipedia.org/wiki/Bootstrapping_(statistics)\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        Suppose we have sampled data from an unknown distribution.\n",
      "        \n",
      "        >>> import numpy as np\n",
      "        >>> rng = np.random.default_rng()\n",
      "        >>> from scipy.stats import norm\n",
      "        >>> dist = norm(loc=2, scale=4)  # our \"unknown\" distribution\n",
      "        >>> data = dist.rvs(size=100, random_state=rng)\n",
      "        \n",
      "        We are interested int the standard deviation of the distribution.\n",
      "        \n",
      "        >>> std_true = dist.std()      # the true value of the statistic\n",
      "        >>> print(std_true)\n",
      "        4.0\n",
      "        >>> std_sample = np.std(data)  # the sample statistic\n",
      "        >>> print(std_sample)\n",
      "        3.9460644295563863\n",
      "        \n",
      "        We can calculate a 90% confidence interval of the statistic using\n",
      "        `bootstrap`.\n",
      "        \n",
      "        >>> from scipy.stats import bootstrap\n",
      "        >>> data = (data,)  # samples must be in a sequence\n",
      "        >>> res = bootstrap(data, np.std, confidence_level=0.9,\n",
      "        ...                 random_state=rng)\n",
      "        >>> print(res.confidence_interval)\n",
      "        ConfidenceInterval(low=3.57655333533867, high=4.382043696342881)\n",
      "        \n",
      "        If we sample from the distribution 1000 times and form a bootstrap\n",
      "        confidence interval for each sample, the confidence interval\n",
      "        contains the true value of the statistic approximately 900 times.\n",
      "        \n",
      "        >>> n_trials = 1000\n",
      "        >>> ci_contains_true_std = 0\n",
      "        >>> for i in range(n_trials):\n",
      "        ...    data = (dist.rvs(size=100, random_state=rng),)\n",
      "        ...    ci = bootstrap(data, np.std, confidence_level=0.9, n_resamples=1000,\n",
      "        ...                   random_state=rng).confidence_interval\n",
      "        ...    if ci[0] < std_true < ci[1]:\n",
      "        ...        ci_contains_true_std += 1\n",
      "        >>> print(ci_contains_true_std)\n",
      "        875\n",
      "        \n",
      "        Rather than writing a loop, we can also determine the confidence intervals\n",
      "        for all 1000 samples at once.\n",
      "        \n",
      "        >>> data = (dist.rvs(size=(n_trials, 100), random_state=rng),)\n",
      "        >>> res = bootstrap(data, np.std, axis=-1, confidence_level=0.9,\n",
      "        ...                 n_resamples=1000, random_state=rng)\n",
      "        >>> ci_l, ci_u = res.confidence_interval\n",
      "        \n",
      "        Here, `ci_l` and `ci_u` contain the confidence interval for each of the\n",
      "        ``n_trials = 1000`` samples.\n",
      "        \n",
      "        >>> print(ci_l[995:])\n",
      "        [3.77729695 3.75090233 3.45829131 3.34078217 3.48072829]\n",
      "        >>> print(ci_u[995:])\n",
      "        [4.88316666 4.86924034 4.32032996 4.2822427  4.59360598]\n",
      "        \n",
      "        And again, approximately 90% contain the true value, ``std_true = 4``.\n",
      "        \n",
      "        >>> print(np.sum((ci_l < std_true) & (std_true < ci_u)))\n",
      "        900\n",
      "        \n",
      "        `bootstrap` can also be used to estimate confidence intervals of\n",
      "        multi-sample statistics, including those calculated by hypothesis\n",
      "        tests. `scipy.stats.mood` perform's Mood's test for equal scale parameters,\n",
      "        and it returns two outputs: a statistic, and a p-value. To get a\n",
      "        confidence interval for the test statistic, we first wrap\n",
      "        `scipy.stats.mood` in a function that accepts two sample arguments,\n",
      "        accepts an `axis` keyword argument, and returns only the statistic.\n",
      "        \n",
      "        >>> from scipy.stats import mood\n",
      "        >>> def my_statistic(sample1, sample2, axis):\n",
      "        ...     statistic, _ = mood(sample1, sample2, axis=-1)\n",
      "        ...     return statistic\n",
      "        \n",
      "        Here, we use the 'percentile' method with the default 95% confidence level.\n",
      "        \n",
      "        >>> sample1 = norm.rvs(scale=1, size=100, random_state=rng)\n",
      "        >>> sample2 = norm.rvs(scale=2, size=100, random_state=rng)\n",
      "        >>> data = (sample1, sample2)\n",
      "        >>> res = bootstrap(data, my_statistic, method='basic', random_state=rng)\n",
      "        >>> print(mood(sample1, sample2)[0])  # element 0 is the statistic\n",
      "        -5.521109549096542\n",
      "        >>> print(res.confidence_interval)\n",
      "        ConfidenceInterval(low=-7.255994487314675, high=-4.016202624747605)\n",
      "        \n",
      "        The bootstrap estimate of the standard error is also available.\n",
      "        \n",
      "        >>> print(res.standard_error)\n",
      "        0.8344963846318795\n",
      "        \n",
      "        Paired-sample statistics work, too. For example, consider the Pearson\n",
      "        correlation coefficient.\n",
      "        \n",
      "        >>> from scipy.stats import pearsonr\n",
      "        >>> n = 100\n",
      "        >>> x = np.linspace(0, 10, n)\n",
      "        >>> y = x + rng.uniform(size=n)\n",
      "        >>> print(pearsonr(x, y)[0])  # element 0 is the statistic\n",
      "        0.9962357936065914\n",
      "        \n",
      "        We wrap `pearsonr` so that it returns only the statistic.\n",
      "        \n",
      "        >>> def my_statistic(x, y):\n",
      "        ...     return pearsonr(x, y)[0]\n",
      "        \n",
      "        We call `bootstrap` using ``paired=True``.\n",
      "        Also, since ``my_statistic`` isn't vectorized to calculate the statistic\n",
      "        along a given axis, we pass in ``vectorized=False``.\n",
      "        \n",
      "        >>> res = bootstrap((x, y), my_statistic, vectorized=False, paired=True,\n",
      "        ...                 random_state=rng)\n",
      "        >>> print(res.confidence_interval)\n",
      "        ConfidenceInterval(low=0.9950085825848624, high=0.9971212407917498)\n",
      "    \n",
      "    boschloo_exact(table, alternative='two-sided', n=32)\n",
      "        Perform Boschloo's exact test on a 2x2 contingency table.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        table : array_like of ints\n",
      "            A 2x2 contingency table.  Elements should be non-negative integers.\n",
      "        \n",
      "        alternative : {'two-sided', 'less', 'greater'}, optional\n",
      "            Defines the null and alternative hypotheses. Default is 'two-sided'.\n",
      "            Please see explanations in the Notes section below.\n",
      "        \n",
      "        n : int, optional\n",
      "            Number of sampling points used in the construction of the sampling\n",
      "            method. Note that this argument will automatically be converted to\n",
      "            the next higher power of 2 since `scipy.stats.qmc.Sobol` is used to\n",
      "            select sample points. Default is 32. Must be positive. In most cases,\n",
      "            32 points is enough to reach good precision. More points comes at\n",
      "            performance cost.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        ber : BoschlooExactResult\n",
      "            A result object with the following attributes.\n",
      "        \n",
      "            statistic : float\n",
      "                The statistic used in Boschloo's test; that is, the p-value\n",
      "                from Fisher's exact test.\n",
      "        \n",
      "            pvalue : float\n",
      "                P-value, the probability of obtaining a distribution at least as\n",
      "                extreme as the one that was actually observed, assuming that the\n",
      "                null hypothesis is true.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        chi2_contingency : Chi-square test of independence of variables in a\n",
      "            contingency table.\n",
      "        fisher_exact : Fisher exact test on a 2x2 contingency table.\n",
      "        barnard_exact : Barnard's exact test, which is a more powerful alternative\n",
      "            than Fisher's exact test for 2x2 contingency tables.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        Boschloo's test is an exact test used in the analysis of contingency\n",
      "        tables. It examines the association of two categorical variables, and\n",
      "        is a uniformly more powerful alternative to Fisher's exact test\n",
      "        for 2x2 contingency tables.\n",
      "        \n",
      "        Let's define :math:`X_0` a 2x2 matrix representing the observed sample,\n",
      "        where each column stores the binomial experiment, as in the example\n",
      "        below. Let's also define :math:`p_1, p_2` the theoretical binomial\n",
      "        probabilities for  :math:`x_{11}` and :math:`x_{12}`. When using\n",
      "        Boschloo exact test, we can assert three different null hypotheses :\n",
      "        \n",
      "        - :math:`H_0 : p_1=p_2` versus :math:`H_1 : p_1 < p_2`,\n",
      "          with `alternative` = \"less\"\n",
      "        \n",
      "        - :math:`H_0 : p_1=p_2` versus :math:`H_1 : p_1 > p_2`,\n",
      "          with `alternative` = \"greater\"\n",
      "        \n",
      "        - :math:`H_0 : p_1=p_2` versus :math:`H_1 : p_1 \\neq p_2`,\n",
      "          with `alternative` = \"two-sided\" (default one)\n",
      "        \n",
      "        Boschloo's exact test uses the p-value of Fisher's exact test as a \n",
      "        statistic, and Boschloo's p-value is the probability under the null \n",
      "        hypothesis of observing such an extreme value of this statistic.\n",
      "        \n",
      "        Boschloo's and Barnard's are both more powerful than Fisher's exact\n",
      "        test.\n",
      "        \n",
      "        .. versionadded:: 1.7.0\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] R.D. Boschloo. \"Raised conditional level of significance for the\n",
      "           2 x 2-table when testing the equality of two probabilities\",\n",
      "           Statistica Neerlandica, 24(1), 1970\n",
      "        \n",
      "        .. [2] \"Boschloo's test\", Wikipedia,\n",
      "           https://en.wikipedia.org/wiki/Boschloo%27s_test\n",
      "        \n",
      "        .. [3] Lise M. Saari et al. \"Employee attitudes and job satisfaction\",\n",
      "           Human Resource Management, 43(4), 395-407, 2004,\n",
      "           :doi:`10.1002/hrm.20032`.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        In the following example, we consider the article \"Employee\n",
      "        attitudes and job satisfaction\" [3]_\n",
      "        which reports the results of a survey from 63 scientists and 117 college\n",
      "        professors. Of the 63 scientists, 31 said they were very satisfied with\n",
      "        their jobs, whereas 74 of the college professors were very satisfied\n",
      "        with their work. Is this significant evidence that college\n",
      "        professors are happier with their work than scientists?\n",
      "        The following table summarizes the data mentioned above::\n",
      "        \n",
      "                             college professors   scientists\n",
      "            Very Satisfied   74                     31\n",
      "            Dissatisfied     43                     32\n",
      "        \n",
      "        When working with statistical hypothesis testing, we usually use a\n",
      "        threshold probability or significance level upon which we decide\n",
      "        to reject the null hypothesis :math:`H_0`. Suppose we choose the common\n",
      "        significance level of 5%.\n",
      "        \n",
      "        Our alternative hypothesis is that college professors are truly more\n",
      "        satisfied with their work than scientists. Therefore, we expect\n",
      "        :math:`p_1` the proportion of very satisfied college professors to be\n",
      "        greater than :math:`p_2`, the proportion of very satisfied scientists.\n",
      "        We thus call `boschloo_exact` with the ``alternative=\"greater\"`` option:\n",
      "        \n",
      "        >>> import scipy.stats as stats\n",
      "        >>> res = stats.boschloo_exact([[74, 31], [43, 32]], alternative=\"greater\")\n",
      "        >>> res.statistic\n",
      "        0.0483...\n",
      "        >>> res.pvalue\n",
      "        0.0355...\n",
      "        \n",
      "        Under the null hypothesis that scientists are happier in their work than\n",
      "        college professors, the probability of obtaining test\n",
      "        results at least as extreme as the observed data is approximately 3.55%.\n",
      "        Since this p-value is less than our chosen significance level, we have\n",
      "        evidence to reject :math:`H_0` in favor of the alternative hypothesis.\n",
      "    \n",
      "    boxcox(x, lmbda=None, alpha=None, optimizer=None)\n",
      "        Return a dataset transformed by a Box-Cox power transformation.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x : ndarray\n",
      "            Input array.  Must be positive 1-dimensional.  Must not be constant.\n",
      "        lmbda : {None, scalar}, optional\n",
      "            If `lmbda` is not None, do the transformation for that value.\n",
      "            If `lmbda` is None, find the lambda that maximizes the log-likelihood\n",
      "            function and return it as the second output argument.\n",
      "        alpha : {None, float}, optional\n",
      "            If ``alpha`` is not None, return the ``100 * (1-alpha)%`` confidence\n",
      "            interval for `lmbda` as the third output argument.\n",
      "            Must be between 0.0 and 1.0.\n",
      "        optimizer : callable, optional\n",
      "            If `lmbda` is None, `optimizer` is the scalar optimizer used to find\n",
      "            the value of `lmbda` that minimizes the negative log-likelihood\n",
      "            function. `optimizer` is a callable that accepts one argument:\n",
      "        \n",
      "            fun : callable\n",
      "                The objective function, which evaluates the negative\n",
      "                log-likelihood function at a provided value of `lmbda`\n",
      "        \n",
      "            and returns an object, such as an instance of\n",
      "            `scipy.optimize.OptimizeResult`, which holds the optimal value of\n",
      "            `lmbda` in an attribute `x`.\n",
      "        \n",
      "            See the example in `boxcox_normmax` or the documentation of\n",
      "            `scipy.optimize.minimize_scalar` for more information.\n",
      "        \n",
      "            If `lmbda` is not None, `optimizer` is ignored.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        boxcox : ndarray\n",
      "            Box-Cox power transformed array.\n",
      "        maxlog : float, optional\n",
      "            If the `lmbda` parameter is None, the second returned argument is\n",
      "            the lambda that maximizes the log-likelihood function.\n",
      "        (min_ci, max_ci) : tuple of float, optional\n",
      "            If `lmbda` parameter is None and ``alpha`` is not None, this returned\n",
      "            tuple of floats represents the minimum and maximum confidence limits\n",
      "            given ``alpha``.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        probplot, boxcox_normplot, boxcox_normmax, boxcox_llf\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        The Box-Cox transform is given by::\n",
      "        \n",
      "            y = (x**lmbda - 1) / lmbda,  for lmbda != 0\n",
      "                log(x),                  for lmbda = 0\n",
      "        \n",
      "        `boxcox` requires the input data to be positive.  Sometimes a Box-Cox\n",
      "        transformation provides a shift parameter to achieve this; `boxcox` does\n",
      "        not.  Such a shift parameter is equivalent to adding a positive constant to\n",
      "        `x` before calling `boxcox`.\n",
      "        \n",
      "        The confidence limits returned when ``alpha`` is provided give the interval\n",
      "        where:\n",
      "        \n",
      "        .. math::\n",
      "        \n",
      "            llf(\\hat{\\lambda}) - llf(\\lambda) < \\frac{1}{2}\\chi^2(1 - \\alpha, 1),\n",
      "        \n",
      "        with ``llf`` the log-likelihood function and :math:`\\chi^2` the chi-squared\n",
      "        function.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        G.E.P. Box and D.R. Cox, \"An Analysis of Transformations\", Journal of the\n",
      "        Royal Statistical Society B, 26, 211-252 (1964).\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy import stats\n",
      "        >>> import matplotlib.pyplot as plt\n",
      "        \n",
      "        We generate some random variates from a non-normal distribution and make a\n",
      "        probability plot for it, to show it is non-normal in the tails:\n",
      "        \n",
      "        >>> fig = plt.figure()\n",
      "        >>> ax1 = fig.add_subplot(211)\n",
      "        >>> x = stats.loggamma.rvs(5, size=500) + 5\n",
      "        >>> prob = stats.probplot(x, dist=stats.norm, plot=ax1)\n",
      "        >>> ax1.set_xlabel('')\n",
      "        >>> ax1.set_title('Probplot against normal distribution')\n",
      "        \n",
      "        We now use `boxcox` to transform the data so it's closest to normal:\n",
      "        \n",
      "        >>> ax2 = fig.add_subplot(212)\n",
      "        >>> xt, _ = stats.boxcox(x)\n",
      "        >>> prob = stats.probplot(xt, dist=stats.norm, plot=ax2)\n",
      "        >>> ax2.set_title('Probplot after Box-Cox transformation')\n",
      "        \n",
      "        >>> plt.show()\n",
      "    \n",
      "    boxcox_llf(lmb, data)\n",
      "        The boxcox log-likelihood function.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        lmb : scalar\n",
      "            Parameter for Box-Cox transformation.  See `boxcox` for details.\n",
      "        data : array_like\n",
      "            Data to calculate Box-Cox log-likelihood for.  If `data` is\n",
      "            multi-dimensional, the log-likelihood is calculated along the first\n",
      "            axis.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        llf : float or ndarray\n",
      "            Box-Cox log-likelihood of `data` given `lmb`.  A float for 1-D `data`,\n",
      "            an array otherwise.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        boxcox, probplot, boxcox_normplot, boxcox_normmax\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        The Box-Cox log-likelihood function is defined here as\n",
      "        \n",
      "        .. math::\n",
      "        \n",
      "            llf = (\\lambda - 1) \\sum_i(\\log(x_i)) -\n",
      "                  N/2 \\log(\\sum_i (y_i - \\bar{y})^2 / N),\n",
      "        \n",
      "        where ``y`` is the Box-Cox transformed input data ``x``.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy import stats\n",
      "        >>> import matplotlib.pyplot as plt\n",
      "        >>> from mpl_toolkits.axes_grid1.inset_locator import inset_axes\n",
      "        \n",
      "        Generate some random variates and calculate Box-Cox log-likelihood values\n",
      "        for them for a range of ``lmbda`` values:\n",
      "        \n",
      "        >>> rng = np.random.default_rng()\n",
      "        >>> x = stats.loggamma.rvs(5, loc=10, size=1000, random_state=rng)\n",
      "        >>> lmbdas = np.linspace(-2, 10)\n",
      "        >>> llf = np.zeros(lmbdas.shape, dtype=float)\n",
      "        >>> for ii, lmbda in enumerate(lmbdas):\n",
      "        ...     llf[ii] = stats.boxcox_llf(lmbda, x)\n",
      "        \n",
      "        Also find the optimal lmbda value with `boxcox`:\n",
      "        \n",
      "        >>> x_most_normal, lmbda_optimal = stats.boxcox(x)\n",
      "        \n",
      "        Plot the log-likelihood as function of lmbda.  Add the optimal lmbda as a\n",
      "        horizontal line to check that that's really the optimum:\n",
      "        \n",
      "        >>> fig = plt.figure()\n",
      "        >>> ax = fig.add_subplot(111)\n",
      "        >>> ax.plot(lmbdas, llf, 'b.-')\n",
      "        >>> ax.axhline(stats.boxcox_llf(lmbda_optimal, x), color='r')\n",
      "        >>> ax.set_xlabel('lmbda parameter')\n",
      "        >>> ax.set_ylabel('Box-Cox log-likelihood')\n",
      "        \n",
      "        Now add some probability plots to show that where the log-likelihood is\n",
      "        maximized the data transformed with `boxcox` looks closest to normal:\n",
      "        \n",
      "        >>> locs = [3, 10, 4]  # 'lower left', 'center', 'lower right'\n",
      "        >>> for lmbda, loc in zip([-1, lmbda_optimal, 9], locs):\n",
      "        ...     xt = stats.boxcox(x, lmbda=lmbda)\n",
      "        ...     (osm, osr), (slope, intercept, r_sq) = stats.probplot(xt)\n",
      "        ...     ax_inset = inset_axes(ax, width=\"20%\", height=\"20%\", loc=loc)\n",
      "        ...     ax_inset.plot(osm, osr, 'c.', osm, slope*osm + intercept, 'k-')\n",
      "        ...     ax_inset.set_xticklabels([])\n",
      "        ...     ax_inset.set_yticklabels([])\n",
      "        ...     ax_inset.set_title(r'$\\lambda=%1.2f$' % lmbda)\n",
      "        \n",
      "        >>> plt.show()\n",
      "    \n",
      "    boxcox_normmax(x, brack=None, method='pearsonr', optimizer=None)\n",
      "        Compute optimal Box-Cox transform parameter for input data.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x : array_like\n",
      "            Input array.\n",
      "        brack : 2-tuple, optional, default (-2.0, 2.0)\n",
      "             The starting interval for a downhill bracket search for the default\n",
      "             `optimize.brent` solver. Note that this is in most cases not\n",
      "             critical; the final result is allowed to be outside this bracket.\n",
      "             If `optimizer` is passed, `brack` must be None.\n",
      "        method : str, optional\n",
      "            The method to determine the optimal transform parameter (`boxcox`\n",
      "            ``lmbda`` parameter). Options are:\n",
      "        \n",
      "            'pearsonr'  (default)\n",
      "                Maximizes the Pearson correlation coefficient between\n",
      "                ``y = boxcox(x)`` and the expected values for ``y`` if `x` would be\n",
      "                normally-distributed.\n",
      "        \n",
      "            'mle'\n",
      "                Minimizes the log-likelihood `boxcox_llf`.  This is the method used\n",
      "                in `boxcox`.\n",
      "        \n",
      "            'all'\n",
      "                Use all optimization methods available, and return all results.\n",
      "                Useful to compare different methods.\n",
      "        optimizer : callable, optional\n",
      "            `optimizer` is a callable that accepts one argument:\n",
      "        \n",
      "            fun : callable\n",
      "                The objective function to be optimized. `fun` accepts one argument,\n",
      "                the Box-Cox transform parameter `lmbda`, and returns the negative\n",
      "                log-likelihood function at the provided value. The job of `optimizer`\n",
      "                is to find the value of `lmbda` that minimizes `fun`.\n",
      "        \n",
      "            and returns an object, such as an instance of\n",
      "            `scipy.optimize.OptimizeResult`, which holds the optimal value of\n",
      "            `lmbda` in an attribute `x`.\n",
      "        \n",
      "            See the example below or the documentation of\n",
      "            `scipy.optimize.minimize_scalar` for more information.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        maxlog : float or ndarray\n",
      "            The optimal transform parameter found.  An array instead of a scalar\n",
      "            for ``method='all'``.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        boxcox, boxcox_llf, boxcox_normplot, scipy.optimize.minimize_scalar\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy import stats\n",
      "        >>> import matplotlib.pyplot as plt\n",
      "        \n",
      "        We can generate some data and determine the optimal ``lmbda`` in various\n",
      "        ways:\n",
      "        \n",
      "        >>> rng = np.random.default_rng()\n",
      "        >>> x = stats.loggamma.rvs(5, size=30, random_state=rng) + 5\n",
      "        >>> y, lmax_mle = stats.boxcox(x)\n",
      "        >>> lmax_pearsonr = stats.boxcox_normmax(x)\n",
      "        \n",
      "        >>> lmax_mle\n",
      "        1.4613865614008015\n",
      "        >>> lmax_pearsonr\n",
      "        1.6685004886804342\n",
      "        >>> stats.boxcox_normmax(x, method='all')\n",
      "        array([1.66850049, 1.46138656])\n",
      "        \n",
      "        >>> fig = plt.figure()\n",
      "        >>> ax = fig.add_subplot(111)\n",
      "        >>> prob = stats.boxcox_normplot(x, -10, 10, plot=ax)\n",
      "        >>> ax.axvline(lmax_mle, color='r')\n",
      "        >>> ax.axvline(lmax_pearsonr, color='g', ls='--')\n",
      "        \n",
      "        >>> plt.show()\n",
      "        \n",
      "        Alternatively, we can define our own `optimizer` function. Suppose we\n",
      "        are only interested in values of `lmbda` on the interval [6, 7], we\n",
      "        want to use `scipy.optimize.minimize_scalar` with ``method='bounded'``,\n",
      "        and we want to use tighter tolerances when optimizing the log-likelihood\n",
      "        function. To do this, we define a function that accepts positional argument\n",
      "        `fun` and uses `scipy.optimize.minimize_scalar` to minimize `fun` subject\n",
      "        to the provided bounds and tolerances:\n",
      "        \n",
      "        >>> from scipy import optimize\n",
      "        >>> options = {'xatol': 1e-12}  # absolute tolerance on `x`\n",
      "        >>> def optimizer(fun):\n",
      "        ...     return optimize.minimize_scalar(fun, bounds=(6, 7),\n",
      "        ...                                     method=\"bounded\", options=options)\n",
      "        >>> stats.boxcox_normmax(x, optimizer=optimizer)\n",
      "        6.000...\n",
      "    \n",
      "    boxcox_normplot(x, la, lb, plot=None, N=80)\n",
      "        Compute parameters for a Box-Cox normality plot, optionally show it.\n",
      "        \n",
      "        A Box-Cox normality plot shows graphically what the best transformation\n",
      "        parameter is to use in `boxcox` to obtain a distribution that is close\n",
      "        to normal.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x : array_like\n",
      "            Input array.\n",
      "        la, lb : scalar\n",
      "            The lower and upper bounds for the ``lmbda`` values to pass to `boxcox`\n",
      "            for Box-Cox transformations.  These are also the limits of the\n",
      "            horizontal axis of the plot if that is generated.\n",
      "        plot : object, optional\n",
      "            If given, plots the quantiles and least squares fit.\n",
      "            `plot` is an object that has to have methods \"plot\" and \"text\".\n",
      "            The `matplotlib.pyplot` module or a Matplotlib Axes object can be used,\n",
      "            or a custom object with the same methods.\n",
      "            Default is None, which means that no plot is created.\n",
      "        N : int, optional\n",
      "            Number of points on the horizontal axis (equally distributed from\n",
      "            `la` to `lb`).\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        lmbdas : ndarray\n",
      "            The ``lmbda`` values for which a Box-Cox transform was done.\n",
      "        ppcc : ndarray\n",
      "            Probability Plot Correlelation Coefficient, as obtained from `probplot`\n",
      "            when fitting the Box-Cox transformed input `x` against a normal\n",
      "            distribution.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        probplot, boxcox, boxcox_normmax, boxcox_llf, ppcc_max\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        Even if `plot` is given, the figure is not shown or saved by\n",
      "        `boxcox_normplot`; ``plt.show()`` or ``plt.savefig('figname.png')``\n",
      "        should be used after calling `probplot`.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy import stats\n",
      "        >>> import matplotlib.pyplot as plt\n",
      "        \n",
      "        Generate some non-normally distributed data, and create a Box-Cox plot:\n",
      "        \n",
      "        >>> x = stats.loggamma.rvs(5, size=500) + 5\n",
      "        >>> fig = plt.figure()\n",
      "        >>> ax = fig.add_subplot(111)\n",
      "        >>> prob = stats.boxcox_normplot(x, -20, 20, plot=ax)\n",
      "        \n",
      "        Determine and plot the optimal ``lmbda`` to transform ``x`` and plot it in\n",
      "        the same plot:\n",
      "        \n",
      "        >>> _, maxlog = stats.boxcox(x)\n",
      "        >>> ax.axvline(maxlog, color='r')\n",
      "        \n",
      "        >>> plt.show()\n",
      "    \n",
      "    brunnermunzel(x, y, alternative='two-sided', distribution='t', nan_policy='propagate')\n",
      "        Compute the Brunner-Munzel test on samples x and y.\n",
      "        \n",
      "        The Brunner-Munzel test is a nonparametric test of the null hypothesis that\n",
      "        when values are taken one by one from each group, the probabilities of\n",
      "        getting large values in both groups are equal.\n",
      "        Unlike the Wilcoxon-Mann-Whitney's U test, this does not require the\n",
      "        assumption of equivariance of two groups. Note that this does not assume\n",
      "        the distributions are same. This test works on two independent samples,\n",
      "        which may have different sizes.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x, y : array_like\n",
      "            Array of samples, should be one-dimensional.\n",
      "        alternative : {'two-sided', 'less', 'greater'}, optional\n",
      "            Defines the alternative hypothesis.\n",
      "            The following options are available (default is 'two-sided'):\n",
      "        \n",
      "              * 'two-sided'\n",
      "              * 'less': one-sided\n",
      "              * 'greater': one-sided\n",
      "        distribution : {'t', 'normal'}, optional\n",
      "            Defines how to get the p-value.\n",
      "            The following options are available (default is 't'):\n",
      "        \n",
      "              * 't': get the p-value by t-distribution\n",
      "              * 'normal': get the p-value by standard normal distribution.\n",
      "        nan_policy : {'propagate', 'raise', 'omit'}, optional\n",
      "            Defines how to handle when input contains nan.\n",
      "            The following options are available (default is 'propagate'):\n",
      "        \n",
      "              * 'propagate': returns nan\n",
      "              * 'raise': throws an error\n",
      "              * 'omit': performs the calculations ignoring nan values\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        statistic : float\n",
      "            The Brunner-Munzer W statistic.\n",
      "        pvalue : float\n",
      "            p-value assuming an t distribution. One-sided or\n",
      "            two-sided, depending on the choice of `alternative` and `distribution`.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        mannwhitneyu : Mann-Whitney rank test on two samples.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        Brunner and Munzel recommended to estimate the p-value by t-distribution\n",
      "        when the size of data is 50 or less. If the size is lower than 10, it would\n",
      "        be better to use permuted Brunner Munzel test (see [2]_).\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] Brunner, E. and Munzel, U. \"The nonparametric Benhrens-Fisher\n",
      "               problem: Asymptotic theory and a small-sample approximation\".\n",
      "               Biometrical Journal. Vol. 42(2000): 17-25.\n",
      "        .. [2] Neubert, K. and Brunner, E. \"A studentized permutation test for the\n",
      "               non-parametric Behrens-Fisher problem\". Computational Statistics and\n",
      "               Data Analysis. Vol. 51(2007): 5192-5204.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy import stats\n",
      "        >>> x1 = [1,2,1,1,1,1,1,1,1,1,2,4,1,1]\n",
      "        >>> x2 = [3,3,4,3,1,2,3,1,1,5,4]\n",
      "        >>> w, p_value = stats.brunnermunzel(x1, x2)\n",
      "        >>> w\n",
      "        3.1374674823029505\n",
      "        >>> p_value\n",
      "        0.0057862086661515377\n",
      "    \n",
      "    chi2_contingency(observed, correction=True, lambda_=None)\n",
      "        Chi-square test of independence of variables in a contingency table.\n",
      "        \n",
      "        This function computes the chi-square statistic and p-value for the\n",
      "        hypothesis test of independence of the observed frequencies in the\n",
      "        contingency table [1]_ `observed`.  The expected frequencies are computed\n",
      "        based on the marginal sums under the assumption of independence; see\n",
      "        `scipy.stats.contingency.expected_freq`.  The number of degrees of\n",
      "        freedom is (expressed using numpy functions and attributes)::\n",
      "        \n",
      "            dof = observed.size - sum(observed.shape) + observed.ndim - 1\n",
      "        \n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        observed : array_like\n",
      "            The contingency table. The table contains the observed frequencies\n",
      "            (i.e. number of occurrences) in each category.  In the two-dimensional\n",
      "            case, the table is often described as an \"R x C table\".\n",
      "        correction : bool, optional\n",
      "            If True, *and* the degrees of freedom is 1, apply Yates' correction\n",
      "            for continuity.  The effect of the correction is to adjust each\n",
      "            observed value by 0.5 towards the corresponding expected value.\n",
      "        lambda_ : float or str, optional\n",
      "            By default, the statistic computed in this test is Pearson's\n",
      "            chi-squared statistic [2]_.  `lambda_` allows a statistic from the\n",
      "            Cressie-Read power divergence family [3]_ to be used instead.  See\n",
      "            `power_divergence` for details.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        chi2 : float\n",
      "            The test statistic.\n",
      "        p : float\n",
      "            The p-value of the test\n",
      "        dof : int\n",
      "            Degrees of freedom\n",
      "        expected : ndarray, same shape as `observed`\n",
      "            The expected frequencies, based on the marginal sums of the table.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        contingency.expected_freq\n",
      "        fisher_exact\n",
      "        chisquare\n",
      "        power_divergence\n",
      "        barnard_exact\n",
      "        boschloo_exact\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        An often quoted guideline for the validity of this calculation is that\n",
      "        the test should be used only if the observed and expected frequencies\n",
      "        in each cell are at least 5.\n",
      "        \n",
      "        This is a test for the independence of different categories of a\n",
      "        population. The test is only meaningful when the dimension of\n",
      "        `observed` is two or more.  Applying the test to a one-dimensional\n",
      "        table will always result in `expected` equal to `observed` and a\n",
      "        chi-square statistic equal to 0.\n",
      "        \n",
      "        This function does not handle masked arrays, because the calculation\n",
      "        does not make sense with missing values.\n",
      "        \n",
      "        Like stats.chisquare, this function computes a chi-square statistic;\n",
      "        the convenience this function provides is to figure out the expected\n",
      "        frequencies and degrees of freedom from the given contingency table.\n",
      "        If these were already known, and if the Yates' correction was not\n",
      "        required, one could use stats.chisquare.  That is, if one calls::\n",
      "        \n",
      "            chi2, p, dof, ex = chi2_contingency(obs, correction=False)\n",
      "        \n",
      "        then the following is true::\n",
      "        \n",
      "            (chi2, p) == stats.chisquare(obs.ravel(), f_exp=ex.ravel(),\n",
      "                                         ddof=obs.size - 1 - dof)\n",
      "        \n",
      "        The `lambda_` argument was added in version 0.13.0 of scipy.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] \"Contingency table\",\n",
      "               https://en.wikipedia.org/wiki/Contingency_table\n",
      "        .. [2] \"Pearson's chi-squared test\",\n",
      "               https://en.wikipedia.org/wiki/Pearson%27s_chi-squared_test\n",
      "        .. [3] Cressie, N. and Read, T. R. C., \"Multinomial Goodness-of-Fit\n",
      "               Tests\", J. Royal Stat. Soc. Series B, Vol. 46, No. 3 (1984),\n",
      "               pp. 440-464.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        A two-way example (2 x 3):\n",
      "        \n",
      "        >>> from scipy.stats import chi2_contingency\n",
      "        >>> obs = np.array([[10, 10, 20], [20, 20, 20]])\n",
      "        >>> chi2_contingency(obs)\n",
      "        (2.7777777777777777,\n",
      "         0.24935220877729619,\n",
      "         2,\n",
      "         array([[ 12.,  12.,  16.],\n",
      "                [ 18.,  18.,  24.]]))\n",
      "        \n",
      "        Perform the test using the log-likelihood ratio (i.e. the \"G-test\")\n",
      "        instead of Pearson's chi-squared statistic.\n",
      "        \n",
      "        >>> g, p, dof, expctd = chi2_contingency(obs, lambda_=\"log-likelihood\")\n",
      "        >>> g, p\n",
      "        (2.7688587616781319, 0.25046668010954165)\n",
      "        \n",
      "        A four-way example (2 x 2 x 2 x 2):\n",
      "        \n",
      "        >>> obs = np.array(\n",
      "        ...     [[[[12, 17],\n",
      "        ...        [11, 16]],\n",
      "        ...       [[11, 12],\n",
      "        ...        [15, 16]]],\n",
      "        ...      [[[23, 15],\n",
      "        ...        [30, 22]],\n",
      "        ...       [[14, 17],\n",
      "        ...        [15, 16]]]])\n",
      "        >>> chi2_contingency(obs)\n",
      "        (8.7584514426741897,\n",
      "         0.64417725029295503,\n",
      "         11,\n",
      "         array([[[[ 14.15462386,  14.15462386],\n",
      "                  [ 16.49423111,  16.49423111]],\n",
      "                 [[ 11.2461395 ,  11.2461395 ],\n",
      "                  [ 13.10500554,  13.10500554]]],\n",
      "                [[[ 19.5591166 ,  19.5591166 ],\n",
      "                  [ 22.79202844,  22.79202844]],\n",
      "                 [[ 15.54012004,  15.54012004],\n",
      "                  [ 18.10873492,  18.10873492]]]]))\n",
      "    \n",
      "    chisquare(f_obs, f_exp=None, ddof=0, axis=0)\n",
      "        Calculate a one-way chi-square test.\n",
      "        \n",
      "        The chi-square test tests the null hypothesis that the categorical data\n",
      "        has the given frequencies.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        f_obs : array_like\n",
      "            Observed frequencies in each category.\n",
      "        f_exp : array_like, optional\n",
      "            Expected frequencies in each category.  By default the categories are\n",
      "            assumed to be equally likely.\n",
      "        ddof : int, optional\n",
      "            \"Delta degrees of freedom\": adjustment to the degrees of freedom\n",
      "            for the p-value.  The p-value is computed using a chi-squared\n",
      "            distribution with ``k - 1 - ddof`` degrees of freedom, where `k`\n",
      "            is the number of observed frequencies.  The default value of `ddof`\n",
      "            is 0.\n",
      "        axis : int or None, optional\n",
      "            The axis of the broadcast result of `f_obs` and `f_exp` along which to\n",
      "            apply the test.  If axis is None, all values in `f_obs` are treated\n",
      "            as a single data set.  Default is 0.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        chisq : float or ndarray\n",
      "            The chi-squared test statistic.  The value is a float if `axis` is\n",
      "            None or `f_obs` and `f_exp` are 1-D.\n",
      "        p : float or ndarray\n",
      "            The p-value of the test.  The value is a float if `ddof` and the\n",
      "            return value `chisq` are scalars.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        scipy.stats.power_divergence\n",
      "        scipy.stats.fisher_exact : Fisher exact test on a 2x2 contingency table.\n",
      "        scipy.stats.barnard_exact : An unconditional exact test. An alternative\n",
      "            to chi-squared test for small sample sizes.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        This test is invalid when the observed or expected frequencies in each\n",
      "        category are too small.  A typical rule is that all of the observed\n",
      "        and expected frequencies should be at least 5. According to [3]_, the\n",
      "        total number of samples is recommended to be greater than 13,\n",
      "        otherwise exact tests (such as Barnard's Exact test) should be used\n",
      "        because they do not overreject.\n",
      "        \n",
      "        Also, the sum of the observed and expected frequencies must be the same\n",
      "        for the test to be valid; `chisquare` raises an error if the sums do not\n",
      "        agree within a relative tolerance of ``1e-8``.\n",
      "        \n",
      "        The default degrees of freedom, k-1, are for the case when no parameters\n",
      "        of the distribution are estimated. If p parameters are estimated by\n",
      "        efficient maximum likelihood then the correct degrees of freedom are\n",
      "        k-1-p. If the parameters are estimated in a different way, then the\n",
      "        dof can be between k-1-p and k-1. However, it is also possible that\n",
      "        the asymptotic distribution is not chi-square, in which case this test\n",
      "        is not appropriate.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] Lowry, Richard.  \"Concepts and Applications of Inferential\n",
      "               Statistics\". Chapter 8.\n",
      "               https://web.archive.org/web/20171022032306/http://vassarstats.net:80/textbook/ch8pt1.html\n",
      "        .. [2] \"Chi-squared test\", https://en.wikipedia.org/wiki/Chi-squared_test\n",
      "        .. [3] Pearson, Karl. \"On the criterion that a given system of deviations from the probable\n",
      "               in the case of a correlated system of variables is such that it can be reasonably\n",
      "               supposed to have arisen from random sampling\", Philosophical Magazine. Series 5. 50\n",
      "               (1900), pp. 157-175.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        When just `f_obs` is given, it is assumed that the expected frequencies\n",
      "        are uniform and given by the mean of the observed frequencies.\n",
      "        \n",
      "        >>> from scipy.stats import chisquare\n",
      "        >>> chisquare([16, 18, 16, 14, 12, 12])\n",
      "        (2.0, 0.84914503608460956)\n",
      "        \n",
      "        With `f_exp` the expected frequencies can be given.\n",
      "        \n",
      "        >>> chisquare([16, 18, 16, 14, 12, 12], f_exp=[16, 16, 16, 16, 16, 8])\n",
      "        (3.5, 0.62338762774958223)\n",
      "        \n",
      "        When `f_obs` is 2-D, by default the test is applied to each column.\n",
      "        \n",
      "        >>> obs = np.array([[16, 18, 16, 14, 12, 12], [32, 24, 16, 28, 20, 24]]).T\n",
      "        >>> obs.shape\n",
      "        (6, 2)\n",
      "        >>> chisquare(obs)\n",
      "        (array([ 2.        ,  6.66666667]), array([ 0.84914504,  0.24663415]))\n",
      "        \n",
      "        By setting ``axis=None``, the test is applied to all data in the array,\n",
      "        which is equivalent to applying the test to the flattened array.\n",
      "        \n",
      "        >>> chisquare(obs, axis=None)\n",
      "        (23.31034482758621, 0.015975692534127565)\n",
      "        >>> chisquare(obs.ravel())\n",
      "        (23.31034482758621, 0.015975692534127565)\n",
      "        \n",
      "        `ddof` is the change to make to the default degrees of freedom.\n",
      "        \n",
      "        >>> chisquare([16, 18, 16, 14, 12, 12], ddof=1)\n",
      "        (2.0, 0.73575888234288467)\n",
      "        \n",
      "        The calculation of the p-values is done by broadcasting the\n",
      "        chi-squared statistic with `ddof`.\n",
      "        \n",
      "        >>> chisquare([16, 18, 16, 14, 12, 12], ddof=[0,1,2])\n",
      "        (2.0, array([ 0.84914504,  0.73575888,  0.5724067 ]))\n",
      "        \n",
      "        `f_obs` and `f_exp` are also broadcast.  In the following, `f_obs` has\n",
      "        shape (6,) and `f_exp` has shape (2, 6), so the result of broadcasting\n",
      "        `f_obs` and `f_exp` has shape (2, 6).  To compute the desired chi-squared\n",
      "        statistics, we use ``axis=1``:\n",
      "        \n",
      "        >>> chisquare([16, 18, 16, 14, 12, 12],\n",
      "        ...           f_exp=[[16, 16, 16, 16, 16, 8], [8, 20, 20, 16, 12, 12]],\n",
      "        ...           axis=1)\n",
      "        (array([ 3.5 ,  9.25]), array([ 0.62338763,  0.09949846]))\n",
      "    \n",
      "    circmean(samples, high=6.283185307179586, low=0, axis=None, nan_policy='propagate')\n",
      "        Compute the circular mean for samples in a range.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        samples : array_like\n",
      "            Input array.\n",
      "        high : float or int, optional\n",
      "            High boundary for circular mean range.  Default is ``2*pi``.\n",
      "        low : float or int, optional\n",
      "            Low boundary for circular mean range.  Default is 0.\n",
      "        axis : int, optional\n",
      "            Axis along which means are computed.  The default is to compute\n",
      "            the mean of the flattened array.\n",
      "        nan_policy : {'propagate', 'raise', 'omit'}, optional\n",
      "            Defines how to handle when input contains nan. 'propagate' returns nan,\n",
      "            'raise' throws an error, 'omit' performs the calculations ignoring nan\n",
      "            values. Default is 'propagate'.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        circmean : float\n",
      "            Circular mean.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy.stats import circmean\n",
      "        >>> circmean([0.1, 2*np.pi+0.2, 6*np.pi+0.3])\n",
      "        0.2\n",
      "        \n",
      "        >>> from scipy.stats import circmean\n",
      "        >>> circmean([0.2, 1.4, 2.6], high = 1, low = 0)\n",
      "        0.4\n",
      "    \n",
      "    circstd(samples, high=6.283185307179586, low=0, axis=None, nan_policy='propagate')\n",
      "        Compute the circular standard deviation for samples assumed to be in the\n",
      "        range [low to high].\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        samples : array_like\n",
      "            Input array.\n",
      "        high : float or int, optional\n",
      "            High boundary for circular standard deviation range.\n",
      "            Default is ``2*pi``.\n",
      "        low : float or int, optional\n",
      "            Low boundary for circular standard deviation range.  Default is 0.\n",
      "        axis : int, optional\n",
      "            Axis along which standard deviations are computed.  The default is\n",
      "            to compute the standard deviation of the flattened array.\n",
      "        nan_policy : {'propagate', 'raise', 'omit'}, optional\n",
      "            Defines how to handle when input contains nan. 'propagate' returns nan,\n",
      "            'raise' throws an error, 'omit' performs the calculations ignoring nan\n",
      "            values. Default is 'propagate'.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        circstd : float\n",
      "            Circular standard deviation.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        This uses a definition of circular standard deviation that in the limit of\n",
      "        small angles returns a number close to the 'linear' standard deviation.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy.stats import circstd\n",
      "        >>> circstd([0, 0.1*np.pi/2, 0.001*np.pi, 0.03*np.pi/2])\n",
      "        0.063564063306\n",
      "    \n",
      "    circvar(samples, high=6.283185307179586, low=0, axis=None, nan_policy='propagate')\n",
      "        Compute the circular variance for samples assumed to be in a range.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        samples : array_like\n",
      "            Input array.\n",
      "        high : float or int, optional\n",
      "            High boundary for circular variance range.  Default is ``2*pi``.\n",
      "        low : float or int, optional\n",
      "            Low boundary for circular variance range.  Default is 0.\n",
      "        axis : int, optional\n",
      "            Axis along which variances are computed.  The default is to compute\n",
      "            the variance of the flattened array.\n",
      "        nan_policy : {'propagate', 'raise', 'omit'}, optional\n",
      "            Defines how to handle when input contains nan. 'propagate' returns nan,\n",
      "            'raise' throws an error, 'omit' performs the calculations ignoring nan\n",
      "            values. Default is 'propagate'.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        circvar : float\n",
      "            Circular variance.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        This uses a definition of circular variance that in the limit of small\n",
      "        angles returns a number close to the 'linear' variance.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy.stats import circvar\n",
      "        >>> circvar([0, 2*np.pi/3, 5*np.pi/3])\n",
      "        2.19722457734\n",
      "    \n",
      "    combine_pvalues(pvalues, method='fisher', weights=None)\n",
      "        Combine p-values from independent tests bearing upon the same hypothesis.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        pvalues : array_like, 1-D\n",
      "            Array of p-values assumed to come from independent tests.\n",
      "        method : {'fisher', 'pearson', 'tippett', 'stouffer',\n",
      "                  'mudholkar_george'}, optional\n",
      "        \n",
      "            Name of method to use to combine p-values.\n",
      "            The following methods are available (default is 'fisher'):\n",
      "        \n",
      "              * 'fisher': Fisher's method (Fisher's combined probability test), the\n",
      "                sum of the logarithm of the p-values\n",
      "              * 'pearson': Pearson's method (similar to Fisher's but uses sum of the\n",
      "                complement of the p-values inside the logarithms)\n",
      "              * 'tippett': Tippett's method (minimum of p-values)\n",
      "              * 'stouffer': Stouffer's Z-score method\n",
      "              * 'mudholkar_george': the difference of Fisher's and Pearson's methods\n",
      "                divided by 2\n",
      "        weights : array_like, 1-D, optional\n",
      "            Optional array of weights used only for Stouffer's Z-score method.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        statistic: float\n",
      "            The statistic calculated by the specified method.\n",
      "        pval: float\n",
      "            The combined p-value.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        Fisher's method (also known as Fisher's combined probability test) [1]_ uses\n",
      "        a chi-squared statistic to compute a combined p-value. The closely related\n",
      "        Stouffer's Z-score method [2]_ uses Z-scores rather than p-values. The\n",
      "        advantage of Stouffer's method is that it is straightforward to introduce\n",
      "        weights, which can make Stouffer's method more powerful than Fisher's\n",
      "        method when the p-values are from studies of different size [6]_ [7]_.\n",
      "        The Pearson's method uses :math:`log(1-p_i)` inside the sum whereas Fisher's\n",
      "        method uses :math:`log(p_i)` [4]_. For Fisher's and Pearson's method, the\n",
      "        sum of the logarithms is multiplied by -2 in the implementation. This\n",
      "        quantity has a chi-square distribution that determines the p-value. The\n",
      "        `mudholkar_george` method is the difference of the Fisher's and Pearson's\n",
      "        test statistics, each of which include the -2 factor [4]_. However, the\n",
      "        `mudholkar_george` method does not include these -2 factors. The test\n",
      "        statistic of `mudholkar_george` is the sum of logisitic random variables and\n",
      "        equation 3.6 in [3]_ is used to approximate the p-value based on Student's\n",
      "        t-distribution.\n",
      "        \n",
      "        Fisher's method may be extended to combine p-values from dependent tests\n",
      "        [5]_. Extensions such as Brown's method and Kost's method are not currently\n",
      "        implemented.\n",
      "        \n",
      "        .. versionadded:: 0.15.0\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] https://en.wikipedia.org/wiki/Fisher%27s_method\n",
      "        .. [2] https://en.wikipedia.org/wiki/Fisher%27s_method#Relation_to_Stouffer.27s_Z-score_method\n",
      "        .. [3] George, E. O., and G. S. Mudholkar. \"On the convolution of logistic\n",
      "               random variables.\" Metrika 30.1 (1983): 1-13.\n",
      "        .. [4] Heard, N. and Rubin-Delanchey, P. \"Choosing between methods of\n",
      "               combining p-values.\"  Biometrika 105.1 (2018): 239-246.\n",
      "        .. [5] Whitlock, M. C. \"Combining probability from independent tests: the\n",
      "               weighted Z-method is superior to Fisher's approach.\" Journal of\n",
      "               Evolutionary Biology 18, no. 5 (2005): 1368-1373.\n",
      "        .. [6] Zaykin, Dmitri V. \"Optimally weighted Z-test is a powerful method\n",
      "               for combining probabilities in meta-analysis.\" Journal of\n",
      "               Evolutionary Biology 24, no. 8 (2011): 1836-1841.\n",
      "        .. [7] https://en.wikipedia.org/wiki/Extensions_of_Fisher%27s_method\n",
      "    \n",
      "    cramervonmises(rvs, cdf, args=())\n",
      "        Perform the one-sample Cramér-von Mises test for goodness of fit.\n",
      "        \n",
      "        This performs a test of the goodness of fit of a cumulative distribution\n",
      "        function (cdf) :math:`F` compared to the empirical distribution function\n",
      "        :math:`F_n` of observed random variates :math:`X_1, ..., X_n` that are\n",
      "        assumed to be independent and identically distributed ([1]_).\n",
      "        The null hypothesis is that the :math:`X_i` have cumulative distribution\n",
      "        :math:`F`.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        rvs : array_like\n",
      "            A 1-D array of observed values of the random variables :math:`X_i`.\n",
      "        cdf : str or callable\n",
      "            The cumulative distribution function :math:`F` to test the\n",
      "            observations against. If a string, it should be the name of a\n",
      "            distribution in `scipy.stats`. If a callable, that callable is used\n",
      "            to calculate the cdf: ``cdf(x, *args) -> float``.\n",
      "        args : tuple, optional\n",
      "            Distribution parameters. These are assumed to be known; see Notes.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        res : object with attributes\n",
      "            statistic : float\n",
      "                Cramér-von Mises statistic.\n",
      "            pvalue : float\n",
      "                The p-value.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        kstest, cramervonmises_2samp\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        .. versionadded:: 1.6.0\n",
      "        \n",
      "        The p-value relies on the approximation given by equation 1.8 in [2]_.\n",
      "        It is important to keep in mind that the p-value is only accurate if\n",
      "        one tests a simple hypothesis, i.e. the parameters of the reference\n",
      "        distribution are known. If the parameters are estimated from the data\n",
      "        (composite hypothesis), the computed p-value is not reliable.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] Cramér-von Mises criterion, Wikipedia,\n",
      "               https://en.wikipedia.org/wiki/Cram%C3%A9r%E2%80%93von_Mises_criterion\n",
      "        .. [2] Csorgo, S. and Faraway, J. (1996). The Exact and Asymptotic\n",
      "               Distribution of Cramér-von Mises Statistics. Journal of the\n",
      "               Royal Statistical Society, pp. 221-234.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        \n",
      "        Suppose we wish to test whether data generated by ``scipy.stats.norm.rvs``\n",
      "        were, in fact, drawn from the standard normal distribution. We choose a\n",
      "        significance level of alpha=0.05.\n",
      "        \n",
      "        >>> from scipy import stats\n",
      "        >>> rng = np.random.default_rng()\n",
      "        >>> x = stats.norm.rvs(size=500, random_state=rng)\n",
      "        >>> res = stats.cramervonmises(x, 'norm')\n",
      "        >>> res.statistic, res.pvalue\n",
      "        (0.49121480855028343, 0.04189256516661377)\n",
      "        \n",
      "        The p-value 0.79 exceeds our chosen significance level, so we do not\n",
      "        reject the null hypothesis that the observed sample is drawn from the\n",
      "        standard normal distribution.\n",
      "        \n",
      "        Now suppose we wish to check whether the same samples shifted by 2.1 is\n",
      "        consistent with being drawn from a normal distribution with a mean of 2.\n",
      "        \n",
      "        >>> y = x + 2.1\n",
      "        >>> res = stats.cramervonmises(y, 'norm', args=(2,))\n",
      "        >>> res.statistic, res.pvalue\n",
      "        (0.07400330012187435, 0.7274595666160468)\n",
      "        \n",
      "        Here we have used the `args` keyword to specify the mean (``loc``)\n",
      "        of the normal distribution to test the data against. This is equivalent\n",
      "        to the following, in which we create a frozen normal distribution with\n",
      "        mean 2.1, then pass its ``cdf`` method as an argument.\n",
      "        \n",
      "        >>> frozen_dist = stats.norm(loc=2)\n",
      "        >>> res = stats.cramervonmises(y, frozen_dist.cdf)\n",
      "        >>> res.statistic, res.pvalue\n",
      "        (0.07400330012187435, 0.7274595666160468)\n",
      "        \n",
      "        In either case, we would reject the null hypothesis that the observed\n",
      "        sample is drawn from a normal distribution with a mean of 2 (and default\n",
      "        variance of 1) because the p-value 0.04 is less than our chosen\n",
      "        significance level.\n",
      "    \n",
      "    cramervonmises_2samp(x, y, method='auto')\n",
      "        Perform the two-sample Cramér-von Mises test for goodness of fit.\n",
      "        \n",
      "        This is the two-sample version of the Cramér-von Mises test ([1]_):\n",
      "        for two independent samples :math:`X_1, ..., X_n` and\n",
      "        :math:`Y_1, ..., Y_m`, the null hypothesis is that the samples\n",
      "        come from the same (unspecified) continuous distribution.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x : array_like\n",
      "            A 1-D array of observed values of the random variables :math:`X_i`.\n",
      "        y : array_like\n",
      "            A 1-D array of observed values of the random variables :math:`Y_i`.\n",
      "        method : {'auto', 'asymptotic', 'exact'}, optional\n",
      "            The method used to compute the p-value, see Notes for details.\n",
      "            The default is 'auto'.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        res : object with attributes\n",
      "            statistic : float\n",
      "                Cramér-von Mises statistic.\n",
      "            pvalue : float\n",
      "                The p-value.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        cramervonmises, anderson_ksamp, epps_singleton_2samp, ks_2samp\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        .. versionadded:: 1.7.0\n",
      "        \n",
      "        The statistic is computed according to equation 9 in [2]_. The\n",
      "        calculation of the p-value depends on the keyword `method`:\n",
      "        \n",
      "        - ``asymptotic``: The p-value is approximated by using the limiting\n",
      "          distribution of the test statistic.\n",
      "        - ``exact``: The exact p-value is computed by enumerating all\n",
      "          possible combinations of the test statistic, see [2]_.\n",
      "        \n",
      "        The exact calculation will be very slow even for moderate sample\n",
      "        sizes as the number of combinations increases rapidly with the\n",
      "        size of the samples. If ``method=='auto'``, the exact approach\n",
      "        is used if both samples contain less than 10 observations,\n",
      "        otherwise the asymptotic distribution is used.\n",
      "        \n",
      "        If the underlying distribution is not continuous, the p-value is likely to\n",
      "        be conservative (Section 6.2 in [3]_). When ranking the data to compute\n",
      "        the test statistic, midranks are used if there are ties.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] https://en.wikipedia.org/wiki/Cramer-von_Mises_criterion\n",
      "        .. [2] Anderson, T.W. (1962). On the distribution of the two-sample\n",
      "               Cramer-von-Mises criterion. The Annals of Mathematical\n",
      "               Statistics, pp. 1148-1159.\n",
      "        .. [3] Conover, W.J., Practical Nonparametric Statistics, 1971.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        \n",
      "        Suppose we wish to test whether two samples generated by\n",
      "        ``scipy.stats.norm.rvs`` have the same distribution. We choose a\n",
      "        significance level of alpha=0.05.\n",
      "        \n",
      "        >>> from scipy import stats\n",
      "        >>> rng = np.random.default_rng()\n",
      "        >>> x = stats.norm.rvs(size=100, random_state=rng)\n",
      "        >>> y = stats.norm.rvs(size=70, random_state=rng)\n",
      "        >>> res = stats.cramervonmises_2samp(x, y)\n",
      "        >>> res.statistic, res.pvalue\n",
      "        (0.29376470588235293, 0.1412873014573014)\n",
      "        \n",
      "        The p-value exceeds our chosen significance level, so we do not\n",
      "        reject the null hypothesis that the observed samples are drawn from the\n",
      "        same distribution.\n",
      "        \n",
      "        For small sample sizes, one can compute the exact p-values:\n",
      "        \n",
      "        >>> x = stats.norm.rvs(size=7, random_state=rng)\n",
      "        >>> y = stats.t.rvs(df=2, size=6, random_state=rng)\n",
      "        >>> res = stats.cramervonmises_2samp(x, y, method='exact')\n",
      "        >>> res.statistic, res.pvalue\n",
      "        (0.197802197802198, 0.31643356643356646)\n",
      "        \n",
      "        The p-value based on the asymptotic distribution is a good approximation\n",
      "        even though the sample size is small.\n",
      "        \n",
      "        >>> res = stats.cramervonmises_2samp(x, y, method='asymptotic')\n",
      "        >>> res.statistic, res.pvalue\n",
      "        (0.197802197802198, 0.2966041181527128)\n",
      "        \n",
      "        Independent of the method, one would not reject the null hypothesis at the\n",
      "        chosen significance level in this example.\n",
      "    \n",
      "    cumfreq(a, numbins=10, defaultreallimits=None, weights=None)\n",
      "        Return a cumulative frequency histogram, using the histogram function.\n",
      "        \n",
      "        A cumulative histogram is a mapping that counts the cumulative number of\n",
      "        observations in all of the bins up to the specified bin.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        a : array_like\n",
      "            Input array.\n",
      "        numbins : int, optional\n",
      "            The number of bins to use for the histogram. Default is 10.\n",
      "        defaultreallimits : tuple (lower, upper), optional\n",
      "            The lower and upper values for the range of the histogram.\n",
      "            If no value is given, a range slightly larger than the range of the\n",
      "            values in `a` is used. Specifically ``(a.min() - s, a.max() + s)``,\n",
      "            where ``s = (1/2)(a.max() - a.min()) / (numbins - 1)``.\n",
      "        weights : array_like, optional\n",
      "            The weights for each value in `a`. Default is None, which gives each\n",
      "            value a weight of 1.0\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        cumcount : ndarray\n",
      "            Binned values of cumulative frequency.\n",
      "        lowerlimit : float\n",
      "            Lower real limit\n",
      "        binsize : float\n",
      "            Width of each bin.\n",
      "        extrapoints : int\n",
      "            Extra points.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> import matplotlib.pyplot as plt\n",
      "        >>> from numpy.random import default_rng\n",
      "        >>> from scipy import stats\n",
      "        >>> rng = default_rng()\n",
      "        >>> x = [1, 4, 2, 1, 3, 1]\n",
      "        >>> res = stats.cumfreq(x, numbins=4, defaultreallimits=(1.5, 5))\n",
      "        >>> res.cumcount\n",
      "        array([ 1.,  2.,  3.,  3.])\n",
      "        >>> res.extrapoints\n",
      "        3\n",
      "        \n",
      "        Create a normal distribution with 1000 random values\n",
      "        \n",
      "        >>> samples = stats.norm.rvs(size=1000, random_state=rng)\n",
      "        \n",
      "        Calculate cumulative frequencies\n",
      "        \n",
      "        >>> res = stats.cumfreq(samples, numbins=25)\n",
      "        \n",
      "        Calculate space of values for x\n",
      "        \n",
      "        >>> x = res.lowerlimit + np.linspace(0, res.binsize*res.cumcount.size,\n",
      "        ...                                  res.cumcount.size)\n",
      "        \n",
      "        Plot histogram and cumulative histogram\n",
      "        \n",
      "        >>> fig = plt.figure(figsize=(10, 4))\n",
      "        >>> ax1 = fig.add_subplot(1, 2, 1)\n",
      "        >>> ax2 = fig.add_subplot(1, 2, 2)\n",
      "        >>> ax1.hist(samples, bins=25)\n",
      "        >>> ax1.set_title('Histogram')\n",
      "        >>> ax2.bar(x, res.cumcount, width=res.binsize)\n",
      "        >>> ax2.set_title('Cumulative histogram')\n",
      "        >>> ax2.set_xlim([x.min(), x.max()])\n",
      "        \n",
      "        >>> plt.show()\n",
      "    \n",
      "    describe(a, axis=0, ddof=1, bias=True, nan_policy='propagate')\n",
      "        Compute several descriptive statistics of the passed array.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        a : array_like\n",
      "            Input data.\n",
      "        axis : int or None, optional\n",
      "            Axis along which statistics are calculated. Default is 0.\n",
      "            If None, compute over the whole array `a`.\n",
      "        ddof : int, optional\n",
      "            Delta degrees of freedom (only for variance).  Default is 1.\n",
      "        bias : bool, optional\n",
      "            If False, then the skewness and kurtosis calculations are corrected\n",
      "            for statistical bias.\n",
      "        nan_policy : {'propagate', 'raise', 'omit'}, optional\n",
      "            Defines how to handle when input contains nan.\n",
      "            The following options are available (default is 'propagate'):\n",
      "        \n",
      "            * 'propagate': returns nan\n",
      "            * 'raise': throws an error\n",
      "            * 'omit': performs the calculations ignoring nan values\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        nobs : int or ndarray of ints\n",
      "            Number of observations (length of data along `axis`).\n",
      "            When 'omit' is chosen as nan_policy, the length along each axis\n",
      "            slice is counted separately.\n",
      "        minmax: tuple of ndarrays or floats\n",
      "            Minimum and maximum value of `a` along the given axis.\n",
      "        mean : ndarray or float\n",
      "            Arithmetic mean of `a` along the given axis.\n",
      "        variance : ndarray or float\n",
      "            Unbiased variance of `a` along the given axis; denominator is number\n",
      "            of observations minus one.\n",
      "        skewness : ndarray or float\n",
      "            Skewness of `a` along the given axis, based on moment calculations\n",
      "            with denominator equal to the number of observations, i.e. no degrees\n",
      "            of freedom correction.\n",
      "        kurtosis : ndarray or float\n",
      "            Kurtosis (Fisher) of `a` along the given axis.  The kurtosis is\n",
      "            normalized so that it is zero for the normal distribution.  No\n",
      "            degrees of freedom are used.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        skew, kurtosis\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy import stats\n",
      "        >>> a = np.arange(10)\n",
      "        >>> stats.describe(a)\n",
      "        DescribeResult(nobs=10, minmax=(0, 9), mean=4.5,\n",
      "                       variance=9.166666666666666, skewness=0.0,\n",
      "                       kurtosis=-1.2242424242424244)\n",
      "        >>> b = [[1, 2], [3, 4]]\n",
      "        >>> stats.describe(b)\n",
      "        DescribeResult(nobs=2, minmax=(array([1, 2]), array([3, 4])),\n",
      "                       mean=array([2., 3.]), variance=array([2., 2.]),\n",
      "                       skewness=array([0., 0.]), kurtosis=array([-2., -2.]))\n",
      "    \n",
      "    differential_entropy(values: 'np.typing.ArrayLike', *, window_length: 'Optional[int]' = None, base: 'Optional[float]' = None, axis: 'int' = 0, method: 'str' = 'auto') -> 'Union[np.number, np.ndarray]'\n",
      "        Given a sample of a distribution, estimate the differential entropy.\n",
      "        \n",
      "        Several estimation methods are available using the `method` parameter. By\n",
      "        default, a method is selected based the size of the sample.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        values : sequence\n",
      "            Sample from a continuous distribution.\n",
      "        window_length : int, optional\n",
      "            Window length for computing Vasicek estimate. Must be an integer\n",
      "            between 1 and half of the sample size. If ``None`` (the default), it\n",
      "            uses the heuristic value\n",
      "        \n",
      "            .. math::\n",
      "                \\left \\lfloor \\sqrt{n} + 0.5 \\right \\rfloor\n",
      "        \n",
      "            where :math:`n` is the sample size. This heuristic was originally\n",
      "            proposed in [2]_ and has become common in the literature.\n",
      "        base : float, optional\n",
      "            The logarithmic base to use, defaults to ``e`` (natural logarithm).\n",
      "        axis : int, optional\n",
      "            The axis along which the differential entropy is calculated.\n",
      "            Default is 0.\n",
      "        method : {'vasicek', 'van es', 'ebrahimi', 'correa', 'auto'}, optional\n",
      "            The method used to estimate the differential entropy from the sample.\n",
      "            Default is ``'auto'``.  See Notes for more information.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        entropy : float\n",
      "            The calculated differential entropy.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        This function will converge to the true differential entropy in the limit\n",
      "        \n",
      "        .. math::\n",
      "            n \\to \\infty, \\quad m \\to \\infty, \\quad \\frac{m}{n} \\to 0\n",
      "        \n",
      "        The optimal choice of ``window_length`` for a given sample size depends on\n",
      "        the (unknown) distribution. Typically, the smoother the density of the\n",
      "        distribution, the larger the optimal value of ``window_length`` [1]_.\n",
      "        \n",
      "        The following options are available for the `method` parameter.\n",
      "        \n",
      "        * ``'vasicek'`` uses the estimator presented in [1]_. This is\n",
      "          one of the first and most influential estimators of differential entropy.\n",
      "        * ``'van es'`` uses the bias-corrected estimator presented in [3]_, which\n",
      "          is not only consistent but, under some conditions, asymptotically normal.\n",
      "        * ``'ebrahimi'`` uses an estimator presented in [4]_, which was shown\n",
      "          in simulation to have smaller bias and mean squared error than\n",
      "          the Vasicek estimator.\n",
      "        * ``'correa'`` uses the estimator presented in [5]_ based on local linear\n",
      "          regression. In a simulation study, it had consistently smaller mean\n",
      "          square error than the Vasiceck estimator, but it is more expensive to\n",
      "          compute.\n",
      "        * ``'auto'`` selects the method automatically (default). Currently,\n",
      "          this selects ``'van es'`` for very small samples (<10), ``'ebrahimi'``\n",
      "          for moderate sample sizes (11-1000), and ``'vasicek'`` for larger\n",
      "          samples, but this behavior is subject to change in future versions.\n",
      "        \n",
      "        All estimators are implemented as described in [6]_.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] Vasicek, O. (1976). A test for normality based on sample entropy.\n",
      "               Journal of the Royal Statistical Society:\n",
      "               Series B (Methodological), 38(1), 54-59.\n",
      "        .. [2] Crzcgorzewski, P., & Wirczorkowski, R. (1999). Entropy-based\n",
      "               goodness-of-fit test for exponentiality. Communications in\n",
      "               Statistics-Theory and Methods, 28(5), 1183-1202.\n",
      "        .. [3] Van Es, B. (1992). Estimating functionals related to a density by a\n",
      "               class of statistics based on spacings. Scandinavian Journal of\n",
      "               Statistics, 61-72.\n",
      "        .. [4] Ebrahimi, N., Pflughoeft, K., & Soofi, E. S. (1994). Two measures\n",
      "               of sample entropy. Statistics & Probability Letters, 20(3), 225-234.\n",
      "        .. [5] Correa, J. C. (1995). A new estimator of entropy. Communications\n",
      "               in Statistics-Theory and Methods, 24(10), 2439-2449.\n",
      "        .. [6] Noughabi, H. A. (2015). Entropy Estimation Using Numerical Methods.\n",
      "               Annals of Data Science, 2(2), 231-241.\n",
      "               https://link.springer.com/article/10.1007/s40745-015-0045-9\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy.stats import differential_entropy, norm\n",
      "        \n",
      "        Entropy of a standard normal distribution:\n",
      "        \n",
      "        >>> rng = np.random.default_rng()\n",
      "        >>> values = rng.standard_normal(100)\n",
      "        >>> differential_entropy(values)\n",
      "        1.3407817436640392\n",
      "        \n",
      "        Compare with the true entropy:\n",
      "        \n",
      "        >>> float(norm.entropy())\n",
      "        1.4189385332046727\n",
      "        \n",
      "        For several sample sizes between 5 and 1000, compare the accuracy of\n",
      "        the ``'vasicek'``, ``'van es'``, and ``'ebrahimi'`` methods. Specifically,\n",
      "        compare the root mean squared error (over 1000 trials) between the estimate\n",
      "        and the true differential entropy of the distribution.\n",
      "        \n",
      "        >>> from scipy import stats\n",
      "        >>> import matplotlib.pyplot as plt\n",
      "        >>>\n",
      "        >>>\n",
      "        >>> def rmse(res, expected):\n",
      "        ...     '''Root mean squared error'''\n",
      "        ...     return np.sqrt(np.mean((res - expected)**2))\n",
      "        >>>\n",
      "        >>>\n",
      "        >>> a, b = np.log10(5), np.log10(1000)\n",
      "        >>> ns = np.round(np.logspace(a, b, 10)).astype(int)\n",
      "        >>> reps = 1000  # number of repetitions for each sample size\n",
      "        >>> expected = stats.expon.entropy()\n",
      "        >>>\n",
      "        >>> method_errors = {'vasicek': [], 'van es': [], 'ebrahimi': []}\n",
      "        >>> for method in method_errors:\n",
      "        ...     for n in ns:\n",
      "        ...        rvs = stats.expon.rvs(size=(reps, n), random_state=rng)\n",
      "        ...        res = stats.differential_entropy(rvs, method=method, axis=-1)\n",
      "        ...        error = rmse(res, expected)\n",
      "        ...        method_errors[method].append(error)\n",
      "        >>>\n",
      "        >>> for method, errors in method_errors.items():\n",
      "        ...     plt.loglog(ns, errors, label=method)\n",
      "        >>>\n",
      "        >>> plt.legend()\n",
      "        >>> plt.xlabel('sample size')\n",
      "        >>> plt.ylabel('RMSE (1000 trials)')\n",
      "        >>> plt.title('Entropy Estimator Error (Exponential Distribution)')\n",
      "    \n",
      "    energy_distance(u_values, v_values, u_weights=None, v_weights=None)\n",
      "        Compute the energy distance between two 1D distributions.\n",
      "        \n",
      "        .. versionadded:: 1.0.0\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        u_values, v_values : array_like\n",
      "            Values observed in the (empirical) distribution.\n",
      "        u_weights, v_weights : array_like, optional\n",
      "            Weight for each value. If unspecified, each value is assigned the same\n",
      "            weight.\n",
      "            `u_weights` (resp. `v_weights`) must have the same length as\n",
      "            `u_values` (resp. `v_values`). If the weight sum differs from 1, it\n",
      "            must still be positive and finite so that the weights can be normalized\n",
      "            to sum to 1.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        distance : float\n",
      "            The computed distance between the distributions.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        The energy distance between two distributions :math:`u` and :math:`v`, whose\n",
      "        respective CDFs are :math:`U` and :math:`V`, equals to:\n",
      "        \n",
      "        .. math::\n",
      "        \n",
      "            D(u, v) = \\left( 2\\mathbb E|X - Y| - \\mathbb E|X - X'| -\n",
      "            \\mathbb E|Y - Y'| \\right)^{1/2}\n",
      "        \n",
      "        where :math:`X` and :math:`X'` (resp. :math:`Y` and :math:`Y'`) are\n",
      "        independent random variables whose probability distribution is :math:`u`\n",
      "        (resp. :math:`v`).\n",
      "        \n",
      "        As shown in [2]_, for one-dimensional real-valued variables, the energy\n",
      "        distance is linked to the non-distribution-free version of the Cramér-von\n",
      "        Mises distance:\n",
      "        \n",
      "        .. math::\n",
      "        \n",
      "            D(u, v) = \\sqrt{2} l_2(u, v) = \\left( 2 \\int_{-\\infty}^{+\\infty} (U-V)^2\n",
      "            \\right)^{1/2}\n",
      "        \n",
      "        Note that the common Cramér-von Mises criterion uses the distribution-free\n",
      "        version of the distance. See [2]_ (section 2), for more details about both\n",
      "        versions of the distance.\n",
      "        \n",
      "        The input distributions can be empirical, therefore coming from samples\n",
      "        whose values are effectively inputs of the function, or they can be seen as\n",
      "        generalized functions, in which case they are weighted sums of Dirac delta\n",
      "        functions located at the specified values.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] \"Energy distance\", https://en.wikipedia.org/wiki/Energy_distance\n",
      "        .. [2] Szekely \"E-statistics: The energy of statistical samples.\" Bowling\n",
      "               Green State University, Department of Mathematics and Statistics,\n",
      "               Technical Report 02-16 (2002).\n",
      "        .. [3] Rizzo, Szekely \"Energy distance.\" Wiley Interdisciplinary Reviews:\n",
      "               Computational Statistics, 8(1):27-38 (2015).\n",
      "        .. [4] Bellemare, Danihelka, Dabney, Mohamed, Lakshminarayanan, Hoyer,\n",
      "               Munos \"The Cramer Distance as a Solution to Biased Wasserstein\n",
      "               Gradients\" (2017). :arXiv:`1705.10743`.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy.stats import energy_distance\n",
      "        >>> energy_distance([0], [2])\n",
      "        2.0000000000000004\n",
      "        >>> energy_distance([0, 8], [0, 8], [3, 1], [2, 2])\n",
      "        1.0000000000000002\n",
      "        >>> energy_distance([0.7, 7.4, 2.4, 6.8], [1.4, 8. ],\n",
      "        ...                 [2.1, 4.2, 7.4, 8. ], [7.6, 8.8])\n",
      "        0.88003340976158217\n",
      "    \n",
      "    entropy(pk, qk=None, base=None, axis=0)\n",
      "        Calculate the entropy of a distribution for given probability values.\n",
      "        \n",
      "        If only probabilities `pk` are given, the entropy is calculated as\n",
      "        ``S = -sum(pk * log(pk), axis=axis)``.\n",
      "        \n",
      "        If `qk` is not None, then compute the Kullback-Leibler divergence\n",
      "        ``S = sum(pk * log(pk / qk), axis=axis)``.\n",
      "        \n",
      "        This routine will normalize `pk` and `qk` if they don't sum to 1.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        pk : sequence\n",
      "            Defines the (discrete) distribution. ``pk[i]`` is the (possibly\n",
      "            unnormalized) probability of event ``i``.\n",
      "        qk : sequence, optional\n",
      "            Sequence against which the relative entropy is computed. Should be in\n",
      "            the same format as `pk`.\n",
      "        base : float, optional\n",
      "            The logarithmic base to use, defaults to ``e`` (natural logarithm).\n",
      "        axis: int, optional\n",
      "            The axis along which the entropy is calculated. Default is 0.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        S : float\n",
      "            The calculated entropy.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        \n",
      "        >>> from scipy.stats import entropy\n",
      "        \n",
      "        Bernoulli trial with different p.\n",
      "        The outcome of a fair coin is the most uncertain:\n",
      "        \n",
      "        >>> entropy([1/2, 1/2], base=2)\n",
      "        1.0\n",
      "        \n",
      "        The outcome of a biased coin is less uncertain:\n",
      "        \n",
      "        >>> entropy([9/10, 1/10], base=2)\n",
      "        0.46899559358928117\n",
      "        \n",
      "        Relative entropy:\n",
      "        \n",
      "        >>> entropy([1/2, 1/2], qk=[9/10, 1/10])\n",
      "        0.5108256237659907\n",
      "    \n",
      "    epps_singleton_2samp(x, y, t=(0.4, 0.8))\n",
      "        Compute the Epps-Singleton (ES) test statistic.\n",
      "        \n",
      "        Test the null hypothesis that two samples have the same underlying\n",
      "        probability distribution.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x, y : array-like\n",
      "            The two samples of observations to be tested. Input must not have more\n",
      "            than one dimension. Samples can have different lengths.\n",
      "        t : array-like, optional\n",
      "            The points (t1, ..., tn) where the empirical characteristic function is\n",
      "            to be evaluated. It should be positive distinct numbers. The default\n",
      "            value (0.4, 0.8) is proposed in [1]_. Input must not have more than\n",
      "            one dimension.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        statistic : float\n",
      "            The test statistic.\n",
      "        pvalue : float\n",
      "            The associated p-value based on the asymptotic chi2-distribution.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        ks_2samp, anderson_ksamp\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        Testing whether two samples are generated by the same underlying\n",
      "        distribution is a classical question in statistics. A widely used test is\n",
      "        the Kolmogorov-Smirnov (KS) test which relies on the empirical\n",
      "        distribution function. Epps and Singleton introduce a test based on the\n",
      "        empirical characteristic function in [1]_.\n",
      "        \n",
      "        One advantage of the ES test compared to the KS test is that is does\n",
      "        not assume a continuous distribution. In [1]_, the authors conclude\n",
      "        that the test also has a higher power than the KS test in many\n",
      "        examples. They recommend the use of the ES test for discrete samples as\n",
      "        well as continuous samples with at least 25 observations each, whereas\n",
      "        `anderson_ksamp` is recommended for smaller sample sizes in the\n",
      "        continuous case.\n",
      "        \n",
      "        The p-value is computed from the asymptotic distribution of the test\n",
      "        statistic which follows a `chi2` distribution. If the sample size of both\n",
      "        `x` and `y` is below 25, the small sample correction proposed in [1]_ is\n",
      "        applied to the test statistic.\n",
      "        \n",
      "        The default values of `t` are determined in [1]_ by considering\n",
      "        various distributions and finding good values that lead to a high power\n",
      "        of the test in general. Table III in [1]_ gives the optimal values for\n",
      "        the distributions tested in that study. The values of `t` are scaled by\n",
      "        the semi-interquartile range in the implementation, see [1]_.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] T. W. Epps and K. J. Singleton, \"An omnibus test for the two-sample\n",
      "           problem using the empirical characteristic function\", Journal of\n",
      "           Statistical Computation and Simulation 26, p. 177--203, 1986.\n",
      "        \n",
      "        .. [2] S. J. Goerg and J. Kaiser, \"Nonparametric testing of distributions\n",
      "           - the Epps-Singleton two-sample test using the empirical characteristic\n",
      "           function\", The Stata Journal 9(3), p. 454--465, 2009.\n",
      "    \n",
      "    f_oneway(*args, axis=0)\n",
      "        Perform one-way ANOVA.\n",
      "        \n",
      "        The one-way ANOVA tests the null hypothesis that two or more groups have\n",
      "        the same population mean.  The test is applied to samples from two or\n",
      "        more groups, possibly with differing sizes.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        sample1, sample2, ... : array_like\n",
      "            The sample measurements for each group.  There must be at least\n",
      "            two arguments.  If the arrays are multidimensional, then all the\n",
      "            dimensions of the array must be the same except for `axis`.\n",
      "        axis : int, optional\n",
      "            Axis of the input arrays along which the test is applied.\n",
      "            Default is 0.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        statistic : float\n",
      "            The computed F statistic of the test.\n",
      "        pvalue : float\n",
      "            The associated p-value from the F distribution.\n",
      "        \n",
      "        Warns\n",
      "        -----\n",
      "        F_onewayConstantInputWarning\n",
      "            Raised if each of the input arrays is constant array.\n",
      "            In this case the F statistic is either infinite or isn't defined,\n",
      "            so ``np.inf`` or ``np.nan`` is returned.\n",
      "        \n",
      "        F_onewayBadInputSizesWarning\n",
      "            Raised if the length of any input array is 0, or if all the input\n",
      "            arrays have length 1.  ``np.nan`` is returned for the F statistic\n",
      "            and the p-value in these cases.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        The ANOVA test has important assumptions that must be satisfied in order\n",
      "        for the associated p-value to be valid.\n",
      "        \n",
      "        1. The samples are independent.\n",
      "        2. Each sample is from a normally distributed population.\n",
      "        3. The population standard deviations of the groups are all equal.  This\n",
      "           property is known as homoscedasticity.\n",
      "        \n",
      "        If these assumptions are not true for a given set of data, it may still\n",
      "        be possible to use the Kruskal-Wallis H-test (`scipy.stats.kruskal`) or\n",
      "        the Alexander-Govern test (`scipy.stats.alexandergovern`) although with\n",
      "        some loss of power.\n",
      "        \n",
      "        The length of each group must be at least one, and there must be at\n",
      "        least one group with length greater than one.  If these conditions\n",
      "        are not satisfied, a warning is generated and (``np.nan``, ``np.nan``)\n",
      "        is returned.\n",
      "        \n",
      "        If each group contains constant values, and there exist at least two\n",
      "        groups with different values, the function generates a warning and\n",
      "        returns (``np.inf``, 0).\n",
      "        \n",
      "        If all values in all groups are the same, function generates a warning\n",
      "        and returns (``np.nan``, ``np.nan``).\n",
      "        \n",
      "        The algorithm is from Heiman [2]_, pp.394-7.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] R. Lowry, \"Concepts and Applications of Inferential Statistics\",\n",
      "               Chapter 14, 2014, http://vassarstats.net/textbook/\n",
      "        \n",
      "        .. [2] G.W. Heiman, \"Understanding research methods and statistics: An\n",
      "               integrated introduction for psychology\", Houghton, Mifflin and\n",
      "               Company, 2001.\n",
      "        \n",
      "        .. [3] G.H. McDonald, \"Handbook of Biological Statistics\", One-way ANOVA.\n",
      "               http://www.biostathandbook.com/onewayanova.html\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy.stats import f_oneway\n",
      "        \n",
      "        Here are some data [3]_ on a shell measurement (the length of the anterior\n",
      "        adductor muscle scar, standardized by dividing by length) in the mussel\n",
      "        Mytilus trossulus from five locations: Tillamook, Oregon; Newport, Oregon;\n",
      "        Petersburg, Alaska; Magadan, Russia; and Tvarminne, Finland, taken from a\n",
      "        much larger data set used in McDonald et al. (1991).\n",
      "        \n",
      "        >>> tillamook = [0.0571, 0.0813, 0.0831, 0.0976, 0.0817, 0.0859, 0.0735,\n",
      "        ...              0.0659, 0.0923, 0.0836]\n",
      "        >>> newport = [0.0873, 0.0662, 0.0672, 0.0819, 0.0749, 0.0649, 0.0835,\n",
      "        ...            0.0725]\n",
      "        >>> petersburg = [0.0974, 0.1352, 0.0817, 0.1016, 0.0968, 0.1064, 0.105]\n",
      "        >>> magadan = [0.1033, 0.0915, 0.0781, 0.0685, 0.0677, 0.0697, 0.0764,\n",
      "        ...            0.0689]\n",
      "        >>> tvarminne = [0.0703, 0.1026, 0.0956, 0.0973, 0.1039, 0.1045]\n",
      "        >>> f_oneway(tillamook, newport, petersburg, magadan, tvarminne)\n",
      "        F_onewayResult(statistic=7.121019471642447, pvalue=0.0002812242314534544)\n",
      "        \n",
      "        `f_oneway` accepts multidimensional input arrays.  When the inputs\n",
      "        are multidimensional and `axis` is not given, the test is performed\n",
      "        along the first axis of the input arrays.  For the following data, the\n",
      "        test is performed three times, once for each column.\n",
      "        \n",
      "        >>> a = np.array([[9.87, 9.03, 6.81],\n",
      "        ...               [7.18, 8.35, 7.00],\n",
      "        ...               [8.39, 7.58, 7.68],\n",
      "        ...               [7.45, 6.33, 9.35],\n",
      "        ...               [6.41, 7.10, 9.33],\n",
      "        ...               [8.00, 8.24, 8.44]])\n",
      "        >>> b = np.array([[6.35, 7.30, 7.16],\n",
      "        ...               [6.65, 6.68, 7.63],\n",
      "        ...               [5.72, 7.73, 6.72],\n",
      "        ...               [7.01, 9.19, 7.41],\n",
      "        ...               [7.75, 7.87, 8.30],\n",
      "        ...               [6.90, 7.97, 6.97]])\n",
      "        >>> c = np.array([[3.31, 8.77, 1.01],\n",
      "        ...               [8.25, 3.24, 3.62],\n",
      "        ...               [6.32, 8.81, 5.19],\n",
      "        ...               [7.48, 8.83, 8.91],\n",
      "        ...               [8.59, 6.01, 6.07],\n",
      "        ...               [3.07, 9.72, 7.48]])\n",
      "        >>> F, p = f_oneway(a, b, c)\n",
      "        >>> F\n",
      "        array([1.75676344, 0.03701228, 3.76439349])\n",
      "        >>> p\n",
      "        array([0.20630784, 0.96375203, 0.04733157])\n",
      "    \n",
      "    find_repeats(arr)\n",
      "        Find repeats and repeat counts.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        arr : array_like\n",
      "            Input array. This is cast to float64.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        values : ndarray\n",
      "            The unique values from the (flattened) input that are repeated.\n",
      "        \n",
      "        counts : ndarray\n",
      "            Number of times the corresponding 'value' is repeated.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        In numpy >= 1.9 `numpy.unique` provides similar functionality. The main\n",
      "        difference is that `find_repeats` only returns repeated values.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy import stats\n",
      "        >>> stats.find_repeats([2, 1, 2, 3, 2, 2, 5])\n",
      "        RepeatedResults(values=array([2.]), counts=array([4]))\n",
      "        \n",
      "        >>> stats.find_repeats([[10, 20, 1, 2], [5, 5, 4, 4]])\n",
      "        RepeatedResults(values=array([4.,  5.]), counts=array([2, 2]))\n",
      "    \n",
      "    fisher_exact(table, alternative='two-sided')\n",
      "        Perform a Fisher exact test on a 2x2 contingency table.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        table : array_like of ints\n",
      "            A 2x2 contingency table.  Elements must be non-negative integers.\n",
      "        alternative : {'two-sided', 'less', 'greater'}, optional\n",
      "            Defines the alternative hypothesis.\n",
      "            The following options are available (default is 'two-sided'):\n",
      "        \n",
      "            * 'two-sided'\n",
      "            * 'less': one-sided\n",
      "            * 'greater': one-sided\n",
      "        \n",
      "            See the Notes for more details.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        oddsratio : float\n",
      "            This is prior odds ratio and not a posterior estimate.\n",
      "        p_value : float\n",
      "            P-value, the probability of obtaining a distribution at least as\n",
      "            extreme as the one that was actually observed, assuming that the\n",
      "            null hypothesis is true.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        chi2_contingency : Chi-square test of independence of variables in a\n",
      "            contingency table.  This can be used as an alternative to\n",
      "            `fisher_exact` when the numbers in the table are large.\n",
      "        barnard_exact : Barnard's exact test, which is a more powerful alternative\n",
      "            than Fisher's exact test for 2x2 contingency tables.\n",
      "        boschloo_exact : Boschloo's exact test, which is a more powerful alternative\n",
      "            than Fisher's exact test for 2x2 contingency tables.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        *Null hypothesis and p-values*\n",
      "        \n",
      "        The null hypothesis is that the input table is from the hypergeometric\n",
      "        distribution with parameters (as used in `hypergeom`)\n",
      "        ``M = a + b + c + d``, ``n = a + b`` and ``N = a + c``, where the\n",
      "        input table is ``[[a, b], [c, d]]``.  This distribution has support\n",
      "        ``max(0, N + n - M) <= x <= min(N, n)``, or, in terms of the values\n",
      "        in the input table, ``min(0, a - d) <= x <= a + min(b, c)``.  ``x``\n",
      "        can be interpreted as the upper-left element of a 2x2 table, so the\n",
      "        tables in the distribution have form::\n",
      "        \n",
      "            [  x           n - x     ]\n",
      "            [N - x    M - (n + N) + x]\n",
      "        \n",
      "        For example, if::\n",
      "        \n",
      "            table = [6  2]\n",
      "                    [1  4]\n",
      "        \n",
      "        then the support is ``2 <= x <= 7``, and the tables in the distribution\n",
      "        are::\n",
      "        \n",
      "            [2 6]   [3 5]   [4 4]   [5 3]   [6 2]  [7 1]\n",
      "            [5 0]   [4 1]   [3 2]   [2 3]   [1 4]  [0 5]\n",
      "        \n",
      "        The probability of each table is given by the hypergeometric distribution\n",
      "        ``hypergeom.pmf(x, M, n, N)``.  For this example, these are (rounded to\n",
      "        three significant digits)::\n",
      "        \n",
      "            x       2      3      4      5       6        7\n",
      "            p  0.0163  0.163  0.408  0.326  0.0816  0.00466\n",
      "        \n",
      "        These can be computed with::\n",
      "        \n",
      "            >>> from scipy.stats import hypergeom\n",
      "            >>> table = np.array([[6, 2], [1, 4]])\n",
      "            >>> M = table.sum()\n",
      "            >>> n = table[0].sum()\n",
      "            >>> N = table[:, 0].sum()\n",
      "            >>> start, end = hypergeom.support(M, n, N)\n",
      "            >>> hypergeom.pmf(np.arange(start, end+1), M, n, N)\n",
      "            array([0.01631702, 0.16317016, 0.40792541, 0.32634033, 0.08158508,\n",
      "                   0.004662  ])\n",
      "        \n",
      "        The two-sided p-value is the probability that, under the null hypothesis,\n",
      "        a random table would have a probability equal to or less than the\n",
      "        probability of the input table.  For our example, the probability of\n",
      "        the input table (where ``x = 6``) is 0.0816.  The x values where the\n",
      "        probability does not exceed this are 2, 6 and 7, so the two-sided p-value\n",
      "        is ``0.0163 + 0.0816 + 0.00466 ~= 0.10256``::\n",
      "        \n",
      "            >>> from scipy.stats import fisher_exact\n",
      "            >>> oddsr, p = fisher_exact(table, alternative='two-sided')\n",
      "            >>> p\n",
      "            0.10256410256410257\n",
      "        \n",
      "        The one-sided p-value for ``alternative='greater'`` is the probability\n",
      "        that a random table has ``x >= a``, which in our example is ``x >= 6``,\n",
      "        or ``0.0816 + 0.00466 ~= 0.08626``::\n",
      "        \n",
      "            >>> oddsr, p = fisher_exact(table, alternative='greater')\n",
      "            >>> p\n",
      "            0.08624708624708627\n",
      "        \n",
      "        This is equivalent to computing the survival function of the\n",
      "        distribution at ``x = 5`` (one less than ``x`` from the input table,\n",
      "        because we want to include the probability of ``x = 6`` in the sum)::\n",
      "        \n",
      "            >>> hypergeom.sf(5, M, n, N)\n",
      "            0.08624708624708627\n",
      "        \n",
      "        For ``alternative='less'``, the one-sided p-value is the probability\n",
      "        that a random table has ``x <= a``, (i.e. ``x <= 6`` in our example),\n",
      "        or ``0.0163 + 0.163 + 0.408 + 0.326 + 0.0816 ~= 0.9949``::\n",
      "        \n",
      "            >>> oddsr, p = fisher_exact(table, alternative='less')\n",
      "            >>> p\n",
      "            0.9953379953379957\n",
      "        \n",
      "        This is equivalent to computing the cumulative distribution function\n",
      "        of the distribution at ``x = 6``:\n",
      "        \n",
      "            >>> hypergeom.cdf(6, M, n, N)\n",
      "            0.9953379953379957\n",
      "        \n",
      "        *Odds ratio*\n",
      "        \n",
      "        The calculated odds ratio is different from the one R uses. This SciPy\n",
      "        implementation returns the (more common) \"unconditional Maximum\n",
      "        Likelihood Estimate\", while R uses the \"conditional Maximum Likelihood\n",
      "        Estimate\".\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        Say we spend a few days counting whales and sharks in the Atlantic and\n",
      "        Indian oceans. In the Atlantic ocean we find 8 whales and 1 shark, in the\n",
      "        Indian ocean 2 whales and 5 sharks. Then our contingency table is::\n",
      "        \n",
      "                    Atlantic  Indian\n",
      "            whales     8        2\n",
      "            sharks     1        5\n",
      "        \n",
      "        We use this table to find the p-value:\n",
      "        \n",
      "        >>> from scipy.stats import fisher_exact\n",
      "        >>> oddsratio, pvalue = fisher_exact([[8, 2], [1, 5]])\n",
      "        >>> pvalue\n",
      "        0.0349...\n",
      "        \n",
      "        The probability that we would observe this or an even more imbalanced ratio\n",
      "        by chance is about 3.5%.  A commonly used significance level is 5%--if we\n",
      "        adopt that, we can therefore conclude that our observed imbalance is\n",
      "        statistically significant; whales prefer the Atlantic while sharks prefer\n",
      "        the Indian ocean.\n",
      "    \n",
      "    fligner(*args, center='median', proportiontocut=0.05)\n",
      "        Perform Fligner-Killeen test for equality of variance.\n",
      "        \n",
      "        Fligner's test tests the null hypothesis that all input samples\n",
      "        are from populations with equal variances.  Fligner-Killeen's test is\n",
      "        distribution free when populations are identical [2]_.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        sample1, sample2, ... : array_like\n",
      "            Arrays of sample data.  Need not be the same length.\n",
      "        center : {'mean', 'median', 'trimmed'}, optional\n",
      "            Keyword argument controlling which function of the data is used in\n",
      "            computing the test statistic.  The default is 'median'.\n",
      "        proportiontocut : float, optional\n",
      "            When `center` is 'trimmed', this gives the proportion of data points\n",
      "            to cut from each end. (See `scipy.stats.trim_mean`.)\n",
      "            Default is 0.05.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        statistic : float\n",
      "            The test statistic.\n",
      "        pvalue : float\n",
      "            The p-value for the hypothesis test.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        bartlett : A parametric test for equality of k variances in normal samples\n",
      "        levene : A robust parametric test for equality of k variances\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        As with Levene's test there are three variants of Fligner's test that\n",
      "        differ by the measure of central tendency used in the test.  See `levene`\n",
      "        for more information.\n",
      "        \n",
      "        Conover et al. (1981) examine many of the existing parametric and\n",
      "        nonparametric tests by extensive simulations and they conclude that the\n",
      "        tests proposed by Fligner and Killeen (1976) and Levene (1960) appear to be\n",
      "        superior in terms of robustness of departures from normality and power [3]_.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] Park, C. and Lindsay, B. G. (1999). Robust Scale Estimation and\n",
      "               Hypothesis Testing based on Quadratic Inference Function. Technical\n",
      "               Report #99-03, Center for Likelihood Studies, Pennsylvania State\n",
      "               University.\n",
      "               https://cecas.clemson.edu/~cspark/cv/paper/qif/draftqif2.pdf\n",
      "        \n",
      "        .. [2] Fligner, M.A. and Killeen, T.J. (1976). Distribution-free two-sample\n",
      "               tests for scale. 'Journal of the American Statistical Association.'\n",
      "               71(353), 210-213.\n",
      "        \n",
      "        .. [3] Park, C. and Lindsay, B. G. (1999). Robust Scale Estimation and\n",
      "               Hypothesis Testing based on Quadratic Inference Function. Technical\n",
      "               Report #99-03, Center for Likelihood Studies, Pennsylvania State\n",
      "               University.\n",
      "        \n",
      "        .. [4] Conover, W. J., Johnson, M. E. and Johnson M. M. (1981). A\n",
      "               comparative study of tests for homogeneity of variances, with\n",
      "               applications to the outer continental shelf biding data.\n",
      "               Technometrics, 23(4), 351-361.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        Test whether or not the lists `a`, `b` and `c` come from populations\n",
      "        with equal variances.\n",
      "        \n",
      "        >>> from scipy.stats import fligner\n",
      "        >>> a = [8.88, 9.12, 9.04, 8.98, 9.00, 9.08, 9.01, 8.85, 9.06, 8.99]\n",
      "        >>> b = [8.88, 8.95, 9.29, 9.44, 9.15, 9.58, 8.36, 9.18, 8.67, 9.05]\n",
      "        >>> c = [8.95, 9.12, 8.95, 8.85, 9.03, 8.84, 9.07, 8.98, 8.86, 8.98]\n",
      "        >>> stat, p = fligner(a, b, c)\n",
      "        >>> p\n",
      "        0.00450826080004775\n",
      "        \n",
      "        The small p-value suggests that the populations do not have equal\n",
      "        variances.\n",
      "        \n",
      "        This is not surprising, given that the sample variance of `b` is much\n",
      "        larger than that of `a` and `c`:\n",
      "        \n",
      "        >>> [np.var(x, ddof=1) for x in [a, b, c]]\n",
      "        [0.007054444444444413, 0.13073888888888888, 0.008890000000000002]\n",
      "    \n",
      "    friedmanchisquare(*args)\n",
      "        Compute the Friedman test for repeated measurements.\n",
      "        \n",
      "        The Friedman test tests the null hypothesis that repeated measurements of\n",
      "        the same individuals have the same distribution.  It is often used\n",
      "        to test for consistency among measurements obtained in different ways.\n",
      "        For example, if two measurement techniques are used on the same set of\n",
      "        individuals, the Friedman test can be used to determine if the two\n",
      "        measurement techniques are consistent.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        measurements1, measurements2, measurements3... : array_like\n",
      "            Arrays of measurements.  All of the arrays must have the same number\n",
      "            of elements.  At least 3 sets of measurements must be given.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        statistic : float\n",
      "            The test statistic, correcting for ties.\n",
      "        pvalue : float\n",
      "            The associated p-value assuming that the test statistic has a chi\n",
      "            squared distribution.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        Due to the assumption that the test statistic has a chi squared\n",
      "        distribution, the p-value is only reliable for n > 10 and more than\n",
      "        6 repeated measurements.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] https://en.wikipedia.org/wiki/Friedman_test\n",
      "    \n",
      "    gmean(a, axis=0, dtype=None, weights=None)\n",
      "        Compute the geometric mean along the specified axis.\n",
      "        \n",
      "        Return the geometric average of the array elements.\n",
      "        That is:  n-th root of (x1 * x2 * ... * xn)\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        a : array_like\n",
      "            Input array or object that can be converted to an array.\n",
      "        axis : int or None, optional\n",
      "            Axis along which the geometric mean is computed. Default is 0.\n",
      "            If None, compute over the whole array `a`.\n",
      "        dtype : dtype, optional\n",
      "            Type of the returned array and of the accumulator in which the\n",
      "            elements are summed. If dtype is not specified, it defaults to the\n",
      "            dtype of a, unless a has an integer dtype with a precision less than\n",
      "            that of the default platform integer. In that case, the default\n",
      "            platform integer is used.\n",
      "        weights : array_like, optional\n",
      "            The weights array can either be 1-D (in which case its length must be\n",
      "            the size of `a` along the given `axis`) or of the same shape as `a`.\n",
      "            Default is None, which gives each value a weight of 1.0.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        gmean : ndarray\n",
      "            See `dtype` parameter above.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        numpy.mean : Arithmetic average\n",
      "        numpy.average : Weighted average\n",
      "        hmean : Harmonic mean\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        The geometric average is computed over a single dimension of the input\n",
      "        array, axis=0 by default, or all values in the array if axis=None.\n",
      "        float64 intermediate and return values are used for integer inputs.\n",
      "        \n",
      "        Use masked arrays to ignore any non-finite values in the input or that\n",
      "        arise in the calculations such as Not a Number and infinity because masked\n",
      "        arrays automatically mask any non-finite values.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] \"Weighted Geometric Mean\", *Wikipedia*, https://en.wikipedia.org/wiki/Weighted_geometric_mean.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy.stats import gmean\n",
      "        >>> gmean([1, 4])\n",
      "        2.0\n",
      "        >>> gmean([1, 2, 3, 4, 5, 6, 7])\n",
      "        3.3800151591412964\n",
      "    \n",
      "    gstd(a, axis=0, ddof=1)\n",
      "        Calculate the geometric standard deviation of an array.\n",
      "        \n",
      "        The geometric standard deviation describes the spread of a set of numbers\n",
      "        where the geometric mean is preferred. It is a multiplicative factor, and\n",
      "        so a dimensionless quantity.\n",
      "        \n",
      "        It is defined as the exponent of the standard deviation of ``log(a)``.\n",
      "        Mathematically the population geometric standard deviation can be\n",
      "        evaluated as::\n",
      "        \n",
      "            gstd = exp(std(log(a)))\n",
      "        \n",
      "        .. versionadded:: 1.3.0\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        a : array_like\n",
      "            An array like object containing the sample data.\n",
      "        axis : int, tuple or None, optional\n",
      "            Axis along which to operate. Default is 0. If None, compute over\n",
      "            the whole array `a`.\n",
      "        ddof : int, optional\n",
      "            Degree of freedom correction in the calculation of the\n",
      "            geometric standard deviation. Default is 1.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        ndarray or float\n",
      "            An array of the geometric standard deviation. If `axis` is None or `a`\n",
      "            is a 1d array a float is returned.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        As the calculation requires the use of logarithms the geometric standard\n",
      "        deviation only supports strictly positive values. Any non-positive or\n",
      "        infinite values will raise a `ValueError`.\n",
      "        The geometric standard deviation is sometimes confused with the exponent of\n",
      "        the standard deviation, ``exp(std(a))``. Instead the geometric standard\n",
      "        deviation is ``exp(std(log(a)))``.\n",
      "        The default value for `ddof` is different to the default value (0) used\n",
      "        by other ddof containing functions, such as ``np.std`` and ``np.nanstd``.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        Find the geometric standard deviation of a log-normally distributed sample.\n",
      "        Note that the standard deviation of the distribution is one, on a\n",
      "        log scale this evaluates to approximately ``exp(1)``.\n",
      "        \n",
      "        >>> from scipy.stats import gstd\n",
      "        >>> rng = np.random.default_rng()\n",
      "        >>> sample = rng.lognormal(mean=0, sigma=1, size=1000)\n",
      "        >>> gstd(sample)\n",
      "        2.810010162475324\n",
      "        \n",
      "        Compute the geometric standard deviation of a multidimensional array and\n",
      "        of a given axis.\n",
      "        \n",
      "        >>> a = np.arange(1, 25).reshape(2, 3, 4)\n",
      "        >>> gstd(a, axis=None)\n",
      "        2.2944076136018947\n",
      "        >>> gstd(a, axis=2)\n",
      "        array([[1.82424757, 1.22436866, 1.13183117],\n",
      "               [1.09348306, 1.07244798, 1.05914985]])\n",
      "        >>> gstd(a, axis=(1,2))\n",
      "        array([2.12939215, 1.22120169])\n",
      "        \n",
      "        The geometric standard deviation further handles masked arrays.\n",
      "        \n",
      "        >>> a = np.arange(1, 25).reshape(2, 3, 4)\n",
      "        >>> ma = np.ma.masked_where(a > 16, a)\n",
      "        >>> ma\n",
      "        masked_array(\n",
      "          data=[[[1, 2, 3, 4],\n",
      "                 [5, 6, 7, 8],\n",
      "                 [9, 10, 11, 12]],\n",
      "                [[13, 14, 15, 16],\n",
      "                 [--, --, --, --],\n",
      "                 [--, --, --, --]]],\n",
      "          mask=[[[False, False, False, False],\n",
      "                 [False, False, False, False],\n",
      "                 [False, False, False, False]],\n",
      "                [[False, False, False, False],\n",
      "                 [ True,  True,  True,  True],\n",
      "                 [ True,  True,  True,  True]]],\n",
      "          fill_value=999999)\n",
      "        >>> gstd(ma, axis=2)\n",
      "        masked_array(\n",
      "          data=[[1.8242475707663655, 1.2243686572447428, 1.1318311657788478],\n",
      "                [1.0934830582350938, --, --]],\n",
      "          mask=[[False, False, False],\n",
      "                [False,  True,  True]],\n",
      "          fill_value=999999)\n",
      "    \n",
      "    hmean(a, axis=0, dtype=None)\n",
      "        Calculate the harmonic mean along the specified axis.\n",
      "        \n",
      "        That is:  n / (1/x1 + 1/x2 + ... + 1/xn)\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        a : array_like\n",
      "            Input array, masked array or object that can be converted to an array.\n",
      "        axis : int or None, optional\n",
      "            Axis along which the harmonic mean is computed. Default is 0.\n",
      "            If None, compute over the whole array `a`.\n",
      "        dtype : dtype, optional\n",
      "            Type of the returned array and of the accumulator in which the\n",
      "            elements are summed. If `dtype` is not specified, it defaults to the\n",
      "            dtype of `a`, unless `a` has an integer `dtype` with a precision less\n",
      "            than that of the default platform integer. In that case, the default\n",
      "            platform integer is used.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        hmean : ndarray\n",
      "            See `dtype` parameter above.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        numpy.mean : Arithmetic average\n",
      "        numpy.average : Weighted average\n",
      "        gmean : Geometric mean\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        The harmonic mean is computed over a single dimension of the input\n",
      "        array, axis=0 by default, or all values in the array if axis=None.\n",
      "        float64 intermediate and return values are used for integer inputs.\n",
      "        \n",
      "        Use masked arrays to ignore any non-finite values in the input or that\n",
      "        arise in the calculations such as Not a Number and infinity.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy.stats import hmean\n",
      "        >>> hmean([1, 4])\n",
      "        1.6000000000000001\n",
      "        >>> hmean([1, 2, 3, 4, 5, 6, 7])\n",
      "        2.6997245179063363\n",
      "    \n",
      "    iqr(x, axis=None, rng=(25, 75), scale=1.0, nan_policy='propagate', interpolation='linear', keepdims=False)\n",
      "        Compute the interquartile range of the data along the specified axis.\n",
      "        \n",
      "        The interquartile range (IQR) is the difference between the 75th and\n",
      "        25th percentile of the data. It is a measure of the dispersion\n",
      "        similar to standard deviation or variance, but is much more robust\n",
      "        against outliers [2]_.\n",
      "        \n",
      "        The ``rng`` parameter allows this function to compute other\n",
      "        percentile ranges than the actual IQR. For example, setting\n",
      "        ``rng=(0, 100)`` is equivalent to `numpy.ptp`.\n",
      "        \n",
      "        The IQR of an empty array is `np.nan`.\n",
      "        \n",
      "        .. versionadded:: 0.18.0\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x : array_like\n",
      "            Input array or object that can be converted to an array.\n",
      "        axis : int or sequence of int, optional\n",
      "            Axis along which the range is computed. The default is to\n",
      "            compute the IQR for the entire array.\n",
      "        rng : Two-element sequence containing floats in range of [0,100] optional\n",
      "            Percentiles over which to compute the range. Each must be\n",
      "            between 0 and 100, inclusive. The default is the true IQR:\n",
      "            `(25, 75)`. The order of the elements is not important.\n",
      "        scale : scalar or str, optional\n",
      "            The numerical value of scale will be divided out of the final\n",
      "            result. The following string values are recognized:\n",
      "        \n",
      "              * 'raw' : No scaling, just return the raw IQR.\n",
      "                **Deprecated!**  Use `scale=1` instead.\n",
      "              * 'normal' : Scale by\n",
      "                :math:`2 \\sqrt{2} erf^{-1}(\\frac{1}{2}) \\approx 1.349`.\n",
      "        \n",
      "            The default is 1.0. The use of scale='raw' is deprecated.\n",
      "            Array-like scale is also allowed, as long\n",
      "            as it broadcasts correctly to the output such that\n",
      "            ``out / scale`` is a valid operation. The output dimensions\n",
      "            depend on the input array, `x`, the `axis` argument, and the\n",
      "            `keepdims` flag.\n",
      "        nan_policy : {'propagate', 'raise', 'omit'}, optional\n",
      "            Defines how to handle when input contains nan.\n",
      "            The following options are available (default is 'propagate'):\n",
      "        \n",
      "              * 'propagate': returns nan\n",
      "              * 'raise': throws an error\n",
      "              * 'omit': performs the calculations ignoring nan values\n",
      "        interpolation : {'linear', 'lower', 'higher', 'midpoint',\n",
      "                         'nearest'}, optional\n",
      "        \n",
      "            Specifies the interpolation method to use when the percentile\n",
      "            boundaries lie between two data points `i` and `j`.\n",
      "            The following options are available (default is 'linear'):\n",
      "        \n",
      "              * 'linear': `i + (j - i) * fraction`, where `fraction` is the\n",
      "                fractional part of the index surrounded by `i` and `j`.\n",
      "              * 'lower': `i`.\n",
      "              * 'higher': `j`.\n",
      "              * 'nearest': `i` or `j` whichever is nearest.\n",
      "              * 'midpoint': `(i + j) / 2`.\n",
      "        \n",
      "        keepdims : bool, optional\n",
      "            If this is set to `True`, the reduced axes are left in the\n",
      "            result as dimensions with size one. With this option, the result\n",
      "            will broadcast correctly against the original array `x`.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        iqr : scalar or ndarray\n",
      "            If ``axis=None``, a scalar is returned. If the input contains\n",
      "            integers or floats of smaller precision than ``np.float64``, then the\n",
      "            output data-type is ``np.float64``. Otherwise, the output data-type is\n",
      "            the same as that of the input.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        numpy.std, numpy.var\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        This function is heavily dependent on the version of `numpy` that is\n",
      "        installed. Versions greater than 1.11.0b3 are highly recommended, as they\n",
      "        include a number of enhancements and fixes to `numpy.percentile` and\n",
      "        `numpy.nanpercentile` that affect the operation of this function. The\n",
      "        following modifications apply:\n",
      "        \n",
      "        Below 1.10.0 : `nan_policy` is poorly defined.\n",
      "            The default behavior of `numpy.percentile` is used for 'propagate'. This\n",
      "            is a hybrid of 'omit' and 'propagate' that mostly yields a skewed\n",
      "            version of 'omit' since NaNs are sorted to the end of the data. A\n",
      "            warning is raised if there are NaNs in the data.\n",
      "        Below 1.9.0: `numpy.nanpercentile` does not exist.\n",
      "            This means that `numpy.percentile` is used regardless of `nan_policy`\n",
      "            and a warning is issued. See previous item for a description of the\n",
      "            behavior.\n",
      "        Below 1.9.0: `keepdims` and `interpolation` are not supported.\n",
      "            The keywords get ignored with a warning if supplied with non-default\n",
      "            values. However, multiple axes are still supported.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] \"Interquartile range\" https://en.wikipedia.org/wiki/Interquartile_range\n",
      "        .. [2] \"Robust measures of scale\" https://en.wikipedia.org/wiki/Robust_measures_of_scale\n",
      "        .. [3] \"Quantile\" https://en.wikipedia.org/wiki/Quantile\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy.stats import iqr\n",
      "        >>> x = np.array([[10, 7, 4], [3, 2, 1]])\n",
      "        >>> x\n",
      "        array([[10,  7,  4],\n",
      "               [ 3,  2,  1]])\n",
      "        >>> iqr(x)\n",
      "        4.0\n",
      "        >>> iqr(x, axis=0)\n",
      "        array([ 3.5,  2.5,  1.5])\n",
      "        >>> iqr(x, axis=1)\n",
      "        array([ 3.,  1.])\n",
      "        >>> iqr(x, axis=1, keepdims=True)\n",
      "        array([[ 3.],\n",
      "               [ 1.]])\n",
      "    \n",
      "    itemfreq(*args, **kwds)\n",
      "        `itemfreq` is deprecated!\n",
      "        `itemfreq` is deprecated and will be removed in a future version. Use instead `np.unique(..., return_counts=True)`\n",
      "        \n",
      "        Return a 2-D array of item frequencies.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        a : (N,) array_like\n",
      "            Input array.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        itemfreq : (K, 2) ndarray\n",
      "            A 2-D frequency table.  Column 1 contains sorted, unique values from\n",
      "            `a`, column 2 contains their respective counts.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy import stats\n",
      "        >>> a = np.array([1, 1, 5, 0, 1, 2, 2, 0, 1, 4])\n",
      "        >>> stats.itemfreq(a)\n",
      "        array([[ 0.,  2.],\n",
      "               [ 1.,  4.],\n",
      "               [ 2.,  2.],\n",
      "               [ 4.,  1.],\n",
      "               [ 5.,  1.]])\n",
      "        >>> np.bincount(a)\n",
      "        array([2, 4, 2, 0, 1, 1])\n",
      "        \n",
      "        >>> stats.itemfreq(a/10.)\n",
      "        array([[ 0. ,  2. ],\n",
      "               [ 0.1,  4. ],\n",
      "               [ 0.2,  2. ],\n",
      "               [ 0.4,  1. ],\n",
      "               [ 0.5,  1. ]])\n",
      "    \n",
      "    jarque_bera(x)\n",
      "        Perform the Jarque-Bera goodness of fit test on sample data.\n",
      "        \n",
      "        The Jarque-Bera test tests whether the sample data has the skewness and\n",
      "        kurtosis matching a normal distribution.\n",
      "        \n",
      "        Note that this test only works for a large enough number of data samples\n",
      "        (>2000) as the test statistic asymptotically has a Chi-squared distribution\n",
      "        with 2 degrees of freedom.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x : array_like\n",
      "            Observations of a random variable.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        jb_value : float\n",
      "            The test statistic.\n",
      "        p : float\n",
      "            The p-value for the hypothesis test.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] Jarque, C. and Bera, A. (1980) \"Efficient tests for normality,\n",
      "               homoscedasticity and serial independence of regression residuals\",\n",
      "               6 Econometric Letters 255-259.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy import stats\n",
      "        >>> rng = np.random.default_rng()\n",
      "        >>> x = rng.normal(0, 1, 100000)\n",
      "        >>> jarque_bera_test = stats.jarque_bera(x)\n",
      "        >>> jarque_bera_test\n",
      "        Jarque_beraResult(statistic=3.3415184718131554, pvalue=0.18810419594996775)\n",
      "        >>> jarque_bera_test.statistic\n",
      "        3.3415184718131554\n",
      "        >>> jarque_bera_test.pvalue\n",
      "        0.18810419594996775\n",
      "    \n",
      "    kendalltau(x, y, initial_lexsort=None, nan_policy='propagate', method='auto', variant='b')\n",
      "        Calculate Kendall's tau, a correlation measure for ordinal data.\n",
      "        \n",
      "        Kendall's tau is a measure of the correspondence between two rankings.\n",
      "        Values close to 1 indicate strong agreement, and values close to -1\n",
      "        indicate strong disagreement. This implements two variants of Kendall's\n",
      "        tau: tau-b (the default) and tau-c (also known as Stuart's tau-c). These\n",
      "        differ only in how they are normalized to lie within the range -1 to 1;\n",
      "        the hypothesis tests (their p-values) are identical. Kendall's original\n",
      "        tau-a is not implemented separately because both tau-b and tau-c reduce\n",
      "        to tau-a in the absence of ties.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x, y : array_like\n",
      "            Arrays of rankings, of the same shape. If arrays are not 1-D, they\n",
      "            will be flattened to 1-D.\n",
      "        initial_lexsort : bool, optional\n",
      "            Unused (deprecated).\n",
      "        nan_policy : {'propagate', 'raise', 'omit'}, optional\n",
      "            Defines how to handle when input contains nan.\n",
      "            The following options are available (default is 'propagate'):\n",
      "        \n",
      "              * 'propagate': returns nan\n",
      "              * 'raise': throws an error\n",
      "              * 'omit': performs the calculations ignoring nan values\n",
      "        \n",
      "        method : {'auto', 'asymptotic', 'exact'}, optional\n",
      "            Defines which method is used to calculate the p-value [5]_.\n",
      "            The following options are available (default is 'auto'):\n",
      "        \n",
      "              * 'auto': selects the appropriate method based on a trade-off\n",
      "                between speed and accuracy\n",
      "              * 'asymptotic': uses a normal approximation valid for large samples\n",
      "              * 'exact': computes the exact p-value, but can only be used if no ties\n",
      "                are present. As the sample size increases, the 'exact' computation\n",
      "                time may grow and the result may lose some precision.\n",
      "        \n",
      "        variant: {'b', 'c'}, optional\n",
      "            Defines which variant of Kendall's tau is returned. Default is 'b'.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        correlation : float\n",
      "           The tau statistic.\n",
      "        pvalue : float\n",
      "           The two-sided p-value for a hypothesis test whose null hypothesis is\n",
      "           an absence of association, tau = 0.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        spearmanr : Calculates a Spearman rank-order correlation coefficient.\n",
      "        theilslopes : Computes the Theil-Sen estimator for a set of points (x, y).\n",
      "        weightedtau : Computes a weighted version of Kendall's tau.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        The definition of Kendall's tau that is used is [2]_::\n",
      "        \n",
      "          tau_b = (P - Q) / sqrt((P + Q + T) * (P + Q + U))\n",
      "        \n",
      "          tau_c = 2 (P - Q) / (n**2 * (m - 1) / m)\n",
      "        \n",
      "        where P is the number of concordant pairs, Q the number of discordant\n",
      "        pairs, T the number of ties only in `x`, and U the number of ties only in\n",
      "        `y`.  If a tie occurs for the same pair in both `x` and `y`, it is not\n",
      "        added to either T or U. n is the total number of samples, and m is the\n",
      "        number of unique values in either `x` or `y`, whichever is smaller.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] Maurice G. Kendall, \"A New Measure of Rank Correlation\", Biometrika\n",
      "               Vol. 30, No. 1/2, pp. 81-93, 1938.\n",
      "        .. [2] Maurice G. Kendall, \"The treatment of ties in ranking problems\",\n",
      "               Biometrika Vol. 33, No. 3, pp. 239-251. 1945.\n",
      "        .. [3] Gottfried E. Noether, \"Elements of Nonparametric Statistics\", John\n",
      "               Wiley & Sons, 1967.\n",
      "        .. [4] Peter M. Fenwick, \"A new data structure for cumulative frequency\n",
      "               tables\", Software: Practice and Experience, Vol. 24, No. 3,\n",
      "               pp. 327-336, 1994.\n",
      "        .. [5] Maurice G. Kendall, \"Rank Correlation Methods\" (4th Edition),\n",
      "               Charles Griffin & Co., 1970.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy import stats\n",
      "        >>> x1 = [12, 2, 1, 12, 2]\n",
      "        >>> x2 = [1, 4, 7, 1, 0]\n",
      "        >>> tau, p_value = stats.kendalltau(x1, x2)\n",
      "        >>> tau\n",
      "        -0.47140452079103173\n",
      "        >>> p_value\n",
      "        0.2827454599327748\n",
      "    \n",
      "    kruskal(*args, nan_policy='propagate')\n",
      "        Compute the Kruskal-Wallis H-test for independent samples.\n",
      "        \n",
      "        The Kruskal-Wallis H-test tests the null hypothesis that the population\n",
      "        median of all of the groups are equal.  It is a non-parametric version of\n",
      "        ANOVA.  The test works on 2 or more independent samples, which may have\n",
      "        different sizes.  Note that rejecting the null hypothesis does not\n",
      "        indicate which of the groups differs.  Post hoc comparisons between\n",
      "        groups are required to determine which groups are different.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        sample1, sample2, ... : array_like\n",
      "           Two or more arrays with the sample measurements can be given as\n",
      "           arguments. Samples must be one-dimensional.\n",
      "        nan_policy : {'propagate', 'raise', 'omit'}, optional\n",
      "            Defines how to handle when input contains nan.\n",
      "            The following options are available (default is 'propagate'):\n",
      "        \n",
      "              * 'propagate': returns nan\n",
      "              * 'raise': throws an error\n",
      "              * 'omit': performs the calculations ignoring nan values\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        statistic : float\n",
      "           The Kruskal-Wallis H statistic, corrected for ties.\n",
      "        pvalue : float\n",
      "           The p-value for the test using the assumption that H has a chi\n",
      "           square distribution. The p-value returned is the survival function of\n",
      "           the chi square distribution evaluated at H.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        f_oneway : 1-way ANOVA.\n",
      "        mannwhitneyu : Mann-Whitney rank test on two samples.\n",
      "        friedmanchisquare : Friedman test for repeated measurements.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        Due to the assumption that H has a chi square distribution, the number\n",
      "        of samples in each group must not be too small.  A typical rule is\n",
      "        that each sample must have at least 5 measurements.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] W. H. Kruskal & W. W. Wallis, \"Use of Ranks in\n",
      "           One-Criterion Variance Analysis\", Journal of the American Statistical\n",
      "           Association, Vol. 47, Issue 260, pp. 583-621, 1952.\n",
      "        .. [2] https://en.wikipedia.org/wiki/Kruskal-Wallis_one-way_analysis_of_variance\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy import stats\n",
      "        >>> x = [1, 3, 5, 7, 9]\n",
      "        >>> y = [2, 4, 6, 8, 10]\n",
      "        >>> stats.kruskal(x, y)\n",
      "        KruskalResult(statistic=0.2727272727272734, pvalue=0.6015081344405895)\n",
      "        \n",
      "        >>> x = [1, 1, 1]\n",
      "        >>> y = [2, 2, 2]\n",
      "        >>> z = [2, 2]\n",
      "        >>> stats.kruskal(x, y, z)\n",
      "        KruskalResult(statistic=7.0, pvalue=0.0301973834223185)\n",
      "    \n",
      "    ks_1samp(x, cdf, args=(), alternative='two-sided', mode='auto')\n",
      "        Performs the one-sample Kolmogorov-Smirnov test for goodness of fit.\n",
      "        \n",
      "        This test compares the underlying distribution F(x) of a sample\n",
      "        against a given continuous distribution G(x). See Notes for a description\n",
      "        of the available null and alternative hypotheses.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x : array_like\n",
      "            a 1-D array of observations of iid random variables.\n",
      "        cdf : callable\n",
      "            callable used to calculate the cdf.\n",
      "        args : tuple, sequence, optional\n",
      "            Distribution parameters, used with `cdf`.\n",
      "        alternative : {'two-sided', 'less', 'greater'}, optional\n",
      "            Defines the null and alternative hypotheses. Default is 'two-sided'.\n",
      "            Please see explanations in the Notes below.\n",
      "        mode : {'auto', 'exact', 'approx', 'asymp'}, optional\n",
      "            Defines the distribution used for calculating the p-value.\n",
      "            The following options are available (default is 'auto'):\n",
      "        \n",
      "              * 'auto' : selects one of the other options.\n",
      "              * 'exact' : uses the exact distribution of test statistic.\n",
      "              * 'approx' : approximates the two-sided probability with twice\n",
      "                the one-sided probability\n",
      "              * 'asymp': uses asymptotic distribution of test statistic\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        statistic : float\n",
      "            KS test statistic, either D, D+ or D- (depending on the value\n",
      "            of 'alternative')\n",
      "        pvalue :  float\n",
      "            One-tailed or two-tailed p-value.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        ks_2samp, kstest\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        There are three options for the null and corresponding alternative\n",
      "        hypothesis that can be selected using the `alternative` parameter.\n",
      "        \n",
      "        - `two-sided`: The null hypothesis is that the two distributions are\n",
      "          identical, F(x)=G(x) for all x; the alternative is that they are not\n",
      "          identical.\n",
      "        \n",
      "        - `less`: The null hypothesis is that F(x) >= G(x) for all x; the\n",
      "          alternative is that F(x) < G(x) for at least one x.\n",
      "        \n",
      "        - `greater`: The null hypothesis is that F(x) <= G(x) for all x; the\n",
      "          alternative is that F(x) > G(x) for at least one x.\n",
      "        \n",
      "        Note that the alternative hypotheses describe the *CDFs* of the\n",
      "        underlying distributions, not the observed values. For example,\n",
      "        suppose x1 ~ F and x2 ~ G. If F(x) > G(x) for all x, the values in\n",
      "        x1 tend to be less than those in x2.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy import stats\n",
      "        >>> rng = np.random.default_rng()\n",
      "        \n",
      "        >>> x = np.linspace(-15, 15, 9)\n",
      "        >>> stats.ks_1samp(x, stats.norm.cdf)\n",
      "        (0.44435602715924361, 0.038850142705171065)\n",
      "        \n",
      "        >>> stats.ks_1samp(stats.norm.rvs(size=100, random_state=rng),\n",
      "        ...                stats.norm.cdf)\n",
      "        KstestResult(statistic=0.165471391799..., pvalue=0.007331283245...)\n",
      "        \n",
      "        *Test against one-sided alternative hypothesis*\n",
      "        \n",
      "        Shift distribution to larger values, so that `` CDF(x) < norm.cdf(x)``:\n",
      "        \n",
      "        >>> x = stats.norm.rvs(loc=0.2, size=100, random_state=rng)\n",
      "        >>> stats.ks_1samp(x, stats.norm.cdf, alternative='less')\n",
      "        KstestResult(statistic=0.100203351482..., pvalue=0.125544644447...)\n",
      "        \n",
      "        Reject null hypothesis in favor of alternative hypothesis: less\n",
      "        \n",
      "        >>> stats.ks_1samp(x, stats.norm.cdf, alternative='greater')\n",
      "        KstestResult(statistic=0.018749806388..., pvalue=0.920581859791...)\n",
      "        \n",
      "        Reject null hypothesis in favor of alternative hypothesis: greater\n",
      "        \n",
      "        >>> stats.ks_1samp(x, stats.norm.cdf)\n",
      "        KstestResult(statistic=0.100203351482..., pvalue=0.250616879765...)\n",
      "        \n",
      "        Don't reject null hypothesis in favor of alternative hypothesis: two-sided\n",
      "        \n",
      "        *Testing t distributed random variables against normal distribution*\n",
      "        \n",
      "        With 100 degrees of freedom the t distribution looks close to the normal\n",
      "        distribution, and the K-S test does not reject the hypothesis that the\n",
      "        sample came from the normal distribution:\n",
      "        \n",
      "        >>> stats.ks_1samp(stats.t.rvs(100,size=100, random_state=rng),\n",
      "        ...                stats.norm.cdf)\n",
      "        KstestResult(statistic=0.064273776544..., pvalue=0.778737758305...)\n",
      "        \n",
      "        With 3 degrees of freedom the t distribution looks sufficiently different\n",
      "        from the normal distribution, that we can reject the hypothesis that the\n",
      "        sample came from the normal distribution at the 10% level:\n",
      "        \n",
      "        >>> stats.ks_1samp(stats.t.rvs(3,size=100, random_state=rng),\n",
      "        ...                stats.norm.cdf)\n",
      "        KstestResult(statistic=0.128678487493..., pvalue=0.066569081515...)\n",
      "    \n",
      "    ks_2samp(data1, data2, alternative='two-sided', mode='auto')\n",
      "        Performs the two-sample Kolmogorov-Smirnov test for goodness of fit.\n",
      "        \n",
      "        This test compares the underlying continuous distributions F(x) and G(x)\n",
      "        of two independent samples.  See Notes for a description\n",
      "        of the available null and alternative hypotheses.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        data1, data2 : array_like, 1-Dimensional\n",
      "            Two arrays of sample observations assumed to be drawn from a continuous\n",
      "            distribution, sample sizes can be different.\n",
      "        alternative : {'two-sided', 'less', 'greater'}, optional\n",
      "            Defines the null and alternative hypotheses. Default is 'two-sided'.\n",
      "            Please see explanations in the Notes below.\n",
      "        mode : {'auto', 'exact', 'asymp'}, optional\n",
      "            Defines the method used for calculating the p-value.\n",
      "            The following options are available (default is 'auto'):\n",
      "        \n",
      "              * 'auto' : use 'exact' for small size arrays, 'asymp' for large\n",
      "              * 'exact' : use exact distribution of test statistic\n",
      "              * 'asymp' : use asymptotic distribution of test statistic\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        statistic : float\n",
      "            KS statistic.\n",
      "        pvalue : float\n",
      "            One-tailed or two-tailed p-value.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        kstest, ks_1samp, epps_singleton_2samp, anderson_ksamp\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        There are three options for the null and corresponding alternative\n",
      "        hypothesis that can be selected using the `alternative` parameter.\n",
      "        \n",
      "        - `two-sided`: The null hypothesis is that the two distributions are\n",
      "          identical, F(x)=G(x) for all x; the alternative is that they are not\n",
      "          identical.\n",
      "        \n",
      "        - `less`: The null hypothesis is that F(x) >= G(x) for all x; the\n",
      "          alternative is that F(x) < G(x) for at least one x.\n",
      "        \n",
      "        - `greater`: The null hypothesis is that F(x) <= G(x) for all x; the\n",
      "          alternative is that F(x) > G(x) for at least one x.\n",
      "        \n",
      "        Note that the alternative hypotheses describe the *CDFs* of the\n",
      "        underlying distributions, not the observed values. For example,\n",
      "        suppose x1 ~ F and x2 ~ G. If F(x) > G(x) for all x, the values in\n",
      "        x1 tend to be less than those in x2.\n",
      "        \n",
      "        \n",
      "        If the KS statistic is small or the p-value is high, then we cannot\n",
      "        reject the null hypothesis in favor of the alternative.\n",
      "        \n",
      "        If the mode is 'auto', the computation is exact if the sample sizes are\n",
      "        less than 10000.  For larger sizes, the computation uses the\n",
      "        Kolmogorov-Smirnov distributions to compute an approximate value.\n",
      "        \n",
      "        The 'two-sided' 'exact' computation computes the complementary probability\n",
      "        and then subtracts from 1.  As such, the minimum probability it can return\n",
      "        is about 1e-16.  While the algorithm itself is exact, numerical\n",
      "        errors may accumulate for large sample sizes.   It is most suited to\n",
      "        situations in which one of the sample sizes is only a few thousand.\n",
      "        \n",
      "        We generally follow Hodges' treatment of Drion/Gnedenko/Korolyuk [1]_.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] Hodges, J.L. Jr.,  \"The Significance Probability of the Smirnov\n",
      "               Two-Sample Test,\" Arkiv fiur Matematik, 3, No. 43 (1958), 469-86.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy import stats\n",
      "        >>> rng = np.random.default_rng()\n",
      "        \n",
      "        >>> n1 = 200  # size of first sample\n",
      "        >>> n2 = 300  # size of second sample\n",
      "        \n",
      "        For a different distribution, we can reject the null hypothesis since the\n",
      "        pvalue is below 1%:\n",
      "        \n",
      "        >>> rvs1 = stats.norm.rvs(size=n1, loc=0., scale=1, random_state=rng)\n",
      "        >>> rvs2 = stats.norm.rvs(size=n2, loc=0.5, scale=1.5, random_state=rng)\n",
      "        >>> stats.ks_2samp(rvs1, rvs2)\n",
      "         KstestResult(statistic=0.24833333333333332, pvalue=5.846586728086578e-07)\n",
      "        \n",
      "        For a slightly different distribution, we cannot reject the null hypothesis\n",
      "        at a 10% or lower alpha since the p-value at 0.144 is higher than 10%\n",
      "        \n",
      "        >>> rvs3 = stats.norm.rvs(size=n2, loc=0.01, scale=1.0, random_state=rng)\n",
      "        >>> stats.ks_2samp(rvs1, rvs3)\n",
      "        KstestResult(statistic=0.07833333333333334, pvalue=0.4379658456442945)\n",
      "        \n",
      "        For an identical distribution, we cannot reject the null hypothesis since\n",
      "        the p-value is high, 41%:\n",
      "        \n",
      "        >>> rvs4 = stats.norm.rvs(size=n2, loc=0.0, scale=1.0, random_state=rng)\n",
      "        >>> stats.ks_2samp(rvs1, rvs4)\n",
      "        KstestResult(statistic=0.12166666666666667, pvalue=0.05401863039081145)\n",
      "    \n",
      "    kstat(data, n=2)\n",
      "        Return the nth k-statistic (1<=n<=4 so far).\n",
      "        \n",
      "        The nth k-statistic k_n is the unique symmetric unbiased estimator of the\n",
      "        nth cumulant kappa_n.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        data : array_like\n",
      "            Input array. Note that n-D input gets flattened.\n",
      "        n : int, {1, 2, 3, 4}, optional\n",
      "            Default is equal to 2.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        kstat : float\n",
      "            The nth k-statistic.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        kstatvar: Returns an unbiased estimator of the variance of the k-statistic.\n",
      "        moment: Returns the n-th central moment about the mean for a sample.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        For a sample size n, the first few k-statistics are given by:\n",
      "        \n",
      "        .. math::\n",
      "        \n",
      "            k_{1} = \\mu\n",
      "            k_{2} = \\frac{n}{n-1} m_{2}\n",
      "            k_{3} = \\frac{ n^{2} } {(n-1) (n-2)} m_{3}\n",
      "            k_{4} = \\frac{ n^{2} [(n + 1)m_{4} - 3(n - 1) m^2_{2}]} {(n-1) (n-2) (n-3)}\n",
      "        \n",
      "        where :math:`\\mu` is the sample mean, :math:`m_2` is the sample\n",
      "        variance, and :math:`m_i` is the i-th sample central moment.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        http://mathworld.wolfram.com/k-Statistic.html\n",
      "        \n",
      "        http://mathworld.wolfram.com/Cumulant.html\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy import stats\n",
      "        >>> from numpy.random import default_rng\n",
      "        >>> rng = default_rng()\n",
      "        \n",
      "        As sample size increases, n-th moment and n-th k-statistic converge to the\n",
      "        same number (although they aren't identical). In the case of the normal\n",
      "        distribution, they converge to zero.\n",
      "        \n",
      "        >>> for n in [2, 3, 4, 5, 6, 7]:\n",
      "        ...     x = rng.normal(size=10**n)\n",
      "        ...     m, k = stats.moment(x, 3), stats.kstat(x, 3)\n",
      "        ...     print(\"%.3g %.3g %.3g\" % (m, k, m-k))\n",
      "        -0.631 -0.651 0.0194  # random\n",
      "        0.0282 0.0283 -8.49e-05\n",
      "        -0.0454 -0.0454 1.36e-05\n",
      "        7.53e-05 7.53e-05 -2.26e-09\n",
      "        0.00166 0.00166 -4.99e-09\n",
      "        -2.88e-06 -2.88e-06 8.63e-13\n",
      "    \n",
      "    kstatvar(data, n=2)\n",
      "        Return an unbiased estimator of the variance of the k-statistic.\n",
      "        \n",
      "        See `kstat` for more details of the k-statistic.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        data : array_like\n",
      "            Input array. Note that n-D input gets flattened.\n",
      "        n : int, {1, 2}, optional\n",
      "            Default is equal to 2.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        kstatvar : float\n",
      "            The nth k-statistic variance.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        kstat: Returns the n-th k-statistic.\n",
      "        moment: Returns the n-th central moment about the mean for a sample.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        The variances of the first few k-statistics are given by:\n",
      "        \n",
      "        .. math::\n",
      "        \n",
      "            var(k_{1}) = \\frac{\\kappa^2}{n}\n",
      "            var(k_{2}) = \\frac{\\kappa^4}{n} + \\frac{2\\kappa^2_{2}}{n - 1}\n",
      "            var(k_{3}) = \\frac{\\kappa^6}{n} + \\frac{9 \\kappa_2 \\kappa_4}{n - 1} +\n",
      "                         \\frac{9 \\kappa^2_{3}}{n - 1} +\n",
      "                         \\frac{6 n \\kappa^3_{2}}{(n-1) (n-2)}\n",
      "            var(k_{4}) = \\frac{\\kappa^8}{n} + \\frac{16 \\kappa_2 \\kappa_6}{n - 1} +\n",
      "                         \\frac{48 \\kappa_{3} \\kappa_5}{n - 1} +\n",
      "                         \\frac{34 \\kappa^2_{4}}{n-1} + \\frac{72 n \\kappa^2_{2} \\kappa_4}{(n - 1) (n - 2)} +\n",
      "                         \\frac{144 n \\kappa_{2} \\kappa^2_{3}}{(n - 1) (n - 2)} +\n",
      "                         \\frac{24 (n + 1) n \\kappa^4_{2}}{(n - 1) (n - 2) (n - 3)}\n",
      "    \n",
      "    kstest(rvs, cdf, args=(), N=20, alternative='two-sided', mode='auto')\n",
      "        Performs the (one-sample or two-sample) Kolmogorov-Smirnov test for\n",
      "        goodness of fit.\n",
      "        \n",
      "        The one-sample test compares the underlying distribution F(x) of a sample\n",
      "        against a given distribution G(x). The two-sample test compares the\n",
      "        underlying distributions of two independent samples. Both tests are valid\n",
      "        only for continuous distributions.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        rvs : str, array_like, or callable\n",
      "            If an array, it should be a 1-D array of observations of random\n",
      "            variables.\n",
      "            If a callable, it should be a function to generate random variables;\n",
      "            it is required to have a keyword argument `size`.\n",
      "            If a string, it should be the name of a distribution in `scipy.stats`,\n",
      "            which will be used to generate random variables.\n",
      "        cdf : str, array_like or callable\n",
      "            If array_like, it should be a 1-D array of observations of random\n",
      "            variables, and the two-sample test is performed\n",
      "            (and rvs must be array_like).\n",
      "            If a callable, that callable is used to calculate the cdf.\n",
      "            If a string, it should be the name of a distribution in `scipy.stats`,\n",
      "            which will be used as the cdf function.\n",
      "        args : tuple, sequence, optional\n",
      "            Distribution parameters, used if `rvs` or `cdf` are strings or\n",
      "            callables.\n",
      "        N : int, optional\n",
      "            Sample size if `rvs` is string or callable.  Default is 20.\n",
      "        alternative : {'two-sided', 'less', 'greater'}, optional\n",
      "            Defines the null and alternative hypotheses. Default is 'two-sided'.\n",
      "            Please see explanations in the Notes below.\n",
      "        mode : {'auto', 'exact', 'approx', 'asymp'}, optional\n",
      "            Defines the distribution used for calculating the p-value.\n",
      "            The following options are available (default is 'auto'):\n",
      "        \n",
      "              * 'auto' : selects one of the other options.\n",
      "              * 'exact' : uses the exact distribution of test statistic.\n",
      "              * 'approx' : approximates the two-sided probability with twice the\n",
      "                one-sided probability\n",
      "              * 'asymp': uses asymptotic distribution of test statistic\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        statistic : float\n",
      "            KS test statistic, either D, D+ or D-.\n",
      "        pvalue :  float\n",
      "            One-tailed or two-tailed p-value.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        ks_2samp\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        There are three options for the null and corresponding alternative\n",
      "        hypothesis that can be selected using the `alternative` parameter.\n",
      "        \n",
      "        - `two-sided`: The null hypothesis is that the two distributions are\n",
      "          identical, F(x)=G(x) for all x; the alternative is that they are not\n",
      "          identical.\n",
      "        \n",
      "        - `less`: The null hypothesis is that F(x) >= G(x) for all x; the\n",
      "          alternative is that F(x) < G(x) for at least one x.\n",
      "        \n",
      "        - `greater`: The null hypothesis is that F(x) <= G(x) for all x; the\n",
      "          alternative is that F(x) > G(x) for at least one x.\n",
      "        \n",
      "        Note that the alternative hypotheses describe the *CDFs* of the\n",
      "        underlying distributions, not the observed values. For example,\n",
      "        suppose x1 ~ F and x2 ~ G. If F(x) > G(x) for all x, the values in\n",
      "        x1 tend to be less than those in x2.\n",
      "        \n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy import stats\n",
      "        >>> rng = np.random.default_rng()\n",
      "        \n",
      "        >>> x = np.linspace(-15, 15, 9)\n",
      "        >>> stats.kstest(x, 'norm')\n",
      "        KstestResult(statistic=0.444356027159..., pvalue=0.038850140086...)\n",
      "        \n",
      "        >>> stats.kstest(stats.norm.rvs(size=100, random_state=rng), stats.norm.cdf)\n",
      "        KstestResult(statistic=0.165471391799..., pvalue=0.007331283245...)\n",
      "        \n",
      "        The above lines are equivalent to:\n",
      "        \n",
      "        >>> stats.kstest(stats.norm.rvs, 'norm', N=100)\n",
      "        KstestResult(statistic=0.113810164200..., pvalue=0.138690052319...)  # may vary\n",
      "        \n",
      "        *Test against one-sided alternative hypothesis*\n",
      "        \n",
      "        Shift distribution to larger values, so that ``CDF(x) < norm.cdf(x)``:\n",
      "        \n",
      "        >>> x = stats.norm.rvs(loc=0.2, size=100, random_state=rng)\n",
      "        >>> stats.kstest(x, 'norm', alternative='less')\n",
      "        KstestResult(statistic=0.1002033514..., pvalue=0.1255446444...)\n",
      "        \n",
      "        Reject null hypothesis in favor of alternative hypothesis: less\n",
      "        \n",
      "        >>> stats.kstest(x, 'norm', alternative='greater')\n",
      "        KstestResult(statistic=0.018749806388..., pvalue=0.920581859791...)\n",
      "        \n",
      "        Don't reject null hypothesis in favor of alternative hypothesis: greater\n",
      "        \n",
      "        >>> stats.kstest(x, 'norm')\n",
      "        KstestResult(statistic=0.100203351482..., pvalue=0.250616879765...)\n",
      "        \n",
      "        *Testing t distributed random variables against normal distribution*\n",
      "        \n",
      "        With 100 degrees of freedom the t distribution looks close to the normal\n",
      "        distribution, and the K-S test does not reject the hypothesis that the\n",
      "        sample came from the normal distribution:\n",
      "        \n",
      "        >>> stats.kstest(stats.t.rvs(100, size=100, random_state=rng), 'norm')\n",
      "        KstestResult(statistic=0.064273776544..., pvalue=0.778737758305...)\n",
      "        \n",
      "        With 3 degrees of freedom the t distribution looks sufficiently different\n",
      "        from the normal distribution, that we can reject the hypothesis that the\n",
      "        sample came from the normal distribution at the 10% level:\n",
      "        \n",
      "        >>> stats.kstest(stats.t.rvs(3, size=100, random_state=rng), 'norm')\n",
      "        KstestResult(statistic=0.128678487493..., pvalue=0.066569081515...)\n",
      "    \n",
      "    kurtosis(a, axis=0, fisher=True, bias=True, nan_policy='propagate')\n",
      "        Compute the kurtosis (Fisher or Pearson) of a dataset.\n",
      "        \n",
      "        Kurtosis is the fourth central moment divided by the square of the\n",
      "        variance. If Fisher's definition is used, then 3.0 is subtracted from\n",
      "        the result to give 0.0 for a normal distribution.\n",
      "        \n",
      "        If bias is False then the kurtosis is calculated using k statistics to\n",
      "        eliminate bias coming from biased moment estimators\n",
      "        \n",
      "        Use `kurtosistest` to see if result is close enough to normal.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        a : array\n",
      "            Data for which the kurtosis is calculated.\n",
      "        axis : int or None, optional\n",
      "            Axis along which the kurtosis is calculated. Default is 0.\n",
      "            If None, compute over the whole array `a`.\n",
      "        fisher : bool, optional\n",
      "            If True, Fisher's definition is used (normal ==> 0.0). If False,\n",
      "            Pearson's definition is used (normal ==> 3.0).\n",
      "        bias : bool, optional\n",
      "            If False, then the calculations are corrected for statistical bias.\n",
      "        nan_policy : {'propagate', 'raise', 'omit'}, optional\n",
      "            Defines how to handle when input contains nan. 'propagate' returns nan,\n",
      "            'raise' throws an error, 'omit' performs the calculations ignoring nan\n",
      "            values. Default is 'propagate'.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        kurtosis : array\n",
      "            The kurtosis of values along an axis. If all values are equal,\n",
      "            return -3 for Fisher's definition and 0 for Pearson's definition.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] Zwillinger, D. and Kokoska, S. (2000). CRC Standard\n",
      "           Probability and Statistics Tables and Formulae. Chapman & Hall: New\n",
      "           York. 2000.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        In Fisher's definiton, the kurtosis of the normal distribution is zero.\n",
      "        In the following example, the kurtosis is close to zero, because it was\n",
      "        calculated from the dataset, not from the continuous distribution.\n",
      "        \n",
      "        >>> from scipy.stats import norm, kurtosis\n",
      "        >>> data = norm.rvs(size=1000, random_state=3)\n",
      "        >>> kurtosis(data)\n",
      "        -0.06928694200380558\n",
      "        \n",
      "        The distribution with a higher kurtosis has a heavier tail.\n",
      "        The zero valued kurtosis of the normal distribution in Fisher's definition\n",
      "        can serve as a reference point.\n",
      "        \n",
      "        >>> import matplotlib.pyplot as plt\n",
      "        >>> import scipy.stats as stats\n",
      "        >>> from scipy.stats import kurtosis\n",
      "        \n",
      "        >>> x = np.linspace(-5, 5, 100)\n",
      "        >>> ax = plt.subplot()\n",
      "        >>> distnames = ['laplace', 'norm', 'uniform']\n",
      "        \n",
      "        >>> for distname in distnames:\n",
      "        ...     if distname == 'uniform':\n",
      "        ...         dist = getattr(stats, distname)(loc=-2, scale=4)\n",
      "        ...     else:\n",
      "        ...         dist = getattr(stats, distname)\n",
      "        ...     data = dist.rvs(size=1000)\n",
      "        ...     kur = kurtosis(data, fisher=True)\n",
      "        ...     y = dist.pdf(x)\n",
      "        ...     ax.plot(x, y, label=\"{}, {}\".format(distname, round(kur, 3)))\n",
      "        ...     ax.legend()\n",
      "        \n",
      "        The Laplace distribution has a heavier tail than the normal distribution.\n",
      "        The uniform distribution (which has negative kurtosis) has the thinnest\n",
      "        tail.\n",
      "    \n",
      "    kurtosistest(a, axis=0, nan_policy='propagate', alternative='two-sided')\n",
      "        Test whether a dataset has normal kurtosis.\n",
      "        \n",
      "        This function tests the null hypothesis that the kurtosis\n",
      "        of the population from which the sample was drawn is that\n",
      "        of the normal distribution.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        a : array\n",
      "            Array of the sample data.\n",
      "        axis : int or None, optional\n",
      "           Axis along which to compute test. Default is 0. If None,\n",
      "           compute over the whole array `a`.\n",
      "        nan_policy : {'propagate', 'raise', 'omit'}, optional\n",
      "            Defines how to handle when input contains nan.\n",
      "            The following options are available (default is 'propagate'):\n",
      "        \n",
      "            * 'propagate': returns nan\n",
      "            * 'raise': throws an error\n",
      "            * 'omit': performs the calculations ignoring nan values\n",
      "        \n",
      "        alternative : {'two-sided', 'less', 'greater'}, optional\n",
      "            Defines the alternative hypothesis.\n",
      "            The following options are available (default is 'two-sided'):\n",
      "        \n",
      "            * 'two-sided': the kurtosis of the distribution underlying the sample\n",
      "              is different from that of the normal distribution\n",
      "            * 'less': the kurtosis of the distribution underlying the sample\n",
      "              is less than that of the normal distribution\n",
      "            * 'greater': the kurtosis of the distribution underlying the sample\n",
      "              is greater than that of the normal distribution\n",
      "        \n",
      "            .. versionadded:: 1.7.0\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        statistic : float\n",
      "            The computed z-score for this test.\n",
      "        pvalue : float\n",
      "            The p-value for the hypothesis test.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        Valid only for n>20. This function uses the method described in [1]_.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] see e.g. F. J. Anscombe, W. J. Glynn, \"Distribution of the kurtosis\n",
      "           statistic b2 for normal samples\", Biometrika, vol. 70, pp. 227-234, 1983.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy.stats import kurtosistest\n",
      "        >>> kurtosistest(list(range(20)))\n",
      "        KurtosistestResult(statistic=-1.7058104152122062, pvalue=0.08804338332528348)\n",
      "        >>> kurtosistest(list(range(20)), alternative='less')\n",
      "        KurtosistestResult(statistic=-1.7058104152122062, pvalue=0.04402169166264174)\n",
      "        >>> kurtosistest(list(range(20)), alternative='greater')\n",
      "        KurtosistestResult(statistic=-1.7058104152122062, pvalue=0.9559783083373583)\n",
      "        \n",
      "        >>> rng = np.random.default_rng()\n",
      "        >>> s = rng.normal(0, 1, 1000)\n",
      "        >>> kurtosistest(s)\n",
      "        KurtosistestResult(statistic=-1.475047944490622, pvalue=0.14019965402996987)\n",
      "    \n",
      "    levene(*args, center='median', proportiontocut=0.05)\n",
      "        Perform Levene test for equal variances.\n",
      "        \n",
      "        The Levene test tests the null hypothesis that all input samples\n",
      "        are from populations with equal variances.  Levene's test is an\n",
      "        alternative to Bartlett's test `bartlett` in the case where\n",
      "        there are significant deviations from normality.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        sample1, sample2, ... : array_like\n",
      "            The sample data, possibly with different lengths. Only one-dimensional\n",
      "            samples are accepted.\n",
      "        center : {'mean', 'median', 'trimmed'}, optional\n",
      "            Which function of the data to use in the test.  The default\n",
      "            is 'median'.\n",
      "        proportiontocut : float, optional\n",
      "            When `center` is 'trimmed', this gives the proportion of data points\n",
      "            to cut from each end. (See `scipy.stats.trim_mean`.)\n",
      "            Default is 0.05.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        statistic : float\n",
      "            The test statistic.\n",
      "        pvalue : float\n",
      "            The p-value for the test.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        Three variations of Levene's test are possible.  The possibilities\n",
      "        and their recommended usages are:\n",
      "        \n",
      "          * 'median' : Recommended for skewed (non-normal) distributions>\n",
      "          * 'mean' : Recommended for symmetric, moderate-tailed distributions.\n",
      "          * 'trimmed' : Recommended for heavy-tailed distributions.\n",
      "        \n",
      "        The test version using the mean was proposed in the original article\n",
      "        of Levene ([2]_) while the median and trimmed mean have been studied by\n",
      "        Brown and Forsythe ([3]_), sometimes also referred to as Brown-Forsythe\n",
      "        test.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] https://www.itl.nist.gov/div898/handbook/eda/section3/eda35a.htm\n",
      "        .. [2] Levene, H. (1960). In Contributions to Probability and Statistics:\n",
      "               Essays in Honor of Harold Hotelling, I. Olkin et al. eds.,\n",
      "               Stanford University Press, pp. 278-292.\n",
      "        .. [3] Brown, M. B. and Forsythe, A. B. (1974), Journal of the American\n",
      "               Statistical Association, 69, 364-367\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        Test whether or not the lists `a`, `b` and `c` come from populations\n",
      "        with equal variances.\n",
      "        \n",
      "        >>> from scipy.stats import levene\n",
      "        >>> a = [8.88, 9.12, 9.04, 8.98, 9.00, 9.08, 9.01, 8.85, 9.06, 8.99]\n",
      "        >>> b = [8.88, 8.95, 9.29, 9.44, 9.15, 9.58, 8.36, 9.18, 8.67, 9.05]\n",
      "        >>> c = [8.95, 9.12, 8.95, 8.85, 9.03, 8.84, 9.07, 8.98, 8.86, 8.98]\n",
      "        >>> stat, p = levene(a, b, c)\n",
      "        >>> p\n",
      "        0.002431505967249681\n",
      "        \n",
      "        The small p-value suggests that the populations do not have equal\n",
      "        variances.\n",
      "        \n",
      "        This is not surprising, given that the sample variance of `b` is much\n",
      "        larger than that of `a` and `c`:\n",
      "        \n",
      "        >>> [np.var(x, ddof=1) for x in [a, b, c]]\n",
      "        [0.007054444444444413, 0.13073888888888888, 0.008890000000000002]\n",
      "    \n",
      "    linregress(x, y=None, alternative='two-sided')\n",
      "        Calculate a linear least-squares regression for two sets of measurements.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x, y : array_like\n",
      "            Two sets of measurements.  Both arrays should have the same length.  If\n",
      "            only `x` is given (and ``y=None``), then it must be a two-dimensional\n",
      "            array where one dimension has length 2.  The two sets of measurements\n",
      "            are then found by splitting the array along the length-2 dimension. In\n",
      "            the case where ``y=None`` and `x` is a 2x2 array, ``linregress(x)`` is\n",
      "            equivalent to ``linregress(x[0], x[1])``.\n",
      "        alternative : {'two-sided', 'less', 'greater'}, optional\n",
      "            Defines the alternative hypothesis. Default is 'two-sided'.\n",
      "            The following options are available:\n",
      "        \n",
      "            * 'two-sided': the slope of the regression line is nonzero\n",
      "            * 'less': the slope of the regression line is less than zero\n",
      "            * 'greater':  the slope of the regression line is greater than zero\n",
      "        \n",
      "            .. versionadded:: 1.7.0\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        result : ``LinregressResult`` instance\n",
      "            The return value is an object with the following attributes:\n",
      "        \n",
      "            slope : float\n",
      "                Slope of the regression line.\n",
      "            intercept : float\n",
      "                Intercept of the regression line.\n",
      "            rvalue : float\n",
      "                Correlation coefficient.\n",
      "            pvalue : float\n",
      "                The p-value for a hypothesis test whose null hypothesis is\n",
      "                that the slope is zero, using Wald Test with t-distribution of\n",
      "                the test statistic. See `alternative` above for alternative\n",
      "                hypotheses.\n",
      "            stderr : float\n",
      "                Standard error of the estimated slope (gradient), under the\n",
      "                assumption of residual normality.\n",
      "            intercept_stderr : float\n",
      "                Standard error of the estimated intercept, under the assumption\n",
      "                of residual normality.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        scipy.optimize.curve_fit :\n",
      "            Use non-linear least squares to fit a function to data.\n",
      "        scipy.optimize.leastsq :\n",
      "            Minimize the sum of squares of a set of equations.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        Missing values are considered pair-wise: if a value is missing in `x`,\n",
      "        the corresponding value in `y` is masked.\n",
      "        \n",
      "        For compatibility with older versions of SciPy, the return value acts\n",
      "        like a ``namedtuple`` of length 5, with fields ``slope``, ``intercept``,\n",
      "        ``rvalue``, ``pvalue`` and ``stderr``, so one can continue to write::\n",
      "        \n",
      "            slope, intercept, r, p, se = linregress(x, y)\n",
      "        \n",
      "        With that style, however, the standard error of the intercept is not\n",
      "        available.  To have access to all the computed values, including the\n",
      "        standard error of the intercept, use the return value as an object\n",
      "        with attributes, e.g.::\n",
      "        \n",
      "            result = linregress(x, y)\n",
      "            print(result.intercept, result.intercept_stderr)\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> import matplotlib.pyplot as plt\n",
      "        >>> from scipy import stats\n",
      "        >>> rng = np.random.default_rng()\n",
      "        \n",
      "        Generate some data:\n",
      "        \n",
      "        >>> x = rng.random(10)\n",
      "        >>> y = 1.6*x + rng.random(10)\n",
      "        \n",
      "        Perform the linear regression:\n",
      "        \n",
      "        >>> res = stats.linregress(x, y)\n",
      "        \n",
      "        Coefficient of determination (R-squared):\n",
      "        \n",
      "        >>> print(f\"R-squared: {res.rvalue**2:.6f}\")\n",
      "        R-squared: 0.717533\n",
      "        \n",
      "        Plot the data along with the fitted line:\n",
      "        \n",
      "        >>> plt.plot(x, y, 'o', label='original data')\n",
      "        >>> plt.plot(x, res.intercept + res.slope*x, 'r', label='fitted line')\n",
      "        >>> plt.legend()\n",
      "        >>> plt.show()\n",
      "        \n",
      "        Calculate 95% confidence interval on slope and intercept:\n",
      "        \n",
      "        >>> # Two-sided inverse Students t-distribution\n",
      "        >>> # p - probability, df - degrees of freedom\n",
      "        >>> from scipy.stats import t\n",
      "        >>> tinv = lambda p, df: abs(t.ppf(p/2, df))\n",
      "        \n",
      "        >>> ts = tinv(0.05, len(x)-2)\n",
      "        >>> print(f\"slope (95%): {res.slope:.6f} +/- {ts*res.stderr:.6f}\")\n",
      "        slope (95%): 1.453392 +/- 0.743465\n",
      "        >>> print(f\"intercept (95%): {res.intercept:.6f}\"\n",
      "        ...       f\" +/- {ts*res.intercept_stderr:.6f}\")\n",
      "        intercept (95%): 0.616950 +/- 0.544475\n",
      "    \n",
      "    mannwhitneyu(x, y, use_continuity=True, alternative='two-sided', axis=0, method='auto')\n",
      "        Perform the Mann-Whitney U rank test on two independent samples.\n",
      "        \n",
      "        The Mann-Whitney U test is a nonparametric test of the null hypothesis\n",
      "        that the distribution underlying sample `x` is the same as the\n",
      "        distribution underlying sample `y`. It is often used as a test of\n",
      "        of difference in location between distributions.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x, y : array-like\n",
      "            N-d arrays of samples. The arrays must be broadcastable except along\n",
      "            the dimension given by `axis`.\n",
      "        use_continuity : bool, optional\n",
      "                Whether a continuity correction (1/2) should be applied.\n",
      "                Default is True when `method` is ``'asymptotic'``; has no effect\n",
      "                otherwise.\n",
      "        alternative : {'two-sided', 'less', 'greater'}, optional\n",
      "            Defines the alternative hypothesis. Default is 'two-sided'.\n",
      "            Let *F(u)* and *G(u)* be the cumulative distribution functions of the\n",
      "            distributions underlying `x` and `y`, respectively. Then the following\n",
      "            alternative hypotheses are available:\n",
      "        \n",
      "            * 'two-sided': the distributions are not equal, i.e. *F(u) ≠ G(u)* for\n",
      "              at least one *u*.\n",
      "            * 'less': the distribution underlying `x` is stochastically less\n",
      "              than the distribution underlying `y`, i.e. *F(u) > G(u)* for all *u*.\n",
      "            * 'greater': the distribution underlying `x` is stochastically greater\n",
      "              than the distribution underlying `y`, i.e. *F(u) < G(u)* for all *u*.\n",
      "        \n",
      "            Under a more restrictive set of assumptions, the alternative hypotheses\n",
      "            can be expressed in terms of the locations of the distributions;\n",
      "            see [5] section 5.1.\n",
      "        axis : int, optional\n",
      "            Axis along which to perform the test. Default is 0.\n",
      "        method : {'auto', 'asymptotic', 'exact'}, optional\n",
      "            Selects the method used to calculate the *p*-value.\n",
      "            Default is 'auto'. The following options are available.\n",
      "        \n",
      "            * ``'asymptotic'``: compares the standardized test statistic\n",
      "              against the normal distribution, correcting for ties.\n",
      "            * ``'exact'``: computes the exact *p*-value by comparing the observed\n",
      "              :math:`U` statistic against the exact distribution of the :math:`U`\n",
      "              statistic under the null hypothesis. No correction is made for ties.\n",
      "            * ``'auto'``: chooses ``'exact'`` when the size of one of the samples\n",
      "              is less than 8 and there are no ties; chooses ``'asymptotic'``\n",
      "              otherwise.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        res : MannwhitneyuResult\n",
      "            An object containing attributes:\n",
      "        \n",
      "            statistic : float\n",
      "                The Mann-Whitney U statistic corresponding with sample `x`. See\n",
      "                Notes for the test statistic corresponding with sample `y`.\n",
      "            pvalue : float\n",
      "                The associated *p*-value for the chosen `alternative`.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        If ``U1`` is the statistic corresponding with sample `x`, then the\n",
      "        statistic corresponding with sample `y` is\n",
      "        `U2 = `x.shape[axis] * y.shape[axis] - U1``.\n",
      "        \n",
      "        `mannwhitneyu` is for independent samples. For related / paired samples,\n",
      "        consider `scipy.stats.wilcoxon`.\n",
      "        \n",
      "        `method` ``'exact'`` is recommended when there are no ties and when either\n",
      "        sample size is less than 8 [1]_. The implementation follows the recurrence\n",
      "        relation originally proposed in [1]_ as it is described in [3]_.\n",
      "        Note that the exact method is *not* corrected for ties, but\n",
      "        `mannwhitneyu` will not raise errors or warnings if there are ties in the\n",
      "        data.\n",
      "        \n",
      "        The Mann-Whitney U test is a non-parametric version of the t-test for\n",
      "        independent samples. When the the means of samples from the populations\n",
      "        are normally distributed, consider `scipy.stats.ttest_ind`.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        scipy.stats.wilcoxon, scipy.stats.ranksums, scipy.stats.ttest_ind\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] H.B. Mann and D.R. Whitney, \"On a test of whether one of two random\n",
      "               variables is stochastically larger than the other\", The Annals of\n",
      "               Mathematical Statistics, Vol. 18, pp. 50-60, 1947.\n",
      "        .. [2] Mann-Whitney U Test, Wikipedia,\n",
      "               http://en.wikipedia.org/wiki/Mann-Whitney_U_test\n",
      "        .. [3] A. Di Bucchianico, \"Combinatorics, computer algebra, and the\n",
      "               Wilcoxon-Mann-Whitney test\", Journal of Statistical Planning and\n",
      "               Inference, Vol. 79, pp. 349-364, 1999.\n",
      "        .. [4] Rosie Shier, \"Statistics: 2.3 The Mann-Whitney U Test\", Mathematics\n",
      "               Learning Support Centre, 2004.\n",
      "        .. [5] Michael P. Fay and Michael A. Proschan. \"Wilcoxon-Mann-Whitney\n",
      "               or t-test? On assumptions for hypothesis tests and multiple \\\n",
      "               interpretations of decision rules.\" Statistics surveys, Vol. 4, pp.\n",
      "               1-39, 2010. https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2857732/\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        We follow the example from [4]_: nine randomly sampled young adults were\n",
      "        diagnosed with type II diabetes at the ages below.\n",
      "        \n",
      "        >>> males = [19, 22, 16, 29, 24]\n",
      "        >>> females = [20, 11, 17, 12]\n",
      "        \n",
      "        We use the Mann-Whitney U test to assess whether there is a statistically\n",
      "        significant difference in the diagnosis age of males and females.\n",
      "        The null hypothesis is that the distribution of male diagnosis ages is\n",
      "        the same as the distribution of female diagnosis ages. We decide\n",
      "        that a confidence level of 95% is required to reject the null hypothesis\n",
      "        in favor of the alternative that that the distributions are different.\n",
      "        Since the number of samples is very small and there are no ties in the\n",
      "        data, we can compare the observed test statistic against the *exact*\n",
      "        distribution of the test statistic under the null hypothesis.\n",
      "        \n",
      "        >>> from scipy.stats import mannwhitneyu\n",
      "        >>> U1, p = mannwhitneyu(males, females, method=\"exact\")\n",
      "        >>> print(U1)\n",
      "        17.0\n",
      "        \n",
      "        `mannwhitneyu` always reports the statistic associated with the first\n",
      "        sample, which, in this case, is males. This agrees with :math:`U_M = 17`\n",
      "        reported in [4]_. The statistic associated with the second statistic\n",
      "        can be calculated:\n",
      "        \n",
      "        >>> nx, ny = len(males), len(females)\n",
      "        >>> U2 = nx*ny - U1\n",
      "        >>> print(U2)\n",
      "        3.0\n",
      "        \n",
      "        This agrees with :math:`U_F = 3` reported in [4]_. The two-sided\n",
      "        *p*-value can be calculated from either statistic, and the value produced\n",
      "        by `mannwhitneyu` agrees with :math:`p = 0.11` reported in [4]_.\n",
      "        \n",
      "        >>> print(p)\n",
      "        0.1111111111111111\n",
      "        \n",
      "        The exact distribution of the test statistic is asymptotically normal, so\n",
      "        the example continues by comparing the exact *p*-value against the\n",
      "        *p*-value produced using the normal approximation.\n",
      "        \n",
      "        >>> _, pnorm = mannwhitneyu(males, females, method=\"asymptotic\")\n",
      "        >>> print(pnorm)\n",
      "        0.11134688653314041\n",
      "        \n",
      "        Here `mannwhitneyu`'s reported *p*-value appears to conflict with the\n",
      "        value :math:`p = 0.09` given in [4]_. The reason is that [4]_\n",
      "        does not apply the continuity correction performed by `mannwhitneyu`;\n",
      "        `mannwhitneyu` reduces the distance between the test statistic and the\n",
      "        mean :math:`\\mu = n_x n_y / 2` by 0.5 to correct for the fact that the\n",
      "        discrete statistic is being compared against a continuous distribution.\n",
      "        Here, the :math:`U` statistic used is less than the mean, so we reduce\n",
      "        the distance by adding 0.5 in the numerator.\n",
      "        \n",
      "        >>> import numpy as np\n",
      "        >>> from scipy.stats import norm\n",
      "        >>> U = min(U1, U2)\n",
      "        >>> N = nx + ny\n",
      "        >>> z = (U - nx*ny/2 + 0.5) / np.sqrt(nx*ny * (N + 1)/ 12)\n",
      "        >>> p = 2 * norm.cdf(z)  # use CDF to get p-value from smaller statistic\n",
      "        >>> print(p)\n",
      "        0.11134688653314041\n",
      "        \n",
      "        If desired, we can disable the continuity correction to get a result\n",
      "        that agrees with that reported in [4]_.\n",
      "        \n",
      "        >>> _, pnorm = mannwhitneyu(males, females, use_continuity=False,\n",
      "        ...                         method=\"asymptotic\")\n",
      "        >>> print(pnorm)\n",
      "        0.0864107329737\n",
      "        \n",
      "        Regardless of whether we perform an exact or asymptotic test, the\n",
      "        probability of the test statistic being as extreme or more extreme by\n",
      "        chance exceeds 5%, so we do not consider the results statistically\n",
      "        significant.\n",
      "        \n",
      "        Suppose that, before seeing the data, we had hypothesized that females\n",
      "        would tend to be diagnosed at a younger age than males.\n",
      "        In that case, it would be natural to provide the female ages as the\n",
      "        first input, and we would have performed a one-sided test using\n",
      "        ``alternative = 'less'``: females are diagnosed at an age that is\n",
      "        stochastically less than that of males.\n",
      "        \n",
      "        >>> res = mannwhitneyu(females, males, alternative=\"less\", method=\"exact\")\n",
      "        >>> print(res)\n",
      "        MannwhitneyuResult(statistic=3.0, pvalue=0.05555555555555555)\n",
      "        \n",
      "        Again, the probability of getting a sufficiently low value of the\n",
      "        test statistic by chance under the null hypothesis is greater than 5%,\n",
      "        so we do not reject the null hypothesis in favor of our alternative.\n",
      "        \n",
      "        If it is reasonable to assume that the means of samples from the\n",
      "        populations are normally distributed, we could have used a t-test to\n",
      "        perform the analysis.\n",
      "        \n",
      "        >>> from scipy.stats import ttest_ind\n",
      "        >>> res = ttest_ind(females, males, alternative=\"less\")\n",
      "        >>> print(res)\n",
      "        Ttest_indResult(statistic=-2.239334696520584, pvalue=0.030068441095757924)\n",
      "        \n",
      "        Under this assumption, the *p*-value would be low enough to reject the\n",
      "        null hypothesis in favor of the alternative.\n",
      "    \n",
      "    median_abs_deviation(x, axis=0, center=<function median at 0x00000167FC2A24C0>, scale=1.0, nan_policy='propagate')\n",
      "        Compute the median absolute deviation of the data along the given axis.\n",
      "        \n",
      "        The median absolute deviation (MAD, [1]_) computes the median over the\n",
      "        absolute deviations from the median. It is a measure of dispersion\n",
      "        similar to the standard deviation but more robust to outliers [2]_.\n",
      "        \n",
      "        The MAD of an empty array is ``np.nan``.\n",
      "        \n",
      "        .. versionadded:: 1.5.0\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x : array_like\n",
      "            Input array or object that can be converted to an array.\n",
      "        axis : int or None, optional\n",
      "            Axis along which the range is computed. Default is 0. If None, compute\n",
      "            the MAD over the entire array.\n",
      "        center : callable, optional\n",
      "            A function that will return the central value. The default is to use\n",
      "            np.median. Any user defined function used will need to have the\n",
      "            function signature ``func(arr, axis)``.\n",
      "        scale : scalar or str, optional\n",
      "            The numerical value of scale will be divided out of the final\n",
      "            result. The default is 1.0. The string \"normal\" is also accepted,\n",
      "            and results in `scale` being the inverse of the standard normal\n",
      "            quantile function at 0.75, which is approximately 0.67449.\n",
      "            Array-like scale is also allowed, as long as it broadcasts correctly\n",
      "            to the output such that ``out / scale`` is a valid operation. The\n",
      "            output dimensions depend on the input array, `x`, and the `axis`\n",
      "            argument.\n",
      "        nan_policy : {'propagate', 'raise', 'omit'}, optional\n",
      "            Defines how to handle when input contains nan.\n",
      "            The following options are available (default is 'propagate'):\n",
      "        \n",
      "            * 'propagate': returns nan\n",
      "            * 'raise': throws an error\n",
      "            * 'omit': performs the calculations ignoring nan values\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        mad : scalar or ndarray\n",
      "            If ``axis=None``, a scalar is returned. If the input contains\n",
      "            integers or floats of smaller precision than ``np.float64``, then the\n",
      "            output data-type is ``np.float64``. Otherwise, the output data-type is\n",
      "            the same as that of the input.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        numpy.std, numpy.var, numpy.median, scipy.stats.iqr, scipy.stats.tmean,\n",
      "        scipy.stats.tstd, scipy.stats.tvar\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        The `center` argument only affects the calculation of the central value\n",
      "        around which the MAD is calculated. That is, passing in ``center=np.mean``\n",
      "        will calculate the MAD around the mean - it will not calculate the *mean*\n",
      "        absolute deviation.\n",
      "        \n",
      "        The input array may contain `inf`, but if `center` returns `inf`, the\n",
      "        corresponding MAD for that data will be `nan`.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] \"Median absolute deviation\",\n",
      "               https://en.wikipedia.org/wiki/Median_absolute_deviation\n",
      "        .. [2] \"Robust measures of scale\",\n",
      "               https://en.wikipedia.org/wiki/Robust_measures_of_scale\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        When comparing the behavior of `median_abs_deviation` with ``np.std``,\n",
      "        the latter is affected when we change a single value of an array to have an\n",
      "        outlier value while the MAD hardly changes:\n",
      "        \n",
      "        >>> from scipy import stats\n",
      "        >>> x = stats.norm.rvs(size=100, scale=1, random_state=123456)\n",
      "        >>> x.std()\n",
      "        0.9973906394005013\n",
      "        >>> stats.median_abs_deviation(x)\n",
      "        0.82832610097857\n",
      "        >>> x[0] = 345.6\n",
      "        >>> x.std()\n",
      "        34.42304872314415\n",
      "        >>> stats.median_abs_deviation(x)\n",
      "        0.8323442311590675\n",
      "        \n",
      "        Axis handling example:\n",
      "        \n",
      "        >>> x = np.array([[10, 7, 4], [3, 2, 1]])\n",
      "        >>> x\n",
      "        array([[10,  7,  4],\n",
      "               [ 3,  2,  1]])\n",
      "        >>> stats.median_abs_deviation(x)\n",
      "        array([3.5, 2.5, 1.5])\n",
      "        >>> stats.median_abs_deviation(x, axis=None)\n",
      "        2.0\n",
      "        \n",
      "        Scale normal example:\n",
      "        \n",
      "        >>> x = stats.norm.rvs(size=1000000, scale=2, random_state=123456)\n",
      "        >>> stats.median_abs_deviation(x)\n",
      "        1.3487398527041636\n",
      "        >>> stats.median_abs_deviation(x, scale='normal')\n",
      "        1.9996446978061115\n",
      "    \n",
      "    median_absolute_deviation(*args, **kwds)\n",
      "        `median_absolute_deviation` is deprecated, use `median_abs_deviation` instead!\n",
      "        \n",
      "        To preserve the existing default behavior, use\n",
      "        `scipy.stats.median_abs_deviation(..., scale=1/1.4826)`.\n",
      "        The value 1.4826 is not numerically precise for scaling\n",
      "        with a normal distribution. For a numerically precise value, use\n",
      "        `scipy.stats.median_abs_deviation(..., scale='normal')`.\n",
      "        \n",
      "        \n",
      "        Compute the median absolute deviation of the data along the given axis.\n",
      "        \n",
      "        The median absolute deviation (MAD, [1]_) computes the median over the\n",
      "        absolute deviations from the median. It is a measure of dispersion\n",
      "        similar to the standard deviation but more robust to outliers [2]_.\n",
      "        \n",
      "        The MAD of an empty array is ``np.nan``.\n",
      "        \n",
      "        .. versionadded:: 1.3.0\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x : array_like\n",
      "            Input array or object that can be converted to an array.\n",
      "        axis : int or None, optional\n",
      "            Axis along which the range is computed. Default is 0. If None, compute\n",
      "            the MAD over the entire array.\n",
      "        center : callable, optional\n",
      "            A function that will return the central value. The default is to use\n",
      "            np.median. Any user defined function used will need to have the function\n",
      "            signature ``func(arr, axis)``.\n",
      "        scale : int, optional\n",
      "            The scaling factor applied to the MAD. The default scale (1.4826)\n",
      "            ensures consistency with the standard deviation for normally distributed\n",
      "            data.\n",
      "        nan_policy : {'propagate', 'raise', 'omit'}, optional\n",
      "            Defines how to handle when input contains nan.\n",
      "            The following options are available (default is 'propagate'):\n",
      "        \n",
      "            * 'propagate': returns nan\n",
      "            * 'raise': throws an error\n",
      "            * 'omit': performs the calculations ignoring nan values\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        mad : scalar or ndarray\n",
      "            If ``axis=None``, a scalar is returned. If the input contains\n",
      "            integers or floats of smaller precision than ``np.float64``, then the\n",
      "            output data-type is ``np.float64``. Otherwise, the output data-type is\n",
      "            the same as that of the input.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        numpy.std, numpy.var, numpy.median, scipy.stats.iqr, scipy.stats.tmean,\n",
      "        scipy.stats.tstd, scipy.stats.tvar\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        The `center` argument only affects the calculation of the central value\n",
      "        around which the MAD is calculated. That is, passing in ``center=np.mean``\n",
      "        will calculate the MAD around the mean - it will not calculate the *mean*\n",
      "        absolute deviation.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] \"Median absolute deviation\",\n",
      "               https://en.wikipedia.org/wiki/Median_absolute_deviation\n",
      "        .. [2] \"Robust measures of scale\",\n",
      "               https://en.wikipedia.org/wiki/Robust_measures_of_scale\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        When comparing the behavior of `median_absolute_deviation` with ``np.std``,\n",
      "        the latter is affected when we change a single value of an array to have an\n",
      "        outlier value while the MAD hardly changes:\n",
      "        \n",
      "        >>> from scipy import stats\n",
      "        >>> x = stats.norm.rvs(size=100, scale=1, random_state=123456)\n",
      "        >>> x.std()\n",
      "        0.9973906394005013\n",
      "        >>> stats.median_absolute_deviation(x)\n",
      "        1.2280762773108278\n",
      "        >>> x[0] = 345.6\n",
      "        >>> x.std()\n",
      "        34.42304872314415\n",
      "        >>> stats.median_absolute_deviation(x)\n",
      "        1.2340335571164334\n",
      "        \n",
      "        Axis handling example:\n",
      "        \n",
      "        >>> x = np.array([[10, 7, 4], [3, 2, 1]])\n",
      "        >>> x\n",
      "        array([[10,  7,  4],\n",
      "               [ 3,  2,  1]])\n",
      "        >>> stats.median_absolute_deviation(x)\n",
      "        array([5.1891, 3.7065, 2.2239])\n",
      "        >>> stats.median_absolute_deviation(x, axis=None)\n",
      "        2.9652\n",
      "    \n",
      "    median_test(*args, ties='below', correction=True, lambda_=1, nan_policy='propagate')\n",
      "        Perform a Mood's median test.\n",
      "        \n",
      "        Test that two or more samples come from populations with the same median.\n",
      "        \n",
      "        Let ``n = len(args)`` be the number of samples.  The \"grand median\" of\n",
      "        all the data is computed, and a contingency table is formed by\n",
      "        classifying the values in each sample as being above or below the grand\n",
      "        median.  The contingency table, along with `correction` and `lambda_`,\n",
      "        are passed to `scipy.stats.chi2_contingency` to compute the test statistic\n",
      "        and p-value.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        sample1, sample2, ... : array_like\n",
      "            The set of samples.  There must be at least two samples.\n",
      "            Each sample must be a one-dimensional sequence containing at least\n",
      "            one value.  The samples are not required to have the same length.\n",
      "        ties : str, optional\n",
      "            Determines how values equal to the grand median are classified in\n",
      "            the contingency table.  The string must be one of::\n",
      "        \n",
      "                \"below\":\n",
      "                    Values equal to the grand median are counted as \"below\".\n",
      "                \"above\":\n",
      "                    Values equal to the grand median are counted as \"above\".\n",
      "                \"ignore\":\n",
      "                    Values equal to the grand median are not counted.\n",
      "        \n",
      "            The default is \"below\".\n",
      "        correction : bool, optional\n",
      "            If True, *and* there are just two samples, apply Yates' correction\n",
      "            for continuity when computing the test statistic associated with\n",
      "            the contingency table.  Default is True.\n",
      "        lambda_ : float or str, optional\n",
      "            By default, the statistic computed in this test is Pearson's\n",
      "            chi-squared statistic.  `lambda_` allows a statistic from the\n",
      "            Cressie-Read power divergence family to be used instead.  See\n",
      "            `power_divergence` for details.\n",
      "            Default is 1 (Pearson's chi-squared statistic).\n",
      "        nan_policy : {'propagate', 'raise', 'omit'}, optional\n",
      "            Defines how to handle when input contains nan. 'propagate' returns nan,\n",
      "            'raise' throws an error, 'omit' performs the calculations ignoring nan\n",
      "            values. Default is 'propagate'.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        stat : float\n",
      "            The test statistic.  The statistic that is returned is determined by\n",
      "            `lambda_`.  The default is Pearson's chi-squared statistic.\n",
      "        p : float\n",
      "            The p-value of the test.\n",
      "        m : float\n",
      "            The grand median.\n",
      "        table : ndarray\n",
      "            The contingency table.  The shape of the table is (2, n), where\n",
      "            n is the number of samples.  The first row holds the counts of the\n",
      "            values above the grand median, and the second row holds the counts\n",
      "            of the values below the grand median.  The table allows further\n",
      "            analysis with, for example, `scipy.stats.chi2_contingency`, or with\n",
      "            `scipy.stats.fisher_exact` if there are two samples, without having\n",
      "            to recompute the table.  If ``nan_policy`` is \"propagate\" and there\n",
      "            are nans in the input, the return value for ``table`` is ``None``.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        kruskal : Compute the Kruskal-Wallis H-test for independent samples.\n",
      "        mannwhitneyu : Computes the Mann-Whitney rank test on samples x and y.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        .. versionadded:: 0.15.0\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] Mood, A. M., Introduction to the Theory of Statistics. McGraw-Hill\n",
      "            (1950), pp. 394-399.\n",
      "        .. [2] Zar, J. H., Biostatistical Analysis, 5th ed. Prentice Hall (2010).\n",
      "            See Sections 8.12 and 10.15.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        A biologist runs an experiment in which there are three groups of plants.\n",
      "        Group 1 has 16 plants, group 2 has 15 plants, and group 3 has 17 plants.\n",
      "        Each plant produces a number of seeds.  The seed counts for each group\n",
      "        are::\n",
      "        \n",
      "            Group 1: 10 14 14 18 20 22 24 25 31 31 32 39 43 43 48 49\n",
      "            Group 2: 28 30 31 33 34 35 36 40 44 55 57 61 91 92 99\n",
      "            Group 3:  0  3  9 22 23 25 25 33 34 34 40 45 46 48 62 67 84\n",
      "        \n",
      "        The following code applies Mood's median test to these samples.\n",
      "        \n",
      "        >>> g1 = [10, 14, 14, 18, 20, 22, 24, 25, 31, 31, 32, 39, 43, 43, 48, 49]\n",
      "        >>> g2 = [28, 30, 31, 33, 34, 35, 36, 40, 44, 55, 57, 61, 91, 92, 99]\n",
      "        >>> g3 = [0, 3, 9, 22, 23, 25, 25, 33, 34, 34, 40, 45, 46, 48, 62, 67, 84]\n",
      "        >>> from scipy.stats import median_test\n",
      "        >>> stat, p, med, tbl = median_test(g1, g2, g3)\n",
      "        \n",
      "        The median is\n",
      "        \n",
      "        >>> med\n",
      "        34.0\n",
      "        \n",
      "        and the contingency table is\n",
      "        \n",
      "        >>> tbl\n",
      "        array([[ 5, 10,  7],\n",
      "               [11,  5, 10]])\n",
      "        \n",
      "        `p` is too large to conclude that the medians are not the same:\n",
      "        \n",
      "        >>> p\n",
      "        0.12609082774093244\n",
      "        \n",
      "        The \"G-test\" can be performed by passing ``lambda_=\"log-likelihood\"`` to\n",
      "        `median_test`.\n",
      "        \n",
      "        >>> g, p, med, tbl = median_test(g1, g2, g3, lambda_=\"log-likelihood\")\n",
      "        >>> p\n",
      "        0.12224779737117837\n",
      "        \n",
      "        The median occurs several times in the data, so we'll get a different\n",
      "        result if, for example, ``ties=\"above\"`` is used:\n",
      "        \n",
      "        >>> stat, p, med, tbl = median_test(g1, g2, g3, ties=\"above\")\n",
      "        >>> p\n",
      "        0.063873276069553273\n",
      "        \n",
      "        >>> tbl\n",
      "        array([[ 5, 11,  9],\n",
      "               [11,  4,  8]])\n",
      "        \n",
      "        This example demonstrates that if the data set is not large and there\n",
      "        are values equal to the median, the p-value can be sensitive to the\n",
      "        choice of `ties`.\n",
      "    \n",
      "    mode(a, axis=0, nan_policy='propagate')\n",
      "        Return an array of the modal (most common) value in the passed array.\n",
      "        \n",
      "        If there is more than one such value, only the smallest is returned.\n",
      "        The bin-count for the modal bins is also returned.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        a : array_like\n",
      "            n-dimensional array of which to find mode(s).\n",
      "        axis : int or None, optional\n",
      "            Axis along which to operate. Default is 0. If None, compute over\n",
      "            the whole array `a`.\n",
      "        nan_policy : {'propagate', 'raise', 'omit'}, optional\n",
      "            Defines how to handle when input contains nan.\n",
      "            The following options are available (default is 'propagate'):\n",
      "        \n",
      "              * 'propagate': returns nan\n",
      "              * 'raise': throws an error\n",
      "              * 'omit': performs the calculations ignoring nan values\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        mode : ndarray\n",
      "            Array of modal values.\n",
      "        count : ndarray\n",
      "            Array of counts for each mode.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> a = np.array([[6, 8, 3, 0],\n",
      "        ...               [3, 2, 1, 7],\n",
      "        ...               [8, 1, 8, 4],\n",
      "        ...               [5, 3, 0, 5],\n",
      "        ...               [4, 7, 5, 9]])\n",
      "        >>> from scipy import stats\n",
      "        >>> stats.mode(a)\n",
      "        ModeResult(mode=array([[3, 1, 0, 0]]), count=array([[1, 1, 1, 1]]))\n",
      "        \n",
      "        To get mode of whole array, specify ``axis=None``:\n",
      "        \n",
      "        >>> stats.mode(a, axis=None)\n",
      "        ModeResult(mode=array([3]), count=array([3]))\n",
      "    \n",
      "    moment(a, moment=1, axis=0, nan_policy='propagate')\n",
      "        Calculate the nth moment about the mean for a sample.\n",
      "        \n",
      "        A moment is a specific quantitative measure of the shape of a set of\n",
      "        points. It is often used to calculate coefficients of skewness and kurtosis\n",
      "        due to its close relationship with them.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        a : array_like\n",
      "           Input array.\n",
      "        moment : int or array_like of ints, optional\n",
      "           Order of central moment that is returned. Default is 1.\n",
      "        axis : int or None, optional\n",
      "           Axis along which the central moment is computed. Default is 0.\n",
      "           If None, compute over the whole array `a`.\n",
      "        nan_policy : {'propagate', 'raise', 'omit'}, optional\n",
      "            Defines how to handle when input contains nan.\n",
      "            The following options are available (default is 'propagate'):\n",
      "        \n",
      "              * 'propagate': returns nan\n",
      "              * 'raise': throws an error\n",
      "              * 'omit': performs the calculations ignoring nan values\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        n-th central moment : ndarray or float\n",
      "           The appropriate moment along the given axis or over all values if axis\n",
      "           is None. The denominator for the moment calculation is the number of\n",
      "           observations, no degrees of freedom correction is done.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        kurtosis, skew, describe\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        The k-th central moment of a data sample is:\n",
      "        \n",
      "        .. math::\n",
      "        \n",
      "            m_k = \\frac{1}{n} \\sum_{i = 1}^n (x_i - \\bar{x})^k\n",
      "        \n",
      "        Where n is the number of samples and x-bar is the mean. This function uses\n",
      "        exponentiation by squares [1]_ for efficiency.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] https://eli.thegreenplace.net/2009/03/21/efficient-integer-exponentiation-algorithms\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy.stats import moment\n",
      "        >>> moment([1, 2, 3, 4, 5], moment=1)\n",
      "        0.0\n",
      "        >>> moment([1, 2, 3, 4, 5], moment=2)\n",
      "        2.0\n",
      "    \n",
      "    mood(x, y, axis=0, alternative='two-sided')\n",
      "        Perform Mood's test for equal scale parameters.\n",
      "        \n",
      "        Mood's two-sample test for scale parameters is a non-parametric\n",
      "        test for the null hypothesis that two samples are drawn from the\n",
      "        same distribution with the same scale parameter.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x, y : array_like\n",
      "            Arrays of sample data.\n",
      "        axis : int, optional\n",
      "            The axis along which the samples are tested.  `x` and `y` can be of\n",
      "            different length along `axis`.\n",
      "            If `axis` is None, `x` and `y` are flattened and the test is done on\n",
      "            all values in the flattened arrays.\n",
      "        alternative : {'two-sided', 'less', 'greater'}, optional\n",
      "            Defines the alternative hypothesis. Default is 'two-sided'.\n",
      "            The following options are available:\n",
      "        \n",
      "            * 'two-sided': the scales of the distributions underlying `x` and `y`\n",
      "              are different.\n",
      "            * 'less': the scale of the distribution underlying `x` is less than\n",
      "              the scale of the distribution underlying `y`.\n",
      "            * 'greater': the scale of the distribution underlying `x` is greater\n",
      "              than the scale of the distribution underlying `y`.\n",
      "        \n",
      "            .. versionadded:: 1.7.0\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        z : scalar or ndarray\n",
      "            The z-score for the hypothesis test.  For 1-D inputs a scalar is\n",
      "            returned.\n",
      "        p-value : scalar ndarray\n",
      "            The p-value for the hypothesis test.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        fligner : A non-parametric test for the equality of k variances\n",
      "        ansari : A non-parametric test for the equality of 2 variances\n",
      "        bartlett : A parametric test for equality of k variances in normal samples\n",
      "        levene : A parametric test for equality of k variances\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        The data are assumed to be drawn from probability distributions ``f(x)``\n",
      "        and ``f(x/s) / s`` respectively, for some probability density function f.\n",
      "        The null hypothesis is that ``s == 1``.\n",
      "        \n",
      "        For multi-dimensional arrays, if the inputs are of shapes\n",
      "        ``(n0, n1, n2, n3)``  and ``(n0, m1, n2, n3)``, then if ``axis=1``, the\n",
      "        resulting z and p values will have shape ``(n0, n2, n3)``.  Note that\n",
      "        ``n1`` and ``m1`` don't have to be equal, but the other dimensions do.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy import stats\n",
      "        >>> rng = np.random.default_rng()\n",
      "        >>> x2 = rng.standard_normal((2, 45, 6, 7))\n",
      "        >>> x1 = rng.standard_normal((2, 30, 6, 7))\n",
      "        >>> z, p = stats.mood(x1, x2, axis=1)\n",
      "        >>> p.shape\n",
      "        (2, 6, 7)\n",
      "        \n",
      "        Find the number of points where the difference in scale is not significant:\n",
      "        \n",
      "        >>> (p > 0.1).sum()\n",
      "        78\n",
      "        \n",
      "        Perform the test with different scales:\n",
      "        \n",
      "        >>> x1 = rng.standard_normal((2, 30))\n",
      "        >>> x2 = rng.standard_normal((2, 35)) * 10.0\n",
      "        >>> stats.mood(x1, x2, axis=1)\n",
      "        (array([-5.76174136, -6.12650783]), array([8.32505043e-09, 8.98287869e-10]))\n",
      "    \n",
      "    multiscale_graphcorr(x, y, compute_distance=<function _euclidean_dist at 0x00000167FE52AE50>, reps=1000, workers=1, is_twosamp=False, random_state=None)\n",
      "        Computes the Multiscale Graph Correlation (MGC) test statistic.\n",
      "        \n",
      "        Specifically, for each point, MGC finds the :math:`k`-nearest neighbors for\n",
      "        one property (e.g. cloud density), and the :math:`l`-nearest neighbors for\n",
      "        the other property (e.g. grass wetness) [1]_. This pair :math:`(k, l)` is\n",
      "        called the \"scale\". A priori, however, it is not know which scales will be\n",
      "        most informative. So, MGC computes all distance pairs, and then efficiently\n",
      "        computes the distance correlations for all scales. The local correlations\n",
      "        illustrate which scales are relatively informative about the relationship.\n",
      "        The key, therefore, to successfully discover and decipher relationships\n",
      "        between disparate data modalities is to adaptively determine which scales\n",
      "        are the most informative, and the geometric implication for the most\n",
      "        informative scales. Doing so not only provides an estimate of whether the\n",
      "        modalities are related, but also provides insight into how the\n",
      "        determination was made. This is especially important in high-dimensional\n",
      "        data, where simple visualizations do not reveal relationships to the\n",
      "        unaided human eye. Characterizations of this implementation in particular\n",
      "        have been derived from and benchmarked within in [2]_.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x, y : ndarray\n",
      "            If ``x`` and ``y`` have shapes ``(n, p)`` and ``(n, q)`` where `n` is\n",
      "            the number of samples and `p` and `q` are the number of dimensions,\n",
      "            then the MGC independence test will be run.  Alternatively, ``x`` and\n",
      "            ``y`` can have shapes ``(n, n)`` if they are distance or similarity\n",
      "            matrices, and ``compute_distance`` must be sent to ``None``. If ``x``\n",
      "            and ``y`` have shapes ``(n, p)`` and ``(m, p)``, an unpaired\n",
      "            two-sample MGC test will be run.\n",
      "        compute_distance : callable, optional\n",
      "            A function that computes the distance or similarity among the samples\n",
      "            within each data matrix. Set to ``None`` if ``x`` and ``y`` are\n",
      "            already distance matrices. The default uses the euclidean norm metric.\n",
      "            If you are calling a custom function, either create the distance\n",
      "            matrix before-hand or create a function of the form\n",
      "            ``compute_distance(x)`` where `x` is the data matrix for which\n",
      "            pairwise distances are calculated.\n",
      "        reps : int, optional\n",
      "            The number of replications used to estimate the null when using the\n",
      "            permutation test. The default is ``1000``.\n",
      "        workers : int or map-like callable, optional\n",
      "            If ``workers`` is an int the population is subdivided into ``workers``\n",
      "            sections and evaluated in parallel (uses ``multiprocessing.Pool\n",
      "            <multiprocessing>``). Supply ``-1`` to use all cores available to the\n",
      "            Process. Alternatively supply a map-like callable, such as\n",
      "            ``multiprocessing.Pool.map`` for evaluating the p-value in parallel.\n",
      "            This evaluation is carried out as ``workers(func, iterable)``.\n",
      "            Requires that `func` be pickleable. The default is ``1``.\n",
      "        is_twosamp : bool, optional\n",
      "            If `True`, a two sample test will be run. If ``x`` and ``y`` have\n",
      "            shapes ``(n, p)`` and ``(m, p)``, this optional will be overriden and\n",
      "            set to ``True``. Set to ``True`` if ``x`` and ``y`` both have shapes\n",
      "            ``(n, p)`` and a two sample test is desired. The default is ``False``.\n",
      "            Note that this will not run if inputs are distance matrices.\n",
      "        random_state : {None, int, `numpy.random.Generator`,\n",
      "                        `numpy.random.RandomState`}, optional\n",
      "        \n",
      "            If `seed` is None (or `np.random`), the `numpy.random.RandomState`\n",
      "            singleton is used.\n",
      "            If `seed` is an int, a new ``RandomState`` instance is used,\n",
      "            seeded with `seed`.\n",
      "            If `seed` is already a ``Generator`` or ``RandomState`` instance then\n",
      "            that instance is used.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        stat : float\n",
      "            The sample MGC test statistic within `[-1, 1]`.\n",
      "        pvalue : float\n",
      "            The p-value obtained via permutation.\n",
      "        mgc_dict : dict\n",
      "            Contains additional useful additional returns containing the following\n",
      "            keys:\n",
      "        \n",
      "                - mgc_map : ndarray\n",
      "                    A 2D representation of the latent geometry of the relationship.\n",
      "                    of the relationship.\n",
      "                - opt_scale : (int, int)\n",
      "                    The estimated optimal scale as a `(x, y)` pair.\n",
      "                - null_dist : list\n",
      "                    The null distribution derived from the permuted matrices\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        pearsonr : Pearson correlation coefficient and p-value for testing\n",
      "                   non-correlation.\n",
      "        kendalltau : Calculates Kendall's tau.\n",
      "        spearmanr : Calculates a Spearman rank-order correlation coefficient.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        A description of the process of MGC and applications on neuroscience data\n",
      "        can be found in [1]_. It is performed using the following steps:\n",
      "        \n",
      "        #. Two distance matrices :math:`D^X` and :math:`D^Y` are computed and\n",
      "           modified to be mean zero columnwise. This results in two\n",
      "           :math:`n \\times n` distance matrices :math:`A` and :math:`B` (the\n",
      "           centering and unbiased modification) [3]_.\n",
      "        \n",
      "        #. For all values :math:`k` and :math:`l` from :math:`1, ..., n`,\n",
      "        \n",
      "           * The :math:`k`-nearest neighbor and :math:`l`-nearest neighbor graphs\n",
      "             are calculated for each property. Here, :math:`G_k (i, j)` indicates\n",
      "             the :math:`k`-smallest values of the :math:`i`-th row of :math:`A`\n",
      "             and :math:`H_l (i, j)` indicates the :math:`l` smallested values of\n",
      "             the :math:`i`-th row of :math:`B`\n",
      "        \n",
      "           * Let :math:`\\circ` denotes the entry-wise matrix product, then local\n",
      "             correlations are summed and normalized using the following statistic:\n",
      "        \n",
      "        .. math::\n",
      "        \n",
      "            c^{kl} = \\frac{\\sum_{ij} A G_k B H_l}\n",
      "                          {\\sqrt{\\sum_{ij} A^2 G_k \\times \\sum_{ij} B^2 H_l}}\n",
      "        \n",
      "        #. The MGC test statistic is the smoothed optimal local correlation of\n",
      "           :math:`\\{ c^{kl} \\}`. Denote the smoothing operation as :math:`R(\\cdot)`\n",
      "           (which essentially set all isolated large correlations) as 0 and\n",
      "           connected large correlations the same as before, see [3]_.) MGC is,\n",
      "        \n",
      "        .. math::\n",
      "        \n",
      "            MGC_n (x, y) = \\max_{(k, l)} R \\left(c^{kl} \\left( x_n, y_n \\right)\n",
      "                                                        \\right)\n",
      "        \n",
      "        The test statistic returns a value between :math:`(-1, 1)` since it is\n",
      "        normalized.\n",
      "        \n",
      "        The p-value returned is calculated using a permutation test. This process\n",
      "        is completed by first randomly permuting :math:`y` to estimate the null\n",
      "        distribution and then calculating the probability of observing a test\n",
      "        statistic, under the null, at least as extreme as the observed test\n",
      "        statistic.\n",
      "        \n",
      "        MGC requires at least 5 samples to run with reliable results. It can also\n",
      "        handle high-dimensional data sets.\n",
      "        In addition, by manipulating the input data matrices, the two-sample\n",
      "        testing problem can be reduced to the independence testing problem [4]_.\n",
      "        Given sample data :math:`U` and :math:`V` of sizes :math:`p \\times n`\n",
      "        :math:`p \\times m`, data matrix :math:`X` and :math:`Y` can be created as\n",
      "        follows:\n",
      "        \n",
      "        .. math::\n",
      "        \n",
      "            X = [U | V] \\in \\mathcal{R}^{p \\times (n + m)}\n",
      "            Y = [0_{1 \\times n} | 1_{1 \\times m}] \\in \\mathcal{R}^{(n + m)}\n",
      "        \n",
      "        Then, the MGC statistic can be calculated as normal. This methodology can\n",
      "        be extended to similar tests such as distance correlation [4]_.\n",
      "        \n",
      "        .. versionadded:: 1.4.0\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] Vogelstein, J. T., Bridgeford, E. W., Wang, Q., Priebe, C. E.,\n",
      "               Maggioni, M., & Shen, C. (2019). Discovering and deciphering\n",
      "               relationships across disparate data modalities. ELife.\n",
      "        .. [2] Panda, S., Palaniappan, S., Xiong, J., Swaminathan, A.,\n",
      "               Ramachandran, S., Bridgeford, E. W., ... Vogelstein, J. T. (2019).\n",
      "               mgcpy: A Comprehensive High Dimensional Independence Testing Python\n",
      "               Package. :arXiv:`1907.02088`\n",
      "        .. [3] Shen, C., Priebe, C.E., & Vogelstein, J. T. (2019). From distance\n",
      "               correlation to multiscale graph correlation. Journal of the American\n",
      "               Statistical Association.\n",
      "        .. [4] Shen, C. & Vogelstein, J. T. (2018). The Exact Equivalence of\n",
      "               Distance and Kernel Methods for Hypothesis Testing.\n",
      "               :arXiv:`1806.05514`\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy.stats import multiscale_graphcorr\n",
      "        >>> x = np.arange(100)\n",
      "        >>> y = x\n",
      "        >>> stat, pvalue, _ = multiscale_graphcorr(x, y, workers=-1)\n",
      "        >>> '%.1f, %.3f' % (stat, pvalue)\n",
      "        '1.0, 0.001'\n",
      "        \n",
      "        Alternatively,\n",
      "        \n",
      "        >>> x = np.arange(100)\n",
      "        >>> y = x\n",
      "        >>> mgc = multiscale_graphcorr(x, y)\n",
      "        >>> '%.1f, %.3f' % (mgc.stat, mgc.pvalue)\n",
      "        '1.0, 0.001'\n",
      "        \n",
      "        To run an unpaired two-sample test,\n",
      "        \n",
      "        >>> x = np.arange(100)\n",
      "        >>> y = np.arange(79)\n",
      "        >>> mgc = multiscale_graphcorr(x, y)\n",
      "        >>> '%.3f, %.2f' % (mgc.stat, mgc.pvalue)  # doctest: +SKIP\n",
      "        '0.033, 0.02'\n",
      "        \n",
      "        or, if shape of the inputs are the same,\n",
      "        \n",
      "        >>> x = np.arange(100)\n",
      "        >>> y = x\n",
      "        >>> mgc = multiscale_graphcorr(x, y, is_twosamp=True)\n",
      "        >>> '%.3f, %.1f' % (mgc.stat, mgc.pvalue)  # doctest: +SKIP\n",
      "        '-0.008, 1.0'\n",
      "    \n",
      "    mvsdist(data)\n",
      "        'Frozen' distributions for mean, variance, and standard deviation of data.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        data : array_like\n",
      "            Input array. Converted to 1-D using ravel.\n",
      "            Requires 2 or more data-points.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        mdist : \"frozen\" distribution object\n",
      "            Distribution object representing the mean of the data.\n",
      "        vdist : \"frozen\" distribution object\n",
      "            Distribution object representing the variance of the data.\n",
      "        sdist : \"frozen\" distribution object\n",
      "            Distribution object representing the standard deviation of the data.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        bayes_mvs\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        The return values from ``bayes_mvs(data)`` is equivalent to\n",
      "        ``tuple((x.mean(), x.interval(0.90)) for x in mvsdist(data))``.\n",
      "        \n",
      "        In other words, calling ``<dist>.mean()`` and ``<dist>.interval(0.90)``\n",
      "        on the three distribution objects returned from this function will give\n",
      "        the same results that are returned from `bayes_mvs`.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        T.E. Oliphant, \"A Bayesian perspective on estimating mean, variance, and\n",
      "        standard-deviation from data\", https://scholarsarchive.byu.edu/facpub/278,\n",
      "        2006.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy import stats\n",
      "        >>> data = [6, 9, 12, 7, 8, 8, 13]\n",
      "        >>> mean, var, std = stats.mvsdist(data)\n",
      "        \n",
      "        We now have frozen distribution objects \"mean\", \"var\" and \"std\" that we can\n",
      "        examine:\n",
      "        \n",
      "        >>> mean.mean()\n",
      "        9.0\n",
      "        >>> mean.interval(0.95)\n",
      "        (6.6120585482655692, 11.387941451734431)\n",
      "        >>> mean.std()\n",
      "        1.1952286093343936\n",
      "    \n",
      "    normaltest(a, axis=0, nan_policy='propagate')\n",
      "        Test whether a sample differs from a normal distribution.\n",
      "        \n",
      "        This function tests the null hypothesis that a sample comes\n",
      "        from a normal distribution.  It is based on D'Agostino and\n",
      "        Pearson's [1]_, [2]_ test that combines skew and kurtosis to\n",
      "        produce an omnibus test of normality.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        a : array_like\n",
      "            The array containing the sample to be tested.\n",
      "        axis : int or None, optional\n",
      "            Axis along which to compute test. Default is 0. If None,\n",
      "            compute over the whole array `a`.\n",
      "        nan_policy : {'propagate', 'raise', 'omit'}, optional\n",
      "            Defines how to handle when input contains nan.\n",
      "            The following options are available (default is 'propagate'):\n",
      "        \n",
      "              * 'propagate': returns nan\n",
      "              * 'raise': throws an error\n",
      "              * 'omit': performs the calculations ignoring nan values\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        statistic : float or array\n",
      "            ``s^2 + k^2``, where ``s`` is the z-score returned by `skewtest` and\n",
      "            ``k`` is the z-score returned by `kurtosistest`.\n",
      "        pvalue : float or array\n",
      "           A 2-sided chi squared probability for the hypothesis test.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] D'Agostino, R. B. (1971), \"An omnibus test of normality for\n",
      "               moderate and large sample size\", Biometrika, 58, 341-348\n",
      "        \n",
      "        .. [2] D'Agostino, R. and Pearson, E. S. (1973), \"Tests for departure from\n",
      "               normality\", Biometrika, 60, 613-622\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy import stats\n",
      "        >>> rng = np.random.default_rng()\n",
      "        >>> pts = 1000\n",
      "        >>> a = rng.normal(0, 1, size=pts)\n",
      "        >>> b = rng.normal(2, 1, size=pts)\n",
      "        >>> x = np.concatenate((a, b))\n",
      "        >>> k2, p = stats.normaltest(x)\n",
      "        >>> alpha = 1e-3\n",
      "        >>> print(\"p = {:g}\".format(p))\n",
      "        p = 8.4713e-19\n",
      "        >>> if p < alpha:  # null hypothesis: x comes from a normal distribution\n",
      "        ...     print(\"The null hypothesis can be rejected\")\n",
      "        ... else:\n",
      "        ...     print(\"The null hypothesis cannot be rejected\")\n",
      "        The null hypothesis can be rejected\n",
      "    \n",
      "    obrientransform(*args)\n",
      "        Compute the O'Brien transform on input data (any number of arrays).\n",
      "        \n",
      "        Used to test for homogeneity of variance prior to running one-way stats.\n",
      "        Each array in ``*args`` is one level of a factor.\n",
      "        If `f_oneway` is run on the transformed data and found significant,\n",
      "        the variances are unequal.  From Maxwell and Delaney [1]_, p.112.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        args : tuple of array_like\n",
      "            Any number of arrays.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        obrientransform : ndarray\n",
      "            Transformed data for use in an ANOVA.  The first dimension\n",
      "            of the result corresponds to the sequence of transformed\n",
      "            arrays.  If the arrays given are all 1-D of the same length,\n",
      "            the return value is a 2-D array; otherwise it is a 1-D array\n",
      "            of type object, with each element being an ndarray.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] S. E. Maxwell and H. D. Delaney, \"Designing Experiments and\n",
      "               Analyzing Data: A Model Comparison Perspective\", Wadsworth, 1990.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        We'll test the following data sets for differences in their variance.\n",
      "        \n",
      "        >>> x = [10, 11, 13, 9, 7, 12, 12, 9, 10]\n",
      "        >>> y = [13, 21, 5, 10, 8, 14, 10, 12, 7, 15]\n",
      "        \n",
      "        Apply the O'Brien transform to the data.\n",
      "        \n",
      "        >>> from scipy.stats import obrientransform\n",
      "        >>> tx, ty = obrientransform(x, y)\n",
      "        \n",
      "        Use `scipy.stats.f_oneway` to apply a one-way ANOVA test to the\n",
      "        transformed data.\n",
      "        \n",
      "        >>> from scipy.stats import f_oneway\n",
      "        >>> F, p = f_oneway(tx, ty)\n",
      "        >>> p\n",
      "        0.1314139477040335\n",
      "        \n",
      "        If we require that ``p < 0.05`` for significance, we cannot conclude\n",
      "        that the variances are different.\n",
      "    \n",
      "    page_trend_test(data, ranked=False, predicted_ranks=None, method='auto')\n",
      "        Perform Page's Test, a measure of trend in observations between treatments.\n",
      "        \n",
      "        Page's Test (also known as Page's :math:`L` test) is useful when:\n",
      "        \n",
      "        * there are :math:`n \\geq 3` treatments,\n",
      "        * :math:`m \\geq 2` subjects are observed for each treatment, and\n",
      "        * the observations are hypothesized to have a particular order.\n",
      "        \n",
      "        Specifically, the test considers the null hypothesis that\n",
      "        \n",
      "        .. math::\n",
      "        \n",
      "            m_1 = m_2 = m_3 \\cdots = m_n,\n",
      "        \n",
      "        where :math:`m_j` is the mean of the observed quantity under treatment\n",
      "        :math:`j`, against the alternative hypothesis that\n",
      "        \n",
      "        .. math::\n",
      "        \n",
      "            m_1 \\leq m_2 \\leq m_3 \\leq \\cdots \\leq m_n,\n",
      "        \n",
      "        where at least one inequality is strict.\n",
      "        \n",
      "        As noted by [4]_, Page's :math:`L` test has greater statistical power than\n",
      "        the Friedman test against the alternative that there is a difference in\n",
      "        trend, as Friedman's test only considers a difference in the means of the\n",
      "        observations without considering their order. Whereas Spearman :math:`\\rho`\n",
      "        considers the correlation between the ranked observations of two variables\n",
      "        (e.g. the airspeed velocity of a swallow vs. the weight of the coconut it\n",
      "        carries), Page's :math:`L` is concerned with a trend in an observation\n",
      "        (e.g. the airspeed velocity of a swallow) across several distinct\n",
      "        treatments (e.g. carrying each of five coconuts of different weight) even\n",
      "        as the observation is repeated with multiple subjects (e.g. one European\n",
      "        swallow and one African swallow).\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        data : array-like\n",
      "            A :math:`m \\times n` array; the element in row :math:`i` and\n",
      "            column :math:`j` is the observation corresponding with subject\n",
      "            :math:`i` and treatment :math:`j`. By default, the columns are\n",
      "            assumed to be arranged in order of increasing predicted mean.\n",
      "        \n",
      "        ranked : boolean, optional\n",
      "            By default, `data` is assumed to be observations rather than ranks;\n",
      "            it will be ranked with `scipy.stats.rankdata` along ``axis=1``. If\n",
      "            `data` is provided in the form of ranks, pass argument ``True``.\n",
      "        \n",
      "        predicted_ranks : array-like, optional\n",
      "            The predicted ranks of the column means. If not specified,\n",
      "            the columns are assumed to be arranged in order of increasing\n",
      "            predicted mean, so the default `predicted_ranks` are\n",
      "            :math:`[1, 2, \\dots, n-1, n]`.\n",
      "        \n",
      "        method : {'auto', 'asymptotic', 'exact'}, optional\n",
      "            Selects the method used to calculate the *p*-value. The following\n",
      "            options are available.\n",
      "        \n",
      "            * 'auto': selects between 'exact' and 'asymptotic' to\n",
      "              achieve reasonably accurate results in reasonable time (default)\n",
      "            * 'asymptotic': compares the standardized test statistic against\n",
      "              the normal distribution\n",
      "            * 'exact': computes the exact *p*-value by comparing the observed\n",
      "              :math:`L` statistic against those realized by all possible\n",
      "              permutations of ranks (under the null hypothesis that each\n",
      "              permutation is equally likely)\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        res : PageTrendTestResult\n",
      "            An object containing attributes:\n",
      "        \n",
      "            statistic : float\n",
      "                Page's :math:`L` test statistic.\n",
      "            pvalue : float\n",
      "                The associated *p*-value\n",
      "            method : {'asymptotic', 'exact'}\n",
      "                The method used to compute the *p*-value\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        rankdata, friedmanchisquare, spearmanr\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        As noted in [1]_, \"the :math:`n` 'treatments' could just as well represent\n",
      "        :math:`n` objects or events or performances or persons or trials ranked.\"\n",
      "        Similarly, the :math:`m` 'subjects' could equally stand for :math:`m`\n",
      "        \"groupings by ability or some other control variable, or judges doing\n",
      "        the ranking, or random replications of some other sort.\"\n",
      "        \n",
      "        The procedure for calculating the :math:`L` statistic, adapted from\n",
      "        [1]_, is:\n",
      "        \n",
      "        1. \"Predetermine with careful logic the appropriate hypotheses\n",
      "           concerning the predicted ording of the experimental results.\n",
      "           If no reasonable basis for ordering any treatments is known, the\n",
      "           :math:`L` test is not appropriate.\"\n",
      "        2. \"As in other experiments, determine at what level of confidence\n",
      "           you will reject the null hypothesis that there is no agreement of\n",
      "           experimental results with the monotonic hypothesis.\"\n",
      "        3. \"Cast the experimental material into a two-way table of :math:`n`\n",
      "           columns (treatments, objects ranked, conditions) and :math:`m`\n",
      "           rows (subjects, replication groups, levels of control variables).\"\n",
      "        4. \"When experimental observations are recorded, rank them across each\n",
      "           row\", e.g. ``ranks = scipy.stats.rankdata(data, axis=1)``.\n",
      "        5. \"Add the ranks in each column\", e.g.\n",
      "           ``colsums = np.sum(ranks, axis=0)``.\n",
      "        6. \"Multiply each sum of ranks by the predicted rank for that same\n",
      "           column\", e.g. ``products = predicted_ranks * colsums``.\n",
      "        7. \"Sum all such products\", e.g. ``L = products.sum()``.\n",
      "        \n",
      "        [1]_ continues by suggesting use of the standardized statistic\n",
      "        \n",
      "        .. math::\n",
      "        \n",
      "            \\chi_L^2 = \\frac{\\left[12L-3mn(n+1)^2\\right]^2}{mn^2(n^2-1)(n+1)}\n",
      "        \n",
      "        \"which is distributed approximately as chi-square with 1 degree of\n",
      "        freedom. The ordinary use of :math:`\\chi^2` tables would be\n",
      "        equivalent to a two-sided test of agreement. If a one-sided test\n",
      "        is desired, *as will almost always be the case*, the probability\n",
      "        discovered in the chi-square table should be *halved*.\"\n",
      "        \n",
      "        However, this standardized statistic does not distinguish between the\n",
      "        observed values being well correlated with the predicted ranks and being\n",
      "        _anti_-correlated with the predicted ranks. Instead, we follow [2]_\n",
      "        and calculate the standardized statistic\n",
      "        \n",
      "        .. math::\n",
      "        \n",
      "            \\Lambda = \\frac{L - E_0}{\\sqrt{V_0}},\n",
      "        \n",
      "        where :math:`E_0 = \\frac{1}{4} mn(n+1)^2` and\n",
      "        :math:`V_0 = \\frac{1}{144} mn^2(n+1)(n^2-1)`, \"which is asymptotically\n",
      "        normal under the null hypothesis\".\n",
      "        \n",
      "        The *p*-value for ``method='exact'`` is generated by comparing the observed\n",
      "        value of :math:`L` against the :math:`L` values generated for all\n",
      "        :math:`(n!)^m` possible permutations of ranks. The calculation is performed\n",
      "        using the recursive method of [5].\n",
      "        \n",
      "        The *p*-values are not adjusted for the possibility of ties. When\n",
      "        ties are present, the reported  ``'exact'`` *p*-values may be somewhat\n",
      "        larger (i.e. more conservative) than the true *p*-value [2]_. The\n",
      "        ``'asymptotic'``` *p*-values, however, tend to be smaller (i.e. less\n",
      "        conservative) than the ``'exact'`` *p*-values.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] Ellis Batten Page, \"Ordered hypotheses for multiple treatments:\n",
      "           a significant test for linear ranks\", *Journal of the American\n",
      "           Statistical Association* 58(301), p. 216--230, 1963.\n",
      "        \n",
      "        .. [2] Markus Neuhauser, *Nonparametric Statistical Test: A computational\n",
      "           approach*, CRC Press, p. 150--152, 2012.\n",
      "        \n",
      "        .. [3] Statext LLC, \"Page's L Trend Test - Easy Statistics\", *Statext -\n",
      "           Statistics Study*, https://www.statext.com/practice/PageTrendTest03.php,\n",
      "           Accessed July 12, 2020.\n",
      "        \n",
      "        .. [4] \"Page's Trend Test\", *Wikipedia*, WikimediaFoundation,\n",
      "           https://en.wikipedia.org/wiki/Page%27s_trend_test,\n",
      "           Accessed July 12, 2020.\n",
      "        \n",
      "        .. [5] Robert E. Odeh, \"The exact distribution of Page's L-statistic in\n",
      "           the two-way layout\", *Communications in Statistics - Simulation and\n",
      "           Computation*,  6(1), p. 49--61, 1977.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        We use the example from [3]_: 10 students are asked to rate three\n",
      "        teaching methods - tutorial, lecture, and seminar - on a scale of 1-5,\n",
      "        with 1 being the lowest and 5 being the highest. We have decided that\n",
      "        a confidence level of 99% is required to reject the null hypothesis in\n",
      "        favor of our alternative: that the seminar will have the highest ratings\n",
      "        and the tutorial will have the lowest. Initially, the data have been\n",
      "        tabulated with each row representing an individual student's ratings of\n",
      "        the three methods in the following order: tutorial, lecture, seminar.\n",
      "        \n",
      "        >>> table = [[3, 4, 3],\n",
      "        ...          [2, 2, 4],\n",
      "        ...          [3, 3, 5],\n",
      "        ...          [1, 3, 2],\n",
      "        ...          [2, 3, 2],\n",
      "        ...          [2, 4, 5],\n",
      "        ...          [1, 2, 4],\n",
      "        ...          [3, 4, 4],\n",
      "        ...          [2, 4, 5],\n",
      "        ...          [1, 3, 4]]\n",
      "        \n",
      "        Because the tutorial is hypothesized to have the lowest ratings, the\n",
      "        column corresponding with tutorial rankings should be first; the seminar\n",
      "        is hypothesized to have the highest ratings, so its column should be last.\n",
      "        Since the columns are already arranged in this order of increasing\n",
      "        predicted mean, we can pass the table directly into `page_trend_test`.\n",
      "        \n",
      "        >>> from scipy.stats import page_trend_test\n",
      "        >>> res = page_trend_test(table)\n",
      "        >>> res\n",
      "        PageTrendTestResult(statistic=133.5, pvalue=0.0018191161948127822,\n",
      "                            method='exact')\n",
      "        \n",
      "        This *p*-value indicates that there is a 0.1819% chance that\n",
      "        the :math:`L` statistic would reach such an extreme value under the null\n",
      "        hypothesis. Because 0.1819% is less than 1%, we have evidence to reject\n",
      "        the null hypothesis in favor of our alternative at a 99% confidence level.\n",
      "        \n",
      "        The value of the :math:`L` statistic is 133.5. To check this manually,\n",
      "        we rank the data such that high scores correspond with high ranks, settling\n",
      "        ties with an average rank:\n",
      "        \n",
      "        >>> from scipy.stats import rankdata\n",
      "        >>> ranks = rankdata(table, axis=1)\n",
      "        >>> ranks\n",
      "        array([[1.5, 3. , 1.5],\n",
      "               [1.5, 1.5, 3. ],\n",
      "               [1.5, 1.5, 3. ],\n",
      "               [1. , 3. , 2. ],\n",
      "               [1.5, 3. , 1.5],\n",
      "               [1. , 2. , 3. ],\n",
      "               [1. , 2. , 3. ],\n",
      "               [1. , 2.5, 2.5],\n",
      "               [1. , 2. , 3. ],\n",
      "               [1. , 2. , 3. ]])\n",
      "        \n",
      "        We add the ranks within each column, multiply the sums by the\n",
      "        predicted ranks, and sum the products.\n",
      "        \n",
      "        >>> import numpy as np\n",
      "        >>> m, n = ranks.shape\n",
      "        >>> predicted_ranks = np.arange(1, n+1)\n",
      "        >>> L = (predicted_ranks * np.sum(ranks, axis=0)).sum()\n",
      "        >>> res.statistic == L\n",
      "        True\n",
      "        \n",
      "        As presented in [3]_, the asymptotic approximation of the *p*-value is the\n",
      "        survival function of the normal distribution evaluated at the standardized\n",
      "        test statistic:\n",
      "        \n",
      "        >>> from scipy.stats import norm\n",
      "        >>> E0 = (m*n*(n+1)**2)/4\n",
      "        >>> V0 = (m*n**2*(n+1)*(n**2-1))/144\n",
      "        >>> Lambda = (L-E0)/np.sqrt(V0)\n",
      "        >>> p = norm.sf(Lambda)\n",
      "        >>> p\n",
      "        0.0012693433690751756\n",
      "        \n",
      "        This does not precisely match the *p*-value reported by `page_trend_test`\n",
      "        above. The asymptotic distribution is not very accurate, nor conservative,\n",
      "        for :math:`m \\leq 12` and :math:`n \\leq 8`, so `page_trend_test` chose to\n",
      "        use ``method='exact'`` based on the dimensions of the table and the\n",
      "        recommendations in Page's original paper [1]_. To override\n",
      "        `page_trend_test`'s choice, provide the `method` argument.\n",
      "        \n",
      "        >>> res = page_trend_test(table, method=\"asymptotic\")\n",
      "        >>> res\n",
      "        PageTrendTestResult(statistic=133.5, pvalue=0.0012693433690751756,\n",
      "                            method='asymptotic')\n",
      "        \n",
      "        If the data are already ranked, we can pass in the ``ranks`` instead of\n",
      "        the ``table`` to save computation time.\n",
      "        \n",
      "        >>> res = page_trend_test(ranks,             # ranks of data\n",
      "        ...                       ranked=True,       # data is already ranked\n",
      "        ...                       )\n",
      "        >>> res\n",
      "        PageTrendTestResult(statistic=133.5, pvalue=0.0018191161948127822,\n",
      "                            method='exact')\n",
      "        \n",
      "        Suppose the raw data had been tabulated in an order different from the\n",
      "        order of predicted means, say lecture, seminar, tutorial.\n",
      "        \n",
      "        >>> table = np.asarray(table)[:, [1, 2, 0]]\n",
      "        \n",
      "        Since the arrangement of this table is not consistent with the assumed\n",
      "        ordering, we can either rearrange the table or provide the\n",
      "        `predicted_ranks`. Remembering that the lecture is predicted\n",
      "        to have the middle rank, the seminar the highest, and tutorial the lowest,\n",
      "        we pass:\n",
      "        \n",
      "        >>> res = page_trend_test(table,             # data as originally tabulated\n",
      "        ...                       predicted_ranks=[2, 3, 1],  # our predicted order\n",
      "        ...                       )\n",
      "        >>> res\n",
      "        PageTrendTestResult(statistic=133.5, pvalue=0.0018191161948127822,\n",
      "                            method='exact')\n",
      "    \n",
      "    pearsonr(x, y)\n",
      "        Pearson correlation coefficient and p-value for testing non-correlation.\n",
      "        \n",
      "        The Pearson correlation coefficient [1]_ measures the linear relationship\n",
      "        between two datasets.  The calculation of the p-value relies on the\n",
      "        assumption that each dataset is normally distributed.  (See Kowalski [3]_\n",
      "        for a discussion of the effects of non-normality of the input on the\n",
      "        distribution of the correlation coefficient.)  Like other correlation\n",
      "        coefficients, this one varies between -1 and +1 with 0 implying no\n",
      "        correlation. Correlations of -1 or +1 imply an exact linear relationship.\n",
      "        Positive correlations imply that as x increases, so does y. Negative\n",
      "        correlations imply that as x increases, y decreases.\n",
      "        \n",
      "        The p-value roughly indicates the probability of an uncorrelated system\n",
      "        producing datasets that have a Pearson correlation at least as extreme\n",
      "        as the one computed from these datasets.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x : (N,) array_like\n",
      "            Input array.\n",
      "        y : (N,) array_like\n",
      "            Input array.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        r : float\n",
      "            Pearson's correlation coefficient.\n",
      "        p-value : float\n",
      "            Two-tailed p-value.\n",
      "        \n",
      "        Warns\n",
      "        -----\n",
      "        PearsonRConstantInputWarning\n",
      "            Raised if an input is a constant array.  The correlation coefficient\n",
      "            is not defined in this case, so ``np.nan`` is returned.\n",
      "        \n",
      "        PearsonRNearConstantInputWarning\n",
      "            Raised if an input is \"nearly\" constant.  The array ``x`` is considered\n",
      "            nearly constant if ``norm(x - mean(x)) < 1e-13 * abs(mean(x))``.\n",
      "            Numerical errors in the calculation ``x - mean(x)`` in this case might\n",
      "            result in an inaccurate calculation of r.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        spearmanr : Spearman rank-order correlation coefficient.\n",
      "        kendalltau : Kendall's tau, a correlation measure for ordinal data.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        The correlation coefficient is calculated as follows:\n",
      "        \n",
      "        .. math::\n",
      "        \n",
      "            r = \\frac{\\sum (x - m_x) (y - m_y)}\n",
      "                     {\\sqrt{\\sum (x - m_x)^2 \\sum (y - m_y)^2}}\n",
      "        \n",
      "        where :math:`m_x` is the mean of the vector :math:`x` and :math:`m_y` is\n",
      "        the mean of the vector :math:`y`.\n",
      "        \n",
      "        Under the assumption that :math:`x` and :math:`m_y` are drawn from\n",
      "        independent normal distributions (so the population correlation coefficient\n",
      "        is 0), the probability density function of the sample correlation\n",
      "        coefficient :math:`r` is ([1]_, [2]_):\n",
      "        \n",
      "        .. math::\n",
      "        \n",
      "            f(r) = \\frac{{(1-r^2)}^{n/2-2}}{\\mathrm{B}(\\frac{1}{2},\\frac{n}{2}-1)}\n",
      "        \n",
      "        where n is the number of samples, and B is the beta function.  This\n",
      "        is sometimes referred to as the exact distribution of r.  This is\n",
      "        the distribution that is used in `pearsonr` to compute the p-value.\n",
      "        The distribution is a beta distribution on the interval [-1, 1],\n",
      "        with equal shape parameters a = b = n/2 - 1.  In terms of SciPy's\n",
      "        implementation of the beta distribution, the distribution of r is::\n",
      "        \n",
      "            dist = scipy.stats.beta(n/2 - 1, n/2 - 1, loc=-1, scale=2)\n",
      "        \n",
      "        The p-value returned by `pearsonr` is a two-sided p-value.  For a\n",
      "        given sample with correlation coefficient r, the p-value is\n",
      "        the probability that abs(r') of a random sample x' and y' drawn from\n",
      "        the population with zero correlation would be greater than or equal\n",
      "        to abs(r).  In terms of the object ``dist`` shown above, the p-value\n",
      "        for a given r and length n can be computed as::\n",
      "        \n",
      "            p = 2*dist.cdf(-abs(r))\n",
      "        \n",
      "        When n is 2, the above continuous distribution is not well-defined.\n",
      "        One can interpret the limit of the beta distribution as the shape\n",
      "        parameters a and b approach a = b = 0 as a discrete distribution with\n",
      "        equal probability masses at r = 1 and r = -1.  More directly, one\n",
      "        can observe that, given the data x = [x1, x2] and y = [y1, y2], and\n",
      "        assuming x1 != x2 and y1 != y2, the only possible values for r are 1\n",
      "        and -1.  Because abs(r') for any sample x' and y' with length 2 will\n",
      "        be 1, the two-sided p-value for a sample of length 2 is always 1.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] \"Pearson correlation coefficient\", Wikipedia,\n",
      "               https://en.wikipedia.org/wiki/Pearson_correlation_coefficient\n",
      "        .. [2] Student, \"Probable error of a correlation coefficient\",\n",
      "               Biometrika, Volume 6, Issue 2-3, 1 September 1908, pp. 302-310.\n",
      "        .. [3] C. J. Kowalski, \"On the Effects of Non-Normality on the Distribution\n",
      "               of the Sample Product-Moment Correlation Coefficient\"\n",
      "               Journal of the Royal Statistical Society. Series C (Applied\n",
      "               Statistics), Vol. 21, No. 1 (1972), pp. 1-12.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy import stats\n",
      "        >>> a = np.array([0, 0, 0, 1, 1, 1, 1])\n",
      "        >>> b = np.arange(7)\n",
      "        >>> stats.pearsonr(a, b)\n",
      "        (0.8660254037844386, 0.011724811003954649)\n",
      "        \n",
      "        >>> stats.pearsonr([1, 2, 3, 4, 5], [10, 9, 2.5, 6, 4])\n",
      "        (-0.7426106572325057, 0.1505558088534455)\n",
      "    \n",
      "    percentileofscore(a, score, kind='rank')\n",
      "        Compute the percentile rank of a score relative to a list of scores.\n",
      "        \n",
      "        A `percentileofscore` of, for example, 80% means that 80% of the\n",
      "        scores in `a` are below the given score. In the case of gaps or\n",
      "        ties, the exact definition depends on the optional keyword, `kind`.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        a : array_like\n",
      "            Array of scores to which `score` is compared.\n",
      "        score : int or float\n",
      "            Score that is compared to the elements in `a`.\n",
      "        kind : {'rank', 'weak', 'strict', 'mean'}, optional\n",
      "            Specifies the interpretation of the resulting score.\n",
      "            The following options are available (default is 'rank'):\n",
      "        \n",
      "              * 'rank': Average percentage ranking of score.  In case of multiple\n",
      "                matches, average the percentage rankings of all matching scores.\n",
      "              * 'weak': This kind corresponds to the definition of a cumulative\n",
      "                distribution function.  A percentileofscore of 80% means that 80%\n",
      "                of values are less than or equal to the provided score.\n",
      "              * 'strict': Similar to \"weak\", except that only values that are\n",
      "                strictly less than the given score are counted.\n",
      "              * 'mean': The average of the \"weak\" and \"strict\" scores, often used\n",
      "                in testing.  See https://en.wikipedia.org/wiki/Percentile_rank\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        pcos : float\n",
      "            Percentile-position of score (0-100) relative to `a`.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        numpy.percentile\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        Three-quarters of the given values lie below a given score:\n",
      "        \n",
      "        >>> from scipy import stats\n",
      "        >>> stats.percentileofscore([1, 2, 3, 4], 3)\n",
      "        75.0\n",
      "        \n",
      "        With multiple matches, note how the scores of the two matches, 0.6\n",
      "        and 0.8 respectively, are averaged:\n",
      "        \n",
      "        >>> stats.percentileofscore([1, 2, 3, 3, 4], 3)\n",
      "        70.0\n",
      "        \n",
      "        Only 2/5 values are strictly less than 3:\n",
      "        \n",
      "        >>> stats.percentileofscore([1, 2, 3, 3, 4], 3, kind='strict')\n",
      "        40.0\n",
      "        \n",
      "        But 4/5 values are less than or equal to 3:\n",
      "        \n",
      "        >>> stats.percentileofscore([1, 2, 3, 3, 4], 3, kind='weak')\n",
      "        80.0\n",
      "        \n",
      "        The average between the weak and the strict scores is:\n",
      "        \n",
      "        >>> stats.percentileofscore([1, 2, 3, 3, 4], 3, kind='mean')\n",
      "        60.0\n",
      "    \n",
      "    pointbiserialr(x, y)\n",
      "        Calculate a point biserial correlation coefficient and its p-value.\n",
      "        \n",
      "        The point biserial correlation is used to measure the relationship\n",
      "        between a binary variable, x, and a continuous variable, y. Like other\n",
      "        correlation coefficients, this one varies between -1 and +1 with 0\n",
      "        implying no correlation. Correlations of -1 or +1 imply a determinative\n",
      "        relationship.\n",
      "        \n",
      "        This function uses a shortcut formula but produces the same result as\n",
      "        `pearsonr`.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x : array_like of bools\n",
      "            Input array.\n",
      "        y : array_like\n",
      "            Input array.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        correlation : float\n",
      "            R value.\n",
      "        pvalue : float\n",
      "            Two-sided p-value.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        `pointbiserialr` uses a t-test with ``n-1`` degrees of freedom.\n",
      "        It is equivalent to `pearsonr`.\n",
      "        \n",
      "        The value of the point-biserial correlation can be calculated from:\n",
      "        \n",
      "        .. math::\n",
      "        \n",
      "            r_{pb} = \\frac{\\overline{Y_{1}} -\n",
      "                     \\overline{Y_{0}}}{s_{y}}\\sqrt{\\frac{N_{1} N_{2}}{N (N - 1))}}\n",
      "        \n",
      "        Where :math:`Y_{0}` and :math:`Y_{1}` are means of the metric\n",
      "        observations coded 0 and 1 respectively; :math:`N_{0}` and :math:`N_{1}`\n",
      "        are number of observations coded 0 and 1 respectively; :math:`N` is the\n",
      "        total number of observations and :math:`s_{y}` is the standard\n",
      "        deviation of all the metric observations.\n",
      "        \n",
      "        A value of :math:`r_{pb}` that is significantly different from zero is\n",
      "        completely equivalent to a significant difference in means between the two\n",
      "        groups. Thus, an independent groups t Test with :math:`N-2` degrees of\n",
      "        freedom may be used to test whether :math:`r_{pb}` is nonzero. The\n",
      "        relation between the t-statistic for comparing two independent groups and\n",
      "        :math:`r_{pb}` is given by:\n",
      "        \n",
      "        .. math::\n",
      "        \n",
      "            t = \\sqrt{N - 2}\\frac{r_{pb}}{\\sqrt{1 - r^{2}_{pb}}}\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] J. Lev, \"The Point Biserial Coefficient of Correlation\", Ann. Math.\n",
      "               Statist., Vol. 20, no.1, pp. 125-126, 1949.\n",
      "        \n",
      "        .. [2] R.F. Tate, \"Correlation Between a Discrete and a Continuous\n",
      "               Variable. Point-Biserial Correlation.\", Ann. Math. Statist., Vol. 25,\n",
      "               np. 3, pp. 603-607, 1954.\n",
      "        \n",
      "        .. [3] D. Kornbrot \"Point Biserial Correlation\", In Wiley StatsRef:\n",
      "               Statistics Reference Online (eds N. Balakrishnan, et al.), 2014.\n",
      "               :doi:`10.1002/9781118445112.stat06227`\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy import stats\n",
      "        >>> a = np.array([0, 0, 0, 1, 1, 1, 1])\n",
      "        >>> b = np.arange(7)\n",
      "        >>> stats.pointbiserialr(a, b)\n",
      "        (0.8660254037844386, 0.011724811003954652)\n",
      "        >>> stats.pearsonr(a, b)\n",
      "        (0.86602540378443871, 0.011724811003954626)\n",
      "        >>> np.corrcoef(a, b)\n",
      "        array([[ 1.       ,  0.8660254],\n",
      "               [ 0.8660254,  1.       ]])\n",
      "    \n",
      "    power_divergence(f_obs, f_exp=None, ddof=0, axis=0, lambda_=None)\n",
      "        Cressie-Read power divergence statistic and goodness of fit test.\n",
      "        \n",
      "        This function tests the null hypothesis that the categorical data\n",
      "        has the given frequencies, using the Cressie-Read power divergence\n",
      "        statistic.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        f_obs : array_like\n",
      "            Observed frequencies in each category.\n",
      "        f_exp : array_like, optional\n",
      "            Expected frequencies in each category.  By default the categories are\n",
      "            assumed to be equally likely.\n",
      "        ddof : int, optional\n",
      "            \"Delta degrees of freedom\": adjustment to the degrees of freedom\n",
      "            for the p-value.  The p-value is computed using a chi-squared\n",
      "            distribution with ``k - 1 - ddof`` degrees of freedom, where `k`\n",
      "            is the number of observed frequencies.  The default value of `ddof`\n",
      "            is 0.\n",
      "        axis : int or None, optional\n",
      "            The axis of the broadcast result of `f_obs` and `f_exp` along which to\n",
      "            apply the test.  If axis is None, all values in `f_obs` are treated\n",
      "            as a single data set.  Default is 0.\n",
      "        lambda_ : float or str, optional\n",
      "            The power in the Cressie-Read power divergence statistic.  The default\n",
      "            is 1.  For convenience, `lambda_` may be assigned one of the following\n",
      "            strings, in which case the corresponding numerical value is used::\n",
      "        \n",
      "                String              Value   Description\n",
      "                \"pearson\"             1     Pearson's chi-squared statistic.\n",
      "                                            In this case, the function is\n",
      "                                            equivalent to `stats.chisquare`.\n",
      "                \"log-likelihood\"      0     Log-likelihood ratio. Also known as\n",
      "                                            the G-test [3]_.\n",
      "                \"freeman-tukey\"      -1/2   Freeman-Tukey statistic.\n",
      "                \"mod-log-likelihood\" -1     Modified log-likelihood ratio.\n",
      "                \"neyman\"             -2     Neyman's statistic.\n",
      "                \"cressie-read\"        2/3   The power recommended in [5]_.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        statistic : float or ndarray\n",
      "            The Cressie-Read power divergence test statistic.  The value is\n",
      "            a float if `axis` is None or if` `f_obs` and `f_exp` are 1-D.\n",
      "        pvalue : float or ndarray\n",
      "            The p-value of the test.  The value is a float if `ddof` and the\n",
      "            return value `stat` are scalars.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        chisquare\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        This test is invalid when the observed or expected frequencies in each\n",
      "        category are too small.  A typical rule is that all of the observed\n",
      "        and expected frequencies should be at least 5.\n",
      "        \n",
      "        Also, the sum of the observed and expected frequencies must be the same\n",
      "        for the test to be valid; `power_divergence` raises an error if the sums\n",
      "        do not agree within a relative tolerance of ``1e-8``.\n",
      "        \n",
      "        When `lambda_` is less than zero, the formula for the statistic involves\n",
      "        dividing by `f_obs`, so a warning or error may be generated if any value\n",
      "        in `f_obs` is 0.\n",
      "        \n",
      "        Similarly, a warning or error may be generated if any value in `f_exp` is\n",
      "        zero when `lambda_` >= 0.\n",
      "        \n",
      "        The default degrees of freedom, k-1, are for the case when no parameters\n",
      "        of the distribution are estimated. If p parameters are estimated by\n",
      "        efficient maximum likelihood then the correct degrees of freedom are\n",
      "        k-1-p. If the parameters are estimated in a different way, then the\n",
      "        dof can be between k-1-p and k-1. However, it is also possible that\n",
      "        the asymptotic distribution is not a chisquare, in which case this\n",
      "        test is not appropriate.\n",
      "        \n",
      "        This function handles masked arrays.  If an element of `f_obs` or `f_exp`\n",
      "        is masked, then data at that position is ignored, and does not count\n",
      "        towards the size of the data set.\n",
      "        \n",
      "        .. versionadded:: 0.13.0\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] Lowry, Richard.  \"Concepts and Applications of Inferential\n",
      "               Statistics\". Chapter 8.\n",
      "               https://web.archive.org/web/20171015035606/http://faculty.vassar.edu/lowry/ch8pt1.html\n",
      "        .. [2] \"Chi-squared test\", https://en.wikipedia.org/wiki/Chi-squared_test\n",
      "        .. [3] \"G-test\", https://en.wikipedia.org/wiki/G-test\n",
      "        .. [4] Sokal, R. R. and Rohlf, F. J. \"Biometry: the principles and\n",
      "               practice of statistics in biological research\", New York: Freeman\n",
      "               (1981)\n",
      "        .. [5] Cressie, N. and Read, T. R. C., \"Multinomial Goodness-of-Fit\n",
      "               Tests\", J. Royal Stat. Soc. Series B, Vol. 46, No. 3 (1984),\n",
      "               pp. 440-464.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        (See `chisquare` for more examples.)\n",
      "        \n",
      "        When just `f_obs` is given, it is assumed that the expected frequencies\n",
      "        are uniform and given by the mean of the observed frequencies.  Here we\n",
      "        perform a G-test (i.e. use the log-likelihood ratio statistic):\n",
      "        \n",
      "        >>> from scipy.stats import power_divergence\n",
      "        >>> power_divergence([16, 18, 16, 14, 12, 12], lambda_='log-likelihood')\n",
      "        (2.006573162632538, 0.84823476779463769)\n",
      "        \n",
      "        The expected frequencies can be given with the `f_exp` argument:\n",
      "        \n",
      "        >>> power_divergence([16, 18, 16, 14, 12, 12],\n",
      "        ...                  f_exp=[16, 16, 16, 16, 16, 8],\n",
      "        ...                  lambda_='log-likelihood')\n",
      "        (3.3281031458963746, 0.6495419288047497)\n",
      "        \n",
      "        When `f_obs` is 2-D, by default the test is applied to each column.\n",
      "        \n",
      "        >>> obs = np.array([[16, 18, 16, 14, 12, 12], [32, 24, 16, 28, 20, 24]]).T\n",
      "        >>> obs.shape\n",
      "        (6, 2)\n",
      "        >>> power_divergence(obs, lambda_=\"log-likelihood\")\n",
      "        (array([ 2.00657316,  6.77634498]), array([ 0.84823477,  0.23781225]))\n",
      "        \n",
      "        By setting ``axis=None``, the test is applied to all data in the array,\n",
      "        which is equivalent to applying the test to the flattened array.\n",
      "        \n",
      "        >>> power_divergence(obs, axis=None)\n",
      "        (23.31034482758621, 0.015975692534127565)\n",
      "        >>> power_divergence(obs.ravel())\n",
      "        (23.31034482758621, 0.015975692534127565)\n",
      "        \n",
      "        `ddof` is the change to make to the default degrees of freedom.\n",
      "        \n",
      "        >>> power_divergence([16, 18, 16, 14, 12, 12], ddof=1)\n",
      "        (2.0, 0.73575888234288467)\n",
      "        \n",
      "        The calculation of the p-values is done by broadcasting the\n",
      "        test statistic with `ddof`.\n",
      "        \n",
      "        >>> power_divergence([16, 18, 16, 14, 12, 12], ddof=[0,1,2])\n",
      "        (2.0, array([ 0.84914504,  0.73575888,  0.5724067 ]))\n",
      "        \n",
      "        `f_obs` and `f_exp` are also broadcast.  In the following, `f_obs` has\n",
      "        shape (6,) and `f_exp` has shape (2, 6), so the result of broadcasting\n",
      "        `f_obs` and `f_exp` has shape (2, 6).  To compute the desired chi-squared\n",
      "        statistics, we must use ``axis=1``:\n",
      "        \n",
      "        >>> power_divergence([16, 18, 16, 14, 12, 12],\n",
      "        ...                  f_exp=[[16, 16, 16, 16, 16, 8],\n",
      "        ...                         [8, 20, 20, 16, 12, 12]],\n",
      "        ...                  axis=1)\n",
      "        (array([ 3.5 ,  9.25]), array([ 0.62338763,  0.09949846]))\n",
      "    \n",
      "    ppcc_max(x, brack=(0.0, 1.0), dist='tukeylambda')\n",
      "        Calculate the shape parameter that maximizes the PPCC.\n",
      "        \n",
      "        The probability plot correlation coefficient (PPCC) plot can be used\n",
      "        to determine the optimal shape parameter for a one-parameter family\n",
      "        of distributions. ``ppcc_max`` returns the shape parameter that would\n",
      "        maximize the probability plot correlation coefficient for the given\n",
      "        data to a one-parameter family of distributions.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x : array_like\n",
      "            Input array.\n",
      "        brack : tuple, optional\n",
      "            Triple (a,b,c) where (a<b<c). If bracket consists of two numbers (a, c)\n",
      "            then they are assumed to be a starting interval for a downhill bracket\n",
      "            search (see `scipy.optimize.brent`).\n",
      "        dist : str or stats.distributions instance, optional\n",
      "            Distribution or distribution function name.  Objects that look enough\n",
      "            like a stats.distributions instance (i.e. they have a ``ppf`` method)\n",
      "            are also accepted.  The default is ``'tukeylambda'``.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        shape_value : float\n",
      "            The shape parameter at which the probability plot correlation\n",
      "            coefficient reaches its max value.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        ppcc_plot, probplot, boxcox\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        The brack keyword serves as a starting point which is useful in corner\n",
      "        cases. One can use a plot to obtain a rough visual estimate of the location\n",
      "        for the maximum to start the search near it.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] J.J. Filliben, \"The Probability Plot Correlation Coefficient Test\n",
      "               for Normality\", Technometrics, Vol. 17, pp. 111-117, 1975.\n",
      "        .. [2] Engineering Statistics Handbook, NIST/SEMATEC,\n",
      "               https://www.itl.nist.gov/div898/handbook/eda/section3/ppccplot.htm\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        First we generate some random data from a Weibull distribution\n",
      "        with shape parameter 2.5:\n",
      "        \n",
      "        >>> from scipy import stats\n",
      "        >>> import matplotlib.pyplot as plt\n",
      "        >>> rng = np.random.default_rng()\n",
      "        >>> c = 2.5\n",
      "        >>> x = stats.weibull_min.rvs(c, scale=4, size=2000, random_state=rng)\n",
      "        \n",
      "        Generate the PPCC plot for this data with the Weibull distribution.\n",
      "        \n",
      "        >>> fig, ax = plt.subplots(figsize=(8, 6))\n",
      "        >>> res = stats.ppcc_plot(x, c/2, 2*c, dist='weibull_min', plot=ax)\n",
      "        \n",
      "        We calculate the value where the shape should reach its maximum and a\n",
      "        red line is drawn there. The line should coincide with the highest\n",
      "        point in the PPCC graph.\n",
      "        \n",
      "        >>> cmax = stats.ppcc_max(x, brack=(c/2, 2*c), dist='weibull_min')\n",
      "        >>> ax.axvline(cmax, color='r')\n",
      "        >>> plt.show()\n",
      "    \n",
      "    ppcc_plot(x, a, b, dist='tukeylambda', plot=None, N=80)\n",
      "        Calculate and optionally plot probability plot correlation coefficient.\n",
      "        \n",
      "        The probability plot correlation coefficient (PPCC) plot can be used to\n",
      "        determine the optimal shape parameter for a one-parameter family of\n",
      "        distributions.  It cannot be used for distributions without shape\n",
      "        parameters\n",
      "        (like the normal distribution) or with multiple shape parameters.\n",
      "        \n",
      "        By default a Tukey-Lambda distribution (`stats.tukeylambda`) is used. A\n",
      "        Tukey-Lambda PPCC plot interpolates from long-tailed to short-tailed\n",
      "        distributions via an approximately normal one, and is therefore\n",
      "        particularly useful in practice.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x : array_like\n",
      "            Input array.\n",
      "        a, b : scalar\n",
      "            Lower and upper bounds of the shape parameter to use.\n",
      "        dist : str or stats.distributions instance, optional\n",
      "            Distribution or distribution function name.  Objects that look enough\n",
      "            like a stats.distributions instance (i.e. they have a ``ppf`` method)\n",
      "            are also accepted.  The default is ``'tukeylambda'``.\n",
      "        plot : object, optional\n",
      "            If given, plots PPCC against the shape parameter.\n",
      "            `plot` is an object that has to have methods \"plot\" and \"text\".\n",
      "            The `matplotlib.pyplot` module or a Matplotlib Axes object can be used,\n",
      "            or a custom object with the same methods.\n",
      "            Default is None, which means that no plot is created.\n",
      "        N : int, optional\n",
      "            Number of points on the horizontal axis (equally distributed from\n",
      "            `a` to `b`).\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        svals : ndarray\n",
      "            The shape values for which `ppcc` was calculated.\n",
      "        ppcc : ndarray\n",
      "            The calculated probability plot correlation coefficient values.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        ppcc_max, probplot, boxcox_normplot, tukeylambda\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        J.J. Filliben, \"The Probability Plot Correlation Coefficient Test for\n",
      "        Normality\", Technometrics, Vol. 17, pp. 111-117, 1975.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        First we generate some random data from a Weibull distribution\n",
      "        with shape parameter 2.5, and plot the histogram of the data:\n",
      "        \n",
      "        >>> from scipy import stats\n",
      "        >>> import matplotlib.pyplot as plt\n",
      "        >>> rng = np.random.default_rng()\n",
      "        >>> c = 2.5\n",
      "        >>> x = stats.weibull_min.rvs(c, scale=4, size=2000, random_state=rng)\n",
      "        \n",
      "        Take a look at the histogram of the data.\n",
      "        \n",
      "        >>> fig1, ax = plt.subplots(figsize=(9, 4))\n",
      "        >>> ax.hist(x, bins=50)\n",
      "        >>> ax.set_title('Histogram of x')\n",
      "        >>> plt.show()\n",
      "        \n",
      "        Now we explore this data with a PPCC plot as well as the related\n",
      "        probability plot and Box-Cox normplot.  A red line is drawn where we\n",
      "        expect the PPCC value to be maximal (at the shape parameter ``c``\n",
      "        used above):\n",
      "        \n",
      "        >>> fig2 = plt.figure(figsize=(12, 4))\n",
      "        >>> ax1 = fig2.add_subplot(1, 3, 1)\n",
      "        >>> ax2 = fig2.add_subplot(1, 3, 2)\n",
      "        >>> ax3 = fig2.add_subplot(1, 3, 3)\n",
      "        >>> res = stats.probplot(x, plot=ax1)\n",
      "        >>> res = stats.boxcox_normplot(x, -4, 4, plot=ax2)\n",
      "        >>> res = stats.ppcc_plot(x, c/2, 2*c, dist='weibull_min', plot=ax3)\n",
      "        >>> ax3.axvline(c, color='r')\n",
      "        >>> plt.show()\n",
      "    \n",
      "    probplot(x, sparams=(), dist='norm', fit=True, plot=None, rvalue=False)\n",
      "        Calculate quantiles for a probability plot, and optionally show the plot.\n",
      "        \n",
      "        Generates a probability plot of sample data against the quantiles of a\n",
      "        specified theoretical distribution (the normal distribution by default).\n",
      "        `probplot` optionally calculates a best-fit line for the data and plots the\n",
      "        results using Matplotlib or a given plot function.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x : array_like\n",
      "            Sample/response data from which `probplot` creates the plot.\n",
      "        sparams : tuple, optional\n",
      "            Distribution-specific shape parameters (shape parameters plus location\n",
      "            and scale).\n",
      "        dist : str or stats.distributions instance, optional\n",
      "            Distribution or distribution function name. The default is 'norm' for a\n",
      "            normal probability plot.  Objects that look enough like a\n",
      "            stats.distributions instance (i.e. they have a ``ppf`` method) are also\n",
      "            accepted.\n",
      "        fit : bool, optional\n",
      "            Fit a least-squares regression (best-fit) line to the sample data if\n",
      "            True (default).\n",
      "        plot : object, optional\n",
      "            If given, plots the quantiles.\n",
      "            If given and `fit` is True, also plots the least squares fit.\n",
      "            `plot` is an object that has to have methods \"plot\" and \"text\".\n",
      "            The `matplotlib.pyplot` module or a Matplotlib Axes object can be used,\n",
      "            or a custom object with the same methods.\n",
      "            Default is None, which means that no plot is created.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        (osm, osr) : tuple of ndarrays\n",
      "            Tuple of theoretical quantiles (osm, or order statistic medians) and\n",
      "            ordered responses (osr).  `osr` is simply sorted input `x`.\n",
      "            For details on how `osm` is calculated see the Notes section.\n",
      "        (slope, intercept, r) : tuple of floats, optional\n",
      "            Tuple  containing the result of the least-squares fit, if that is\n",
      "            performed by `probplot`. `r` is the square root of the coefficient of\n",
      "            determination.  If ``fit=False`` and ``plot=None``, this tuple is not\n",
      "            returned.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        Even if `plot` is given, the figure is not shown or saved by `probplot`;\n",
      "        ``plt.show()`` or ``plt.savefig('figname.png')`` should be used after\n",
      "        calling `probplot`.\n",
      "        \n",
      "        `probplot` generates a probability plot, which should not be confused with\n",
      "        a Q-Q or a P-P plot.  Statsmodels has more extensive functionality of this\n",
      "        type, see ``statsmodels.api.ProbPlot``.\n",
      "        \n",
      "        The formula used for the theoretical quantiles (horizontal axis of the\n",
      "        probability plot) is Filliben's estimate::\n",
      "        \n",
      "            quantiles = dist.ppf(val), for\n",
      "        \n",
      "                    0.5**(1/n),                  for i = n\n",
      "              val = (i - 0.3175) / (n + 0.365),  for i = 2, ..., n-1\n",
      "                    1 - 0.5**(1/n),              for i = 1\n",
      "        \n",
      "        where ``i`` indicates the i-th ordered value and ``n`` is the total number\n",
      "        of values.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy import stats\n",
      "        >>> import matplotlib.pyplot as plt\n",
      "        >>> nsample = 100\n",
      "        >>> rng = np.random.default_rng()\n",
      "        \n",
      "        A t distribution with small degrees of freedom:\n",
      "        \n",
      "        >>> ax1 = plt.subplot(221)\n",
      "        >>> x = stats.t.rvs(3, size=nsample, random_state=rng)\n",
      "        >>> res = stats.probplot(x, plot=plt)\n",
      "        \n",
      "        A t distribution with larger degrees of freedom:\n",
      "        \n",
      "        >>> ax2 = plt.subplot(222)\n",
      "        >>> x = stats.t.rvs(25, size=nsample, random_state=rng)\n",
      "        >>> res = stats.probplot(x, plot=plt)\n",
      "        \n",
      "        A mixture of two normal distributions with broadcasting:\n",
      "        \n",
      "        >>> ax3 = plt.subplot(223)\n",
      "        >>> x = stats.norm.rvs(loc=[0,5], scale=[1,1.5],\n",
      "        ...                    size=(nsample//2,2), random_state=rng).ravel()\n",
      "        >>> res = stats.probplot(x, plot=plt)\n",
      "        \n",
      "        A standard normal distribution:\n",
      "        \n",
      "        >>> ax4 = plt.subplot(224)\n",
      "        >>> x = stats.norm.rvs(loc=0, scale=1, size=nsample, random_state=rng)\n",
      "        >>> res = stats.probplot(x, plot=plt)\n",
      "        \n",
      "        Produce a new figure with a loggamma distribution, using the ``dist`` and\n",
      "        ``sparams`` keywords:\n",
      "        \n",
      "        >>> fig = plt.figure()\n",
      "        >>> ax = fig.add_subplot(111)\n",
      "        >>> x = stats.loggamma.rvs(c=2.5, size=500, random_state=rng)\n",
      "        >>> res = stats.probplot(x, dist=stats.loggamma, sparams=(2.5,), plot=ax)\n",
      "        >>> ax.set_title(\"Probplot for loggamma dist with shape parameter 2.5\")\n",
      "        \n",
      "        Show the results with Matplotlib:\n",
      "        \n",
      "        >>> plt.show()\n",
      "    \n",
      "    rankdata(a, method='average', *, axis=None)\n",
      "        Assign ranks to data, dealing with ties appropriately.\n",
      "        \n",
      "        By default (``axis=None``), the data array is first flattened, and a flat\n",
      "        array of ranks is returned. Separately reshape the rank array to the\n",
      "        shape of the data array if desired (see Examples).\n",
      "        \n",
      "        Ranks begin at 1.  The `method` argument controls how ranks are assigned\n",
      "        to equal values.  See [1]_ for further discussion of ranking methods.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        a : array_like\n",
      "            The array of values to be ranked.\n",
      "        method : {'average', 'min', 'max', 'dense', 'ordinal'}, optional\n",
      "            The method used to assign ranks to tied elements.\n",
      "            The following methods are available (default is 'average'):\n",
      "        \n",
      "              * 'average': The average of the ranks that would have been assigned to\n",
      "                all the tied values is assigned to each value.\n",
      "              * 'min': The minimum of the ranks that would have been assigned to all\n",
      "                the tied values is assigned to each value.  (This is also\n",
      "                referred to as \"competition\" ranking.)\n",
      "              * 'max': The maximum of the ranks that would have been assigned to all\n",
      "                the tied values is assigned to each value.\n",
      "              * 'dense': Like 'min', but the rank of the next highest element is\n",
      "                assigned the rank immediately after those assigned to the tied\n",
      "                elements.\n",
      "              * 'ordinal': All values are given a distinct rank, corresponding to\n",
      "                the order that the values occur in `a`.\n",
      "        axis : {None, int}, optional\n",
      "            Axis along which to perform the ranking. If ``None``, the data array\n",
      "            is first flattened.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        ranks : ndarray\n",
      "             An array of size equal to the size of `a`, containing rank\n",
      "             scores.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] \"Ranking\", https://en.wikipedia.org/wiki/Ranking\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy.stats import rankdata\n",
      "        >>> rankdata([0, 2, 3, 2])\n",
      "        array([ 1. ,  2.5,  4. ,  2.5])\n",
      "        >>> rankdata([0, 2, 3, 2], method='min')\n",
      "        array([ 1,  2,  4,  2])\n",
      "        >>> rankdata([0, 2, 3, 2], method='max')\n",
      "        array([ 1,  3,  4,  3])\n",
      "        >>> rankdata([0, 2, 3, 2], method='dense')\n",
      "        array([ 1,  2,  3,  2])\n",
      "        >>> rankdata([0, 2, 3, 2], method='ordinal')\n",
      "        array([ 1,  2,  4,  3])\n",
      "        >>> rankdata([[0, 2], [3, 2]]).reshape(2,2)\n",
      "        array([[1. , 2.5],\n",
      "              [4. , 2.5]])\n",
      "        >>> rankdata([[0, 2, 2], [3, 2, 5]], axis=1)\n",
      "        array([[1. , 2.5, 2.5],\n",
      "               [2. , 1. , 3. ]])\n",
      "    \n",
      "    ranksums(x, y, alternative='two-sided')\n",
      "        Compute the Wilcoxon rank-sum statistic for two samples.\n",
      "        \n",
      "        The Wilcoxon rank-sum test tests the null hypothesis that two sets\n",
      "        of measurements are drawn from the same distribution.  The alternative\n",
      "        hypothesis is that values in one sample are more likely to be\n",
      "        larger than the values in the other sample.\n",
      "        \n",
      "        This test should be used to compare two samples from continuous\n",
      "        distributions.  It does not handle ties between measurements\n",
      "        in x and y.  For tie-handling and an optional continuity correction\n",
      "        see `scipy.stats.mannwhitneyu`.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x,y : array_like\n",
      "            The data from the two samples.\n",
      "        alternative : {'two-sided', 'less', 'greater'}, optional\n",
      "            Defines the alternative hypothesis. Default is 'two-sided'.\n",
      "            The following options are available:\n",
      "        \n",
      "            * 'two-sided': one of the distributions (underlying `x` or `y`) is\n",
      "              stochastically greater than the other.\n",
      "            * 'less': the distribution underlying `x` is stochastically less\n",
      "              than the distribution underlying `y`.\n",
      "            * 'greater': the distribution underlying `x` is stochastically greater\n",
      "              than the distribution underlying `y`.\n",
      "        \n",
      "            .. versionadded:: 1.7.0\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        statistic : float\n",
      "            The test statistic under the large-sample approximation that the\n",
      "            rank sum statistic is normally distributed.\n",
      "        pvalue : float\n",
      "            The p-value of the test.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] https://en.wikipedia.org/wiki/Wilcoxon_rank-sum_test\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        We can test the hypothesis that two independent unequal-sized samples are\n",
      "        drawn from the same distribution with computing the Wilcoxon rank-sum\n",
      "        statistic.\n",
      "        \n",
      "        >>> from scipy.stats import ranksums\n",
      "        >>> rng = np.random.default_rng()\n",
      "        >>> sample1 = rng.uniform(-1, 1, 200)\n",
      "        >>> sample2 = rng.uniform(-0.5, 1.5, 300) # a shifted distribution\n",
      "        >>> ranksums(sample1, sample2)\n",
      "        RanksumsResult(statistic=-7.887059, pvalue=3.09390448e-15)  # may vary\n",
      "        >>> ranksums(sample1, sample2, alternative='less')\n",
      "        RanksumsResult(statistic=-7.750585297581713, pvalue=4.573497606342543e-15) # may vary\n",
      "        >>> ranksums(sample1, sample2, alternative='greater')\n",
      "        RanksumsResult(statistic=-7.750585297581713, pvalue=0.9999999999999954) # may vary\n",
      "        \n",
      "        The p-value of less than ``0.05`` indicates that this test rejects the\n",
      "        hypothesis at the 5% significance level.\n",
      "    \n",
      "    relfreq(a, numbins=10, defaultreallimits=None, weights=None)\n",
      "        Return a relative frequency histogram, using the histogram function.\n",
      "        \n",
      "        A relative frequency  histogram is a mapping of the number of\n",
      "        observations in each of the bins relative to the total of observations.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        a : array_like\n",
      "            Input array.\n",
      "        numbins : int, optional\n",
      "            The number of bins to use for the histogram. Default is 10.\n",
      "        defaultreallimits : tuple (lower, upper), optional\n",
      "            The lower and upper values for the range of the histogram.\n",
      "            If no value is given, a range slightly larger than the range of the\n",
      "            values in a is used. Specifically ``(a.min() - s, a.max() + s)``,\n",
      "            where ``s = (1/2)(a.max() - a.min()) / (numbins - 1)``.\n",
      "        weights : array_like, optional\n",
      "            The weights for each value in `a`. Default is None, which gives each\n",
      "            value a weight of 1.0\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        frequency : ndarray\n",
      "            Binned values of relative frequency.\n",
      "        lowerlimit : float\n",
      "            Lower real limit.\n",
      "        binsize : float\n",
      "            Width of each bin.\n",
      "        extrapoints : int\n",
      "            Extra points.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> import matplotlib.pyplot as plt\n",
      "        >>> from numpy.random import default_rng\n",
      "        >>> from scipy import stats\n",
      "        >>> rng = default_rng()\n",
      "        >>> a = np.array([2, 4, 1, 2, 3, 2])\n",
      "        >>> res = stats.relfreq(a, numbins=4)\n",
      "        >>> res.frequency\n",
      "        array([ 0.16666667, 0.5       , 0.16666667,  0.16666667])\n",
      "        >>> np.sum(res.frequency)  # relative frequencies should add up to 1\n",
      "        1.0\n",
      "        \n",
      "        Create a normal distribution with 1000 random values\n",
      "        \n",
      "        >>> samples = stats.norm.rvs(size=1000, random_state=rng)\n",
      "        \n",
      "        Calculate relative frequencies\n",
      "        \n",
      "        >>> res = stats.relfreq(samples, numbins=25)\n",
      "        \n",
      "        Calculate space of values for x\n",
      "        \n",
      "        >>> x = res.lowerlimit + np.linspace(0, res.binsize*res.frequency.size,\n",
      "        ...                                  res.frequency.size)\n",
      "        \n",
      "        Plot relative frequency histogram\n",
      "        \n",
      "        >>> fig = plt.figure(figsize=(5, 4))\n",
      "        >>> ax = fig.add_subplot(1, 1, 1)\n",
      "        >>> ax.bar(x, res.frequency, width=res.binsize)\n",
      "        >>> ax.set_title('Relative frequency histogram')\n",
      "        >>> ax.set_xlim([x.min(), x.max()])\n",
      "        \n",
      "        >>> plt.show()\n",
      "    \n",
      "    rvs_ratio_uniforms(pdf, umax, vmin, vmax, size=1, c=0, random_state=None)\n",
      "        Generate random samples from a probability density function using the\n",
      "        ratio-of-uniforms method.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        pdf : callable\n",
      "            A function with signature `pdf(x)` that is proportional to the\n",
      "            probability density function of the distribution.\n",
      "        umax : float\n",
      "            The upper bound of the bounding rectangle in the u-direction.\n",
      "        vmin : float\n",
      "            The lower bound of the bounding rectangle in the v-direction.\n",
      "        vmax : float\n",
      "            The upper bound of the bounding rectangle in the v-direction.\n",
      "        size : int or tuple of ints, optional\n",
      "            Defining number of random variates (default is 1).\n",
      "        c : float, optional.\n",
      "            Shift parameter of ratio-of-uniforms method, see Notes. Default is 0.\n",
      "        random_state : {None, int, `numpy.random.Generator`,\n",
      "                        `numpy.random.RandomState`}, optional\n",
      "        \n",
      "            If `seed` is None (or `np.random`), the `numpy.random.RandomState`\n",
      "            singleton is used.\n",
      "            If `seed` is an int, a new ``RandomState`` instance is used,\n",
      "            seeded with `seed`.\n",
      "            If `seed` is already a ``Generator`` or ``RandomState`` instance then\n",
      "            that instance is used.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        rvs : ndarray\n",
      "            The random variates distributed according to the probability\n",
      "            distribution defined by the pdf.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        Given a univariate probability density function `pdf` and a constant `c`,\n",
      "        define the set ``A = {(u, v) : 0 < u <= sqrt(pdf(v/u + c))}``.\n",
      "        If `(U, V)` is a random vector uniformly distributed over `A`,\n",
      "        then `V/U + c` follows a distribution according to `pdf`.\n",
      "        \n",
      "        The above result (see [1]_, [2]_) can be used to sample random variables\n",
      "        using only the pdf, i.e. no inversion of the cdf is required. Typical\n",
      "        choices of `c` are zero or the mode of `pdf`. The set `A` is a subset of\n",
      "        the rectangle ``R = [0, umax] x [vmin, vmax]`` where\n",
      "        \n",
      "        - ``umax = sup sqrt(pdf(x))``\n",
      "        - ``vmin = inf (x - c) sqrt(pdf(x))``\n",
      "        - ``vmax = sup (x - c) sqrt(pdf(x))``\n",
      "        \n",
      "        In particular, these values are finite if `pdf` is bounded and\n",
      "        ``x**2 * pdf(x)`` is bounded (i.e. subquadratic tails).\n",
      "        One can generate `(U, V)` uniformly on `R` and return\n",
      "        `V/U + c` if `(U, V)` are also in `A` which can be directly\n",
      "        verified.\n",
      "        \n",
      "        The algorithm is not changed if one replaces `pdf` by k * `pdf` for any\n",
      "        constant k > 0. Thus, it is often convenient to work with a function\n",
      "        that is proportional to the probability density function by dropping\n",
      "        unneccessary normalization factors.\n",
      "        \n",
      "        Intuitively, the method works well if `A` fills up most of the\n",
      "        enclosing rectangle such that the probability is high that `(U, V)`\n",
      "        lies in `A` whenever it lies in `R` as the number of required\n",
      "        iterations becomes too large otherwise. To be more precise, note that\n",
      "        the expected number of iterations to draw `(U, V)` uniformly\n",
      "        distributed on `R` such that `(U, V)` is also in `A` is given by\n",
      "        the ratio ``area(R) / area(A) = 2 * umax * (vmax - vmin) / area(pdf)``,\n",
      "        where `area(pdf)` is the integral of `pdf` (which is equal to one if the\n",
      "        probability density function is used but can take on other values if a\n",
      "        function proportional to the density is used). The equality holds since\n",
      "        the area of `A` is equal to 0.5 * area(pdf) (Theorem 7.1 in [1]_).\n",
      "        If the sampling fails to generate a single random variate after 50000\n",
      "        iterations (i.e. not a single draw is in `A`), an exception is raised.\n",
      "        \n",
      "        If the bounding rectangle is not correctly specified (i.e. if it does not\n",
      "        contain `A`), the algorithm samples from a distribution different from\n",
      "        the one given by `pdf`. It is therefore recommended to perform a\n",
      "        test such as `~scipy.stats.kstest` as a check.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] L. Devroye, \"Non-Uniform Random Variate Generation\",\n",
      "           Springer-Verlag, 1986.\n",
      "        \n",
      "        .. [2] W. Hoermann and J. Leydold, \"Generating generalized inverse Gaussian\n",
      "           random variates\", Statistics and Computing, 24(4), p. 547--557, 2014.\n",
      "        \n",
      "        .. [3] A.J. Kinderman and J.F. Monahan, \"Computer Generation of Random\n",
      "           Variables Using the Ratio of Uniform Deviates\",\n",
      "           ACM Transactions on Mathematical Software, 3(3), p. 257--260, 1977.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy import stats\n",
      "        >>> rng = np.random.default_rng()\n",
      "        \n",
      "        Simulate normally distributed random variables. It is easy to compute the\n",
      "        bounding rectangle explicitly in that case. For simplicity, we drop the\n",
      "        normalization factor of the density.\n",
      "        \n",
      "        >>> f = lambda x: np.exp(-x**2 / 2)\n",
      "        >>> v_bound = np.sqrt(f(np.sqrt(2))) * np.sqrt(2)\n",
      "        >>> umax, vmin, vmax = np.sqrt(f(0)), -v_bound, v_bound\n",
      "        >>> rvs = stats.rvs_ratio_uniforms(f, umax, vmin, vmax, size=2500,\n",
      "        ...                                random_state=rng)\n",
      "        \n",
      "        The K-S test confirms that the random variates are indeed normally\n",
      "        distributed (normality is not rejected at 5% significance level):\n",
      "        \n",
      "        >>> stats.kstest(rvs, 'norm')[1]\n",
      "        0.250634764150542\n",
      "        \n",
      "        The exponential distribution provides another example where the bounding\n",
      "        rectangle can be determined explicitly.\n",
      "        \n",
      "        >>> rvs = stats.rvs_ratio_uniforms(lambda x: np.exp(-x), umax=1,\n",
      "        ...                                vmin=0, vmax=2*np.exp(-1), size=1000,\n",
      "        ...                                random_state=rng)\n",
      "        >>> stats.kstest(rvs, 'expon')[1]\n",
      "        0.21121052054580314\n",
      "    \n",
      "    scoreatpercentile(a, per, limit=(), interpolation_method='fraction', axis=None)\n",
      "        Calculate the score at a given percentile of the input sequence.\n",
      "        \n",
      "        For example, the score at `per=50` is the median. If the desired quantile\n",
      "        lies between two data points, we interpolate between them, according to\n",
      "        the value of `interpolation`. If the parameter `limit` is provided, it\n",
      "        should be a tuple (lower, upper) of two values.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        a : array_like\n",
      "            A 1-D array of values from which to extract score.\n",
      "        per : array_like\n",
      "            Percentile(s) at which to extract score.  Values should be in range\n",
      "            [0,100].\n",
      "        limit : tuple, optional\n",
      "            Tuple of two scalars, the lower and upper limits within which to\n",
      "            compute the percentile. Values of `a` outside\n",
      "            this (closed) interval will be ignored.\n",
      "        interpolation_method : {'fraction', 'lower', 'higher'}, optional\n",
      "            Specifies the interpolation method to use,\n",
      "            when the desired quantile lies between two data points `i` and `j`\n",
      "            The following options are available (default is 'fraction'):\n",
      "        \n",
      "              * 'fraction': ``i + (j - i) * fraction`` where ``fraction`` is the\n",
      "                fractional part of the index surrounded by ``i`` and ``j``\n",
      "              * 'lower': ``i``\n",
      "              * 'higher': ``j``\n",
      "        \n",
      "        axis : int, optional\n",
      "            Axis along which the percentiles are computed. Default is None. If\n",
      "            None, compute over the whole array `a`.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        score : float or ndarray\n",
      "            Score at percentile(s).\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        percentileofscore, numpy.percentile\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        This function will become obsolete in the future.\n",
      "        For NumPy 1.9 and higher, `numpy.percentile` provides all the functionality\n",
      "        that `scoreatpercentile` provides.  And it's significantly faster.\n",
      "        Therefore it's recommended to use `numpy.percentile` for users that have\n",
      "        numpy >= 1.9.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy import stats\n",
      "        >>> a = np.arange(100)\n",
      "        >>> stats.scoreatpercentile(a, 50)\n",
      "        49.5\n",
      "    \n",
      "    sem(a, axis=0, ddof=1, nan_policy='propagate')\n",
      "        Compute standard error of the mean.\n",
      "        \n",
      "        Calculate the standard error of the mean (or standard error of\n",
      "        measurement) of the values in the input array.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        a : array_like\n",
      "            An array containing the values for which the standard error is\n",
      "            returned.\n",
      "        axis : int or None, optional\n",
      "            Axis along which to operate. Default is 0. If None, compute over\n",
      "            the whole array `a`.\n",
      "        ddof : int, optional\n",
      "            Delta degrees-of-freedom. How many degrees of freedom to adjust\n",
      "            for bias in limited samples relative to the population estimate\n",
      "            of variance. Defaults to 1.\n",
      "        nan_policy : {'propagate', 'raise', 'omit'}, optional\n",
      "            Defines how to handle when input contains nan.\n",
      "            The following options are available (default is 'propagate'):\n",
      "        \n",
      "              * 'propagate': returns nan\n",
      "              * 'raise': throws an error\n",
      "              * 'omit': performs the calculations ignoring nan values\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        s : ndarray or float\n",
      "            The standard error of the mean in the sample(s), along the input axis.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        The default value for `ddof` is different to the default (0) used by other\n",
      "        ddof containing routines, such as np.std and np.nanstd.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        Find standard error along the first axis:\n",
      "        \n",
      "        >>> from scipy import stats\n",
      "        >>> a = np.arange(20).reshape(5,4)\n",
      "        >>> stats.sem(a)\n",
      "        array([ 2.8284,  2.8284,  2.8284,  2.8284])\n",
      "        \n",
      "        Find standard error across the whole array, using n degrees of freedom:\n",
      "        \n",
      "        >>> stats.sem(a, axis=None, ddof=0)\n",
      "        1.2893796958227628\n",
      "    \n",
      "    shapiro(x)\n",
      "        Perform the Shapiro-Wilk test for normality.\n",
      "        \n",
      "        The Shapiro-Wilk test tests the null hypothesis that the\n",
      "        data was drawn from a normal distribution.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x : array_like\n",
      "            Array of sample data.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        statistic : float\n",
      "            The test statistic.\n",
      "        p-value : float\n",
      "            The p-value for the hypothesis test.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        anderson : The Anderson-Darling test for normality\n",
      "        kstest : The Kolmogorov-Smirnov test for goodness of fit.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        The algorithm used is described in [4]_ but censoring parameters as\n",
      "        described are not implemented. For N > 5000 the W test statistic is accurate\n",
      "        but the p-value may not be.\n",
      "        \n",
      "        The chance of rejecting the null hypothesis when it is true is close to 5%\n",
      "        regardless of sample size.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] https://www.itl.nist.gov/div898/handbook/prc/section2/prc213.htm\n",
      "        .. [2] Shapiro, S. S. & Wilk, M.B (1965). An analysis of variance test for\n",
      "               normality (complete samples), Biometrika, Vol. 52, pp. 591-611.\n",
      "        .. [3] Razali, N. M. & Wah, Y. B. (2011) Power comparisons of Shapiro-Wilk,\n",
      "               Kolmogorov-Smirnov, Lilliefors and Anderson-Darling tests, Journal of\n",
      "               Statistical Modeling and Analytics, Vol. 2, pp. 21-33.\n",
      "        .. [4] ALGORITHM AS R94 APPL. STATIST. (1995) VOL. 44, NO. 4.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy import stats\n",
      "        >>> rng = np.random.default_rng()\n",
      "        >>> x = stats.norm.rvs(loc=5, scale=3, size=100, random_state=rng)\n",
      "        >>> shapiro_test = stats.shapiro(x)\n",
      "        >>> shapiro_test\n",
      "        ShapiroResult(statistic=0.9813305735588074, pvalue=0.16855233907699585)\n",
      "        >>> shapiro_test.statistic\n",
      "        0.9813305735588074\n",
      "        >>> shapiro_test.pvalue\n",
      "        0.16855233907699585\n",
      "    \n",
      "    siegelslopes(y, x=None, method='hierarchical')\n",
      "        Computes the Siegel estimator for a set of points (x, y).\n",
      "        \n",
      "        `siegelslopes` implements a method for robust linear regression\n",
      "        using repeated medians (see [1]_) to fit a line to the points (x, y).\n",
      "        The method is robust to outliers with an asymptotic breakdown point\n",
      "        of 50%.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        y : array_like\n",
      "            Dependent variable.\n",
      "        x : array_like or None, optional\n",
      "            Independent variable. If None, use ``arange(len(y))`` instead.\n",
      "        method : {'hierarchical', 'separate'}\n",
      "            If 'hierarchical', estimate the intercept using the estimated\n",
      "            slope ``medslope`` (default option).\n",
      "            If 'separate', estimate the intercept independent of the estimated\n",
      "            slope. See Notes for details.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        medslope : float\n",
      "            Estimate of the slope of the regression line.\n",
      "        medintercept : float\n",
      "            Estimate of the intercept of the regression line.\n",
      "        \n",
      "        See also\n",
      "        --------\n",
      "        theilslopes : a similar technique without repeated medians\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        With ``n = len(y)``, compute ``m_j`` as the median of\n",
      "        the slopes from the point ``(x[j], y[j])`` to all other `n-1` points.\n",
      "        ``medslope`` is then the median of all slopes ``m_j``.\n",
      "        Two ways are given to estimate the intercept in [1]_ which can be chosen\n",
      "        via the parameter ``method``.\n",
      "        The hierarchical approach uses the estimated slope ``medslope``\n",
      "        and computes ``medintercept`` as the median of ``y - medslope*x``.\n",
      "        The other approach estimates the intercept separately as follows: for\n",
      "        each point ``(x[j], y[j])``, compute the intercepts of all the `n-1`\n",
      "        lines through the remaining points and take the median ``i_j``.\n",
      "        ``medintercept`` is the median of the ``i_j``.\n",
      "        \n",
      "        The implementation computes `n` times the median of a vector of size `n`\n",
      "        which can be slow for large vectors. There are more efficient algorithms\n",
      "        (see [2]_) which are not implemented here.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] A. Siegel, \"Robust Regression Using Repeated Medians\",\n",
      "               Biometrika, Vol. 69, pp. 242-244, 1982.\n",
      "        \n",
      "        .. [2] A. Stein and M. Werman, \"Finding the repeated median regression\n",
      "               line\", Proceedings of the Third Annual ACM-SIAM Symposium on\n",
      "               Discrete Algorithms, pp. 409-413, 1992.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy import stats\n",
      "        >>> import matplotlib.pyplot as plt\n",
      "        \n",
      "        >>> x = np.linspace(-5, 5, num=150)\n",
      "        >>> y = x + np.random.normal(size=x.size)\n",
      "        >>> y[11:15] += 10  # add outliers\n",
      "        >>> y[-5:] -= 7\n",
      "        \n",
      "        Compute the slope and intercept.  For comparison, also compute the\n",
      "        least-squares fit with `linregress`:\n",
      "        \n",
      "        >>> res = stats.siegelslopes(y, x)\n",
      "        >>> lsq_res = stats.linregress(x, y)\n",
      "        \n",
      "        Plot the results. The Siegel regression line is shown in red. The green\n",
      "        line shows the least-squares fit for comparison.\n",
      "        \n",
      "        >>> fig = plt.figure()\n",
      "        >>> ax = fig.add_subplot(111)\n",
      "        >>> ax.plot(x, y, 'b.')\n",
      "        >>> ax.plot(x, res[1] + res[0] * x, 'r-')\n",
      "        >>> ax.plot(x, lsq_res[1] + lsq_res[0] * x, 'g-')\n",
      "        >>> plt.show()\n",
      "    \n",
      "    sigmaclip(a, low=4.0, high=4.0)\n",
      "        Perform iterative sigma-clipping of array elements.\n",
      "        \n",
      "        Starting from the full sample, all elements outside the critical range are\n",
      "        removed, i.e. all elements of the input array `c` that satisfy either of\n",
      "        the following conditions::\n",
      "        \n",
      "            c < mean(c) - std(c)*low\n",
      "            c > mean(c) + std(c)*high\n",
      "        \n",
      "        The iteration continues with the updated sample until no\n",
      "        elements are outside the (updated) range.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        a : array_like\n",
      "            Data array, will be raveled if not 1-D.\n",
      "        low : float, optional\n",
      "            Lower bound factor of sigma clipping. Default is 4.\n",
      "        high : float, optional\n",
      "            Upper bound factor of sigma clipping. Default is 4.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        clipped : ndarray\n",
      "            Input array with clipped elements removed.\n",
      "        lower : float\n",
      "            Lower threshold value use for clipping.\n",
      "        upper : float\n",
      "            Upper threshold value use for clipping.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy.stats import sigmaclip\n",
      "        >>> a = np.concatenate((np.linspace(9.5, 10.5, 31),\n",
      "        ...                     np.linspace(0, 20, 5)))\n",
      "        >>> fact = 1.5\n",
      "        >>> c, low, upp = sigmaclip(a, fact, fact)\n",
      "        >>> c\n",
      "        array([  9.96666667,  10.        ,  10.03333333,  10.        ])\n",
      "        >>> c.var(), c.std()\n",
      "        (0.00055555555555555165, 0.023570226039551501)\n",
      "        >>> low, c.mean() - fact*c.std(), c.min()\n",
      "        (9.9646446609406727, 9.9646446609406727, 9.9666666666666668)\n",
      "        >>> upp, c.mean() + fact*c.std(), c.max()\n",
      "        (10.035355339059327, 10.035355339059327, 10.033333333333333)\n",
      "        \n",
      "        >>> a = np.concatenate((np.linspace(9.5, 10.5, 11),\n",
      "        ...                     np.linspace(-100, -50, 3)))\n",
      "        >>> c, low, upp = sigmaclip(a, 1.8, 1.8)\n",
      "        >>> (c == np.linspace(9.5, 10.5, 11)).all()\n",
      "        True\n",
      "    \n",
      "    skew(a, axis=0, bias=True, nan_policy='propagate')\n",
      "        Compute the sample skewness of a data set.\n",
      "        \n",
      "        For normally distributed data, the skewness should be about zero. For\n",
      "        unimodal continuous distributions, a skewness value greater than zero means\n",
      "        that there is more weight in the right tail of the distribution. The\n",
      "        function `skewtest` can be used to determine if the skewness value\n",
      "        is close enough to zero, statistically speaking.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        a : ndarray\n",
      "            Input array.\n",
      "        axis : int or None, optional\n",
      "            Axis along which skewness is calculated. Default is 0.\n",
      "            If None, compute over the whole array `a`.\n",
      "        bias : bool, optional\n",
      "            If False, then the calculations are corrected for statistical bias.\n",
      "        nan_policy : {'propagate', 'raise', 'omit'}, optional\n",
      "            Defines how to handle when input contains nan.\n",
      "            The following options are available (default is 'propagate'):\n",
      "        \n",
      "              * 'propagate': returns nan\n",
      "              * 'raise': throws an error\n",
      "              * 'omit': performs the calculations ignoring nan values\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        skewness : ndarray\n",
      "            The skewness of values along an axis, returning 0 where all values are\n",
      "            equal.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        The sample skewness is computed as the Fisher-Pearson coefficient\n",
      "        of skewness, i.e.\n",
      "        \n",
      "        .. math::\n",
      "        \n",
      "            g_1=\\frac{m_3}{m_2^{3/2}}\n",
      "        \n",
      "        where\n",
      "        \n",
      "        .. math::\n",
      "        \n",
      "            m_i=\\frac{1}{N}\\sum_{n=1}^N(x[n]-\\bar{x})^i\n",
      "        \n",
      "        is the biased sample :math:`i\\texttt{th}` central moment, and\n",
      "        :math:`\\bar{x}` is\n",
      "        the sample mean.  If ``bias`` is False, the calculations are\n",
      "        corrected for bias and the value computed is the adjusted\n",
      "        Fisher-Pearson standardized moment coefficient, i.e.\n",
      "        \n",
      "        .. math::\n",
      "        \n",
      "            G_1=\\frac{k_3}{k_2^{3/2}}=\n",
      "                \\frac{\\sqrt{N(N-1)}}{N-2}\\frac{m_3}{m_2^{3/2}}.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] Zwillinger, D. and Kokoska, S. (2000). CRC Standard\n",
      "           Probability and Statistics Tables and Formulae. Chapman & Hall: New\n",
      "           York. 2000.\n",
      "           Section 2.2.24.1\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy.stats import skew\n",
      "        >>> skew([1, 2, 3, 4, 5])\n",
      "        0.0\n",
      "        >>> skew([2, 8, 0, 4, 1, 9, 9, 0])\n",
      "        0.2650554122698573\n",
      "    \n",
      "    skewtest(a, axis=0, nan_policy='propagate', alternative='two-sided')\n",
      "        Test whether the skew is different from the normal distribution.\n",
      "        \n",
      "        This function tests the null hypothesis that the skewness of\n",
      "        the population that the sample was drawn from is the same\n",
      "        as that of a corresponding normal distribution.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        a : array\n",
      "            The data to be tested.\n",
      "        axis : int or None, optional\n",
      "           Axis along which statistics are calculated. Default is 0.\n",
      "           If None, compute over the whole array `a`.\n",
      "        nan_policy : {'propagate', 'raise', 'omit'}, optional\n",
      "            Defines how to handle when input contains nan.\n",
      "            The following options are available (default is 'propagate'):\n",
      "        \n",
      "            * 'propagate': returns nan\n",
      "            * 'raise': throws an error\n",
      "            * 'omit': performs the calculations ignoring nan values\n",
      "        \n",
      "        alternative : {'two-sided', 'less', 'greater'}, optional\n",
      "            Defines the alternative hypothesis. Default is 'two-sided'.\n",
      "            The following options are available:\n",
      "        \n",
      "            * 'two-sided': the skewness of the distribution underlying the sample\n",
      "              is different from that of the normal distribution (i.e. 0)\n",
      "            * 'less': the skewness of the distribution underlying the sample\n",
      "              is less than that of the normal distribution\n",
      "            * 'greater': the skewness of the distribution underlying the sample\n",
      "              is greater than that of the normal distribution\n",
      "        \n",
      "            .. versionadded:: 1.7.0\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        statistic : float\n",
      "            The computed z-score for this test.\n",
      "        pvalue : float\n",
      "            The p-value for the hypothesis test.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        The sample size must be at least 8.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] R. B. D'Agostino, A. J. Belanger and R. B. D'Agostino Jr.,\n",
      "                \"A suggestion for using powerful and informative tests of\n",
      "                normality\", American Statistician 44, pp. 316-321, 1990.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy.stats import skewtest\n",
      "        >>> skewtest([1, 2, 3, 4, 5, 6, 7, 8])\n",
      "        SkewtestResult(statistic=1.0108048609177787, pvalue=0.3121098361421897)\n",
      "        >>> skewtest([2, 8, 0, 4, 1, 9, 9, 0])\n",
      "        SkewtestResult(statistic=0.44626385374196975, pvalue=0.6554066631275459)\n",
      "        >>> skewtest([1, 2, 3, 4, 5, 6, 7, 8000])\n",
      "        SkewtestResult(statistic=3.571773510360407, pvalue=0.0003545719905823133)\n",
      "        >>> skewtest([100, 100, 100, 100, 100, 100, 100, 101])\n",
      "        SkewtestResult(statistic=3.5717766638478072, pvalue=0.000354567720281634)\n",
      "        >>> skewtest([1, 2, 3, 4, 5, 6, 7, 8], alternative='less')\n",
      "        SkewtestResult(statistic=1.0108048609177787, pvalue=0.8439450819289052)\n",
      "        >>> skewtest([1, 2, 3, 4, 5, 6, 7, 8], alternative='greater')\n",
      "        SkewtestResult(statistic=1.0108048609177787, pvalue=0.15605491807109484)\n",
      "    \n",
      "    somersd(x, y=None)\n",
      "        Calculates Somers' D, an asymmetric measure of ordinal association.\n",
      "        \n",
      "        Like Kendall's :math:`\\tau`, Somers' :math:`D` is a measure of the\n",
      "        correspondence between two rankings. Both statistics consider the\n",
      "        difference between the number of concordant and discordant pairs in two\n",
      "        rankings :math:`X` and :math:`Y`, and both are normalized such that values\n",
      "        close  to 1 indicate strong agreement and values close to -1 indicate\n",
      "        strong disagreement. They differ in how they are normalized. To show the\n",
      "        relationship, Somers' :math:`D` can be defined in terms of Kendall's\n",
      "        :math:`\\tau_a`:\n",
      "        \n",
      "        .. math::\n",
      "            D(Y|X) = \\frac{\\tau_a(X, Y)}{\\tau_a(X, X)}\n",
      "        \n",
      "        Suppose the first ranking :math:`X` has :math:`r` distinct ranks and the\n",
      "        second ranking :math:`Y` has :math:`s` distinct ranks. These two lists of\n",
      "        :math:`n` rankings can also be viewed as an :math:`r \\times s` contingency\n",
      "        table in which element :math:`i, j` is the number of rank pairs with rank\n",
      "        :math:`i` in ranking :math:`X` and rank :math:`j` in ranking :math:`Y`.\n",
      "        Accordingly, `somersd` also allows the input data to be supplied as a\n",
      "        single, 2D contingency table instead of as two separate, 1D rankings.\n",
      "        \n",
      "        Note that the definition of Somers' :math:`D` is asymmetric: in general,\n",
      "        :math:`D(Y|X) \\neq D(X|Y)`. ``somersd(x, y)`` calculates Somers'\n",
      "        :math:`D(Y|X)`: the \"row\" variable :math:`X` is treated as an independent\n",
      "        variable, and the \"column\" variable :math:`Y` is dependent. For Somers'\n",
      "        :math:`D(X|Y)`, swap the input lists or transpose the input table.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x: array_like\n",
      "            1D array of rankings, treated as the (row) independent variable.\n",
      "            Alternatively, a 2D contingency table.\n",
      "        y: array_like\n",
      "            If `x` is a 1D array of rankings, `y` is a 1D array of rankings of the\n",
      "            same length, treated as the (column) dependent variable.\n",
      "            If `x` is 2D, `y` is ignored.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        res : SomersDResult\n",
      "            A `SomersDResult` object with the following fields:\n",
      "        \n",
      "                correlation : float\n",
      "                   The Somers' :math:`D` statistic.\n",
      "                pvalue : float\n",
      "                   The two-sided p-value for a hypothesis test whose null\n",
      "                   hypothesis is an absence of association, :math:`D=0`.\n",
      "                   See notes for more information.\n",
      "                table : 2D array\n",
      "                   The contingency table formed from rankings `x` and `y` (or the\n",
      "                   provided contingency table, if `x` is a 2D array)\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        kendalltau : Calculates Kendall's tau, another correlation measure.\n",
      "        weightedtau : Computes a weighted version of Kendall's tau.\n",
      "        spearmanr : Calculates a Spearman rank-order correlation coefficient.\n",
      "        pearsonr : Calculates a Pearson correlation coefficient.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        This function follows the contingency table approach of [2]_ and\n",
      "        [3]_. *p*-values are computed based on an asymptotic approximation of\n",
      "        the test statistic distribution under the null hypothesis :math:`D=0`.\n",
      "        \n",
      "        Theoretically, hypothesis tests based on Kendall's :math:`tau` and Somers'\n",
      "        :math:`D` should be identical.\n",
      "        However, the *p*-values returned by `kendalltau` are based\n",
      "        on the null hypothesis of *independence* between :math:`X` and :math:`Y`\n",
      "        (i.e. the population from which pairs in :math:`X` and :math:`Y` are\n",
      "        sampled contains equal numbers of all possible pairs), which is more\n",
      "        specific than the null hypothesis :math:`D=0` used here. If the null\n",
      "        hypothesis of independence is desired, it is acceptable to use the\n",
      "        *p*-value returned by `kendalltau` with the statistic returned by\n",
      "        `somersd` and vice versa. For more information, see [2]_.\n",
      "        \n",
      "        Contingency tables are formatted according to the convention used by\n",
      "        SAS and R: the first ranking supplied (``x``) is the \"row\" variable, and\n",
      "        the second ranking supplied (``y``) is the \"column\" variable. This is\n",
      "        opposite the convention of Somers' original paper [1]_.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] Robert H. Somers, \"A New Asymmetric Measure of Association for\n",
      "               Ordinal Variables\", *American Sociological Review*, Vol. 27, No. 6,\n",
      "               pp. 799--811, 1962.\n",
      "        \n",
      "        .. [2] Morton B. Brown and Jacqueline K. Benedetti, \"Sampling Behavior of\n",
      "               Tests for Correlation in Two-Way Contingency Tables\", *Journal of\n",
      "               the American Statistical Association* Vol. 72, No. 358, pp.\n",
      "               309--315, 1977.\n",
      "        \n",
      "        .. [3] SAS Institute, Inc., \"The FREQ Procedure (Book Excerpt)\",\n",
      "               *SAS/STAT 9.2 User's Guide, Second Edition*, SAS Publishing, 2009.\n",
      "        \n",
      "        .. [4] Laerd Statistics, \"Somers' d using SPSS Statistics\", *SPSS\n",
      "               Statistics Tutorials and Statistical Guides*,\n",
      "               https://statistics.laerd.com/spss-tutorials/somers-d-using-spss-statistics.php,\n",
      "               Accessed July 31, 2020.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        We calculate Somers' D for the example given in [4]_, in which a hotel\n",
      "        chain owner seeks to determine the association between hotel room\n",
      "        cleanliness and customer satisfaction. The independent variable, hotel\n",
      "        room cleanliness, is ranked on an ordinal scale: \"below average (1)\",\n",
      "        \"average (2)\", or \"above average (3)\". The dependent variable, customer\n",
      "        satisfaction, is ranked on a second scale: \"very dissatisfied (1)\",\n",
      "        \"moderately dissatisfied (2)\", \"neither dissatisfied nor satisfied (3)\",\n",
      "        \"moderately satisfied (4)\", or \"very satisfied (5)\". 189 customers\n",
      "        respond to the survey, and the results are cast into a contingency table\n",
      "        with the hotel room cleanliness as the \"row\" variable and customer\n",
      "        satisfaction as the \"column\" variable.\n",
      "        \n",
      "        +-----+-----+-----+-----+-----+-----+\n",
      "        |     | (1) | (2) | (3) | (4) | (5) |\n",
      "        +=====+=====+=====+=====+=====+=====+\n",
      "        | (1) | 27  | 25  | 14  | 7   | 0   |\n",
      "        +-----+-----+-----+-----+-----+-----+\n",
      "        | (2) | 7   | 14  | 18  | 35  | 12  |\n",
      "        +-----+-----+-----+-----+-----+-----+\n",
      "        | (3) | 1   | 3   | 2   | 7   | 17  |\n",
      "        +-----+-----+-----+-----+-----+-----+\n",
      "        \n",
      "        For example, 27 customers assigned their room a cleanliness ranking of\n",
      "        \"below average (1)\" and a corresponding satisfaction of \"very\n",
      "        dissatisfied (1)\". We perform the analysis as follows.\n",
      "        \n",
      "        >>> from scipy.stats import somersd\n",
      "        >>> table = [[27, 25, 14, 7, 0], [7, 14, 18, 35, 12], [1, 3, 2, 7, 17]]\n",
      "        >>> res = somersd(table)\n",
      "        >>> res.statistic\n",
      "        0.6032766111513396\n",
      "        >>> res.pvalue\n",
      "        1.0007091191074533e-27\n",
      "        \n",
      "        The value of the Somers' D statistic is approximately 0.6, indicating\n",
      "        a positive correlation between room cleanliness and customer satisfaction\n",
      "        in the sample.\n",
      "        The *p*-value is very small, indicating a very small probability of\n",
      "        observing such an extreme value of the statistic under the null\n",
      "        hypothesis that the statistic of the entire population (from which\n",
      "        our sample of 189 customers is drawn) is zero. This supports the\n",
      "        alternative hypothesis that the true value of Somers' D for the population\n",
      "        is nonzero.\n",
      "    \n",
      "    spearmanr(a, b=None, axis=0, nan_policy='propagate', alternative='two-sided')\n",
      "        Calculate a Spearman correlation coefficient with associated p-value.\n",
      "        \n",
      "        The Spearman rank-order correlation coefficient is a nonparametric measure\n",
      "        of the monotonicity of the relationship between two datasets. Unlike the\n",
      "        Pearson correlation, the Spearman correlation does not assume that both\n",
      "        datasets are normally distributed. Like other correlation coefficients,\n",
      "        this one varies between -1 and +1 with 0 implying no correlation.\n",
      "        Correlations of -1 or +1 imply an exact monotonic relationship. Positive\n",
      "        correlations imply that as x increases, so does y. Negative correlations\n",
      "        imply that as x increases, y decreases.\n",
      "        \n",
      "        The p-value roughly indicates the probability of an uncorrelated system\n",
      "        producing datasets that have a Spearman correlation at least as extreme\n",
      "        as the one computed from these datasets. The p-values are not entirely\n",
      "        reliable but are probably reasonable for datasets larger than 500 or so.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        a, b : 1D or 2D array_like, b is optional\n",
      "            One or two 1-D or 2-D arrays containing multiple variables and\n",
      "            observations. When these are 1-D, each represents a vector of\n",
      "            observations of a single variable. For the behavior in the 2-D case,\n",
      "            see under ``axis``, below.\n",
      "            Both arrays need to have the same length in the ``axis`` dimension.\n",
      "        axis : int or None, optional\n",
      "            If axis=0 (default), then each column represents a variable, with\n",
      "            observations in the rows. If axis=1, the relationship is transposed:\n",
      "            each row represents a variable, while the columns contain observations.\n",
      "            If axis=None, then both arrays will be raveled.\n",
      "        nan_policy : {'propagate', 'raise', 'omit'}, optional\n",
      "            Defines how to handle when input contains nan.\n",
      "            The following options are available (default is 'propagate'):\n",
      "        \n",
      "            * 'propagate': returns nan\n",
      "            * 'raise': throws an error\n",
      "            * 'omit': performs the calculations ignoring nan values\n",
      "        \n",
      "        alternative : {'two-sided', 'less', 'greater'}, optional\n",
      "            Defines the alternative hypothesis. Default is 'two-sided'.\n",
      "            The following options are available:\n",
      "        \n",
      "            * 'two-sided': the correlation is nonzero\n",
      "            * 'less': the correlation is negative (less than zero)\n",
      "            * 'greater':  the correlation is positive (greater than zero)\n",
      "        \n",
      "            .. versionadded:: 1.7.0\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        correlation : float or ndarray (2-D square)\n",
      "            Spearman correlation matrix or correlation coefficient (if only 2\n",
      "            variables are given as parameters. Correlation matrix is square with\n",
      "            length equal to total number of variables (columns or rows) in ``a``\n",
      "            and ``b`` combined.\n",
      "        pvalue : float\n",
      "            The p-value for a hypothesis test whose null hypotheisis\n",
      "            is that two sets of data are uncorrelated. See `alternative` above\n",
      "            for alternative hypotheses. `pvalue` has the same\n",
      "            shape as `correlation`.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] Zwillinger, D. and Kokoska, S. (2000). CRC Standard\n",
      "           Probability and Statistics Tables and Formulae. Chapman & Hall: New\n",
      "           York. 2000.\n",
      "           Section  14.7\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy import stats\n",
      "        >>> stats.spearmanr([1,2,3,4,5], [5,6,7,8,7])\n",
      "        SpearmanrResult(correlation=0.82078..., pvalue=0.08858...)\n",
      "        >>> rng = np.random.default_rng()\n",
      "        >>> x2n = rng.standard_normal((100, 2))\n",
      "        >>> y2n = rng.standard_normal((100, 2))\n",
      "        >>> stats.spearmanr(x2n)\n",
      "        SpearmanrResult(correlation=-0.07960396039603959, pvalue=0.4311168705769747)\n",
      "        >>> stats.spearmanr(x2n[:,0], x2n[:,1])\n",
      "        SpearmanrResult(correlation=-0.07960396039603959, pvalue=0.4311168705769747)\n",
      "        >>> rho, pval = stats.spearmanr(x2n, y2n)\n",
      "        >>> rho\n",
      "        array([[ 1.        , -0.07960396, -0.08314431,  0.09662166],\n",
      "               [-0.07960396,  1.        , -0.14448245,  0.16738074],\n",
      "               [-0.08314431, -0.14448245,  1.        ,  0.03234323],\n",
      "               [ 0.09662166,  0.16738074,  0.03234323,  1.        ]])\n",
      "        >>> pval\n",
      "        array([[0.        , 0.43111687, 0.41084066, 0.33891628],\n",
      "               [0.43111687, 0.        , 0.15151618, 0.09600687],\n",
      "               [0.41084066, 0.15151618, 0.        , 0.74938561],\n",
      "               [0.33891628, 0.09600687, 0.74938561, 0.        ]])\n",
      "        >>> rho, pval = stats.spearmanr(x2n.T, y2n.T, axis=1)\n",
      "        >>> rho\n",
      "        array([[ 1.        , -0.07960396, -0.08314431,  0.09662166],\n",
      "               [-0.07960396,  1.        , -0.14448245,  0.16738074],\n",
      "               [-0.08314431, -0.14448245,  1.        ,  0.03234323],\n",
      "               [ 0.09662166,  0.16738074,  0.03234323,  1.        ]])\n",
      "        >>> stats.spearmanr(x2n, y2n, axis=None)\n",
      "        SpearmanrResult(correlation=0.044981624540613524, pvalue=0.5270803651336189)\n",
      "        >>> stats.spearmanr(x2n.ravel(), y2n.ravel())\n",
      "        SpearmanrResult(correlation=0.044981624540613524, pvalue=0.5270803651336189)\n",
      "        \n",
      "        >>> rng = np.random.default_rng()\n",
      "        >>> xint = rng.integers(10, size=(100, 2))\n",
      "        >>> stats.spearmanr(xint)\n",
      "        SpearmanrResult(correlation=0.09800224850707953, pvalue=0.3320271757932076)\n",
      "    \n",
      "    theilslopes(y, x=None, alpha=0.95)\n",
      "        Computes the Theil-Sen estimator for a set of points (x, y).\n",
      "        \n",
      "        `theilslopes` implements a method for robust linear regression.  It\n",
      "        computes the slope as the median of all slopes between paired values.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        y : array_like\n",
      "            Dependent variable.\n",
      "        x : array_like or None, optional\n",
      "            Independent variable. If None, use ``arange(len(y))`` instead.\n",
      "        alpha : float, optional\n",
      "            Confidence degree between 0 and 1. Default is 95% confidence.\n",
      "            Note that `alpha` is symmetric around 0.5, i.e. both 0.1 and 0.9 are\n",
      "            interpreted as \"find the 90% confidence interval\".\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        medslope : float\n",
      "            Theil slope.\n",
      "        medintercept : float\n",
      "            Intercept of the Theil line, as ``median(y) - medslope*median(x)``.\n",
      "        lo_slope : float\n",
      "            Lower bound of the confidence interval on `medslope`.\n",
      "        up_slope : float\n",
      "            Upper bound of the confidence interval on `medslope`.\n",
      "        \n",
      "        See also\n",
      "        --------\n",
      "        siegelslopes : a similar technique using repeated medians\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        The implementation of `theilslopes` follows [1]_. The intercept is\n",
      "        not defined in [1]_, and here it is defined as ``median(y) -\n",
      "        medslope*median(x)``, which is given in [3]_. Other definitions of\n",
      "        the intercept exist in the literature. A confidence interval for\n",
      "        the intercept is not given as this question is not addressed in\n",
      "        [1]_.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] P.K. Sen, \"Estimates of the regression coefficient based on\n",
      "               Kendall's tau\", J. Am. Stat. Assoc., Vol. 63, pp. 1379-1389, 1968.\n",
      "        .. [2] H. Theil, \"A rank-invariant method of linear and polynomial\n",
      "               regression analysis I, II and III\",  Nederl. Akad. Wetensch., Proc.\n",
      "               53:, pp. 386-392, pp. 521-525, pp. 1397-1412, 1950.\n",
      "        .. [3] W.L. Conover, \"Practical nonparametric statistics\", 2nd ed.,\n",
      "               John Wiley and Sons, New York, pp. 493.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy import stats\n",
      "        >>> import matplotlib.pyplot as plt\n",
      "        \n",
      "        >>> x = np.linspace(-5, 5, num=150)\n",
      "        >>> y = x + np.random.normal(size=x.size)\n",
      "        >>> y[11:15] += 10  # add outliers\n",
      "        >>> y[-5:] -= 7\n",
      "        \n",
      "        Compute the slope, intercept and 90% confidence interval.  For comparison,\n",
      "        also compute the least-squares fit with `linregress`:\n",
      "        \n",
      "        >>> res = stats.theilslopes(y, x, 0.90)\n",
      "        >>> lsq_res = stats.linregress(x, y)\n",
      "        \n",
      "        Plot the results. The Theil-Sen regression line is shown in red, with the\n",
      "        dashed red lines illustrating the confidence interval of the slope (note\n",
      "        that the dashed red lines are not the confidence interval of the regression\n",
      "        as the confidence interval of the intercept is not included). The green\n",
      "        line shows the least-squares fit for comparison.\n",
      "        \n",
      "        >>> fig = plt.figure()\n",
      "        >>> ax = fig.add_subplot(111)\n",
      "        >>> ax.plot(x, y, 'b.')\n",
      "        >>> ax.plot(x, res[1] + res[0] * x, 'r-')\n",
      "        >>> ax.plot(x, res[1] + res[2] * x, 'r--')\n",
      "        >>> ax.plot(x, res[1] + res[3] * x, 'r--')\n",
      "        >>> ax.plot(x, lsq_res[1] + lsq_res[0] * x, 'g-')\n",
      "        >>> plt.show()\n",
      "    \n",
      "    tiecorrect(rankvals)\n",
      "        Tie correction factor for Mann-Whitney U and Kruskal-Wallis H tests.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        rankvals : array_like\n",
      "            A 1-D sequence of ranks.  Typically this will be the array\n",
      "            returned by `~scipy.stats.rankdata`.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        factor : float\n",
      "            Correction factor for U or H.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        rankdata : Assign ranks to the data\n",
      "        mannwhitneyu : Mann-Whitney rank test\n",
      "        kruskal : Kruskal-Wallis H test\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] Siegel, S. (1956) Nonparametric Statistics for the Behavioral\n",
      "               Sciences.  New York: McGraw-Hill.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy.stats import tiecorrect, rankdata\n",
      "        >>> tiecorrect([1, 2.5, 2.5, 4])\n",
      "        0.9\n",
      "        >>> ranks = rankdata([1, 3, 2, 4, 5, 7, 2, 8, 4])\n",
      "        >>> ranks\n",
      "        array([ 1. ,  4. ,  2.5,  5.5,  7. ,  8. ,  2.5,  9. ,  5.5])\n",
      "        >>> tiecorrect(ranks)\n",
      "        0.9833333333333333\n",
      "    \n",
      "    tmax(a, upperlimit=None, axis=0, inclusive=True, nan_policy='propagate')\n",
      "        Compute the trimmed maximum.\n",
      "        \n",
      "        This function computes the maximum value of an array along a given axis,\n",
      "        while ignoring values larger than a specified upper limit.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        a : array_like\n",
      "            Array of values.\n",
      "        upperlimit : None or float, optional\n",
      "            Values in the input array greater than the given limit will be ignored.\n",
      "            When upperlimit is None, then all values are used. The default value\n",
      "            is None.\n",
      "        axis : int or None, optional\n",
      "            Axis along which to operate. Default is 0. If None, compute over the\n",
      "            whole array `a`.\n",
      "        inclusive : {True, False}, optional\n",
      "            This flag determines whether values exactly equal to the upper limit\n",
      "            are included.  The default value is True.\n",
      "        nan_policy : {'propagate', 'raise', 'omit'}, optional\n",
      "            Defines how to handle when input contains nan.\n",
      "            The following options are available (default is 'propagate'):\n",
      "        \n",
      "              * 'propagate': returns nan\n",
      "              * 'raise': throws an error\n",
      "              * 'omit': performs the calculations ignoring nan values\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        tmax : float, int or ndarray\n",
      "            Trimmed maximum.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy import stats\n",
      "        >>> x = np.arange(20)\n",
      "        >>> stats.tmax(x)\n",
      "        19\n",
      "        \n",
      "        >>> stats.tmax(x, 13)\n",
      "        13\n",
      "        \n",
      "        >>> stats.tmax(x, 13, inclusive=False)\n",
      "        12\n",
      "    \n",
      "    tmean(a, limits=None, inclusive=(True, True), axis=None)\n",
      "        Compute the trimmed mean.\n",
      "        \n",
      "        This function finds the arithmetic mean of given values, ignoring values\n",
      "        outside the given `limits`.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        a : array_like\n",
      "            Array of values.\n",
      "        limits : None or (lower limit, upper limit), optional\n",
      "            Values in the input array less than the lower limit or greater than the\n",
      "            upper limit will be ignored.  When limits is None (default), then all\n",
      "            values are used.  Either of the limit values in the tuple can also be\n",
      "            None representing a half-open interval.\n",
      "        inclusive : (bool, bool), optional\n",
      "            A tuple consisting of the (lower flag, upper flag).  These flags\n",
      "            determine whether values exactly equal to the lower or upper limits\n",
      "            are included.  The default value is (True, True).\n",
      "        axis : int or None, optional\n",
      "            Axis along which to compute test. Default is None.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        tmean : float\n",
      "            Trimmed mean.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        trim_mean : Returns mean after trimming a proportion from both tails.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy import stats\n",
      "        >>> x = np.arange(20)\n",
      "        >>> stats.tmean(x)\n",
      "        9.5\n",
      "        >>> stats.tmean(x, (3,17))\n",
      "        10.0\n",
      "    \n",
      "    tmin(a, lowerlimit=None, axis=0, inclusive=True, nan_policy='propagate')\n",
      "        Compute the trimmed minimum.\n",
      "        \n",
      "        This function finds the miminum value of an array `a` along the\n",
      "        specified axis, but only considering values greater than a specified\n",
      "        lower limit.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        a : array_like\n",
      "            Array of values.\n",
      "        lowerlimit : None or float, optional\n",
      "            Values in the input array less than the given limit will be ignored.\n",
      "            When lowerlimit is None, then all values are used. The default value\n",
      "            is None.\n",
      "        axis : int or None, optional\n",
      "            Axis along which to operate. Default is 0. If None, compute over the\n",
      "            whole array `a`.\n",
      "        inclusive : {True, False}, optional\n",
      "            This flag determines whether values exactly equal to the lower limit\n",
      "            are included.  The default value is True.\n",
      "        nan_policy : {'propagate', 'raise', 'omit'}, optional\n",
      "            Defines how to handle when input contains nan.\n",
      "            The following options are available (default is 'propagate'):\n",
      "        \n",
      "              * 'propagate': returns nan\n",
      "              * 'raise': throws an error\n",
      "              * 'omit': performs the calculations ignoring nan values\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        tmin : float, int or ndarray\n",
      "            Trimmed minimum.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy import stats\n",
      "        >>> x = np.arange(20)\n",
      "        >>> stats.tmin(x)\n",
      "        0\n",
      "        \n",
      "        >>> stats.tmin(x, 13)\n",
      "        13\n",
      "        \n",
      "        >>> stats.tmin(x, 13, inclusive=False)\n",
      "        14\n",
      "    \n",
      "    trim1(a, proportiontocut, tail='right', axis=0)\n",
      "        Slice off a proportion from ONE end of the passed array distribution.\n",
      "        \n",
      "        If `proportiontocut` = 0.1, slices off 'leftmost' or 'rightmost'\n",
      "        10% of scores. The lowest or highest values are trimmed (depending on\n",
      "        the tail).\n",
      "        Slice off less if proportion results in a non-integer slice index\n",
      "        (i.e. conservatively slices off `proportiontocut` ).\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        a : array_like\n",
      "            Input array.\n",
      "        proportiontocut : float\n",
      "            Fraction to cut off of 'left' or 'right' of distribution.\n",
      "        tail : {'left', 'right'}, optional\n",
      "            Defaults to 'right'.\n",
      "        axis : int or None, optional\n",
      "            Axis along which to trim data. Default is 0. If None, compute over\n",
      "            the whole array `a`.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        trim1 : ndarray\n",
      "            Trimmed version of array `a`. The order of the trimmed content is\n",
      "            undefined.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy import stats\n",
      "        >>> a = np.arange(20)\n",
      "        >>> b = stats.trim1(a, 0.5, 'left')\n",
      "        >>> b\n",
      "        array([10, 11, 12, 13, 14, 16, 15, 17, 18, 19])\n",
      "    \n",
      "    trim_mean(a, proportiontocut, axis=0)\n",
      "        Return mean of array after trimming distribution from both tails.\n",
      "        \n",
      "        If `proportiontocut` = 0.1, slices off 'leftmost' and 'rightmost' 10% of\n",
      "        scores. The input is sorted before slicing. Slices off less if proportion\n",
      "        results in a non-integer slice index (i.e., conservatively slices off\n",
      "        `proportiontocut` ).\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        a : array_like\n",
      "            Input array.\n",
      "        proportiontocut : float\n",
      "            Fraction to cut off of both tails of the distribution.\n",
      "        axis : int or None, optional\n",
      "            Axis along which the trimmed means are computed. Default is 0.\n",
      "            If None, compute over the whole array `a`.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        trim_mean : ndarray\n",
      "            Mean of trimmed array.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        trimboth\n",
      "        tmean : Compute the trimmed mean ignoring values outside given `limits`.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy import stats\n",
      "        >>> x = np.arange(20)\n",
      "        >>> stats.trim_mean(x, 0.1)\n",
      "        9.5\n",
      "        >>> x2 = x.reshape(5, 4)\n",
      "        >>> x2\n",
      "        array([[ 0,  1,  2,  3],\n",
      "               [ 4,  5,  6,  7],\n",
      "               [ 8,  9, 10, 11],\n",
      "               [12, 13, 14, 15],\n",
      "               [16, 17, 18, 19]])\n",
      "        >>> stats.trim_mean(x2, 0.25)\n",
      "        array([  8.,   9.,  10.,  11.])\n",
      "        >>> stats.trim_mean(x2, 0.25, axis=1)\n",
      "        array([  1.5,   5.5,   9.5,  13.5,  17.5])\n",
      "    \n",
      "    trimboth(a, proportiontocut, axis=0)\n",
      "        Slice off a proportion of items from both ends of an array.\n",
      "        \n",
      "        Slice off the passed proportion of items from both ends of the passed\n",
      "        array (i.e., with `proportiontocut` = 0.1, slices leftmost 10% **and**\n",
      "        rightmost 10% of scores). The trimmed values are the lowest and\n",
      "        highest ones.\n",
      "        Slice off less if proportion results in a non-integer slice index (i.e.\n",
      "        conservatively slices off `proportiontocut`).\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        a : array_like\n",
      "            Data to trim.\n",
      "        proportiontocut : float\n",
      "            Proportion (in range 0-1) of total data set to trim of each end.\n",
      "        axis : int or None, optional\n",
      "            Axis along which to trim data. Default is 0. If None, compute over\n",
      "            the whole array `a`.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        out : ndarray\n",
      "            Trimmed version of array `a`. The order of the trimmed content\n",
      "            is undefined.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        trim_mean\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy import stats\n",
      "        >>> a = np.arange(20)\n",
      "        >>> b = stats.trimboth(a, 0.1)\n",
      "        >>> b.shape\n",
      "        (16,)\n",
      "    \n",
      "    tsem(a, limits=None, inclusive=(True, True), axis=0, ddof=1)\n",
      "        Compute the trimmed standard error of the mean.\n",
      "        \n",
      "        This function finds the standard error of the mean for given\n",
      "        values, ignoring values outside the given `limits`.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        a : array_like\n",
      "            Array of values.\n",
      "        limits : None or (lower limit, upper limit), optional\n",
      "            Values in the input array less than the lower limit or greater than the\n",
      "            upper limit will be ignored. When limits is None, then all values are\n",
      "            used. Either of the limit values in the tuple can also be None\n",
      "            representing a half-open interval.  The default value is None.\n",
      "        inclusive : (bool, bool), optional\n",
      "            A tuple consisting of the (lower flag, upper flag).  These flags\n",
      "            determine whether values exactly equal to the lower or upper limits\n",
      "            are included.  The default value is (True, True).\n",
      "        axis : int or None, optional\n",
      "            Axis along which to operate. Default is 0. If None, compute over the\n",
      "            whole array `a`.\n",
      "        ddof : int, optional\n",
      "            Delta degrees of freedom.  Default is 1.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        tsem : float\n",
      "            Trimmed standard error of the mean.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        `tsem` uses unbiased sample standard deviation, i.e. it uses a\n",
      "        correction factor ``n / (n - 1)``.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy import stats\n",
      "        >>> x = np.arange(20)\n",
      "        >>> stats.tsem(x)\n",
      "        1.3228756555322954\n",
      "        >>> stats.tsem(x, (3,17))\n",
      "        1.1547005383792515\n",
      "    \n",
      "    tstd(a, limits=None, inclusive=(True, True), axis=0, ddof=1)\n",
      "        Compute the trimmed sample standard deviation.\n",
      "        \n",
      "        This function finds the sample standard deviation of given values,\n",
      "        ignoring values outside the given `limits`.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        a : array_like\n",
      "            Array of values.\n",
      "        limits : None or (lower limit, upper limit), optional\n",
      "            Values in the input array less than the lower limit or greater than the\n",
      "            upper limit will be ignored. When limits is None, then all values are\n",
      "            used. Either of the limit values in the tuple can also be None\n",
      "            representing a half-open interval.  The default value is None.\n",
      "        inclusive : (bool, bool), optional\n",
      "            A tuple consisting of the (lower flag, upper flag).  These flags\n",
      "            determine whether values exactly equal to the lower or upper limits\n",
      "            are included.  The default value is (True, True).\n",
      "        axis : int or None, optional\n",
      "            Axis along which to operate. Default is 0. If None, compute over the\n",
      "            whole array `a`.\n",
      "        ddof : int, optional\n",
      "            Delta degrees of freedom.  Default is 1.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        tstd : float\n",
      "            Trimmed sample standard deviation.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        `tstd` computes the unbiased sample standard deviation, i.e. it uses a\n",
      "        correction factor ``n / (n - 1)``.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy import stats\n",
      "        >>> x = np.arange(20)\n",
      "        >>> stats.tstd(x)\n",
      "        5.9160797830996161\n",
      "        >>> stats.tstd(x, (3,17))\n",
      "        4.4721359549995796\n",
      "    \n",
      "    ttest_1samp(a, popmean, axis=0, nan_policy='propagate', alternative='two-sided')\n",
      "        Calculate the T-test for the mean of ONE group of scores.\n",
      "        \n",
      "        This is a two-sided test for the null hypothesis that the expected value\n",
      "        (mean) of a sample of independent observations `a` is equal to the given\n",
      "        population mean, `popmean`.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        a : array_like\n",
      "            Sample observation.\n",
      "        popmean : float or array_like\n",
      "            Expected value in null hypothesis. If array_like, then it must have the\n",
      "            same shape as `a` excluding the axis dimension.\n",
      "        axis : int or None, optional\n",
      "            Axis along which to compute test; default is 0. If None, compute over\n",
      "            the whole array `a`.\n",
      "        nan_policy : {'propagate', 'raise', 'omit'}, optional\n",
      "            Defines how to handle when input contains nan.\n",
      "            The following options are available (default is 'propagate'):\n",
      "        \n",
      "              * 'propagate': returns nan\n",
      "              * 'raise': throws an error\n",
      "              * 'omit': performs the calculations ignoring nan values\n",
      "        alternative : {'two-sided', 'less', 'greater'}, optional\n",
      "            Defines the alternative hypothesis.\n",
      "            The following options are available (default is 'two-sided'):\n",
      "        \n",
      "              * 'two-sided'\n",
      "              * 'less': one-sided\n",
      "              * 'greater': one-sided\n",
      "        \n",
      "            .. versionadded:: 1.6.0\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        statistic : float or array\n",
      "            t-statistic.\n",
      "        pvalue : float or array\n",
      "            Two-sided p-value.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy import stats\n",
      "        >>> rng = np.random.default_rng()\n",
      "        >>> rvs = stats.norm.rvs(loc=5, scale=10, size=(50, 2), random_state=rng)\n",
      "        \n",
      "        Test if mean of random sample is equal to true mean, and different mean.\n",
      "        We reject the null hypothesis in the second case and don't reject it in\n",
      "        the first case.\n",
      "        \n",
      "        >>> stats.ttest_1samp(rvs, 5.0)\n",
      "        Ttest_1sampResult(statistic=array([-2.09794637, -1.75977004]), pvalue=array([0.04108952, 0.08468867]))\n",
      "        >>> stats.ttest_1samp(rvs, 0.0)\n",
      "        Ttest_1sampResult(statistic=array([1.64495065, 1.62095307]), pvalue=array([0.10638103, 0.11144602]))\n",
      "        \n",
      "        Examples using axis and non-scalar dimension for population mean.\n",
      "        \n",
      "        >>> result = stats.ttest_1samp(rvs, [5.0, 0.0])\n",
      "        >>> result.statistic\n",
      "        array([-2.09794637,  1.62095307])\n",
      "        >>> result.pvalue\n",
      "        array([0.04108952, 0.11144602])\n",
      "        \n",
      "        >>> result = stats.ttest_1samp(rvs.T, [5.0, 0.0], axis=1)\n",
      "        >>> result.statistic\n",
      "        array([-2.09794637,  1.62095307])\n",
      "        >>> result.pvalue\n",
      "        array([0.04108952, 0.11144602])\n",
      "        \n",
      "        >>> result = stats.ttest_1samp(rvs, [[5.0], [0.0]])\n",
      "        >>> result.statistic\n",
      "        array([[-2.09794637, -1.75977004],\n",
      "               [ 1.64495065,  1.62095307]])\n",
      "        >>> result.pvalue\n",
      "        array([[0.04108952, 0.08468867],\n",
      "               [0.10638103, 0.11144602]])\n",
      "    \n",
      "    ttest_ind(a, b, axis=0, equal_var=True, nan_policy='propagate', permutations=None, random_state=None, alternative='two-sided', trim=0)\n",
      "        Calculate the T-test for the means of *two independent* samples of scores.\n",
      "        \n",
      "        This is a two-sided test for the null hypothesis that 2 independent samples\n",
      "        have identical average (expected) values. This test assumes that the\n",
      "        populations have identical variances by default.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        a, b : array_like\n",
      "            The arrays must have the same shape, except in the dimension\n",
      "            corresponding to `axis` (the first, by default).\n",
      "        axis : int or None, optional\n",
      "            Axis along which to compute test. If None, compute over the whole\n",
      "            arrays, `a`, and `b`.\n",
      "        equal_var : bool, optional\n",
      "            If True (default), perform a standard independent 2 sample test\n",
      "            that assumes equal population variances [1]_.\n",
      "            If False, perform Welch's t-test, which does not assume equal\n",
      "            population variance [2]_.\n",
      "        \n",
      "            .. versionadded:: 0.11.0\n",
      "        \n",
      "        nan_policy : {'propagate', 'raise', 'omit'}, optional\n",
      "            Defines how to handle when input contains nan.\n",
      "            The following options are available (default is 'propagate'):\n",
      "        \n",
      "              * 'propagate': returns nan\n",
      "              * 'raise': throws an error\n",
      "              * 'omit': performs the calculations ignoring nan values\n",
      "        \n",
      "            The 'omit' option is not currently available for permutation tests or\n",
      "            one-sided asympyotic tests.\n",
      "        \n",
      "        permutations : non-negative int, np.inf, or None (default), optional\n",
      "            If 0 or None (default), use the t-distribution to calculate p-values.\n",
      "            Otherwise, `permutations` is  the number of random permutations that\n",
      "            will be used to estimate p-values using a permutation test. If\n",
      "            `permutations` equals or exceeds the number of distinct partitions of\n",
      "            the pooled data, an exact test is performed instead (i.e. each\n",
      "            distinct partition is used exactly once). See Notes for details.\n",
      "        \n",
      "            .. versionadded:: 1.7.0\n",
      "        \n",
      "        random_state : {None, int, `numpy.random.Generator`,\n",
      "                `numpy.random.RandomState`}, optional\n",
      "        \n",
      "            If `seed` is None (or `np.random`), the `numpy.random.RandomState`\n",
      "            singleton is used.\n",
      "            If `seed` is an int, a new ``RandomState`` instance is used,\n",
      "            seeded with `seed`.\n",
      "            If `seed` is already a ``Generator`` or ``RandomState`` instance then\n",
      "            that instance is used.\n",
      "        \n",
      "            Pseudorandom number generator state used to generate permutations\n",
      "            (used only when `permutations` is not None).\n",
      "        \n",
      "            .. versionadded:: 1.7.0\n",
      "        \n",
      "        alternative : {'two-sided', 'less', 'greater'}, optional\n",
      "            Defines the alternative hypothesis.\n",
      "            The following options are available (default is 'two-sided'):\n",
      "        \n",
      "              * 'two-sided'\n",
      "              * 'less': one-sided\n",
      "              * 'greater': one-sided\n",
      "        \n",
      "            .. versionadded:: 1.6.0\n",
      "        \n",
      "        trim : float, optional\n",
      "            If nonzero, performs a trimmed (Yuen's) t-test.\n",
      "            Defines the fraction of elements to be trimmed from each end of the\n",
      "            input samples. If 0 (default), no elements will be trimmed from either\n",
      "            side. The number of trimmed elements from each tail is the floor of the\n",
      "            trim times the number of elements. Valid range is [0, .5).\n",
      "        \n",
      "            .. versionadded:: 1.7\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        statistic : float or array\n",
      "            The calculated t-statistic.\n",
      "        pvalue : float or array\n",
      "            The two-tailed p-value.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        Suppose we observe two independent samples, e.g. flower petal lengths, and\n",
      "        we are considering whether the two samples were drawn from the same\n",
      "        population (e.g. the same species of flower or two species with similar\n",
      "        petal characteristics) or two different populations.\n",
      "        \n",
      "        The t-test quantifies the difference between the arithmetic means\n",
      "        of the two samples. The p-value quantifies the probability of observing\n",
      "        as or more extreme values assuming the null hypothesis, that the\n",
      "        samples are drawn from populations with the same population means, is true.\n",
      "        A p-value larger than a chosen threshold (e.g. 5% or 1%) indicates that\n",
      "        our observation is not so unlikely to have occurred by chance. Therefore,\n",
      "        we do not reject the null hypothesis of equal population means.\n",
      "        If the p-value is smaller than our threshold, then we have evidence\n",
      "        against the null hypothesis of equal population means.\n",
      "        \n",
      "        By default, the p-value is determined by comparing the t-statistic of the\n",
      "        observed data against a theoretical t-distribution.\n",
      "        When ``1 < permutations < binom(n, k)``, where\n",
      "        \n",
      "        * ``k`` is the number of observations in `a`,\n",
      "        * ``n`` is the total number of observations in `a` and `b`, and\n",
      "        * ``binom(n, k)`` is the binomial coefficient (``n`` choose ``k``),\n",
      "        \n",
      "        the data are pooled (concatenated), randomly assigned to either group `a`\n",
      "        or `b`, and the t-statistic is calculated. This process is performed\n",
      "        repeatedly (`permutation` times), generating a distribution of the\n",
      "        t-statistic under the null hypothesis, and the t-statistic of the observed\n",
      "        data is compared to this distribution to determine the p-value. When\n",
      "        ``permutations >= binom(n, k)``, an exact test is performed: the data are\n",
      "        partitioned between the groups in each distinct way exactly once.\n",
      "        \n",
      "        The permutation test can be computationally expensive and not necessarily\n",
      "        more accurate than the analytical test, but it does not make strong\n",
      "        assumptions about the shape of the underlying distribution.\n",
      "        \n",
      "        Use of trimming is commonly referred to as the trimmed t-test. At times\n",
      "        called Yuen's t-test, this is an extension of Welch's t-test, with the\n",
      "        difference being the use of winsorized means in calculation of the variance\n",
      "        and the trimmed sample size in calculation of the statistic. Trimming is\n",
      "        reccomended if the underlying distribution is long-tailed or contaminated\n",
      "        with outliers [4]_.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] https://en.wikipedia.org/wiki/T-test#Independent_two-sample_t-test\n",
      "        \n",
      "        .. [2] https://en.wikipedia.org/wiki/Welch%27s_t-test\n",
      "        \n",
      "        .. [3] http://en.wikipedia.org/wiki/Resampling_%28statistics%29\n",
      "        \n",
      "        .. [4] Yuen, Karen K. \"The Two-Sample Trimmed t for Unequal Population\n",
      "               Variances.\" Biometrika, vol. 61, no. 1, 1974, pp. 165-170. JSTOR,\n",
      "               www.jstor.org/stable/2334299. Accessed 30 Mar. 2021.\n",
      "        \n",
      "        .. [5] Yuen, Karen K., and W. J. Dixon. \"The Approximate Behaviour and\n",
      "               Performance of the Two-Sample Trimmed t.\" Biometrika, vol. 60,\n",
      "               no. 2, 1973, pp. 369-374. JSTOR, www.jstor.org/stable/2334550.\n",
      "               Accessed 30 Mar. 2021.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy import stats\n",
      "        >>> rng = np.random.default_rng()\n",
      "        \n",
      "        Test with sample with identical means:\n",
      "        \n",
      "        >>> rvs1 = stats.norm.rvs(loc=5, scale=10, size=500, random_state=rng)\n",
      "        >>> rvs2 = stats.norm.rvs(loc=5, scale=10, size=500, random_state=rng)\n",
      "        >>> stats.ttest_ind(rvs1, rvs2)\n",
      "        Ttest_indResult(statistic=-0.4390847099199348, pvalue=0.6606952038870015)\n",
      "        >>> stats.ttest_ind(rvs1, rvs2, equal_var=False)\n",
      "        Ttest_indResult(statistic=-0.4390847099199348, pvalue=0.6606952553131064)\n",
      "        \n",
      "        `ttest_ind` underestimates p for unequal variances:\n",
      "        \n",
      "        >>> rvs3 = stats.norm.rvs(loc=5, scale=20, size=500, random_state=rng)\n",
      "        >>> stats.ttest_ind(rvs1, rvs3)\n",
      "        Ttest_indResult(statistic=-1.6370984482905417, pvalue=0.1019251574705033)\n",
      "        >>> stats.ttest_ind(rvs1, rvs3, equal_var=False)\n",
      "        Ttest_indResult(statistic=-1.637098448290542, pvalue=0.10202110497954867)\n",
      "        \n",
      "        When ``n1 != n2``, the equal variance t-statistic is no longer equal to the\n",
      "        unequal variance t-statistic:\n",
      "        \n",
      "        >>> rvs4 = stats.norm.rvs(loc=5, scale=20, size=100, random_state=rng)\n",
      "        >>> stats.ttest_ind(rvs1, rvs4)\n",
      "        Ttest_indResult(statistic=-1.9481646859513422, pvalue=0.05186270935842703)\n",
      "        >>> stats.ttest_ind(rvs1, rvs4, equal_var=False)\n",
      "        Ttest_indResult(statistic=-1.3146566100751664, pvalue=0.1913495266513811)\n",
      "        \n",
      "        T-test with different means, variance, and n:\n",
      "        \n",
      "        >>> rvs5 = stats.norm.rvs(loc=8, scale=20, size=100, random_state=rng)\n",
      "        >>> stats.ttest_ind(rvs1, rvs5)\n",
      "        Ttest_indResult(statistic=-2.8415950600298774, pvalue=0.0046418707568707885)\n",
      "        >>> stats.ttest_ind(rvs1, rvs5, equal_var=False)\n",
      "        Ttest_indResult(statistic=-1.8686598649188084, pvalue=0.06434714193919686)\n",
      "        \n",
      "        When performing a permutation test, more permutations typically yields\n",
      "        more accurate results. Use a ``np.random.Generator`` to ensure\n",
      "        reproducibility:\n",
      "        \n",
      "        >>> stats.ttest_ind(rvs1, rvs5, permutations=10000,\n",
      "        ...                 random_state=rng)\n",
      "        Ttest_indResult(statistic=-2.8415950600298774, pvalue=0.0052)\n",
      "        \n",
      "        Take these two samples, one of which has an extreme tail.\n",
      "        \n",
      "        >>> a = (56, 128.6, 12, 123.8, 64.34, 78, 763.3)\n",
      "        >>> b = (1.1, 2.9, 4.2)\n",
      "        \n",
      "        Use the `trim` keyword to perform a trimmed (Yuen) t-test. For example,\n",
      "        using 20% trimming, ``trim=.2``, the test will reduce the impact of one\n",
      "        (``np.floor(trim*len(a))``) element from each tail of sample `a`. It will\n",
      "        have no effect on sample `b` because ``np.floor(trim*len(b))`` is 0.\n",
      "        \n",
      "        >>> stats.ttest_ind(a, b, trim=.2)\n",
      "        Ttest_indResult(statistic=3.4463884028073513,\n",
      "                        pvalue=0.01369338726499547)\n",
      "    \n",
      "    ttest_ind_from_stats(mean1, std1, nobs1, mean2, std2, nobs2, equal_var=True, alternative='two-sided')\n",
      "        T-test for means of two independent samples from descriptive statistics.\n",
      "        \n",
      "        This is a two-sided test for the null hypothesis that two independent\n",
      "        samples have identical average (expected) values.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        mean1 : array_like\n",
      "            The mean(s) of sample 1.\n",
      "        std1 : array_like\n",
      "            The standard deviation(s) of sample 1.\n",
      "        nobs1 : array_like\n",
      "            The number(s) of observations of sample 1.\n",
      "        mean2 : array_like\n",
      "            The mean(s) of sample 2.\n",
      "        std2 : array_like\n",
      "            The standard deviations(s) of sample 2.\n",
      "        nobs2 : array_like\n",
      "            The number(s) of observations of sample 2.\n",
      "        equal_var : bool, optional\n",
      "            If True (default), perform a standard independent 2 sample test\n",
      "            that assumes equal population variances [1]_.\n",
      "            If False, perform Welch's t-test, which does not assume equal\n",
      "            population variance [2]_.\n",
      "        alternative : {'two-sided', 'less', 'greater'}, optional\n",
      "            Defines the alternative hypothesis.\n",
      "            The following options are available (default is 'two-sided'):\n",
      "        \n",
      "              * 'two-sided'\n",
      "              * 'less': one-sided\n",
      "              * 'greater': one-sided\n",
      "        \n",
      "            .. versionadded:: 1.6.0\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        statistic : float or array\n",
      "            The calculated t-statistics.\n",
      "        pvalue : float or array\n",
      "            The two-tailed p-value.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        scipy.stats.ttest_ind\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        .. versionadded:: 0.16.0\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] https://en.wikipedia.org/wiki/T-test#Independent_two-sample_t-test\n",
      "        \n",
      "        .. [2] https://en.wikipedia.org/wiki/Welch%27s_t-test\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        Suppose we have the summary data for two samples, as follows::\n",
      "        \n",
      "                             Sample   Sample\n",
      "                       Size   Mean   Variance\n",
      "            Sample 1    13    15.0     87.5\n",
      "            Sample 2    11    12.0     39.0\n",
      "        \n",
      "        Apply the t-test to this data (with the assumption that the population\n",
      "        variances are equal):\n",
      "        \n",
      "        >>> from scipy.stats import ttest_ind_from_stats\n",
      "        >>> ttest_ind_from_stats(mean1=15.0, std1=np.sqrt(87.5), nobs1=13,\n",
      "        ...                      mean2=12.0, std2=np.sqrt(39.0), nobs2=11)\n",
      "        Ttest_indResult(statistic=0.9051358093310269, pvalue=0.3751996797581487)\n",
      "        \n",
      "        For comparison, here is the data from which those summary statistics\n",
      "        were taken.  With this data, we can compute the same result using\n",
      "        `scipy.stats.ttest_ind`:\n",
      "        \n",
      "        >>> a = np.array([1, 3, 4, 6, 11, 13, 15, 19, 22, 24, 25, 26, 26])\n",
      "        >>> b = np.array([2, 4, 6, 9, 11, 13, 14, 15, 18, 19, 21])\n",
      "        >>> from scipy.stats import ttest_ind\n",
      "        >>> ttest_ind(a, b)\n",
      "        Ttest_indResult(statistic=0.905135809331027, pvalue=0.3751996797581486)\n",
      "        \n",
      "        Suppose we instead have binary data and would like to apply a t-test to\n",
      "        compare the proportion of 1s in two independent groups::\n",
      "        \n",
      "                              Number of    Sample     Sample\n",
      "                        Size    ones        Mean     Variance\n",
      "            Sample 1    150      30         0.2        0.16\n",
      "            Sample 2    200      45         0.225      0.174375\n",
      "        \n",
      "        The sample mean :math:`\\hat{p}` is the proportion of ones in the sample\n",
      "        and the variance for a binary observation is estimated by\n",
      "        :math:`\\hat{p}(1-\\hat{p})`.\n",
      "        \n",
      "        >>> ttest_ind_from_stats(mean1=0.2, std1=np.sqrt(0.16), nobs1=150,\n",
      "        ...                      mean2=0.225, std2=np.sqrt(0.17437), nobs2=200)\n",
      "        Ttest_indResult(statistic=-0.564327545549774, pvalue=0.5728947691244874)\n",
      "        \n",
      "        For comparison, we could compute the t statistic and p-value using\n",
      "        arrays of 0s and 1s and `scipy.stat.ttest_ind`, as above.\n",
      "        \n",
      "        >>> group1 = np.array([1]*30 + [0]*(150-30))\n",
      "        >>> group2 = np.array([1]*45 + [0]*(200-45))\n",
      "        >>> ttest_ind(group1, group2)\n",
      "        Ttest_indResult(statistic=-0.5627179589855622, pvalue=0.573989277115258)\n",
      "    \n",
      "    ttest_rel(a, b, axis=0, nan_policy='propagate', alternative='two-sided')\n",
      "        Calculate the t-test on TWO RELATED samples of scores, a and b.\n",
      "        \n",
      "        This is a two-sided test for the null hypothesis that 2 related or\n",
      "        repeated samples have identical average (expected) values.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        a, b : array_like\n",
      "            The arrays must have the same shape.\n",
      "        axis : int or None, optional\n",
      "            Axis along which to compute test. If None, compute over the whole\n",
      "            arrays, `a`, and `b`.\n",
      "        nan_policy : {'propagate', 'raise', 'omit'}, optional\n",
      "            Defines how to handle when input contains nan.\n",
      "            The following options are available (default is 'propagate'):\n",
      "        \n",
      "              * 'propagate': returns nan\n",
      "              * 'raise': throws an error\n",
      "              * 'omit': performs the calculations ignoring nan values\n",
      "        alternative : {'two-sided', 'less', 'greater'}, optional\n",
      "            Defines the alternative hypothesis.\n",
      "            The following options are available (default is 'two-sided'):\n",
      "        \n",
      "              * 'two-sided'\n",
      "              * 'less': one-sided\n",
      "              * 'greater': one-sided\n",
      "        \n",
      "              .. versionadded:: 1.6.0\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        statistic : float or array\n",
      "            t-statistic.\n",
      "        pvalue : float or array\n",
      "            Two-sided p-value.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        Examples for use are scores of the same set of student in\n",
      "        different exams, or repeated sampling from the same units. The\n",
      "        test measures whether the average score differs significantly\n",
      "        across samples (e.g. exams). If we observe a large p-value, for\n",
      "        example greater than 0.05 or 0.1 then we cannot reject the null\n",
      "        hypothesis of identical average scores. If the p-value is smaller\n",
      "        than the threshold, e.g. 1%, 5% or 10%, then we reject the null\n",
      "        hypothesis of equal averages. Small p-values are associated with\n",
      "        large t-statistics.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        https://en.wikipedia.org/wiki/T-test#Dependent_t-test_for_paired_samples\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy import stats\n",
      "        >>> rng = np.random.default_rng()\n",
      "        \n",
      "        >>> rvs1 = stats.norm.rvs(loc=5, scale=10, size=500, random_state=rng)\n",
      "        >>> rvs2 = (stats.norm.rvs(loc=5, scale=10, size=500, random_state=rng)\n",
      "        ...         + stats.norm.rvs(scale=0.2, size=500, random_state=rng))\n",
      "        >>> stats.ttest_rel(rvs1, rvs2)\n",
      "        Ttest_relResult(statistic=-0.4549717054410304, pvalue=0.6493274702088672)\n",
      "        >>> rvs3 = (stats.norm.rvs(loc=8, scale=10, size=500, random_state=rng)\n",
      "        ...         + stats.norm.rvs(scale=0.2, size=500, random_state=rng))\n",
      "        >>> stats.ttest_rel(rvs1, rvs3)\n",
      "        Ttest_relResult(statistic=-5.879467544540889, pvalue=7.540777129099917e-09)\n",
      "    \n",
      "    tvar(a, limits=None, inclusive=(True, True), axis=0, ddof=1)\n",
      "        Compute the trimmed variance.\n",
      "        \n",
      "        This function computes the sample variance of an array of values,\n",
      "        while ignoring values which are outside of given `limits`.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        a : array_like\n",
      "            Array of values.\n",
      "        limits : None or (lower limit, upper limit), optional\n",
      "            Values in the input array less than the lower limit or greater than the\n",
      "            upper limit will be ignored. When limits is None, then all values are\n",
      "            used. Either of the limit values in the tuple can also be None\n",
      "            representing a half-open interval.  The default value is None.\n",
      "        inclusive : (bool, bool), optional\n",
      "            A tuple consisting of the (lower flag, upper flag).  These flags\n",
      "            determine whether values exactly equal to the lower or upper limits\n",
      "            are included.  The default value is (True, True).\n",
      "        axis : int or None, optional\n",
      "            Axis along which to operate. Default is 0. If None, compute over the\n",
      "            whole array `a`.\n",
      "        ddof : int, optional\n",
      "            Delta degrees of freedom.  Default is 1.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        tvar : float\n",
      "            Trimmed variance.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        `tvar` computes the unbiased sample variance, i.e. it uses a correction\n",
      "        factor ``n / (n - 1)``.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy import stats\n",
      "        >>> x = np.arange(20)\n",
      "        >>> stats.tvar(x)\n",
      "        35.0\n",
      "        >>> stats.tvar(x, (3,17))\n",
      "        20.0\n",
      "    \n",
      "    variation(a, axis=0, nan_policy='propagate', ddof=0)\n",
      "        Compute the coefficient of variation.\n",
      "        \n",
      "        The coefficient of variation is the standard deviation divided by the\n",
      "        mean.  This function is equivalent to::\n",
      "        \n",
      "            np.std(x, axis=axis, ddof=ddof) / np.mean(x)\n",
      "        \n",
      "        The default for ``ddof`` is 0, but many definitions of the coefficient\n",
      "        of variation use the square root of the unbiased sample variance\n",
      "        for the sample standard deviation, which corresponds to ``ddof=1``.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        a : array_like\n",
      "            Input array.\n",
      "        axis : int or None, optional\n",
      "            Axis along which to calculate the coefficient of variation. Default\n",
      "            is 0. If None, compute over the whole array `a`.\n",
      "        nan_policy : {'propagate', 'raise', 'omit'}, optional\n",
      "            Defines how to handle when input contains nan.\n",
      "            The following options are available (default is 'propagate'):\n",
      "        \n",
      "            * 'propagate': returns nan\n",
      "            * 'raise': throws an error\n",
      "            * 'omit': performs the calculations ignoring nan values\n",
      "        \n",
      "        ddof : int, optional\n",
      "            Delta degrees of freedom.  Default is 0.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        variation : ndarray\n",
      "            The calculated variation along the requested axis.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] Zwillinger, D. and Kokoska, S. (2000). CRC Standard\n",
      "           Probability and Statistics Tables and Formulae. Chapman & Hall: New\n",
      "           York. 2000.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy.stats import variation\n",
      "        >>> variation([1, 2, 3, 4, 5])\n",
      "        0.47140452079103173\n",
      "    \n",
      "    wasserstein_distance(u_values, v_values, u_weights=None, v_weights=None)\n",
      "        Compute the first Wasserstein distance between two 1D distributions.\n",
      "        \n",
      "        This distance is also known as the earth mover's distance, since it can be\n",
      "        seen as the minimum amount of \"work\" required to transform :math:`u` into\n",
      "        :math:`v`, where \"work\" is measured as the amount of distribution weight\n",
      "        that must be moved, multiplied by the distance it has to be moved.\n",
      "        \n",
      "        .. versionadded:: 1.0.0\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        u_values, v_values : array_like\n",
      "            Values observed in the (empirical) distribution.\n",
      "        u_weights, v_weights : array_like, optional\n",
      "            Weight for each value. If unspecified, each value is assigned the same\n",
      "            weight.\n",
      "            `u_weights` (resp. `v_weights`) must have the same length as\n",
      "            `u_values` (resp. `v_values`). If the weight sum differs from 1, it\n",
      "            must still be positive and finite so that the weights can be normalized\n",
      "            to sum to 1.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        distance : float\n",
      "            The computed distance between the distributions.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        The first Wasserstein distance between the distributions :math:`u` and\n",
      "        :math:`v` is:\n",
      "        \n",
      "        .. math::\n",
      "        \n",
      "            l_1 (u, v) = \\inf_{\\pi \\in \\Gamma (u, v)} \\int_{\\mathbb{R} \\times\n",
      "            \\mathbb{R}} |x-y| \\mathrm{d} \\pi (x, y)\n",
      "        \n",
      "        where :math:`\\Gamma (u, v)` is the set of (probability) distributions on\n",
      "        :math:`\\mathbb{R} \\times \\mathbb{R}` whose marginals are :math:`u` and\n",
      "        :math:`v` on the first and second factors respectively.\n",
      "        \n",
      "        If :math:`U` and :math:`V` are the respective CDFs of :math:`u` and\n",
      "        :math:`v`, this distance also equals to:\n",
      "        \n",
      "        .. math::\n",
      "        \n",
      "            l_1(u, v) = \\int_{-\\infty}^{+\\infty} |U-V|\n",
      "        \n",
      "        See [2]_ for a proof of the equivalence of both definitions.\n",
      "        \n",
      "        The input distributions can be empirical, therefore coming from samples\n",
      "        whose values are effectively inputs of the function, or they can be seen as\n",
      "        generalized functions, in which case they are weighted sums of Dirac delta\n",
      "        functions located at the specified values.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] \"Wasserstein metric\", https://en.wikipedia.org/wiki/Wasserstein_metric\n",
      "        .. [2] Ramdas, Garcia, Cuturi \"On Wasserstein Two Sample Testing and Related\n",
      "               Families of Nonparametric Tests\" (2015). :arXiv:`1509.02237`.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy.stats import wasserstein_distance\n",
      "        >>> wasserstein_distance([0, 1, 3], [5, 6, 8])\n",
      "        5.0\n",
      "        >>> wasserstein_distance([0, 1], [0, 1], [3, 1], [2, 2])\n",
      "        0.25\n",
      "        >>> wasserstein_distance([3.4, 3.9, 7.5, 7.8], [4.5, 1.4],\n",
      "        ...                      [1.4, 0.9, 3.1, 7.2], [3.2, 3.5])\n",
      "        4.0781331438047861\n",
      "    \n",
      "    weightedtau(x, y, rank=True, weigher=None, additive=True)\n",
      "        Compute a weighted version of Kendall's :math:`\\tau`.\n",
      "        \n",
      "        The weighted :math:`\\tau` is a weighted version of Kendall's\n",
      "        :math:`\\tau` in which exchanges of high weight are more influential than\n",
      "        exchanges of low weight. The default parameters compute the additive\n",
      "        hyperbolic version of the index, :math:`\\tau_\\mathrm h`, which has\n",
      "        been shown to provide the best balance between important and\n",
      "        unimportant elements [1]_.\n",
      "        \n",
      "        The weighting is defined by means of a rank array, which assigns a\n",
      "        nonnegative rank to each element (higher importance ranks being\n",
      "        associated with smaller values, e.g., 0 is the highest possible rank),\n",
      "        and a weigher function, which assigns a weight based on the rank to\n",
      "        each element. The weight of an exchange is then the sum or the product\n",
      "        of the weights of the ranks of the exchanged elements. The default\n",
      "        parameters compute :math:`\\tau_\\mathrm h`: an exchange between\n",
      "        elements with rank :math:`r` and :math:`s` (starting from zero) has\n",
      "        weight :math:`1/(r+1) + 1/(s+1)`.\n",
      "        \n",
      "        Specifying a rank array is meaningful only if you have in mind an\n",
      "        external criterion of importance. If, as it usually happens, you do\n",
      "        not have in mind a specific rank, the weighted :math:`\\tau` is\n",
      "        defined by averaging the values obtained using the decreasing\n",
      "        lexicographical rank by (`x`, `y`) and by (`y`, `x`). This is the\n",
      "        behavior with default parameters. Note that the convention used\n",
      "        here for ranking (lower values imply higher importance) is opposite\n",
      "        to that used by other SciPy statistical functions.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x, y : array_like\n",
      "            Arrays of scores, of the same shape. If arrays are not 1-D, they will\n",
      "            be flattened to 1-D.\n",
      "        rank : array_like of ints or bool, optional\n",
      "            A nonnegative rank assigned to each element. If it is None, the\n",
      "            decreasing lexicographical rank by (`x`, `y`) will be used: elements of\n",
      "            higher rank will be those with larger `x`-values, using `y`-values to\n",
      "            break ties (in particular, swapping `x` and `y` will give a different\n",
      "            result). If it is False, the element indices will be used\n",
      "            directly as ranks. The default is True, in which case this\n",
      "            function returns the average of the values obtained using the\n",
      "            decreasing lexicographical rank by (`x`, `y`) and by (`y`, `x`).\n",
      "        weigher : callable, optional\n",
      "            The weigher function. Must map nonnegative integers (zero\n",
      "            representing the most important element) to a nonnegative weight.\n",
      "            The default, None, provides hyperbolic weighing, that is,\n",
      "            rank :math:`r` is mapped to weight :math:`1/(r+1)`.\n",
      "        additive : bool, optional\n",
      "            If True, the weight of an exchange is computed by adding the\n",
      "            weights of the ranks of the exchanged elements; otherwise, the weights\n",
      "            are multiplied. The default is True.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        correlation : float\n",
      "           The weighted :math:`\\tau` correlation index.\n",
      "        pvalue : float\n",
      "           Presently ``np.nan``, as the null statistics is unknown (even in the\n",
      "           additive hyperbolic case).\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        kendalltau : Calculates Kendall's tau.\n",
      "        spearmanr : Calculates a Spearman rank-order correlation coefficient.\n",
      "        theilslopes : Computes the Theil-Sen estimator for a set of points (x, y).\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        This function uses an :math:`O(n \\log n)`, mergesort-based algorithm\n",
      "        [1]_ that is a weighted extension of Knight's algorithm for Kendall's\n",
      "        :math:`\\tau` [2]_. It can compute Shieh's weighted :math:`\\tau` [3]_\n",
      "        between rankings without ties (i.e., permutations) by setting\n",
      "        `additive` and `rank` to False, as the definition given in [1]_ is a\n",
      "        generalization of Shieh's.\n",
      "        \n",
      "        NaNs are considered the smallest possible score.\n",
      "        \n",
      "        .. versionadded:: 0.19.0\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] Sebastiano Vigna, \"A weighted correlation index for rankings with\n",
      "               ties\", Proceedings of the 24th international conference on World\n",
      "               Wide Web, pp. 1166-1176, ACM, 2015.\n",
      "        .. [2] W.R. Knight, \"A Computer Method for Calculating Kendall's Tau with\n",
      "               Ungrouped Data\", Journal of the American Statistical Association,\n",
      "               Vol. 61, No. 314, Part 1, pp. 436-439, 1966.\n",
      "        .. [3] Grace S. Shieh. \"A weighted Kendall's tau statistic\", Statistics &\n",
      "               Probability Letters, Vol. 39, No. 1, pp. 17-24, 1998.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy import stats\n",
      "        >>> x = [12, 2, 1, 12, 2]\n",
      "        >>> y = [1, 4, 7, 1, 0]\n",
      "        >>> tau, p_value = stats.weightedtau(x, y)\n",
      "        >>> tau\n",
      "        -0.56694968153682723\n",
      "        >>> p_value\n",
      "        nan\n",
      "        >>> tau, p_value = stats.weightedtau(x, y, additive=False)\n",
      "        >>> tau\n",
      "        -0.62205716951801038\n",
      "        \n",
      "        NaNs are considered the smallest possible score:\n",
      "        \n",
      "        >>> x = [12, 2, 1, 12, 2]\n",
      "        >>> y = [1, 4, 7, 1, np.nan]\n",
      "        >>> tau, _ = stats.weightedtau(x, y)\n",
      "        >>> tau\n",
      "        -0.56694968153682723\n",
      "        \n",
      "        This is exactly Kendall's tau:\n",
      "        \n",
      "        >>> x = [12, 2, 1, 12, 2]\n",
      "        >>> y = [1, 4, 7, 1, 0]\n",
      "        >>> tau, _ = stats.weightedtau(x, y, weigher=lambda x: 1)\n",
      "        >>> tau\n",
      "        -0.47140452079103173\n",
      "        \n",
      "        >>> x = [12, 2, 1, 12, 2]\n",
      "        >>> y = [1, 4, 7, 1, 0]\n",
      "        >>> stats.weightedtau(x, y, rank=None)\n",
      "        WeightedTauResult(correlation=-0.4157652301037516, pvalue=nan)\n",
      "        >>> stats.weightedtau(y, x, rank=None)\n",
      "        WeightedTauResult(correlation=-0.7181341329699028, pvalue=nan)\n",
      "    \n",
      "    wilcoxon(x, y=None, zero_method='wilcox', correction=False, alternative='two-sided', mode='auto')\n",
      "        Calculate the Wilcoxon signed-rank test.\n",
      "        \n",
      "        The Wilcoxon signed-rank test tests the null hypothesis that two\n",
      "        related paired samples come from the same distribution. In particular,\n",
      "        it tests whether the distribution of the differences x - y is symmetric\n",
      "        about zero. It is a non-parametric version of the paired T-test.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x : array_like\n",
      "            Either the first set of measurements (in which case ``y`` is the second\n",
      "            set of measurements), or the differences between two sets of\n",
      "            measurements (in which case ``y`` is not to be specified.)  Must be\n",
      "            one-dimensional.\n",
      "        y : array_like, optional\n",
      "            Either the second set of measurements (if ``x`` is the first set of\n",
      "            measurements), or not specified (if ``x`` is the differences between\n",
      "            two sets of measurements.)  Must be one-dimensional.\n",
      "        zero_method : {\"pratt\", \"wilcox\", \"zsplit\"}, optional\n",
      "            The following options are available (default is \"wilcox\"):\n",
      "        \n",
      "              * \"pratt\": Includes zero-differences in the ranking process,\n",
      "                but drops the ranks of the zeros, see [4]_, (more conservative).\n",
      "              * \"wilcox\": Discards all zero-differences, the default.\n",
      "              * \"zsplit\": Includes zero-differences in the ranking process and\n",
      "                split the zero rank between positive and negative ones.\n",
      "        correction : bool, optional\n",
      "            If True, apply continuity correction by adjusting the Wilcoxon rank\n",
      "            statistic by 0.5 towards the mean value when computing the\n",
      "            z-statistic if a normal approximation is used.  Default is False.\n",
      "        alternative : {\"two-sided\", \"greater\", \"less\"}, optional\n",
      "            The alternative hypothesis to be tested, see Notes. Default is\n",
      "            \"two-sided\".\n",
      "        mode : {\"auto\", \"exact\", \"approx\"}\n",
      "            Method to calculate the p-value, see Notes. Default is \"auto\".\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        statistic : float\n",
      "            If ``alternative`` is \"two-sided\", the sum of the ranks of the\n",
      "            differences above or below zero, whichever is smaller.\n",
      "            Otherwise the sum of the ranks of the differences above zero.\n",
      "        pvalue : float\n",
      "            The p-value for the test depending on ``alternative`` and ``mode``.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        kruskal, mannwhitneyu\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        The test has been introduced in [4]_. Given n independent samples\n",
      "        (xi, yi) from a bivariate distribution (i.e. paired samples),\n",
      "        it computes the differences di = xi - yi. One assumption of the test\n",
      "        is that the differences are symmetric, see [2]_.\n",
      "        The two-sided test has the null hypothesis that the median of the\n",
      "        differences is zero against the alternative that it is different from\n",
      "        zero. The one-sided test has the null hypothesis that the median is\n",
      "        positive against the alternative that it is negative\n",
      "        (``alternative == 'less'``), or vice versa (``alternative == 'greater.'``).\n",
      "        \n",
      "        To derive the p-value, the exact distribution (``mode == 'exact'``)\n",
      "        can be used for sample sizes of up to 25. The default ``mode == 'auto'``\n",
      "        uses the exact distribution if there are at most 25 observations and no\n",
      "        ties, otherwise a normal approximation is used (``mode == 'approx'``).\n",
      "        \n",
      "        The treatment of ties can be controlled by the parameter `zero_method`.\n",
      "        If ``zero_method == 'pratt'``, the normal approximation is adjusted as in\n",
      "        [5]_. A typical rule is to require that n > 20 ([2]_, p. 383).\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] https://en.wikipedia.org/wiki/Wilcoxon_signed-rank_test\n",
      "        .. [2] Conover, W.J., Practical Nonparametric Statistics, 1971.\n",
      "        .. [3] Pratt, J.W., Remarks on Zeros and Ties in the Wilcoxon Signed\n",
      "           Rank Procedures, Journal of the American Statistical Association,\n",
      "           Vol. 54, 1959, pp. 655-667. :doi:`10.1080/01621459.1959.10501526`\n",
      "        .. [4] Wilcoxon, F., Individual Comparisons by Ranking Methods,\n",
      "           Biometrics Bulletin, Vol. 1, 1945, pp. 80-83. :doi:`10.2307/3001968`\n",
      "        .. [5] Cureton, E.E., The Normal Approximation to the Signed-Rank\n",
      "           Sampling Distribution When Zero Differences are Present,\n",
      "           Journal of the American Statistical Association, Vol. 62, 1967,\n",
      "           pp. 1068-1069. :doi:`10.1080/01621459.1967.10500917`\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        In [4]_, the differences in height between cross- and self-fertilized\n",
      "        corn plants is given as follows:\n",
      "        \n",
      "        >>> d = [6, 8, 14, 16, 23, 24, 28, 29, 41, -48, 49, 56, 60, -67, 75]\n",
      "        \n",
      "        Cross-fertilized plants appear to be be higher. To test the null\n",
      "        hypothesis that there is no height difference, we can apply the\n",
      "        two-sided test:\n",
      "        \n",
      "        >>> from scipy.stats import wilcoxon\n",
      "        >>> w, p = wilcoxon(d)\n",
      "        >>> w, p\n",
      "        (24.0, 0.041259765625)\n",
      "        \n",
      "        Hence, we would reject the null hypothesis at a confidence level of 5%,\n",
      "        concluding that there is a difference in height between the groups.\n",
      "        To confirm that the median of the differences can be assumed to be\n",
      "        positive, we use:\n",
      "        \n",
      "        >>> w, p = wilcoxon(d, alternative='greater')\n",
      "        >>> w, p\n",
      "        (96.0, 0.0206298828125)\n",
      "        \n",
      "        This shows that the null hypothesis that the median is negative can be\n",
      "        rejected at a confidence level of 5% in favor of the alternative that\n",
      "        the median is greater than zero. The p-values above are exact. Using the\n",
      "        normal approximation gives very similar values:\n",
      "        \n",
      "        >>> w, p = wilcoxon(d, mode='approx')\n",
      "        >>> w, p\n",
      "        (24.0, 0.04088813291185591)\n",
      "        \n",
      "        Note that the statistic changed to 96 in the one-sided case (the sum\n",
      "        of ranks of positive differences) whereas it is 24 in the two-sided\n",
      "        case (the minimum of sum of ranks above and below zero).\n",
      "    \n",
      "    yeojohnson(x, lmbda=None)\n",
      "        Return a dataset transformed by a Yeo-Johnson power transformation.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x : ndarray\n",
      "            Input array.  Should be 1-dimensional.\n",
      "        lmbda : float, optional\n",
      "            If ``lmbda`` is ``None``, find the lambda that maximizes the\n",
      "            log-likelihood function and return it as the second output argument.\n",
      "            Otherwise the transformation is done for the given value.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        yeojohnson: ndarray\n",
      "            Yeo-Johnson power transformed array.\n",
      "        maxlog : float, optional\n",
      "            If the `lmbda` parameter is None, the second returned argument is\n",
      "            the lambda that maximizes the log-likelihood function.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        probplot, yeojohnson_normplot, yeojohnson_normmax, yeojohnson_llf, boxcox\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        The Yeo-Johnson transform is given by::\n",
      "        \n",
      "            y = ((x + 1)**lmbda - 1) / lmbda,                for x >= 0, lmbda != 0\n",
      "                log(x + 1),                                  for x >= 0, lmbda = 0\n",
      "                -((-x + 1)**(2 - lmbda) - 1) / (2 - lmbda),  for x < 0, lmbda != 2\n",
      "                -log(-x + 1),                                for x < 0, lmbda = 2\n",
      "        \n",
      "        Unlike `boxcox`, `yeojohnson` does not require the input data to be\n",
      "        positive.\n",
      "        \n",
      "        .. versionadded:: 1.2.0\n",
      "        \n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        I. Yeo and R.A. Johnson, \"A New Family of Power Transformations to\n",
      "        Improve Normality or Symmetry\", Biometrika 87.4 (2000):\n",
      "        \n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy import stats\n",
      "        >>> import matplotlib.pyplot as plt\n",
      "        \n",
      "        We generate some random variates from a non-normal distribution and make a\n",
      "        probability plot for it, to show it is non-normal in the tails:\n",
      "        \n",
      "        >>> fig = plt.figure()\n",
      "        >>> ax1 = fig.add_subplot(211)\n",
      "        >>> x = stats.loggamma.rvs(5, size=500) + 5\n",
      "        >>> prob = stats.probplot(x, dist=stats.norm, plot=ax1)\n",
      "        >>> ax1.set_xlabel('')\n",
      "        >>> ax1.set_title('Probplot against normal distribution')\n",
      "        \n",
      "        We now use `yeojohnson` to transform the data so it's closest to normal:\n",
      "        \n",
      "        >>> ax2 = fig.add_subplot(212)\n",
      "        >>> xt, lmbda = stats.yeojohnson(x)\n",
      "        >>> prob = stats.probplot(xt, dist=stats.norm, plot=ax2)\n",
      "        >>> ax2.set_title('Probplot after Yeo-Johnson transformation')\n",
      "        \n",
      "        >>> plt.show()\n",
      "    \n",
      "    yeojohnson_llf(lmb, data)\n",
      "        The yeojohnson log-likelihood function.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        lmb : scalar\n",
      "            Parameter for Yeo-Johnson transformation. See `yeojohnson` for\n",
      "            details.\n",
      "        data : array_like\n",
      "            Data to calculate Yeo-Johnson log-likelihood for. If `data` is\n",
      "            multi-dimensional, the log-likelihood is calculated along the first\n",
      "            axis.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        llf : float\n",
      "            Yeo-Johnson log-likelihood of `data` given `lmb`.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        yeojohnson, probplot, yeojohnson_normplot, yeojohnson_normmax\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        The Yeo-Johnson log-likelihood function is defined here as\n",
      "        \n",
      "        .. math::\n",
      "        \n",
      "            llf = -N/2 \\log(\\hat{\\sigma}^2) + (\\lambda - 1)\n",
      "                  \\sum_i \\text{ sign }(x_i)\\log(|x_i| + 1)\n",
      "        \n",
      "        where :math:`\\hat{\\sigma}^2` is estimated variance of the the Yeo-Johnson\n",
      "        transformed input data ``x``.\n",
      "        \n",
      "        .. versionadded:: 1.2.0\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy import stats\n",
      "        >>> import matplotlib.pyplot as plt\n",
      "        >>> from mpl_toolkits.axes_grid1.inset_locator import inset_axes\n",
      "        \n",
      "        Generate some random variates and calculate Yeo-Johnson log-likelihood\n",
      "        values for them for a range of ``lmbda`` values:\n",
      "        \n",
      "        >>> x = stats.loggamma.rvs(5, loc=10, size=1000)\n",
      "        >>> lmbdas = np.linspace(-2, 10)\n",
      "        >>> llf = np.zeros(lmbdas.shape, dtype=float)\n",
      "        >>> for ii, lmbda in enumerate(lmbdas):\n",
      "        ...     llf[ii] = stats.yeojohnson_llf(lmbda, x)\n",
      "        \n",
      "        Also find the optimal lmbda value with `yeojohnson`:\n",
      "        \n",
      "        >>> x_most_normal, lmbda_optimal = stats.yeojohnson(x)\n",
      "        \n",
      "        Plot the log-likelihood as function of lmbda.  Add the optimal lmbda as a\n",
      "        horizontal line to check that that's really the optimum:\n",
      "        \n",
      "        >>> fig = plt.figure()\n",
      "        >>> ax = fig.add_subplot(111)\n",
      "        >>> ax.plot(lmbdas, llf, 'b.-')\n",
      "        >>> ax.axhline(stats.yeojohnson_llf(lmbda_optimal, x), color='r')\n",
      "        >>> ax.set_xlabel('lmbda parameter')\n",
      "        >>> ax.set_ylabel('Yeo-Johnson log-likelihood')\n",
      "        \n",
      "        Now add some probability plots to show that where the log-likelihood is\n",
      "        maximized the data transformed with `yeojohnson` looks closest to normal:\n",
      "        \n",
      "        >>> locs = [3, 10, 4]  # 'lower left', 'center', 'lower right'\n",
      "        >>> for lmbda, loc in zip([-1, lmbda_optimal, 9], locs):\n",
      "        ...     xt = stats.yeojohnson(x, lmbda=lmbda)\n",
      "        ...     (osm, osr), (slope, intercept, r_sq) = stats.probplot(xt)\n",
      "        ...     ax_inset = inset_axes(ax, width=\"20%\", height=\"20%\", loc=loc)\n",
      "        ...     ax_inset.plot(osm, osr, 'c.', osm, slope*osm + intercept, 'k-')\n",
      "        ...     ax_inset.set_xticklabels([])\n",
      "        ...     ax_inset.set_yticklabels([])\n",
      "        ...     ax_inset.set_title(r'$\\lambda=%1.2f$' % lmbda)\n",
      "        \n",
      "        >>> plt.show()\n",
      "    \n",
      "    yeojohnson_normmax(x, brack=(-2, 2))\n",
      "        Compute optimal Yeo-Johnson transform parameter.\n",
      "        \n",
      "        Compute optimal Yeo-Johnson transform parameter for input data, using\n",
      "        maximum likelihood estimation.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x : array_like\n",
      "            Input array.\n",
      "        brack : 2-tuple, optional\n",
      "            The starting interval for a downhill bracket search with\n",
      "            `optimize.brent`. Note that this is in most cases not critical; the\n",
      "            final result is allowed to be outside this bracket.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        maxlog : float\n",
      "            The optimal transform parameter found.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        yeojohnson, yeojohnson_llf, yeojohnson_normplot\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        .. versionadded:: 1.2.0\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy import stats\n",
      "        >>> import matplotlib.pyplot as plt\n",
      "        \n",
      "        Generate some data and determine optimal ``lmbda``\n",
      "        \n",
      "        >>> rng = np.random.default_rng()\n",
      "        >>> x = stats.loggamma.rvs(5, size=30, random_state=rng) + 5\n",
      "        >>> lmax = stats.yeojohnson_normmax(x)\n",
      "        \n",
      "        >>> fig = plt.figure()\n",
      "        >>> ax = fig.add_subplot(111)\n",
      "        >>> prob = stats.yeojohnson_normplot(x, -10, 10, plot=ax)\n",
      "        >>> ax.axvline(lmax, color='r')\n",
      "        \n",
      "        >>> plt.show()\n",
      "    \n",
      "    yeojohnson_normplot(x, la, lb, plot=None, N=80)\n",
      "        Compute parameters for a Yeo-Johnson normality plot, optionally show it.\n",
      "        \n",
      "        A Yeo-Johnson normality plot shows graphically what the best\n",
      "        transformation parameter is to use in `yeojohnson` to obtain a\n",
      "        distribution that is close to normal.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x : array_like\n",
      "            Input array.\n",
      "        la, lb : scalar\n",
      "            The lower and upper bounds for the ``lmbda`` values to pass to\n",
      "            `yeojohnson` for Yeo-Johnson transformations. These are also the\n",
      "            limits of the horizontal axis of the plot if that is generated.\n",
      "        plot : object, optional\n",
      "            If given, plots the quantiles and least squares fit.\n",
      "            `plot` is an object that has to have methods \"plot\" and \"text\".\n",
      "            The `matplotlib.pyplot` module or a Matplotlib Axes object can be used,\n",
      "            or a custom object with the same methods.\n",
      "            Default is None, which means that no plot is created.\n",
      "        N : int, optional\n",
      "            Number of points on the horizontal axis (equally distributed from\n",
      "            `la` to `lb`).\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        lmbdas : ndarray\n",
      "            The ``lmbda`` values for which a Yeo-Johnson transform was done.\n",
      "        ppcc : ndarray\n",
      "            Probability Plot Correlelation Coefficient, as obtained from `probplot`\n",
      "            when fitting the Box-Cox transformed input `x` against a normal\n",
      "            distribution.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        probplot, yeojohnson, yeojohnson_normmax, yeojohnson_llf, ppcc_max\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        Even if `plot` is given, the figure is not shown or saved by\n",
      "        `boxcox_normplot`; ``plt.show()`` or ``plt.savefig('figname.png')``\n",
      "        should be used after calling `probplot`.\n",
      "        \n",
      "        .. versionadded:: 1.2.0\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy import stats\n",
      "        >>> import matplotlib.pyplot as plt\n",
      "        \n",
      "        Generate some non-normally distributed data, and create a Yeo-Johnson plot:\n",
      "        \n",
      "        >>> x = stats.loggamma.rvs(5, size=500) + 5\n",
      "        >>> fig = plt.figure()\n",
      "        >>> ax = fig.add_subplot(111)\n",
      "        >>> prob = stats.yeojohnson_normplot(x, -20, 20, plot=ax)\n",
      "        \n",
      "        Determine and plot the optimal ``lmbda`` to transform ``x`` and plot it in\n",
      "        the same plot:\n",
      "        \n",
      "        >>> _, maxlog = stats.yeojohnson(x)\n",
      "        >>> ax.axvline(maxlog, color='r')\n",
      "        \n",
      "        >>> plt.show()\n",
      "    \n",
      "    zmap(scores, compare, axis=0, ddof=0, nan_policy='propagate')\n",
      "        Calculate the relative z-scores.\n",
      "        \n",
      "        Return an array of z-scores, i.e., scores that are standardized to\n",
      "        zero mean and unit variance, where mean and variance are calculated\n",
      "        from the comparison array.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        scores : array_like\n",
      "            The input for which z-scores are calculated.\n",
      "        compare : array_like\n",
      "            The input from which the mean and standard deviation of the\n",
      "            normalization are taken; assumed to have the same dimension as\n",
      "            `scores`.\n",
      "        axis : int or None, optional\n",
      "            Axis over which mean and variance of `compare` are calculated.\n",
      "            Default is 0. If None, compute over the whole array `scores`.\n",
      "        ddof : int, optional\n",
      "            Degrees of freedom correction in the calculation of the\n",
      "            standard deviation. Default is 0.\n",
      "        nan_policy : {'propagate', 'raise', 'omit'}, optional\n",
      "            Defines how to handle the occurrence of nans in `compare`.\n",
      "            'propagate' returns nan, 'raise' raises an exception, 'omit'\n",
      "            performs the calculations ignoring nan values. Default is\n",
      "            'propagate'. Note that when the value is 'omit', nans in `scores`\n",
      "            also propagate to the output, but they do not affect the z-scores\n",
      "            computed for the non-nan values.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        zscore : array_like\n",
      "            Z-scores, in the same shape as `scores`.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        This function preserves ndarray subclasses, and works also with\n",
      "        matrices and masked arrays (it uses `asanyarray` instead of\n",
      "        `asarray` for parameters).\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy.stats import zmap\n",
      "        >>> a = [0.5, 2.0, 2.5, 3]\n",
      "        >>> b = [0, 1, 2, 3, 4]\n",
      "        >>> zmap(a, b)\n",
      "        array([-1.06066017,  0.        ,  0.35355339,  0.70710678])\n",
      "    \n",
      "    zscore(a, axis=0, ddof=0, nan_policy='propagate')\n",
      "        Compute the z score.\n",
      "        \n",
      "        Compute the z score of each value in the sample, relative to the\n",
      "        sample mean and standard deviation.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        a : array_like\n",
      "            An array like object containing the sample data.\n",
      "        axis : int or None, optional\n",
      "            Axis along which to operate. Default is 0. If None, compute over\n",
      "            the whole array `a`.\n",
      "        ddof : int, optional\n",
      "            Degrees of freedom correction in the calculation of the\n",
      "            standard deviation. Default is 0.\n",
      "        nan_policy : {'propagate', 'raise', 'omit'}, optional\n",
      "            Defines how to handle when input contains nan. 'propagate' returns nan,\n",
      "            'raise' throws an error, 'omit' performs the calculations ignoring nan\n",
      "            values. Default is 'propagate'.  Note that when the value is 'omit',\n",
      "            nans in the input also propagate to the output, but they do not affect\n",
      "            the z-scores computed for the non-nan values.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        zscore : array_like\n",
      "            The z-scores, standardized by mean and standard deviation of\n",
      "            input array `a`.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        This function preserves ndarray subclasses, and works also with\n",
      "        matrices and masked arrays (it uses `asanyarray` instead of\n",
      "        `asarray` for parameters).\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> a = np.array([ 0.7972,  0.0767,  0.4383,  0.7866,  0.8091,\n",
      "        ...                0.1954,  0.6307,  0.6599,  0.1065,  0.0508])\n",
      "        >>> from scipy import stats\n",
      "        >>> stats.zscore(a)\n",
      "        array([ 1.1273, -1.247 , -0.0552,  1.0923,  1.1664, -0.8559,  0.5786,\n",
      "                0.6748, -1.1488, -1.3324])\n",
      "        \n",
      "        Computing along a specified axis, using n-1 degrees of freedom\n",
      "        (``ddof=1``) to calculate the standard deviation:\n",
      "        \n",
      "        >>> b = np.array([[ 0.3148,  0.0478,  0.6243,  0.4608],\n",
      "        ...               [ 0.7149,  0.0775,  0.6072,  0.9656],\n",
      "        ...               [ 0.6341,  0.1403,  0.9759,  0.4064],\n",
      "        ...               [ 0.5918,  0.6948,  0.904 ,  0.3721],\n",
      "        ...               [ 0.0921,  0.2481,  0.1188,  0.1366]])\n",
      "        >>> stats.zscore(b, axis=1, ddof=1)\n",
      "        array([[-0.19264823, -1.28415119,  1.07259584,  0.40420358],\n",
      "               [ 0.33048416, -1.37380874,  0.04251374,  1.00081084],\n",
      "               [ 0.26796377, -1.12598418,  1.23283094, -0.37481053],\n",
      "               [-0.22095197,  0.24468594,  1.19042819, -1.21416216],\n",
      "               [-0.82780366,  1.4457416 , -0.43867764, -0.1792603 ]])\n",
      "        \n",
      "        An example with `nan_policy='omit'`:\n",
      "        \n",
      "        >>> x = np.array([[25.11, 30.10, np.nan, 32.02, 43.15],\n",
      "        ...               [14.95, 16.06, 121.25, 94.35, 29.81]])\n",
      "        >>> stats.zscore(x, axis=1, nan_policy='omit')\n",
      "        array([[-1.13490897, -0.37830299,         nan, -0.08718406,  1.60039602],\n",
      "               [-0.91611681, -0.89090508,  1.4983032 ,  0.88731639, -0.5785977 ]])\n",
      "\n",
      "DATA\n",
      "    __all__ = ['F_onewayBadInputSizesWarning', 'F_onewayConstantInputWarni...\n",
      "    alpha = <scipy.stats._continuous_distns.alpha_gen object>\n",
      "    anglit = <scipy.stats._continuous_distns.anglit_gen object>\n",
      "    arcsine = <scipy.stats._continuous_distns.arcsine_gen object>\n",
      "    argus = <scipy.stats._continuous_distns.argus_gen object>\n",
      "    bernoulli = <scipy.stats._discrete_distns.bernoulli_gen object>\n",
      "    beta = <scipy.stats._continuous_distns.beta_gen object>\n",
      "    betabinom = <scipy.stats._discrete_distns.betabinom_gen object>\n",
      "    betaprime = <scipy.stats._continuous_distns.betaprime_gen object>\n",
      "    binom = <scipy.stats._discrete_distns.binom_gen object>\n",
      "    boltzmann = <scipy.stats._discrete_distns.boltzmann_gen object>\n",
      "    bradford = <scipy.stats._continuous_distns.bradford_gen object>\n",
      "    burr = <scipy.stats._continuous_distns.burr_gen object>\n",
      "    burr12 = <scipy.stats._continuous_distns.burr12_gen object>\n",
      "    cauchy = <scipy.stats._continuous_distns.cauchy_gen object>\n",
      "    chi = <scipy.stats._continuous_distns.chi_gen object>\n",
      "    chi2 = <scipy.stats._continuous_distns.chi2_gen object>\n",
      "    cosine = <scipy.stats._continuous_distns.cosine_gen object>\n",
      "    crystalball = <scipy.stats._continuous_distns.crystalball_gen object>\n",
      "    dgamma = <scipy.stats._continuous_distns.dgamma_gen object>\n",
      "    dirichlet = <scipy.stats._multivariate.dirichlet_gen object>\n",
      "    dlaplace = <scipy.stats._discrete_distns.dlaplace_gen object>\n",
      "    dweibull = <scipy.stats._continuous_distns.dweibull_gen object>\n",
      "    erlang = <scipy.stats._continuous_distns.erlang_gen object>\n",
      "    expon = <scipy.stats._continuous_distns.expon_gen object>\n",
      "    exponnorm = <scipy.stats._continuous_distns.exponnorm_gen object>\n",
      "    exponpow = <scipy.stats._continuous_distns.exponpow_gen object>\n",
      "    exponweib = <scipy.stats._continuous_distns.exponweib_gen object>\n",
      "    f = <scipy.stats._continuous_distns.f_gen object>\n",
      "    fatiguelife = <scipy.stats._continuous_distns.fatiguelife_gen object>\n",
      "    fisk = <scipy.stats._continuous_distns.fisk_gen object>\n",
      "    foldcauchy = <scipy.stats._continuous_distns.foldcauchy_gen object>\n",
      "    foldnorm = <scipy.stats._continuous_distns.foldnorm_gen object>\n",
      "    gamma = <scipy.stats._continuous_distns.gamma_gen object>\n",
      "    gausshyper = <scipy.stats._continuous_distns.gausshyper_gen object>\n",
      "    genexpon = <scipy.stats._continuous_distns.genexpon_gen object>\n",
      "    genextreme = <scipy.stats._continuous_distns.genextreme_gen object>\n",
      "    gengamma = <scipy.stats._continuous_distns.gengamma_gen object>\n",
      "    genhalflogistic = <scipy.stats._continuous_distns.genhalflogistic_gen ...\n",
      "    genhyperbolic = <scipy.stats._continuous_distns.genhyperbolic_gen obje...\n",
      "    geninvgauss = <scipy.stats._continuous_distns.geninvgauss_gen object>\n",
      "    genlogistic = <scipy.stats._continuous_distns.genlogistic_gen object>\n",
      "    gennorm = <scipy.stats._continuous_distns.gennorm_gen object>\n",
      "    genpareto = <scipy.stats._continuous_distns.genpareto_gen object>\n",
      "    geom = <scipy.stats._discrete_distns.geom_gen object>\n",
      "    gilbrat = <scipy.stats._continuous_distns.gilbrat_gen object>\n",
      "    gompertz = <scipy.stats._continuous_distns.gompertz_gen object>\n",
      "    gumbel_l = <scipy.stats._continuous_distns.gumbel_l_gen object>\n",
      "    gumbel_r = <scipy.stats._continuous_distns.gumbel_r_gen object>\n",
      "    halfcauchy = <scipy.stats._continuous_distns.halfcauchy_gen object>\n",
      "    halfgennorm = <scipy.stats._continuous_distns.halfgennorm_gen object>\n",
      "    halflogistic = <scipy.stats._continuous_distns.halflogistic_gen object...\n",
      "    halfnorm = <scipy.stats._continuous_distns.halfnorm_gen object>\n",
      "    hypergeom = <scipy.stats._discrete_distns.hypergeom_gen object>\n",
      "    hypsecant = <scipy.stats._continuous_distns.hypsecant_gen object>\n",
      "    invgamma = <scipy.stats._continuous_distns.invgamma_gen object>\n",
      "    invgauss = <scipy.stats._continuous_distns.invgauss_gen object>\n",
      "    invweibull = <scipy.stats._continuous_distns.invweibull_gen object>\n",
      "    invwishart = <scipy.stats._multivariate.invwishart_gen object>\n",
      "    johnsonsb = <scipy.stats._continuous_distns.johnsonsb_gen object>\n",
      "    johnsonsu = <scipy.stats._continuous_distns.johnsonsu_gen object>\n",
      "    kappa3 = <scipy.stats._continuous_distns.kappa3_gen object>\n",
      "    kappa4 = <scipy.stats._continuous_distns.kappa4_gen object>\n",
      "    ksone = <scipy.stats._continuous_distns.ksone_gen object>\n",
      "    kstwo = <scipy.stats._continuous_distns.kstwo_gen object>\n",
      "    kstwobign = <scipy.stats._continuous_distns.kstwobign_gen object>\n",
      "    laplace = <scipy.stats._continuous_distns.laplace_gen object>\n",
      "    laplace_asymmetric = <scipy.stats._continuous_distns.laplace_asymmetri...\n",
      "    levy = <scipy.stats._continuous_distns.levy_gen object>\n",
      "    levy_l = <scipy.stats._continuous_distns.levy_l_gen object>\n",
      "    levy_stable = <scipy.stats._continuous_distns.levy_stable_gen object>\n",
      "    loggamma = <scipy.stats._continuous_distns.loggamma_gen object>\n",
      "    logistic = <scipy.stats._continuous_distns.logistic_gen object>\n",
      "    loglaplace = <scipy.stats._continuous_distns.loglaplace_gen object>\n",
      "    lognorm = <scipy.stats._continuous_distns.lognorm_gen object>\n",
      "    logser = <scipy.stats._discrete_distns.logser_gen object>\n",
      "    loguniform = <scipy.stats._continuous_distns.reciprocal_gen object>\n",
      "    lomax = <scipy.stats._continuous_distns.lomax_gen object>\n",
      "    matrix_normal = <scipy.stats._multivariate.matrix_normal_gen object>\n",
      "    maxwell = <scipy.stats._continuous_distns.maxwell_gen object>\n",
      "    mielke = <scipy.stats._continuous_distns.mielke_gen object>\n",
      "    moyal = <scipy.stats._continuous_distns.moyal_gen object>\n",
      "    multinomial = <scipy.stats._multivariate.multinomial_gen object>\n",
      "    multivariate_hypergeom = <scipy.stats._multivariate.multivariate_hyper...\n",
      "    multivariate_normal = <scipy.stats._multivariate.multivariate_normal_g...\n",
      "    multivariate_t = <scipy.stats._multivariate.multivariate_t_gen object>\n",
      "    nakagami = <scipy.stats._continuous_distns.nakagami_gen object>\n",
      "    nbinom = <scipy.stats._discrete_distns.nbinom_gen object>\n",
      "    ncf = <scipy.stats._continuous_distns.ncf_gen object>\n",
      "    nchypergeom_fisher = <scipy.stats._discrete_distns.nchypergeom_fisher_...\n",
      "    nchypergeom_wallenius = <scipy.stats._discrete_distns.nchypergeom_wall...\n",
      "    nct = <scipy.stats._continuous_distns.nct_gen object>\n",
      "    ncx2 = <scipy.stats._continuous_distns.ncx2_gen object>\n",
      "    nhypergeom = <scipy.stats._discrete_distns.nhypergeom_gen object>\n",
      "    norm = <scipy.stats._continuous_distns.norm_gen object>\n",
      "    norminvgauss = <scipy.stats._continuous_distns.norminvgauss_gen object...\n",
      "    ortho_group = <scipy.stats._multivariate.ortho_group_gen object>\n",
      "    pareto = <scipy.stats._continuous_distns.pareto_gen object>\n",
      "    pearson3 = <scipy.stats._continuous_distns.pearson3_gen object>\n",
      "    planck = <scipy.stats._discrete_distns.planck_gen object>\n",
      "    poisson = <scipy.stats._discrete_distns.poisson_gen object>\n",
      "    powerlaw = <scipy.stats._continuous_distns.powerlaw_gen object>\n",
      "    powerlognorm = <scipy.stats._continuous_distns.powerlognorm_gen object...\n",
      "    powernorm = <scipy.stats._continuous_distns.powernorm_gen object>\n",
      "    randint = <scipy.stats._discrete_distns.randint_gen object>\n",
      "    random_correlation = <scipy.stats._multivariate.random_correlation_gen...\n",
      "    rayleigh = <scipy.stats._continuous_distns.rayleigh_gen object>\n",
      "    rdist = <scipy.stats._continuous_distns.rdist_gen object>\n",
      "    recipinvgauss = <scipy.stats._continuous_distns.recipinvgauss_gen obje...\n",
      "    reciprocal = <scipy.stats._continuous_distns.reciprocal_gen object>\n",
      "    rice = <scipy.stats._continuous_distns.rice_gen object>\n",
      "    semicircular = <scipy.stats._continuous_distns.semicircular_gen object...\n",
      "    skellam = <scipy.stats._discrete_distns.skellam_gen object>\n",
      "    skewcauchy = <scipy.stats._continuous_distns.skewcauchy_gen object>\n",
      "    skewnorm = <scipy.stats._continuous_distns.skew_norm_gen object>\n",
      "    special_ortho_group = <scipy.stats._multivariate.special_ortho_group_g...\n",
      "    studentized_range = <scipy.stats._continuous_distns.studentized_range_...\n",
      "    t = <scipy.stats._continuous_distns.t_gen object>\n",
      "    trapezoid = <scipy.stats._continuous_distns.trapezoid_gen object>\n",
      "    trapz = <scipy.stats._continuous_distns.trapezoid_gen object>\n",
      "    triang = <scipy.stats._continuous_distns.triang_gen object>\n",
      "    truncexpon = <scipy.stats._continuous_distns.truncexpon_gen object>\n",
      "    truncnorm = <scipy.stats._continuous_distns.truncnorm_gen object>\n",
      "    tukeylambda = <scipy.stats._continuous_distns.tukeylambda_gen object>\n",
      "    uniform = <scipy.stats._continuous_distns.uniform_gen object>\n",
      "    unitary_group = <scipy.stats._multivariate.unitary_group_gen object>\n",
      "    vonmises = <scipy.stats._continuous_distns.vonmises_gen object>\n",
      "    vonmises_line = <scipy.stats._continuous_distns.vonmises_gen object>\n",
      "    wald = <scipy.stats._continuous_distns.wald_gen object>\n",
      "    weibull_max = <scipy.stats._continuous_distns.weibull_max_gen object>\n",
      "    weibull_min = <scipy.stats._continuous_distns.weibull_min_gen object>\n",
      "    wishart = <scipy.stats._multivariate.wishart_gen object>\n",
      "    wrapcauchy = <scipy.stats._continuous_distns.wrapcauchy_gen object>\n",
      "    yulesimon = <scipy.stats._discrete_distns.yulesimon_gen object>\n",
      "    zipf = <scipy.stats._discrete_distns.zipf_gen object>\n",
      "    zipfian = <scipy.stats._discrete_distns.zipfian_gen object>\n",
      "\n",
      "FILE\n",
      "    c:\\programdata\\anaconda3\\lib\\site-packages\\scipy\\stats\\__init__.py\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(stat)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "a8267153",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scipy.optimize as spop"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "a788cbf3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on package scipy.optimize in scipy:\n",
      "\n",
      "NAME\n",
      "    scipy.optimize\n",
      "\n",
      "DESCRIPTION\n",
      "    =====================================================\n",
      "    Optimization and root finding (:mod:`scipy.optimize`)\n",
      "    =====================================================\n",
      "    \n",
      "    .. currentmodule:: scipy.optimize\n",
      "    \n",
      "    SciPy ``optimize`` provides functions for minimizing (or maximizing)\n",
      "    objective functions, possibly subject to constraints. It includes\n",
      "    solvers for nonlinear problems (with support for both local and global\n",
      "    optimization algorithms), linear programing, constrained\n",
      "    and nonlinear least-squares, root finding, and curve fitting.\n",
      "    \n",
      "    Common functions and objects, shared across different solvers, are:\n",
      "    \n",
      "    .. autosummary::\n",
      "       :toctree: generated/\n",
      "    \n",
      "       show_options - Show specific options optimization solvers.\n",
      "       OptimizeResult - The optimization result returned by some optimizers.\n",
      "       OptimizeWarning - The optimization encountered problems.\n",
      "    \n",
      "    \n",
      "    Optimization\n",
      "    ============\n",
      "    \n",
      "    Scalar functions optimization\n",
      "    -----------------------------\n",
      "    \n",
      "    .. autosummary::\n",
      "       :toctree: generated/\n",
      "    \n",
      "       minimize_scalar - Interface for minimizers of univariate functions\n",
      "    \n",
      "    The `minimize_scalar` function supports the following methods:\n",
      "    \n",
      "    .. toctree::\n",
      "    \n",
      "       optimize.minimize_scalar-brent\n",
      "       optimize.minimize_scalar-bounded\n",
      "       optimize.minimize_scalar-golden\n",
      "    \n",
      "    Local (multivariate) optimization\n",
      "    ---------------------------------\n",
      "    \n",
      "    .. autosummary::\n",
      "       :toctree: generated/\n",
      "    \n",
      "       minimize - Interface for minimizers of multivariate functions.\n",
      "    \n",
      "    The `minimize` function supports the following methods:\n",
      "    \n",
      "    .. toctree::\n",
      "    \n",
      "       optimize.minimize-neldermead\n",
      "       optimize.minimize-powell\n",
      "       optimize.minimize-cg\n",
      "       optimize.minimize-bfgs\n",
      "       optimize.minimize-newtoncg\n",
      "       optimize.minimize-lbfgsb\n",
      "       optimize.minimize-tnc\n",
      "       optimize.minimize-cobyla\n",
      "       optimize.minimize-slsqp\n",
      "       optimize.minimize-trustconstr\n",
      "       optimize.minimize-dogleg\n",
      "       optimize.minimize-trustncg\n",
      "       optimize.minimize-trustkrylov\n",
      "       optimize.minimize-trustexact\n",
      "    \n",
      "    Constraints are passed to `minimize` function as a single object or\n",
      "    as a list of objects from the following classes:\n",
      "    \n",
      "    .. autosummary::\n",
      "       :toctree: generated/\n",
      "    \n",
      "       NonlinearConstraint - Class defining general nonlinear constraints.\n",
      "       LinearConstraint - Class defining general linear constraints.\n",
      "    \n",
      "    Simple bound constraints are handled separately and there is a special class\n",
      "    for them:\n",
      "    \n",
      "    .. autosummary::\n",
      "       :toctree: generated/\n",
      "    \n",
      "       Bounds - Bound constraints.\n",
      "    \n",
      "    Quasi-Newton strategies implementing `HessianUpdateStrategy`\n",
      "    interface can be used to approximate the Hessian in `minimize`\n",
      "    function (available only for the 'trust-constr' method). Available\n",
      "    quasi-Newton methods implementing this interface are:\n",
      "    \n",
      "    .. autosummary::\n",
      "       :toctree: generated/\n",
      "    \n",
      "       BFGS - Broyden-Fletcher-Goldfarb-Shanno (BFGS) Hessian update strategy.\n",
      "       SR1 - Symmetric-rank-1 Hessian update strategy.\n",
      "    \n",
      "    Global optimization\n",
      "    -------------------\n",
      "    \n",
      "    .. autosummary::\n",
      "       :toctree: generated/\n",
      "    \n",
      "       basinhopping - Basinhopping stochastic optimizer.\n",
      "       brute - Brute force searching optimizer.\n",
      "       differential_evolution - stochastic minimization using differential evolution.\n",
      "    \n",
      "       shgo - simplicial homology global optimisation\n",
      "       dual_annealing - Dual annealing stochastic optimizer.\n",
      "    \n",
      "    \n",
      "    Least-squares and curve fitting\n",
      "    ===============================\n",
      "    \n",
      "    Nonlinear least-squares\n",
      "    -----------------------\n",
      "    \n",
      "    .. autosummary::\n",
      "       :toctree: generated/\n",
      "    \n",
      "       least_squares - Solve a nonlinear least-squares problem with bounds on the variables.\n",
      "    \n",
      "    Linear least-squares\n",
      "    --------------------\n",
      "    \n",
      "    .. autosummary::\n",
      "       :toctree: generated/\n",
      "    \n",
      "       nnls - Linear least-squares problem with non-negativity constraint.\n",
      "       lsq_linear - Linear least-squares problem with bound constraints.\n",
      "    \n",
      "    Curve fitting\n",
      "    -------------\n",
      "    \n",
      "    .. autosummary::\n",
      "       :toctree: generated/\n",
      "    \n",
      "       curve_fit -- Fit curve to a set of points.\n",
      "    \n",
      "    Root finding\n",
      "    ============\n",
      "    \n",
      "    Scalar functions\n",
      "    ----------------\n",
      "    .. autosummary::\n",
      "       :toctree: generated/\n",
      "    \n",
      "       root_scalar - Unified interface for nonlinear solvers of scalar functions.\n",
      "       brentq - quadratic interpolation Brent method.\n",
      "       brenth - Brent method, modified by Harris with hyperbolic extrapolation.\n",
      "       ridder - Ridder's method.\n",
      "       bisect - Bisection method.\n",
      "       newton - Newton's method (also Secant and Halley's methods).\n",
      "       toms748 - Alefeld, Potra & Shi Algorithm 748.\n",
      "       RootResults - The root finding result returned by some root finders.\n",
      "    \n",
      "    The `root_scalar` function supports the following methods:\n",
      "    \n",
      "    .. toctree::\n",
      "    \n",
      "       optimize.root_scalar-brentq\n",
      "       optimize.root_scalar-brenth\n",
      "       optimize.root_scalar-bisect\n",
      "       optimize.root_scalar-ridder\n",
      "       optimize.root_scalar-newton\n",
      "       optimize.root_scalar-toms748\n",
      "       optimize.root_scalar-secant\n",
      "       optimize.root_scalar-halley\n",
      "    \n",
      "    \n",
      "    \n",
      "    The table below lists situations and appropriate methods, along with\n",
      "    *asymptotic* convergence rates per iteration (and per function evaluation)\n",
      "    for successful convergence to a simple root(*).\n",
      "    Bisection is the slowest of them all, adding one bit of accuracy for each\n",
      "    function evaluation, but is guaranteed to converge.\n",
      "    The other bracketing methods all (eventually) increase the number of accurate\n",
      "    bits by about 50% for every function evaluation.\n",
      "    The derivative-based methods, all built on `newton`, can converge quite quickly\n",
      "    if the initial value is close to the root.  They can also be applied to\n",
      "    functions defined on (a subset of) the complex plane.\n",
      "    \n",
      "    +-------------+----------+----------+-----------+-------------+-------------+----------------+\n",
      "    | Domain of f | Bracket? |    Derivatives?      | Solvers     |        Convergence           |\n",
      "    +             +          +----------+-----------+             +-------------+----------------+\n",
      "    |             |          | `fprime` | `fprime2` |             | Guaranteed? |  Rate(s)(*)    |\n",
      "    +=============+==========+==========+===========+=============+=============+================+\n",
      "    | `R`         | Yes      | N/A      | N/A       | - bisection | - Yes       | - 1 \"Linear\"   |\n",
      "    |             |          |          |           | - brentq    | - Yes       | - >=1, <= 1.62 |\n",
      "    |             |          |          |           | - brenth    | - Yes       | - >=1, <= 1.62 |\n",
      "    |             |          |          |           | - ridder    | - Yes       | - 2.0 (1.41)   |\n",
      "    |             |          |          |           | - toms748   | - Yes       | - 2.7 (1.65)   |\n",
      "    +-------------+----------+----------+-----------+-------------+-------------+----------------+\n",
      "    | `R` or `C`  | No       | No       | No        | secant      | No          | 1.62 (1.62)    |\n",
      "    +-------------+----------+----------+-----------+-------------+-------------+----------------+\n",
      "    | `R` or `C`  | No       | Yes      | No        | newton      | No          | 2.00 (1.41)    |\n",
      "    +-------------+----------+----------+-----------+-------------+-------------+----------------+\n",
      "    | `R` or `C`  | No       | Yes      | Yes       | halley      | No          | 3.00 (1.44)    |\n",
      "    +-------------+----------+----------+-----------+-------------+-------------+----------------+\n",
      "    \n",
      "    .. seealso::\n",
      "    \n",
      "       `scipy.optimize.cython_optimize` -- Typed Cython versions of zeros functions\n",
      "    \n",
      "    Fixed point finding:\n",
      "    \n",
      "    .. autosummary::\n",
      "       :toctree: generated/\n",
      "    \n",
      "       fixed_point - Single-variable fixed-point solver.\n",
      "    \n",
      "    Multidimensional\n",
      "    ----------------\n",
      "    \n",
      "    .. autosummary::\n",
      "       :toctree: generated/\n",
      "    \n",
      "       root - Unified interface for nonlinear solvers of multivariate functions.\n",
      "    \n",
      "    The `root` function supports the following methods:\n",
      "    \n",
      "    .. toctree::\n",
      "    \n",
      "       optimize.root-hybr\n",
      "       optimize.root-lm\n",
      "       optimize.root-broyden1\n",
      "       optimize.root-broyden2\n",
      "       optimize.root-anderson\n",
      "       optimize.root-linearmixing\n",
      "       optimize.root-diagbroyden\n",
      "       optimize.root-excitingmixing\n",
      "       optimize.root-krylov\n",
      "       optimize.root-dfsane\n",
      "    \n",
      "    Linear programming\n",
      "    ==================\n",
      "    \n",
      "    .. autosummary::\n",
      "       :toctree: generated/\n",
      "    \n",
      "       linprog -- Unified interface for minimizers of linear programming problems.\n",
      "    \n",
      "    The `linprog` function supports the following methods:\n",
      "    \n",
      "    .. toctree::\n",
      "    \n",
      "       optimize.linprog-simplex\n",
      "       optimize.linprog-interior-point\n",
      "       optimize.linprog-revised_simplex\n",
      "       optimize.linprog-highs-ipm\n",
      "       optimize.linprog-highs-ds\n",
      "       optimize.linprog-highs\n",
      "    \n",
      "    The simplex, interior-point, and revised simplex methods support callback\n",
      "    functions, such as:\n",
      "    \n",
      "    .. autosummary::\n",
      "       :toctree: generated/\n",
      "    \n",
      "       linprog_verbose_callback -- Sample callback function for linprog (simplex).\n",
      "    \n",
      "    Assignment problems\n",
      "    ===================\n",
      "    \n",
      "    .. autosummary::\n",
      "       :toctree: generated/\n",
      "    \n",
      "       linear_sum_assignment -- Solves the linear-sum assignment problem.\n",
      "       quadratic_assignment -- Solves the quadratic assignment problem.\n",
      "    \n",
      "    The `quadratic_assignment` function supports the following methods:\n",
      "    \n",
      "    .. toctree::\n",
      "    \n",
      "       optimize.qap-faq\n",
      "       optimize.qap-2opt\n",
      "    \n",
      "    Utilities\n",
      "    =========\n",
      "    \n",
      "    Finite-difference approximation\n",
      "    -------------------------------\n",
      "    \n",
      "    .. autosummary::\n",
      "       :toctree: generated/\n",
      "    \n",
      "       approx_fprime - Approximate the gradient of a scalar function.\n",
      "       check_grad - Check the supplied derivative using finite differences.\n",
      "    \n",
      "    \n",
      "    Line search\n",
      "    -----------\n",
      "    \n",
      "    .. autosummary::\n",
      "       :toctree: generated/\n",
      "    \n",
      "       bracket - Bracket a minimum, given two starting points.\n",
      "       line_search - Return a step that satisfies the strong Wolfe conditions.\n",
      "    \n",
      "    Hessian approximation\n",
      "    ---------------------\n",
      "    \n",
      "    .. autosummary::\n",
      "       :toctree: generated/\n",
      "    \n",
      "       LbfgsInvHessProduct - Linear operator for L-BFGS approximate inverse Hessian.\n",
      "       HessianUpdateStrategy - Interface for implementing Hessian update strategies\n",
      "    \n",
      "    Benchmark problems\n",
      "    ------------------\n",
      "    \n",
      "    .. autosummary::\n",
      "       :toctree: generated/\n",
      "    \n",
      "       rosen - The Rosenbrock function.\n",
      "       rosen_der - The derivative of the Rosenbrock function.\n",
      "       rosen_hess - The Hessian matrix of the Rosenbrock function.\n",
      "       rosen_hess_prod - Product of the Rosenbrock Hessian with a vector.\n",
      "    \n",
      "    Legacy functions\n",
      "    ================\n",
      "    \n",
      "    The functions below are not recommended for use in new scripts;\n",
      "    all of these methods are accessible via a newer, more consistent\n",
      "    interfaces, provided by the interfaces above.\n",
      "    \n",
      "    Optimization\n",
      "    ------------\n",
      "    \n",
      "    General-purpose multivariate methods:\n",
      "    \n",
      "    .. autosummary::\n",
      "       :toctree: generated/\n",
      "    \n",
      "       fmin - Nelder-Mead Simplex algorithm.\n",
      "       fmin_powell - Powell's (modified) level set method.\n",
      "       fmin_cg - Non-linear (Polak-Ribiere) conjugate gradient algorithm.\n",
      "       fmin_bfgs - Quasi-Newton method (Broydon-Fletcher-Goldfarb-Shanno).\n",
      "       fmin_ncg - Line-search Newton Conjugate Gradient.\n",
      "    \n",
      "    Constrained multivariate methods:\n",
      "    \n",
      "    .. autosummary::\n",
      "       :toctree: generated/\n",
      "    \n",
      "       fmin_l_bfgs_b - Zhu, Byrd, and Nocedal's constrained optimizer.\n",
      "       fmin_tnc - Truncated Newton code.\n",
      "       fmin_cobyla - Constrained optimization by linear approximation.\n",
      "       fmin_slsqp - Minimization using sequential least-squares programming.\n",
      "    \n",
      "    Univariate (scalar) minimization methods:\n",
      "    \n",
      "    .. autosummary::\n",
      "       :toctree: generated/\n",
      "    \n",
      "       fminbound - Bounded minimization of a scalar function.\n",
      "       brent - 1-D function minimization using Brent method.\n",
      "       golden - 1-D function minimization using Golden Section method.\n",
      "    \n",
      "    Least-squares\n",
      "    -------------\n",
      "    \n",
      "    .. autosummary::\n",
      "       :toctree: generated/\n",
      "    \n",
      "       leastsq - Minimize the sum of squares of M equations in N unknowns.\n",
      "    \n",
      "    Root finding\n",
      "    ------------\n",
      "    \n",
      "    General nonlinear solvers:\n",
      "    \n",
      "    .. autosummary::\n",
      "       :toctree: generated/\n",
      "    \n",
      "       fsolve - Non-linear multivariable equation solver.\n",
      "       broyden1 - Broyden's first method.\n",
      "       broyden2 - Broyden's second method.\n",
      "    \n",
      "    Large-scale nonlinear solvers:\n",
      "    \n",
      "    .. autosummary::\n",
      "       :toctree: generated/\n",
      "    \n",
      "       newton_krylov\n",
      "       anderson\n",
      "    \n",
      "    Simple iteration solvers:\n",
      "    \n",
      "    .. autosummary::\n",
      "       :toctree: generated/\n",
      "    \n",
      "       excitingmixing\n",
      "       linearmixing\n",
      "       diagbroyden\n",
      "    \n",
      "    :mod:`Additional information on the nonlinear solvers <scipy.optimize.nonlin>`\n",
      "\n",
      "PACKAGE CONTENTS\n",
      "    __nnls\n",
      "    _basinhopping\n",
      "    _bglu_dense\n",
      "    _cobyla\n",
      "    _constraints\n",
      "    _differentiable_functions\n",
      "    _differentialevolution\n",
      "    _dual_annealing\n",
      "    _group_columns\n",
      "    _hessian_update_strategy\n",
      "    _highs (package)\n",
      "    _lbfgsb\n",
      "    _linprog\n",
      "    _linprog_doc\n",
      "    _linprog_highs\n",
      "    _linprog_ip\n",
      "    _linprog_rs\n",
      "    _linprog_simplex\n",
      "    _linprog_util\n",
      "    _lsap\n",
      "    _lsap_module\n",
      "    _lsq (package)\n",
      "    _minimize\n",
      "    _minpack\n",
      "    _nnls\n",
      "    _numdiff\n",
      "    _qap\n",
      "    _remove_redundancy\n",
      "    _root\n",
      "    _root_scalar\n",
      "    _shgo\n",
      "    _shgo_lib (package)\n",
      "    _slsqp\n",
      "    _spectral\n",
      "    _trlib (package)\n",
      "    _trustregion\n",
      "    _trustregion_constr (package)\n",
      "    _trustregion_dogleg\n",
      "    _trustregion_exact\n",
      "    _trustregion_krylov\n",
      "    _trustregion_ncg\n",
      "    _tstutils\n",
      "    _zeros\n",
      "    cobyla\n",
      "    cython_optimize (package)\n",
      "    lbfgsb\n",
      "    linesearch\n",
      "    minpack\n",
      "    minpack2\n",
      "    moduleTNC\n",
      "    nonlin\n",
      "    optimize\n",
      "    setup\n",
      "    slsqp\n",
      "    tests (package)\n",
      "    tnc\n",
      "    zeros\n",
      "\n",
      "CLASSES\n",
      "    builtins.UserWarning(builtins.Warning)\n",
      "        scipy.optimize.optimize.OptimizeWarning\n",
      "    builtins.dict(builtins.object)\n",
      "        scipy.optimize.optimize.OptimizeResult\n",
      "    builtins.object\n",
      "        scipy.optimize._constraints.Bounds\n",
      "        scipy.optimize._constraints.LinearConstraint\n",
      "        scipy.optimize._constraints.NonlinearConstraint\n",
      "        scipy.optimize._hessian_update_strategy.HessianUpdateStrategy\n",
      "        scipy.optimize.zeros.RootResults\n",
      "    scipy.optimize._hessian_update_strategy.FullHessianUpdateStrategy(scipy.optimize._hessian_update_strategy.HessianUpdateStrategy)\n",
      "        scipy.optimize._hessian_update_strategy.BFGS\n",
      "        scipy.optimize._hessian_update_strategy.SR1\n",
      "    scipy.sparse.linalg.interface.LinearOperator(builtins.object)\n",
      "        scipy.optimize.lbfgsb.LbfgsInvHessProduct\n",
      "    \n",
      "    class BFGS(FullHessianUpdateStrategy)\n",
      "     |  BFGS(exception_strategy='skip_update', min_curvature=None, init_scale='auto')\n",
      "     |  \n",
      "     |  Broyden-Fletcher-Goldfarb-Shanno (BFGS) Hessian update strategy.\n",
      "     |  \n",
      "     |  Parameters\n",
      "     |  ----------\n",
      "     |  exception_strategy : {'skip_update', 'damp_update'}, optional\n",
      "     |      Define how to proceed when the curvature condition is violated.\n",
      "     |      Set it to 'skip_update' to just skip the update. Or, alternatively,\n",
      "     |      set it to 'damp_update' to interpolate between the actual BFGS\n",
      "     |      result and the unmodified matrix. Both exceptions strategies\n",
      "     |      are explained  in [1]_, p.536-537.\n",
      "     |  min_curvature : float\n",
      "     |      This number, scaled by a normalization factor, defines the\n",
      "     |      minimum curvature ``dot(delta_grad, delta_x)`` allowed to go\n",
      "     |      unaffected by the exception strategy. By default is equal to\n",
      "     |      1e-8 when ``exception_strategy = 'skip_update'`` and equal\n",
      "     |      to 0.2 when ``exception_strategy = 'damp_update'``.\n",
      "     |  init_scale : {float, 'auto'}\n",
      "     |      Matrix scale at first iteration. At the first\n",
      "     |      iteration the Hessian matrix or its inverse will be initialized\n",
      "     |      with ``init_scale*np.eye(n)``, where ``n`` is the problem dimension.\n",
      "     |      Set it to 'auto' in order to use an automatic heuristic for choosing\n",
      "     |      the initial scale. The heuristic is described in [1]_, p.143.\n",
      "     |      By default uses 'auto'.\n",
      "     |  \n",
      "     |  Notes\n",
      "     |  -----\n",
      "     |  The update is based on the description in [1]_, p.140.\n",
      "     |  \n",
      "     |  References\n",
      "     |  ----------\n",
      "     |  .. [1] Nocedal, Jorge, and Stephen J. Wright. \"Numerical optimization\"\n",
      "     |         Second Edition (2006).\n",
      "     |  \n",
      "     |  Method resolution order:\n",
      "     |      BFGS\n",
      "     |      FullHessianUpdateStrategy\n",
      "     |      HessianUpdateStrategy\n",
      "     |      builtins.object\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __init__(self, exception_strategy='skip_update', min_curvature=None, init_scale='auto')\n",
      "     |      Initialize self.  See help(type(self)) for accurate signature.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from FullHessianUpdateStrategy:\n",
      "     |  \n",
      "     |  dot(self, p)\n",
      "     |      Compute the product of the internal matrix with the given vector.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      p : array_like\n",
      "     |          1-D array representing a vector.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      Hp : array\n",
      "     |          1-D represents the result of multiplying the approximation matrix\n",
      "     |          by vector p.\n",
      "     |  \n",
      "     |  get_matrix(self)\n",
      "     |      Return the current internal matrix.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      M : ndarray, shape (n, n)\n",
      "     |          Dense matrix containing either the Hessian or its inverse\n",
      "     |          (depending on how `approx_type` was defined).\n",
      "     |  \n",
      "     |  initialize(self, n, approx_type)\n",
      "     |      Initialize internal matrix.\n",
      "     |      \n",
      "     |      Allocate internal memory for storing and updating\n",
      "     |      the Hessian or its inverse.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      n : int\n",
      "     |          Problem dimension.\n",
      "     |      approx_type : {'hess', 'inv_hess'}\n",
      "     |          Selects either the Hessian or the inverse Hessian.\n",
      "     |          When set to 'hess' the Hessian will be stored and updated.\n",
      "     |          When set to 'inv_hess' its inverse will be used instead.\n",
      "     |  \n",
      "     |  update(self, delta_x, delta_grad)\n",
      "     |      Update internal matrix.\n",
      "     |      \n",
      "     |      Update Hessian matrix or its inverse (depending on how 'approx_type'\n",
      "     |      is defined) using information about the last evaluated points.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      delta_x : ndarray\n",
      "     |          The difference between two points the gradient\n",
      "     |          function have been evaluated at: ``delta_x = x2 - x1``.\n",
      "     |      delta_grad : ndarray\n",
      "     |          The difference between the gradients:\n",
      "     |          ``delta_grad = grad(x2) - grad(x1)``.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from HessianUpdateStrategy:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |      dictionary for instance variables (if defined)\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "    \n",
      "    class Bounds(builtins.object)\n",
      "     |  Bounds(lb, ub, keep_feasible=False)\n",
      "     |  \n",
      "     |  Bounds constraint on the variables.\n",
      "     |  \n",
      "     |  The constraint has the general inequality form::\n",
      "     |  \n",
      "     |      lb <= x <= ub\n",
      "     |  \n",
      "     |  It is possible to use equal bounds to represent an equality constraint or\n",
      "     |  infinite bounds to represent a one-sided constraint.\n",
      "     |  \n",
      "     |  Parameters\n",
      "     |  ----------\n",
      "     |  lb, ub : array_like\n",
      "     |      Lower and upper bounds on independent variables. Each array must\n",
      "     |      have the same size as x or be a scalar, in which case a bound will be\n",
      "     |      the same for all the variables. Set components of `lb` and `ub` equal\n",
      "     |      to fix a variable. Use ``np.inf`` with an appropriate sign to disable\n",
      "     |      bounds on all or some variables. Note that you can mix constraints of\n",
      "     |      different types: interval, one-sided or equality, by setting different\n",
      "     |      components of `lb` and `ub` as necessary.\n",
      "     |  keep_feasible : array_like of bool, optional\n",
      "     |      Whether to keep the constraint components feasible throughout\n",
      "     |      iterations. A single value set this property for all components.\n",
      "     |      Default is False. Has no effect for equality constraints.\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __init__(self, lb, ub, keep_feasible=False)\n",
      "     |      Initialize self.  See help(type(self)) for accurate signature.\n",
      "     |  \n",
      "     |  __repr__(self)\n",
      "     |      Return repr(self).\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors defined here:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |      dictionary for instance variables (if defined)\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "    \n",
      "    class HessianUpdateStrategy(builtins.object)\n",
      "     |  Interface for implementing Hessian update strategies.\n",
      "     |  \n",
      "     |  Many optimization methods make use of Hessian (or inverse Hessian)\n",
      "     |  approximations, such as the quasi-Newton methods BFGS, SR1, L-BFGS.\n",
      "     |  Some of these  approximations, however, do not actually need to store\n",
      "     |  the entire matrix or can compute the internal matrix product with a\n",
      "     |  given vector in a very efficiently manner. This class serves as an\n",
      "     |  abstract interface between the optimization algorithm and the\n",
      "     |  quasi-Newton update strategies, giving freedom of implementation\n",
      "     |  to store and update the internal matrix as efficiently as possible.\n",
      "     |  Different choices of initialization and update procedure will result\n",
      "     |  in different quasi-Newton strategies.\n",
      "     |  \n",
      "     |  Four methods should be implemented in derived classes: ``initialize``,\n",
      "     |  ``update``, ``dot`` and ``get_matrix``.\n",
      "     |  \n",
      "     |  Notes\n",
      "     |  -----\n",
      "     |  Any instance of a class that implements this interface,\n",
      "     |  can be accepted by the method ``minimize`` and used by\n",
      "     |  the compatible solvers to approximate the Hessian (or\n",
      "     |  inverse Hessian) used by the optimization algorithms.\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  dot(self, p)\n",
      "     |      Compute the product of the internal matrix with the given vector.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      p : array_like\n",
      "     |          1-D array representing a vector.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      Hp : array\n",
      "     |          1-D represents the result of multiplying the approximation matrix\n",
      "     |          by vector p.\n",
      "     |  \n",
      "     |  get_matrix(self)\n",
      "     |      Return current internal matrix.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      H : ndarray, shape (n, n)\n",
      "     |          Dense matrix containing either the Hessian\n",
      "     |          or its inverse (depending on how 'approx_type'\n",
      "     |          is defined).\n",
      "     |  \n",
      "     |  initialize(self, n, approx_type)\n",
      "     |      Initialize internal matrix.\n",
      "     |      \n",
      "     |      Allocate internal memory for storing and updating\n",
      "     |      the Hessian or its inverse.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      n : int\n",
      "     |          Problem dimension.\n",
      "     |      approx_type : {'hess', 'inv_hess'}\n",
      "     |          Selects either the Hessian or the inverse Hessian.\n",
      "     |          When set to 'hess' the Hessian will be stored and updated.\n",
      "     |          When set to 'inv_hess' its inverse will be used instead.\n",
      "     |  \n",
      "     |  update(self, delta_x, delta_grad)\n",
      "     |      Update internal matrix.\n",
      "     |      \n",
      "     |      Update Hessian matrix or its inverse (depending on how 'approx_type'\n",
      "     |      is defined) using information about the last evaluated points.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      delta_x : ndarray\n",
      "     |          The difference between two points the gradient\n",
      "     |          function have been evaluated at: ``delta_x = x2 - x1``.\n",
      "     |      delta_grad : ndarray\n",
      "     |          The difference between the gradients:\n",
      "     |          ``delta_grad = grad(x2) - grad(x1)``.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors defined here:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |      dictionary for instance variables (if defined)\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "    \n",
      "    class LbfgsInvHessProduct(scipy.sparse.linalg.interface.LinearOperator)\n",
      "     |  LbfgsInvHessProduct(*args, **kwargs)\n",
      "     |  \n",
      "     |  Linear operator for the L-BFGS approximate inverse Hessian.\n",
      "     |  \n",
      "     |  This operator computes the product of a vector with the approximate inverse\n",
      "     |  of the Hessian of the objective function, using the L-BFGS limited\n",
      "     |  memory approximation to the inverse Hessian, accumulated during the\n",
      "     |  optimization.\n",
      "     |  \n",
      "     |  Objects of this class implement the ``scipy.sparse.linalg.LinearOperator``\n",
      "     |  interface.\n",
      "     |  \n",
      "     |  Parameters\n",
      "     |  ----------\n",
      "     |  sk : array_like, shape=(n_corr, n)\n",
      "     |      Array of `n_corr` most recent updates to the solution vector.\n",
      "     |      (See [1]).\n",
      "     |  yk : array_like, shape=(n_corr, n)\n",
      "     |      Array of `n_corr` most recent updates to the gradient. (See [1]).\n",
      "     |  \n",
      "     |  References\n",
      "     |  ----------\n",
      "     |  .. [1] Nocedal, Jorge. \"Updating quasi-Newton matrices with limited\n",
      "     |     storage.\" Mathematics of computation 35.151 (1980): 773-782.\n",
      "     |  \n",
      "     |  Method resolution order:\n",
      "     |      LbfgsInvHessProduct\n",
      "     |      scipy.sparse.linalg.interface.LinearOperator\n",
      "     |      builtins.object\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __init__(self, sk, yk)\n",
      "     |      Construct the operator.\n",
      "     |  \n",
      "     |  todense(self)\n",
      "     |      Return a dense array representation of this operator.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      arr : ndarray, shape=(n, n)\n",
      "     |          An array with the same shape and containing\n",
      "     |          the same data represented by this `LinearOperator`.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from scipy.sparse.linalg.interface.LinearOperator:\n",
      "     |  \n",
      "     |  __add__(self, x)\n",
      "     |  \n",
      "     |  __call__(self, x)\n",
      "     |      Call self as a function.\n",
      "     |  \n",
      "     |  __matmul__(self, other)\n",
      "     |  \n",
      "     |  __mul__(self, x)\n",
      "     |  \n",
      "     |  __neg__(self)\n",
      "     |  \n",
      "     |  __pow__(self, p)\n",
      "     |  \n",
      "     |  __repr__(self)\n",
      "     |      Return repr(self).\n",
      "     |  \n",
      "     |  __rmatmul__(self, other)\n",
      "     |  \n",
      "     |  __rmul__(self, x)\n",
      "     |  \n",
      "     |  __sub__(self, x)\n",
      "     |  \n",
      "     |  adjoint(self)\n",
      "     |      Hermitian adjoint.\n",
      "     |      \n",
      "     |      Returns the Hermitian adjoint of self, aka the Hermitian\n",
      "     |      conjugate or Hermitian transpose. For a complex matrix, the\n",
      "     |      Hermitian adjoint is equal to the conjugate transpose.\n",
      "     |      \n",
      "     |      Can be abbreviated self.H instead of self.adjoint().\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      A_H : LinearOperator\n",
      "     |          Hermitian adjoint of self.\n",
      "     |  \n",
      "     |  dot(self, x)\n",
      "     |      Matrix-matrix or matrix-vector multiplication.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      x : array_like\n",
      "     |          1-d or 2-d array, representing a vector or matrix.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      Ax : array\n",
      "     |          1-d or 2-d array (depending on the shape of x) that represents\n",
      "     |          the result of applying this linear operator on x.\n",
      "     |  \n",
      "     |  matmat(self, X)\n",
      "     |      Matrix-matrix multiplication.\n",
      "     |      \n",
      "     |      Performs the operation y=A*X where A is an MxN linear\n",
      "     |      operator and X dense N*K matrix or ndarray.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      X : {matrix, ndarray}\n",
      "     |          An array with shape (N,K).\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      Y : {matrix, ndarray}\n",
      "     |          A matrix or ndarray with shape (M,K) depending on\n",
      "     |          the type of the X argument.\n",
      "     |      \n",
      "     |      Notes\n",
      "     |      -----\n",
      "     |      This matmat wraps any user-specified matmat routine or overridden\n",
      "     |      _matmat method to ensure that y has the correct type.\n",
      "     |  \n",
      "     |  matvec(self, x)\n",
      "     |      Matrix-vector multiplication.\n",
      "     |      \n",
      "     |      Performs the operation y=A*x where A is an MxN linear\n",
      "     |      operator and x is a column vector or 1-d array.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      x : {matrix, ndarray}\n",
      "     |          An array with shape (N,) or (N,1).\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      y : {matrix, ndarray}\n",
      "     |          A matrix or ndarray with shape (M,) or (M,1) depending\n",
      "     |          on the type and shape of the x argument.\n",
      "     |      \n",
      "     |      Notes\n",
      "     |      -----\n",
      "     |      This matvec wraps the user-specified matvec routine or overridden\n",
      "     |      _matvec method to ensure that y has the correct shape and type.\n",
      "     |  \n",
      "     |  rmatmat(self, X)\n",
      "     |      Adjoint matrix-matrix multiplication.\n",
      "     |      \n",
      "     |      Performs the operation y = A^H * x where A is an MxN linear\n",
      "     |      operator and x is a column vector or 1-d array, or 2-d array.\n",
      "     |      The default implementation defers to the adjoint.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      X : {matrix, ndarray}\n",
      "     |          A matrix or 2D array.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      Y : {matrix, ndarray}\n",
      "     |          A matrix or 2D array depending on the type of the input.\n",
      "     |      \n",
      "     |      Notes\n",
      "     |      -----\n",
      "     |      This rmatmat wraps the user-specified rmatmat routine.\n",
      "     |  \n",
      "     |  rmatvec(self, x)\n",
      "     |      Adjoint matrix-vector multiplication.\n",
      "     |      \n",
      "     |      Performs the operation y = A^H * x where A is an MxN linear\n",
      "     |      operator and x is a column vector or 1-d array.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      x : {matrix, ndarray}\n",
      "     |          An array with shape (M,) or (M,1).\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      y : {matrix, ndarray}\n",
      "     |          A matrix or ndarray with shape (N,) or (N,1) depending\n",
      "     |          on the type and shape of the x argument.\n",
      "     |      \n",
      "     |      Notes\n",
      "     |      -----\n",
      "     |      This rmatvec wraps the user-specified rmatvec routine or overridden\n",
      "     |      _rmatvec method to ensure that y has the correct shape and type.\n",
      "     |  \n",
      "     |  transpose(self)\n",
      "     |      Transpose this linear operator.\n",
      "     |      \n",
      "     |      Returns a LinearOperator that represents the transpose of this one.\n",
      "     |      Can be abbreviated self.T instead of self.transpose().\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Static methods inherited from scipy.sparse.linalg.interface.LinearOperator:\n",
      "     |  \n",
      "     |  __new__(cls, *args, **kwargs)\n",
      "     |      Create and return a new object.  See help(type) for accurate signature.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Readonly properties inherited from scipy.sparse.linalg.interface.LinearOperator:\n",
      "     |  \n",
      "     |  H\n",
      "     |      Hermitian adjoint.\n",
      "     |      \n",
      "     |      Returns the Hermitian adjoint of self, aka the Hermitian\n",
      "     |      conjugate or Hermitian transpose. For a complex matrix, the\n",
      "     |      Hermitian adjoint is equal to the conjugate transpose.\n",
      "     |      \n",
      "     |      Can be abbreviated self.H instead of self.adjoint().\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      A_H : LinearOperator\n",
      "     |          Hermitian adjoint of self.\n",
      "     |  \n",
      "     |  T\n",
      "     |      Transpose this linear operator.\n",
      "     |      \n",
      "     |      Returns a LinearOperator that represents the transpose of this one.\n",
      "     |      Can be abbreviated self.T instead of self.transpose().\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from scipy.sparse.linalg.interface.LinearOperator:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |      dictionary for instance variables (if defined)\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data and other attributes inherited from scipy.sparse.linalg.interface.LinearOperator:\n",
      "     |  \n",
      "     |  ndim = 2\n",
      "    \n",
      "    class LinearConstraint(builtins.object)\n",
      "     |  LinearConstraint(A, lb, ub, keep_feasible=False)\n",
      "     |  \n",
      "     |  Linear constraint on the variables.\n",
      "     |  \n",
      "     |  The constraint has the general inequality form::\n",
      "     |  \n",
      "     |      lb <= A.dot(x) <= ub\n",
      "     |  \n",
      "     |  Here the vector of independent variables x is passed as ndarray of shape\n",
      "     |  (n,) and the matrix A has shape (m, n).\n",
      "     |  \n",
      "     |  It is possible to use equal bounds to represent an equality constraint or\n",
      "     |  infinite bounds to represent a one-sided constraint.\n",
      "     |  \n",
      "     |  Parameters\n",
      "     |  ----------\n",
      "     |  A : {array_like, sparse matrix}, shape (m, n)\n",
      "     |      Matrix defining the constraint.\n",
      "     |  lb, ub : array_like\n",
      "     |      Lower and upper bounds on the constraint. Each array must have the\n",
      "     |      shape (m,) or be a scalar, in the latter case a bound will be the same\n",
      "     |      for all components of the constraint. Use ``np.inf`` with an\n",
      "     |      appropriate sign to specify a one-sided constraint.\n",
      "     |      Set components of `lb` and `ub` equal to represent an equality\n",
      "     |      constraint. Note that you can mix constraints of different types:\n",
      "     |      interval, one-sided or equality, by setting different components of\n",
      "     |      `lb` and `ub` as  necessary.\n",
      "     |  keep_feasible : array_like of bool, optional\n",
      "     |      Whether to keep the constraint components feasible throughout\n",
      "     |      iterations. A single value set this property for all components.\n",
      "     |      Default is False. Has no effect for equality constraints.\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __init__(self, A, lb, ub, keep_feasible=False)\n",
      "     |      Initialize self.  See help(type(self)) for accurate signature.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors defined here:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |      dictionary for instance variables (if defined)\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "    \n",
      "    class NonlinearConstraint(builtins.object)\n",
      "     |  NonlinearConstraint(fun, lb, ub, jac='2-point', hess=<scipy.optimize._hessian_update_strategy.BFGS object at 0x00000167FE009B80>, keep_feasible=False, finite_diff_rel_step=None, finite_diff_jac_sparsity=None)\n",
      "     |  \n",
      "     |  Nonlinear constraint on the variables.\n",
      "     |  \n",
      "     |  The constraint has the general inequality form::\n",
      "     |  \n",
      "     |      lb <= fun(x) <= ub\n",
      "     |  \n",
      "     |  Here the vector of independent variables x is passed as ndarray of shape\n",
      "     |  (n,) and ``fun`` returns a vector with m components.\n",
      "     |  \n",
      "     |  It is possible to use equal bounds to represent an equality constraint or\n",
      "     |  infinite bounds to represent a one-sided constraint.\n",
      "     |  \n",
      "     |  Parameters\n",
      "     |  ----------\n",
      "     |  fun : callable\n",
      "     |      The function defining the constraint.\n",
      "     |      The signature is ``fun(x) -> array_like, shape (m,)``.\n",
      "     |  lb, ub : array_like\n",
      "     |      Lower and upper bounds on the constraint. Each array must have the\n",
      "     |      shape (m,) or be a scalar, in the latter case a bound will be the same\n",
      "     |      for all components of the constraint. Use ``np.inf`` with an\n",
      "     |      appropriate sign to specify a one-sided constraint.\n",
      "     |      Set components of `lb` and `ub` equal to represent an equality\n",
      "     |      constraint. Note that you can mix constraints of different types:\n",
      "     |      interval, one-sided or equality, by setting different components of\n",
      "     |      `lb` and `ub` as  necessary.\n",
      "     |  jac : {callable,  '2-point', '3-point', 'cs'}, optional\n",
      "     |      Method of computing the Jacobian matrix (an m-by-n matrix,\n",
      "     |      where element (i, j) is the partial derivative of f[i] with\n",
      "     |      respect to x[j]).  The keywords {'2-point', '3-point',\n",
      "     |      'cs'} select a finite difference scheme for the numerical estimation.\n",
      "     |      A callable must have the following signature:\n",
      "     |      ``jac(x) -> {ndarray, sparse matrix}, shape (m, n)``.\n",
      "     |      Default is '2-point'.\n",
      "     |  hess : {callable, '2-point', '3-point', 'cs', HessianUpdateStrategy, None}, optional\n",
      "     |      Method for computing the Hessian matrix. The keywords\n",
      "     |      {'2-point', '3-point', 'cs'} select a finite difference scheme for\n",
      "     |      numerical  estimation.  Alternatively, objects implementing\n",
      "     |      `HessianUpdateStrategy` interface can be used to approximate the\n",
      "     |      Hessian. Currently available implementations are:\n",
      "     |  \n",
      "     |          - `BFGS` (default option)\n",
      "     |          - `SR1`\n",
      "     |  \n",
      "     |      A callable must return the Hessian matrix of ``dot(fun, v)`` and\n",
      "     |      must have the following signature:\n",
      "     |      ``hess(x, v) -> {LinearOperator, sparse matrix, array_like}, shape (n, n)``.\n",
      "     |      Here ``v`` is ndarray with shape (m,) containing Lagrange multipliers.\n",
      "     |  keep_feasible : array_like of bool, optional\n",
      "     |      Whether to keep the constraint components feasible throughout\n",
      "     |      iterations. A single value set this property for all components.\n",
      "     |      Default is False. Has no effect for equality constraints.\n",
      "     |  finite_diff_rel_step: None or array_like, optional\n",
      "     |      Relative step size for the finite difference approximation. Default is\n",
      "     |      None, which will select a reasonable value automatically depending\n",
      "     |      on a finite difference scheme.\n",
      "     |  finite_diff_jac_sparsity: {None, array_like, sparse matrix}, optional\n",
      "     |      Defines the sparsity structure of the Jacobian matrix for finite\n",
      "     |      difference estimation, its shape must be (m, n). If the Jacobian has\n",
      "     |      only few non-zero elements in *each* row, providing the sparsity\n",
      "     |      structure will greatly speed up the computations. A zero entry means\n",
      "     |      that a corresponding element in the Jacobian is identically zero.\n",
      "     |      If provided, forces the use of 'lsmr' trust-region solver.\n",
      "     |      If None (default) then dense differencing will be used.\n",
      "     |  \n",
      "     |  Notes\n",
      "     |  -----\n",
      "     |  Finite difference schemes {'2-point', '3-point', 'cs'} may be used for\n",
      "     |  approximating either the Jacobian or the Hessian. We, however, do not allow\n",
      "     |  its use for approximating both simultaneously. Hence whenever the Jacobian\n",
      "     |  is estimated via finite-differences, we require the Hessian to be estimated\n",
      "     |  using one of the quasi-Newton strategies.\n",
      "     |  \n",
      "     |  The scheme 'cs' is potentially the most accurate, but requires the function\n",
      "     |  to correctly handles complex inputs and be analytically continuable to the\n",
      "     |  complex plane. The scheme '3-point' is more accurate than '2-point' but\n",
      "     |  requires twice as many operations.\n",
      "     |  \n",
      "     |  Examples\n",
      "     |  --------\n",
      "     |  Constrain ``x[0] < sin(x[1]) + 1.9``\n",
      "     |  \n",
      "     |  >>> from scipy.optimize import NonlinearConstraint\n",
      "     |  >>> con = lambda x: x[0] - np.sin(x[1])\n",
      "     |  >>> nlc = NonlinearConstraint(con, -np.inf, 1.9)\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __init__(self, fun, lb, ub, jac='2-point', hess=<scipy.optimize._hessian_update_strategy.BFGS object at 0x00000167FE009B80>, keep_feasible=False, finite_diff_rel_step=None, finite_diff_jac_sparsity=None)\n",
      "     |      Initialize self.  See help(type(self)) for accurate signature.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors defined here:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |      dictionary for instance variables (if defined)\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "    \n",
      "    class OptimizeResult(builtins.dict)\n",
      "     |  Represents the optimization result.\n",
      "     |  \n",
      "     |  Attributes\n",
      "     |  ----------\n",
      "     |  x : ndarray\n",
      "     |      The solution of the optimization.\n",
      "     |  success : bool\n",
      "     |      Whether or not the optimizer exited successfully.\n",
      "     |  status : int\n",
      "     |      Termination status of the optimizer. Its value depends on the\n",
      "     |      underlying solver. Refer to `message` for details.\n",
      "     |  message : str\n",
      "     |      Description of the cause of the termination.\n",
      "     |  fun, jac, hess: ndarray\n",
      "     |      Values of objective function, its Jacobian and its Hessian (if\n",
      "     |      available). The Hessians may be approximations, see the documentation\n",
      "     |      of the function in question.\n",
      "     |  hess_inv : object\n",
      "     |      Inverse of the objective function's Hessian; may be an approximation.\n",
      "     |      Not available for all solvers. The type of this attribute may be\n",
      "     |      either np.ndarray or scipy.sparse.linalg.LinearOperator.\n",
      "     |  nfev, njev, nhev : int\n",
      "     |      Number of evaluations of the objective functions and of its\n",
      "     |      Jacobian and Hessian.\n",
      "     |  nit : int\n",
      "     |      Number of iterations performed by the optimizer.\n",
      "     |  maxcv : float\n",
      "     |      The maximum constraint violation.\n",
      "     |  \n",
      "     |  Notes\n",
      "     |  -----\n",
      "     |  There may be additional attributes not listed above depending of the\n",
      "     |  specific solver. Since this class is essentially a subclass of dict\n",
      "     |  with attribute accessors, one can see which attributes are available\n",
      "     |  using the `keys()` method.\n",
      "     |  \n",
      "     |  Method resolution order:\n",
      "     |      OptimizeResult\n",
      "     |      builtins.dict\n",
      "     |      builtins.object\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __delattr__ = __delitem__(self, key, /)\n",
      "     |  \n",
      "     |  __dir__(self)\n",
      "     |      Default dir() implementation.\n",
      "     |  \n",
      "     |  __getattr__(self, name)\n",
      "     |  \n",
      "     |  __repr__(self)\n",
      "     |      Return repr(self).\n",
      "     |  \n",
      "     |  __setattr__ = __setitem__(self, key, value, /)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors defined here:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |      dictionary for instance variables (if defined)\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from builtins.dict:\n",
      "     |  \n",
      "     |  __contains__(self, key, /)\n",
      "     |      True if the dictionary has the specified key, else False.\n",
      "     |  \n",
      "     |  __delitem__(self, key, /)\n",
      "     |      Delete self[key].\n",
      "     |  \n",
      "     |  __eq__(self, value, /)\n",
      "     |      Return self==value.\n",
      "     |  \n",
      "     |  __ge__(self, value, /)\n",
      "     |      Return self>=value.\n",
      "     |  \n",
      "     |  __getattribute__(self, name, /)\n",
      "     |      Return getattr(self, name).\n",
      "     |  \n",
      "     |  __getitem__(...)\n",
      "     |      x.__getitem__(y) <==> x[y]\n",
      "     |  \n",
      "     |  __gt__(self, value, /)\n",
      "     |      Return self>value.\n",
      "     |  \n",
      "     |  __init__(self, /, *args, **kwargs)\n",
      "     |      Initialize self.  See help(type(self)) for accurate signature.\n",
      "     |  \n",
      "     |  __iter__(self, /)\n",
      "     |      Implement iter(self).\n",
      "     |  \n",
      "     |  __le__(self, value, /)\n",
      "     |      Return self<=value.\n",
      "     |  \n",
      "     |  __len__(self, /)\n",
      "     |      Return len(self).\n",
      "     |  \n",
      "     |  __lt__(self, value, /)\n",
      "     |      Return self<value.\n",
      "     |  \n",
      "     |  __ne__(self, value, /)\n",
      "     |      Return self!=value.\n",
      "     |  \n",
      "     |  __reversed__(self, /)\n",
      "     |      Return a reverse iterator over the dict keys.\n",
      "     |  \n",
      "     |  __setitem__(self, key, value, /)\n",
      "     |      Set self[key] to value.\n",
      "     |  \n",
      "     |  __sizeof__(...)\n",
      "     |      D.__sizeof__() -> size of D in memory, in bytes\n",
      "     |  \n",
      "     |  clear(...)\n",
      "     |      D.clear() -> None.  Remove all items from D.\n",
      "     |  \n",
      "     |  copy(...)\n",
      "     |      D.copy() -> a shallow copy of D\n",
      "     |  \n",
      "     |  get(self, key, default=None, /)\n",
      "     |      Return the value for key if key is in the dictionary, else default.\n",
      "     |  \n",
      "     |  items(...)\n",
      "     |      D.items() -> a set-like object providing a view on D's items\n",
      "     |  \n",
      "     |  keys(...)\n",
      "     |      D.keys() -> a set-like object providing a view on D's keys\n",
      "     |  \n",
      "     |  pop(...)\n",
      "     |      D.pop(k[,d]) -> v, remove specified key and return the corresponding value.\n",
      "     |      If key is not found, d is returned if given, otherwise KeyError is raised\n",
      "     |  \n",
      "     |  popitem(self, /)\n",
      "     |      Remove and return a (key, value) pair as a 2-tuple.\n",
      "     |      \n",
      "     |      Pairs are returned in LIFO (last-in, first-out) order.\n",
      "     |      Raises KeyError if the dict is empty.\n",
      "     |  \n",
      "     |  setdefault(self, key, default=None, /)\n",
      "     |      Insert key with a value of default if key is not in the dictionary.\n",
      "     |      \n",
      "     |      Return the value for key if key is in the dictionary, else default.\n",
      "     |  \n",
      "     |  update(...)\n",
      "     |      D.update([E, ]**F) -> None.  Update D from dict/iterable E and F.\n",
      "     |      If E is present and has a .keys() method, then does:  for k in E: D[k] = E[k]\n",
      "     |      If E is present and lacks a .keys() method, then does:  for k, v in E: D[k] = v\n",
      "     |      In either case, this is followed by: for k in F:  D[k] = F[k]\n",
      "     |  \n",
      "     |  values(...)\n",
      "     |      D.values() -> an object providing a view on D's values\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Class methods inherited from builtins.dict:\n",
      "     |  \n",
      "     |  fromkeys(iterable, value=None, /) from builtins.type\n",
      "     |      Create a new dictionary with keys from iterable and values set to value.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Static methods inherited from builtins.dict:\n",
      "     |  \n",
      "     |  __new__(*args, **kwargs) from builtins.type\n",
      "     |      Create and return a new object.  See help(type) for accurate signature.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data and other attributes inherited from builtins.dict:\n",
      "     |  \n",
      "     |  __hash__ = None\n",
      "    \n",
      "    class OptimizeWarning(builtins.UserWarning)\n",
      "     |  Base class for warnings generated by user code.\n",
      "     |  \n",
      "     |  Method resolution order:\n",
      "     |      OptimizeWarning\n",
      "     |      builtins.UserWarning\n",
      "     |      builtins.Warning\n",
      "     |      builtins.Exception\n",
      "     |      builtins.BaseException\n",
      "     |      builtins.object\n",
      "     |  \n",
      "     |  Data descriptors defined here:\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from builtins.UserWarning:\n",
      "     |  \n",
      "     |  __init__(self, /, *args, **kwargs)\n",
      "     |      Initialize self.  See help(type(self)) for accurate signature.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Static methods inherited from builtins.UserWarning:\n",
      "     |  \n",
      "     |  __new__(*args, **kwargs) from builtins.type\n",
      "     |      Create and return a new object.  See help(type) for accurate signature.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from builtins.BaseException:\n",
      "     |  \n",
      "     |  __delattr__(self, name, /)\n",
      "     |      Implement delattr(self, name).\n",
      "     |  \n",
      "     |  __getattribute__(self, name, /)\n",
      "     |      Return getattr(self, name).\n",
      "     |  \n",
      "     |  __reduce__(...)\n",
      "     |      Helper for pickle.\n",
      "     |  \n",
      "     |  __repr__(self, /)\n",
      "     |      Return repr(self).\n",
      "     |  \n",
      "     |  __setattr__(self, name, value, /)\n",
      "     |      Implement setattr(self, name, value).\n",
      "     |  \n",
      "     |  __setstate__(...)\n",
      "     |  \n",
      "     |  __str__(self, /)\n",
      "     |      Return str(self).\n",
      "     |  \n",
      "     |  with_traceback(...)\n",
      "     |      Exception.with_traceback(tb) --\n",
      "     |      set self.__traceback__ to tb and return self.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from builtins.BaseException:\n",
      "     |  \n",
      "     |  __cause__\n",
      "     |      exception cause\n",
      "     |  \n",
      "     |  __context__\n",
      "     |      exception context\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |  \n",
      "     |  __suppress_context__\n",
      "     |  \n",
      "     |  __traceback__\n",
      "     |  \n",
      "     |  args\n",
      "    \n",
      "    class RootResults(builtins.object)\n",
      "     |  RootResults(root, iterations, function_calls, flag)\n",
      "     |  \n",
      "     |  Represents the root finding result.\n",
      "     |  \n",
      "     |  Attributes\n",
      "     |  ----------\n",
      "     |  root : float\n",
      "     |      Estimated root location.\n",
      "     |  iterations : int\n",
      "     |      Number of iterations needed to find the root.\n",
      "     |  function_calls : int\n",
      "     |      Number of times the function was called.\n",
      "     |  converged : bool\n",
      "     |      True if the routine converged.\n",
      "     |  flag : str\n",
      "     |      Description of the cause of termination.\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __init__(self, root, iterations, function_calls, flag)\n",
      "     |      Initialize self.  See help(type(self)) for accurate signature.\n",
      "     |  \n",
      "     |  __repr__(self)\n",
      "     |      Return repr(self).\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors defined here:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |      dictionary for instance variables (if defined)\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "    \n",
      "    class SR1(FullHessianUpdateStrategy)\n",
      "     |  SR1(min_denominator=1e-08, init_scale='auto')\n",
      "     |  \n",
      "     |  Symmetric-rank-1 Hessian update strategy.\n",
      "     |  \n",
      "     |  Parameters\n",
      "     |  ----------\n",
      "     |  min_denominator : float\n",
      "     |      This number, scaled by a normalization factor,\n",
      "     |      defines the minimum denominator magnitude allowed\n",
      "     |      in the update. When the condition is violated we skip\n",
      "     |      the update. By default uses ``1e-8``.\n",
      "     |  init_scale : {float, 'auto'}, optional\n",
      "     |      Matrix scale at first iteration. At the first\n",
      "     |      iteration the Hessian matrix or its inverse will be initialized\n",
      "     |      with ``init_scale*np.eye(n)``, where ``n`` is the problem dimension.\n",
      "     |      Set it to 'auto' in order to use an automatic heuristic for choosing\n",
      "     |      the initial scale. The heuristic is described in [1]_, p.143.\n",
      "     |      By default uses 'auto'.\n",
      "     |  \n",
      "     |  Notes\n",
      "     |  -----\n",
      "     |  The update is based on the description in [1]_, p.144-146.\n",
      "     |  \n",
      "     |  References\n",
      "     |  ----------\n",
      "     |  .. [1] Nocedal, Jorge, and Stephen J. Wright. \"Numerical optimization\"\n",
      "     |         Second Edition (2006).\n",
      "     |  \n",
      "     |  Method resolution order:\n",
      "     |      SR1\n",
      "     |      FullHessianUpdateStrategy\n",
      "     |      HessianUpdateStrategy\n",
      "     |      builtins.object\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __init__(self, min_denominator=1e-08, init_scale='auto')\n",
      "     |      Initialize self.  See help(type(self)) for accurate signature.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from FullHessianUpdateStrategy:\n",
      "     |  \n",
      "     |  dot(self, p)\n",
      "     |      Compute the product of the internal matrix with the given vector.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      p : array_like\n",
      "     |          1-D array representing a vector.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      Hp : array\n",
      "     |          1-D represents the result of multiplying the approximation matrix\n",
      "     |          by vector p.\n",
      "     |  \n",
      "     |  get_matrix(self)\n",
      "     |      Return the current internal matrix.\n",
      "     |      \n",
      "     |      Returns\n",
      "     |      -------\n",
      "     |      M : ndarray, shape (n, n)\n",
      "     |          Dense matrix containing either the Hessian or its inverse\n",
      "     |          (depending on how `approx_type` was defined).\n",
      "     |  \n",
      "     |  initialize(self, n, approx_type)\n",
      "     |      Initialize internal matrix.\n",
      "     |      \n",
      "     |      Allocate internal memory for storing and updating\n",
      "     |      the Hessian or its inverse.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      n : int\n",
      "     |          Problem dimension.\n",
      "     |      approx_type : {'hess', 'inv_hess'}\n",
      "     |          Selects either the Hessian or the inverse Hessian.\n",
      "     |          When set to 'hess' the Hessian will be stored and updated.\n",
      "     |          When set to 'inv_hess' its inverse will be used instead.\n",
      "     |  \n",
      "     |  update(self, delta_x, delta_grad)\n",
      "     |      Update internal matrix.\n",
      "     |      \n",
      "     |      Update Hessian matrix or its inverse (depending on how 'approx_type'\n",
      "     |      is defined) using information about the last evaluated points.\n",
      "     |      \n",
      "     |      Parameters\n",
      "     |      ----------\n",
      "     |      delta_x : ndarray\n",
      "     |          The difference between two points the gradient\n",
      "     |          function have been evaluated at: ``delta_x = x2 - x1``.\n",
      "     |      delta_grad : ndarray\n",
      "     |          The difference between the gradients:\n",
      "     |          ``delta_grad = grad(x2) - grad(x1)``.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from HessianUpdateStrategy:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |      dictionary for instance variables (if defined)\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "\n",
      "FUNCTIONS\n",
      "    anderson(F, xin, iter=None, alpha=None, w0=0.01, M=5, verbose=False, maxiter=None, f_tol=None, f_rtol=None, x_tol=None, x_rtol=None, tol_norm=None, line_search='armijo', callback=None, **kw)\n",
      "        Find a root of a function, using (extended) Anderson mixing.\n",
      "        \n",
      "        The Jacobian is formed by for a 'best' solution in the space\n",
      "        spanned by last `M` vectors. As a result, only a MxM matrix\n",
      "        inversions and MxN multiplications are required. [Ey]_\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        F : function(x) -> f\n",
      "            Function whose root to find; should take and return an array-like\n",
      "            object.\n",
      "        xin : array_like\n",
      "            Initial guess for the solution\n",
      "        alpha : float, optional\n",
      "            Initial guess for the Jacobian is (-1/alpha).\n",
      "        M : float, optional\n",
      "            Number of previous vectors to retain. Defaults to 5.\n",
      "        w0 : float, optional\n",
      "            Regularization parameter for numerical stability.\n",
      "            Compared to unity, good values of the order of 0.01.\n",
      "        iter : int, optional\n",
      "            Number of iterations to make. If omitted (default), make as many\n",
      "            as required to meet tolerances.\n",
      "        verbose : bool, optional\n",
      "            Print status to stdout on every iteration.\n",
      "        maxiter : int, optional\n",
      "            Maximum number of iterations to make. If more are needed to\n",
      "            meet convergence, `NoConvergence` is raised.\n",
      "        f_tol : float, optional\n",
      "            Absolute tolerance (in max-norm) for the residual.\n",
      "            If omitted, default is 6e-6.\n",
      "        f_rtol : float, optional\n",
      "            Relative tolerance for the residual. If omitted, not used.\n",
      "        x_tol : float, optional\n",
      "            Absolute minimum step size, as determined from the Jacobian\n",
      "            approximation. If the step size is smaller than this, optimization\n",
      "            is terminated as successful. If omitted, not used.\n",
      "        x_rtol : float, optional\n",
      "            Relative minimum step size. If omitted, not used.\n",
      "        tol_norm : function(vector) -> scalar, optional\n",
      "            Norm to use in convergence check. Default is the maximum norm.\n",
      "        line_search : {None, 'armijo' (default), 'wolfe'}, optional\n",
      "            Which type of a line search to use to determine the step size in the\n",
      "            direction given by the Jacobian approximation. Defaults to 'armijo'.\n",
      "        callback : function, optional\n",
      "            Optional callback function. It is called on every iteration as\n",
      "            ``callback(x, f)`` where `x` is the current solution and `f`\n",
      "            the corresponding residual.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        sol : ndarray\n",
      "            An array (of similar array type as `x0`) containing the final solution.\n",
      "        \n",
      "        Raises\n",
      "        ------\n",
      "        NoConvergence\n",
      "            When a solution was not found.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        root : Interface to root finding algorithms for multivariate\n",
      "               functions. See ``method=='anderson'`` in particular.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [Ey] V. Eyert, J. Comp. Phys., 124, 271 (1996).\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        The following functions define a system of nonlinear equations\n",
      "        \n",
      "        >>> def fun(x):\n",
      "        ...     return [x[0]  + 0.5 * (x[0] - x[1])**3 - 1.0,\n",
      "        ...             0.5 * (x[1] - x[0])**3 + x[1]]\n",
      "        \n",
      "        A solution can be obtained as follows.\n",
      "        \n",
      "        >>> from scipy import optimize\n",
      "        >>> sol = optimize.anderson(fun, [0, 0])\n",
      "        >>> sol\n",
      "        array([0.84116588, 0.15883789])\n",
      "    \n",
      "    approx_fprime(xk, f, epsilon, *args)\n",
      "        Finite-difference approximation of the gradient of a scalar function.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        xk : array_like\n",
      "            The coordinate vector at which to determine the gradient of `f`.\n",
      "        f : callable\n",
      "            The function of which to determine the gradient (partial derivatives).\n",
      "            Should take `xk` as first argument, other arguments to `f` can be\n",
      "            supplied in ``*args``. Should return a scalar, the value of the\n",
      "            function at `xk`.\n",
      "        epsilon : array_like\n",
      "            Increment to `xk` to use for determining the function gradient.\n",
      "            If a scalar, uses the same finite difference delta for all partial\n",
      "            derivatives. If an array, should contain one value per element of\n",
      "            `xk`.\n",
      "        \\*args : args, optional\n",
      "            Any other arguments that are to be passed to `f`.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        grad : ndarray\n",
      "            The partial derivatives of `f` to `xk`.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        check_grad : Check correctness of gradient function against approx_fprime.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        The function gradient is determined by the forward finite difference\n",
      "        formula::\n",
      "        \n",
      "                     f(xk[i] + epsilon[i]) - f(xk[i])\n",
      "            f'[i] = ---------------------------------\n",
      "                                epsilon[i]\n",
      "        \n",
      "        The main use of `approx_fprime` is in scalar function optimizers like\n",
      "        `fmin_bfgs`, to determine numerically the Jacobian of a function.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy import optimize\n",
      "        >>> def func(x, c0, c1):\n",
      "        ...     \"Coordinate vector `x` should be an array of size two.\"\n",
      "        ...     return c0 * x[0]**2 + c1*x[1]**2\n",
      "        \n",
      "        >>> x = np.ones(2)\n",
      "        >>> c0, c1 = (1, 200)\n",
      "        >>> eps = np.sqrt(np.finfo(float).eps)\n",
      "        >>> optimize.approx_fprime(x, func, [eps, np.sqrt(200) * eps], c0, c1)\n",
      "        array([   2.        ,  400.00004198])\n",
      "    \n",
      "    basinhopping(func, x0, niter=100, T=1.0, stepsize=0.5, minimizer_kwargs=None, take_step=None, accept_test=None, callback=None, interval=50, disp=False, niter_success=None, seed=None)\n",
      "        Find the global minimum of a function using the basin-hopping algorithm.\n",
      "        \n",
      "        Basin-hopping is a two-phase method that combines a global stepping\n",
      "        algorithm with local minimization at each step. Designed to mimic\n",
      "        the natural process of energy minimization of clusters of atoms, it works\n",
      "        well for similar problems with \"funnel-like, but rugged\" energy landscapes\n",
      "        [5]_.\n",
      "        \n",
      "        As the step-taking, step acceptance, and minimization methods are all\n",
      "        customizable, this function can also be used to implement other two-phase\n",
      "        methods.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        func : callable ``f(x, *args)``\n",
      "            Function to be optimized.  ``args`` can be passed as an optional item\n",
      "            in the dict ``minimizer_kwargs``\n",
      "        x0 : array_like\n",
      "            Initial guess.\n",
      "        niter : integer, optional\n",
      "            The number of basin-hopping iterations. There will be a total of\n",
      "            ``niter + 1`` runs of the local minimizer.\n",
      "        T : float, optional\n",
      "            The \"temperature\" parameter for the accept or reject criterion. Higher\n",
      "            \"temperatures\" mean that larger jumps in function value will be\n",
      "            accepted.  For best results ``T`` should be comparable to the\n",
      "            separation (in function value) between local minima.\n",
      "        stepsize : float, optional\n",
      "            Maximum step size for use in the random displacement.\n",
      "        minimizer_kwargs : dict, optional\n",
      "            Extra keyword arguments to be passed to the local minimizer\n",
      "            ``scipy.optimize.minimize()`` Some important options could be:\n",
      "        \n",
      "                method : str\n",
      "                    The minimization method (e.g. ``\"L-BFGS-B\"``)\n",
      "                args : tuple\n",
      "                    Extra arguments passed to the objective function (``func``) and\n",
      "                    its derivatives (Jacobian, Hessian).\n",
      "        \n",
      "        take_step : callable ``take_step(x)``, optional\n",
      "            Replace the default step-taking routine with this routine. The default\n",
      "            step-taking routine is a random displacement of the coordinates, but\n",
      "            other step-taking algorithms may be better for some systems.\n",
      "            ``take_step`` can optionally have the attribute ``take_step.stepsize``.\n",
      "            If this attribute exists, then ``basinhopping`` will adjust\n",
      "            ``take_step.stepsize`` in order to try to optimize the global minimum\n",
      "            search.\n",
      "        accept_test : callable, ``accept_test(f_new=f_new, x_new=x_new, f_old=fold, x_old=x_old)``, optional\n",
      "            Define a test which will be used to judge whether or not to accept the\n",
      "            step.  This will be used in addition to the Metropolis test based on\n",
      "            \"temperature\" ``T``.  The acceptable return values are True,\n",
      "            False, or ``\"force accept\"``. If any of the tests return False\n",
      "            then the step is rejected. If the latter, then this will override any\n",
      "            other tests in order to accept the step. This can be used, for example,\n",
      "            to forcefully escape from a local minimum that ``basinhopping`` is\n",
      "            trapped in.\n",
      "        callback : callable, ``callback(x, f, accept)``, optional\n",
      "            A callback function which will be called for all minima found. ``x``\n",
      "            and ``f`` are the coordinates and function value of the trial minimum,\n",
      "            and ``accept`` is whether or not that minimum was accepted. This can\n",
      "            be used, for example, to save the lowest N minima found. Also,\n",
      "            ``callback`` can be used to specify a user defined stop criterion by\n",
      "            optionally returning True to stop the ``basinhopping`` routine.\n",
      "        interval : integer, optional\n",
      "            interval for how often to update the ``stepsize``\n",
      "        disp : bool, optional\n",
      "            Set to True to print status messages\n",
      "        niter_success : integer, optional\n",
      "            Stop the run if the global minimum candidate remains the same for this\n",
      "            number of iterations.\n",
      "        seed : {None, int, `numpy.random.Generator`,\n",
      "                `numpy.random.RandomState`}, optional\n",
      "        \n",
      "            If `seed` is None (or `np.random`), the `numpy.random.RandomState`\n",
      "            singleton is used.\n",
      "            If `seed` is an int, a new ``RandomState`` instance is used,\n",
      "            seeded with `seed`.\n",
      "            If `seed` is already a ``Generator`` or ``RandomState`` instance then\n",
      "            that instance is used.\n",
      "            Specify `seed` for repeatable minimizations. The random numbers\n",
      "            generated with this seed only affect the default Metropolis\n",
      "            `accept_test` and the default `take_step`. If you supply your own\n",
      "            `take_step` and `accept_test`, and these functions use random\n",
      "            number generation, then those functions are responsible for the state\n",
      "            of their random number generator.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        res : OptimizeResult\n",
      "            The optimization result represented as a ``OptimizeResult`` object.\n",
      "            Important attributes are: ``x`` the solution array, ``fun`` the value\n",
      "            of the function at the solution, and ``message`` which describes the\n",
      "            cause of the termination. The ``OptimizeResult`` object returned by the\n",
      "            selected minimizer at the lowest minimum is also contained within this\n",
      "            object and can be accessed through the ``lowest_optimization_result``\n",
      "            attribute.  See `OptimizeResult` for a description of other attributes.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        minimize :\n",
      "            The local minimization function called once for each basinhopping step.\n",
      "            ``minimizer_kwargs`` is passed to this routine.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        Basin-hopping is a stochastic algorithm which attempts to find the global\n",
      "        minimum of a smooth scalar function of one or more variables [1]_ [2]_ [3]_\n",
      "        [4]_. The algorithm in its current form was described by David Wales and\n",
      "        Jonathan Doye [2]_ http://www-wales.ch.cam.ac.uk/.\n",
      "        \n",
      "        The algorithm is iterative with each cycle composed of the following\n",
      "        features\n",
      "        \n",
      "        1) random perturbation of the coordinates\n",
      "        \n",
      "        2) local minimization\n",
      "        \n",
      "        3) accept or reject the new coordinates based on the minimized function\n",
      "           value\n",
      "        \n",
      "        The acceptance test used here is the Metropolis criterion of standard Monte\n",
      "        Carlo algorithms, although there are many other possibilities [3]_.\n",
      "        \n",
      "        This global minimization method has been shown to be extremely efficient\n",
      "        for a wide variety of problems in physics and chemistry. It is\n",
      "        particularly useful when the function has many minima separated by large\n",
      "        barriers. See the Cambridge Cluster Database\n",
      "        http://www-wales.ch.cam.ac.uk/CCD.html for databases of molecular systems\n",
      "        that have been optimized primarily using basin-hopping. This database\n",
      "        includes minimization problems exceeding 300 degrees of freedom.\n",
      "        \n",
      "        See the free software program GMIN (http://www-wales.ch.cam.ac.uk/GMIN) for\n",
      "        a Fortran implementation of basin-hopping. This implementation has many\n",
      "        different variations of the procedure described above, including more\n",
      "        advanced step taking algorithms and alternate acceptance criterion.\n",
      "        \n",
      "        For stochastic global optimization there is no way to determine if the true\n",
      "        global minimum has actually been found. Instead, as a consistency check,\n",
      "        the algorithm can be run from a number of different random starting points\n",
      "        to ensure the lowest minimum found in each example has converged to the\n",
      "        global minimum. For this reason, ``basinhopping`` will by default simply\n",
      "        run for the number of iterations ``niter`` and return the lowest minimum\n",
      "        found. It is left to the user to ensure that this is in fact the global\n",
      "        minimum.\n",
      "        \n",
      "        Choosing ``stepsize``:  This is a crucial parameter in ``basinhopping`` and\n",
      "        depends on the problem being solved. The step is chosen uniformly in the\n",
      "        region from x0-stepsize to x0+stepsize, in each dimension. Ideally, it\n",
      "        should be comparable to the typical separation (in argument values) between\n",
      "        local minima of the function being optimized. ``basinhopping`` will, by\n",
      "        default, adjust ``stepsize`` to find an optimal value, but this may take\n",
      "        many iterations. You will get quicker results if you set a sensible\n",
      "        initial value for ``stepsize``.\n",
      "        \n",
      "        Choosing ``T``: The parameter ``T`` is the \"temperature\" used in the\n",
      "        Metropolis criterion. Basinhopping steps are always accepted if\n",
      "        ``func(xnew) < func(xold)``. Otherwise, they are accepted with\n",
      "        probability::\n",
      "        \n",
      "            exp( -(func(xnew) - func(xold)) / T )\n",
      "        \n",
      "        So, for best results, ``T`` should to be comparable to the typical\n",
      "        difference (in function values) between local minima. (The height of\n",
      "        \"walls\" between local minima is irrelevant.)\n",
      "        \n",
      "        If ``T`` is 0, the algorithm becomes Monotonic Basin-Hopping, in which all\n",
      "        steps that increase energy are rejected.\n",
      "        \n",
      "        .. versionadded:: 0.12.0\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] Wales, David J. 2003, Energy Landscapes, Cambridge University Press,\n",
      "            Cambridge, UK.\n",
      "        .. [2] Wales, D J, and Doye J P K, Global Optimization by Basin-Hopping and\n",
      "            the Lowest Energy Structures of Lennard-Jones Clusters Containing up to\n",
      "            110 Atoms.  Journal of Physical Chemistry A, 1997, 101, 5111.\n",
      "        .. [3] Li, Z. and Scheraga, H. A., Monte Carlo-minimization approach to the\n",
      "            multiple-minima problem in protein folding, Proc. Natl. Acad. Sci. USA,\n",
      "            1987, 84, 6611.\n",
      "        .. [4] Wales, D. J. and Scheraga, H. A., Global optimization of clusters,\n",
      "            crystals, and biomolecules, Science, 1999, 285, 1368.\n",
      "        .. [5] Olson, B., Hashmi, I., Molloy, K., and Shehu1, A., Basin Hopping as\n",
      "            a General and Versatile Optimization Framework for the Characterization\n",
      "            of Biological Macromolecules, Advances in Artificial Intelligence,\n",
      "            Volume 2012 (2012), Article ID 674832, :doi:`10.1155/2012/674832`\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        The following example is a 1-D minimization problem, with many\n",
      "        local minima superimposed on a parabola.\n",
      "        \n",
      "        >>> from scipy.optimize import basinhopping\n",
      "        >>> func = lambda x: np.cos(14.5 * x - 0.3) + (x + 0.2) * x\n",
      "        >>> x0=[1.]\n",
      "        \n",
      "        Basinhopping, internally, uses a local minimization algorithm. We will use\n",
      "        the parameter ``minimizer_kwargs`` to tell basinhopping which algorithm to\n",
      "        use and how to set up that minimizer. This parameter will be passed to\n",
      "        ``scipy.optimize.minimize()``.\n",
      "        \n",
      "        >>> minimizer_kwargs = {\"method\": \"BFGS\"}\n",
      "        >>> ret = basinhopping(func, x0, minimizer_kwargs=minimizer_kwargs,\n",
      "        ...                    niter=200)\n",
      "        >>> print(\"global minimum: x = %.4f, f(x0) = %.4f\" % (ret.x, ret.fun))\n",
      "        global minimum: x = -0.1951, f(x0) = -1.0009\n",
      "        \n",
      "        Next consider a 2-D minimization problem. Also, this time, we\n",
      "        will use gradient information to significantly speed up the search.\n",
      "        \n",
      "        >>> def func2d(x):\n",
      "        ...     f = np.cos(14.5 * x[0] - 0.3) + (x[1] + 0.2) * x[1] + (x[0] +\n",
      "        ...                                                            0.2) * x[0]\n",
      "        ...     df = np.zeros(2)\n",
      "        ...     df[0] = -14.5 * np.sin(14.5 * x[0] - 0.3) + 2. * x[0] + 0.2\n",
      "        ...     df[1] = 2. * x[1] + 0.2\n",
      "        ...     return f, df\n",
      "        \n",
      "        We'll also use a different local minimization algorithm. Also, we must tell\n",
      "        the minimizer that our function returns both energy and gradient (Jacobian).\n",
      "        \n",
      "        >>> minimizer_kwargs = {\"method\":\"L-BFGS-B\", \"jac\":True}\n",
      "        >>> x0 = [1.0, 1.0]\n",
      "        >>> ret = basinhopping(func2d, x0, minimizer_kwargs=minimizer_kwargs,\n",
      "        ...                    niter=200)\n",
      "        >>> print(\"global minimum: x = [%.4f, %.4f], f(x0) = %.4f\" % (ret.x[0],\n",
      "        ...                                                           ret.x[1],\n",
      "        ...                                                           ret.fun))\n",
      "        global minimum: x = [-0.1951, -0.1000], f(x0) = -1.0109\n",
      "        \n",
      "        \n",
      "        Here is an example using a custom step-taking routine. Imagine you want\n",
      "        the first coordinate to take larger steps than the rest of the coordinates.\n",
      "        This can be implemented like so:\n",
      "        \n",
      "        >>> class MyTakeStep:\n",
      "        ...    def __init__(self, stepsize=0.5):\n",
      "        ...        self.stepsize = stepsize\n",
      "        ...        self.rng = np.random.default_rng()\n",
      "        ...    def __call__(self, x):\n",
      "        ...        s = self.stepsize\n",
      "        ...        x[0] += self.rng.uniform(-2.*s, 2.*s)\n",
      "        ...        x[1:] += self.rng.uniform(-s, s, x[1:].shape)\n",
      "        ...        return x\n",
      "        \n",
      "        Since ``MyTakeStep.stepsize`` exists basinhopping will adjust the magnitude\n",
      "        of ``stepsize`` to optimize the search. We'll use the same 2-D function as\n",
      "        before\n",
      "        \n",
      "        >>> mytakestep = MyTakeStep()\n",
      "        >>> ret = basinhopping(func2d, x0, minimizer_kwargs=minimizer_kwargs,\n",
      "        ...                    niter=200, take_step=mytakestep)\n",
      "        >>> print(\"global minimum: x = [%.4f, %.4f], f(x0) = %.4f\" % (ret.x[0],\n",
      "        ...                                                           ret.x[1],\n",
      "        ...                                                           ret.fun))\n",
      "        global minimum: x = [-0.1951, -0.1000], f(x0) = -1.0109\n",
      "        \n",
      "        \n",
      "        Now, let's do an example using a custom callback function which prints the\n",
      "        value of every minimum found\n",
      "        \n",
      "        >>> def print_fun(x, f, accepted):\n",
      "        ...         print(\"at minimum %.4f accepted %d\" % (f, int(accepted)))\n",
      "        \n",
      "        We'll run it for only 10 basinhopping steps this time.\n",
      "        \n",
      "        >>> rng = np.random.default_rng()\n",
      "        >>> ret = basinhopping(func2d, x0, minimizer_kwargs=minimizer_kwargs,\n",
      "        ...                    niter=10, callback=print_fun, seed=rng)\n",
      "        at minimum 0.4159 accepted 1\n",
      "        at minimum -0.4317 accepted 1\n",
      "        at minimum -1.0109 accepted 1\n",
      "        at minimum -0.9073 accepted 1\n",
      "        at minimum -0.4317 accepted 0\n",
      "        at minimum -0.1021 accepted 1\n",
      "        at minimum -0.7425 accepted 1\n",
      "        at minimum -0.9073 accepted 1\n",
      "        at minimum -0.4317 accepted 0\n",
      "        at minimum -0.7425 accepted 1\n",
      "        at minimum -0.9073 accepted 1\n",
      "        \n",
      "        \n",
      "        The minimum at -1.0109 is actually the global minimum, found already on the\n",
      "        8th iteration.\n",
      "        \n",
      "        Now let's implement bounds on the problem using a custom ``accept_test``:\n",
      "        \n",
      "        >>> class MyBounds:\n",
      "        ...     def __init__(self, xmax=[1.1,1.1], xmin=[-1.1,-1.1] ):\n",
      "        ...         self.xmax = np.array(xmax)\n",
      "        ...         self.xmin = np.array(xmin)\n",
      "        ...     def __call__(self, **kwargs):\n",
      "        ...         x = kwargs[\"x_new\"]\n",
      "        ...         tmax = bool(np.all(x <= self.xmax))\n",
      "        ...         tmin = bool(np.all(x >= self.xmin))\n",
      "        ...         return tmax and tmin\n",
      "        \n",
      "        >>> mybounds = MyBounds()\n",
      "        >>> ret = basinhopping(func2d, x0, minimizer_kwargs=minimizer_kwargs,\n",
      "        ...                    niter=10, accept_test=mybounds)\n",
      "    \n",
      "    bisect(f, a, b, args=(), xtol=2e-12, rtol=8.881784197001252e-16, maxiter=100, full_output=False, disp=True)\n",
      "        Find root of a function within an interval using bisection.\n",
      "        \n",
      "        Basic bisection routine to find a zero of the function `f` between the\n",
      "        arguments `a` and `b`. `f(a)` and `f(b)` cannot have the same signs.\n",
      "        Slow but sure.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        f : function\n",
      "            Python function returning a number.  `f` must be continuous, and\n",
      "            f(a) and f(b) must have opposite signs.\n",
      "        a : scalar\n",
      "            One end of the bracketing interval [a,b].\n",
      "        b : scalar\n",
      "            The other end of the bracketing interval [a,b].\n",
      "        xtol : number, optional\n",
      "            The computed root ``x0`` will satisfy ``np.allclose(x, x0,\n",
      "            atol=xtol, rtol=rtol)``, where ``x`` is the exact root. The\n",
      "            parameter must be nonnegative.\n",
      "        rtol : number, optional\n",
      "            The computed root ``x0`` will satisfy ``np.allclose(x, x0,\n",
      "            atol=xtol, rtol=rtol)``, where ``x`` is the exact root. The\n",
      "            parameter cannot be smaller than its default value of\n",
      "            ``4*np.finfo(float).eps``.\n",
      "        maxiter : int, optional\n",
      "            If convergence is not achieved in `maxiter` iterations, an error is\n",
      "            raised. Must be >= 0.\n",
      "        args : tuple, optional\n",
      "            Containing extra arguments for the function `f`.\n",
      "            `f` is called by ``apply(f, (x)+args)``.\n",
      "        full_output : bool, optional\n",
      "            If `full_output` is False, the root is returned. If `full_output` is\n",
      "            True, the return value is ``(x, r)``, where x is the root, and r is\n",
      "            a `RootResults` object.\n",
      "        disp : bool, optional\n",
      "            If True, raise RuntimeError if the algorithm didn't converge.\n",
      "            Otherwise, the convergence status is recorded in a `RootResults`\n",
      "            return object.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        x0 : float\n",
      "            Zero of `f` between `a` and `b`.\n",
      "        r : `RootResults` (present if ``full_output = True``)\n",
      "            Object containing information about the convergence. In particular,\n",
      "            ``r.converged`` is True if the routine converged.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        \n",
      "        >>> def f(x):\n",
      "        ...     return (x**2 - 1)\n",
      "        \n",
      "        >>> from scipy import optimize\n",
      "        \n",
      "        >>> root = optimize.bisect(f, 0, 2)\n",
      "        >>> root\n",
      "        1.0\n",
      "        \n",
      "        >>> root = optimize.bisect(f, -2, 0)\n",
      "        >>> root\n",
      "        -1.0\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        brentq, brenth, bisect, newton\n",
      "        fixed_point : scalar fixed-point finder\n",
      "        fsolve : n-dimensional root-finding\n",
      "    \n",
      "    bracket(func, xa=0.0, xb=1.0, args=(), grow_limit=110.0, maxiter=1000)\n",
      "        Bracket the minimum of the function.\n",
      "        \n",
      "        Given a function and distinct initial points, search in the\n",
      "        downhill direction (as defined by the initial points) and return\n",
      "        new points xa, xb, xc that bracket the minimum of the function\n",
      "        f(xa) > f(xb) < f(xc). It doesn't always mean that obtained\n",
      "        solution will satisfy xa<=x<=xb.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        func : callable f(x,*args)\n",
      "            Objective function to minimize.\n",
      "        xa, xb : float, optional\n",
      "            Bracketing interval. Defaults `xa` to 0.0, and `xb` to 1.0.\n",
      "        args : tuple, optional\n",
      "            Additional arguments (if present), passed to `func`.\n",
      "        grow_limit : float, optional\n",
      "            Maximum grow limit.  Defaults to 110.0\n",
      "        maxiter : int, optional\n",
      "            Maximum number of iterations to perform. Defaults to 1000.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        xa, xb, xc : float\n",
      "            Bracket.\n",
      "        fa, fb, fc : float\n",
      "            Objective function values in bracket.\n",
      "        funcalls : int\n",
      "            Number of function evaluations made.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        This function can find a downward convex region of a function:\n",
      "        \n",
      "        >>> import matplotlib.pyplot as plt\n",
      "        >>> from scipy.optimize import bracket\n",
      "        >>> def f(x):\n",
      "        ...     return 10*x**2 + 3*x + 5\n",
      "        >>> x = np.linspace(-2, 2)\n",
      "        >>> y = f(x)\n",
      "        >>> init_xa, init_xb = 0, 1\n",
      "        >>> xa, xb, xc, fa, fb, fc, funcalls = bracket(f, xa=init_xa, xb=init_xb)\n",
      "        >>> plt.axvline(x=init_xa, color=\"k\", linestyle=\"--\")\n",
      "        >>> plt.axvline(x=init_xb, color=\"k\", linestyle=\"--\")\n",
      "        >>> plt.plot(x, y, \"-k\")\n",
      "        >>> plt.plot(xa, fa, \"bx\")\n",
      "        >>> plt.plot(xb, fb, \"rx\")\n",
      "        >>> plt.plot(xc, fc, \"bx\")\n",
      "        >>> plt.show()\n",
      "    \n",
      "    brent(func, args=(), brack=None, tol=1.48e-08, full_output=0, maxiter=500)\n",
      "        Given a function of one variable and a possible bracket, return\n",
      "        the local minimum of the function isolated to a fractional precision\n",
      "        of tol.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        func : callable f(x,*args)\n",
      "            Objective function.\n",
      "        args : tuple, optional\n",
      "            Additional arguments (if present).\n",
      "        brack : tuple, optional\n",
      "            Either a triple (xa,xb,xc) where xa<xb<xc and func(xb) <\n",
      "            func(xa), func(xc) or a pair (xa,xb) which are used as a\n",
      "            starting interval for a downhill bracket search (see\n",
      "            `bracket`). Providing the pair (xa,xb) does not always mean\n",
      "            the obtained solution will satisfy xa<=x<=xb.\n",
      "        tol : float, optional\n",
      "            Stop if between iteration change is less than `tol`.\n",
      "        full_output : bool, optional\n",
      "            If True, return all output args (xmin, fval, iter,\n",
      "            funcalls).\n",
      "        maxiter : int, optional\n",
      "            Maximum number of iterations in solution.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        xmin : ndarray\n",
      "            Optimum point.\n",
      "        fval : float\n",
      "            Optimum value.\n",
      "        iter : int\n",
      "            Number of iterations.\n",
      "        funcalls : int\n",
      "            Number of objective function evaluations made.\n",
      "        \n",
      "        See also\n",
      "        --------\n",
      "        minimize_scalar: Interface to minimization algorithms for scalar\n",
      "            univariate functions. See the 'Brent' `method` in particular.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        Uses inverse parabolic interpolation when possible to speed up\n",
      "        convergence of golden section method.\n",
      "        \n",
      "        Does not ensure that the minimum lies in the range specified by\n",
      "        `brack`. See `fminbound`.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        We illustrate the behaviour of the function when `brack` is of\n",
      "        size 2 and 3 respectively. In the case where `brack` is of the\n",
      "        form (xa,xb), we can see for the given values, the output need\n",
      "        not necessarily lie in the range (xa,xb).\n",
      "        \n",
      "        >>> def f(x):\n",
      "        ...     return x**2\n",
      "        \n",
      "        >>> from scipy import optimize\n",
      "        \n",
      "        >>> minimum = optimize.brent(f,brack=(1,2))\n",
      "        >>> minimum\n",
      "        0.0\n",
      "        >>> minimum = optimize.brent(f,brack=(-1,0.5,2))\n",
      "        >>> minimum\n",
      "        -2.7755575615628914e-17\n",
      "    \n",
      "    brenth(f, a, b, args=(), xtol=2e-12, rtol=8.881784197001252e-16, maxiter=100, full_output=False, disp=True)\n",
      "        Find a root of a function in a bracketing interval using Brent's\n",
      "        method with hyperbolic extrapolation.\n",
      "        \n",
      "        A variation on the classic Brent routine to find a zero of the function f\n",
      "        between the arguments a and b that uses hyperbolic extrapolation instead of\n",
      "        inverse quadratic extrapolation. There was a paper back in the 1980's ...\n",
      "        f(a) and f(b) cannot have the same signs. Generally, on a par with the\n",
      "        brent routine, but not as heavily tested. It is a safe version of the\n",
      "        secant method that uses hyperbolic extrapolation. The version here is by\n",
      "        Chuck Harris.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        f : function\n",
      "            Python function returning a number. f must be continuous, and f(a) and\n",
      "            f(b) must have opposite signs.\n",
      "        a : scalar\n",
      "            One end of the bracketing interval [a,b].\n",
      "        b : scalar\n",
      "            The other end of the bracketing interval [a,b].\n",
      "        xtol : number, optional\n",
      "            The computed root ``x0`` will satisfy ``np.allclose(x, x0,\n",
      "            atol=xtol, rtol=rtol)``, where ``x`` is the exact root. The\n",
      "            parameter must be nonnegative. As with `brentq`, for nice\n",
      "            functions the method will often satisfy the above condition\n",
      "            with ``xtol/2`` and ``rtol/2``.\n",
      "        rtol : number, optional\n",
      "            The computed root ``x0`` will satisfy ``np.allclose(x, x0,\n",
      "            atol=xtol, rtol=rtol)``, where ``x`` is the exact root. The\n",
      "            parameter cannot be smaller than its default value of\n",
      "            ``4*np.finfo(float).eps``. As with `brentq`, for nice functions\n",
      "            the method will often satisfy the above condition with\n",
      "            ``xtol/2`` and ``rtol/2``.\n",
      "        maxiter : int, optional\n",
      "            If convergence is not achieved in `maxiter` iterations, an error is\n",
      "            raised. Must be >= 0.\n",
      "        args : tuple, optional\n",
      "            Containing extra arguments for the function `f`.\n",
      "            `f` is called by ``apply(f, (x)+args)``.\n",
      "        full_output : bool, optional\n",
      "            If `full_output` is False, the root is returned. If `full_output` is\n",
      "            True, the return value is ``(x, r)``, where `x` is the root, and `r` is\n",
      "            a `RootResults` object.\n",
      "        disp : bool, optional\n",
      "            If True, raise RuntimeError if the algorithm didn't converge.\n",
      "            Otherwise, the convergence status is recorded in any `RootResults`\n",
      "            return object.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        x0 : float\n",
      "            Zero of `f` between `a` and `b`.\n",
      "        r : `RootResults` (present if ``full_output = True``)\n",
      "            Object containing information about the convergence. In particular,\n",
      "            ``r.converged`` is True if the routine converged.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> def f(x):\n",
      "        ...     return (x**2 - 1)\n",
      "        \n",
      "        >>> from scipy import optimize\n",
      "        \n",
      "        >>> root = optimize.brenth(f, -2, 0)\n",
      "        >>> root\n",
      "        -1.0\n",
      "        \n",
      "        >>> root = optimize.brenth(f, 0, 2)\n",
      "        >>> root\n",
      "        1.0\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        fmin, fmin_powell, fmin_cg,\n",
      "               fmin_bfgs, fmin_ncg : multivariate local optimizers\n",
      "        \n",
      "        leastsq : nonlinear least squares minimizer\n",
      "        \n",
      "        fmin_l_bfgs_b, fmin_tnc, fmin_cobyla : constrained multivariate optimizers\n",
      "        \n",
      "        basinhopping, differential_evolution, brute : global optimizers\n",
      "        \n",
      "        fminbound, brent, golden, bracket : local scalar minimizers\n",
      "        \n",
      "        fsolve : N-D root-finding\n",
      "        \n",
      "        brentq, brenth, ridder, bisect, newton : 1-D root-finding\n",
      "        \n",
      "        fixed_point : scalar fixed-point finder\n",
      "    \n",
      "    brentq(f, a, b, args=(), xtol=2e-12, rtol=8.881784197001252e-16, maxiter=100, full_output=False, disp=True)\n",
      "        Find a root of a function in a bracketing interval using Brent's method.\n",
      "        \n",
      "        Uses the classic Brent's method to find a zero of the function `f` on\n",
      "        the sign changing interval [a , b]. Generally considered the best of the\n",
      "        rootfinding routines here. It is a safe version of the secant method that\n",
      "        uses inverse quadratic extrapolation. Brent's method combines root\n",
      "        bracketing, interval bisection, and inverse quadratic interpolation. It is\n",
      "        sometimes known as the van Wijngaarden-Dekker-Brent method. Brent (1973)\n",
      "        claims convergence is guaranteed for functions computable within [a,b].\n",
      "        \n",
      "        [Brent1973]_ provides the classic description of the algorithm. Another\n",
      "        description can be found in a recent edition of Numerical Recipes, including\n",
      "        [PressEtal1992]_. A third description is at\n",
      "        http://mathworld.wolfram.com/BrentsMethod.html. It should be easy to\n",
      "        understand the algorithm just by reading our code. Our code diverges a bit\n",
      "        from standard presentations: we choose a different formula for the\n",
      "        extrapolation step.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        f : function\n",
      "            Python function returning a number. The function :math:`f`\n",
      "            must be continuous, and :math:`f(a)` and :math:`f(b)` must\n",
      "            have opposite signs.\n",
      "        a : scalar\n",
      "            One end of the bracketing interval :math:`[a, b]`.\n",
      "        b : scalar\n",
      "            The other end of the bracketing interval :math:`[a, b]`.\n",
      "        xtol : number, optional\n",
      "            The computed root ``x0`` will satisfy ``np.allclose(x, x0,\n",
      "            atol=xtol, rtol=rtol)``, where ``x`` is the exact root. The\n",
      "            parameter must be nonnegative. For nice functions, Brent's\n",
      "            method will often satisfy the above condition with ``xtol/2``\n",
      "            and ``rtol/2``. [Brent1973]_\n",
      "        rtol : number, optional\n",
      "            The computed root ``x0`` will satisfy ``np.allclose(x, x0,\n",
      "            atol=xtol, rtol=rtol)``, where ``x`` is the exact root. The\n",
      "            parameter cannot be smaller than its default value of\n",
      "            ``4*np.finfo(float).eps``. For nice functions, Brent's\n",
      "            method will often satisfy the above condition with ``xtol/2``\n",
      "            and ``rtol/2``. [Brent1973]_\n",
      "        maxiter : int, optional\n",
      "            If convergence is not achieved in `maxiter` iterations, an error is\n",
      "            raised. Must be >= 0.\n",
      "        args : tuple, optional\n",
      "            Containing extra arguments for the function `f`.\n",
      "            `f` is called by ``apply(f, (x)+args)``.\n",
      "        full_output : bool, optional\n",
      "            If `full_output` is False, the root is returned. If `full_output` is\n",
      "            True, the return value is ``(x, r)``, where `x` is the root, and `r` is\n",
      "            a `RootResults` object.\n",
      "        disp : bool, optional\n",
      "            If True, raise RuntimeError if the algorithm didn't converge.\n",
      "            Otherwise, the convergence status is recorded in any `RootResults`\n",
      "            return object.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        x0 : float\n",
      "            Zero of `f` between `a` and `b`.\n",
      "        r : `RootResults` (present if ``full_output = True``)\n",
      "            Object containing information about the convergence. In particular,\n",
      "            ``r.converged`` is True if the routine converged.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        `f` must be continuous.  f(a) and f(b) must have opposite signs.\n",
      "        \n",
      "        Related functions fall into several classes:\n",
      "        \n",
      "        multivariate local optimizers\n",
      "          `fmin`, `fmin_powell`, `fmin_cg`, `fmin_bfgs`, `fmin_ncg`\n",
      "        nonlinear least squares minimizer\n",
      "          `leastsq`\n",
      "        constrained multivariate optimizers\n",
      "          `fmin_l_bfgs_b`, `fmin_tnc`, `fmin_cobyla`\n",
      "        global optimizers\n",
      "          `basinhopping`, `brute`, `differential_evolution`\n",
      "        local scalar minimizers\n",
      "          `fminbound`, `brent`, `golden`, `bracket`\n",
      "        N-D root-finding\n",
      "          `fsolve`\n",
      "        1-D root-finding\n",
      "          `brenth`, `ridder`, `bisect`, `newton`\n",
      "        scalar fixed-point finder\n",
      "          `fixed_point`\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [Brent1973]\n",
      "           Brent, R. P.,\n",
      "           *Algorithms for Minimization Without Derivatives*.\n",
      "           Englewood Cliffs, NJ: Prentice-Hall, 1973. Ch. 3-4.\n",
      "        \n",
      "        .. [PressEtal1992]\n",
      "           Press, W. H.; Flannery, B. P.; Teukolsky, S. A.; and Vetterling, W. T.\n",
      "           *Numerical Recipes in FORTRAN: The Art of Scientific Computing*, 2nd ed.\n",
      "           Cambridge, England: Cambridge University Press, pp. 352-355, 1992.\n",
      "           Section 9.3:  \"Van Wijngaarden-Dekker-Brent Method.\"\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> def f(x):\n",
      "        ...     return (x**2 - 1)\n",
      "        \n",
      "        >>> from scipy import optimize\n",
      "        \n",
      "        >>> root = optimize.brentq(f, -2, 0)\n",
      "        >>> root\n",
      "        -1.0\n",
      "        \n",
      "        >>> root = optimize.brentq(f, 0, 2)\n",
      "        >>> root\n",
      "        1.0\n",
      "    \n",
      "    broyden1(F, xin, iter=None, alpha=None, reduction_method='restart', max_rank=None, verbose=False, maxiter=None, f_tol=None, f_rtol=None, x_tol=None, x_rtol=None, tol_norm=None, line_search='armijo', callback=None, **kw)\n",
      "        Find a root of a function, using Broyden's first Jacobian approximation.\n",
      "        \n",
      "        This method is also known as \\\"Broyden's good method\\\".\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        F : function(x) -> f\n",
      "            Function whose root to find; should take and return an array-like\n",
      "            object.\n",
      "        xin : array_like\n",
      "            Initial guess for the solution\n",
      "        alpha : float, optional\n",
      "            Initial guess for the Jacobian is ``(-1/alpha)``.\n",
      "        reduction_method : str or tuple, optional\n",
      "            Method used in ensuring that the rank of the Broyden matrix\n",
      "            stays low. Can either be a string giving the name of the method,\n",
      "            or a tuple of the form ``(method, param1, param2, ...)``\n",
      "            that gives the name of the method and values for additional parameters.\n",
      "        \n",
      "            Methods available:\n",
      "        \n",
      "                - ``restart``: drop all matrix columns. Has no extra parameters.\n",
      "                - ``simple``: drop oldest matrix column. Has no extra parameters.\n",
      "                - ``svd``: keep only the most significant SVD components.\n",
      "                  Takes an extra parameter, ``to_retain``, which determines the\n",
      "                  number of SVD components to retain when rank reduction is done.\n",
      "                  Default is ``max_rank - 2``.\n",
      "        \n",
      "        max_rank : int, optional\n",
      "            Maximum rank for the Broyden matrix.\n",
      "            Default is infinity (i.e., no rank reduction).\n",
      "        iter : int, optional\n",
      "            Number of iterations to make. If omitted (default), make as many\n",
      "            as required to meet tolerances.\n",
      "        verbose : bool, optional\n",
      "            Print status to stdout on every iteration.\n",
      "        maxiter : int, optional\n",
      "            Maximum number of iterations to make. If more are needed to\n",
      "            meet convergence, `NoConvergence` is raised.\n",
      "        f_tol : float, optional\n",
      "            Absolute tolerance (in max-norm) for the residual.\n",
      "            If omitted, default is 6e-6.\n",
      "        f_rtol : float, optional\n",
      "            Relative tolerance for the residual. If omitted, not used.\n",
      "        x_tol : float, optional\n",
      "            Absolute minimum step size, as determined from the Jacobian\n",
      "            approximation. If the step size is smaller than this, optimization\n",
      "            is terminated as successful. If omitted, not used.\n",
      "        x_rtol : float, optional\n",
      "            Relative minimum step size. If omitted, not used.\n",
      "        tol_norm : function(vector) -> scalar, optional\n",
      "            Norm to use in convergence check. Default is the maximum norm.\n",
      "        line_search : {None, 'armijo' (default), 'wolfe'}, optional\n",
      "            Which type of a line search to use to determine the step size in the\n",
      "            direction given by the Jacobian approximation. Defaults to 'armijo'.\n",
      "        callback : function, optional\n",
      "            Optional callback function. It is called on every iteration as\n",
      "            ``callback(x, f)`` where `x` is the current solution and `f`\n",
      "            the corresponding residual.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        sol : ndarray\n",
      "            An array (of similar array type as `x0`) containing the final solution.\n",
      "        \n",
      "        Raises\n",
      "        ------\n",
      "        NoConvergence\n",
      "            When a solution was not found.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        root : Interface to root finding algorithms for multivariate\n",
      "               functions. See ``method=='broyden1'`` in particular.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        This algorithm implements the inverse Jacobian Quasi-Newton update\n",
      "        \n",
      "        .. math:: H_+ = H + (dx - H df) dx^\\dagger H / ( dx^\\dagger H df)\n",
      "        \n",
      "        which corresponds to Broyden's first Jacobian update\n",
      "        \n",
      "        .. math:: J_+ = J + (df - J dx) dx^\\dagger / dx^\\dagger dx\n",
      "        \n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] B.A. van der Rotten, PhD thesis,\n",
      "           \\\"A limited memory Broyden method to solve high-dimensional\n",
      "           systems of nonlinear equations\\\". Mathematisch Instituut,\n",
      "           Universiteit Leiden, The Netherlands (2003).\n",
      "        \n",
      "           https://web.archive.org/web/20161022015821/http://www.math.leidenuniv.nl/scripties/Rotten.pdf\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        The following functions define a system of nonlinear equations\n",
      "        \n",
      "        >>> def fun(x):\n",
      "        ...     return [x[0]  + 0.5 * (x[0] - x[1])**3 - 1.0,\n",
      "        ...             0.5 * (x[1] - x[0])**3 + x[1]]\n",
      "        \n",
      "        A solution can be obtained as follows.\n",
      "        \n",
      "        >>> from scipy import optimize\n",
      "        >>> sol = optimize.broyden1(fun, [0, 0])\n",
      "        >>> sol\n",
      "        array([0.84116396, 0.15883641])\n",
      "    \n",
      "    broyden2(F, xin, iter=None, alpha=None, reduction_method='restart', max_rank=None, verbose=False, maxiter=None, f_tol=None, f_rtol=None, x_tol=None, x_rtol=None, tol_norm=None, line_search='armijo', callback=None, **kw)\n",
      "        Find a root of a function, using Broyden's second Jacobian approximation.\n",
      "        \n",
      "        This method is also known as \"Broyden's bad method\".\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        F : function(x) -> f\n",
      "            Function whose root to find; should take and return an array-like\n",
      "            object.\n",
      "        xin : array_like\n",
      "            Initial guess for the solution\n",
      "        alpha : float, optional\n",
      "            Initial guess for the Jacobian is ``(-1/alpha)``.\n",
      "        reduction_method : str or tuple, optional\n",
      "            Method used in ensuring that the rank of the Broyden matrix\n",
      "            stays low. Can either be a string giving the name of the method,\n",
      "            or a tuple of the form ``(method, param1, param2, ...)``\n",
      "            that gives the name of the method and values for additional parameters.\n",
      "        \n",
      "            Methods available:\n",
      "        \n",
      "                - ``restart``: drop all matrix columns. Has no extra parameters.\n",
      "                - ``simple``: drop oldest matrix column. Has no extra parameters.\n",
      "                - ``svd``: keep only the most significant SVD components.\n",
      "                  Takes an extra parameter, ``to_retain``, which determines the\n",
      "                  number of SVD components to retain when rank reduction is done.\n",
      "                  Default is ``max_rank - 2``.\n",
      "        \n",
      "        max_rank : int, optional\n",
      "            Maximum rank for the Broyden matrix.\n",
      "            Default is infinity (i.e., no rank reduction).\n",
      "        iter : int, optional\n",
      "            Number of iterations to make. If omitted (default), make as many\n",
      "            as required to meet tolerances.\n",
      "        verbose : bool, optional\n",
      "            Print status to stdout on every iteration.\n",
      "        maxiter : int, optional\n",
      "            Maximum number of iterations to make. If more are needed to\n",
      "            meet convergence, `NoConvergence` is raised.\n",
      "        f_tol : float, optional\n",
      "            Absolute tolerance (in max-norm) for the residual.\n",
      "            If omitted, default is 6e-6.\n",
      "        f_rtol : float, optional\n",
      "            Relative tolerance for the residual. If omitted, not used.\n",
      "        x_tol : float, optional\n",
      "            Absolute minimum step size, as determined from the Jacobian\n",
      "            approximation. If the step size is smaller than this, optimization\n",
      "            is terminated as successful. If omitted, not used.\n",
      "        x_rtol : float, optional\n",
      "            Relative minimum step size. If omitted, not used.\n",
      "        tol_norm : function(vector) -> scalar, optional\n",
      "            Norm to use in convergence check. Default is the maximum norm.\n",
      "        line_search : {None, 'armijo' (default), 'wolfe'}, optional\n",
      "            Which type of a line search to use to determine the step size in the\n",
      "            direction given by the Jacobian approximation. Defaults to 'armijo'.\n",
      "        callback : function, optional\n",
      "            Optional callback function. It is called on every iteration as\n",
      "            ``callback(x, f)`` where `x` is the current solution and `f`\n",
      "            the corresponding residual.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        sol : ndarray\n",
      "            An array (of similar array type as `x0`) containing the final solution.\n",
      "        \n",
      "        Raises\n",
      "        ------\n",
      "        NoConvergence\n",
      "            When a solution was not found.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        root : Interface to root finding algorithms for multivariate\n",
      "               functions. See ``method=='broyden2'`` in particular.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        This algorithm implements the inverse Jacobian Quasi-Newton update\n",
      "        \n",
      "        .. math:: H_+ = H + (dx - H df) df^\\dagger / ( df^\\dagger df)\n",
      "        \n",
      "        corresponding to Broyden's second method.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] B.A. van der Rotten, PhD thesis,\n",
      "           \"A limited memory Broyden method to solve high-dimensional\n",
      "           systems of nonlinear equations\". Mathematisch Instituut,\n",
      "           Universiteit Leiden, The Netherlands (2003).\n",
      "        \n",
      "           https://web.archive.org/web/20161022015821/http://www.math.leidenuniv.nl/scripties/Rotten.pdf\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        The following functions define a system of nonlinear equations\n",
      "        \n",
      "        >>> def fun(x):\n",
      "        ...     return [x[0]  + 0.5 * (x[0] - x[1])**3 - 1.0,\n",
      "        ...             0.5 * (x[1] - x[0])**3 + x[1]]\n",
      "        \n",
      "        A solution can be obtained as follows.\n",
      "        \n",
      "        >>> from scipy import optimize\n",
      "        >>> sol = optimize.broyden2(fun, [0, 0])\n",
      "        >>> sol\n",
      "        array([0.84116365, 0.15883529])\n",
      "    \n",
      "    brute(func, ranges, args=(), Ns=20, full_output=0, finish=<function fmin at 0x00000167FDFF3E50>, disp=False, workers=1)\n",
      "        Minimize a function over a given range by brute force.\n",
      "        \n",
      "        Uses the \"brute force\" method, i.e., computes the function's value\n",
      "        at each point of a multidimensional grid of points, to find the global\n",
      "        minimum of the function.\n",
      "        \n",
      "        The function is evaluated everywhere in the range with the datatype of the\n",
      "        first call to the function, as enforced by the ``vectorize`` NumPy\n",
      "        function. The value and type of the function evaluation returned when\n",
      "        ``full_output=True`` are affected in addition by the ``finish`` argument\n",
      "        (see Notes).\n",
      "        \n",
      "        The brute force approach is inefficient because the number of grid points\n",
      "        increases exponentially - the number of grid points to evaluate is\n",
      "        ``Ns ** len(x)``. Consequently, even with coarse grid spacing, even\n",
      "        moderately sized problems can take a long time to run, and/or run into\n",
      "        memory limitations.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        func : callable\n",
      "            The objective function to be minimized. Must be in the\n",
      "            form ``f(x, *args)``, where ``x`` is the argument in\n",
      "            the form of a 1-D array and ``args`` is a tuple of any\n",
      "            additional fixed parameters needed to completely specify\n",
      "            the function.\n",
      "        ranges : tuple\n",
      "            Each component of the `ranges` tuple must be either a\n",
      "            \"slice object\" or a range tuple of the form ``(low, high)``.\n",
      "            The program uses these to create the grid of points on which\n",
      "            the objective function will be computed. See `Note 2` for\n",
      "            more detail.\n",
      "        args : tuple, optional\n",
      "            Any additional fixed parameters needed to completely specify\n",
      "            the function.\n",
      "        Ns : int, optional\n",
      "            Number of grid points along the axes, if not otherwise\n",
      "            specified. See `Note2`.\n",
      "        full_output : bool, optional\n",
      "            If True, return the evaluation grid and the objective function's\n",
      "            values on it.\n",
      "        finish : callable, optional\n",
      "            An optimization function that is called with the result of brute force\n",
      "            minimization as initial guess. `finish` should take `func` and\n",
      "            the initial guess as positional arguments, and take `args` as\n",
      "            keyword arguments. It may additionally take `full_output`\n",
      "            and/or `disp` as keyword arguments. Use None if no \"polishing\"\n",
      "            function is to be used. See Notes for more details.\n",
      "        disp : bool, optional\n",
      "            Set to True to print convergence messages from the `finish` callable.\n",
      "        workers : int or map-like callable, optional\n",
      "            If `workers` is an int the grid is subdivided into `workers`\n",
      "            sections and evaluated in parallel (uses\n",
      "            `multiprocessing.Pool <multiprocessing>`).\n",
      "            Supply `-1` to use all cores available to the Process.\n",
      "            Alternatively supply a map-like callable, such as\n",
      "            `multiprocessing.Pool.map` for evaluating the grid in parallel.\n",
      "            This evaluation is carried out as ``workers(func, iterable)``.\n",
      "            Requires that `func` be pickleable.\n",
      "        \n",
      "            .. versionadded:: 1.3.0\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        x0 : ndarray\n",
      "            A 1-D array containing the coordinates of a point at which the\n",
      "            objective function had its minimum value. (See `Note 1` for\n",
      "            which point is returned.)\n",
      "        fval : float\n",
      "            Function value at the point `x0`. (Returned when `full_output` is\n",
      "            True.)\n",
      "        grid : tuple\n",
      "            Representation of the evaluation grid. It has the same\n",
      "            length as `x0`. (Returned when `full_output` is True.)\n",
      "        Jout : ndarray\n",
      "            Function values at each point of the evaluation\n",
      "            grid, i.e., ``Jout = func(*grid)``. (Returned\n",
      "            when `full_output` is True.)\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        basinhopping, differential_evolution\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        *Note 1*: The program finds the gridpoint at which the lowest value\n",
      "        of the objective function occurs. If `finish` is None, that is the\n",
      "        point returned. When the global minimum occurs within (or not very far\n",
      "        outside) the grid's boundaries, and the grid is fine enough, that\n",
      "        point will be in the neighborhood of the global minimum.\n",
      "        \n",
      "        However, users often employ some other optimization program to\n",
      "        \"polish\" the gridpoint values, i.e., to seek a more precise\n",
      "        (local) minimum near `brute's` best gridpoint.\n",
      "        The `brute` function's `finish` option provides a convenient way to do\n",
      "        that. Any polishing program used must take `brute's` output as its\n",
      "        initial guess as a positional argument, and take `brute's` input values\n",
      "        for `args` as keyword arguments, otherwise an error will be raised.\n",
      "        It may additionally take `full_output` and/or `disp` as keyword arguments.\n",
      "        \n",
      "        `brute` assumes that the `finish` function returns either an\n",
      "        `OptimizeResult` object or a tuple in the form:\n",
      "        ``(xmin, Jmin, ... , statuscode)``, where ``xmin`` is the minimizing\n",
      "        value of the argument, ``Jmin`` is the minimum value of the objective\n",
      "        function, \"...\" may be some other returned values (which are not used\n",
      "        by `brute`), and ``statuscode`` is the status code of the `finish` program.\n",
      "        \n",
      "        Note that when `finish` is not None, the values returned are those\n",
      "        of the `finish` program, *not* the gridpoint ones. Consequently,\n",
      "        while `brute` confines its search to the input grid points,\n",
      "        the `finish` program's results usually will not coincide with any\n",
      "        gridpoint, and may fall outside the grid's boundary. Thus, if a\n",
      "        minimum only needs to be found over the provided grid points, make\n",
      "        sure to pass in `finish=None`.\n",
      "        \n",
      "        *Note 2*: The grid of points is a `numpy.mgrid` object.\n",
      "        For `brute` the `ranges` and `Ns` inputs have the following effect.\n",
      "        Each component of the `ranges` tuple can be either a slice object or a\n",
      "        two-tuple giving a range of values, such as (0, 5). If the component is a\n",
      "        slice object, `brute` uses it directly. If the component is a two-tuple\n",
      "        range, `brute` internally converts it to a slice object that interpolates\n",
      "        `Ns` points from its low-value to its high-value, inclusive.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        We illustrate the use of `brute` to seek the global minimum of a function\n",
      "        of two variables that is given as the sum of a positive-definite\n",
      "        quadratic and two deep \"Gaussian-shaped\" craters. Specifically, define\n",
      "        the objective function `f` as the sum of three other functions,\n",
      "        ``f = f1 + f2 + f3``. We suppose each of these has a signature\n",
      "        ``(z, *params)``, where ``z = (x, y)``,  and ``params`` and the functions\n",
      "        are as defined below.\n",
      "        \n",
      "        >>> params = (2, 3, 7, 8, 9, 10, 44, -1, 2, 26, 1, -2, 0.5)\n",
      "        >>> def f1(z, *params):\n",
      "        ...     x, y = z\n",
      "        ...     a, b, c, d, e, f, g, h, i, j, k, l, scale = params\n",
      "        ...     return (a * x**2 + b * x * y + c * y**2 + d*x + e*y + f)\n",
      "        \n",
      "        >>> def f2(z, *params):\n",
      "        ...     x, y = z\n",
      "        ...     a, b, c, d, e, f, g, h, i, j, k, l, scale = params\n",
      "        ...     return (-g*np.exp(-((x-h)**2 + (y-i)**2) / scale))\n",
      "        \n",
      "        >>> def f3(z, *params):\n",
      "        ...     x, y = z\n",
      "        ...     a, b, c, d, e, f, g, h, i, j, k, l, scale = params\n",
      "        ...     return (-j*np.exp(-((x-k)**2 + (y-l)**2) / scale))\n",
      "        \n",
      "        >>> def f(z, *params):\n",
      "        ...     return f1(z, *params) + f2(z, *params) + f3(z, *params)\n",
      "        \n",
      "        Thus, the objective function may have local minima near the minimum\n",
      "        of each of the three functions of which it is composed. To\n",
      "        use `fmin` to polish its gridpoint result, we may then continue as\n",
      "        follows:\n",
      "        \n",
      "        >>> rranges = (slice(-4, 4, 0.25), slice(-4, 4, 0.25))\n",
      "        >>> from scipy import optimize\n",
      "        >>> resbrute = optimize.brute(f, rranges, args=params, full_output=True,\n",
      "        ...                           finish=optimize.fmin)\n",
      "        >>> resbrute[0]  # global minimum\n",
      "        array([-1.05665192,  1.80834843])\n",
      "        >>> resbrute[1]  # function value at global minimum\n",
      "        -3.4085818767\n",
      "        \n",
      "        Note that if `finish` had been set to None, we would have gotten the\n",
      "        gridpoint [-1.0 1.75] where the rounded function value is -2.892.\n",
      "    \n",
      "    check_grad(func, grad, x0, *args, **kwargs)\n",
      "        Check the correctness of a gradient function by comparing it against a\n",
      "        (forward) finite-difference approximation of the gradient.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        func : callable ``func(x0, *args)``\n",
      "            Function whose derivative is to be checked.\n",
      "        grad : callable ``grad(x0, *args)``\n",
      "            Gradient of `func`.\n",
      "        x0 : ndarray\n",
      "            Points to check `grad` against forward difference approximation of grad\n",
      "            using `func`.\n",
      "        args : \\*args, optional\n",
      "            Extra arguments passed to `func` and `grad`.\n",
      "        epsilon : float, optional\n",
      "            Step size used for the finite difference approximation. It defaults to\n",
      "            ``sqrt(np.finfo(float).eps)``, which is approximately 1.49e-08.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        err : float\n",
      "            The square root of the sum of squares (i.e., the 2-norm) of the\n",
      "            difference between ``grad(x0, *args)`` and the finite difference\n",
      "            approximation of `grad` using func at the points `x0`.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        approx_fprime\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> def func(x):\n",
      "        ...     return x[0]**2 - 0.5 * x[1]**3\n",
      "        >>> def grad(x):\n",
      "        ...     return [2 * x[0], -1.5 * x[1]**2]\n",
      "        >>> from scipy.optimize import check_grad\n",
      "        >>> check_grad(func, grad, [1.5, -1.5])\n",
      "        2.9802322387695312e-08\n",
      "    \n",
      "    curve_fit(f, xdata, ydata, p0=None, sigma=None, absolute_sigma=False, check_finite=True, bounds=(-inf, inf), method=None, jac=None, **kwargs)\n",
      "        Use non-linear least squares to fit a function, f, to data.\n",
      "        \n",
      "        Assumes ``ydata = f(xdata, *params) + eps``.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        f : callable\n",
      "            The model function, f(x, ...). It must take the independent\n",
      "            variable as the first argument and the parameters to fit as\n",
      "            separate remaining arguments.\n",
      "        xdata : array_like or object\n",
      "            The independent variable where the data is measured.\n",
      "            Should usually be an M-length sequence or an (k,M)-shaped array for\n",
      "            functions with k predictors, but can actually be any object.\n",
      "        ydata : array_like\n",
      "            The dependent data, a length M array - nominally ``f(xdata, ...)``.\n",
      "        p0 : array_like, optional\n",
      "            Initial guess for the parameters (length N). If None, then the\n",
      "            initial values will all be 1 (if the number of parameters for the\n",
      "            function can be determined using introspection, otherwise a\n",
      "            ValueError is raised).\n",
      "        sigma : None or M-length sequence or MxM array, optional\n",
      "            Determines the uncertainty in `ydata`. If we define residuals as\n",
      "            ``r = ydata - f(xdata, *popt)``, then the interpretation of `sigma`\n",
      "            depends on its number of dimensions:\n",
      "        \n",
      "                - A 1-D `sigma` should contain values of standard deviations of\n",
      "                  errors in `ydata`. In this case, the optimized function is\n",
      "                  ``chisq = sum((r / sigma) ** 2)``.\n",
      "        \n",
      "                - A 2-D `sigma` should contain the covariance matrix of\n",
      "                  errors in `ydata`. In this case, the optimized function is\n",
      "                  ``chisq = r.T @ inv(sigma) @ r``.\n",
      "        \n",
      "                  .. versionadded:: 0.19\n",
      "        \n",
      "            None (default) is equivalent of 1-D `sigma` filled with ones.\n",
      "        absolute_sigma : bool, optional\n",
      "            If True, `sigma` is used in an absolute sense and the estimated parameter\n",
      "            covariance `pcov` reflects these absolute values.\n",
      "        \n",
      "            If False (default), only the relative magnitudes of the `sigma` values matter.\n",
      "            The returned parameter covariance matrix `pcov` is based on scaling\n",
      "            `sigma` by a constant factor. This constant is set by demanding that the\n",
      "            reduced `chisq` for the optimal parameters `popt` when using the\n",
      "            *scaled* `sigma` equals unity. In other words, `sigma` is scaled to\n",
      "            match the sample variance of the residuals after the fit. Default is False.\n",
      "            Mathematically,\n",
      "            ``pcov(absolute_sigma=False) = pcov(absolute_sigma=True) * chisq(popt)/(M-N)``\n",
      "        check_finite : bool, optional\n",
      "            If True, check that the input arrays do not contain nans of infs,\n",
      "            and raise a ValueError if they do. Setting this parameter to\n",
      "            False may silently produce nonsensical results if the input arrays\n",
      "            do contain nans. Default is True.\n",
      "        bounds : 2-tuple of array_like, optional\n",
      "            Lower and upper bounds on parameters. Defaults to no bounds.\n",
      "            Each element of the tuple must be either an array with the length equal\n",
      "            to the number of parameters, or a scalar (in which case the bound is\n",
      "            taken to be the same for all parameters). Use ``np.inf`` with an\n",
      "            appropriate sign to disable bounds on all or some parameters.\n",
      "        \n",
      "            .. versionadded:: 0.17\n",
      "        method : {'lm', 'trf', 'dogbox'}, optional\n",
      "            Method to use for optimization. See `least_squares` for more details.\n",
      "            Default is 'lm' for unconstrained problems and 'trf' if `bounds` are\n",
      "            provided. The method 'lm' won't work when the number of observations\n",
      "            is less than the number of variables, use 'trf' or 'dogbox' in this\n",
      "            case.\n",
      "        \n",
      "            .. versionadded:: 0.17\n",
      "        jac : callable, string or None, optional\n",
      "            Function with signature ``jac(x, ...)`` which computes the Jacobian\n",
      "            matrix of the model function with respect to parameters as a dense\n",
      "            array_like structure. It will be scaled according to provided `sigma`.\n",
      "            If None (default), the Jacobian will be estimated numerically.\n",
      "            String keywords for 'trf' and 'dogbox' methods can be used to select\n",
      "            a finite difference scheme, see `least_squares`.\n",
      "        \n",
      "            .. versionadded:: 0.18\n",
      "        kwargs\n",
      "            Keyword arguments passed to `leastsq` for ``method='lm'`` or\n",
      "            `least_squares` otherwise.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        popt : array\n",
      "            Optimal values for the parameters so that the sum of the squared\n",
      "            residuals of ``f(xdata, *popt) - ydata`` is minimized.\n",
      "        pcov : 2-D array\n",
      "            The estimated covariance of popt. The diagonals provide the variance\n",
      "            of the parameter estimate. To compute one standard deviation errors\n",
      "            on the parameters use ``perr = np.sqrt(np.diag(pcov))``.\n",
      "        \n",
      "            How the `sigma` parameter affects the estimated covariance\n",
      "            depends on `absolute_sigma` argument, as described above.\n",
      "        \n",
      "            If the Jacobian matrix at the solution doesn't have a full rank, then\n",
      "            'lm' method returns a matrix filled with ``np.inf``, on the other hand\n",
      "            'trf'  and 'dogbox' methods use Moore-Penrose pseudoinverse to compute\n",
      "            the covariance matrix.\n",
      "        \n",
      "        Raises\n",
      "        ------\n",
      "        ValueError\n",
      "            if either `ydata` or `xdata` contain NaNs, or if incompatible options\n",
      "            are used.\n",
      "        \n",
      "        RuntimeError\n",
      "            if the least-squares minimization fails.\n",
      "        \n",
      "        OptimizeWarning\n",
      "            if covariance of the parameters can not be estimated.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        least_squares : Minimize the sum of squares of nonlinear functions.\n",
      "        scipy.stats.linregress : Calculate a linear least squares regression for\n",
      "                                 two sets of measurements.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        With ``method='lm'``, the algorithm uses the Levenberg-Marquardt algorithm\n",
      "        through `leastsq`. Note that this algorithm can only deal with\n",
      "        unconstrained problems.\n",
      "        \n",
      "        Box constraints can be handled by methods 'trf' and 'dogbox'. Refer to\n",
      "        the docstring of `least_squares` for more information.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> import matplotlib.pyplot as plt\n",
      "        >>> from scipy.optimize import curve_fit\n",
      "        \n",
      "        >>> def func(x, a, b, c):\n",
      "        ...     return a * np.exp(-b * x) + c\n",
      "        \n",
      "        Define the data to be fit with some noise:\n",
      "        \n",
      "        >>> xdata = np.linspace(0, 4, 50)\n",
      "        >>> y = func(xdata, 2.5, 1.3, 0.5)\n",
      "        >>> rng = np.random.default_rng()\n",
      "        >>> y_noise = 0.2 * rng.normal(size=xdata.size)\n",
      "        >>> ydata = y + y_noise\n",
      "        >>> plt.plot(xdata, ydata, 'b-', label='data')\n",
      "        \n",
      "        Fit for the parameters a, b, c of the function `func`:\n",
      "        \n",
      "        >>> popt, pcov = curve_fit(func, xdata, ydata)\n",
      "        >>> popt\n",
      "        array([2.56274217, 1.37268521, 0.47427475])\n",
      "        >>> plt.plot(xdata, func(xdata, *popt), 'r-',\n",
      "        ...          label='fit: a=%5.3f, b=%5.3f, c=%5.3f' % tuple(popt))\n",
      "        \n",
      "        Constrain the optimization to the region of ``0 <= a <= 3``,\n",
      "        ``0 <= b <= 1`` and ``0 <= c <= 0.5``:\n",
      "        \n",
      "        >>> popt, pcov = curve_fit(func, xdata, ydata, bounds=(0, [3., 1., 0.5]))\n",
      "        >>> popt\n",
      "        array([2.43736712, 1.        , 0.34463856])\n",
      "        >>> plt.plot(xdata, func(xdata, *popt), 'g--',\n",
      "        ...          label='fit: a=%5.3f, b=%5.3f, c=%5.3f' % tuple(popt))\n",
      "        \n",
      "        >>> plt.xlabel('x')\n",
      "        >>> plt.ylabel('y')\n",
      "        >>> plt.legend()\n",
      "        >>> plt.show()\n",
      "    \n",
      "    diagbroyden(F, xin, iter=None, alpha=None, verbose=False, maxiter=None, f_tol=None, f_rtol=None, x_tol=None, x_rtol=None, tol_norm=None, line_search='armijo', callback=None, **kw)\n",
      "        Find a root of a function, using diagonal Broyden Jacobian approximation.\n",
      "        \n",
      "        The Jacobian approximation is derived from previous iterations, by\n",
      "        retaining only the diagonal of Broyden matrices.\n",
      "        \n",
      "        .. warning::\n",
      "        \n",
      "           This algorithm may be useful for specific problems, but whether\n",
      "           it will work may depend strongly on the problem.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        F : function(x) -> f\n",
      "            Function whose root to find; should take and return an array-like\n",
      "            object.\n",
      "        xin : array_like\n",
      "            Initial guess for the solution\n",
      "        alpha : float, optional\n",
      "            Initial guess for the Jacobian is (-1/alpha).\n",
      "        iter : int, optional\n",
      "            Number of iterations to make. If omitted (default), make as many\n",
      "            as required to meet tolerances.\n",
      "        verbose : bool, optional\n",
      "            Print status to stdout on every iteration.\n",
      "        maxiter : int, optional\n",
      "            Maximum number of iterations to make. If more are needed to\n",
      "            meet convergence, `NoConvergence` is raised.\n",
      "        f_tol : float, optional\n",
      "            Absolute tolerance (in max-norm) for the residual.\n",
      "            If omitted, default is 6e-6.\n",
      "        f_rtol : float, optional\n",
      "            Relative tolerance for the residual. If omitted, not used.\n",
      "        x_tol : float, optional\n",
      "            Absolute minimum step size, as determined from the Jacobian\n",
      "            approximation. If the step size is smaller than this, optimization\n",
      "            is terminated as successful. If omitted, not used.\n",
      "        x_rtol : float, optional\n",
      "            Relative minimum step size. If omitted, not used.\n",
      "        tol_norm : function(vector) -> scalar, optional\n",
      "            Norm to use in convergence check. Default is the maximum norm.\n",
      "        line_search : {None, 'armijo' (default), 'wolfe'}, optional\n",
      "            Which type of a line search to use to determine the step size in the\n",
      "            direction given by the Jacobian approximation. Defaults to 'armijo'.\n",
      "        callback : function, optional\n",
      "            Optional callback function. It is called on every iteration as\n",
      "            ``callback(x, f)`` where `x` is the current solution and `f`\n",
      "            the corresponding residual.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        sol : ndarray\n",
      "            An array (of similar array type as `x0`) containing the final solution.\n",
      "        \n",
      "        Raises\n",
      "        ------\n",
      "        NoConvergence\n",
      "            When a solution was not found.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        root : Interface to root finding algorithms for multivariate\n",
      "               functions. See ``method=='diagbroyden'`` in particular.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        The following functions define a system of nonlinear equations\n",
      "        \n",
      "        >>> def fun(x):\n",
      "        ...     return [x[0]  + 0.5 * (x[0] - x[1])**3 - 1.0,\n",
      "        ...             0.5 * (x[1] - x[0])**3 + x[1]]\n",
      "        \n",
      "        A solution can be obtained as follows.\n",
      "        \n",
      "        >>> from scipy import optimize\n",
      "        >>> sol = optimize.diagbroyden(fun, [0, 0])\n",
      "        >>> sol\n",
      "        array([0.84116403, 0.15883384])\n",
      "    \n",
      "    differential_evolution(func, bounds, args=(), strategy='best1bin', maxiter=1000, popsize=15, tol=0.01, mutation=(0.5, 1), recombination=0.7, seed=None, callback=None, disp=False, polish=True, init='latinhypercube', atol=0, updating='immediate', workers=1, constraints=(), x0=None)\n",
      "        Finds the global minimum of a multivariate function.\n",
      "        \n",
      "        Differential Evolution is stochastic in nature (does not use gradient\n",
      "        methods) to find the minimum, and can search large areas of candidate\n",
      "        space, but often requires larger numbers of function evaluations than\n",
      "        conventional gradient-based techniques.\n",
      "        \n",
      "        The algorithm is due to Storn and Price [1]_.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        func : callable\n",
      "            The objective function to be minimized. Must be in the form\n",
      "            ``f(x, *args)``, where ``x`` is the argument in the form of a 1-D array\n",
      "            and ``args`` is a  tuple of any additional fixed parameters needed to\n",
      "            completely specify the function.\n",
      "        bounds : sequence or `Bounds`\n",
      "            Bounds for variables. There are two ways to specify the bounds:\n",
      "            1. Instance of `Bounds` class.\n",
      "            2. ``(min, max)`` pairs for each element in ``x``, defining the finite\n",
      "            lower and upper bounds for the optimizing argument of `func`. It is\n",
      "            required to have ``len(bounds) == len(x)``. ``len(bounds)`` is used\n",
      "            to determine the number of parameters in ``x``.\n",
      "        args : tuple, optional\n",
      "            Any additional fixed parameters needed to\n",
      "            completely specify the objective function.\n",
      "        strategy : str, optional\n",
      "            The differential evolution strategy to use. Should be one of:\n",
      "        \n",
      "                - 'best1bin'\n",
      "                - 'best1exp'\n",
      "                - 'rand1exp'\n",
      "                - 'randtobest1exp'\n",
      "                - 'currenttobest1exp'\n",
      "                - 'best2exp'\n",
      "                - 'rand2exp'\n",
      "                - 'randtobest1bin'\n",
      "                - 'currenttobest1bin'\n",
      "                - 'best2bin'\n",
      "                - 'rand2bin'\n",
      "                - 'rand1bin'\n",
      "        \n",
      "            The default is 'best1bin'.\n",
      "        maxiter : int, optional\n",
      "            The maximum number of generations over which the entire population is\n",
      "            evolved. The maximum number of function evaluations (with no polishing)\n",
      "            is: ``(maxiter + 1) * popsize * len(x)``\n",
      "        popsize : int, optional\n",
      "            A multiplier for setting the total population size. The population has\n",
      "            ``popsize * len(x)`` individuals. This keyword is overridden if an\n",
      "            initial population is supplied via the `init` keyword. When using\n",
      "            ``init='sobol'`` the population size is calculated as the next power\n",
      "            of 2 after ``popsize * len(x)``.\n",
      "        tol : float, optional\n",
      "            Relative tolerance for convergence, the solving stops when\n",
      "            ``np.std(pop) <= atol + tol * np.abs(np.mean(population_energies))``,\n",
      "            where and `atol` and `tol` are the absolute and relative tolerance\n",
      "            respectively.\n",
      "        mutation : float or tuple(float, float), optional\n",
      "            The mutation constant. In the literature this is also known as\n",
      "            differential weight, being denoted by F.\n",
      "            If specified as a float it should be in the range [0, 2].\n",
      "            If specified as a tuple ``(min, max)`` dithering is employed. Dithering\n",
      "            randomly changes the mutation constant on a generation by generation\n",
      "            basis. The mutation constant for that generation is taken from\n",
      "            ``U[min, max)``. Dithering can help speed convergence significantly.\n",
      "            Increasing the mutation constant increases the search radius, but will\n",
      "            slow down convergence.\n",
      "        recombination : float, optional\n",
      "            The recombination constant, should be in the range [0, 1]. In the\n",
      "            literature this is also known as the crossover probability, being\n",
      "            denoted by CR. Increasing this value allows a larger number of mutants\n",
      "            to progress into the next generation, but at the risk of population\n",
      "            stability.\n",
      "        seed : {None, int, `numpy.random.Generator`,\n",
      "                `numpy.random.RandomState`}, optional\n",
      "        \n",
      "            If `seed` is None (or `np.random`), the `numpy.random.RandomState`\n",
      "            singleton is used.\n",
      "            If `seed` is an int, a new ``RandomState`` instance is used,\n",
      "            seeded with `seed`.\n",
      "            If `seed` is already a ``Generator`` or ``RandomState`` instance then\n",
      "            that instance is used.\n",
      "            Specify `seed` for repeatable minimizations.\n",
      "        disp : bool, optional\n",
      "            Prints the evaluated `func` at every iteration.\n",
      "        callback : callable, `callback(xk, convergence=val)`, optional\n",
      "            A function to follow the progress of the minimization. ``xk`` is\n",
      "            the current value of ``x0``. ``val`` represents the fractional\n",
      "            value of the population convergence.  When ``val`` is greater than one\n",
      "            the function halts. If callback returns `True`, then the minimization\n",
      "            is halted (any polishing is still carried out).\n",
      "        polish : bool, optional\n",
      "            If True (default), then `scipy.optimize.minimize` with the `L-BFGS-B`\n",
      "            method is used to polish the best population member at the end, which\n",
      "            can improve the minimization slightly. If a constrained problem is\n",
      "            being studied then the `trust-constr` method is used instead.\n",
      "        init : str or array-like, optional\n",
      "            Specify which type of population initialization is performed. Should be\n",
      "            one of:\n",
      "        \n",
      "                - 'latinhypercube'\n",
      "                - 'sobol'\n",
      "                - 'halton'\n",
      "                - 'random'\n",
      "                - array specifying the initial population. The array should have\n",
      "                  shape ``(M, len(x))``, where M is the total population size and\n",
      "                  len(x) is the number of parameters.\n",
      "                  `init` is clipped to `bounds` before use.\n",
      "        \n",
      "            The default is 'latinhypercube'. Latin Hypercube sampling tries to\n",
      "            maximize coverage of the available parameter space.\n",
      "        \n",
      "            'sobol' and 'halton' are superior alternatives and maximize even more\n",
      "            the parameter space. 'sobol' will enforce an initial population\n",
      "            size which is calculated as the next power of 2 after\n",
      "            ``popsize * len(x)``. 'halton' has no requirements but is a bit less\n",
      "            efficient. See `scipy.stats.qmc` for more details.\n",
      "        \n",
      "            'random' initializes the population randomly - this has the drawback\n",
      "            that clustering can occur, preventing the whole of parameter space\n",
      "            being covered. Use of an array to specify a population could be used,\n",
      "            for example, to create a tight bunch of initial guesses in an location\n",
      "            where the solution is known to exist, thereby reducing time for\n",
      "            convergence.\n",
      "        atol : float, optional\n",
      "            Absolute tolerance for convergence, the solving stops when\n",
      "            ``np.std(pop) <= atol + tol * np.abs(np.mean(population_energies))``,\n",
      "            where and `atol` and `tol` are the absolute and relative tolerance\n",
      "            respectively.\n",
      "        updating : {'immediate', 'deferred'}, optional\n",
      "            If ``'immediate'``, the best solution vector is continuously updated\n",
      "            within a single generation [4]_. This can lead to faster convergence as\n",
      "            trial vectors can take advantage of continuous improvements in the best\n",
      "            solution.\n",
      "            With ``'deferred'``, the best solution vector is updated once per\n",
      "            generation. Only ``'deferred'`` is compatible with parallelization, and\n",
      "            the `workers` keyword can over-ride this option.\n",
      "        \n",
      "            .. versionadded:: 1.2.0\n",
      "        \n",
      "        workers : int or map-like callable, optional\n",
      "            If `workers` is an int the population is subdivided into `workers`\n",
      "            sections and evaluated in parallel\n",
      "            (uses `multiprocessing.Pool <multiprocessing>`).\n",
      "            Supply -1 to use all available CPU cores.\n",
      "            Alternatively supply a map-like callable, such as\n",
      "            `multiprocessing.Pool.map` for evaluating the population in parallel.\n",
      "            This evaluation is carried out as ``workers(func, iterable)``.\n",
      "            This option will override the `updating` keyword to\n",
      "            ``updating='deferred'`` if ``workers != 1``.\n",
      "            Requires that `func` be pickleable.\n",
      "        \n",
      "            .. versionadded:: 1.2.0\n",
      "        \n",
      "        constraints : {NonLinearConstraint, LinearConstraint, Bounds}\n",
      "            Constraints on the solver, over and above those applied by the `bounds`\n",
      "            kwd. Uses the approach by Lampinen [5]_.\n",
      "        \n",
      "            .. versionadded:: 1.4.0\n",
      "        \n",
      "        x0 : None or array-like, optional\n",
      "            Provides an initial guess to the minimization. Once the population has\n",
      "            been initialized this vector replaces the first (best) member. This\n",
      "            replacement is done even if `init` is given an initial population.\n",
      "        \n",
      "            .. versionadded:: 1.7.0\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        res : OptimizeResult\n",
      "            The optimization result represented as a `OptimizeResult` object.\n",
      "            Important attributes are: ``x`` the solution array, ``success`` a\n",
      "            Boolean flag indicating if the optimizer exited successfully and\n",
      "            ``message`` which describes the cause of the termination. See\n",
      "            `OptimizeResult` for a description of other attributes. If `polish`\n",
      "            was employed, and a lower minimum was obtained by the polishing, then\n",
      "            OptimizeResult also contains the ``jac`` attribute.\n",
      "            If the eventual solution does not satisfy the applied constraints\n",
      "            ``success`` will be `False`.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        Differential evolution is a stochastic population based method that is\n",
      "        useful for global optimization problems. At each pass through the population\n",
      "        the algorithm mutates each candidate solution by mixing with other candidate\n",
      "        solutions to create a trial candidate. There are several strategies [2]_ for\n",
      "        creating trial candidates, which suit some problems more than others. The\n",
      "        'best1bin' strategy is a good starting point for many systems. In this\n",
      "        strategy two members of the population are randomly chosen. Their difference\n",
      "        is used to mutate the best member (the 'best' in 'best1bin'), :math:`b_0`,\n",
      "        so far:\n",
      "        \n",
      "        .. math::\n",
      "        \n",
      "            b' = b_0 + mutation * (population[rand0] - population[rand1])\n",
      "        \n",
      "        A trial vector is then constructed. Starting with a randomly chosen ith\n",
      "        parameter the trial is sequentially filled (in modulo) with parameters from\n",
      "        ``b'`` or the original candidate. The choice of whether to use ``b'`` or the\n",
      "        original candidate is made with a binomial distribution (the 'bin' in\n",
      "        'best1bin') - a random number in [0, 1) is generated. If this number is\n",
      "        less than the `recombination` constant then the parameter is loaded from\n",
      "        ``b'``, otherwise it is loaded from the original candidate. The final\n",
      "        parameter is always loaded from ``b'``. Once the trial candidate is built\n",
      "        its fitness is assessed. If the trial is better than the original candidate\n",
      "        then it takes its place. If it is also better than the best overall\n",
      "        candidate it also replaces that.\n",
      "        To improve your chances of finding a global minimum use higher `popsize`\n",
      "        values, with higher `mutation` and (dithering), but lower `recombination`\n",
      "        values. This has the effect of widening the search radius, but slowing\n",
      "        convergence.\n",
      "        By default the best solution vector is updated continuously within a single\n",
      "        iteration (``updating='immediate'``). This is a modification [4]_ of the\n",
      "        original differential evolution algorithm which can lead to faster\n",
      "        convergence as trial vectors can immediately benefit from improved\n",
      "        solutions. To use the original Storn and Price behaviour, updating the best\n",
      "        solution once per iteration, set ``updating='deferred'``.\n",
      "        \n",
      "        .. versionadded:: 0.15.0\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        Let us consider the problem of minimizing the Rosenbrock function. This\n",
      "        function is implemented in `rosen` in `scipy.optimize`.\n",
      "        \n",
      "        >>> from scipy.optimize import rosen, differential_evolution\n",
      "        >>> bounds = [(0,2), (0, 2), (0, 2), (0, 2), (0, 2)]\n",
      "        >>> result = differential_evolution(rosen, bounds)\n",
      "        >>> result.x, result.fun\n",
      "        (array([1., 1., 1., 1., 1.]), 1.9216496320061384e-19)\n",
      "        \n",
      "        Now repeat, but with parallelization.\n",
      "        \n",
      "        >>> bounds = [(0,2), (0, 2), (0, 2), (0, 2), (0, 2)]\n",
      "        >>> result = differential_evolution(rosen, bounds, updating='deferred',\n",
      "        ...                                 workers=2)\n",
      "        >>> result.x, result.fun\n",
      "        (array([1., 1., 1., 1., 1.]), 1.9216496320061384e-19)\n",
      "        \n",
      "        Let's try and do a constrained minimization\n",
      "        \n",
      "        >>> from scipy.optimize import NonlinearConstraint, Bounds\n",
      "        >>> def constr_f(x):\n",
      "        ...     return np.array(x[0] + x[1])\n",
      "        >>>\n",
      "        >>> # the sum of x[0] and x[1] must be less than 1.9\n",
      "        >>> nlc = NonlinearConstraint(constr_f, -np.inf, 1.9)\n",
      "        >>> # specify limits using a `Bounds` object.\n",
      "        >>> bounds = Bounds([0., 0.], [2., 2.])\n",
      "        >>> result = differential_evolution(rosen, bounds, constraints=(nlc),\n",
      "        ...                                 seed=1)\n",
      "        >>> result.x, result.fun\n",
      "        (array([0.96633867, 0.93363577]), 0.0011361355854792312)\n",
      "        \n",
      "        Next find the minimum of the Ackley function\n",
      "        (https://en.wikipedia.org/wiki/Test_functions_for_optimization).\n",
      "        \n",
      "        >>> from scipy.optimize import differential_evolution\n",
      "        >>> import numpy as np\n",
      "        >>> def ackley(x):\n",
      "        ...     arg1 = -0.2 * np.sqrt(0.5 * (x[0] ** 2 + x[1] ** 2))\n",
      "        ...     arg2 = 0.5 * (np.cos(2. * np.pi * x[0]) + np.cos(2. * np.pi * x[1]))\n",
      "        ...     return -20. * np.exp(arg1) - np.exp(arg2) + 20. + np.e\n",
      "        >>> bounds = [(-5, 5), (-5, 5)]\n",
      "        >>> result = differential_evolution(ackley, bounds)\n",
      "        >>> result.x, result.fun\n",
      "        (array([ 0.,  0.]), 4.4408920985006262e-16)\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] Storn, R and Price, K, Differential Evolution - a Simple and\n",
      "               Efficient Heuristic for Global Optimization over Continuous Spaces,\n",
      "               Journal of Global Optimization, 1997, 11, 341 - 359.\n",
      "        .. [2] http://www1.icsi.berkeley.edu/~storn/code.html\n",
      "        .. [3] http://en.wikipedia.org/wiki/Differential_evolution\n",
      "        .. [4] Wormington, M., Panaccione, C., Matney, K. M., Bowen, D. K., -\n",
      "               Characterization of structures from X-ray scattering data using\n",
      "               genetic algorithms, Phil. Trans. R. Soc. Lond. A, 1999, 357,\n",
      "               2827-2848\n",
      "        .. [5] Lampinen, J., A constraint handling approach for the differential\n",
      "               evolution algorithm. Proceedings of the 2002 Congress on\n",
      "               Evolutionary Computation. CEC'02 (Cat. No. 02TH8600). Vol. 2. IEEE,\n",
      "               2002.\n",
      "    \n",
      "    dual_annealing(func, bounds, args=(), maxiter=1000, local_search_options={}, initial_temp=5230.0, restart_temp_ratio=2e-05, visit=2.62, accept=-5.0, maxfun=10000000.0, seed=None, no_local_search=False, callback=None, x0=None)\n",
      "        Find the global minimum of a function using Dual Annealing.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        func : callable\n",
      "            The objective function to be minimized. Must be in the form\n",
      "            ``f(x, *args)``, where ``x`` is the argument in the form of a 1-D array\n",
      "            and ``args`` is a  tuple of any additional fixed parameters needed to\n",
      "            completely specify the function.\n",
      "        bounds : sequence, shape (n, 2)\n",
      "            Bounds for variables.  ``(min, max)`` pairs for each element in ``x``,\n",
      "            defining bounds for the objective function parameter.\n",
      "        args : tuple, optional\n",
      "            Any additional fixed parameters needed to completely specify the\n",
      "            objective function.\n",
      "        maxiter : int, optional\n",
      "            The maximum number of global search iterations. Default value is 1000.\n",
      "        local_search_options : dict, optional\n",
      "            Extra keyword arguments to be passed to the local minimizer\n",
      "            (`minimize`). Some important options could be:\n",
      "            ``method`` for the minimizer method to use and ``args`` for\n",
      "            objective function additional arguments.\n",
      "        initial_temp : float, optional\n",
      "            The initial temperature, use higher values to facilitates a wider\n",
      "            search of the energy landscape, allowing dual_annealing to escape\n",
      "            local minima that it is trapped in. Default value is 5230. Range is\n",
      "            (0.01, 5.e4].\n",
      "        restart_temp_ratio : float, optional\n",
      "            During the annealing process, temperature is decreasing, when it\n",
      "            reaches ``initial_temp * restart_temp_ratio``, the reannealing process\n",
      "            is triggered. Default value of the ratio is 2e-5. Range is (0, 1).\n",
      "        visit : float, optional\n",
      "            Parameter for visiting distribution. Default value is 2.62. Higher\n",
      "            values give the visiting distribution a heavier tail, this makes\n",
      "            the algorithm jump to a more distant region. The value range is (1, 3].\n",
      "        accept : float, optional\n",
      "            Parameter for acceptance distribution. It is used to control the\n",
      "            probability of acceptance. The lower the acceptance parameter, the\n",
      "            smaller the probability of acceptance. Default value is -5.0 with\n",
      "            a range (-1e4, -5].\n",
      "        maxfun : int, optional\n",
      "            Soft limit for the number of objective function calls. If the\n",
      "            algorithm is in the middle of a local search, this number will be\n",
      "            exceeded, the algorithm will stop just after the local search is\n",
      "            done. Default value is 1e7.\n",
      "        seed : {None, int, `numpy.random.Generator`,\n",
      "                `numpy.random.RandomState`}, optional\n",
      "        \n",
      "            If `seed` is None (or `np.random`), the `numpy.random.RandomState`\n",
      "            singleton is used.\n",
      "            If `seed` is an int, a new ``RandomState`` instance is used,\n",
      "            seeded with `seed`.\n",
      "            If `seed` is already a ``Generator`` or ``RandomState`` instance then\n",
      "            that instance is used.\n",
      "            Specify `seed` for repeatable minimizations. The random numbers\n",
      "            generated with this seed only affect the visiting distribution function\n",
      "            and new coordinates generation.\n",
      "        no_local_search : bool, optional\n",
      "            If `no_local_search` is set to True, a traditional Generalized\n",
      "            Simulated Annealing will be performed with no local search\n",
      "            strategy applied.\n",
      "        callback : callable, optional\n",
      "            A callback function with signature ``callback(x, f, context)``,\n",
      "            which will be called for all minima found.\n",
      "            ``x`` and ``f`` are the coordinates and function value of the\n",
      "            latest minimum found, and ``context`` has value in [0, 1, 2], with the\n",
      "            following meaning:\n",
      "        \n",
      "                - 0: minimum detected in the annealing process.\n",
      "                - 1: detection occurred in the local search process.\n",
      "                - 2: detection done in the dual annealing process.\n",
      "        \n",
      "            If the callback implementation returns True, the algorithm will stop.\n",
      "        x0 : ndarray, shape(n,), optional\n",
      "            Coordinates of a single N-D starting point.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        res : OptimizeResult\n",
      "            The optimization result represented as a `OptimizeResult` object.\n",
      "            Important attributes are: ``x`` the solution array, ``fun`` the value\n",
      "            of the function at the solution, and ``message`` which describes the\n",
      "            cause of the termination.\n",
      "            See `OptimizeResult` for a description of other attributes.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        This function implements the Dual Annealing optimization. This stochastic\n",
      "        approach derived from [3]_ combines the generalization of CSA (Classical\n",
      "        Simulated Annealing) and FSA (Fast Simulated Annealing) [1]_ [2]_ coupled\n",
      "        to a strategy for applying a local search on accepted locations [4]_.\n",
      "        An alternative implementation of this same algorithm is described in [5]_\n",
      "        and benchmarks are presented in [6]_. This approach introduces an advanced\n",
      "        method to refine the solution found by the generalized annealing\n",
      "        process. This algorithm uses a distorted Cauchy-Lorentz visiting\n",
      "        distribution, with its shape controlled by the parameter :math:`q_{v}`\n",
      "        \n",
      "        .. math::\n",
      "        \n",
      "            g_{q_{v}}(\\Delta x(t)) \\propto \\frac{ \\\n",
      "            \\left[T_{q_{v}}(t) \\right]^{-\\frac{D}{3-q_{v}}}}{ \\\n",
      "            \\left[{1+(q_{v}-1)\\frac{(\\Delta x(t))^{2}} { \\\n",
      "            \\left[T_{q_{v}}(t)\\right]^{\\frac{2}{3-q_{v}}}}}\\right]^{ \\\n",
      "            \\frac{1}{q_{v}-1}+\\frac{D-1}{2}}}\n",
      "        \n",
      "        Where :math:`t` is the artificial time. This visiting distribution is used\n",
      "        to generate a trial jump distance :math:`\\Delta x(t)` of variable\n",
      "        :math:`x(t)` under artificial temperature :math:`T_{q_{v}}(t)`.\n",
      "        \n",
      "        From the starting point, after calling the visiting distribution\n",
      "        function, the acceptance probability is computed as follows:\n",
      "        \n",
      "        .. math::\n",
      "        \n",
      "            p_{q_{a}} = \\min{\\{1,\\left[1-(1-q_{a}) \\beta \\Delta E \\right]^{ \\\n",
      "            \\frac{1}{1-q_{a}}}\\}}\n",
      "        \n",
      "        Where :math:`q_{a}` is a acceptance parameter. For :math:`q_{a}<1`, zero\n",
      "        acceptance probability is assigned to the cases where\n",
      "        \n",
      "        .. math::\n",
      "        \n",
      "            [1-(1-q_{a}) \\beta \\Delta E] < 0\n",
      "        \n",
      "        The artificial temperature :math:`T_{q_{v}}(t)` is decreased according to\n",
      "        \n",
      "        .. math::\n",
      "        \n",
      "            T_{q_{v}}(t) = T_{q_{v}}(1) \\frac{2^{q_{v}-1}-1}{\\left( \\\n",
      "            1 + t\\right)^{q_{v}-1}-1}\n",
      "        \n",
      "        Where :math:`q_{v}` is the visiting parameter.\n",
      "        \n",
      "        .. versionadded:: 1.2.0\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] Tsallis C. Possible generalization of Boltzmann-Gibbs\n",
      "            statistics. Journal of Statistical Physics, 52, 479-487 (1998).\n",
      "        .. [2] Tsallis C, Stariolo DA. Generalized Simulated Annealing.\n",
      "            Physica A, 233, 395-406 (1996).\n",
      "        .. [3] Xiang Y, Sun DY, Fan W, Gong XG. Generalized Simulated\n",
      "            Annealing Algorithm and Its Application to the Thomson Model.\n",
      "            Physics Letters A, 233, 216-220 (1997).\n",
      "        .. [4] Xiang Y, Gong XG. Efficiency of Generalized Simulated\n",
      "            Annealing. Physical Review E, 62, 4473 (2000).\n",
      "        .. [5] Xiang Y, Gubian S, Suomela B, Hoeng J. Generalized\n",
      "            Simulated Annealing for Efficient Global Optimization: the GenSA\n",
      "            Package for R. The R Journal, Volume 5/1 (2013).\n",
      "        .. [6] Mullen, K. Continuous Global Optimization in R. Journal of\n",
      "            Statistical Software, 60(6), 1 - 45, (2014).\n",
      "            :doi:`10.18637/jss.v060.i06`\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        The following example is a 10-D problem, with many local minima.\n",
      "        The function involved is called Rastrigin\n",
      "        (https://en.wikipedia.org/wiki/Rastrigin_function)\n",
      "        \n",
      "        >>> from scipy.optimize import dual_annealing\n",
      "        >>> func = lambda x: np.sum(x*x - 10*np.cos(2*np.pi*x)) + 10*np.size(x)\n",
      "        >>> lw = [-5.12] * 10\n",
      "        >>> up = [5.12] * 10\n",
      "        >>> ret = dual_annealing(func, bounds=list(zip(lw, up)))\n",
      "        >>> ret.x\n",
      "        array([-4.26437714e-09, -3.91699361e-09, -1.86149218e-09, -3.97165720e-09,\n",
      "               -6.29151648e-09, -6.53145322e-09, -3.93616815e-09, -6.55623025e-09,\n",
      "               -6.05775280e-09, -5.00668935e-09]) # random\n",
      "        >>> ret.fun\n",
      "        0.000000\n",
      "    \n",
      "    excitingmixing(F, xin, iter=None, alpha=None, alphamax=1.0, verbose=False, maxiter=None, f_tol=None, f_rtol=None, x_tol=None, x_rtol=None, tol_norm=None, line_search='armijo', callback=None, **kw)\n",
      "        Find a root of a function, using a tuned diagonal Jacobian approximation.\n",
      "        \n",
      "        The Jacobian matrix is diagonal and is tuned on each iteration.\n",
      "        \n",
      "        .. warning::\n",
      "        \n",
      "           This algorithm may be useful for specific problems, but whether\n",
      "           it will work may depend strongly on the problem.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        root : Interface to root finding algorithms for multivariate\n",
      "               functions. See ``method=='excitingmixing'`` in particular.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        F : function(x) -> f\n",
      "            Function whose root to find; should take and return an array-like\n",
      "            object.\n",
      "        xin : array_like\n",
      "            Initial guess for the solution\n",
      "        alpha : float, optional\n",
      "            Initial Jacobian approximation is (-1/alpha).\n",
      "        alphamax : float, optional\n",
      "            The entries of the diagonal Jacobian are kept in the range\n",
      "            ``[alpha, alphamax]``.\n",
      "        iter : int, optional\n",
      "            Number of iterations to make. If omitted (default), make as many\n",
      "            as required to meet tolerances.\n",
      "        verbose : bool, optional\n",
      "            Print status to stdout on every iteration.\n",
      "        maxiter : int, optional\n",
      "            Maximum number of iterations to make. If more are needed to\n",
      "            meet convergence, `NoConvergence` is raised.\n",
      "        f_tol : float, optional\n",
      "            Absolute tolerance (in max-norm) for the residual.\n",
      "            If omitted, default is 6e-6.\n",
      "        f_rtol : float, optional\n",
      "            Relative tolerance for the residual. If omitted, not used.\n",
      "        x_tol : float, optional\n",
      "            Absolute minimum step size, as determined from the Jacobian\n",
      "            approximation. If the step size is smaller than this, optimization\n",
      "            is terminated as successful. If omitted, not used.\n",
      "        x_rtol : float, optional\n",
      "            Relative minimum step size. If omitted, not used.\n",
      "        tol_norm : function(vector) -> scalar, optional\n",
      "            Norm to use in convergence check. Default is the maximum norm.\n",
      "        line_search : {None, 'armijo' (default), 'wolfe'}, optional\n",
      "            Which type of a line search to use to determine the step size in the\n",
      "            direction given by the Jacobian approximation. Defaults to 'armijo'.\n",
      "        callback : function, optional\n",
      "            Optional callback function. It is called on every iteration as\n",
      "            ``callback(x, f)`` where `x` is the current solution and `f`\n",
      "            the corresponding residual.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        sol : ndarray\n",
      "            An array (of similar array type as `x0`) containing the final solution.\n",
      "        \n",
      "        Raises\n",
      "        ------\n",
      "        NoConvergence\n",
      "            When a solution was not found.\n",
      "    \n",
      "    fixed_point(func, x0, args=(), xtol=1e-08, maxiter=500, method='del2')\n",
      "        Find a fixed point of the function.\n",
      "        \n",
      "        Given a function of one or more variables and a starting point, find a\n",
      "        fixed point of the function: i.e., where ``func(x0) == x0``.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        func : function\n",
      "            Function to evaluate.\n",
      "        x0 : array_like\n",
      "            Fixed point of function.\n",
      "        args : tuple, optional\n",
      "            Extra arguments to `func`.\n",
      "        xtol : float, optional\n",
      "            Convergence tolerance, defaults to 1e-08.\n",
      "        maxiter : int, optional\n",
      "            Maximum number of iterations, defaults to 500.\n",
      "        method : {\"del2\", \"iteration\"}, optional\n",
      "            Method of finding the fixed-point, defaults to \"del2\",\n",
      "            which uses Steffensen's Method with Aitken's ``Del^2``\n",
      "            convergence acceleration [1]_. The \"iteration\" method simply iterates\n",
      "            the function until convergence is detected, without attempting to\n",
      "            accelerate the convergence.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] Burden, Faires, \"Numerical Analysis\", 5th edition, pg. 80\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy import optimize\n",
      "        >>> def func(x, c1, c2):\n",
      "        ...    return np.sqrt(c1/(x+c2))\n",
      "        >>> c1 = np.array([10,12.])\n",
      "        >>> c2 = np.array([3, 5.])\n",
      "        >>> optimize.fixed_point(func, [1.2, 1.3], args=(c1,c2))\n",
      "        array([ 1.4920333 ,  1.37228132])\n",
      "    \n",
      "    fmin(func, x0, args=(), xtol=0.0001, ftol=0.0001, maxiter=None, maxfun=None, full_output=0, disp=1, retall=0, callback=None, initial_simplex=None)\n",
      "        Minimize a function using the downhill simplex algorithm.\n",
      "        \n",
      "        This algorithm only uses function values, not derivatives or second\n",
      "        derivatives.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        func : callable func(x,*args)\n",
      "            The objective function to be minimized.\n",
      "        x0 : ndarray\n",
      "            Initial guess.\n",
      "        args : tuple, optional\n",
      "            Extra arguments passed to func, i.e., ``f(x,*args)``.\n",
      "        xtol : float, optional\n",
      "            Absolute error in xopt between iterations that is acceptable for\n",
      "            convergence.\n",
      "        ftol : number, optional\n",
      "            Absolute error in func(xopt) between iterations that is acceptable for\n",
      "            convergence.\n",
      "        maxiter : int, optional\n",
      "            Maximum number of iterations to perform.\n",
      "        maxfun : number, optional\n",
      "            Maximum number of function evaluations to make.\n",
      "        full_output : bool, optional\n",
      "            Set to True if fopt and warnflag outputs are desired.\n",
      "        disp : bool, optional\n",
      "            Set to True to print convergence messages.\n",
      "        retall : bool, optional\n",
      "            Set to True to return list of solutions at each iteration.\n",
      "        callback : callable, optional\n",
      "            Called after each iteration, as callback(xk), where xk is the\n",
      "            current parameter vector.\n",
      "        initial_simplex : array_like of shape (N + 1, N), optional\n",
      "            Initial simplex. If given, overrides `x0`.\n",
      "            ``initial_simplex[j,:]`` should contain the coordinates of\n",
      "            the jth vertex of the ``N+1`` vertices in the simplex, where\n",
      "            ``N`` is the dimension.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        xopt : ndarray\n",
      "            Parameter that minimizes function.\n",
      "        fopt : float\n",
      "            Value of function at minimum: ``fopt = func(xopt)``.\n",
      "        iter : int\n",
      "            Number of iterations performed.\n",
      "        funcalls : int\n",
      "            Number of function calls made.\n",
      "        warnflag : int\n",
      "            1 : Maximum number of function evaluations made.\n",
      "            2 : Maximum number of iterations reached.\n",
      "        allvecs : list\n",
      "            Solution at each iteration.\n",
      "        \n",
      "        See also\n",
      "        --------\n",
      "        minimize: Interface to minimization algorithms for multivariate\n",
      "            functions. See the 'Nelder-Mead' `method` in particular.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        Uses a Nelder-Mead simplex algorithm to find the minimum of function of\n",
      "        one or more variables.\n",
      "        \n",
      "        This algorithm has a long history of successful use in applications.\n",
      "        But it will usually be slower than an algorithm that uses first or\n",
      "        second derivative information. In practice, it can have poor\n",
      "        performance in high-dimensional problems and is not robust to\n",
      "        minimizing complicated functions. Additionally, there currently is no\n",
      "        complete theory describing when the algorithm will successfully\n",
      "        converge to the minimum, or how fast it will if it does. Both the ftol and\n",
      "        xtol criteria must be met for convergence.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> def f(x):\n",
      "        ...     return x**2\n",
      "        \n",
      "        >>> from scipy import optimize\n",
      "        \n",
      "        >>> minimum = optimize.fmin(f, 1)\n",
      "        Optimization terminated successfully.\n",
      "                 Current function value: 0.000000\n",
      "                 Iterations: 17\n",
      "                 Function evaluations: 34\n",
      "        >>> minimum[0]\n",
      "        -8.8817841970012523e-16\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] Nelder, J.A. and Mead, R. (1965), \"A simplex method for function\n",
      "               minimization\", The Computer Journal, 7, pp. 308-313\n",
      "        \n",
      "        .. [2] Wright, M.H. (1996), \"Direct Search Methods: Once Scorned, Now\n",
      "               Respectable\", in Numerical Analysis 1995, Proceedings of the\n",
      "               1995 Dundee Biennial Conference in Numerical Analysis, D.F.\n",
      "               Griffiths and G.A. Watson (Eds.), Addison Wesley Longman,\n",
      "               Harlow, UK, pp. 191-208.\n",
      "    \n",
      "    fmin_bfgs(f, x0, fprime=None, args=(), gtol=1e-05, norm=inf, epsilon=1.4901161193847656e-08, maxiter=None, full_output=0, disp=1, retall=0, callback=None)\n",
      "        Minimize a function using the BFGS algorithm.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        f : callable ``f(x,*args)``\n",
      "            Objective function to be minimized.\n",
      "        x0 : ndarray\n",
      "            Initial guess.\n",
      "        fprime : callable ``f'(x,*args)``, optional\n",
      "            Gradient of f.\n",
      "        args : tuple, optional\n",
      "            Extra arguments passed to f and fprime.\n",
      "        gtol : float, optional\n",
      "            Gradient norm must be less than `gtol` before successful termination.\n",
      "        norm : float, optional\n",
      "            Order of norm (Inf is max, -Inf is min)\n",
      "        epsilon : int or ndarray, optional\n",
      "            If `fprime` is approximated, use this value for the step size.\n",
      "        callback : callable, optional\n",
      "            An optional user-supplied function to call after each\n",
      "            iteration. Called as ``callback(xk)``, where ``xk`` is the\n",
      "            current parameter vector.\n",
      "        maxiter : int, optional\n",
      "            Maximum number of iterations to perform.\n",
      "        full_output : bool, optional\n",
      "            If True, return ``fopt``, ``func_calls``, ``grad_calls``, and\n",
      "            ``warnflag`` in addition to ``xopt``.\n",
      "        disp : bool, optional\n",
      "            Print convergence message if True.\n",
      "        retall : bool, optional\n",
      "            Return a list of results at each iteration if True.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        xopt : ndarray\n",
      "            Parameters which minimize f, i.e., ``f(xopt) == fopt``.\n",
      "        fopt : float\n",
      "            Minimum value.\n",
      "        gopt : ndarray\n",
      "            Value of gradient at minimum, f'(xopt), which should be near 0.\n",
      "        Bopt : ndarray\n",
      "            Value of 1/f''(xopt), i.e., the inverse Hessian matrix.\n",
      "        func_calls : int\n",
      "            Number of function_calls made.\n",
      "        grad_calls : int\n",
      "            Number of gradient calls made.\n",
      "        warnflag : integer\n",
      "            1 : Maximum number of iterations exceeded.\n",
      "            2 : Gradient and/or function calls not changing.\n",
      "            3 : NaN result encountered.\n",
      "        allvecs : list\n",
      "            The value of `xopt` at each iteration. Only returned if `retall` is\n",
      "            True.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        Optimize the function, `f`, whose gradient is given by `fprime`\n",
      "        using the quasi-Newton method of Broyden, Fletcher, Goldfarb,\n",
      "        and Shanno (BFGS).\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        minimize: Interface to minimization algorithms for multivariate\n",
      "            functions. See ``method='BFGS'`` in particular.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        Wright, and Nocedal 'Numerical Optimization', 1999, p. 198.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy.optimize import fmin_bfgs\n",
      "        >>> def quadratic_cost(x, Q):\n",
      "        ...     return x @ Q @ x\n",
      "        ...\n",
      "        >>> x0 = np.array([-3, -4])\n",
      "        >>> cost_weight =  np.diag([1., 10.])\n",
      "        >>> # Note that a trailing comma is necessary for a tuple with single element\n",
      "        >>> fmin_bfgs(quadratic_cost, x0, args=(cost_weight,))\n",
      "        Optimization terminated successfully.\n",
      "                Current function value: 0.000000\n",
      "                Iterations: 7                   # may vary\n",
      "                Function evaluations: 24        # may vary\n",
      "                Gradient evaluations: 8         # may vary\n",
      "        array([ 2.85169950e-06, -4.61820139e-07])\n",
      "        \n",
      "        >>> def quadratic_cost_grad(x, Q):\n",
      "        ...     return 2 * Q @ x\n",
      "        ...\n",
      "        >>> fmin_bfgs(quadratic_cost, x0, quadratic_cost_grad, args=(cost_weight,))\n",
      "        Optimization terminated successfully.\n",
      "                Current function value: 0.000000\n",
      "                Iterations: 7\n",
      "                Function evaluations: 8\n",
      "                Gradient evaluations: 8\n",
      "        array([ 2.85916637e-06, -4.54371951e-07])\n",
      "    \n",
      "    fmin_cg(f, x0, fprime=None, args=(), gtol=1e-05, norm=inf, epsilon=1.4901161193847656e-08, maxiter=None, full_output=0, disp=1, retall=0, callback=None)\n",
      "        Minimize a function using a nonlinear conjugate gradient algorithm.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        f : callable, ``f(x, *args)``\n",
      "            Objective function to be minimized. Here `x` must be a 1-D array of\n",
      "            the variables that are to be changed in the search for a minimum, and\n",
      "            `args` are the other (fixed) parameters of `f`.\n",
      "        x0 : ndarray\n",
      "            A user-supplied initial estimate of `xopt`, the optimal value of `x`.\n",
      "            It must be a 1-D array of values.\n",
      "        fprime : callable, ``fprime(x, *args)``, optional\n",
      "            A function that returns the gradient of `f` at `x`. Here `x` and `args`\n",
      "            are as described above for `f`. The returned value must be a 1-D array.\n",
      "            Defaults to None, in which case the gradient is approximated\n",
      "            numerically (see `epsilon`, below).\n",
      "        args : tuple, optional\n",
      "            Parameter values passed to `f` and `fprime`. Must be supplied whenever\n",
      "            additional fixed parameters are needed to completely specify the\n",
      "            functions `f` and `fprime`.\n",
      "        gtol : float, optional\n",
      "            Stop when the norm of the gradient is less than `gtol`.\n",
      "        norm : float, optional\n",
      "            Order to use for the norm of the gradient\n",
      "            (``-np.Inf`` is min, ``np.Inf`` is max).\n",
      "        epsilon : float or ndarray, optional\n",
      "            Step size(s) to use when `fprime` is approximated numerically. Can be a\n",
      "            scalar or a 1-D array. Defaults to ``sqrt(eps)``, with eps the\n",
      "            floating point machine precision.  Usually ``sqrt(eps)`` is about\n",
      "            1.5e-8.\n",
      "        maxiter : int, optional\n",
      "            Maximum number of iterations to perform. Default is ``200 * len(x0)``.\n",
      "        full_output : bool, optional\n",
      "            If True, return `fopt`, `func_calls`, `grad_calls`, and `warnflag` in\n",
      "            addition to `xopt`.  See the Returns section below for additional\n",
      "            information on optional return values.\n",
      "        disp : bool, optional\n",
      "            If True, return a convergence message, followed by `xopt`.\n",
      "        retall : bool, optional\n",
      "            If True, add to the returned values the results of each iteration.\n",
      "        callback : callable, optional\n",
      "            An optional user-supplied function, called after each iteration.\n",
      "            Called as ``callback(xk)``, where ``xk`` is the current value of `x0`.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        xopt : ndarray\n",
      "            Parameters which minimize f, i.e., ``f(xopt) == fopt``.\n",
      "        fopt : float, optional\n",
      "            Minimum value found, f(xopt). Only returned if `full_output` is True.\n",
      "        func_calls : int, optional\n",
      "            The number of function_calls made. Only returned if `full_output`\n",
      "            is True.\n",
      "        grad_calls : int, optional\n",
      "            The number of gradient calls made. Only returned if `full_output` is\n",
      "            True.\n",
      "        warnflag : int, optional\n",
      "            Integer value with warning status, only returned if `full_output` is\n",
      "            True.\n",
      "        \n",
      "            0 : Success.\n",
      "        \n",
      "            1 : The maximum number of iterations was exceeded.\n",
      "        \n",
      "            2 : Gradient and/or function calls were not changing. May indicate\n",
      "                that precision was lost, i.e., the routine did not converge.\n",
      "        \n",
      "            3 : NaN result encountered.\n",
      "        \n",
      "        allvecs : list of ndarray, optional\n",
      "            List of arrays, containing the results at each iteration.\n",
      "            Only returned if `retall` is True.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        minimize : common interface to all `scipy.optimize` algorithms for\n",
      "                   unconstrained and constrained minimization of multivariate\n",
      "                   functions. It provides an alternative way to call\n",
      "                   ``fmin_cg``, by specifying ``method='CG'``.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        This conjugate gradient algorithm is based on that of Polak and Ribiere\n",
      "        [1]_.\n",
      "        \n",
      "        Conjugate gradient methods tend to work better when:\n",
      "        \n",
      "        1. `f` has a unique global minimizing point, and no local minima or\n",
      "           other stationary points,\n",
      "        2. `f` is, at least locally, reasonably well approximated by a\n",
      "           quadratic function of the variables,\n",
      "        3. `f` is continuous and has a continuous gradient,\n",
      "        4. `fprime` is not too large, e.g., has a norm less than 1000,\n",
      "        5. The initial guess, `x0`, is reasonably close to `f` 's global\n",
      "           minimizing point, `xopt`.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] Wright & Nocedal, \"Numerical Optimization\", 1999, pp. 120-122.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        Example 1: seek the minimum value of the expression\n",
      "        ``a*u**2 + b*u*v + c*v**2 + d*u + e*v + f`` for given values\n",
      "        of the parameters and an initial guess ``(u, v) = (0, 0)``.\n",
      "        \n",
      "        >>> args = (2, 3, 7, 8, 9, 10)  # parameter values\n",
      "        >>> def f(x, *args):\n",
      "        ...     u, v = x\n",
      "        ...     a, b, c, d, e, f = args\n",
      "        ...     return a*u**2 + b*u*v + c*v**2 + d*u + e*v + f\n",
      "        >>> def gradf(x, *args):\n",
      "        ...     u, v = x\n",
      "        ...     a, b, c, d, e, f = args\n",
      "        ...     gu = 2*a*u + b*v + d     # u-component of the gradient\n",
      "        ...     gv = b*u + 2*c*v + e     # v-component of the gradient\n",
      "        ...     return np.asarray((gu, gv))\n",
      "        >>> x0 = np.asarray((0, 0))  # Initial guess.\n",
      "        >>> from scipy import optimize\n",
      "        >>> res1 = optimize.fmin_cg(f, x0, fprime=gradf, args=args)\n",
      "        Optimization terminated successfully.\n",
      "                 Current function value: 1.617021\n",
      "                 Iterations: 4\n",
      "                 Function evaluations: 8\n",
      "                 Gradient evaluations: 8\n",
      "        >>> res1\n",
      "        array([-1.80851064, -0.25531915])\n",
      "        \n",
      "        Example 2: solve the same problem using the `minimize` function.\n",
      "        (This `myopts` dictionary shows all of the available options,\n",
      "        although in practice only non-default values would be needed.\n",
      "        The returned value will be a dictionary.)\n",
      "        \n",
      "        >>> opts = {'maxiter' : None,    # default value.\n",
      "        ...         'disp' : True,    # non-default value.\n",
      "        ...         'gtol' : 1e-5,    # default value.\n",
      "        ...         'norm' : np.inf,  # default value.\n",
      "        ...         'eps' : 1.4901161193847656e-08}  # default value.\n",
      "        >>> res2 = optimize.minimize(f, x0, jac=gradf, args=args,\n",
      "        ...                          method='CG', options=opts)\n",
      "        Optimization terminated successfully.\n",
      "                Current function value: 1.617021\n",
      "                Iterations: 4\n",
      "                Function evaluations: 8\n",
      "                Gradient evaluations: 8\n",
      "        >>> res2.x  # minimum found\n",
      "        array([-1.80851064, -0.25531915])\n",
      "    \n",
      "    fmin_cobyla(func, x0, cons, args=(), consargs=None, rhobeg=1.0, rhoend=0.0001, maxfun=1000, disp=None, catol=0.0002)\n",
      "        Minimize a function using the Constrained Optimization By Linear\n",
      "        Approximation (COBYLA) method. This method wraps a FORTRAN\n",
      "        implementation of the algorithm.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        func : callable\n",
      "            Function to minimize. In the form func(x, \\*args).\n",
      "        x0 : ndarray\n",
      "            Initial guess.\n",
      "        cons : sequence\n",
      "            Constraint functions; must all be ``>=0`` (a single function\n",
      "            if only 1 constraint). Each function takes the parameters `x`\n",
      "            as its first argument, and it can return either a single number or\n",
      "            an array or list of numbers.\n",
      "        args : tuple, optional\n",
      "            Extra arguments to pass to function.\n",
      "        consargs : tuple, optional\n",
      "            Extra arguments to pass to constraint functions (default of None means\n",
      "            use same extra arguments as those passed to func).\n",
      "            Use ``()`` for no extra arguments.\n",
      "        rhobeg : float, optional\n",
      "            Reasonable initial changes to the variables.\n",
      "        rhoend : float, optional\n",
      "            Final accuracy in the optimization (not precisely guaranteed). This\n",
      "            is a lower bound on the size of the trust region.\n",
      "        disp : {0, 1, 2, 3}, optional\n",
      "            Controls the frequency of output; 0 implies no output.\n",
      "        maxfun : int, optional\n",
      "            Maximum number of function evaluations.\n",
      "        catol : float, optional\n",
      "            Absolute tolerance for constraint violations.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        x : ndarray\n",
      "            The argument that minimises `f`.\n",
      "        \n",
      "        See also\n",
      "        --------\n",
      "        minimize: Interface to minimization algorithms for multivariate\n",
      "            functions. See the 'COBYLA' `method` in particular.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        This algorithm is based on linear approximations to the objective\n",
      "        function and each constraint. We briefly describe the algorithm.\n",
      "        \n",
      "        Suppose the function is being minimized over k variables. At the\n",
      "        jth iteration the algorithm has k+1 points v_1, ..., v_(k+1),\n",
      "        an approximate solution x_j, and a radius RHO_j.\n",
      "        (i.e., linear plus a constant) approximations to the objective\n",
      "        function and constraint functions such that their function values\n",
      "        agree with the linear approximation on the k+1 points v_1,.., v_(k+1).\n",
      "        This gives a linear program to solve (where the linear approximations\n",
      "        of the constraint functions are constrained to be non-negative).\n",
      "        \n",
      "        However, the linear approximations are likely only good\n",
      "        approximations near the current simplex, so the linear program is\n",
      "        given the further requirement that the solution, which\n",
      "        will become x_(j+1), must be within RHO_j from x_j. RHO_j only\n",
      "        decreases, never increases. The initial RHO_j is rhobeg and the\n",
      "        final RHO_j is rhoend. In this way COBYLA's iterations behave\n",
      "        like a trust region algorithm.\n",
      "        \n",
      "        Additionally, the linear program may be inconsistent, or the\n",
      "        approximation may give poor improvement. For details about\n",
      "        how these issues are resolved, as well as how the points v_i are\n",
      "        updated, refer to the source code or the references below.\n",
      "        \n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        Powell M.J.D. (1994), \"A direct search optimization method that models\n",
      "        the objective and constraint functions by linear interpolation.\", in\n",
      "        Advances in Optimization and Numerical Analysis, eds. S. Gomez and\n",
      "        J-P Hennart, Kluwer Academic (Dordrecht), pp. 51-67\n",
      "        \n",
      "        Powell M.J.D. (1998), \"Direct search algorithms for optimization\n",
      "        calculations\", Acta Numerica 7, 287-336\n",
      "        \n",
      "        Powell M.J.D. (2007), \"A view of algorithms for optimization without\n",
      "        derivatives\", Cambridge University Technical Report DAMTP 2007/NA03\n",
      "        \n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        Minimize the objective function f(x,y) = x*y subject\n",
      "        to the constraints x**2 + y**2 < 1 and y > 0::\n",
      "        \n",
      "            >>> def objective(x):\n",
      "            ...     return x[0]*x[1]\n",
      "            ...\n",
      "            >>> def constr1(x):\n",
      "            ...     return 1 - (x[0]**2 + x[1]**2)\n",
      "            ...\n",
      "            >>> def constr2(x):\n",
      "            ...     return x[1]\n",
      "            ...\n",
      "            >>> from scipy.optimize import fmin_cobyla\n",
      "            >>> fmin_cobyla(objective, [0.0, 0.1], [constr1, constr2], rhoend=1e-7)\n",
      "            array([-0.70710685,  0.70710671])\n",
      "        \n",
      "        The exact solution is (-sqrt(2)/2, sqrt(2)/2).\n",
      "    \n",
      "    fmin_l_bfgs_b(func, x0, fprime=None, args=(), approx_grad=0, bounds=None, m=10, factr=10000000.0, pgtol=1e-05, epsilon=1e-08, iprint=-1, maxfun=15000, maxiter=15000, disp=None, callback=None, maxls=20)\n",
      "        Minimize a function func using the L-BFGS-B algorithm.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        func : callable f(x,*args)\n",
      "            Function to minimize.\n",
      "        x0 : ndarray\n",
      "            Initial guess.\n",
      "        fprime : callable fprime(x,*args), optional\n",
      "            The gradient of `func`. If None, then `func` returns the function\n",
      "            value and the gradient (``f, g = func(x, *args)``), unless\n",
      "            `approx_grad` is True in which case `func` returns only ``f``.\n",
      "        args : sequence, optional\n",
      "            Arguments to pass to `func` and `fprime`.\n",
      "        approx_grad : bool, optional\n",
      "            Whether to approximate the gradient numerically (in which case\n",
      "            `func` returns only the function value).\n",
      "        bounds : list, optional\n",
      "            ``(min, max)`` pairs for each element in ``x``, defining\n",
      "            the bounds on that parameter. Use None or +-inf for one of ``min`` or\n",
      "            ``max`` when there is no bound in that direction.\n",
      "        m : int, optional\n",
      "            The maximum number of variable metric corrections\n",
      "            used to define the limited memory matrix. (The limited memory BFGS\n",
      "            method does not store the full hessian but uses this many terms in an\n",
      "            approximation to it.)\n",
      "        factr : float, optional\n",
      "            The iteration stops when\n",
      "            ``(f^k - f^{k+1})/max{|f^k|,|f^{k+1}|,1} <= factr * eps``,\n",
      "            where ``eps`` is the machine precision, which is automatically\n",
      "            generated by the code. Typical values for `factr` are: 1e12 for\n",
      "            low accuracy; 1e7 for moderate accuracy; 10.0 for extremely\n",
      "            high accuracy. See Notes for relationship to `ftol`, which is exposed\n",
      "            (instead of `factr`) by the `scipy.optimize.minimize` interface to\n",
      "            L-BFGS-B.\n",
      "        pgtol : float, optional\n",
      "            The iteration will stop when\n",
      "            ``max{|proj g_i | i = 1, ..., n} <= pgtol``\n",
      "            where ``pg_i`` is the i-th component of the projected gradient.\n",
      "        epsilon : float, optional\n",
      "            Step size used when `approx_grad` is True, for numerically\n",
      "            calculating the gradient\n",
      "        iprint : int, optional\n",
      "            Controls the frequency of output. ``iprint < 0`` means no output;\n",
      "            ``iprint = 0``    print only one line at the last iteration;\n",
      "            ``0 < iprint < 99`` print also f and ``|proj g|`` every iprint iterations;\n",
      "            ``iprint = 99``   print details of every iteration except n-vectors;\n",
      "            ``iprint = 100``  print also the changes of active set and final x;\n",
      "            ``iprint > 100``  print details of every iteration including x and g.\n",
      "        disp : int, optional\n",
      "            If zero, then no output. If a positive number, then this over-rides\n",
      "            `iprint` (i.e., `iprint` gets the value of `disp`).\n",
      "        maxfun : int, optional\n",
      "            Maximum number of function evaluations.\n",
      "        maxiter : int, optional\n",
      "            Maximum number of iterations.\n",
      "        callback : callable, optional\n",
      "            Called after each iteration, as ``callback(xk)``, where ``xk`` is the\n",
      "            current parameter vector.\n",
      "        maxls : int, optional\n",
      "            Maximum number of line search steps (per iteration). Default is 20.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        x : array_like\n",
      "            Estimated position of the minimum.\n",
      "        f : float\n",
      "            Value of `func` at the minimum.\n",
      "        d : dict\n",
      "            Information dictionary.\n",
      "        \n",
      "            * d['warnflag'] is\n",
      "        \n",
      "              - 0 if converged,\n",
      "              - 1 if too many function evaluations or too many iterations,\n",
      "              - 2 if stopped for another reason, given in d['task']\n",
      "        \n",
      "            * d['grad'] is the gradient at the minimum (should be 0 ish)\n",
      "            * d['funcalls'] is the number of function calls made.\n",
      "            * d['nit'] is the number of iterations.\n",
      "        \n",
      "        See also\n",
      "        --------\n",
      "        minimize: Interface to minimization algorithms for multivariate\n",
      "            functions. See the 'L-BFGS-B' `method` in particular. Note that the\n",
      "            `ftol` option is made available via that interface, while `factr` is\n",
      "            provided via this interface, where `factr` is the factor multiplying\n",
      "            the default machine floating-point precision to arrive at `ftol`:\n",
      "            ``ftol = factr * numpy.finfo(float).eps``.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        License of L-BFGS-B (FORTRAN code):\n",
      "        \n",
      "        The version included here (in fortran code) is 3.0\n",
      "        (released April 25, 2011). It was written by Ciyou Zhu, Richard Byrd,\n",
      "        and Jorge Nocedal <nocedal@ece.nwu.edu>. It carries the following\n",
      "        condition for use:\n",
      "        \n",
      "        This software is freely available, but we expect that all publications\n",
      "        describing work using this software, or all commercial products using it,\n",
      "        quote at least one of the references given below. This software is released\n",
      "        under the BSD License.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        * R. H. Byrd, P. Lu and J. Nocedal. A Limited Memory Algorithm for Bound\n",
      "          Constrained Optimization, (1995), SIAM Journal on Scientific and\n",
      "          Statistical Computing, 16, 5, pp. 1190-1208.\n",
      "        * C. Zhu, R. H. Byrd and J. Nocedal. L-BFGS-B: Algorithm 778: L-BFGS-B,\n",
      "          FORTRAN routines for large scale bound constrained optimization (1997),\n",
      "          ACM Transactions on Mathematical Software, 23, 4, pp. 550 - 560.\n",
      "        * J.L. Morales and J. Nocedal. L-BFGS-B: Remark on Algorithm 778: L-BFGS-B,\n",
      "          FORTRAN routines for large scale bound constrained optimization (2011),\n",
      "          ACM Transactions on Mathematical Software, 38, 1.\n",
      "    \n",
      "    fmin_ncg(f, x0, fprime, fhess_p=None, fhess=None, args=(), avextol=1e-05, epsilon=1.4901161193847656e-08, maxiter=None, full_output=0, disp=1, retall=0, callback=None)\n",
      "        Unconstrained minimization of a function using the Newton-CG method.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        f : callable ``f(x, *args)``\n",
      "            Objective function to be minimized.\n",
      "        x0 : ndarray\n",
      "            Initial guess.\n",
      "        fprime : callable ``f'(x, *args)``\n",
      "            Gradient of f.\n",
      "        fhess_p : callable ``fhess_p(x, p, *args)``, optional\n",
      "            Function which computes the Hessian of f times an\n",
      "            arbitrary vector, p.\n",
      "        fhess : callable ``fhess(x, *args)``, optional\n",
      "            Function to compute the Hessian matrix of f.\n",
      "        args : tuple, optional\n",
      "            Extra arguments passed to f, fprime, fhess_p, and fhess\n",
      "            (the same set of extra arguments is supplied to all of\n",
      "            these functions).\n",
      "        epsilon : float or ndarray, optional\n",
      "            If fhess is approximated, use this value for the step size.\n",
      "        callback : callable, optional\n",
      "            An optional user-supplied function which is called after\n",
      "            each iteration. Called as callback(xk), where xk is the\n",
      "            current parameter vector.\n",
      "        avextol : float, optional\n",
      "            Convergence is assumed when the average relative error in\n",
      "            the minimizer falls below this amount.\n",
      "        maxiter : int, optional\n",
      "            Maximum number of iterations to perform.\n",
      "        full_output : bool, optional\n",
      "            If True, return the optional outputs.\n",
      "        disp : bool, optional\n",
      "            If True, print convergence message.\n",
      "        retall : bool, optional\n",
      "            If True, return a list of results at each iteration.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        xopt : ndarray\n",
      "            Parameters which minimize f, i.e., ``f(xopt) == fopt``.\n",
      "        fopt : float\n",
      "            Value of the function at xopt, i.e., ``fopt = f(xopt)``.\n",
      "        fcalls : int\n",
      "            Number of function calls made.\n",
      "        gcalls : int\n",
      "            Number of gradient calls made.\n",
      "        hcalls : int\n",
      "            Number of Hessian calls made.\n",
      "        warnflag : int\n",
      "            Warnings generated by the algorithm.\n",
      "            1 : Maximum number of iterations exceeded.\n",
      "            2 : Line search failure (precision loss).\n",
      "            3 : NaN result encountered.\n",
      "        allvecs : list\n",
      "            The result at each iteration, if retall is True (see below).\n",
      "        \n",
      "        See also\n",
      "        --------\n",
      "        minimize: Interface to minimization algorithms for multivariate\n",
      "            functions. See the 'Newton-CG' `method` in particular.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        Only one of `fhess_p` or `fhess` need to be given.  If `fhess`\n",
      "        is provided, then `fhess_p` will be ignored. If neither `fhess`\n",
      "        nor `fhess_p` is provided, then the hessian product will be\n",
      "        approximated using finite differences on `fprime`. `fhess_p`\n",
      "        must compute the hessian times an arbitrary vector. If it is not\n",
      "        given, finite-differences on `fprime` are used to compute\n",
      "        it.\n",
      "        \n",
      "        Newton-CG methods are also called truncated Newton methods. This\n",
      "        function differs from scipy.optimize.fmin_tnc because\n",
      "        \n",
      "        1. scipy.optimize.fmin_ncg is written purely in Python using NumPy\n",
      "            and scipy while scipy.optimize.fmin_tnc calls a C function.\n",
      "        2. scipy.optimize.fmin_ncg is only for unconstrained minimization\n",
      "            while scipy.optimize.fmin_tnc is for unconstrained minimization\n",
      "            or box constrained minimization. (Box constraints give\n",
      "            lower and upper bounds for each variable separately.)\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        Wright & Nocedal, 'Numerical Optimization', 1999, p. 140.\n",
      "    \n",
      "    fmin_powell(func, x0, args=(), xtol=0.0001, ftol=0.0001, maxiter=None, maxfun=None, full_output=0, disp=1, retall=0, callback=None, direc=None)\n",
      "        Minimize a function using modified Powell's method.\n",
      "        \n",
      "        This method only uses function values, not derivatives.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        func : callable f(x,*args)\n",
      "            Objective function to be minimized.\n",
      "        x0 : ndarray\n",
      "            Initial guess.\n",
      "        args : tuple, optional\n",
      "            Extra arguments passed to func.\n",
      "        xtol : float, optional\n",
      "            Line-search error tolerance.\n",
      "        ftol : float, optional\n",
      "            Relative error in ``func(xopt)`` acceptable for convergence.\n",
      "        maxiter : int, optional\n",
      "            Maximum number of iterations to perform.\n",
      "        maxfun : int, optional\n",
      "            Maximum number of function evaluations to make.\n",
      "        full_output : bool, optional\n",
      "            If True, ``fopt``, ``xi``, ``direc``, ``iter``, ``funcalls``, and\n",
      "            ``warnflag`` are returned.\n",
      "        disp : bool, optional\n",
      "            If True, print convergence messages.\n",
      "        retall : bool, optional\n",
      "            If True, return a list of the solution at each iteration.\n",
      "        callback : callable, optional\n",
      "            An optional user-supplied function, called after each\n",
      "            iteration.  Called as ``callback(xk)``, where ``xk`` is the\n",
      "            current parameter vector.\n",
      "        direc : ndarray, optional\n",
      "            Initial fitting step and parameter order set as an (N, N) array, where N\n",
      "            is the number of fitting parameters in `x0`. Defaults to step size 1.0\n",
      "            fitting all parameters simultaneously (``np.eye((N, N))``). To\n",
      "            prevent initial consideration of values in a step or to change initial\n",
      "            step size, set to 0 or desired step size in the Jth position in the Mth\n",
      "            block, where J is the position in `x0` and M is the desired evaluation\n",
      "            step, with steps being evaluated in index order. Step size and ordering\n",
      "            will change freely as minimization proceeds.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        xopt : ndarray\n",
      "            Parameter which minimizes `func`.\n",
      "        fopt : number\n",
      "            Value of function at minimum: ``fopt = func(xopt)``.\n",
      "        direc : ndarray\n",
      "            Current direction set.\n",
      "        iter : int\n",
      "            Number of iterations.\n",
      "        funcalls : int\n",
      "            Number of function calls made.\n",
      "        warnflag : int\n",
      "            Integer warning flag:\n",
      "                1 : Maximum number of function evaluations.\n",
      "                2 : Maximum number of iterations.\n",
      "                3 : NaN result encountered.\n",
      "                4 : The result is out of the provided bounds.\n",
      "        allvecs : list\n",
      "            List of solutions at each iteration.\n",
      "        \n",
      "        See also\n",
      "        --------\n",
      "        minimize: Interface to unconstrained minimization algorithms for\n",
      "            multivariate functions. See the 'Powell' method in particular.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        Uses a modification of Powell's method to find the minimum of\n",
      "        a function of N variables. Powell's method is a conjugate\n",
      "        direction method.\n",
      "        \n",
      "        The algorithm has two loops. The outer loop merely iterates over the inner\n",
      "        loop. The inner loop minimizes over each current direction in the direction\n",
      "        set. At the end of the inner loop, if certain conditions are met, the\n",
      "        direction that gave the largest decrease is dropped and replaced with the\n",
      "        difference between the current estimated x and the estimated x from the\n",
      "        beginning of the inner-loop.\n",
      "        \n",
      "        The technical conditions for replacing the direction of greatest\n",
      "        increase amount to checking that\n",
      "        \n",
      "        1. No further gain can be made along the direction of greatest increase\n",
      "           from that iteration.\n",
      "        2. The direction of greatest increase accounted for a large sufficient\n",
      "           fraction of the decrease in the function value from that iteration of\n",
      "           the inner loop.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        Powell M.J.D. (1964) An efficient method for finding the minimum of a\n",
      "        function of several variables without calculating derivatives,\n",
      "        Computer Journal, 7 (2):155-162.\n",
      "        \n",
      "        Press W., Teukolsky S.A., Vetterling W.T., and Flannery B.P.:\n",
      "        Numerical Recipes (any edition), Cambridge University Press\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> def f(x):\n",
      "        ...     return x**2\n",
      "        \n",
      "        >>> from scipy import optimize\n",
      "        \n",
      "        >>> minimum = optimize.fmin_powell(f, -1)\n",
      "        Optimization terminated successfully.\n",
      "                 Current function value: 0.000000\n",
      "                 Iterations: 2\n",
      "                 Function evaluations: 18\n",
      "        >>> minimum\n",
      "        array(0.0)\n",
      "    \n",
      "    fmin_slsqp(func, x0, eqcons=(), f_eqcons=None, ieqcons=(), f_ieqcons=None, bounds=(), fprime=None, fprime_eqcons=None, fprime_ieqcons=None, args=(), iter=100, acc=1e-06, iprint=1, disp=None, full_output=0, epsilon=1.4901161193847656e-08, callback=None)\n",
      "        Minimize a function using Sequential Least Squares Programming\n",
      "        \n",
      "        Python interface function for the SLSQP Optimization subroutine\n",
      "        originally implemented by Dieter Kraft.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        func : callable f(x,*args)\n",
      "            Objective function.  Must return a scalar.\n",
      "        x0 : 1-D ndarray of float\n",
      "            Initial guess for the independent variable(s).\n",
      "        eqcons : list, optional\n",
      "            A list of functions of length n such that\n",
      "            eqcons[j](x,*args) == 0.0 in a successfully optimized\n",
      "            problem.\n",
      "        f_eqcons : callable f(x,*args), optional\n",
      "            Returns a 1-D array in which each element must equal 0.0 in a\n",
      "            successfully optimized problem. If f_eqcons is specified,\n",
      "            eqcons is ignored.\n",
      "        ieqcons : list, optional\n",
      "            A list of functions of length n such that\n",
      "            ieqcons[j](x,*args) >= 0.0 in a successfully optimized\n",
      "            problem.\n",
      "        f_ieqcons : callable f(x,*args), optional\n",
      "            Returns a 1-D ndarray in which each element must be greater or\n",
      "            equal to 0.0 in a successfully optimized problem. If\n",
      "            f_ieqcons is specified, ieqcons is ignored.\n",
      "        bounds : list, optional\n",
      "            A list of tuples specifying the lower and upper bound\n",
      "            for each independent variable [(xl0, xu0),(xl1, xu1),...]\n",
      "            Infinite values will be interpreted as large floating values.\n",
      "        fprime : callable `f(x,*args)`, optional\n",
      "            A function that evaluates the partial derivatives of func.\n",
      "        fprime_eqcons : callable `f(x,*args)`, optional\n",
      "            A function of the form `f(x, *args)` that returns the m by n\n",
      "            array of equality constraint normals. If not provided,\n",
      "            the normals will be approximated. The array returned by\n",
      "            fprime_eqcons should be sized as ( len(eqcons), len(x0) ).\n",
      "        fprime_ieqcons : callable `f(x,*args)`, optional\n",
      "            A function of the form `f(x, *args)` that returns the m by n\n",
      "            array of inequality constraint normals. If not provided,\n",
      "            the normals will be approximated. The array returned by\n",
      "            fprime_ieqcons should be sized as ( len(ieqcons), len(x0) ).\n",
      "        args : sequence, optional\n",
      "            Additional arguments passed to func and fprime.\n",
      "        iter : int, optional\n",
      "            The maximum number of iterations.\n",
      "        acc : float, optional\n",
      "            Requested accuracy.\n",
      "        iprint : int, optional\n",
      "            The verbosity of fmin_slsqp :\n",
      "        \n",
      "            * iprint <= 0 : Silent operation\n",
      "            * iprint == 1 : Print summary upon completion (default)\n",
      "            * iprint >= 2 : Print status of each iterate and summary\n",
      "        disp : int, optional\n",
      "            Overrides the iprint interface (preferred).\n",
      "        full_output : bool, optional\n",
      "            If False, return only the minimizer of func (default).\n",
      "            Otherwise, output final objective function and summary\n",
      "            information.\n",
      "        epsilon : float, optional\n",
      "            The step size for finite-difference derivative estimates.\n",
      "        callback : callable, optional\n",
      "            Called after each iteration, as ``callback(x)``, where ``x`` is the\n",
      "            current parameter vector.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        out : ndarray of float\n",
      "            The final minimizer of func.\n",
      "        fx : ndarray of float, if full_output is true\n",
      "            The final value of the objective function.\n",
      "        its : int, if full_output is true\n",
      "            The number of iterations.\n",
      "        imode : int, if full_output is true\n",
      "            The exit mode from the optimizer (see below).\n",
      "        smode : string, if full_output is true\n",
      "            Message describing the exit mode from the optimizer.\n",
      "        \n",
      "        See also\n",
      "        --------\n",
      "        minimize: Interface to minimization algorithms for multivariate\n",
      "            functions. See the 'SLSQP' `method` in particular.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        Exit modes are defined as follows ::\n",
      "        \n",
      "            -1 : Gradient evaluation required (g & a)\n",
      "             0 : Optimization terminated successfully\n",
      "             1 : Function evaluation required (f & c)\n",
      "             2 : More equality constraints than independent variables\n",
      "             3 : More than 3*n iterations in LSQ subproblem\n",
      "             4 : Inequality constraints incompatible\n",
      "             5 : Singular matrix E in LSQ subproblem\n",
      "             6 : Singular matrix C in LSQ subproblem\n",
      "             7 : Rank-deficient equality constraint subproblem HFTI\n",
      "             8 : Positive directional derivative for linesearch\n",
      "             9 : Iteration limit reached\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        Examples are given :ref:`in the tutorial <tutorial-sqlsp>`.\n",
      "    \n",
      "    fmin_tnc(func, x0, fprime=None, args=(), approx_grad=0, bounds=None, epsilon=1e-08, scale=None, offset=None, messages=15, maxCGit=-1, maxfun=None, eta=-1, stepmx=0, accuracy=0, fmin=0, ftol=-1, xtol=-1, pgtol=-1, rescale=-1, disp=None, callback=None)\n",
      "        Minimize a function with variables subject to bounds, using\n",
      "        gradient information in a truncated Newton algorithm. This\n",
      "        method wraps a C implementation of the algorithm.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        func : callable ``func(x, *args)``\n",
      "            Function to minimize.  Must do one of:\n",
      "        \n",
      "            1. Return f and g, where f is the value of the function and g its\n",
      "               gradient (a list of floats).\n",
      "        \n",
      "            2. Return the function value but supply gradient function\n",
      "               separately as `fprime`.\n",
      "        \n",
      "            3. Return the function value and set ``approx_grad=True``.\n",
      "        \n",
      "            If the function returns None, the minimization\n",
      "            is aborted.\n",
      "        x0 : array_like\n",
      "            Initial estimate of minimum.\n",
      "        fprime : callable ``fprime(x, *args)``, optional\n",
      "            Gradient of `func`. If None, then either `func` must return the\n",
      "            function value and the gradient (``f,g = func(x, *args)``)\n",
      "            or `approx_grad` must be True.\n",
      "        args : tuple, optional\n",
      "            Arguments to pass to function.\n",
      "        approx_grad : bool, optional\n",
      "            If true, approximate the gradient numerically.\n",
      "        bounds : list, optional\n",
      "            (min, max) pairs for each element in x0, defining the\n",
      "            bounds on that parameter. Use None or +/-inf for one of\n",
      "            min or max when there is no bound in that direction.\n",
      "        epsilon : float, optional\n",
      "            Used if approx_grad is True. The stepsize in a finite\n",
      "            difference approximation for fprime.\n",
      "        scale : array_like, optional\n",
      "            Scaling factors to apply to each variable. If None, the\n",
      "            factors are up-low for interval bounded variables and\n",
      "            1+|x| for the others. Defaults to None.\n",
      "        offset : array_like, optional\n",
      "            Value to subtract from each variable. If None, the\n",
      "            offsets are (up+low)/2 for interval bounded variables\n",
      "            and x for the others.\n",
      "        messages : int, optional\n",
      "            Bit mask used to select messages display during\n",
      "            minimization values defined in the MSGS dict. Defaults to\n",
      "            MGS_ALL.\n",
      "        disp : int, optional\n",
      "            Integer interface to messages. 0 = no message, 5 = all messages\n",
      "        maxCGit : int, optional\n",
      "            Maximum number of hessian*vector evaluations per main\n",
      "            iteration. If maxCGit == 0, the direction chosen is\n",
      "            -gradient if maxCGit < 0, maxCGit is set to\n",
      "            max(1,min(50,n/2)). Defaults to -1.\n",
      "        maxfun : int, optional\n",
      "            Maximum number of function evaluation. If None, maxfun is\n",
      "            set to max(100, 10*len(x0)). Defaults to None.\n",
      "        eta : float, optional\n",
      "            Severity of the line search. If < 0 or > 1, set to 0.25.\n",
      "            Defaults to -1.\n",
      "        stepmx : float, optional\n",
      "            Maximum step for the line search. May be increased during\n",
      "            call. If too small, it will be set to 10.0. Defaults to 0.\n",
      "        accuracy : float, optional\n",
      "            Relative precision for finite difference calculations. If\n",
      "            <= machine_precision, set to sqrt(machine_precision).\n",
      "            Defaults to 0.\n",
      "        fmin : float, optional\n",
      "            Minimum function value estimate. Defaults to 0.\n",
      "        ftol : float, optional\n",
      "            Precision goal for the value of f in the stopping criterion.\n",
      "            If ftol < 0.0, ftol is set to 0.0 defaults to -1.\n",
      "        xtol : float, optional\n",
      "            Precision goal for the value of x in the stopping\n",
      "            criterion (after applying x scaling factors). If xtol <\n",
      "            0.0, xtol is set to sqrt(machine_precision). Defaults to\n",
      "            -1.\n",
      "        pgtol : float, optional\n",
      "            Precision goal for the value of the projected gradient in\n",
      "            the stopping criterion (after applying x scaling factors).\n",
      "            If pgtol < 0.0, pgtol is set to 1e-2 * sqrt(accuracy).\n",
      "            Setting it to 0.0 is not recommended. Defaults to -1.\n",
      "        rescale : float, optional\n",
      "            Scaling factor (in log10) used to trigger f value\n",
      "            rescaling. If 0, rescale at each iteration. If a large\n",
      "            value, never rescale. If < 0, rescale is set to 1.3.\n",
      "        callback : callable, optional\n",
      "            Called after each iteration, as callback(xk), where xk is the\n",
      "            current parameter vector.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        x : ndarray\n",
      "            The solution.\n",
      "        nfeval : int\n",
      "            The number of function evaluations.\n",
      "        rc : int\n",
      "            Return code, see below\n",
      "        \n",
      "        See also\n",
      "        --------\n",
      "        minimize: Interface to minimization algorithms for multivariate\n",
      "            functions. See the 'TNC' `method` in particular.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        The underlying algorithm is truncated Newton, also called\n",
      "        Newton Conjugate-Gradient. This method differs from\n",
      "        scipy.optimize.fmin_ncg in that\n",
      "        \n",
      "        1. it wraps a C implementation of the algorithm\n",
      "        2. it allows each variable to be given an upper and lower bound.\n",
      "        \n",
      "        The algorithm incorporates the bound constraints by determining\n",
      "        the descent direction as in an unconstrained truncated Newton,\n",
      "        but never taking a step-size large enough to leave the space\n",
      "        of feasible x's. The algorithm keeps track of a set of\n",
      "        currently active constraints, and ignores them when computing\n",
      "        the minimum allowable step size. (The x's associated with the\n",
      "        active constraint are kept fixed.) If the maximum allowable\n",
      "        step size is zero then a new constraint is added. At the end\n",
      "        of each iteration one of the constraints may be deemed no\n",
      "        longer active and removed. A constraint is considered\n",
      "        no longer active is if it is currently active\n",
      "        but the gradient for that variable points inward from the\n",
      "        constraint. The specific constraint removed is the one\n",
      "        associated with the variable of largest index whose\n",
      "        constraint is no longer active.\n",
      "        \n",
      "        Return codes are defined as follows::\n",
      "        \n",
      "            -1 : Infeasible (lower bound > upper bound)\n",
      "             0 : Local minimum reached (|pg| ~= 0)\n",
      "             1 : Converged (|f_n-f_(n-1)| ~= 0)\n",
      "             2 : Converged (|x_n-x_(n-1)| ~= 0)\n",
      "             3 : Max. number of function evaluations reached\n",
      "             4 : Linear search failed\n",
      "             5 : All lower bounds are equal to the upper bounds\n",
      "             6 : Unable to progress\n",
      "             7 : User requested end of minimization\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        Wright S., Nocedal J. (2006), 'Numerical Optimization'\n",
      "        \n",
      "        Nash S.G. (1984), \"Newton-Type Minimization Via the Lanczos Method\",\n",
      "        SIAM Journal of Numerical Analysis 21, pp. 770-778\n",
      "    \n",
      "    fminbound(func, x1, x2, args=(), xtol=1e-05, maxfun=500, full_output=0, disp=1)\n",
      "        Bounded minimization for scalar functions.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        func : callable f(x,*args)\n",
      "            Objective function to be minimized (must accept and return scalars).\n",
      "        x1, x2 : float or array scalar\n",
      "            The optimization bounds.\n",
      "        args : tuple, optional\n",
      "            Extra arguments passed to function.\n",
      "        xtol : float, optional\n",
      "            The convergence tolerance.\n",
      "        maxfun : int, optional\n",
      "            Maximum number of function evaluations allowed.\n",
      "        full_output : bool, optional\n",
      "            If True, return optional outputs.\n",
      "        disp : int, optional\n",
      "            If non-zero, print messages.\n",
      "                0 : no message printing.\n",
      "                1 : non-convergence notification messages only.\n",
      "                2 : print a message on convergence too.\n",
      "                3 : print iteration results.\n",
      "        \n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        xopt : ndarray\n",
      "            Parameters (over given interval) which minimize the\n",
      "            objective function.\n",
      "        fval : number\n",
      "            The function value at the minimum point.\n",
      "        ierr : int\n",
      "            An error flag (0 if converged, 1 if maximum number of\n",
      "            function calls reached).\n",
      "        numfunc : int\n",
      "          The number of function calls made.\n",
      "        \n",
      "        See also\n",
      "        --------\n",
      "        minimize_scalar: Interface to minimization algorithms for scalar\n",
      "            univariate functions. See the 'Bounded' `method` in particular.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        Finds a local minimizer of the scalar function `func` in the\n",
      "        interval x1 < xopt < x2 using Brent's method. (See `brent`\n",
      "        for auto-bracketing.)\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        `fminbound` finds the minimum of the function in the given range.\n",
      "        The following examples illustrate the same\n",
      "        \n",
      "        >>> def f(x):\n",
      "        ...     return x**2\n",
      "        \n",
      "        >>> from scipy import optimize\n",
      "        \n",
      "        >>> minimum = optimize.fminbound(f, -1, 2)\n",
      "        >>> minimum\n",
      "        0.0\n",
      "        >>> minimum = optimize.fminbound(f, 1, 2)\n",
      "        >>> minimum\n",
      "        1.0000059608609866\n",
      "    \n",
      "    fsolve(func, x0, args=(), fprime=None, full_output=0, col_deriv=0, xtol=1.49012e-08, maxfev=0, band=None, epsfcn=None, factor=100, diag=None)\n",
      "        Find the roots of a function.\n",
      "        \n",
      "        Return the roots of the (non-linear) equations defined by\n",
      "        ``func(x) = 0`` given a starting estimate.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        func : callable ``f(x, *args)``\n",
      "            A function that takes at least one (possibly vector) argument,\n",
      "            and returns a value of the same length.\n",
      "        x0 : ndarray\n",
      "            The starting estimate for the roots of ``func(x) = 0``.\n",
      "        args : tuple, optional\n",
      "            Any extra arguments to `func`.\n",
      "        fprime : callable ``f(x, *args)``, optional\n",
      "            A function to compute the Jacobian of `func` with derivatives\n",
      "            across the rows. By default, the Jacobian will be estimated.\n",
      "        full_output : bool, optional\n",
      "            If True, return optional outputs.\n",
      "        col_deriv : bool, optional\n",
      "            Specify whether the Jacobian function computes derivatives down\n",
      "            the columns (faster, because there is no transpose operation).\n",
      "        xtol : float, optional\n",
      "            The calculation will terminate if the relative error between two\n",
      "            consecutive iterates is at most `xtol`.\n",
      "        maxfev : int, optional\n",
      "            The maximum number of calls to the function. If zero, then\n",
      "            ``100*(N+1)`` is the maximum where N is the number of elements\n",
      "            in `x0`.\n",
      "        band : tuple, optional\n",
      "            If set to a two-sequence containing the number of sub- and\n",
      "            super-diagonals within the band of the Jacobi matrix, the\n",
      "            Jacobi matrix is considered banded (only for ``fprime=None``).\n",
      "        epsfcn : float, optional\n",
      "            A suitable step length for the forward-difference\n",
      "            approximation of the Jacobian (for ``fprime=None``). If\n",
      "            `epsfcn` is less than the machine precision, it is assumed\n",
      "            that the relative errors in the functions are of the order of\n",
      "            the machine precision.\n",
      "        factor : float, optional\n",
      "            A parameter determining the initial step bound\n",
      "            (``factor * || diag * x||``). Should be in the interval\n",
      "            ``(0.1, 100)``.\n",
      "        diag : sequence, optional\n",
      "            N positive entries that serve as a scale factors for the\n",
      "            variables.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        x : ndarray\n",
      "            The solution (or the result of the last iteration for\n",
      "            an unsuccessful call).\n",
      "        infodict : dict\n",
      "            A dictionary of optional outputs with the keys:\n",
      "        \n",
      "            ``nfev``\n",
      "                number of function calls\n",
      "            ``njev``\n",
      "                number of Jacobian calls\n",
      "            ``fvec``\n",
      "                function evaluated at the output\n",
      "            ``fjac``\n",
      "                the orthogonal matrix, q, produced by the QR\n",
      "                factorization of the final approximate Jacobian\n",
      "                matrix, stored column wise\n",
      "            ``r``\n",
      "                upper triangular matrix produced by QR factorization\n",
      "                of the same matrix\n",
      "            ``qtf``\n",
      "                the vector ``(transpose(q) * fvec)``\n",
      "        \n",
      "        ier : int\n",
      "            An integer flag.  Set to 1 if a solution was found, otherwise refer\n",
      "            to `mesg` for more information.\n",
      "        mesg : str\n",
      "            If no solution is found, `mesg` details the cause of failure.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        root : Interface to root finding algorithms for multivariate\n",
      "               functions. See the ``method=='hybr'`` in particular.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        ``fsolve`` is a wrapper around MINPACK's hybrd and hybrj algorithms.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        Find a solution to the system of equations:\n",
      "        ``x0*cos(x1) = 4,  x1*x0 - x1 = 5``.\n",
      "        \n",
      "        >>> from scipy.optimize import fsolve\n",
      "        >>> def func(x):\n",
      "        ...     return [x[0] * np.cos(x[1]) - 4,\n",
      "        ...             x[1] * x[0] - x[1] - 5]\n",
      "        >>> root = fsolve(func, [1, 1])\n",
      "        >>> root\n",
      "        array([6.50409711, 0.90841421])\n",
      "        >>> np.isclose(func(root), [0.0, 0.0])  # func(root) should be almost 0.0.\n",
      "        array([ True,  True])\n",
      "    \n",
      "    golden(func, args=(), brack=None, tol=1.4901161193847656e-08, full_output=0, maxiter=5000)\n",
      "        Return the minimum of a function of one variable using golden section\n",
      "        method.\n",
      "        \n",
      "        Given a function of one variable and a possible bracketing interval,\n",
      "        return the minimum of the function isolated to a fractional precision of\n",
      "        tol.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        func : callable func(x,*args)\n",
      "            Objective function to minimize.\n",
      "        args : tuple, optional\n",
      "            Additional arguments (if present), passed to func.\n",
      "        brack : tuple, optional\n",
      "            Triple (a,b,c), where (a<b<c) and func(b) <\n",
      "            func(a),func(c). If bracket consists of two numbers (a,\n",
      "            c), then they are assumed to be a starting interval for a\n",
      "            downhill bracket search (see `bracket`); it doesn't always\n",
      "            mean that obtained solution will satisfy a<=x<=c.\n",
      "        tol : float, optional\n",
      "            x tolerance stop criterion\n",
      "        full_output : bool, optional\n",
      "            If True, return optional outputs.\n",
      "        maxiter : int\n",
      "            Maximum number of iterations to perform.\n",
      "        \n",
      "        See also\n",
      "        --------\n",
      "        minimize_scalar: Interface to minimization algorithms for scalar\n",
      "            univariate functions. See the 'Golden' `method` in particular.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        Uses analog of bisection method to decrease the bracketed\n",
      "        interval.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        We illustrate the behaviour of the function when `brack` is of\n",
      "        size 2 and 3, respectively. In the case where `brack` is of the\n",
      "        form (xa,xb), we can see for the given values, the output need\n",
      "        not necessarily lie in the range ``(xa, xb)``.\n",
      "        \n",
      "        >>> def f(x):\n",
      "        ...     return x**2\n",
      "        \n",
      "        >>> from scipy import optimize\n",
      "        \n",
      "        >>> minimum = optimize.golden(f, brack=(1, 2))\n",
      "        >>> minimum\n",
      "        1.5717277788484873e-162\n",
      "        >>> minimum = optimize.golden(f, brack=(-1, 0.5, 2))\n",
      "        >>> minimum\n",
      "        -1.5717277788484873e-162\n",
      "    \n",
      "    least_squares(fun, x0, jac='2-point', bounds=(-inf, inf), method='trf', ftol=1e-08, xtol=1e-08, gtol=1e-08, x_scale=1.0, loss='linear', f_scale=1.0, diff_step=None, tr_solver=None, tr_options={}, jac_sparsity=None, max_nfev=None, verbose=0, args=(), kwargs={})\n",
      "        Solve a nonlinear least-squares problem with bounds on the variables.\n",
      "        \n",
      "        Given the residuals f(x) (an m-D real function of n real\n",
      "        variables) and the loss function rho(s) (a scalar function), `least_squares`\n",
      "        finds a local minimum of the cost function F(x)::\n",
      "        \n",
      "            minimize F(x) = 0.5 * sum(rho(f_i(x)**2), i = 0, ..., m - 1)\n",
      "            subject to lb <= x <= ub\n",
      "        \n",
      "        The purpose of the loss function rho(s) is to reduce the influence of\n",
      "        outliers on the solution.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        fun : callable\n",
      "            Function which computes the vector of residuals, with the signature\n",
      "            ``fun(x, *args, **kwargs)``, i.e., the minimization proceeds with\n",
      "            respect to its first argument. The argument ``x`` passed to this\n",
      "            function is an ndarray of shape (n,) (never a scalar, even for n=1).\n",
      "            It must allocate and return a 1-D array_like of shape (m,) or a scalar.\n",
      "            If the argument ``x`` is complex or the function ``fun`` returns\n",
      "            complex residuals, it must be wrapped in a real function of real\n",
      "            arguments, as shown at the end of the Examples section.\n",
      "        x0 : array_like with shape (n,) or float\n",
      "            Initial guess on independent variables. If float, it will be treated\n",
      "            as a 1-D array with one element.\n",
      "        jac : {'2-point', '3-point', 'cs', callable}, optional\n",
      "            Method of computing the Jacobian matrix (an m-by-n matrix, where\n",
      "            element (i, j) is the partial derivative of f[i] with respect to\n",
      "            x[j]). The keywords select a finite difference scheme for numerical\n",
      "            estimation. The scheme '3-point' is more accurate, but requires\n",
      "            twice as many operations as '2-point' (default). The scheme 'cs'\n",
      "            uses complex steps, and while potentially the most accurate, it is\n",
      "            applicable only when `fun` correctly handles complex inputs and\n",
      "            can be analytically continued to the complex plane. Method 'lm'\n",
      "            always uses the '2-point' scheme. If callable, it is used as\n",
      "            ``jac(x, *args, **kwargs)`` and should return a good approximation\n",
      "            (or the exact value) for the Jacobian as an array_like (np.atleast_2d\n",
      "            is applied), a sparse matrix (csr_matrix preferred for performance) or\n",
      "            a `scipy.sparse.linalg.LinearOperator`.\n",
      "        bounds : 2-tuple of array_like, optional\n",
      "            Lower and upper bounds on independent variables. Defaults to no bounds.\n",
      "            Each array must match the size of `x0` or be a scalar, in the latter\n",
      "            case a bound will be the same for all variables. Use ``np.inf`` with\n",
      "            an appropriate sign to disable bounds on all or some variables.\n",
      "        method : {'trf', 'dogbox', 'lm'}, optional\n",
      "            Algorithm to perform minimization.\n",
      "        \n",
      "                * 'trf' : Trust Region Reflective algorithm, particularly suitable\n",
      "                  for large sparse problems with bounds. Generally robust method.\n",
      "                * 'dogbox' : dogleg algorithm with rectangular trust regions,\n",
      "                  typical use case is small problems with bounds. Not recommended\n",
      "                  for problems with rank-deficient Jacobian.\n",
      "                * 'lm' : Levenberg-Marquardt algorithm as implemented in MINPACK.\n",
      "                  Doesn't handle bounds and sparse Jacobians. Usually the most\n",
      "                  efficient method for small unconstrained problems.\n",
      "        \n",
      "            Default is 'trf'. See Notes for more information.\n",
      "        ftol : float or None, optional\n",
      "            Tolerance for termination by the change of the cost function. Default\n",
      "            is 1e-8. The optimization process is stopped when ``dF < ftol * F``,\n",
      "            and there was an adequate agreement between a local quadratic model and\n",
      "            the true model in the last step.\n",
      "        \n",
      "            If None and 'method' is not 'lm', the termination by this condition is\n",
      "            disabled. If 'method' is 'lm', this tolerance must be higher than\n",
      "            machine epsilon.\n",
      "        xtol : float or None, optional\n",
      "            Tolerance for termination by the change of the independent variables.\n",
      "            Default is 1e-8. The exact condition depends on the `method` used:\n",
      "        \n",
      "                * For 'trf' and 'dogbox' : ``norm(dx) < xtol * (xtol + norm(x))``.\n",
      "                * For 'lm' : ``Delta < xtol * norm(xs)``, where ``Delta`` is\n",
      "                  a trust-region radius and ``xs`` is the value of ``x``\n",
      "                  scaled according to `x_scale` parameter (see below).\n",
      "        \n",
      "            If None and 'method' is not 'lm', the termination by this condition is\n",
      "            disabled. If 'method' is 'lm', this tolerance must be higher than\n",
      "            machine epsilon.\n",
      "        gtol : float or None, optional\n",
      "            Tolerance for termination by the norm of the gradient. Default is 1e-8.\n",
      "            The exact condition depends on a `method` used:\n",
      "        \n",
      "                * For 'trf' : ``norm(g_scaled, ord=np.inf) < gtol``, where\n",
      "                  ``g_scaled`` is the value of the gradient scaled to account for\n",
      "                  the presence of the bounds [STIR]_.\n",
      "                * For 'dogbox' : ``norm(g_free, ord=np.inf) < gtol``, where\n",
      "                  ``g_free`` is the gradient with respect to the variables which\n",
      "                  are not in the optimal state on the boundary.\n",
      "                * For 'lm' : the maximum absolute value of the cosine of angles\n",
      "                  between columns of the Jacobian and the residual vector is less\n",
      "                  than `gtol`, or the residual vector is zero.\n",
      "        \n",
      "            If None and 'method' is not 'lm', the termination by this condition is\n",
      "            disabled. If 'method' is 'lm', this tolerance must be higher than\n",
      "            machine epsilon.\n",
      "        x_scale : array_like or 'jac', optional\n",
      "            Characteristic scale of each variable. Setting `x_scale` is equivalent\n",
      "            to reformulating the problem in scaled variables ``xs = x / x_scale``.\n",
      "            An alternative view is that the size of a trust region along jth\n",
      "            dimension is proportional to ``x_scale[j]``. Improved convergence may\n",
      "            be achieved by setting `x_scale` such that a step of a given size\n",
      "            along any of the scaled variables has a similar effect on the cost\n",
      "            function. If set to 'jac', the scale is iteratively updated using the\n",
      "            inverse norms of the columns of the Jacobian matrix (as described in\n",
      "            [JJMore]_).\n",
      "        loss : str or callable, optional\n",
      "            Determines the loss function. The following keyword values are allowed:\n",
      "        \n",
      "                * 'linear' (default) : ``rho(z) = z``. Gives a standard\n",
      "                  least-squares problem.\n",
      "                * 'soft_l1' : ``rho(z) = 2 * ((1 + z)**0.5 - 1)``. The smooth\n",
      "                  approximation of l1 (absolute value) loss. Usually a good\n",
      "                  choice for robust least squares.\n",
      "                * 'huber' : ``rho(z) = z if z <= 1 else 2*z**0.5 - 1``. Works\n",
      "                  similarly to 'soft_l1'.\n",
      "                * 'cauchy' : ``rho(z) = ln(1 + z)``. Severely weakens outliers\n",
      "                  influence, but may cause difficulties in optimization process.\n",
      "                * 'arctan' : ``rho(z) = arctan(z)``. Limits a maximum loss on\n",
      "                  a single residual, has properties similar to 'cauchy'.\n",
      "        \n",
      "            If callable, it must take a 1-D ndarray ``z=f**2`` and return an\n",
      "            array_like with shape (3, m) where row 0 contains function values,\n",
      "            row 1 contains first derivatives and row 2 contains second\n",
      "            derivatives. Method 'lm' supports only 'linear' loss.\n",
      "        f_scale : float, optional\n",
      "            Value of soft margin between inlier and outlier residuals, default\n",
      "            is 1.0. The loss function is evaluated as follows\n",
      "            ``rho_(f**2) = C**2 * rho(f**2 / C**2)``, where ``C`` is `f_scale`,\n",
      "            and ``rho`` is determined by `loss` parameter. This parameter has\n",
      "            no effect with ``loss='linear'``, but for other `loss` values it is\n",
      "            of crucial importance.\n",
      "        max_nfev : None or int, optional\n",
      "            Maximum number of function evaluations before the termination.\n",
      "            If None (default), the value is chosen automatically:\n",
      "        \n",
      "                * For 'trf' and 'dogbox' : 100 * n.\n",
      "                * For 'lm' :  100 * n if `jac` is callable and 100 * n * (n + 1)\n",
      "                  otherwise (because 'lm' counts function calls in Jacobian\n",
      "                  estimation).\n",
      "        \n",
      "        diff_step : None or array_like, optional\n",
      "            Determines the relative step size for the finite difference\n",
      "            approximation of the Jacobian. The actual step is computed as\n",
      "            ``x * diff_step``. If None (default), then `diff_step` is taken to be\n",
      "            a conventional \"optimal\" power of machine epsilon for the finite\n",
      "            difference scheme used [NR]_.\n",
      "        tr_solver : {None, 'exact', 'lsmr'}, optional\n",
      "            Method for solving trust-region subproblems, relevant only for 'trf'\n",
      "            and 'dogbox' methods.\n",
      "        \n",
      "                * 'exact' is suitable for not very large problems with dense\n",
      "                  Jacobian matrices. The computational complexity per iteration is\n",
      "                  comparable to a singular value decomposition of the Jacobian\n",
      "                  matrix.\n",
      "                * 'lsmr' is suitable for problems with sparse and large Jacobian\n",
      "                  matrices. It uses the iterative procedure\n",
      "                  `scipy.sparse.linalg.lsmr` for finding a solution of a linear\n",
      "                  least-squares problem and only requires matrix-vector product\n",
      "                  evaluations.\n",
      "        \n",
      "            If None (default), the solver is chosen based on the type of Jacobian\n",
      "            returned on the first iteration.\n",
      "        tr_options : dict, optional\n",
      "            Keyword options passed to trust-region solver.\n",
      "        \n",
      "                * ``tr_solver='exact'``: `tr_options` are ignored.\n",
      "                * ``tr_solver='lsmr'``: options for `scipy.sparse.linalg.lsmr`.\n",
      "                  Additionally,  ``method='trf'`` supports  'regularize' option\n",
      "                  (bool, default is True), which adds a regularization term to the\n",
      "                  normal equation, which improves convergence if the Jacobian is\n",
      "                  rank-deficient [Byrd]_ (eq. 3.4).\n",
      "        \n",
      "        jac_sparsity : {None, array_like, sparse matrix}, optional\n",
      "            Defines the sparsity structure of the Jacobian matrix for finite\n",
      "            difference estimation, its shape must be (m, n). If the Jacobian has\n",
      "            only few non-zero elements in *each* row, providing the sparsity\n",
      "            structure will greatly speed up the computations [Curtis]_. A zero\n",
      "            entry means that a corresponding element in the Jacobian is identically\n",
      "            zero. If provided, forces the use of 'lsmr' trust-region solver.\n",
      "            If None (default), then dense differencing will be used. Has no effect\n",
      "            for 'lm' method.\n",
      "        verbose : {0, 1, 2}, optional\n",
      "            Level of algorithm's verbosity:\n",
      "        \n",
      "                * 0 (default) : work silently.\n",
      "                * 1 : display a termination report.\n",
      "                * 2 : display progress during iterations (not supported by 'lm'\n",
      "                  method).\n",
      "        \n",
      "        args, kwargs : tuple and dict, optional\n",
      "            Additional arguments passed to `fun` and `jac`. Both empty by default.\n",
      "            The calling signature is ``fun(x, *args, **kwargs)`` and the same for\n",
      "            `jac`.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        result : OptimizeResult\n",
      "            `OptimizeResult` with the following fields defined:\n",
      "        \n",
      "                x : ndarray, shape (n,)\n",
      "                    Solution found.\n",
      "                cost : float\n",
      "                    Value of the cost function at the solution.\n",
      "                fun : ndarray, shape (m,)\n",
      "                    Vector of residuals at the solution.\n",
      "                jac : ndarray, sparse matrix or LinearOperator, shape (m, n)\n",
      "                    Modified Jacobian matrix at the solution, in the sense that J^T J\n",
      "                    is a Gauss-Newton approximation of the Hessian of the cost function.\n",
      "                    The type is the same as the one used by the algorithm.\n",
      "                grad : ndarray, shape (m,)\n",
      "                    Gradient of the cost function at the solution.\n",
      "                optimality : float\n",
      "                    First-order optimality measure. In unconstrained problems, it is\n",
      "                    always the uniform norm of the gradient. In constrained problems,\n",
      "                    it is the quantity which was compared with `gtol` during iterations.\n",
      "                active_mask : ndarray of int, shape (n,)\n",
      "                    Each component shows whether a corresponding constraint is active\n",
      "                    (that is, whether a variable is at the bound):\n",
      "        \n",
      "                        *  0 : a constraint is not active.\n",
      "                        * -1 : a lower bound is active.\n",
      "                        *  1 : an upper bound is active.\n",
      "        \n",
      "                    Might be somewhat arbitrary for 'trf' method as it generates a\n",
      "                    sequence of strictly feasible iterates and `active_mask` is\n",
      "                    determined within a tolerance threshold.\n",
      "                nfev : int\n",
      "                    Number of function evaluations done. Methods 'trf' and 'dogbox' do\n",
      "                    not count function calls for numerical Jacobian approximation, as\n",
      "                    opposed to 'lm' method.\n",
      "                njev : int or None\n",
      "                    Number of Jacobian evaluations done. If numerical Jacobian\n",
      "                    approximation is used in 'lm' method, it is set to None.\n",
      "                status : int\n",
      "                    The reason for algorithm termination:\n",
      "        \n",
      "                        * -1 : improper input parameters status returned from MINPACK.\n",
      "                        *  0 : the maximum number of function evaluations is exceeded.\n",
      "                        *  1 : `gtol` termination condition is satisfied.\n",
      "                        *  2 : `ftol` termination condition is satisfied.\n",
      "                        *  3 : `xtol` termination condition is satisfied.\n",
      "                        *  4 : Both `ftol` and `xtol` termination conditions are satisfied.\n",
      "        \n",
      "                message : str\n",
      "                    Verbal description of the termination reason.\n",
      "                success : bool\n",
      "                    True if one of the convergence criteria is satisfied (`status` > 0).\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        leastsq : A legacy wrapper for the MINPACK implementation of the\n",
      "                  Levenberg-Marquadt algorithm.\n",
      "        curve_fit : Least-squares minimization applied to a curve-fitting problem.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        Method 'lm' (Levenberg-Marquardt) calls a wrapper over least-squares\n",
      "        algorithms implemented in MINPACK (lmder, lmdif). It runs the\n",
      "        Levenberg-Marquardt algorithm formulated as a trust-region type algorithm.\n",
      "        The implementation is based on paper [JJMore]_, it is very robust and\n",
      "        efficient with a lot of smart tricks. It should be your first choice\n",
      "        for unconstrained problems. Note that it doesn't support bounds. Also,\n",
      "        it doesn't work when m < n.\n",
      "        \n",
      "        Method 'trf' (Trust Region Reflective) is motivated by the process of\n",
      "        solving a system of equations, which constitute the first-order optimality\n",
      "        condition for a bound-constrained minimization problem as formulated in\n",
      "        [STIR]_. The algorithm iteratively solves trust-region subproblems\n",
      "        augmented by a special diagonal quadratic term and with trust-region shape\n",
      "        determined by the distance from the bounds and the direction of the\n",
      "        gradient. This enhancements help to avoid making steps directly into bounds\n",
      "        and efficiently explore the whole space of variables. To further improve\n",
      "        convergence, the algorithm considers search directions reflected from the\n",
      "        bounds. To obey theoretical requirements, the algorithm keeps iterates\n",
      "        strictly feasible. With dense Jacobians trust-region subproblems are\n",
      "        solved by an exact method very similar to the one described in [JJMore]_\n",
      "        (and implemented in MINPACK). The difference from the MINPACK\n",
      "        implementation is that a singular value decomposition of a Jacobian\n",
      "        matrix is done once per iteration, instead of a QR decomposition and series\n",
      "        of Givens rotation eliminations. For large sparse Jacobians a 2-D subspace\n",
      "        approach of solving trust-region subproblems is used [STIR]_, [Byrd]_.\n",
      "        The subspace is spanned by a scaled gradient and an approximate\n",
      "        Gauss-Newton solution delivered by `scipy.sparse.linalg.lsmr`. When no\n",
      "        constraints are imposed the algorithm is very similar to MINPACK and has\n",
      "        generally comparable performance. The algorithm works quite robust in\n",
      "        unbounded and bounded problems, thus it is chosen as a default algorithm.\n",
      "        \n",
      "        Method 'dogbox' operates in a trust-region framework, but considers\n",
      "        rectangular trust regions as opposed to conventional ellipsoids [Voglis]_.\n",
      "        The intersection of a current trust region and initial bounds is again\n",
      "        rectangular, so on each iteration a quadratic minimization problem subject\n",
      "        to bound constraints is solved approximately by Powell's dogleg method\n",
      "        [NumOpt]_. The required Gauss-Newton step can be computed exactly for\n",
      "        dense Jacobians or approximately by `scipy.sparse.linalg.lsmr` for large\n",
      "        sparse Jacobians. The algorithm is likely to exhibit slow convergence when\n",
      "        the rank of Jacobian is less than the number of variables. The algorithm\n",
      "        often outperforms 'trf' in bounded problems with a small number of\n",
      "        variables.\n",
      "        \n",
      "        Robust loss functions are implemented as described in [BA]_. The idea\n",
      "        is to modify a residual vector and a Jacobian matrix on each iteration\n",
      "        such that computed gradient and Gauss-Newton Hessian approximation match\n",
      "        the true gradient and Hessian approximation of the cost function. Then\n",
      "        the algorithm proceeds in a normal way, i.e., robust loss functions are\n",
      "        implemented as a simple wrapper over standard least-squares algorithms.\n",
      "        \n",
      "        .. versionadded:: 0.17.0\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [STIR] M. A. Branch, T. F. Coleman, and Y. Li, \"A Subspace, Interior,\n",
      "                  and Conjugate Gradient Method for Large-Scale Bound-Constrained\n",
      "                  Minimization Problems,\" SIAM Journal on Scientific Computing,\n",
      "                  Vol. 21, Number 1, pp 1-23, 1999.\n",
      "        .. [NR] William H. Press et. al., \"Numerical Recipes. The Art of Scientific\n",
      "                Computing. 3rd edition\", Sec. 5.7.\n",
      "        .. [Byrd] R. H. Byrd, R. B. Schnabel and G. A. Shultz, \"Approximate\n",
      "                  solution of the trust region problem by minimization over\n",
      "                  two-dimensional subspaces\", Math. Programming, 40, pp. 247-263,\n",
      "                  1988.\n",
      "        .. [Curtis] A. Curtis, M. J. D. Powell, and J. Reid, \"On the estimation of\n",
      "                    sparse Jacobian matrices\", Journal of the Institute of\n",
      "                    Mathematics and its Applications, 13, pp. 117-120, 1974.\n",
      "        .. [JJMore] J. J. More, \"The Levenberg-Marquardt Algorithm: Implementation\n",
      "                    and Theory,\" Numerical Analysis, ed. G. A. Watson, Lecture\n",
      "                    Notes in Mathematics 630, Springer Verlag, pp. 105-116, 1977.\n",
      "        .. [Voglis] C. Voglis and I. E. Lagaris, \"A Rectangular Trust Region\n",
      "                    Dogleg Approach for Unconstrained and Bound Constrained\n",
      "                    Nonlinear Optimization\", WSEAS International Conference on\n",
      "                    Applied Mathematics, Corfu, Greece, 2004.\n",
      "        .. [NumOpt] J. Nocedal and S. J. Wright, \"Numerical optimization,\n",
      "                    2nd edition\", Chapter 4.\n",
      "        .. [BA] B. Triggs et. al., \"Bundle Adjustment - A Modern Synthesis\",\n",
      "                Proceedings of the International Workshop on Vision Algorithms:\n",
      "                Theory and Practice, pp. 298-372, 1999.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        In this example we find a minimum of the Rosenbrock function without bounds\n",
      "        on independent variables.\n",
      "        \n",
      "        >>> def fun_rosenbrock(x):\n",
      "        ...     return np.array([10 * (x[1] - x[0]**2), (1 - x[0])])\n",
      "        \n",
      "        Notice that we only provide the vector of the residuals. The algorithm\n",
      "        constructs the cost function as a sum of squares of the residuals, which\n",
      "        gives the Rosenbrock function. The exact minimum is at ``x = [1.0, 1.0]``.\n",
      "        \n",
      "        >>> from scipy.optimize import least_squares\n",
      "        >>> x0_rosenbrock = np.array([2, 2])\n",
      "        >>> res_1 = least_squares(fun_rosenbrock, x0_rosenbrock)\n",
      "        >>> res_1.x\n",
      "        array([ 1.,  1.])\n",
      "        >>> res_1.cost\n",
      "        9.8669242910846867e-30\n",
      "        >>> res_1.optimality\n",
      "        8.8928864934219529e-14\n",
      "        \n",
      "        We now constrain the variables, in such a way that the previous solution\n",
      "        becomes infeasible. Specifically, we require that ``x[1] >= 1.5``, and\n",
      "        ``x[0]`` left unconstrained. To this end, we specify the `bounds` parameter\n",
      "        to `least_squares` in the form ``bounds=([-np.inf, 1.5], np.inf)``.\n",
      "        \n",
      "        We also provide the analytic Jacobian:\n",
      "        \n",
      "        >>> def jac_rosenbrock(x):\n",
      "        ...     return np.array([\n",
      "        ...         [-20 * x[0], 10],\n",
      "        ...         [-1, 0]])\n",
      "        \n",
      "        Putting this all together, we see that the new solution lies on the bound:\n",
      "        \n",
      "        >>> res_2 = least_squares(fun_rosenbrock, x0_rosenbrock, jac_rosenbrock,\n",
      "        ...                       bounds=([-np.inf, 1.5], np.inf))\n",
      "        >>> res_2.x\n",
      "        array([ 1.22437075,  1.5       ])\n",
      "        >>> res_2.cost\n",
      "        0.025213093946805685\n",
      "        >>> res_2.optimality\n",
      "        1.5885401433157753e-07\n",
      "        \n",
      "        Now we solve a system of equations (i.e., the cost function should be zero\n",
      "        at a minimum) for a Broyden tridiagonal vector-valued function of 100000\n",
      "        variables:\n",
      "        \n",
      "        >>> def fun_broyden(x):\n",
      "        ...     f = (3 - x) * x + 1\n",
      "        ...     f[1:] -= x[:-1]\n",
      "        ...     f[:-1] -= 2 * x[1:]\n",
      "        ...     return f\n",
      "        \n",
      "        The corresponding Jacobian matrix is sparse. We tell the algorithm to\n",
      "        estimate it by finite differences and provide the sparsity structure of\n",
      "        Jacobian to significantly speed up this process.\n",
      "        \n",
      "        >>> from scipy.sparse import lil_matrix\n",
      "        >>> def sparsity_broyden(n):\n",
      "        ...     sparsity = lil_matrix((n, n), dtype=int)\n",
      "        ...     i = np.arange(n)\n",
      "        ...     sparsity[i, i] = 1\n",
      "        ...     i = np.arange(1, n)\n",
      "        ...     sparsity[i, i - 1] = 1\n",
      "        ...     i = np.arange(n - 1)\n",
      "        ...     sparsity[i, i + 1] = 1\n",
      "        ...     return sparsity\n",
      "        ...\n",
      "        >>> n = 100000\n",
      "        >>> x0_broyden = -np.ones(n)\n",
      "        ...\n",
      "        >>> res_3 = least_squares(fun_broyden, x0_broyden,\n",
      "        ...                       jac_sparsity=sparsity_broyden(n))\n",
      "        >>> res_3.cost\n",
      "        4.5687069299604613e-23\n",
      "        >>> res_3.optimality\n",
      "        1.1650454296851518e-11\n",
      "        \n",
      "        Let's also solve a curve fitting problem using robust loss function to\n",
      "        take care of outliers in the data. Define the model function as\n",
      "        ``y = a + b * exp(c * t)``, where t is a predictor variable, y is an\n",
      "        observation and a, b, c are parameters to estimate.\n",
      "        \n",
      "        First, define the function which generates the data with noise and\n",
      "        outliers, define the model parameters, and generate data:\n",
      "        \n",
      "        >>> from numpy.random import default_rng\n",
      "        >>> rng = default_rng()\n",
      "        >>> def gen_data(t, a, b, c, noise=0., n_outliers=0, seed=None):\n",
      "        ...     rng = default_rng(seed)\n",
      "        ...\n",
      "        ...     y = a + b * np.exp(t * c)\n",
      "        ...\n",
      "        ...     error = noise * rng.standard_normal(t.size)\n",
      "        ...     outliers = rng.integers(0, t.size, n_outliers)\n",
      "        ...     error[outliers] *= 10\n",
      "        ...\n",
      "        ...     return y + error\n",
      "        ...\n",
      "        >>> a = 0.5\n",
      "        >>> b = 2.0\n",
      "        >>> c = -1\n",
      "        >>> t_min = 0\n",
      "        >>> t_max = 10\n",
      "        >>> n_points = 15\n",
      "        ...\n",
      "        >>> t_train = np.linspace(t_min, t_max, n_points)\n",
      "        >>> y_train = gen_data(t_train, a, b, c, noise=0.1, n_outliers=3)\n",
      "        \n",
      "        Define function for computing residuals and initial estimate of\n",
      "        parameters.\n",
      "        \n",
      "        >>> def fun(x, t, y):\n",
      "        ...     return x[0] + x[1] * np.exp(x[2] * t) - y\n",
      "        ...\n",
      "        >>> x0 = np.array([1.0, 1.0, 0.0])\n",
      "        \n",
      "        Compute a standard least-squares solution:\n",
      "        \n",
      "        >>> res_lsq = least_squares(fun, x0, args=(t_train, y_train))\n",
      "        \n",
      "        Now compute two solutions with two different robust loss functions. The\n",
      "        parameter `f_scale` is set to 0.1, meaning that inlier residuals should\n",
      "        not significantly exceed 0.1 (the noise level used).\n",
      "        \n",
      "        >>> res_soft_l1 = least_squares(fun, x0, loss='soft_l1', f_scale=0.1,\n",
      "        ...                             args=(t_train, y_train))\n",
      "        >>> res_log = least_squares(fun, x0, loss='cauchy', f_scale=0.1,\n",
      "        ...                         args=(t_train, y_train))\n",
      "        \n",
      "        And, finally, plot all the curves. We see that by selecting an appropriate\n",
      "        `loss`  we can get estimates close to optimal even in the presence of\n",
      "        strong outliers. But keep in mind that generally it is recommended to try\n",
      "        'soft_l1' or 'huber' losses first (if at all necessary) as the other two\n",
      "        options may cause difficulties in optimization process.\n",
      "        \n",
      "        >>> t_test = np.linspace(t_min, t_max, n_points * 10)\n",
      "        >>> y_true = gen_data(t_test, a, b, c)\n",
      "        >>> y_lsq = gen_data(t_test, *res_lsq.x)\n",
      "        >>> y_soft_l1 = gen_data(t_test, *res_soft_l1.x)\n",
      "        >>> y_log = gen_data(t_test, *res_log.x)\n",
      "        ...\n",
      "        >>> import matplotlib.pyplot as plt\n",
      "        >>> plt.plot(t_train, y_train, 'o')\n",
      "        >>> plt.plot(t_test, y_true, 'k', linewidth=2, label='true')\n",
      "        >>> plt.plot(t_test, y_lsq, label='linear loss')\n",
      "        >>> plt.plot(t_test, y_soft_l1, label='soft_l1 loss')\n",
      "        >>> plt.plot(t_test, y_log, label='cauchy loss')\n",
      "        >>> plt.xlabel(\"t\")\n",
      "        >>> plt.ylabel(\"y\")\n",
      "        >>> plt.legend()\n",
      "        >>> plt.show()\n",
      "        \n",
      "        In the next example, we show how complex-valued residual functions of\n",
      "        complex variables can be optimized with ``least_squares()``. Consider the\n",
      "        following function:\n",
      "        \n",
      "        >>> def f(z):\n",
      "        ...     return z - (0.5 + 0.5j)\n",
      "        \n",
      "        We wrap it into a function of real variables that returns real residuals\n",
      "        by simply handling the real and imaginary parts as independent variables:\n",
      "        \n",
      "        >>> def f_wrap(x):\n",
      "        ...     fx = f(x[0] + 1j*x[1])\n",
      "        ...     return np.array([fx.real, fx.imag])\n",
      "        \n",
      "        Thus, instead of the original m-D complex function of n complex\n",
      "        variables we optimize a 2m-D real function of 2n real variables:\n",
      "        \n",
      "        >>> from scipy.optimize import least_squares\n",
      "        >>> res_wrapped = least_squares(f_wrap, (0.1, 0.1), bounds=([0, 0], [1, 1]))\n",
      "        >>> z = res_wrapped.x[0] + res_wrapped.x[1]*1j\n",
      "        >>> z\n",
      "        (0.49999999999925893+0.49999999999925893j)\n",
      "    \n",
      "    leastsq(func, x0, args=(), Dfun=None, full_output=0, col_deriv=0, ftol=1.49012e-08, xtol=1.49012e-08, gtol=0.0, maxfev=0, epsfcn=None, factor=100, diag=None)\n",
      "        Minimize the sum of squares of a set of equations.\n",
      "        \n",
      "        ::\n",
      "        \n",
      "            x = arg min(sum(func(y)**2,axis=0))\n",
      "                     y\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        func : callable\n",
      "            Should take at least one (possibly length ``N`` vector) argument and\n",
      "            returns ``M`` floating point numbers. It must not return NaNs or\n",
      "            fitting might fail. ``M`` must be greater than or equal to ``N``.\n",
      "        x0 : ndarray\n",
      "            The starting estimate for the minimization.\n",
      "        args : tuple, optional\n",
      "            Any extra arguments to func are placed in this tuple.\n",
      "        Dfun : callable, optional\n",
      "            A function or method to compute the Jacobian of func with derivatives\n",
      "            across the rows. If this is None, the Jacobian will be estimated.\n",
      "        full_output : bool, optional\n",
      "            non-zero to return all optional outputs.\n",
      "        col_deriv : bool, optional\n",
      "            non-zero to specify that the Jacobian function computes derivatives\n",
      "            down the columns (faster, because there is no transpose operation).\n",
      "        ftol : float, optional\n",
      "            Relative error desired in the sum of squares.\n",
      "        xtol : float, optional\n",
      "            Relative error desired in the approximate solution.\n",
      "        gtol : float, optional\n",
      "            Orthogonality desired between the function vector and the columns of\n",
      "            the Jacobian.\n",
      "        maxfev : int, optional\n",
      "            The maximum number of calls to the function. If `Dfun` is provided,\n",
      "            then the default `maxfev` is 100*(N+1) where N is the number of elements\n",
      "            in x0, otherwise the default `maxfev` is 200*(N+1).\n",
      "        epsfcn : float, optional\n",
      "            A variable used in determining a suitable step length for the forward-\n",
      "            difference approximation of the Jacobian (for Dfun=None).\n",
      "            Normally the actual step length will be sqrt(epsfcn)*x\n",
      "            If epsfcn is less than the machine precision, it is assumed that the\n",
      "            relative errors are of the order of the machine precision.\n",
      "        factor : float, optional\n",
      "            A parameter determining the initial step bound\n",
      "            (``factor * || diag * x||``). Should be in interval ``(0.1, 100)``.\n",
      "        diag : sequence, optional\n",
      "            N positive entries that serve as a scale factors for the variables.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        x : ndarray\n",
      "            The solution (or the result of the last iteration for an unsuccessful\n",
      "            call).\n",
      "        cov_x : ndarray\n",
      "            The inverse of the Hessian. `fjac` and `ipvt` are used to construct an\n",
      "            estimate of the Hessian. A value of None indicates a singular matrix,\n",
      "            which means the curvature in parameters `x` is numerically flat. To\n",
      "            obtain the covariance matrix of the parameters `x`, `cov_x` must be\n",
      "            multiplied by the variance of the residuals -- see curve_fit.\n",
      "        infodict : dict\n",
      "            a dictionary of optional outputs with the keys:\n",
      "        \n",
      "            ``nfev``\n",
      "                The number of function calls\n",
      "            ``fvec``\n",
      "                The function evaluated at the output\n",
      "            ``fjac``\n",
      "                A permutation of the R matrix of a QR\n",
      "                factorization of the final approximate\n",
      "                Jacobian matrix, stored column wise.\n",
      "                Together with ipvt, the covariance of the\n",
      "                estimate can be approximated.\n",
      "            ``ipvt``\n",
      "                An integer array of length N which defines\n",
      "                a permutation matrix, p, such that\n",
      "                fjac*p = q*r, where r is upper triangular\n",
      "                with diagonal elements of nonincreasing\n",
      "                magnitude. Column j of p is column ipvt(j)\n",
      "                of the identity matrix.\n",
      "            ``qtf``\n",
      "                The vector (transpose(q) * fvec).\n",
      "        \n",
      "        mesg : str\n",
      "            A string message giving information about the cause of failure.\n",
      "        ier : int\n",
      "            An integer flag. If it is equal to 1, 2, 3 or 4, the solution was\n",
      "            found. Otherwise, the solution was not found. In either case, the\n",
      "            optional output variable 'mesg' gives more information.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        least_squares : Newer interface to solve nonlinear least-squares problems\n",
      "            with bounds on the variables. See ``method=='lm'`` in particular.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        \"leastsq\" is a wrapper around MINPACK's lmdif and lmder algorithms.\n",
      "        \n",
      "        cov_x is a Jacobian approximation to the Hessian of the least squares\n",
      "        objective function.\n",
      "        This approximation assumes that the objective function is based on the\n",
      "        difference between some observed target data (ydata) and a (non-linear)\n",
      "        function of the parameters `f(xdata, params)` ::\n",
      "        \n",
      "               func(params) = ydata - f(xdata, params)\n",
      "        \n",
      "        so that the objective function is ::\n",
      "        \n",
      "               min   sum((ydata - f(xdata, params))**2, axis=0)\n",
      "             params\n",
      "        \n",
      "        The solution, `x`, is always a 1-D array, regardless of the shape of `x0`,\n",
      "        or whether `x0` is a scalar.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy.optimize import leastsq\n",
      "        >>> def func(x):\n",
      "        ...     return 2*(x-3)**2+1\n",
      "        >>> leastsq(func, 0)\n",
      "        (array([2.99999999]), 1)\n",
      "    \n",
      "    line_search = line_search_wolfe2(f, myfprime, xk, pk, gfk=None, old_fval=None, old_old_fval=None, args=(), c1=0.0001, c2=0.9, amax=None, extra_condition=None, maxiter=10)\n",
      "        Find alpha that satisfies strong Wolfe conditions.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        f : callable f(x,*args)\n",
      "            Objective function.\n",
      "        myfprime : callable f'(x,*args)\n",
      "            Objective function gradient.\n",
      "        xk : ndarray\n",
      "            Starting point.\n",
      "        pk : ndarray\n",
      "            Search direction.\n",
      "        gfk : ndarray, optional\n",
      "            Gradient value for x=xk (xk being the current parameter\n",
      "            estimate). Will be recomputed if omitted.\n",
      "        old_fval : float, optional\n",
      "            Function value for x=xk. Will be recomputed if omitted.\n",
      "        old_old_fval : float, optional\n",
      "            Function value for the point preceding x=xk.\n",
      "        args : tuple, optional\n",
      "            Additional arguments passed to objective function.\n",
      "        c1 : float, optional\n",
      "            Parameter for Armijo condition rule.\n",
      "        c2 : float, optional\n",
      "            Parameter for curvature condition rule.\n",
      "        amax : float, optional\n",
      "            Maximum step size\n",
      "        extra_condition : callable, optional\n",
      "            A callable of the form ``extra_condition(alpha, x, f, g)``\n",
      "            returning a boolean. Arguments are the proposed step ``alpha``\n",
      "            and the corresponding ``x``, ``f`` and ``g`` values. The line search\n",
      "            accepts the value of ``alpha`` only if this\n",
      "            callable returns ``True``. If the callable returns ``False``\n",
      "            for the step length, the algorithm will continue with\n",
      "            new iterates. The callable is only called for iterates\n",
      "            satisfying the strong Wolfe conditions.\n",
      "        maxiter : int, optional\n",
      "            Maximum number of iterations to perform.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        alpha : float or None\n",
      "            Alpha for which ``x_new = x0 + alpha * pk``,\n",
      "            or None if the line search algorithm did not converge.\n",
      "        fc : int\n",
      "            Number of function evaluations made.\n",
      "        gc : int\n",
      "            Number of gradient evaluations made.\n",
      "        new_fval : float or None\n",
      "            New function value ``f(x_new)=f(x0+alpha*pk)``,\n",
      "            or None if the line search algorithm did not converge.\n",
      "        old_fval : float\n",
      "            Old function value ``f(x0)``.\n",
      "        new_slope : float or None\n",
      "            The local slope along the search direction at the\n",
      "            new value ``<myfprime(x_new), pk>``,\n",
      "            or None if the line search algorithm did not converge.\n",
      "        \n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        Uses the line search algorithm to enforce strong Wolfe\n",
      "        conditions. See Wright and Nocedal, 'Numerical Optimization',\n",
      "        1999, pp. 59-61.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy.optimize import line_search\n",
      "        \n",
      "        A objective function and its gradient are defined.\n",
      "        \n",
      "        >>> def obj_func(x):\n",
      "        ...     return (x[0])**2+(x[1])**2\n",
      "        >>> def obj_grad(x):\n",
      "        ...     return [2*x[0], 2*x[1]]\n",
      "        \n",
      "        We can find alpha that satisfies strong Wolfe conditions.\n",
      "        \n",
      "        >>> start_point = np.array([1.8, 1.7])\n",
      "        >>> search_gradient = np.array([-1.0, -1.0])\n",
      "        >>> line_search(obj_func, obj_grad, start_point, search_gradient)\n",
      "        (1.0, 2, 1, 1.1300000000000001, 6.13, [1.6, 1.4])\n",
      "    \n",
      "    linear_sum_assignment(cost_matrix, maximize=False)\n",
      "        Solve the linear sum assignment problem.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        cost_matrix : array\n",
      "            The cost matrix of the bipartite graph.\n",
      "        \n",
      "        maximize : bool (default: False)\n",
      "            Calculates a maximum weight matching if true.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        row_ind, col_ind : array\n",
      "            An array of row indices and one of corresponding column indices giving\n",
      "            the optimal assignment. The cost of the assignment can be computed\n",
      "            as ``cost_matrix[row_ind, col_ind].sum()``. The row indices will be\n",
      "            sorted; in the case of a square cost matrix they will be equal to\n",
      "            ``numpy.arange(cost_matrix.shape[0])``.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        scipy.sparse.csgraph.min_weight_full_bipartite_matching : for sparse inputs\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        \n",
      "        The linear sum assignment problem [1]_ is also known as minimum weight\n",
      "        matching in bipartite graphs. A problem instance is described by a matrix\n",
      "        C, where each C[i,j] is the cost of matching vertex i of the first partite\n",
      "        set (a \"worker\") and vertex j of the second set (a \"job\"). The goal is to\n",
      "        find a complete assignment of workers to jobs of minimal cost.\n",
      "        \n",
      "        Formally, let X be a boolean matrix where :math:`X[i,j] = 1` iff row i is\n",
      "        assigned to column j. Then the optimal assignment has cost\n",
      "        \n",
      "        .. math::\n",
      "            \\min \\sum_i \\sum_j C_{i,j} X_{i,j}\n",
      "        \n",
      "        where, in the case where the matrix X is square, each row is assigned to\n",
      "        exactly one column, and each column to exactly one row.\n",
      "        \n",
      "        This function can also solve a generalization of the classic assignment\n",
      "        problem where the cost matrix is rectangular. If it has more rows than\n",
      "        columns, then not every row needs to be assigned to a column, and vice\n",
      "        versa.\n",
      "        \n",
      "        This implementation is a modified Jonker-Volgenant algorithm with no\n",
      "        initialization, described in ref. [2]_.\n",
      "        \n",
      "        .. versionadded:: 0.17.0\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        \n",
      "        .. [1] https://en.wikipedia.org/wiki/Assignment_problem\n",
      "        \n",
      "        .. [2] DF Crouse. On implementing 2D rectangular assignment algorithms.\n",
      "               *IEEE Transactions on Aerospace and Electronic Systems*,\n",
      "               52(4):1679-1696, August 2016, :doi:`10.1109/TAES.2016.140952`\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> cost = np.array([[4, 1, 3], [2, 0, 5], [3, 2, 2]])\n",
      "        >>> from scipy.optimize import linear_sum_assignment\n",
      "        >>> row_ind, col_ind = linear_sum_assignment(cost)\n",
      "        >>> col_ind\n",
      "        array([1, 0, 2])\n",
      "        >>> cost[row_ind, col_ind].sum()\n",
      "        5\n",
      "    \n",
      "    linearmixing(F, xin, iter=None, alpha=None, verbose=False, maxiter=None, f_tol=None, f_rtol=None, x_tol=None, x_rtol=None, tol_norm=None, line_search='armijo', callback=None, **kw)\n",
      "        Find a root of a function, using a scalar Jacobian approximation.\n",
      "        \n",
      "        .. warning::\n",
      "        \n",
      "           This algorithm may be useful for specific problems, but whether\n",
      "           it will work may depend strongly on the problem.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        F : function(x) -> f\n",
      "            Function whose root to find; should take and return an array-like\n",
      "            object.\n",
      "        xin : array_like\n",
      "            Initial guess for the solution\n",
      "        alpha : float, optional\n",
      "            The Jacobian approximation is (-1/alpha).\n",
      "        iter : int, optional\n",
      "            Number of iterations to make. If omitted (default), make as many\n",
      "            as required to meet tolerances.\n",
      "        verbose : bool, optional\n",
      "            Print status to stdout on every iteration.\n",
      "        maxiter : int, optional\n",
      "            Maximum number of iterations to make. If more are needed to\n",
      "            meet convergence, `NoConvergence` is raised.\n",
      "        f_tol : float, optional\n",
      "            Absolute tolerance (in max-norm) for the residual.\n",
      "            If omitted, default is 6e-6.\n",
      "        f_rtol : float, optional\n",
      "            Relative tolerance for the residual. If omitted, not used.\n",
      "        x_tol : float, optional\n",
      "            Absolute minimum step size, as determined from the Jacobian\n",
      "            approximation. If the step size is smaller than this, optimization\n",
      "            is terminated as successful. If omitted, not used.\n",
      "        x_rtol : float, optional\n",
      "            Relative minimum step size. If omitted, not used.\n",
      "        tol_norm : function(vector) -> scalar, optional\n",
      "            Norm to use in convergence check. Default is the maximum norm.\n",
      "        line_search : {None, 'armijo' (default), 'wolfe'}, optional\n",
      "            Which type of a line search to use to determine the step size in the\n",
      "            direction given by the Jacobian approximation. Defaults to 'armijo'.\n",
      "        callback : function, optional\n",
      "            Optional callback function. It is called on every iteration as\n",
      "            ``callback(x, f)`` where `x` is the current solution and `f`\n",
      "            the corresponding residual.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        sol : ndarray\n",
      "            An array (of similar array type as `x0`) containing the final solution.\n",
      "        \n",
      "        Raises\n",
      "        ------\n",
      "        NoConvergence\n",
      "            When a solution was not found.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        root : Interface to root finding algorithms for multivariate\n",
      "               functions. See ``method=='linearmixing'`` in particular.\n",
      "    \n",
      "    linprog(c, A_ub=None, b_ub=None, A_eq=None, b_eq=None, bounds=None, method='interior-point', callback=None, options=None, x0=None)\n",
      "        Linear programming: minimize a linear objective function subject to linear\n",
      "        equality and inequality constraints.\n",
      "        \n",
      "        Linear programming solves problems of the following form:\n",
      "        \n",
      "        .. math::\n",
      "        \n",
      "            \\min_x \\ & c^T x \\\\\n",
      "            \\mbox{such that} \\ & A_{ub} x \\leq b_{ub},\\\\\n",
      "            & A_{eq} x = b_{eq},\\\\\n",
      "            & l \\leq x \\leq u ,\n",
      "        \n",
      "        where :math:`x` is a vector of decision variables; :math:`c`,\n",
      "        :math:`b_{ub}`, :math:`b_{eq}`, :math:`l`, and :math:`u` are vectors; and\n",
      "        :math:`A_{ub}` and :math:`A_{eq}` are matrices.\n",
      "        \n",
      "        Alternatively, that's:\n",
      "        \n",
      "        minimize::\n",
      "        \n",
      "            c @ x\n",
      "        \n",
      "        such that::\n",
      "        \n",
      "            A_ub @ x <= b_ub\n",
      "            A_eq @ x == b_eq\n",
      "            lb <= x <= ub\n",
      "        \n",
      "        Note that by default ``lb = 0`` and ``ub = None`` unless specified with\n",
      "        ``bounds``.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        c : 1-D array\n",
      "            The coefficients of the linear objective function to be minimized.\n",
      "        A_ub : 2-D array, optional\n",
      "            The inequality constraint matrix. Each row of ``A_ub`` specifies the\n",
      "            coefficients of a linear inequality constraint on ``x``.\n",
      "        b_ub : 1-D array, optional\n",
      "            The inequality constraint vector. Each element represents an\n",
      "            upper bound on the corresponding value of ``A_ub @ x``.\n",
      "        A_eq : 2-D array, optional\n",
      "            The equality constraint matrix. Each row of ``A_eq`` specifies the\n",
      "            coefficients of a linear equality constraint on ``x``.\n",
      "        b_eq : 1-D array, optional\n",
      "            The equality constraint vector. Each element of ``A_eq @ x`` must equal\n",
      "            the corresponding element of ``b_eq``.\n",
      "        bounds : sequence, optional\n",
      "            A sequence of ``(min, max)`` pairs for each element in ``x``, defining\n",
      "            the minimum and maximum values of that decision variable. Use ``None``\n",
      "            to indicate that there is no bound. By default, bounds are\n",
      "            ``(0, None)`` (all decision variables are non-negative).\n",
      "            If a single tuple ``(min, max)`` is provided, then ``min`` and\n",
      "            ``max`` will serve as bounds for all decision variables.\n",
      "        method : str, optional\n",
      "            The algorithm used to solve the standard form problem.\n",
      "            :ref:`'highs-ds' <optimize.linprog-highs-ds>`,\n",
      "            :ref:`'highs-ipm' <optimize.linprog-highs-ipm>`,\n",
      "            :ref:`'highs' <optimize.linprog-highs>`,\n",
      "            :ref:`'interior-point' <optimize.linprog-interior-point>` (default),\n",
      "            :ref:`'revised simplex' <optimize.linprog-revised_simplex>`, and\n",
      "            :ref:`'simplex' <optimize.linprog-simplex>` (legacy)\n",
      "            are supported.\n",
      "        callback : callable, optional\n",
      "            If a callback function is provided, it will be called at least once per\n",
      "            iteration of the algorithm. The callback function must accept a single\n",
      "            `scipy.optimize.OptimizeResult` consisting of the following fields:\n",
      "        \n",
      "            x : 1-D array\n",
      "                The current solution vector.\n",
      "            fun : float\n",
      "                The current value of the objective function ``c @ x``.\n",
      "            success : bool\n",
      "                ``True`` when the algorithm has completed successfully.\n",
      "            slack : 1-D array\n",
      "                The (nominally positive) values of the slack,\n",
      "                ``b_ub - A_ub @ x``.\n",
      "            con : 1-D array\n",
      "                The (nominally zero) residuals of the equality constraints,\n",
      "                ``b_eq - A_eq @ x``.\n",
      "            phase : int\n",
      "                The phase of the algorithm being executed.\n",
      "            status : int\n",
      "                An integer representing the status of the algorithm.\n",
      "        \n",
      "                ``0`` : Optimization proceeding nominally.\n",
      "        \n",
      "                ``1`` : Iteration limit reached.\n",
      "        \n",
      "                ``2`` : Problem appears to be infeasible.\n",
      "        \n",
      "                ``3`` : Problem appears to be unbounded.\n",
      "        \n",
      "                ``4`` : Numerical difficulties encountered.\n",
      "        \n",
      "                nit : int\n",
      "                    The current iteration number.\n",
      "                message : str\n",
      "                    A string descriptor of the algorithm status.\n",
      "        \n",
      "            Callback functions are not currently supported by the HiGHS methods.\n",
      "        \n",
      "        options : dict, optional\n",
      "            A dictionary of solver options. All methods accept the following\n",
      "            options:\n",
      "        \n",
      "            maxiter : int\n",
      "                Maximum number of iterations to perform.\n",
      "                Default: see method-specific documentation.\n",
      "            disp : bool\n",
      "                Set to ``True`` to print convergence messages.\n",
      "                Default: ``False``.\n",
      "            presolve : bool\n",
      "                Set to ``False`` to disable automatic presolve.\n",
      "                Default: ``True``.\n",
      "        \n",
      "            All methods except the HiGHS solvers also accept:\n",
      "        \n",
      "            tol : float\n",
      "                A tolerance which determines when a residual is \"close enough\" to\n",
      "                zero to be considered exactly zero.\n",
      "            autoscale : bool\n",
      "                Set to ``True`` to automatically perform equilibration.\n",
      "                Consider using this option if the numerical values in the\n",
      "                constraints are separated by several orders of magnitude.\n",
      "                Default: ``False``.\n",
      "            rr : bool\n",
      "                Set to ``False`` to disable automatic redundancy removal.\n",
      "                Default: ``True``.\n",
      "            rr_method : string\n",
      "                Method used to identify and remove redundant rows from the\n",
      "                equality constraint matrix after presolve. For problems with\n",
      "                dense input, the available methods for redundancy removal are:\n",
      "        \n",
      "                \"SVD\":\n",
      "                    Repeatedly performs singular value decomposition on\n",
      "                    the matrix, detecting redundant rows based on nonzeros\n",
      "                    in the left singular vectors that correspond with\n",
      "                    zero singular values. May be fast when the matrix is\n",
      "                    nearly full rank.\n",
      "                \"pivot\":\n",
      "                    Uses the algorithm presented in [5]_ to identify\n",
      "                    redundant rows.\n",
      "                \"ID\":\n",
      "                    Uses a randomized interpolative decomposition.\n",
      "                    Identifies columns of the matrix transpose not used in\n",
      "                    a full-rank interpolative decomposition of the matrix.\n",
      "                None:\n",
      "                    Uses \"svd\" if the matrix is nearly full rank, that is,\n",
      "                    the difference between the matrix rank and the number\n",
      "                    of rows is less than five. If not, uses \"pivot\". The\n",
      "                    behavior of this default is subject to change without\n",
      "                    prior notice.\n",
      "        \n",
      "                Default: None.\n",
      "                For problems with sparse input, this option is ignored, and the\n",
      "                pivot-based algorithm presented in [5]_ is used.\n",
      "        \n",
      "            For method-specific options, see\n",
      "            :func:`show_options('linprog') <show_options>`.\n",
      "        \n",
      "        x0 : 1-D array, optional\n",
      "            Guess values of the decision variables, which will be refined by\n",
      "            the optimization algorithm. This argument is currently used only by the\n",
      "            'revised simplex' method, and can only be used if `x0` represents a\n",
      "            basic feasible solution.\n",
      "        \n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        res : OptimizeResult\n",
      "            A :class:`scipy.optimize.OptimizeResult` consisting of the fields:\n",
      "        \n",
      "            x : 1-D array\n",
      "                The values of the decision variables that minimizes the\n",
      "                objective function while satisfying the constraints.\n",
      "            fun : float\n",
      "                The optimal value of the objective function ``c @ x``.\n",
      "            slack : 1-D array\n",
      "                The (nominally positive) values of the slack variables,\n",
      "                ``b_ub - A_ub @ x``.\n",
      "            con : 1-D array\n",
      "                The (nominally zero) residuals of the equality constraints,\n",
      "                ``b_eq - A_eq @ x``.\n",
      "            success : bool\n",
      "                ``True`` when the algorithm succeeds in finding an optimal\n",
      "                solution.\n",
      "            status : int\n",
      "                An integer representing the exit status of the algorithm.\n",
      "        \n",
      "                ``0`` : Optimization terminated successfully.\n",
      "        \n",
      "                ``1`` : Iteration limit reached.\n",
      "        \n",
      "                ``2`` : Problem appears to be infeasible.\n",
      "        \n",
      "                ``3`` : Problem appears to be unbounded.\n",
      "        \n",
      "                ``4`` : Numerical difficulties encountered.\n",
      "        \n",
      "            nit : int\n",
      "                The total number of iterations performed in all phases.\n",
      "            message : str\n",
      "                A string descriptor of the exit status of the algorithm.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        show_options : Additional options accepted by the solvers.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        This section describes the available solvers that can be selected by the\n",
      "        'method' parameter.\n",
      "        \n",
      "        `'highs-ds'` and\n",
      "        `'highs-ipm'` are interfaces to the\n",
      "        HiGHS simplex and interior-point method solvers [13]_, respectively.\n",
      "        `'highs'` chooses between\n",
      "        the two automatically. These are the fastest linear\n",
      "        programming solvers in SciPy, especially for large, sparse problems;\n",
      "        which of these two is faster is problem-dependent.\n",
      "        `'interior-point'` is the default\n",
      "        as it was the fastest and most robust method before the recent\n",
      "        addition of the HiGHS solvers.\n",
      "        `'revised simplex'` is more\n",
      "        accurate than interior-point for the problems it solves.\n",
      "        `'simplex'` is the legacy method and is\n",
      "        included for backwards compatibility and educational purposes.\n",
      "        \n",
      "        Method *highs-ds* is a wrapper of the C++ high performance dual\n",
      "        revised simplex implementation (HSOL) [13]_, [14]_. Method *highs-ipm*\n",
      "        is a wrapper of a C++ implementation of an **i**\\ nterior-\\ **p**\\ oint\n",
      "        **m**\\ ethod [13]_; it features a crossover routine, so it is as accurate\n",
      "        as a simplex solver. Method *highs* chooses between the two automatically.\n",
      "        For new code involving `linprog`, we recommend explicitly choosing one of\n",
      "        these three method values.\n",
      "        \n",
      "        .. versionadded:: 1.6.0\n",
      "        \n",
      "        Method *interior-point* uses the primal-dual path following algorithm\n",
      "        as outlined in [4]_. This algorithm supports sparse constraint matrices and\n",
      "        is typically faster than the simplex methods, especially for large, sparse\n",
      "        problems. Note, however, that the solution returned may be slightly less\n",
      "        accurate than those of the simplex methods and will not, in general,\n",
      "        correspond with a vertex of the polytope defined by the constraints.\n",
      "        \n",
      "        .. versionadded:: 1.0.0\n",
      "        \n",
      "        Method *revised simplex* uses the revised simplex method as described in\n",
      "        [9]_, except that a factorization [11]_ of the basis matrix, rather than\n",
      "        its inverse, is efficiently maintained and used to solve the linear systems\n",
      "        at each iteration of the algorithm.\n",
      "        \n",
      "        .. versionadded:: 1.3.0\n",
      "        \n",
      "        Method *simplex* uses a traditional, full-tableau implementation of\n",
      "        Dantzig's simplex algorithm [1]_, [2]_ (*not* the\n",
      "        Nelder-Mead simplex). This algorithm is included for backwards\n",
      "        compatibility and educational purposes.\n",
      "        \n",
      "        .. versionadded:: 0.15.0\n",
      "        \n",
      "        Before applying *interior-point*, *revised simplex*, or *simplex*,\n",
      "        a presolve procedure based on [8]_ attempts\n",
      "        to identify trivial infeasibilities, trivial unboundedness, and potential\n",
      "        problem simplifications. Specifically, it checks for:\n",
      "        \n",
      "        - rows of zeros in ``A_eq`` or ``A_ub``, representing trivial constraints;\n",
      "        - columns of zeros in ``A_eq`` `and` ``A_ub``, representing unconstrained\n",
      "          variables;\n",
      "        - column singletons in ``A_eq``, representing fixed variables; and\n",
      "        - column singletons in ``A_ub``, representing simple bounds.\n",
      "        \n",
      "        If presolve reveals that the problem is unbounded (e.g. an unconstrained\n",
      "        and unbounded variable has negative cost) or infeasible (e.g., a row of\n",
      "        zeros in ``A_eq`` corresponds with a nonzero in ``b_eq``), the solver\n",
      "        terminates with the appropriate status code. Note that presolve terminates\n",
      "        as soon as any sign of unboundedness is detected; consequently, a problem\n",
      "        may be reported as unbounded when in reality the problem is infeasible\n",
      "        (but infeasibility has not been detected yet). Therefore, if it is\n",
      "        important to know whether the problem is actually infeasible, solve the\n",
      "        problem again with option ``presolve=False``.\n",
      "        \n",
      "        If neither infeasibility nor unboundedness are detected in a single pass\n",
      "        of the presolve, bounds are tightened where possible and fixed\n",
      "        variables are removed from the problem. Then, linearly dependent rows\n",
      "        of the ``A_eq`` matrix are removed, (unless they represent an\n",
      "        infeasibility) to avoid numerical difficulties in the primary solve\n",
      "        routine. Note that rows that are nearly linearly dependent (within a\n",
      "        prescribed tolerance) may also be removed, which can change the optimal\n",
      "        solution in rare cases. If this is a concern, eliminate redundancy from\n",
      "        your problem formulation and run with option ``rr=False`` or\n",
      "        ``presolve=False``.\n",
      "        \n",
      "        Several potential improvements can be made here: additional presolve\n",
      "        checks outlined in [8]_ should be implemented, the presolve routine should\n",
      "        be run multiple times (until no further simplifications can be made), and\n",
      "        more of the efficiency improvements from [5]_ should be implemented in the\n",
      "        redundancy removal routines.\n",
      "        \n",
      "        After presolve, the problem is transformed to standard form by converting\n",
      "        the (tightened) simple bounds to upper bound constraints, introducing\n",
      "        non-negative slack variables for inequality constraints, and expressing\n",
      "        unbounded variables as the difference between two non-negative variables.\n",
      "        Optionally, the problem is automatically scaled via equilibration [12]_.\n",
      "        The selected algorithm solves the standard form problem, and a\n",
      "        postprocessing routine converts the result to a solution to the original\n",
      "        problem.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] Dantzig, George B., Linear programming and extensions. Rand\n",
      "               Corporation Research Study Princeton Univ. Press, Princeton, NJ,\n",
      "               1963\n",
      "        .. [2] Hillier, S.H. and Lieberman, G.J. (1995), \"Introduction to\n",
      "               Mathematical Programming\", McGraw-Hill, Chapter 4.\n",
      "        .. [3] Bland, Robert G. New finite pivoting rules for the simplex method.\n",
      "               Mathematics of Operations Research (2), 1977: pp. 103-107.\n",
      "        .. [4] Andersen, Erling D., and Knud D. Andersen. \"The MOSEK interior point\n",
      "               optimizer for linear programming: an implementation of the\n",
      "               homogeneous algorithm.\" High performance optimization. Springer US,\n",
      "               2000. 197-232.\n",
      "        .. [5] Andersen, Erling D. \"Finding all linearly dependent rows in\n",
      "               large-scale linear programming.\" Optimization Methods and Software\n",
      "               6.3 (1995): 219-227.\n",
      "        .. [6] Freund, Robert M. \"Primal-Dual Interior-Point Methods for Linear\n",
      "               Programming based on Newton's Method.\" Unpublished Course Notes,\n",
      "               March 2004. Available 2/25/2017 at\n",
      "               https://ocw.mit.edu/courses/sloan-school-of-management/15-084j-nonlinear-programming-spring-2004/lecture-notes/lec14_int_pt_mthd.pdf\n",
      "        .. [7] Fourer, Robert. \"Solving Linear Programs by Interior-Point Methods.\"\n",
      "               Unpublished Course Notes, August 26, 2005. Available 2/25/2017 at\n",
      "               http://www.4er.org/CourseNotes/Book%20B/B-III.pdf\n",
      "        .. [8] Andersen, Erling D., and Knud D. Andersen. \"Presolving in linear\n",
      "               programming.\" Mathematical Programming 71.2 (1995): 221-245.\n",
      "        .. [9] Bertsimas, Dimitris, and J. Tsitsiklis. \"Introduction to linear\n",
      "               programming.\" Athena Scientific 1 (1997): 997.\n",
      "        .. [10] Andersen, Erling D., et al. Implementation of interior point\n",
      "                methods for large scale linear programming. HEC/Universite de\n",
      "                Geneve, 1996.\n",
      "        .. [11] Bartels, Richard H. \"A stabilization of the simplex method.\"\n",
      "                Journal in  Numerische Mathematik 16.5 (1971): 414-434.\n",
      "        .. [12] Tomlin, J. A. \"On scaling linear programming problems.\"\n",
      "                Mathematical Programming Study 4 (1975): 146-166.\n",
      "        .. [13] Huangfu, Q., Galabova, I., Feldmeier, M., and Hall, J. A. J.\n",
      "                \"HiGHS - high performance software for linear optimization.\"\n",
      "                Accessed 4/16/2020 at https://www.maths.ed.ac.uk/hall/HiGHS/#guide\n",
      "        .. [14] Huangfu, Q. and Hall, J. A. J. \"Parallelizing the dual revised\n",
      "                simplex method.\" Mathematical Programming Computation, 10 (1),\n",
      "                119-142, 2018. DOI: 10.1007/s12532-017-0130-5\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        Consider the following problem:\n",
      "        \n",
      "        .. math::\n",
      "        \n",
      "            \\min_{x_0, x_1} \\ -x_0 + 4x_1 & \\\\\n",
      "            \\mbox{such that} \\ -3x_0 + x_1 & \\leq 6,\\\\\n",
      "            -x_0 - 2x_1 & \\geq -4,\\\\\n",
      "            x_1 & \\geq -3.\n",
      "        \n",
      "        The problem is not presented in the form accepted by `linprog`. This is\n",
      "        easily remedied by converting the \"greater than\" inequality\n",
      "        constraint to a \"less than\" inequality constraint by\n",
      "        multiplying both sides by a factor of :math:`-1`. Note also that the last\n",
      "        constraint is really the simple bound :math:`-3 \\leq x_1 \\leq \\infty`.\n",
      "        Finally, since there are no bounds on :math:`x_0`, we must explicitly\n",
      "        specify the bounds :math:`-\\infty \\leq x_0 \\leq \\infty`, as the\n",
      "        default is for variables to be non-negative. After collecting coeffecients\n",
      "        into arrays and tuples, the input for this problem is:\n",
      "        \n",
      "        >>> c = [-1, 4]\n",
      "        >>> A = [[-3, 1], [1, 2]]\n",
      "        >>> b = [6, 4]\n",
      "        >>> x0_bounds = (None, None)\n",
      "        >>> x1_bounds = (-3, None)\n",
      "        >>> from scipy.optimize import linprog\n",
      "        >>> res = linprog(c, A_ub=A, b_ub=b, bounds=[x0_bounds, x1_bounds])\n",
      "        \n",
      "        Note that the default method for `linprog` is 'interior-point', which is\n",
      "        approximate by nature.\n",
      "        \n",
      "        >>> print(res)\n",
      "             con: array([], dtype=float64)\n",
      "             fun: -21.99999984082494 # may vary\n",
      "         message: 'Optimization terminated successfully.'\n",
      "             nit: 6 # may vary\n",
      "           slack: array([3.89999997e+01, 8.46872439e-08] # may vary\n",
      "          status: 0\n",
      "         success: True\n",
      "               x: array([ 9.99999989, -2.99999999]) # may vary\n",
      "        \n",
      "        If you need greater accuracy, try 'revised simplex'.\n",
      "        \n",
      "        >>> res = linprog(c, A_ub=A, b_ub=b, bounds=[x0_bounds, x1_bounds], method='revised simplex')\n",
      "        >>> print(res)\n",
      "             con: array([], dtype=float64)\n",
      "             fun: -22.0 # may vary\n",
      "         message: 'Optimization terminated successfully.'\n",
      "             nit: 1 # may vary\n",
      "           slack: array([39.,  0.]) # may vary\n",
      "          status: 0\n",
      "         success: True\n",
      "               x: array([10., -3.]) # may vary\n",
      "    \n",
      "    linprog_verbose_callback(res)\n",
      "        A sample callback function demonstrating the linprog callback interface.\n",
      "        This callback produces detailed output to sys.stdout before each iteration\n",
      "        and after the final iteration of the simplex algorithm.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        res : A `scipy.optimize.OptimizeResult` consisting of the following fields:\n",
      "        \n",
      "            x : 1-D array\n",
      "                The independent variable vector which optimizes the linear\n",
      "                programming problem.\n",
      "            fun : float\n",
      "                Value of the objective function.\n",
      "            success : bool\n",
      "                True if the algorithm succeeded in finding an optimal solution.\n",
      "            slack : 1-D array\n",
      "                The values of the slack variables. Each slack variable corresponds\n",
      "                to an inequality constraint. If the slack is zero, then the\n",
      "                corresponding constraint is active.\n",
      "            con : 1-D array\n",
      "                The (nominally zero) residuals of the equality constraints, that is,\n",
      "                ``b - A_eq @ x``\n",
      "            phase : int\n",
      "                The phase of the optimization being executed. In phase 1 a basic\n",
      "                feasible solution is sought and the T has an additional row\n",
      "                representing an alternate objective function.\n",
      "            status : int\n",
      "                An integer representing the exit status of the optimization::\n",
      "        \n",
      "                     0 : Optimization terminated successfully\n",
      "                     1 : Iteration limit reached\n",
      "                     2 : Problem appears to be infeasible\n",
      "                     3 : Problem appears to be unbounded\n",
      "                     4 : Serious numerical difficulties encountered\n",
      "        \n",
      "            nit : int\n",
      "                The number of iterations performed.\n",
      "            message : str\n",
      "                A string descriptor of the exit status of the optimization.\n",
      "    \n",
      "    lsq_linear(A, b, bounds=(-inf, inf), method='trf', tol=1e-10, lsq_solver=None, lsmr_tol=None, max_iter=None, verbose=0)\n",
      "        Solve a linear least-squares problem with bounds on the variables.\n",
      "        \n",
      "        Given a m-by-n design matrix A and a target vector b with m elements,\n",
      "        `lsq_linear` solves the following optimization problem::\n",
      "        \n",
      "            minimize 0.5 * ||A x - b||**2\n",
      "            subject to lb <= x <= ub\n",
      "        \n",
      "        This optimization problem is convex, hence a found minimum (if iterations\n",
      "        have converged) is guaranteed to be global.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        A : array_like, sparse matrix of LinearOperator, shape (m, n)\n",
      "            Design matrix. Can be `scipy.sparse.linalg.LinearOperator`.\n",
      "        b : array_like, shape (m,)\n",
      "            Target vector.\n",
      "        bounds : 2-tuple of array_like, optional\n",
      "            Lower and upper bounds on independent variables. Defaults to no bounds.\n",
      "            Each array must have shape (n,) or be a scalar, in the latter\n",
      "            case a bound will be the same for all variables. Use ``np.inf`` with\n",
      "            an appropriate sign to disable bounds on all or some variables.\n",
      "        method : 'trf' or 'bvls', optional\n",
      "            Method to perform minimization.\n",
      "        \n",
      "                * 'trf' : Trust Region Reflective algorithm adapted for a linear\n",
      "                  least-squares problem. This is an interior-point-like method\n",
      "                  and the required number of iterations is weakly correlated with\n",
      "                  the number of variables.\n",
      "                * 'bvls' : Bounded-variable least-squares algorithm. This is\n",
      "                  an active set method, which requires the number of iterations\n",
      "                  comparable to the number of variables. Can't be used when `A` is\n",
      "                  sparse or LinearOperator.\n",
      "        \n",
      "            Default is 'trf'.\n",
      "        tol : float, optional\n",
      "            Tolerance parameter. The algorithm terminates if a relative change\n",
      "            of the cost function is less than `tol` on the last iteration.\n",
      "            Additionally, the first-order optimality measure is considered:\n",
      "        \n",
      "                * ``method='trf'`` terminates if the uniform norm of the gradient,\n",
      "                  scaled to account for the presence of the bounds, is less than\n",
      "                  `tol`.\n",
      "                * ``method='bvls'`` terminates if Karush-Kuhn-Tucker conditions\n",
      "                  are satisfied within `tol` tolerance.\n",
      "        \n",
      "        lsq_solver : {None, 'exact', 'lsmr'}, optional\n",
      "            Method of solving unbounded least-squares problems throughout\n",
      "            iterations:\n",
      "        \n",
      "                * 'exact' : Use dense QR or SVD decomposition approach. Can't be\n",
      "                  used when `A` is sparse or LinearOperator.\n",
      "                * 'lsmr' : Use `scipy.sparse.linalg.lsmr` iterative procedure\n",
      "                  which requires only matrix-vector product evaluations. Can't\n",
      "                  be used with ``method='bvls'``.\n",
      "        \n",
      "            If None (default), the solver is chosen based on type of `A`.\n",
      "        lsmr_tol : None, float or 'auto', optional\n",
      "            Tolerance parameters 'atol' and 'btol' for `scipy.sparse.linalg.lsmr`\n",
      "            If None (default), it is set to ``1e-2 * tol``. If 'auto', the\n",
      "            tolerance will be adjusted based on the optimality of the current\n",
      "            iterate, which can speed up the optimization process, but is not always\n",
      "            reliable.\n",
      "        max_iter : None or int, optional\n",
      "            Maximum number of iterations before termination. If None (default), it\n",
      "            is set to 100 for ``method='trf'`` or to the number of variables for\n",
      "            ``method='bvls'`` (not counting iterations for 'bvls' initialization).\n",
      "        verbose : {0, 1, 2}, optional\n",
      "            Level of algorithm's verbosity:\n",
      "        \n",
      "                * 0 : work silently (default).\n",
      "                * 1 : display a termination report.\n",
      "                * 2 : display progress during iterations.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        OptimizeResult with the following fields defined:\n",
      "        x : ndarray, shape (n,)\n",
      "            Solution found.\n",
      "        cost : float\n",
      "            Value of the cost function at the solution.\n",
      "        fun : ndarray, shape (m,)\n",
      "            Vector of residuals at the solution.\n",
      "        optimality : float\n",
      "            First-order optimality measure. The exact meaning depends on `method`,\n",
      "            refer to the description of `tol` parameter.\n",
      "        active_mask : ndarray of int, shape (n,)\n",
      "            Each component shows whether a corresponding constraint is active\n",
      "            (that is, whether a variable is at the bound):\n",
      "        \n",
      "                *  0 : a constraint is not active.\n",
      "                * -1 : a lower bound is active.\n",
      "                *  1 : an upper bound is active.\n",
      "        \n",
      "            Might be somewhat arbitrary for the `trf` method as it generates a\n",
      "            sequence of strictly feasible iterates and active_mask is determined\n",
      "            within a tolerance threshold.\n",
      "        nit : int\n",
      "            Number of iterations. Zero if the unconstrained solution is optimal.\n",
      "        status : int\n",
      "            Reason for algorithm termination:\n",
      "        \n",
      "                * -1 : the algorithm was not able to make progress on the last\n",
      "                  iteration.\n",
      "                *  0 : the maximum number of iterations is exceeded.\n",
      "                *  1 : the first-order optimality measure is less than `tol`.\n",
      "                *  2 : the relative change of the cost function is less than `tol`.\n",
      "                *  3 : the unconstrained solution is optimal.\n",
      "        \n",
      "        message : str\n",
      "            Verbal description of the termination reason.\n",
      "        success : bool\n",
      "            True if one of the convergence criteria is satisfied (`status` > 0).\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        nnls : Linear least squares with non-negativity constraint.\n",
      "        least_squares : Nonlinear least squares with bounds on the variables.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        The algorithm first computes the unconstrained least-squares solution by\n",
      "        `numpy.linalg.lstsq` or `scipy.sparse.linalg.lsmr` depending on\n",
      "        `lsq_solver`. This solution is returned as optimal if it lies within the\n",
      "        bounds.\n",
      "        \n",
      "        Method 'trf' runs the adaptation of the algorithm described in [STIR]_ for\n",
      "        a linear least-squares problem. The iterations are essentially the same as\n",
      "        in the nonlinear least-squares algorithm, but as the quadratic function\n",
      "        model is always accurate, we don't need to track or modify the radius of\n",
      "        a trust region. The line search (backtracking) is used as a safety net\n",
      "        when a selected step does not decrease the cost function. Read more\n",
      "        detailed description of the algorithm in `scipy.optimize.least_squares`.\n",
      "        \n",
      "        Method 'bvls' runs a Python implementation of the algorithm described in\n",
      "        [BVLS]_. The algorithm maintains active and free sets of variables, on\n",
      "        each iteration chooses a new variable to move from the active set to the\n",
      "        free set and then solves the unconstrained least-squares problem on free\n",
      "        variables. This algorithm is guaranteed to give an accurate solution\n",
      "        eventually, but may require up to n iterations for a problem with n\n",
      "        variables. Additionally, an ad-hoc initialization procedure is\n",
      "        implemented, that determines which variables to set free or active\n",
      "        initially. It takes some number of iterations before actual BVLS starts,\n",
      "        but can significantly reduce the number of further iterations.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [STIR] M. A. Branch, T. F. Coleman, and Y. Li, \"A Subspace, Interior,\n",
      "                  and Conjugate Gradient Method for Large-Scale Bound-Constrained\n",
      "                  Minimization Problems,\" SIAM Journal on Scientific Computing,\n",
      "                  Vol. 21, Number 1, pp 1-23, 1999.\n",
      "        .. [BVLS] P. B. Start and R. L. Parker, \"Bounded-Variable Least-Squares:\n",
      "                  an Algorithm and Applications\", Computational Statistics, 10,\n",
      "                  129-141, 1995.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        In this example, a problem with a large sparse matrix and bounds on the\n",
      "        variables is solved.\n",
      "        \n",
      "        >>> from scipy.sparse import rand\n",
      "        >>> from scipy.optimize import lsq_linear\n",
      "        >>> rng = np.random.default_rng()\n",
      "        ...\n",
      "        >>> m = 20000\n",
      "        >>> n = 10000\n",
      "        ...\n",
      "        >>> A = rand(m, n, density=1e-4, random_state=rng)\n",
      "        >>> b = rng.standard_normal(m)\n",
      "        ...\n",
      "        >>> lb = rng.standard_normal(n)\n",
      "        >>> ub = lb + 1\n",
      "        ...\n",
      "        >>> res = lsq_linear(A, b, bounds=(lb, ub), lsmr_tol='auto', verbose=1)\n",
      "        # may vary\n",
      "        The relative change of the cost function is less than `tol`.\n",
      "        Number of iterations 16, initial cost 1.5039e+04, final cost 1.1112e+04,\n",
      "        first-order optimality 4.66e-08.\n",
      "    \n",
      "    minimize(fun, x0, args=(), method=None, jac=None, hess=None, hessp=None, bounds=None, constraints=(), tol=None, callback=None, options=None)\n",
      "        Minimization of scalar function of one or more variables.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        fun : callable\n",
      "            The objective function to be minimized.\n",
      "        \n",
      "                ``fun(x, *args) -> float``\n",
      "        \n",
      "            where ``x`` is an 1-D array with shape (n,) and ``args``\n",
      "            is a tuple of the fixed parameters needed to completely\n",
      "            specify the function.\n",
      "        x0 : ndarray, shape (n,)\n",
      "            Initial guess. Array of real elements of size (n,),\n",
      "            where 'n' is the number of independent variables.\n",
      "        args : tuple, optional\n",
      "            Extra arguments passed to the objective function and its\n",
      "            derivatives (`fun`, `jac` and `hess` functions).\n",
      "        method : str or callable, optional\n",
      "            Type of solver.  Should be one of\n",
      "        \n",
      "                - 'Nelder-Mead' :ref:`(see here) <optimize.minimize-neldermead>`\n",
      "                - 'Powell'      :ref:`(see here) <optimize.minimize-powell>`\n",
      "                - 'CG'          :ref:`(see here) <optimize.minimize-cg>`\n",
      "                - 'BFGS'        :ref:`(see here) <optimize.minimize-bfgs>`\n",
      "                - 'Newton-CG'   :ref:`(see here) <optimize.minimize-newtoncg>`\n",
      "                - 'L-BFGS-B'    :ref:`(see here) <optimize.minimize-lbfgsb>`\n",
      "                - 'TNC'         :ref:`(see here) <optimize.minimize-tnc>`\n",
      "                - 'COBYLA'      :ref:`(see here) <optimize.minimize-cobyla>`\n",
      "                - 'SLSQP'       :ref:`(see here) <optimize.minimize-slsqp>`\n",
      "                - 'trust-constr':ref:`(see here) <optimize.minimize-trustconstr>`\n",
      "                - 'dogleg'      :ref:`(see here) <optimize.minimize-dogleg>`\n",
      "                - 'trust-ncg'   :ref:`(see here) <optimize.minimize-trustncg>`\n",
      "                - 'trust-exact' :ref:`(see here) <optimize.minimize-trustexact>`\n",
      "                - 'trust-krylov' :ref:`(see here) <optimize.minimize-trustkrylov>`\n",
      "                - custom - a callable object (added in version 0.14.0),\n",
      "                  see below for description.\n",
      "        \n",
      "            If not given, chosen to be one of ``BFGS``, ``L-BFGS-B``, ``SLSQP``,\n",
      "            depending if the problem has constraints or bounds.\n",
      "        jac : {callable,  '2-point', '3-point', 'cs', bool}, optional\n",
      "            Method for computing the gradient vector. Only for CG, BFGS,\n",
      "            Newton-CG, L-BFGS-B, TNC, SLSQP, dogleg, trust-ncg, trust-krylov,\n",
      "            trust-exact and trust-constr.\n",
      "            If it is a callable, it should be a function that returns the gradient\n",
      "            vector:\n",
      "        \n",
      "                ``jac(x, *args) -> array_like, shape (n,)``\n",
      "        \n",
      "            where ``x`` is an array with shape (n,) and ``args`` is a tuple with\n",
      "            the fixed parameters. If `jac` is a Boolean and is True, `fun` is\n",
      "            assumed to return a tuple ``(f, g)`` containing the objective\n",
      "            function and the gradient.\n",
      "            Methods 'Newton-CG', 'trust-ncg', 'dogleg', 'trust-exact', and\n",
      "            'trust-krylov' require that either a callable be supplied, or that\n",
      "            `fun` return the objective and gradient.\n",
      "            If None or False, the gradient will be estimated using 2-point finite\n",
      "            difference estimation with an absolute step size.\n",
      "            Alternatively, the keywords  {'2-point', '3-point', 'cs'} can be used\n",
      "            to select a finite difference scheme for numerical estimation of the\n",
      "            gradient with a relative step size. These finite difference schemes\n",
      "            obey any specified `bounds`.\n",
      "        hess : {callable, '2-point', '3-point', 'cs', HessianUpdateStrategy}, optional\n",
      "            Method for computing the Hessian matrix. Only for Newton-CG, dogleg,\n",
      "            trust-ncg, trust-krylov, trust-exact and trust-constr. If it is\n",
      "            callable, it should return the Hessian matrix:\n",
      "        \n",
      "                ``hess(x, *args) -> {LinearOperator, spmatrix, array}, (n, n)``\n",
      "        \n",
      "            where x is a (n,) ndarray and `args` is a tuple with the fixed\n",
      "            parameters. LinearOperator and sparse matrix returns are only allowed\n",
      "            for 'trust-constr' method. Alternatively, the keywords\n",
      "            {'2-point', '3-point', 'cs'} select a finite difference scheme\n",
      "            for numerical estimation. Or, objects implementing the\n",
      "            `HessianUpdateStrategy` interface can be used to approximate\n",
      "            the Hessian. Available quasi-Newton methods implementing\n",
      "            this interface are:\n",
      "        \n",
      "                - `BFGS`;\n",
      "                - `SR1`.\n",
      "        \n",
      "            Whenever the gradient is estimated via finite-differences,\n",
      "            the Hessian cannot be estimated with options\n",
      "            {'2-point', '3-point', 'cs'} and needs to be\n",
      "            estimated using one of the quasi-Newton strategies.\n",
      "            'trust-exact' cannot use a finite-difference scheme, and must be used\n",
      "            with a callable returning an (n, n) array.\n",
      "        hessp : callable, optional\n",
      "            Hessian of objective function times an arbitrary vector p. Only for\n",
      "            Newton-CG, trust-ncg, trust-krylov, trust-constr.\n",
      "            Only one of `hessp` or `hess` needs to be given.  If `hess` is\n",
      "            provided, then `hessp` will be ignored.  `hessp` must compute the\n",
      "            Hessian times an arbitrary vector:\n",
      "        \n",
      "                ``hessp(x, p, *args) ->  ndarray shape (n,)``\n",
      "        \n",
      "            where x is a (n,) ndarray, p is an arbitrary vector with\n",
      "            dimension (n,) and `args` is a tuple with the fixed\n",
      "            parameters.\n",
      "        bounds : sequence or `Bounds`, optional\n",
      "            Bounds on variables for Nelder-Mead, L-BFGS-B, TNC, SLSQP, Powell, and\n",
      "            trust-constr methods. There are two ways to specify the bounds:\n",
      "        \n",
      "                1. Instance of `Bounds` class.\n",
      "                2. Sequence of ``(min, max)`` pairs for each element in `x`. None\n",
      "                   is used to specify no bound.\n",
      "        \n",
      "        constraints : {Constraint, dict} or List of {Constraint, dict}, optional\n",
      "            Constraints definition (only for COBYLA, SLSQP and trust-constr).\n",
      "        \n",
      "            Constraints for 'trust-constr' are defined as a single object or a\n",
      "            list of objects specifying constraints to the optimization problem.\n",
      "            Available constraints are:\n",
      "        \n",
      "                - `LinearConstraint`\n",
      "                - `NonlinearConstraint`\n",
      "        \n",
      "            Constraints for COBYLA, SLSQP are defined as a list of dictionaries.\n",
      "            Each dictionary with fields:\n",
      "        \n",
      "                type : str\n",
      "                    Constraint type: 'eq' for equality, 'ineq' for inequality.\n",
      "                fun : callable\n",
      "                    The function defining the constraint.\n",
      "                jac : callable, optional\n",
      "                    The Jacobian of `fun` (only for SLSQP).\n",
      "                args : sequence, optional\n",
      "                    Extra arguments to be passed to the function and Jacobian.\n",
      "        \n",
      "            Equality constraint means that the constraint function result is to\n",
      "            be zero whereas inequality means that it is to be non-negative.\n",
      "            Note that COBYLA only supports inequality constraints.\n",
      "        tol : float, optional\n",
      "            Tolerance for termination. When `tol` is specified, the selected\n",
      "            minimization algorithm sets some relevant solver-specific tolerance(s)\n",
      "            equal to `tol`. For detailed control, use solver-specific\n",
      "            options.\n",
      "        options : dict, optional\n",
      "            A dictionary of solver options. All methods accept the following\n",
      "            generic options:\n",
      "        \n",
      "                maxiter : int\n",
      "                    Maximum number of iterations to perform. Depending on the\n",
      "                    method each iteration may use several function evaluations.\n",
      "                disp : bool\n",
      "                    Set to True to print convergence messages.\n",
      "        \n",
      "            For method-specific options, see :func:`show_options()`.\n",
      "        callback : callable, optional\n",
      "            Called after each iteration. For 'trust-constr' it is a callable with\n",
      "            the signature:\n",
      "        \n",
      "                ``callback(xk, OptimizeResult state) -> bool``\n",
      "        \n",
      "            where ``xk`` is the current parameter vector. and ``state``\n",
      "            is an `OptimizeResult` object, with the same fields\n",
      "            as the ones from the return. If callback returns True\n",
      "            the algorithm execution is terminated.\n",
      "            For all the other methods, the signature is:\n",
      "        \n",
      "                ``callback(xk)``\n",
      "        \n",
      "            where ``xk`` is the current parameter vector.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        res : OptimizeResult\n",
      "            The optimization result represented as a ``OptimizeResult`` object.\n",
      "            Important attributes are: ``x`` the solution array, ``success`` a\n",
      "            Boolean flag indicating if the optimizer exited successfully and\n",
      "            ``message`` which describes the cause of the termination. See\n",
      "            `OptimizeResult` for a description of other attributes.\n",
      "        \n",
      "        See also\n",
      "        --------\n",
      "        minimize_scalar : Interface to minimization algorithms for scalar\n",
      "            univariate functions\n",
      "        show_options : Additional options accepted by the solvers\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        This section describes the available solvers that can be selected by the\n",
      "        'method' parameter. The default method is *BFGS*.\n",
      "        \n",
      "        **Unconstrained minimization**\n",
      "        \n",
      "        Method :ref:`CG <optimize.minimize-cg>` uses a nonlinear conjugate\n",
      "        gradient algorithm by Polak and Ribiere, a variant of the\n",
      "        Fletcher-Reeves method described in [5]_ pp.120-122. Only the\n",
      "        first derivatives are used.\n",
      "        \n",
      "        Method :ref:`BFGS <optimize.minimize-bfgs>` uses the quasi-Newton\n",
      "        method of Broyden, Fletcher, Goldfarb, and Shanno (BFGS) [5]_\n",
      "        pp. 136. It uses the first derivatives only. BFGS has proven good\n",
      "        performance even for non-smooth optimizations. This method also\n",
      "        returns an approximation of the Hessian inverse, stored as\n",
      "        `hess_inv` in the OptimizeResult object.\n",
      "        \n",
      "        Method :ref:`Newton-CG <optimize.minimize-newtoncg>` uses a\n",
      "        Newton-CG algorithm [5]_ pp. 168 (also known as the truncated\n",
      "        Newton method). It uses a CG method to the compute the search\n",
      "        direction. See also *TNC* method for a box-constrained\n",
      "        minimization with a similar algorithm. Suitable for large-scale\n",
      "        problems.\n",
      "        \n",
      "        Method :ref:`dogleg <optimize.minimize-dogleg>` uses the dog-leg\n",
      "        trust-region algorithm [5]_ for unconstrained minimization. This\n",
      "        algorithm requires the gradient and Hessian; furthermore the\n",
      "        Hessian is required to be positive definite.\n",
      "        \n",
      "        Method :ref:`trust-ncg <optimize.minimize-trustncg>` uses the\n",
      "        Newton conjugate gradient trust-region algorithm [5]_ for\n",
      "        unconstrained minimization. This algorithm requires the gradient\n",
      "        and either the Hessian or a function that computes the product of\n",
      "        the Hessian with a given vector. Suitable for large-scale problems.\n",
      "        \n",
      "        Method :ref:`trust-krylov <optimize.minimize-trustkrylov>` uses\n",
      "        the Newton GLTR trust-region algorithm [14]_, [15]_ for unconstrained\n",
      "        minimization. This algorithm requires the gradient\n",
      "        and either the Hessian or a function that computes the product of\n",
      "        the Hessian with a given vector. Suitable for large-scale problems.\n",
      "        On indefinite problems it requires usually less iterations than the\n",
      "        `trust-ncg` method and is recommended for medium and large-scale problems.\n",
      "        \n",
      "        Method :ref:`trust-exact <optimize.minimize-trustexact>`\n",
      "        is a trust-region method for unconstrained minimization in which\n",
      "        quadratic subproblems are solved almost exactly [13]_. This\n",
      "        algorithm requires the gradient and the Hessian (which is\n",
      "        *not* required to be positive definite). It is, in many\n",
      "        situations, the Newton method to converge in fewer iteraction\n",
      "        and the most recommended for small and medium-size problems.\n",
      "        \n",
      "        **Bound-Constrained minimization**\n",
      "        \n",
      "        Method :ref:`Nelder-Mead <optimize.minimize-neldermead>` uses the\n",
      "        Simplex algorithm [1]_, [2]_. This algorithm is robust in many\n",
      "        applications. However, if numerical computation of derivative can be\n",
      "        trusted, other algorithms using the first and/or second derivatives\n",
      "        information might be preferred for their better performance in\n",
      "        general.\n",
      "        \n",
      "        Method :ref:`L-BFGS-B <optimize.minimize-lbfgsb>` uses the L-BFGS-B\n",
      "        algorithm [6]_, [7]_ for bound constrained minimization.\n",
      "        \n",
      "        Method :ref:`Powell <optimize.minimize-powell>` is a modification\n",
      "        of Powell's method [3]_, [4]_ which is a conjugate direction\n",
      "        method. It performs sequential one-dimensional minimizations along\n",
      "        each vector of the directions set (`direc` field in `options` and\n",
      "        `info`), which is updated at each iteration of the main\n",
      "        minimization loop. The function need not be differentiable, and no\n",
      "        derivatives are taken. If bounds are not provided, then an\n",
      "        unbounded line search will be used. If bounds are provided and\n",
      "        the initial guess is within the bounds, then every function\n",
      "        evaluation throughout the minimization procedure will be within\n",
      "        the bounds. If bounds are provided, the initial guess is outside\n",
      "        the bounds, and `direc` is full rank (default has full rank), then\n",
      "        some function evaluations during the first iteration may be\n",
      "        outside the bounds, but every function evaluation after the first\n",
      "        iteration will be within the bounds. If `direc` is not full rank,\n",
      "        then some parameters may not be optimized and the solution is not\n",
      "        guaranteed to be within the bounds.\n",
      "        \n",
      "        Method :ref:`TNC <optimize.minimize-tnc>` uses a truncated Newton\n",
      "        algorithm [5]_, [8]_ to minimize a function with variables subject\n",
      "        to bounds. This algorithm uses gradient information; it is also\n",
      "        called Newton Conjugate-Gradient. It differs from the *Newton-CG*\n",
      "        method described above as it wraps a C implementation and allows\n",
      "        each variable to be given upper and lower bounds.\n",
      "        \n",
      "        **Constrained Minimization**\n",
      "        \n",
      "        Method :ref:`COBYLA <optimize.minimize-cobyla>` uses the\n",
      "        Constrained Optimization BY Linear Approximation (COBYLA) method\n",
      "        [9]_, [10]_, [11]_. The algorithm is based on linear\n",
      "        approximations to the objective function and each constraint. The\n",
      "        method wraps a FORTRAN implementation of the algorithm. The\n",
      "        constraints functions 'fun' may return either a single number\n",
      "        or an array or list of numbers.\n",
      "        \n",
      "        Method :ref:`SLSQP <optimize.minimize-slsqp>` uses Sequential\n",
      "        Least SQuares Programming to minimize a function of several\n",
      "        variables with any combination of bounds, equality and inequality\n",
      "        constraints. The method wraps the SLSQP Optimization subroutine\n",
      "        originally implemented by Dieter Kraft [12]_. Note that the\n",
      "        wrapper handles infinite values in bounds by converting them into\n",
      "        large floating values.\n",
      "        \n",
      "        Method :ref:`trust-constr <optimize.minimize-trustconstr>` is a\n",
      "        trust-region algorithm for constrained optimization. It swiches\n",
      "        between two implementations depending on the problem definition.\n",
      "        It is the most versatile constrained minimization algorithm\n",
      "        implemented in SciPy and the most appropriate for large-scale problems.\n",
      "        For equality constrained problems it is an implementation of Byrd-Omojokun\n",
      "        Trust-Region SQP method described in [17]_ and in [5]_, p. 549. When\n",
      "        inequality constraints  are imposed as well, it swiches to the trust-region\n",
      "        interior point  method described in [16]_. This interior point algorithm,\n",
      "        in turn, solves inequality constraints by introducing slack variables\n",
      "        and solving a sequence of equality-constrained barrier problems\n",
      "        for progressively smaller values of the barrier parameter.\n",
      "        The previously described equality constrained SQP method is\n",
      "        used to solve the subproblems with increasing levels of accuracy\n",
      "        as the iterate gets closer to a solution.\n",
      "        \n",
      "        **Finite-Difference Options**\n",
      "        \n",
      "        For Method :ref:`trust-constr <optimize.minimize-trustconstr>`\n",
      "        the gradient and the Hessian may be approximated using\n",
      "        three finite-difference schemes: {'2-point', '3-point', 'cs'}.\n",
      "        The scheme 'cs' is, potentially, the most accurate but it\n",
      "        requires the function to correctly handles complex inputs and to\n",
      "        be differentiable in the complex plane. The scheme '3-point' is more\n",
      "        accurate than '2-point' but requires twice as many operations.\n",
      "        \n",
      "        **Custom minimizers**\n",
      "        \n",
      "        It may be useful to pass a custom minimization method, for example\n",
      "        when using a frontend to this method such as `scipy.optimize.basinhopping`\n",
      "        or a different library.  You can simply pass a callable as the ``method``\n",
      "        parameter.\n",
      "        \n",
      "        The callable is called as ``method(fun, x0, args, **kwargs, **options)``\n",
      "        where ``kwargs`` corresponds to any other parameters passed to `minimize`\n",
      "        (such as `callback`, `hess`, etc.), except the `options` dict, which has\n",
      "        its contents also passed as `method` parameters pair by pair.  Also, if\n",
      "        `jac` has been passed as a bool type, `jac` and `fun` are mangled so that\n",
      "        `fun` returns just the function values and `jac` is converted to a function\n",
      "        returning the Jacobian.  The method shall return an `OptimizeResult`\n",
      "        object.\n",
      "        \n",
      "        The provided `method` callable must be able to accept (and possibly ignore)\n",
      "        arbitrary parameters; the set of parameters accepted by `minimize` may\n",
      "        expand in future versions and then these parameters will be passed to\n",
      "        the method.  You can find an example in the scipy.optimize tutorial.\n",
      "        \n",
      "        .. versionadded:: 0.11.0\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] Nelder, J A, and R Mead. 1965. A Simplex Method for Function\n",
      "            Minimization. The Computer Journal 7: 308-13.\n",
      "        .. [2] Wright M H. 1996. Direct search methods: Once scorned, now\n",
      "            respectable, in Numerical Analysis 1995: Proceedings of the 1995\n",
      "            Dundee Biennial Conference in Numerical Analysis (Eds. D F\n",
      "            Griffiths and G A Watson). Addison Wesley Longman, Harlow, UK.\n",
      "            191-208.\n",
      "        .. [3] Powell, M J D. 1964. An efficient method for finding the minimum of\n",
      "           a function of several variables without calculating derivatives. The\n",
      "           Computer Journal 7: 155-162.\n",
      "        .. [4] Press W, S A Teukolsky, W T Vetterling and B P Flannery.\n",
      "           Numerical Recipes (any edition), Cambridge University Press.\n",
      "        .. [5] Nocedal, J, and S J Wright. 2006. Numerical Optimization.\n",
      "           Springer New York.\n",
      "        .. [6] Byrd, R H and P Lu and J. Nocedal. 1995. A Limited Memory\n",
      "           Algorithm for Bound Constrained Optimization. SIAM Journal on\n",
      "           Scientific and Statistical Computing 16 (5): 1190-1208.\n",
      "        .. [7] Zhu, C and R H Byrd and J Nocedal. 1997. L-BFGS-B: Algorithm\n",
      "           778: L-BFGS-B, FORTRAN routines for large scale bound constrained\n",
      "           optimization. ACM Transactions on Mathematical Software 23 (4):\n",
      "           550-560.\n",
      "        .. [8] Nash, S G. Newton-Type Minimization Via the Lanczos Method.\n",
      "           1984. SIAM Journal of Numerical Analysis 21: 770-778.\n",
      "        .. [9] Powell, M J D. A direct search optimization method that models\n",
      "           the objective and constraint functions by linear interpolation.\n",
      "           1994. Advances in Optimization and Numerical Analysis, eds. S. Gomez\n",
      "           and J-P Hennart, Kluwer Academic (Dordrecht), 51-67.\n",
      "        .. [10] Powell M J D. Direct search algorithms for optimization\n",
      "           calculations. 1998. Acta Numerica 7: 287-336.\n",
      "        .. [11] Powell M J D. A view of algorithms for optimization without\n",
      "           derivatives. 2007.Cambridge University Technical Report DAMTP\n",
      "           2007/NA03\n",
      "        .. [12] Kraft, D. A software package for sequential quadratic\n",
      "           programming. 1988. Tech. Rep. DFVLR-FB 88-28, DLR German Aerospace\n",
      "           Center -- Institute for Flight Mechanics, Koln, Germany.\n",
      "        .. [13] Conn, A. R., Gould, N. I., and Toint, P. L.\n",
      "           Trust region methods. 2000. Siam. pp. 169-200.\n",
      "        .. [14] F. Lenders, C. Kirches, A. Potschka: \"trlib: A vector-free\n",
      "           implementation of the GLTR method for iterative solution of\n",
      "           the trust region problem\", :arxiv:`1611.04718`\n",
      "        .. [15] N. Gould, S. Lucidi, M. Roma, P. Toint: \"Solving the\n",
      "           Trust-Region Subproblem using the Lanczos Method\",\n",
      "           SIAM J. Optim., 9(2), 504--525, (1999).\n",
      "        .. [16] Byrd, Richard H., Mary E. Hribar, and Jorge Nocedal. 1999.\n",
      "            An interior point algorithm for large-scale nonlinear  programming.\n",
      "            SIAM Journal on Optimization 9.4: 877-900.\n",
      "        .. [17] Lalee, Marucha, Jorge Nocedal, and Todd Plantega. 1998. On the\n",
      "            implementation of an algorithm for large-scale equality constrained\n",
      "            optimization. SIAM Journal on Optimization 8.3: 682-706.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        Let us consider the problem of minimizing the Rosenbrock function. This\n",
      "        function (and its respective derivatives) is implemented in `rosen`\n",
      "        (resp. `rosen_der`, `rosen_hess`) in the `scipy.optimize`.\n",
      "        \n",
      "        >>> from scipy.optimize import minimize, rosen, rosen_der\n",
      "        \n",
      "        A simple application of the *Nelder-Mead* method is:\n",
      "        \n",
      "        >>> x0 = [1.3, 0.7, 0.8, 1.9, 1.2]\n",
      "        >>> res = minimize(rosen, x0, method='Nelder-Mead', tol=1e-6)\n",
      "        >>> res.x\n",
      "        array([ 1.,  1.,  1.,  1.,  1.])\n",
      "        \n",
      "        Now using the *BFGS* algorithm, using the first derivative and a few\n",
      "        options:\n",
      "        \n",
      "        >>> res = minimize(rosen, x0, method='BFGS', jac=rosen_der,\n",
      "        ...                options={'gtol': 1e-6, 'disp': True})\n",
      "        Optimization terminated successfully.\n",
      "                 Current function value: 0.000000\n",
      "                 Iterations: 26\n",
      "                 Function evaluations: 31\n",
      "                 Gradient evaluations: 31\n",
      "        >>> res.x\n",
      "        array([ 1.,  1.,  1.,  1.,  1.])\n",
      "        >>> print(res.message)\n",
      "        Optimization terminated successfully.\n",
      "        >>> res.hess_inv\n",
      "        array([[ 0.00749589,  0.01255155,  0.02396251,  0.04750988,  0.09495377],  # may vary\n",
      "               [ 0.01255155,  0.02510441,  0.04794055,  0.09502834,  0.18996269],\n",
      "               [ 0.02396251,  0.04794055,  0.09631614,  0.19092151,  0.38165151],\n",
      "               [ 0.04750988,  0.09502834,  0.19092151,  0.38341252,  0.7664427 ],\n",
      "               [ 0.09495377,  0.18996269,  0.38165151,  0.7664427,   1.53713523]])\n",
      "        \n",
      "        \n",
      "        Next, consider a minimization problem with several constraints (namely\n",
      "        Example 16.4 from [5]_). The objective function is:\n",
      "        \n",
      "        >>> fun = lambda x: (x[0] - 1)**2 + (x[1] - 2.5)**2\n",
      "        \n",
      "        There are three constraints defined as:\n",
      "        \n",
      "        >>> cons = ({'type': 'ineq', 'fun': lambda x:  x[0] - 2 * x[1] + 2},\n",
      "        ...         {'type': 'ineq', 'fun': lambda x: -x[0] - 2 * x[1] + 6},\n",
      "        ...         {'type': 'ineq', 'fun': lambda x: -x[0] + 2 * x[1] + 2})\n",
      "        \n",
      "        And variables must be positive, hence the following bounds:\n",
      "        \n",
      "        >>> bnds = ((0, None), (0, None))\n",
      "        \n",
      "        The optimization problem is solved using the SLSQP method as:\n",
      "        \n",
      "        >>> res = minimize(fun, (2, 0), method='SLSQP', bounds=bnds,\n",
      "        ...                constraints=cons)\n",
      "        \n",
      "        It should converge to the theoretical solution (1.4 ,1.7).\n",
      "    \n",
      "    minimize_scalar(fun, bracket=None, bounds=None, args=(), method='brent', tol=None, options=None)\n",
      "        Minimization of scalar function of one variable.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        fun : callable\n",
      "            Objective function.\n",
      "            Scalar function, must return a scalar.\n",
      "        bracket : sequence, optional\n",
      "            For methods 'brent' and 'golden', `bracket` defines the bracketing\n",
      "            interval and can either have three items ``(a, b, c)`` so that\n",
      "            ``a < b < c`` and ``fun(b) < fun(a), fun(c)`` or two items ``a`` and\n",
      "            ``c`` which are assumed to be a starting interval for a downhill\n",
      "            bracket search (see `bracket`); it doesn't always mean that the\n",
      "            obtained solution will satisfy ``a <= x <= c``.\n",
      "        bounds : sequence, optional\n",
      "            For method 'bounded', `bounds` is mandatory and must have two items\n",
      "            corresponding to the optimization bounds.\n",
      "        args : tuple, optional\n",
      "            Extra arguments passed to the objective function.\n",
      "        method : str or callable, optional\n",
      "            Type of solver.  Should be one of:\n",
      "        \n",
      "                - 'Brent'     :ref:`(see here) <optimize.minimize_scalar-brent>`\n",
      "                - 'Bounded'   :ref:`(see here) <optimize.minimize_scalar-bounded>`\n",
      "                - 'Golden'    :ref:`(see here) <optimize.minimize_scalar-golden>`\n",
      "                - custom - a callable object (added in version 0.14.0), see below\n",
      "        \n",
      "        tol : float, optional\n",
      "            Tolerance for termination. For detailed control, use solver-specific\n",
      "            options.\n",
      "        options : dict, optional\n",
      "            A dictionary of solver options.\n",
      "        \n",
      "                maxiter : int\n",
      "                    Maximum number of iterations to perform.\n",
      "                disp : bool\n",
      "                    Set to True to print convergence messages.\n",
      "        \n",
      "            See :func:`show_options()` for solver-specific options.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        res : OptimizeResult\n",
      "            The optimization result represented as a ``OptimizeResult`` object.\n",
      "            Important attributes are: ``x`` the solution array, ``success`` a\n",
      "            Boolean flag indicating if the optimizer exited successfully and\n",
      "            ``message`` which describes the cause of the termination. See\n",
      "            `OptimizeResult` for a description of other attributes.\n",
      "        \n",
      "        See also\n",
      "        --------\n",
      "        minimize : Interface to minimization algorithms for scalar multivariate\n",
      "            functions\n",
      "        show_options : Additional options accepted by the solvers\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        This section describes the available solvers that can be selected by the\n",
      "        'method' parameter. The default method is *Brent*.\n",
      "        \n",
      "        Method :ref:`Brent <optimize.minimize_scalar-brent>` uses Brent's\n",
      "        algorithm to find a local minimum.  The algorithm uses inverse\n",
      "        parabolic interpolation when possible to speed up convergence of\n",
      "        the golden section method.\n",
      "        \n",
      "        Method :ref:`Golden <optimize.minimize_scalar-golden>` uses the\n",
      "        golden section search technique. It uses analog of the bisection\n",
      "        method to decrease the bracketed interval. It is usually\n",
      "        preferable to use the *Brent* method.\n",
      "        \n",
      "        Method :ref:`Bounded <optimize.minimize_scalar-bounded>` can\n",
      "        perform bounded minimization. It uses the Brent method to find a\n",
      "        local minimum in the interval x1 < xopt < x2.\n",
      "        \n",
      "        **Custom minimizers**\n",
      "        \n",
      "        It may be useful to pass a custom minimization method, for example\n",
      "        when using some library frontend to minimize_scalar. You can simply\n",
      "        pass a callable as the ``method`` parameter.\n",
      "        \n",
      "        The callable is called as ``method(fun, args, **kwargs, **options)``\n",
      "        where ``kwargs`` corresponds to any other parameters passed to `minimize`\n",
      "        (such as `bracket`, `tol`, etc.), except the `options` dict, which has\n",
      "        its contents also passed as `method` parameters pair by pair.  The method\n",
      "        shall return an `OptimizeResult` object.\n",
      "        \n",
      "        The provided `method` callable must be able to accept (and possibly ignore)\n",
      "        arbitrary parameters; the set of parameters accepted by `minimize` may\n",
      "        expand in future versions and then these parameters will be passed to\n",
      "        the method. You can find an example in the scipy.optimize tutorial.\n",
      "        \n",
      "        .. versionadded:: 0.11.0\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        Consider the problem of minimizing the following function.\n",
      "        \n",
      "        >>> def f(x):\n",
      "        ...     return (x - 2) * x * (x + 2)**2\n",
      "        \n",
      "        Using the *Brent* method, we find the local minimum as:\n",
      "        \n",
      "        >>> from scipy.optimize import minimize_scalar\n",
      "        >>> res = minimize_scalar(f)\n",
      "        >>> res.x\n",
      "        1.28077640403\n",
      "        \n",
      "        Using the *Bounded* method, we find a local minimum with specified\n",
      "        bounds as:\n",
      "        \n",
      "        >>> res = minimize_scalar(f, bounds=(-3, -1), method='bounded')\n",
      "        >>> res.x\n",
      "        -2.0000002026\n",
      "    \n",
      "    newton(func, x0, fprime=None, args=(), tol=1.48e-08, maxiter=50, fprime2=None, x1=None, rtol=0.0, full_output=False, disp=True)\n",
      "        Find a zero of a real or complex function using the Newton-Raphson\n",
      "        (or secant or Halley's) method.\n",
      "        \n",
      "        Find a zero of the function `func` given a nearby starting point `x0`.\n",
      "        The Newton-Raphson method is used if the derivative `fprime` of `func`\n",
      "        is provided, otherwise the secant method is used. If the second order\n",
      "        derivative `fprime2` of `func` is also provided, then Halley's method is\n",
      "        used.\n",
      "        \n",
      "        If `x0` is a sequence with more than one item, then `newton` returns an\n",
      "        array, and `func` must be vectorized and return a sequence or array of the\n",
      "        same shape as its first argument. If `fprime` or `fprime2` is given, then\n",
      "        its return must also have the same shape.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        func : callable\n",
      "            The function whose zero is wanted. It must be a function of a\n",
      "            single variable of the form ``f(x,a,b,c...)``, where ``a,b,c...``\n",
      "            are extra arguments that can be passed in the `args` parameter.\n",
      "        x0 : float, sequence, or ndarray\n",
      "            An initial estimate of the zero that should be somewhere near the\n",
      "            actual zero. If not scalar, then `func` must be vectorized and return\n",
      "            a sequence or array of the same shape as its first argument.\n",
      "        fprime : callable, optional\n",
      "            The derivative of the function when available and convenient. If it\n",
      "            is None (default), then the secant method is used.\n",
      "        args : tuple, optional\n",
      "            Extra arguments to be used in the function call.\n",
      "        tol : float, optional\n",
      "            The allowable error of the zero value. If `func` is complex-valued,\n",
      "            a larger `tol` is recommended as both the real and imaginary parts\n",
      "            of `x` contribute to ``|x - x0|``.\n",
      "        maxiter : int, optional\n",
      "            Maximum number of iterations.\n",
      "        fprime2 : callable, optional\n",
      "            The second order derivative of the function when available and\n",
      "            convenient. If it is None (default), then the normal Newton-Raphson\n",
      "            or the secant method is used. If it is not None, then Halley's method\n",
      "            is used.\n",
      "        x1 : float, optional\n",
      "            Another estimate of the zero that should be somewhere near the\n",
      "            actual zero. Used if `fprime` is not provided.\n",
      "        rtol : float, optional\n",
      "            Tolerance (relative) for termination.\n",
      "        full_output : bool, optional\n",
      "            If `full_output` is False (default), the root is returned.\n",
      "            If True and `x0` is scalar, the return value is ``(x, r)``, where ``x``\n",
      "            is the root and ``r`` is a `RootResults` object.\n",
      "            If True and `x0` is non-scalar, the return value is ``(x, converged,\n",
      "            zero_der)`` (see Returns section for details).\n",
      "        disp : bool, optional\n",
      "            If True, raise a RuntimeError if the algorithm didn't converge, with\n",
      "            the error message containing the number of iterations and current\n",
      "            function value. Otherwise, the convergence status is recorded in a\n",
      "            `RootResults` return object.\n",
      "            Ignored if `x0` is not scalar.\n",
      "            *Note: this has little to do with displaying, however,\n",
      "            the `disp` keyword cannot be renamed for backwards compatibility.*\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        root : float, sequence, or ndarray\n",
      "            Estimated location where function is zero.\n",
      "        r : `RootResults`, optional\n",
      "            Present if ``full_output=True`` and `x0` is scalar.\n",
      "            Object containing information about the convergence. In particular,\n",
      "            ``r.converged`` is True if the routine converged.\n",
      "        converged : ndarray of bool, optional\n",
      "            Present if ``full_output=True`` and `x0` is non-scalar.\n",
      "            For vector functions, indicates which elements converged successfully.\n",
      "        zero_der : ndarray of bool, optional\n",
      "            Present if ``full_output=True`` and `x0` is non-scalar.\n",
      "            For vector functions, indicates which elements had a zero derivative.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        brentq, brenth, ridder, bisect\n",
      "        fsolve : find zeros in N dimensions.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        The convergence rate of the Newton-Raphson method is quadratic,\n",
      "        the Halley method is cubic, and the secant method is\n",
      "        sub-quadratic. This means that if the function is well-behaved\n",
      "        the actual error in the estimated zero after the nth iteration\n",
      "        is approximately the square (cube for Halley) of the error\n",
      "        after the (n-1)th step. However, the stopping criterion used\n",
      "        here is the step size and there is no guarantee that a zero\n",
      "        has been found. Consequently, the result should be verified.\n",
      "        Safer algorithms are brentq, brenth, ridder, and bisect,\n",
      "        but they all require that the root first be bracketed in an\n",
      "        interval where the function changes sign. The brentq algorithm\n",
      "        is recommended for general use in one dimensional problems\n",
      "        when such an interval has been found.\n",
      "        \n",
      "        When `newton` is used with arrays, it is best suited for the following\n",
      "        types of problems:\n",
      "        \n",
      "        * The initial guesses, `x0`, are all relatively the same distance from\n",
      "          the roots.\n",
      "        * Some or all of the extra arguments, `args`, are also arrays so that a\n",
      "          class of similar problems can be solved together.\n",
      "        * The size of the initial guesses, `x0`, is larger than O(100) elements.\n",
      "          Otherwise, a naive loop may perform as well or better than a vector.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy import optimize\n",
      "        >>> import matplotlib.pyplot as plt\n",
      "        \n",
      "        >>> def f(x):\n",
      "        ...     return (x**3 - 1)  # only one real root at x = 1\n",
      "        \n",
      "        ``fprime`` is not provided, use the secant method:\n",
      "        \n",
      "        >>> root = optimize.newton(f, 1.5)\n",
      "        >>> root\n",
      "        1.0000000000000016\n",
      "        >>> root = optimize.newton(f, 1.5, fprime2=lambda x: 6 * x)\n",
      "        >>> root\n",
      "        1.0000000000000016\n",
      "        \n",
      "        Only ``fprime`` is provided, use the Newton-Raphson method:\n",
      "        \n",
      "        >>> root = optimize.newton(f, 1.5, fprime=lambda x: 3 * x**2)\n",
      "        >>> root\n",
      "        1.0\n",
      "        \n",
      "        Both ``fprime2`` and ``fprime`` are provided, use Halley's method:\n",
      "        \n",
      "        >>> root = optimize.newton(f, 1.5, fprime=lambda x: 3 * x**2,\n",
      "        ...                        fprime2=lambda x: 6 * x)\n",
      "        >>> root\n",
      "        1.0\n",
      "        \n",
      "        When we want to find zeros for a set of related starting values and/or\n",
      "        function parameters, we can provide both of those as an array of inputs:\n",
      "        \n",
      "        >>> f = lambda x, a: x**3 - a\n",
      "        >>> fder = lambda x, a: 3 * x**2\n",
      "        >>> rng = np.random.default_rng()\n",
      "        >>> x = rng.standard_normal(100)\n",
      "        >>> a = np.arange(-50, 50)\n",
      "        >>> vec_res = optimize.newton(f, x, fprime=fder, args=(a, ), maxiter=200)\n",
      "        \n",
      "        The above is the equivalent of solving for each value in ``(x, a)``\n",
      "        separately in a for-loop, just faster:\n",
      "        \n",
      "        >>> loop_res = [optimize.newton(f, x0, fprime=fder, args=(a0,))\n",
      "        ...             for x0, a0 in zip(x, a)]\n",
      "        >>> np.allclose(vec_res, loop_res)\n",
      "        True\n",
      "        \n",
      "        Plot the results found for all values of ``a``:\n",
      "        \n",
      "        >>> analytical_result = np.sign(a) * np.abs(a)**(1/3)\n",
      "        >>> fig = plt.figure()\n",
      "        >>> ax = fig.add_subplot(111)\n",
      "        >>> ax.plot(a, analytical_result, 'o')\n",
      "        >>> ax.plot(a, vec_res, '.')\n",
      "        >>> ax.set_xlabel('$a$')\n",
      "        >>> ax.set_ylabel('$x$ where $f(x, a)=0$')\n",
      "        >>> plt.show()\n",
      "    \n",
      "    newton_krylov(F, xin, iter=None, rdiff=None, method='lgmres', inner_maxiter=20, inner_M=None, outer_k=10, verbose=False, maxiter=None, f_tol=None, f_rtol=None, x_tol=None, x_rtol=None, tol_norm=None, line_search='armijo', callback=None, **kw)\n",
      "        Find a root of a function, using Krylov approximation for inverse Jacobian.\n",
      "        \n",
      "        This method is suitable for solving large-scale problems.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        F : function(x) -> f\n",
      "            Function whose root to find; should take and return an array-like\n",
      "            object.\n",
      "        xin : array_like\n",
      "            Initial guess for the solution\n",
      "        rdiff : float, optional\n",
      "            Relative step size to use in numerical differentiation.\n",
      "        method : {'lgmres', 'gmres', 'bicgstab', 'cgs', 'minres'} or function\n",
      "            Krylov method to use to approximate the Jacobian.\n",
      "            Can be a string, or a function implementing the same interface as\n",
      "            the iterative solvers in `scipy.sparse.linalg`.\n",
      "        \n",
      "            The default is `scipy.sparse.linalg.lgmres`.\n",
      "        inner_maxiter : int, optional\n",
      "            Parameter to pass to the \"inner\" Krylov solver: maximum number of\n",
      "            iterations. Iteration will stop after maxiter steps even if the\n",
      "            specified tolerance has not been achieved.\n",
      "        inner_M : LinearOperator or InverseJacobian\n",
      "            Preconditioner for the inner Krylov iteration.\n",
      "            Note that you can use also inverse Jacobians as (adaptive)\n",
      "            preconditioners. For example,\n",
      "        \n",
      "            >>> from scipy.optimize.nonlin import BroydenFirst, KrylovJacobian\n",
      "            >>> from scipy.optimize.nonlin import InverseJacobian\n",
      "            >>> jac = BroydenFirst()\n",
      "            >>> kjac = KrylovJacobian(inner_M=InverseJacobian(jac))\n",
      "        \n",
      "            If the preconditioner has a method named 'update', it will be called\n",
      "            as ``update(x, f)`` after each nonlinear step, with ``x`` giving\n",
      "            the current point, and ``f`` the current function value.\n",
      "        outer_k : int, optional\n",
      "            Size of the subspace kept across LGMRES nonlinear iterations.\n",
      "            See `scipy.sparse.linalg.lgmres` for details.\n",
      "        inner_kwargs : kwargs\n",
      "            Keyword parameters for the \"inner\" Krylov solver\n",
      "            (defined with `method`). Parameter names must start with\n",
      "            the `inner_` prefix which will be stripped before passing on\n",
      "            the inner method. See, e.g., `scipy.sparse.linalg.gmres` for details.\n",
      "        iter : int, optional\n",
      "            Number of iterations to make. If omitted (default), make as many\n",
      "            as required to meet tolerances.\n",
      "        verbose : bool, optional\n",
      "            Print status to stdout on every iteration.\n",
      "        maxiter : int, optional\n",
      "            Maximum number of iterations to make. If more are needed to\n",
      "            meet convergence, `NoConvergence` is raised.\n",
      "        f_tol : float, optional\n",
      "            Absolute tolerance (in max-norm) for the residual.\n",
      "            If omitted, default is 6e-6.\n",
      "        f_rtol : float, optional\n",
      "            Relative tolerance for the residual. If omitted, not used.\n",
      "        x_tol : float, optional\n",
      "            Absolute minimum step size, as determined from the Jacobian\n",
      "            approximation. If the step size is smaller than this, optimization\n",
      "            is terminated as successful. If omitted, not used.\n",
      "        x_rtol : float, optional\n",
      "            Relative minimum step size. If omitted, not used.\n",
      "        tol_norm : function(vector) -> scalar, optional\n",
      "            Norm to use in convergence check. Default is the maximum norm.\n",
      "        line_search : {None, 'armijo' (default), 'wolfe'}, optional\n",
      "            Which type of a line search to use to determine the step size in the\n",
      "            direction given by the Jacobian approximation. Defaults to 'armijo'.\n",
      "        callback : function, optional\n",
      "            Optional callback function. It is called on every iteration as\n",
      "            ``callback(x, f)`` where `x` is the current solution and `f`\n",
      "            the corresponding residual.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        sol : ndarray\n",
      "            An array (of similar array type as `x0`) containing the final solution.\n",
      "        \n",
      "        Raises\n",
      "        ------\n",
      "        NoConvergence\n",
      "            When a solution was not found.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        root : Interface to root finding algorithms for multivariate\n",
      "               functions. See ``method=='krylov'`` in particular.\n",
      "        scipy.sparse.linalg.gmres\n",
      "        scipy.sparse.linalg.lgmres\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        This function implements a Newton-Krylov solver. The basic idea is\n",
      "        to compute the inverse of the Jacobian with an iterative Krylov\n",
      "        method. These methods require only evaluating the Jacobian-vector\n",
      "        products, which are conveniently approximated by a finite difference:\n",
      "        \n",
      "        .. math:: J v \\approx (f(x + \\omega*v/|v|) - f(x)) / \\omega\n",
      "        \n",
      "        Due to the use of iterative matrix inverses, these methods can\n",
      "        deal with large nonlinear problems.\n",
      "        \n",
      "        SciPy's `scipy.sparse.linalg` module offers a selection of Krylov\n",
      "        solvers to choose from. The default here is `lgmres`, which is a\n",
      "        variant of restarted GMRES iteration that reuses some of the\n",
      "        information obtained in the previous Newton steps to invert\n",
      "        Jacobians in subsequent steps.\n",
      "        \n",
      "        For a review on Newton-Krylov methods, see for example [1]_,\n",
      "        and for the LGMRES sparse inverse method, see [2]_.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] D.A. Knoll and D.E. Keyes, J. Comp. Phys. 193, 357 (2004).\n",
      "               :doi:`10.1016/j.jcp.2003.08.010`\n",
      "        .. [2] A.H. Baker and E.R. Jessup and T. Manteuffel,\n",
      "               SIAM J. Matrix Anal. Appl. 26, 962 (2005).\n",
      "               :doi:`10.1137/S0895479803422014`\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        The following functions define a system of nonlinear equations\n",
      "        \n",
      "        >>> def fun(x):\n",
      "        ...     return [x[0] + 0.5 * x[1] - 1.0,\n",
      "        ...             0.5 * (x[1] - x[0]) ** 2]\n",
      "        \n",
      "        A solution can be obtained as follows.\n",
      "        \n",
      "        >>> from scipy import optimize\n",
      "        >>> sol = optimize.newton_krylov(fun, [0, 0])\n",
      "        >>> sol\n",
      "        array([0.66731771, 0.66536458])\n",
      "    \n",
      "    nnls(A, b, maxiter=None)\n",
      "        Solve ``argmin_x || Ax - b ||_2`` for ``x>=0``. This is a wrapper\n",
      "        for a FORTRAN non-negative least squares solver.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        A : ndarray\n",
      "            Matrix ``A`` as shown above.\n",
      "        b : ndarray\n",
      "            Right-hand side vector.\n",
      "        maxiter: int, optional\n",
      "            Maximum number of iterations, optional.\n",
      "            Default is ``3 * A.shape[1]``.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        x : ndarray\n",
      "            Solution vector.\n",
      "        rnorm : float\n",
      "            The residual, ``|| Ax-b ||_2``.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        lsq_linear : Linear least squares with bounds on the variables\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        The FORTRAN code was published in the book below. The algorithm\n",
      "        is an active set method. It solves the KKT (Karush-Kuhn-Tucker)\n",
      "        conditions for the non-negative least squares problem.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        Lawson C., Hanson R.J., (1987) Solving Least Squares Problems, SIAM\n",
      "        \n",
      "         Examples\n",
      "        --------\n",
      "        >>> from scipy.optimize import nnls\n",
      "        ...\n",
      "        >>> A = np.array([[1, 0], [1, 0], [0, 1]])\n",
      "        >>> b = np.array([2, 1, 1])\n",
      "        >>> nnls(A, b)\n",
      "        (array([1.5, 1. ]), 0.7071067811865475)\n",
      "        \n",
      "        >>> b = np.array([-1, -1, -1])\n",
      "        >>> nnls(A, b)\n",
      "        (array([0., 0.]), 1.7320508075688772)\n",
      "    \n",
      "    quadratic_assignment(A, B, method='faq', options=None)\n",
      "        Approximates solution to the quadratic assignment problem and\n",
      "        the graph matching problem.\n",
      "        \n",
      "        Quadratic assignment solves problems of the following form:\n",
      "        \n",
      "        .. math::\n",
      "        \n",
      "            \\min_P & \\ {\\ \\text{trace}(A^T P B P^T)}\\\\\n",
      "            \\mbox{s.t. } & {P \\ \\epsilon \\ \\mathcal{P}}\\\\\n",
      "        \n",
      "        where :math:`\\mathcal{P}` is the set of all permutation matrices,\n",
      "        and :math:`A` and :math:`B` are square matrices.\n",
      "        \n",
      "        Graph matching tries to *maximize* the same objective function.\n",
      "        This algorithm can be thought of as finding the alignment of the\n",
      "        nodes of two graphs that minimizes the number of induced edge\n",
      "        disagreements, or, in the case of weighted graphs, the sum of squared\n",
      "        edge weight differences.\n",
      "        \n",
      "        Note that the quadratic assignment problem is NP-hard. The results given\n",
      "        here are approximations and are not guaranteed to be optimal.\n",
      "        \n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        A : 2-D array, square\n",
      "            The square matrix :math:`A` in the objective function above.\n",
      "        \n",
      "        B : 2-D array, square\n",
      "            The square matrix :math:`B` in the objective function above.\n",
      "        \n",
      "        method :  str in {'faq', '2opt'} (default: 'faq')\n",
      "            The algorithm used to solve the problem.\n",
      "            :ref:`'faq' <optimize.qap-faq>` (default) and\n",
      "            :ref:`'2opt' <optimize.qap-2opt>` are available.\n",
      "        \n",
      "        options : dict, optional\n",
      "            A dictionary of solver options. All solvers support the following:\n",
      "        \n",
      "            maximize : bool (default: False)\n",
      "                Maximizes the objective function if ``True``.\n",
      "        \n",
      "            partial_match : 2-D array of integers, optional (default: None)\n",
      "                Fixes part of the matching. Also known as a \"seed\" [2]_.\n",
      "        \n",
      "                Each row of `partial_match` specifies a pair of matched nodes:\n",
      "                node ``partial_match[i, 0]`` of `A` is matched to node\n",
      "                ``partial_match[i, 1]`` of `B`. The array has shape ``(m, 2)``,\n",
      "                where ``m`` is not greater than the number of nodes, :math:`n`.\n",
      "        \n",
      "            rng : {None, int, `numpy.random.Generator`,\n",
      "                   `numpy.random.RandomState`}, optional\n",
      "        \n",
      "                If `seed` is None (or `np.random`), the `numpy.random.RandomState`\n",
      "                singleton is used.\n",
      "                If `seed` is an int, a new ``RandomState`` instance is used,\n",
      "                seeded with `seed`.\n",
      "                If `seed` is already a ``Generator`` or ``RandomState`` instance then\n",
      "                that instance is used.\n",
      "        \n",
      "            For method-specific options, see\n",
      "            :func:`show_options('quadratic_assignment') <show_options>`.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        res : OptimizeResult\n",
      "            `OptimizeResult` containing the following fields.\n",
      "        \n",
      "            col_ind : 1-D array\n",
      "                Column indices corresponding to the best permutation found of the\n",
      "                nodes of `B`.\n",
      "            fun : float\n",
      "                The objective value of the solution.\n",
      "            nit : int\n",
      "                The number of iterations performed during optimization.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        The default method :ref:`'faq' <optimize.qap-faq>` uses the Fast\n",
      "        Approximate QAP algorithm [1]_; it typically offers the best combination of\n",
      "        speed and accuracy.\n",
      "        Method :ref:`'2opt' <optimize.qap-2opt>` can be computationally expensive,\n",
      "        but may be a useful alternative, or it can be used to refine the solution\n",
      "        returned by another method.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] J.T. Vogelstein, J.M. Conroy, V. Lyzinski, L.J. Podrazik,\n",
      "               S.G. Kratzer, E.T. Harley, D.E. Fishkind, R.J. Vogelstein, and\n",
      "               C.E. Priebe, \"Fast approximate quadratic programming for graph\n",
      "               matching,\" PLOS one, vol. 10, no. 4, p. e0121002, 2015,\n",
      "               :doi:`10.1371/journal.pone.0121002`\n",
      "        \n",
      "        .. [2] D. Fishkind, S. Adali, H. Patsolic, L. Meng, D. Singh, V. Lyzinski,\n",
      "               C. Priebe, \"Seeded graph matching\", Pattern Recognit. 87 (2019):\n",
      "               203-215, :doi:`10.1016/j.patcog.2018.09.014`\n",
      "        \n",
      "        .. [3] \"2-opt,\" Wikipedia.\n",
      "               https://en.wikipedia.org/wiki/2-opt\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy.optimize import quadratic_assignment\n",
      "        >>> A = np.array([[0, 80, 150, 170], [80, 0, 130, 100],\n",
      "        ...               [150, 130, 0, 120], [170, 100, 120, 0]])\n",
      "        >>> B = np.array([[0, 5, 2, 7], [0, 0, 3, 8],\n",
      "        ...               [0, 0, 0, 3], [0, 0, 0, 0]])\n",
      "        >>> res = quadratic_assignment(A, B)\n",
      "        >>> print(res)\n",
      "         col_ind: array([0, 3, 2, 1])\n",
      "             fun: 3260\n",
      "             nit: 9\n",
      "        \n",
      "        The see the relationship between the returned ``col_ind`` and ``fun``,\n",
      "        use ``col_ind`` to form the best permutation matrix found, then evaluate\n",
      "        the objective function :math:`f(P) = trace(A^T P B P^T )`.\n",
      "        \n",
      "        >>> perm = res['col_ind']\n",
      "        >>> P = np.eye(len(A), dtype=int)[perm]\n",
      "        >>> fun = np.trace(A.T @ P @ B @ P.T)\n",
      "        >>> print(fun)\n",
      "        3260\n",
      "        \n",
      "        Alternatively, to avoid constructing the permutation matrix explicitly,\n",
      "        directly permute the rows and columns of the distance matrix.\n",
      "        \n",
      "        >>> fun = np.trace(A.T @ B[perm][:, perm])\n",
      "        >>> print(fun)\n",
      "        3260\n",
      "        \n",
      "        Although not guaranteed in general, ``quadratic_assignment`` happens to\n",
      "        have found the globally optimal solution.\n",
      "        \n",
      "        >>> from itertools import permutations\n",
      "        >>> perm_opt, fun_opt = None, np.inf\n",
      "        >>> for perm in permutations([0, 1, 2, 3]):\n",
      "        ...     perm = np.array(perm)\n",
      "        ...     fun = np.trace(A.T @ B[perm][:, perm])\n",
      "        ...     if fun < fun_opt:\n",
      "        ...         fun_opt, perm_opt = fun, perm\n",
      "        >>> print(np.array_equal(perm_opt, res['col_ind']))\n",
      "        True\n",
      "        \n",
      "        Here is an example for which the default method,\n",
      "        :ref:`'faq' <optimize.qap-faq>`, does not find the global optimum.\n",
      "        \n",
      "        >>> A = np.array([[0, 5, 8, 6], [5, 0, 5, 1],\n",
      "        ...               [8, 5, 0, 2], [6, 1, 2, 0]])\n",
      "        >>> B = np.array([[0, 1, 8, 4], [1, 0, 5, 2],\n",
      "        ...               [8, 5, 0, 5], [4, 2, 5, 0]])\n",
      "        >>> res = quadratic_assignment(A, B)\n",
      "        >>> print(res)\n",
      "         col_ind: array([1, 0, 3, 2])\n",
      "             fun: 178\n",
      "             nit: 13\n",
      "        \n",
      "        If accuracy is important, consider using  :ref:`'2opt' <optimize.qap-2opt>`\n",
      "        to refine the solution.\n",
      "        \n",
      "        >>> guess = np.array([np.arange(len(A)), res.col_ind]).T\n",
      "        >>> res = quadratic_assignment(A, B, method=\"2opt\",\n",
      "        ...                            options = {'partial_guess': guess})\n",
      "        >>> print(res)\n",
      "         col_ind: array([1, 2, 3, 0])\n",
      "             fun: 176\n",
      "             nit: 17\n",
      "    \n",
      "    ridder(f, a, b, args=(), xtol=2e-12, rtol=8.881784197001252e-16, maxiter=100, full_output=False, disp=True)\n",
      "        Find a root of a function in an interval using Ridder's method.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        f : function\n",
      "            Python function returning a number. f must be continuous, and f(a) and\n",
      "            f(b) must have opposite signs.\n",
      "        a : scalar\n",
      "            One end of the bracketing interval [a,b].\n",
      "        b : scalar\n",
      "            The other end of the bracketing interval [a,b].\n",
      "        xtol : number, optional\n",
      "            The computed root ``x0`` will satisfy ``np.allclose(x, x0,\n",
      "            atol=xtol, rtol=rtol)``, where ``x`` is the exact root. The\n",
      "            parameter must be nonnegative.\n",
      "        rtol : number, optional\n",
      "            The computed root ``x0`` will satisfy ``np.allclose(x, x0,\n",
      "            atol=xtol, rtol=rtol)``, where ``x`` is the exact root. The\n",
      "            parameter cannot be smaller than its default value of\n",
      "            ``4*np.finfo(float).eps``.\n",
      "        maxiter : int, optional\n",
      "            If convergence is not achieved in `maxiter` iterations, an error is\n",
      "            raised. Must be >= 0.\n",
      "        args : tuple, optional\n",
      "            Containing extra arguments for the function `f`.\n",
      "            `f` is called by ``apply(f, (x)+args)``.\n",
      "        full_output : bool, optional\n",
      "            If `full_output` is False, the root is returned. If `full_output` is\n",
      "            True, the return value is ``(x, r)``, where `x` is the root, and `r` is\n",
      "            a `RootResults` object.\n",
      "        disp : bool, optional\n",
      "            If True, raise RuntimeError if the algorithm didn't converge.\n",
      "            Otherwise, the convergence status is recorded in any `RootResults`\n",
      "            return object.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        x0 : float\n",
      "            Zero of `f` between `a` and `b`.\n",
      "        r : `RootResults` (present if ``full_output = True``)\n",
      "            Object containing information about the convergence.\n",
      "            In particular, ``r.converged`` is True if the routine converged.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        brentq, brenth, bisect, newton : 1-D root-finding\n",
      "        fixed_point : scalar fixed-point finder\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        Uses [Ridders1979]_ method to find a zero of the function `f` between the\n",
      "        arguments `a` and `b`. Ridders' method is faster than bisection, but not\n",
      "        generally as fast as the Brent routines. [Ridders1979]_ provides the\n",
      "        classic description and source of the algorithm. A description can also be\n",
      "        found in any recent edition of Numerical Recipes.\n",
      "        \n",
      "        The routine used here diverges slightly from standard presentations in\n",
      "        order to be a bit more careful of tolerance.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [Ridders1979]\n",
      "           Ridders, C. F. J. \"A New Algorithm for Computing a\n",
      "           Single Root of a Real Continuous Function.\"\n",
      "           IEEE Trans. Circuits Systems 26, 979-980, 1979.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        \n",
      "        >>> def f(x):\n",
      "        ...     return (x**2 - 1)\n",
      "        \n",
      "        >>> from scipy import optimize\n",
      "        \n",
      "        >>> root = optimize.ridder(f, 0, 2)\n",
      "        >>> root\n",
      "        1.0\n",
      "        \n",
      "        >>> root = optimize.ridder(f, -2, 0)\n",
      "        >>> root\n",
      "        -1.0\n",
      "    \n",
      "    root(fun, x0, args=(), method='hybr', jac=None, tol=None, callback=None, options=None)\n",
      "        Find a root of a vector function.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        fun : callable\n",
      "            A vector function to find a root of.\n",
      "        x0 : ndarray\n",
      "            Initial guess.\n",
      "        args : tuple, optional\n",
      "            Extra arguments passed to the objective function and its Jacobian.\n",
      "        method : str, optional\n",
      "            Type of solver. Should be one of\n",
      "        \n",
      "                - 'hybr'             :ref:`(see here) <optimize.root-hybr>`\n",
      "                - 'lm'               :ref:`(see here) <optimize.root-lm>`\n",
      "                - 'broyden1'         :ref:`(see here) <optimize.root-broyden1>`\n",
      "                - 'broyden2'         :ref:`(see here) <optimize.root-broyden2>`\n",
      "                - 'anderson'         :ref:`(see here) <optimize.root-anderson>`\n",
      "                - 'linearmixing'     :ref:`(see here) <optimize.root-linearmixing>`\n",
      "                - 'diagbroyden'      :ref:`(see here) <optimize.root-diagbroyden>`\n",
      "                - 'excitingmixing'   :ref:`(see here) <optimize.root-excitingmixing>`\n",
      "                - 'krylov'           :ref:`(see here) <optimize.root-krylov>`\n",
      "                - 'df-sane'          :ref:`(see here) <optimize.root-dfsane>`\n",
      "        \n",
      "        jac : bool or callable, optional\n",
      "            If `jac` is a Boolean and is True, `fun` is assumed to return the\n",
      "            value of Jacobian along with the objective function. If False, the\n",
      "            Jacobian will be estimated numerically.\n",
      "            `jac` can also be a callable returning the Jacobian of `fun`. In\n",
      "            this case, it must accept the same arguments as `fun`.\n",
      "        tol : float, optional\n",
      "            Tolerance for termination. For detailed control, use solver-specific\n",
      "            options.\n",
      "        callback : function, optional\n",
      "            Optional callback function. It is called on every iteration as\n",
      "            ``callback(x, f)`` where `x` is the current solution and `f`\n",
      "            the corresponding residual. For all methods but 'hybr' and 'lm'.\n",
      "        options : dict, optional\n",
      "            A dictionary of solver options. E.g., `xtol` or `maxiter`, see\n",
      "            :obj:`show_options()` for details.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        sol : OptimizeResult\n",
      "            The solution represented as a ``OptimizeResult`` object.\n",
      "            Important attributes are: ``x`` the solution array, ``success`` a\n",
      "            Boolean flag indicating if the algorithm exited successfully and\n",
      "            ``message`` which describes the cause of the termination. See\n",
      "            `OptimizeResult` for a description of other attributes.\n",
      "        \n",
      "        See also\n",
      "        --------\n",
      "        show_options : Additional options accepted by the solvers\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        This section describes the available solvers that can be selected by the\n",
      "        'method' parameter. The default method is *hybr*.\n",
      "        \n",
      "        Method *hybr* uses a modification of the Powell hybrid method as\n",
      "        implemented in MINPACK [1]_.\n",
      "        \n",
      "        Method *lm* solves the system of nonlinear equations in a least squares\n",
      "        sense using a modification of the Levenberg-Marquardt algorithm as\n",
      "        implemented in MINPACK [1]_.\n",
      "        \n",
      "        Method *df-sane* is a derivative-free spectral method. [3]_\n",
      "        \n",
      "        Methods *broyden1*, *broyden2*, *anderson*, *linearmixing*,\n",
      "        *diagbroyden*, *excitingmixing*, *krylov* are inexact Newton methods,\n",
      "        with backtracking or full line searches [2]_. Each method corresponds\n",
      "        to a particular Jacobian approximations. See `nonlin` for details.\n",
      "        \n",
      "        - Method *broyden1* uses Broyden's first Jacobian approximation, it is\n",
      "          known as Broyden's good method.\n",
      "        - Method *broyden2* uses Broyden's second Jacobian approximation, it\n",
      "          is known as Broyden's bad method.\n",
      "        - Method *anderson* uses (extended) Anderson mixing.\n",
      "        - Method *Krylov* uses Krylov approximation for inverse Jacobian. It\n",
      "          is suitable for large-scale problem.\n",
      "        - Method *diagbroyden* uses diagonal Broyden Jacobian approximation.\n",
      "        - Method *linearmixing* uses a scalar Jacobian approximation.\n",
      "        - Method *excitingmixing* uses a tuned diagonal Jacobian\n",
      "          approximation.\n",
      "        \n",
      "        .. warning::\n",
      "        \n",
      "            The algorithms implemented for methods *diagbroyden*,\n",
      "            *linearmixing* and *excitingmixing* may be useful for specific\n",
      "            problems, but whether they will work may depend strongly on the\n",
      "            problem.\n",
      "        \n",
      "        .. versionadded:: 0.11.0\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] More, Jorge J., Burton S. Garbow, and Kenneth E. Hillstrom.\n",
      "           1980. User Guide for MINPACK-1.\n",
      "        .. [2] C. T. Kelley. 1995. Iterative Methods for Linear and Nonlinear\n",
      "           Equations. Society for Industrial and Applied Mathematics.\n",
      "           <https://archive.siam.org/books/kelley/fr16/>\n",
      "        .. [3] W. La Cruz, J.M. Martinez, M. Raydan. Math. Comp. 75, 1429 (2006).\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        The following functions define a system of nonlinear equations and its\n",
      "        jacobian.\n",
      "        \n",
      "        >>> def fun(x):\n",
      "        ...     return [x[0]  + 0.5 * (x[0] - x[1])**3 - 1.0,\n",
      "        ...             0.5 * (x[1] - x[0])**3 + x[1]]\n",
      "        \n",
      "        >>> def jac(x):\n",
      "        ...     return np.array([[1 + 1.5 * (x[0] - x[1])**2,\n",
      "        ...                       -1.5 * (x[0] - x[1])**2],\n",
      "        ...                      [-1.5 * (x[1] - x[0])**2,\n",
      "        ...                       1 + 1.5 * (x[1] - x[0])**2]])\n",
      "        \n",
      "        A solution can be obtained as follows.\n",
      "        \n",
      "        >>> from scipy import optimize\n",
      "        >>> sol = optimize.root(fun, [0, 0], jac=jac, method='hybr')\n",
      "        >>> sol.x\n",
      "        array([ 0.8411639,  0.1588361])\n",
      "    \n",
      "    root_scalar(f, args=(), method=None, bracket=None, fprime=None, fprime2=None, x0=None, x1=None, xtol=None, rtol=None, maxiter=None, options=None)\n",
      "        Find a root of a scalar function.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        f : callable\n",
      "            A function to find a root of.\n",
      "        args : tuple, optional\n",
      "            Extra arguments passed to the objective function and its derivative(s).\n",
      "        method : str, optional\n",
      "            Type of solver.  Should be one of\n",
      "        \n",
      "                - 'bisect'    :ref:`(see here) <optimize.root_scalar-bisect>`\n",
      "                - 'brentq'    :ref:`(see here) <optimize.root_scalar-brentq>`\n",
      "                - 'brenth'    :ref:`(see here) <optimize.root_scalar-brenth>`\n",
      "                - 'ridder'    :ref:`(see here) <optimize.root_scalar-ridder>`\n",
      "                - 'toms748'    :ref:`(see here) <optimize.root_scalar-toms748>`\n",
      "                - 'newton'    :ref:`(see here) <optimize.root_scalar-newton>`\n",
      "                - 'secant'    :ref:`(see here) <optimize.root_scalar-secant>`\n",
      "                - 'halley'    :ref:`(see here) <optimize.root_scalar-halley>`\n",
      "        \n",
      "        bracket: A sequence of 2 floats, optional\n",
      "            An interval bracketing a root.  `f(x, *args)` must have different\n",
      "            signs at the two endpoints.\n",
      "        x0 : float, optional\n",
      "            Initial guess.\n",
      "        x1 : float, optional\n",
      "            A second guess.\n",
      "        fprime : bool or callable, optional\n",
      "            If `fprime` is a boolean and is True, `f` is assumed to return the\n",
      "            value of the objective function and of the derivative.\n",
      "            `fprime` can also be a callable returning the derivative of `f`. In\n",
      "            this case, it must accept the same arguments as `f`.\n",
      "        fprime2 : bool or callable, optional\n",
      "            If `fprime2` is a boolean and is True, `f` is assumed to return the\n",
      "            value of the objective function and of the\n",
      "            first and second derivatives.\n",
      "            `fprime2` can also be a callable returning the second derivative of `f`.\n",
      "            In this case, it must accept the same arguments as `f`.\n",
      "        xtol : float, optional\n",
      "            Tolerance (absolute) for termination.\n",
      "        rtol : float, optional\n",
      "            Tolerance (relative) for termination.\n",
      "        maxiter : int, optional\n",
      "            Maximum number of iterations.\n",
      "        options : dict, optional\n",
      "            A dictionary of solver options. E.g., ``k``, see\n",
      "            :obj:`show_options()` for details.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        sol : RootResults\n",
      "            The solution represented as a ``RootResults`` object.\n",
      "            Important attributes are: ``root`` the solution , ``converged`` a\n",
      "            boolean flag indicating if the algorithm exited successfully and\n",
      "            ``flag`` which describes the cause of the termination. See\n",
      "            `RootResults` for a description of other attributes.\n",
      "        \n",
      "        See also\n",
      "        --------\n",
      "        show_options : Additional options accepted by the solvers\n",
      "        root : Find a root of a vector function.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        This section describes the available solvers that can be selected by the\n",
      "        'method' parameter.\n",
      "        \n",
      "        The default is to use the best method available for the situation\n",
      "        presented.\n",
      "        If a bracket is provided, it may use one of the bracketing methods.\n",
      "        If a derivative and an initial value are specified, it may\n",
      "        select one of the derivative-based methods.\n",
      "        If no method is judged applicable, it will raise an Exception.\n",
      "        \n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        \n",
      "        Find the root of a simple cubic\n",
      "        \n",
      "        >>> from scipy import optimize\n",
      "        >>> def f(x):\n",
      "        ...     return (x**3 - 1)  # only one real root at x = 1\n",
      "        \n",
      "        >>> def fprime(x):\n",
      "        ...     return 3*x**2\n",
      "        \n",
      "        The `brentq` method takes as input a bracket\n",
      "        \n",
      "        >>> sol = optimize.root_scalar(f, bracket=[0, 3], method='brentq')\n",
      "        >>> sol.root, sol.iterations, sol.function_calls\n",
      "        (1.0, 10, 11)\n",
      "        \n",
      "        The `newton` method takes as input a single point and uses the derivative(s)\n",
      "        \n",
      "        >>> sol = optimize.root_scalar(f, x0=0.2, fprime=fprime, method='newton')\n",
      "        >>> sol.root, sol.iterations, sol.function_calls\n",
      "        (1.0, 11, 22)\n",
      "        \n",
      "        The function can provide the value and derivative(s) in a single call.\n",
      "        \n",
      "        >>> def f_p_pp(x):\n",
      "        ...     return (x**3 - 1), 3*x**2, 6*x\n",
      "        \n",
      "        >>> sol = optimize.root_scalar(f_p_pp, x0=0.2, fprime=True, method='newton')\n",
      "        >>> sol.root, sol.iterations, sol.function_calls\n",
      "        (1.0, 11, 11)\n",
      "        \n",
      "        >>> sol = optimize.root_scalar(f_p_pp, x0=0.2, fprime=True, fprime2=True, method='halley')\n",
      "        >>> sol.root, sol.iterations, sol.function_calls\n",
      "        (1.0, 7, 8)\n",
      "    \n",
      "    rosen(x)\n",
      "        The Rosenbrock function.\n",
      "        \n",
      "        The function computed is::\n",
      "        \n",
      "            sum(100.0*(x[1:] - x[:-1]**2.0)**2.0 + (1 - x[:-1])**2.0)\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x : array_like\n",
      "            1-D array of points at which the Rosenbrock function is to be computed.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        f : float\n",
      "            The value of the Rosenbrock function.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        rosen_der, rosen_hess, rosen_hess_prod\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy.optimize import rosen\n",
      "        >>> X = 0.1 * np.arange(10)\n",
      "        >>> rosen(X)\n",
      "        76.56\n",
      "        \n",
      "        For higher-dimensional input ``rosen`` broadcasts.\n",
      "        In the following example, we use this to plot a 2D landscape.\n",
      "        Note that ``rosen_hess`` does not broadcast in this manner.\n",
      "        \n",
      "        >>> import matplotlib.pyplot as plt\n",
      "        >>> from mpl_toolkits.mplot3d import Axes3D\n",
      "        >>> x = np.linspace(-1, 1, 50)\n",
      "        >>> X, Y = np.meshgrid(x, x)\n",
      "        >>> ax = plt.subplot(111, projection='3d')\n",
      "        >>> ax.plot_surface(X, Y, rosen([X, Y]))\n",
      "        >>> plt.show()\n",
      "    \n",
      "    rosen_der(x)\n",
      "        The derivative (i.e. gradient) of the Rosenbrock function.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x : array_like\n",
      "            1-D array of points at which the derivative is to be computed.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        rosen_der : (N,) ndarray\n",
      "            The gradient of the Rosenbrock function at `x`.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        rosen, rosen_hess, rosen_hess_prod\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy.optimize import rosen_der\n",
      "        >>> X = 0.1 * np.arange(9)\n",
      "        >>> rosen_der(X)\n",
      "        array([ -2. ,  10.6,  15.6,  13.4,   6.4,  -3. , -12.4, -19.4,  62. ])\n",
      "    \n",
      "    rosen_hess(x)\n",
      "        The Hessian matrix of the Rosenbrock function.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x : array_like\n",
      "            1-D array of points at which the Hessian matrix is to be computed.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        rosen_hess : ndarray\n",
      "            The Hessian matrix of the Rosenbrock function at `x`.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        rosen, rosen_der, rosen_hess_prod\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy.optimize import rosen_hess\n",
      "        >>> X = 0.1 * np.arange(4)\n",
      "        >>> rosen_hess(X)\n",
      "        array([[-38.,   0.,   0.,   0.],\n",
      "               [  0., 134., -40.,   0.],\n",
      "               [  0., -40., 130., -80.],\n",
      "               [  0.,   0., -80., 200.]])\n",
      "    \n",
      "    rosen_hess_prod(x, p)\n",
      "        Product of the Hessian matrix of the Rosenbrock function with a vector.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        x : array_like\n",
      "            1-D array of points at which the Hessian matrix is to be computed.\n",
      "        p : array_like\n",
      "            1-D array, the vector to be multiplied by the Hessian matrix.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        rosen_hess_prod : ndarray\n",
      "            The Hessian matrix of the Rosenbrock function at `x` multiplied\n",
      "            by the vector `p`.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        rosen, rosen_der, rosen_hess\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> from scipy.optimize import rosen_hess_prod\n",
      "        >>> X = 0.1 * np.arange(9)\n",
      "        >>> p = 0.5 * np.arange(9)\n",
      "        >>> rosen_hess_prod(X, p)\n",
      "        array([  -0.,   27.,  -10.,  -95., -192., -265., -278., -195., -180.])\n",
      "    \n",
      "    shgo(func, bounds, args=(), constraints=None, n=None, iters=1, callback=None, minimizer_kwargs=None, options=None, sampling_method='simplicial')\n",
      "        Finds the global minimum of a function using SHG optimization.\n",
      "        \n",
      "        SHGO stands for \"simplicial homology global optimization\".\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        func : callable\n",
      "            The objective function to be minimized.  Must be in the form\n",
      "            ``f(x, *args)``, where ``x`` is the argument in the form of a 1-D array\n",
      "            and ``args`` is a tuple of any additional fixed parameters needed to\n",
      "            completely specify the function.\n",
      "        bounds : sequence\n",
      "            Bounds for variables.  ``(min, max)`` pairs for each element in ``x``,\n",
      "            defining the lower and upper bounds for the optimizing argument of\n",
      "            `func`. It is required to have ``len(bounds) == len(x)``.\n",
      "            ``len(bounds)`` is used to determine the number of parameters in ``x``.\n",
      "            Use ``None`` for one of min or max when there is no bound in that\n",
      "            direction. By default bounds are ``(None, None)``.\n",
      "        args : tuple, optional\n",
      "            Any additional fixed parameters needed to completely specify the\n",
      "            objective function.\n",
      "        constraints : dict or sequence of dict, optional\n",
      "            Constraints definition.\n",
      "            Function(s) ``R**n`` in the form::\n",
      "        \n",
      "                g(x) >= 0 applied as g : R^n -> R^m\n",
      "                h(x) == 0 applied as h : R^n -> R^p\n",
      "        \n",
      "            Each constraint is defined in a dictionary with fields:\n",
      "        \n",
      "                type : str\n",
      "                    Constraint type: 'eq' for equality, 'ineq' for inequality.\n",
      "                fun : callable\n",
      "                    The function defining the constraint.\n",
      "                jac : callable, optional\n",
      "                    The Jacobian of `fun` (only for SLSQP).\n",
      "                args : sequence, optional\n",
      "                    Extra arguments to be passed to the function and Jacobian.\n",
      "        \n",
      "            Equality constraint means that the constraint function result is to\n",
      "            be zero whereas inequality means that it is to be non-negative.\n",
      "            Note that COBYLA only supports inequality constraints.\n",
      "        \n",
      "            .. note::\n",
      "        \n",
      "               Only the COBYLA and SLSQP local minimize methods currently\n",
      "               support constraint arguments. If the ``constraints`` sequence\n",
      "               used in the local optimization problem is not defined in\n",
      "               ``minimizer_kwargs`` and a constrained method is used then the\n",
      "               global ``constraints`` will be used.\n",
      "               (Defining a ``constraints`` sequence in ``minimizer_kwargs``\n",
      "               means that ``constraints`` will not be added so if equality\n",
      "               constraints and so forth need to be added then the inequality\n",
      "               functions in ``constraints`` need to be added to\n",
      "               ``minimizer_kwargs`` too).\n",
      "        \n",
      "        n : int, optional\n",
      "            Number of sampling points used in the construction of the simplicial\n",
      "            complex. Note that this argument is only used for ``sobol`` and other\n",
      "            arbitrary `sampling_methods`. In case of ``sobol``, it must be a\n",
      "            power of 2: ``n=2**m``, and the argument will automatically be\n",
      "            converted to the next higher power of 2. Default is 100 for\n",
      "            ``sampling_method='simplicial'`` and 128 for\n",
      "            ``sampling_method='sobol'``.\n",
      "        iters : int, optional\n",
      "            Number of iterations used in the construction of the simplicial\n",
      "            complex. Default is 1.\n",
      "        callback : callable, optional\n",
      "            Called after each iteration, as ``callback(xk)``, where ``xk`` is the\n",
      "            current parameter vector.\n",
      "        minimizer_kwargs : dict, optional\n",
      "            Extra keyword arguments to be passed to the minimizer\n",
      "            ``scipy.optimize.minimize`` Some important options could be:\n",
      "        \n",
      "                * method : str\n",
      "                    The minimization method, the default is ``SLSQP``.\n",
      "                * args : tuple\n",
      "                    Extra arguments passed to the objective function (``func``) and\n",
      "                    its derivatives (Jacobian, Hessian).\n",
      "                * options : dict, optional\n",
      "                    Note that by default the tolerance is specified as\n",
      "                    ``{ftol: 1e-12}``\n",
      "        \n",
      "        options : dict, optional\n",
      "            A dictionary of solver options. Many of the options specified for the\n",
      "            global routine are also passed to the scipy.optimize.minimize routine.\n",
      "            The options that are also passed to the local routine are marked with\n",
      "            \"(L)\".\n",
      "        \n",
      "            Stopping criteria, the algorithm will terminate if any of the specified\n",
      "            criteria are met. However, the default algorithm does not require any to\n",
      "            be specified:\n",
      "        \n",
      "            * maxfev : int (L)\n",
      "                Maximum number of function evaluations in the feasible domain.\n",
      "                (Note only methods that support this option will terminate\n",
      "                the routine at precisely exact specified value. Otherwise the\n",
      "                criterion will only terminate during a global iteration)\n",
      "            * f_min\n",
      "                Specify the minimum objective function value, if it is known.\n",
      "            * f_tol : float\n",
      "                Precision goal for the value of f in the stopping\n",
      "                criterion. Note that the global routine will also\n",
      "                terminate if a sampling point in the global routine is\n",
      "                within this tolerance.\n",
      "            * maxiter : int\n",
      "                Maximum number of iterations to perform.\n",
      "            * maxev : int\n",
      "                Maximum number of sampling evaluations to perform (includes\n",
      "                searching in infeasible points).\n",
      "            * maxtime : float\n",
      "                Maximum processing runtime allowed\n",
      "            * minhgrd : int\n",
      "                Minimum homology group rank differential. The homology group of the\n",
      "                objective function is calculated (approximately) during every\n",
      "                iteration. The rank of this group has a one-to-one correspondence\n",
      "                with the number of locally convex subdomains in the objective\n",
      "                function (after adequate sampling points each of these subdomains\n",
      "                contain a unique global minimum). If the difference in the hgr is 0\n",
      "                between iterations for ``maxhgrd`` specified iterations the\n",
      "                algorithm will terminate.\n",
      "        \n",
      "            Objective function knowledge:\n",
      "        \n",
      "            * symmetry : bool\n",
      "                Specify True if the objective function contains symmetric variables.\n",
      "                The search space (and therefore performance) is decreased by O(n!).\n",
      "        \n",
      "            * jac : bool or callable, optional\n",
      "                Jacobian (gradient) of objective function. Only for CG, BFGS,\n",
      "                Newton-CG, L-BFGS-B, TNC, SLSQP, dogleg, trust-ncg. If ``jac`` is a\n",
      "                boolean and is True, ``fun`` is assumed to return the gradient along\n",
      "                with the objective function. If False, the gradient will be\n",
      "                estimated numerically. ``jac`` can also be a callable returning the\n",
      "                gradient of the objective. In this case, it must accept the same\n",
      "                arguments as ``fun``. (Passed to `scipy.optimize.minmize` automatically)\n",
      "        \n",
      "            * hess, hessp : callable, optional\n",
      "                Hessian (matrix of second-order derivatives) of objective function\n",
      "                or Hessian of objective function times an arbitrary vector p.\n",
      "                Only for Newton-CG, dogleg, trust-ncg. Only one of ``hessp`` or\n",
      "                ``hess`` needs to be given. If ``hess`` is provided, then\n",
      "                ``hessp`` will be ignored. If neither ``hess`` nor ``hessp`` is\n",
      "                provided, then the Hessian product will be approximated using\n",
      "                finite differences on ``jac``. ``hessp`` must compute the Hessian\n",
      "                times an arbitrary vector. (Passed to `scipy.optimize.minmize`\n",
      "                automatically)\n",
      "        \n",
      "            Algorithm settings:\n",
      "        \n",
      "            * minimize_every_iter : bool\n",
      "                If True then promising global sampling points will be passed to a\n",
      "                local minimization routine every iteration. If False then only the\n",
      "                final minimizer pool will be run. Defaults to False.\n",
      "            * local_iter : int\n",
      "                Only evaluate a few of the best minimizer pool candidates every\n",
      "                iteration. If False all potential points are passed to the local\n",
      "                minimization routine.\n",
      "            * infty_constraints: bool\n",
      "                If True then any sampling points generated which are outside will\n",
      "                the feasible domain will be saved and given an objective function\n",
      "                value of ``inf``. If False then these points will be discarded.\n",
      "                Using this functionality could lead to higher performance with\n",
      "                respect to function evaluations before the global minimum is found,\n",
      "                specifying False will use less memory at the cost of a slight\n",
      "                decrease in performance. Defaults to True.\n",
      "        \n",
      "            Feedback:\n",
      "        \n",
      "            * disp : bool (L)\n",
      "                Set to True to print convergence messages.\n",
      "        \n",
      "        sampling_method : str or function, optional\n",
      "            Current built in sampling method options are ``halton``, ``sobol`` and\n",
      "            ``simplicial``. The default ``simplicial`` provides\n",
      "            the theoretical guarantee of convergence to the global minimum in finite\n",
      "            time. ``halton`` and ``sobol`` method are faster in terms of sampling\n",
      "            point generation at the cost of the loss of\n",
      "            guaranteed convergence. It is more appropriate for most \"easier\"\n",
      "            problems where the convergence is relatively fast.\n",
      "            User defined sampling functions must accept two arguments of ``n``\n",
      "            sampling points of dimension ``dim`` per call and output an array of\n",
      "            sampling points with shape `n x dim`.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        res : OptimizeResult\n",
      "            The optimization result represented as a `OptimizeResult` object.\n",
      "            Important attributes are:\n",
      "            ``x`` the solution array corresponding to the global minimum,\n",
      "            ``fun`` the function output at the global solution,\n",
      "            ``xl`` an ordered list of local minima solutions,\n",
      "            ``funl`` the function output at the corresponding local solutions,\n",
      "            ``success`` a Boolean flag indicating if the optimizer exited\n",
      "            successfully,\n",
      "            ``message`` which describes the cause of the termination,\n",
      "            ``nfev`` the total number of objective function evaluations including\n",
      "            the sampling calls,\n",
      "            ``nlfev`` the total number of objective function evaluations\n",
      "            culminating from all local search optimizations,\n",
      "            ``nit`` number of iterations performed by the global routine.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        Global optimization using simplicial homology global optimization [1]_.\n",
      "        Appropriate for solving general purpose NLP and blackbox optimization\n",
      "        problems to global optimality (low-dimensional problems).\n",
      "        \n",
      "        In general, the optimization problems are of the form::\n",
      "        \n",
      "            minimize f(x) subject to\n",
      "        \n",
      "            g_i(x) >= 0,  i = 1,...,m\n",
      "            h_j(x)  = 0,  j = 1,...,p\n",
      "        \n",
      "        where x is a vector of one or more variables. ``f(x)`` is the objective\n",
      "        function ``R^n -> R``, ``g_i(x)`` are the inequality constraints, and\n",
      "        ``h_j(x)`` are the equality constraints.\n",
      "        \n",
      "        Optionally, the lower and upper bounds for each element in x can also be\n",
      "        specified using the `bounds` argument.\n",
      "        \n",
      "        While most of the theoretical advantages of SHGO are only proven for when\n",
      "        ``f(x)`` is a Lipschitz smooth function, the algorithm is also proven to\n",
      "        converge to the global optimum for the more general case where ``f(x)`` is\n",
      "        non-continuous, non-convex and non-smooth, if the default sampling method\n",
      "        is used [1]_.\n",
      "        \n",
      "        The local search method may be specified using the ``minimizer_kwargs``\n",
      "        parameter which is passed on to ``scipy.optimize.minimize``. By default,\n",
      "        the ``SLSQP`` method is used. In general, it is recommended to use the\n",
      "        ``SLSQP`` or ``COBYLA`` local minimization if inequality constraints\n",
      "        are defined for the problem since the other methods do not use constraints.\n",
      "        \n",
      "        The ``halton`` and ``sobol`` method points are generated using\n",
      "        `scipy.stats.qmc`. Any other QMC method could be used.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [1] Endres, SC, Sandrock, C, Focke, WW (2018) \"A simplicial homology\n",
      "               algorithm for lipschitz optimisation\", Journal of Global Optimization.\n",
      "        .. [2] Joe, SW and Kuo, FY (2008) \"Constructing Sobol' sequences with\n",
      "               better  two-dimensional projections\", SIAM J. Sci. Comput. 30,\n",
      "               2635-2654.\n",
      "        .. [3] Hoch, W and Schittkowski, K (1981) \"Test examples for nonlinear\n",
      "               programming codes\", Lecture Notes in Economics and Mathematical\n",
      "               Systems, 187. Springer-Verlag, New York.\n",
      "               http://www.ai7.uni-bayreuth.de/test_problem_coll.pdf\n",
      "        .. [4] Wales, DJ (2015) \"Perspective: Insight into reaction coordinates and\n",
      "               dynamics from the potential energy landscape\",\n",
      "               Journal of Chemical Physics, 142(13), 2015.\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        First consider the problem of minimizing the Rosenbrock function, `rosen`:\n",
      "        \n",
      "        >>> from scipy.optimize import rosen, shgo\n",
      "        >>> bounds = [(0,2), (0, 2), (0, 2), (0, 2), (0, 2)]\n",
      "        >>> result = shgo(rosen, bounds)\n",
      "        >>> result.x, result.fun\n",
      "        (array([1., 1., 1., 1., 1.]), 2.920392374190081e-18)\n",
      "        \n",
      "        Note that bounds determine the dimensionality of the objective\n",
      "        function and is therefore a required input, however you can specify\n",
      "        empty bounds using ``None`` or objects like ``np.inf`` which will be\n",
      "        converted to large float numbers.\n",
      "        \n",
      "        >>> bounds = [(None, None), ]*4\n",
      "        >>> result = shgo(rosen, bounds)\n",
      "        >>> result.x\n",
      "        array([0.99999851, 0.99999704, 0.99999411, 0.9999882 ])\n",
      "        \n",
      "        Next, we consider the Eggholder function, a problem with several local\n",
      "        minima and one global minimum. We will demonstrate the use of arguments and\n",
      "        the capabilities of `shgo`.\n",
      "        (https://en.wikipedia.org/wiki/Test_functions_for_optimization)\n",
      "        \n",
      "        >>> def eggholder(x):\n",
      "        ...     return (-(x[1] + 47.0)\n",
      "        ...             * np.sin(np.sqrt(abs(x[0]/2.0 + (x[1] + 47.0))))\n",
      "        ...             - x[0] * np.sin(np.sqrt(abs(x[0] - (x[1] + 47.0))))\n",
      "        ...             )\n",
      "        ...\n",
      "        >>> bounds = [(-512, 512), (-512, 512)]\n",
      "        \n",
      "        `shgo` has built-in low discrepancy sampling sequences. First, we will\n",
      "        input 64 initial sampling points of the *Sobol'* sequence:\n",
      "        \n",
      "        >>> result = shgo(eggholder, bounds, n=64, sampling_method='sobol')\n",
      "        >>> result.x, result.fun\n",
      "        (array([512.        , 404.23180824]), -959.6406627208397)\n",
      "        \n",
      "        `shgo` also has a return for any other local minima that was found, these\n",
      "        can be called using:\n",
      "        \n",
      "        >>> result.xl\n",
      "        array([[ 512.        ,  404.23180824],\n",
      "               [ 283.0759062 , -487.12565635],\n",
      "               [-294.66820039, -462.01964031],\n",
      "               [-105.87688911,  423.15323845],\n",
      "               [-242.97926   ,  274.38030925],\n",
      "               [-506.25823477,    6.3131022 ],\n",
      "               [-408.71980731, -156.10116949],\n",
      "               [ 150.23207937,  301.31376595],\n",
      "               [  91.00920901, -391.283763  ],\n",
      "               [ 202.89662724, -269.38043241],\n",
      "               [ 361.66623976, -106.96493868],\n",
      "               [-219.40612786, -244.06020508]])\n",
      "        \n",
      "        >>> result.funl\n",
      "        array([-959.64066272, -718.16745962, -704.80659592, -565.99778097,\n",
      "               -559.78685655, -557.36868733, -507.87385942, -493.9605115 ,\n",
      "               -426.48799655, -421.15571437, -419.31194957, -410.98477763])\n",
      "        \n",
      "        These results are useful in applications where there are many global minima\n",
      "        and the values of other global minima are desired or where the local minima\n",
      "        can provide insight into the system (for example morphologies\n",
      "        in physical chemistry [4]_).\n",
      "        \n",
      "        If we want to find a larger number of local minima, we can increase the\n",
      "        number of sampling points or the number of iterations. We'll increase the\n",
      "        number of sampling points to 64 and the number of iterations from the\n",
      "        default of 1 to 3. Using ``simplicial`` this would have given us\n",
      "        64 x 3 = 192 initial sampling points.\n",
      "        \n",
      "        >>> result_2 = shgo(eggholder, bounds, n=64, iters=3, sampling_method='sobol')\n",
      "        >>> len(result.xl), len(result_2.xl)\n",
      "        (12, 20)\n",
      "        \n",
      "        Note the difference between, e.g., ``n=192, iters=1`` and ``n=64,\n",
      "        iters=3``.\n",
      "        In the first case the promising points contained in the minimiser pool\n",
      "        are processed only once. In the latter case it is processed every 64\n",
      "        sampling points for a total of 3 times.\n",
      "        \n",
      "        To demonstrate solving problems with non-linear constraints consider the\n",
      "        following example from Hock and Schittkowski problem 73 (cattle-feed) [3]_::\n",
      "        \n",
      "            minimize: f = 24.55 * x_1 + 26.75 * x_2 + 39 * x_3 + 40.50 * x_4\n",
      "        \n",
      "            subject to: 2.3 * x_1 + 5.6 * x_2 + 11.1 * x_3 + 1.3 * x_4 - 5     >= 0,\n",
      "        \n",
      "                        12 * x_1 + 11.9 * x_2 + 41.8 * x_3 + 52.1 * x_4 - 21\n",
      "                            -1.645 * sqrt(0.28 * x_1**2 + 0.19 * x_2**2 +\n",
      "                                          20.5 * x_3**2 + 0.62 * x_4**2)       >= 0,\n",
      "        \n",
      "                        x_1 + x_2 + x_3 + x_4 - 1                              == 0,\n",
      "        \n",
      "                        1 >= x_i >= 0 for all i\n",
      "        \n",
      "        The approximate answer given in [3]_ is::\n",
      "        \n",
      "            f([0.6355216, -0.12e-11, 0.3127019, 0.05177655]) = 29.894378\n",
      "        \n",
      "        >>> def f(x):  # (cattle-feed)\n",
      "        ...     return 24.55*x[0] + 26.75*x[1] + 39*x[2] + 40.50*x[3]\n",
      "        ...\n",
      "        >>> def g1(x):\n",
      "        ...     return 2.3*x[0] + 5.6*x[1] + 11.1*x[2] + 1.3*x[3] - 5  # >=0\n",
      "        ...\n",
      "        >>> def g2(x):\n",
      "        ...     return (12*x[0] + 11.9*x[1] +41.8*x[2] + 52.1*x[3] - 21\n",
      "        ...             - 1.645 * np.sqrt(0.28*x[0]**2 + 0.19*x[1]**2\n",
      "        ...                             + 20.5*x[2]**2 + 0.62*x[3]**2)\n",
      "        ...             ) # >=0\n",
      "        ...\n",
      "        >>> def h1(x):\n",
      "        ...     return x[0] + x[1] + x[2] + x[3] - 1  # == 0\n",
      "        ...\n",
      "        >>> cons = ({'type': 'ineq', 'fun': g1},\n",
      "        ...         {'type': 'ineq', 'fun': g2},\n",
      "        ...         {'type': 'eq', 'fun': h1})\n",
      "        >>> bounds = [(0, 1.0),]*4\n",
      "        >>> res = shgo(f, bounds, iters=3, constraints=cons)\n",
      "        >>> res\n",
      "             fun: 29.894378159142136\n",
      "            funl: array([29.89437816])\n",
      "         message: 'Optimization terminated successfully.'\n",
      "            nfev: 114\n",
      "             nit: 3\n",
      "           nlfev: 35\n",
      "           nlhev: 0\n",
      "           nljev: 5\n",
      "         success: True\n",
      "               x: array([6.35521569e-01, 1.13700270e-13, 3.12701881e-01, 5.17765506e-02])\n",
      "              xl: array([[6.35521569e-01, 1.13700270e-13, 3.12701881e-01, 5.17765506e-02]])\n",
      "        \n",
      "        >>> g1(res.x), g2(res.x), h1(res.x)\n",
      "        (-5.062616992290714e-14, -2.9594104944408173e-12, 0.0)\n",
      "    \n",
      "    show_options(solver=None, method=None, disp=True)\n",
      "        Show documentation for additional options of optimization solvers.\n",
      "        \n",
      "        These are method-specific options that can be supplied through the\n",
      "        ``options`` dict.\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        solver : str\n",
      "            Type of optimization solver. One of 'minimize', 'minimize_scalar',\n",
      "            'root', 'root_scalar', 'linprog', or 'quadratic_assignment'.\n",
      "        method : str, optional\n",
      "            If not given, shows all methods of the specified solver. Otherwise,\n",
      "            show only the options for the specified method. Valid values\n",
      "            corresponds to methods' names of respective solver (e.g., 'BFGS' for\n",
      "            'minimize').\n",
      "        disp : bool, optional\n",
      "            Whether to print the result rather than returning it.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        text\n",
      "            Either None (for disp=True) or the text string (disp=False)\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        The solver-specific methods are:\n",
      "        \n",
      "        `scipy.optimize.minimize`\n",
      "        \n",
      "        - :ref:`Nelder-Mead <optimize.minimize-neldermead>`\n",
      "        - :ref:`Powell      <optimize.minimize-powell>`\n",
      "        - :ref:`CG          <optimize.minimize-cg>`\n",
      "        - :ref:`BFGS        <optimize.minimize-bfgs>`\n",
      "        - :ref:`Newton-CG   <optimize.minimize-newtoncg>`\n",
      "        - :ref:`L-BFGS-B    <optimize.minimize-lbfgsb>`\n",
      "        - :ref:`TNC         <optimize.minimize-tnc>`\n",
      "        - :ref:`COBYLA      <optimize.minimize-cobyla>`\n",
      "        - :ref:`SLSQP       <optimize.minimize-slsqp>`\n",
      "        - :ref:`dogleg      <optimize.minimize-dogleg>`\n",
      "        - :ref:`trust-ncg   <optimize.minimize-trustncg>`\n",
      "        \n",
      "        `scipy.optimize.root`\n",
      "        \n",
      "        - :ref:`hybr              <optimize.root-hybr>`\n",
      "        - :ref:`lm                <optimize.root-lm>`\n",
      "        - :ref:`broyden1          <optimize.root-broyden1>`\n",
      "        - :ref:`broyden2          <optimize.root-broyden2>`\n",
      "        - :ref:`anderson          <optimize.root-anderson>`\n",
      "        - :ref:`linearmixing      <optimize.root-linearmixing>`\n",
      "        - :ref:`diagbroyden       <optimize.root-diagbroyden>`\n",
      "        - :ref:`excitingmixing    <optimize.root-excitingmixing>`\n",
      "        - :ref:`krylov            <optimize.root-krylov>`\n",
      "        - :ref:`df-sane           <optimize.root-dfsane>`\n",
      "        \n",
      "        `scipy.optimize.minimize_scalar`\n",
      "        \n",
      "        - :ref:`brent       <optimize.minimize_scalar-brent>`\n",
      "        - :ref:`golden      <optimize.minimize_scalar-golden>`\n",
      "        - :ref:`bounded     <optimize.minimize_scalar-bounded>`\n",
      "        \n",
      "        `scipy.optimize.root_scalar`\n",
      "        \n",
      "        - :ref:`bisect  <optimize.root_scalar-bisect>`\n",
      "        - :ref:`brentq  <optimize.root_scalar-brentq>`\n",
      "        - :ref:`brenth  <optimize.root_scalar-brenth>`\n",
      "        - :ref:`ridder  <optimize.root_scalar-ridder>`\n",
      "        - :ref:`toms748 <optimize.root_scalar-toms748>`\n",
      "        - :ref:`newton  <optimize.root_scalar-newton>`\n",
      "        - :ref:`secant  <optimize.root_scalar-secant>`\n",
      "        - :ref:`halley  <optimize.root_scalar-halley>`\n",
      "        \n",
      "        `scipy.optimize.linprog`\n",
      "        \n",
      "        - :ref:`simplex           <optimize.linprog-simplex>`\n",
      "        - :ref:`interior-point    <optimize.linprog-interior-point>`\n",
      "        - :ref:`revised simplex   <optimize.linprog-revised_simplex>`\n",
      "        - :ref:`highs             <optimize.linprog-highs>`\n",
      "        - :ref:`highs-ds          <optimize.linprog-highs-ds>`\n",
      "        - :ref:`highs-ipm         <optimize.linprog-highs-ipm>`\n",
      "        \n",
      "        `scipy.optimize.quadratic_assignment`\n",
      "        \n",
      "        - :ref:`faq             <optimize.qap-faq>`\n",
      "        - :ref:`2opt            <optimize.qap-2opt>`\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        We can print documentations of a solver in stdout:\n",
      "        \n",
      "        >>> from scipy.optimize import show_options\n",
      "        >>> show_options(solver=\"minimize\")\n",
      "        ...\n",
      "        \n",
      "        Specifying a method is possible:\n",
      "        \n",
      "        >>> show_options(solver=\"minimize\", method=\"Nelder-Mead\")\n",
      "        ...\n",
      "        \n",
      "        We can also get the documentations as a string:\n",
      "        \n",
      "        >>> show_options(solver=\"minimize\", method=\"Nelder-Mead\", disp=False)\n",
      "        Minimization of scalar function of one or more variables using the ...\n",
      "    \n",
      "    toms748(f, a, b, args=(), k=1, xtol=2e-12, rtol=8.881784197001252e-16, maxiter=100, full_output=False, disp=True)\n",
      "        Find a zero using TOMS Algorithm 748 method.\n",
      "        \n",
      "        Implements the Algorithm 748 method of Alefeld, Potro and Shi to find a\n",
      "        zero of the function `f` on the interval `[a , b]`, where `f(a)` and\n",
      "        `f(b)` must have opposite signs.\n",
      "        \n",
      "        It uses a mixture of inverse cubic interpolation and\n",
      "        \"Newton-quadratic\" steps. [APS1995].\n",
      "        \n",
      "        Parameters\n",
      "        ----------\n",
      "        f : function\n",
      "            Python function returning a scalar. The function :math:`f`\n",
      "            must be continuous, and :math:`f(a)` and :math:`f(b)`\n",
      "            have opposite signs.\n",
      "        a : scalar,\n",
      "            lower boundary of the search interval\n",
      "        b : scalar,\n",
      "            upper boundary of the search interval\n",
      "        args : tuple, optional\n",
      "            containing extra arguments for the function `f`.\n",
      "            `f` is called by ``f(x, *args)``.\n",
      "        k : int, optional\n",
      "            The number of Newton quadratic steps to perform each\n",
      "            iteration. ``k>=1``.\n",
      "        xtol : scalar, optional\n",
      "            The computed root ``x0`` will satisfy ``np.allclose(x, x0,\n",
      "            atol=xtol, rtol=rtol)``, where ``x`` is the exact root. The\n",
      "            parameter must be nonnegative.\n",
      "        rtol : scalar, optional\n",
      "            The computed root ``x0`` will satisfy ``np.allclose(x, x0,\n",
      "            atol=xtol, rtol=rtol)``, where ``x`` is the exact root.\n",
      "        maxiter : int, optional\n",
      "            If convergence is not achieved in `maxiter` iterations, an error is\n",
      "            raised. Must be >= 0.\n",
      "        full_output : bool, optional\n",
      "            If `full_output` is False, the root is returned. If `full_output` is\n",
      "            True, the return value is ``(x, r)``, where `x` is the root, and `r` is\n",
      "            a `RootResults` object.\n",
      "        disp : bool, optional\n",
      "            If True, raise RuntimeError if the algorithm didn't converge.\n",
      "            Otherwise, the convergence status is recorded in the `RootResults`\n",
      "            return object.\n",
      "        \n",
      "        Returns\n",
      "        -------\n",
      "        x0 : float\n",
      "            Approximate Zero of `f`\n",
      "        r : `RootResults` (present if ``full_output = True``)\n",
      "            Object containing information about the convergence. In particular,\n",
      "            ``r.converged`` is True if the routine converged.\n",
      "        \n",
      "        See Also\n",
      "        --------\n",
      "        brentq, brenth, ridder, bisect, newton\n",
      "        fsolve : find zeroes in N dimensions.\n",
      "        \n",
      "        Notes\n",
      "        -----\n",
      "        `f` must be continuous.\n",
      "        Algorithm 748 with ``k=2`` is asymptotically the most efficient\n",
      "        algorithm known for finding roots of a four times continuously\n",
      "        differentiable function.\n",
      "        In contrast with Brent's algorithm, which may only decrease the length of\n",
      "        the enclosing bracket on the last step, Algorithm 748 decreases it each\n",
      "        iteration with the same asymptotic efficiency as it finds the root.\n",
      "        \n",
      "        For easy statement of efficiency indices, assume that `f` has 4\n",
      "        continuouous deriviatives.\n",
      "        For ``k=1``, the convergence order is at least 2.7, and with about\n",
      "        asymptotically 2 function evaluations per iteration, the efficiency\n",
      "        index is approximately 1.65.\n",
      "        For ``k=2``, the order is about 4.6 with asymptotically 3 function\n",
      "        evaluations per iteration, and the efficiency index 1.66.\n",
      "        For higher values of `k`, the efficiency index approaches\n",
      "        the kth root of ``(3k-2)``, hence ``k=1`` or ``k=2`` are\n",
      "        usually appropriate.\n",
      "        \n",
      "        References\n",
      "        ----------\n",
      "        .. [APS1995]\n",
      "           Alefeld, G. E. and Potra, F. A. and Shi, Yixun,\n",
      "           *Algorithm 748: Enclosing Zeros of Continuous Functions*,\n",
      "           ACM Trans. Math. Softw. Volume 221(1995)\n",
      "           doi = {10.1145/210089.210111}\n",
      "        \n",
      "        Examples\n",
      "        --------\n",
      "        >>> def f(x):\n",
      "        ...     return (x**3 - 1)  # only one real root at x = 1\n",
      "        \n",
      "        >>> from scipy import optimize\n",
      "        >>> root, results = optimize.toms748(f, 0, 2, full_output=True)\n",
      "        >>> root\n",
      "        1.0\n",
      "        >>> results\n",
      "              converged: True\n",
      "                   flag: 'converged'\n",
      "         function_calls: 11\n",
      "             iterations: 5\n",
      "                   root: 1.0\n",
      "\n",
      "DATA\n",
      "    __all__ = ['BFGS', 'Bounds', 'HessianUpdateStrategy', 'LbfgsInvHessPro...\n",
      "\n",
      "FILE\n",
      "    c:\\programdata\\anaconda3\\lib\\site-packages\\scipy\\optimize\\__init__.py\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(spop)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "568a1253",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scipy.integrate as si"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "335b5b34",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[3.07702305 1.46882473]\n",
      "[[ 0.03060261 -0.00028282]\n",
      " [-0.00028282  0.00031984]]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<matplotlib.legend.Legend at 0x16781078c40>"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXIAAAD4CAYAAADxeG0DAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjQuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/MnkTPAAAACXBIWXMAAAsTAAALEwEAmpwYAAA2rElEQVR4nO3deXhU1fnA8e9JWBI2UTZlk00RihAgooIgixoIqAS0CHWjUEXrggsVbbHa1p8IKLjVrVatCxGVRBJIWUVAQA2EHUEExAQKiASiBMhyfn8cJmbPzOTO3Ll33s/z8EQnM/eezPLOuee85z1Ka40QQgjnirC7AUIIIapHArkQQjicBHIhhHA4CeRCCOFwEsiFEMLhathx0saNG+s2bdrYcWohhHCsdevW/ai1blL6dlsCeZs2bUhPT7fj1EII4VhKqe/Lu12GVoQQwuEkkAshhMNJIBdCCIezZYxcCBHa8vLyyMzM5OTJk3Y3JSxFRUXRsmVLatas6dX9JZALIcrIzMykfv36tGnTBqWU3c0JK1prjhw5QmZmJm3btvXqMRLIhbBBckYW0xfuYH92Ls0bRjMpriPDu7ewu1lFTp48KUHcJkopGjVqxOHDh71+jARyIYIsOSOLR+duJjevAICs7FwenbsZIKSCuQRx+/j63MtkpxBBNn3hjqIg7pGbV8D0hTtsapFwOgnkQgTZ/uxcn24PV5GRkcTExPCb3/yGbt268dxzz1FYWFjpY/bu3csHH3wQpBaGDhlaESLImjeMJqucoN28YbQNrbFGIMb8o6Oj2bBhAwCHDh1izJgxHDt2jCeffLLCx3gC+ZgxY6p1bqeRHrkQQTYpriPRNSNL3BZdM5JJcR1talH1eMb8s7Jz0fw65p+ckWXZOZo2bcrrr7/OSy+9hNaavXv30rdvX3r06EGPHj1YvXo1AJMnT2blypXExMQwc+bMCu/nNtIjFyLIPD3VUM5a8UVlY/5W/k3t2rWjsLCQQ4cO0bRpUxYvXkxUVBTffvsto0ePJj09nalTpzJjxgxSU1MBOHHiRLn3cxsJ5ELYYHj3Fo4N3KUFc8zfs8dwXl4e99xzDxs2bCAyMpKdO3eWe39v7+d0EsiFENUSrDH/3bt3ExkZSdOmTXnyySdp1qwZGzdupLCwkKioqHIfM3PmTK/u53QyRi6EqJZgjPkfPnyYCRMmcM8996CU4tixY5x33nlERETw7rvvUlBghnbq169PTk5O0eMqup/bSI9cCFEtgRrzz83NJSYmhry8PGrUqMEtt9zCgw8+CMDdd9/NyJEj+eijjxgwYAB169YFoGvXrtSoUYNu3bpx++23V3g/t1GeMadgio2N1W6ccBDCLbZv306nTp3sbkZYK+81UEqt01rHlr6vZUMrSqlIpVSGUirVqmMKIYSompVDK/cD24EGFh4zbIR6ESUhROiypEeulGoJDAX+ZcXxwk0wFlQIIdzLqqGVWcCfgMoLIYhySRElUVxyRhZ9pi6j7eT59Jm6TL7QRZWqHciVUsOAQ1rrdVXc7w6lVLpSKt2XOrvhQIooCQ+5OhP+sKJH3ge4Tim1F0gEBiql3it9J63161rrWK11bJMmTSw4rXtUtHDCyUWUhH/k6kz4o9qBXGv9qNa6pda6DXATsExrfXO1WxZG3FZESfhPrs5+5Slj6/m3d+9eevfuDZQtV7thwwYWLFjg8zn69+9fbu2VvLw8Jk+ezAUXXECXLl3o1asXaWlp/v8xASYLgkKA24ooCf+5scStv4qXsfXwVC8sXa52w4YNpKenEx8fb8m5p0yZwoEDB9iyZQu1a9fm4MGDfP75514/Pj8/nxo1ghdeLT2T1no5sNzKY4YLNxVREv6bFNexxDZwIFdnxdWrV4+ff/6ZyZMns337dmJiYhg9ejQvv/wyubm5rFq1ikcffZRhw4Zx7733snnzZvLz83niiSe4/vrryc3NZezYsWzbto1OnTqRm1v2S/PEiRO88cYb7Nmzh9q1awPQrFkzfvvb35ZoA8DHH39Mamoqb7/9NrfffjvnnHMOGRkZxMTEkJSUxIYNG2jYsCEAHTp04IsvviAiIoIJEyawb98+AGbNmkWfPn2q9bxIj1yIEBKSV2cTJ0KpnnG1xcTArFmV3sWzRB+gbdu2JCUlFf2udLnaZs2akZ6ezksvvQTAY489xsCBA/n3v/9NdnY2vXr14qqrruK1116jTp06bNq0iU2bNtGjR48y5921axetW7emQQPfl8Ts3LmTJUuWEBkZSWFhIUlJSYwdO5Yvv/ySNm3a0KxZM8aMGcMDDzzAFVdcwb59+4iLi2P79u0+n6s4CeRChBi5OjPKG1rx1qJFi5g3bx4zZswA4OTJk+zbt48VK1Zw3333AaYuS9euXa1qLgA33ngjkZFmvmvUqFH87W9/Y+zYsSQmJjJq1CgAlixZwrZt24oec/z4cXJycqhfv77f55VALoRD2Lb6t4qecyjSWvPJJ5/QsWPZIamqdqjv0KED+/btqzC4Fn/8yZMnS/yueFGuyy+/nF27dnH48GGSk5P5y1/+AkBhYSFr1qwhOtq6eQ8pYyuEA0h+eUmly9WW/v+4uDhefPHFoo0oMjIyAOjXrx/vv/8+AFu2bGHTpk1ljl2nTh3GjRvHfffdx+nTpwE4cOAA771nsqqbNWvG9u3bi4ZOKqKUIiEhgQcffJBOnTrRqFEjAK655pqiISDA76uO4iSQu5CsDHQfyS8vqXi52pkzZzJgwAC2bdtGTEwMH374IVOmTCEvL4+uXbvSpUsXpkyZAsBdd93Fzz//TNeuXZk2bRq9evUq9/j/+Mc/aNKkCZ07d6ZLly4MHz4cz/qXqVOnMmzYMAYOHMh5551XaTtHjRrFe++9VzSsAvDCCy+Qnp5O165d6dy5M6+++mq1nw8pY+synp5b6ayHp0dcLOOuDtZ28nzK+6QqYM/UoZafT8rY2s+WMrYiNEjPzZ1k9a+ojARyl5GVge4kq39FZSSQu4z03NxpePcWPD3iYlo0jEYBLRpGB3y4zI5hV2H4+txL+qHLyMpA9wpmfnlUVBRHjhyhUaNGVabrCWtprTly5AhRUVFeP0YCucuE5MpA4TgtW7YkMzMTKTltj6ioKFq2bOn1/SVrRQghHEKyVoQQwqUkkAshhMNJIBdCCIeTQB5KMjNBJpdEfj7s3w+S/ie8JIHcTvn5sHIlTJ4MXbtCq1bQti28+67dLRPBVFAA69bBjBkwbBg0agQtWsAf/gClqusJUR5JP7TD4sXw73/DwoVw9CjUqAFXXAHTpkFqKtx6KyxfDi++CHXq2N1aEShbtsBjj8GKFXDsmLmtY0cYPRpq1TKv/8aN8Mkn0Lp1hYexrbytCBmSfhggFX64Vq6EAQPgnHMgPh6GDoVrroGzzjIPzM+HJ5+Ep56Czp3ho49Aihe5z48/Qmws/PwzjBwJ/fubf8Wr6X36qflSr1UL5swx75tSpEiaPez68qwo/VACeQBU9OF67spzGXJrPNSrB+npUNlWUosWwc03wy+/wCuvmA+0cIeCAhg82PTEV62CSy6p+L47dsCIEfDNN+aK7cEHodhKyz5Tl5W7WXOLhtF8MXlgIFof9uz88pQ88iAqrwLhqVOnaXrnWDOU8vHHlQdxML30DRvMh/y228x4qUx+ucOUKbBkCfzzn5UHcTBDLWvXmmD+8MNw003my/0MKZIWfKFYYVQCeQCU9yF6YNUH9Ny9wfSuvd0nsHlz84F/6CH4178gMdHahorgmzsXnn7afDGPG+fdY+rXN0Mr06aZTsDEiUW/kiJpwReKX54SyAOg9Ieo/3fp3LvmQ1Jih8Dtt/t2sBo14JlnoEcPmDSpRG9MOMw335irq169zESmL5Qyr//998Obb5pJUKS8rR1C8ctTAnkAFP9wtTh2iJmpz7K9WTv0Cy/4d8DISHj+ecjKMkFdOE9ODiQkQHS06VXXru3fcaZMMRPlDzwAWttS3jbcheSXp9Y66P969uyp3S5pfabu9/f/6ozzLtQ5tevqRckrqn/Q0aO1jorSes+e6h9LBE9hodYjR2odEaH1smXVP95LL2kNWicnV/9Ywi9J6zN176eX6jaPpOreTy/VSeszq35QYaHWa9ean34C0nU5MVUCeSDde695ij/5xJrj7dundXS01jfcYM3xRHDMnGneBzNmWHO8vDytO3XSukMHrU+dsuaYIvCWLjXvgzlz/D5ERYFchlYCJT3djINOnGgyDqzQqhU8+qi5NP/sM2uOKQLrxAn4xz8gLs6kDlqhRg147jnYtQteesmaY4rAe+YZaNYMrr3W8kNLIA+UGTNMiuGTT1p73IcfhvPPN18Q+fnWHltY76234MgRM7Zt5U47gwebf3/7m1lcJEJbRoZZGzJxIviw84+3JJAHwt69ZkXmHXdUnS/uq+ho8yWxaRO88Ya1xxbWys+HZ5+Fyy+HPn2sP/6zz5qVoX/9q/XHFtZ65hkTC+66KyCHl0AeCLNmQUQE3HdfYI4/ciRceaXp5f30U2DOIarvk09gzx74058Cc/zOnWHCBHjtNdi6NTDnENX33XemYzdhwq+lOCwmgdxqR4+axTs33WTGtANBKZOOePSo9UM3whpaw/TpcOGFcN11gTvPE0+YBUMPPRS4c4jqmTEDatQgbdAo+kxdRtvJ8+kzdRnJGVmWnUICudVef90s2rHog5WckVX+i9+tmxm6efll+PZbS84lLPTZZ6Y07aRJ5uosUBo3hscfN5U009ICdx7hn4MH4a232Dv0Bh5ceYis7Fw0kJWdy6NzN1sWzCWQW+n0adNTvuoqiImp9uE8xXkqfPH/+lfTO3/ttWqfS1hs2jSToXDzzYE/1x//CB06yNVZKHr+eTh9mkfaxQW0PosEcivNng0HDpjMEgtUVpwnOSOLPm9vI61dL46++i/mfbnbknMKC2zaZHrI998fkAyFMmrVgrvvhi+/NDXORWg4ftwURhs5kq9qNCr3LlbVZ5FAbhWtzVhYly6mcqEFKnqRPT3zrOxcErvFcfYvx1g27U1Lx9xENUyfbkoVT5gQvHPecgvUrGnqsIjQ8NprZsOQRx4JeH0WCeRWWbTI9IYeftiyfOGKXuRIpYp66ivbxJDZoAkj1qfZWkZTnPH99+bK7I474Oyzg3fexo1h+HCzTeCpU8E7ryjfqVMwcyYMGgSxsQGvzyKB3CozZpiys6NHW3bIil78gmJ1yQsjIvno4qvptzeDiL17LDu38NOsWeaLvFip2aAZN84sPvr00+CfW5T07rtmmHXyZICAFzeTHYKssGEDdO8OU6fCI49YeujytpSavnBHiV1hzjt+mFWvjuPd/mO4fZls3Gybo0dNyumIEfCf/wBB3hKsoADatYOLLjJj9MIeBQUmx9+zE5iFK3or2iFINl+2wrPPmhftzjstP/Tw7i3K/eAX32rqQIMmrGrfk1FbFpvVhDXkZbXFK6+Y1NNJk4CyW4J55jaAwATzyEgYO9Ys2//+e1PKQQRfSgrs3Gk2A7GyLEMlqj20opRqpZT6TCm1XSm1VSl1vxUNc4wffjA794wfDw0bBuWU5V2mRU24g+jDB+G//w1KG0QpBQUmpz8uDi6+GLBpS7CxY83Pt94K3DlE5d55x2yibVWxPC9Y0XXLBx7SWq9XStUH1imlFmutt1lw7ND37rumFxyo5fgVKNNTz+sL0/9i6q8MGxbUtoSTCodKVqyA/fvNGPkZtmwJdv75cPXVJpBPmWJ66SJ4srNhwQKT2x/E577aPXKt9QGt9foz/50DbAfCZ3uSxETo3RvatrW3HTVrmt7Y/PkmoAjLVbpAa/ZsM7w2dGjR/W3bEmz8eNi3z+z3KoJr7lyzMNDCpAdvWJq1opRqA3QHviznd3copdKVUumHDx+28rT22boVNm82dVVCwbhx5hL/7bftbokrVTRUMnP+FlMg6/rroU6dot/ZtiXYdddBo0aSU26HDz4wq2xjy8xHBpRlgVwpVQ/4BJiotT5e+vda69e11rFa69gmTZpYdVp7ffihqaNx441eP6TC2ilW6NABBgwwRbsKC607rgAqHhJpv3GNqUJZqhdm236atWvDrbdCcjK4pdPkBAcOmBo7o0cHbZLTw5L0BqVUTUwQf19rPdeKY4Y8rc2wSv/+cO65Xj0kKFkM48fD735n3lCDBllzTAGYIZGscoL5qF2rzOKfq68u87uKso4Cbtw4mDmTF8b+lZldhgY+9VGYLJXCwqAPq4A1WSsKeBPYrrV+rvpNcoiMDPj2W6aeFeN17zooWQwjRphd1mXTCcuVN1RyNnkM2rEGbrjB1DwJEcmnG7KhxUUMXjsfrbXl1fZEOWbPNsXyOnUK+qmtGFrpA9wCDFRKbTjzL96C44a0b59/g7yISBJbxXpdljIoWQxRUabuRlKSbAFmsfKGSl475yA1ck/Y0gurzPSFO5h98dVceGQf3febjkLAUx/D2e7dpmiZTe8DK7JWVmmtlda6q9Y65sy/BVY0LmQVFlJ/3lxWtulOdvSvW7lV9UEJWhbD+PFm5vz99609rmB49xZ8MXkge6YO5YvJA+m1dqHJGe7Xz+6mlbA/O5fUi/ryS80oRm1aVOJ2EQCzZ5ufNiU+SK0Vf6xdy7nZh0jpVPbDW9kHJWhZDF26mEUpc8NjusI2x46ZnOHf/jbk8rWbN4zml9p1mH/RFQz9ZiW18vOKbhcW09pkq1xxBbRubUsTJJD7IzGRUzVqsviCy8r8qrIPSlCzGEaMgJUr4dAh648tjKQkU+UuxIZV4NdOw4KOfah/Opfe328ITupjONq8GbZts/V9IEU5fFVQAHPmcKTfVRTUqw/FJi+9+aAELYshIcHsGDNvnhlqEdZLTDQLwXr1srslZXjeY7PmQ868aEZ8/zXDH/uDZK0EwuzZ5orMhzRkq0mP3Feffw4HD9J8wlh7coS91bWrqYQnwyuBcfiwWTl5001Bzxn21vDuLVj+l8HUT7iO6/Z+zfCu3qXJCh9obQL51VeDjetjpEfuq8REqFsXhg5leJ06oRO4S1PK9MpfeMGM5Z51lt0tcpePPjJXZyE4rFLGiBFm8drq1dC3r92tcZc1a0ylyb//3dZmSI/cF6dPl7sUO2SNGAF5eWZCTlhr9mxTc7pLF7tbUrUhQ8xqz6Qku1viPh98YFJ+hw+3tRkSyH2xZIlZih0qtVWqctllZtWpDK9Y64cfYNUqW5Zi+6V+fbjqKhPIbdhIxrXy881qzmHDzHNsIwnkvkhMNDXHLdpcOeAiIkxPIS0NciV/2DIffmh+OuULHcww2969sHGj3S1xj6VLzVzJmDFFNwW0llIlJJB7KzfXFCEaMcJcpjpFQoLZtWbxYrtb4h6zZ5vqdh062N0S7113nflil+EV63z4ITRoYIauqKLMcYBJIPdWWhrk5DirFwamqFfDhvIBtsru3bB+vfPeB02amAUrMsxmjYICs6Xb0KFmjBybdoQ6QwK5t+bONTWeBwywuyW+qVXLjOHNm2cmPkX1pKSYnzZPbvklIQG2bIFdu+xuifN99ZWpZXTttUU32bIj1BkSyL2Rn28yP4YOdebGxiNGmEnaFSvsbonzpaaaXerbt7e7Jb7zfPnI1Vn1paSYRUCDBxfdZNuOUEgg986aNXD0qHP3woyLg+ho+QBX1/HjZkFYsV6Yh12TXD5p0wa6d5f3gRVSUkxO/tlnF91k245QSCD3Tmqq6YnHxdndEv/UqWN6DklJsnNQdSxcaIanSgVyOye5fJaQYDomBw7Y3RLn2rvXDFGVeh/YtiMUEsi9k5ICV15pZqidKiHBbMr89dd2t8S5UlNND+zyy0vcbOckl88SEszPTz+1tx1O5pknKefKrHSZ42Ct/JZAXpXvvoPt28t90Rxl2DBzVSFZC0V8Gg4pKDDzJPHxZeZJ7Jzk8tlvfmPSJmV4xX8pKdCxI1xwgd0tKSKBvCqpqeanU8fHPc4+22TczJ0rq/vwYzhk7doyWQoedk5y+cxTg2fZMsjOtrs1znP8OCxfHnIdOwnkVXFylkJpI0aY1LOtW+1uie18Hg6pZJ7EzkkuvyQkmEys+fPtbonzLFpU7jyJ3SSQV6aSLAVHuv560yOT4RXfh0M8WQoNG5b5lZ2TXH659FKzPZ0Mr/guJcVc3fbubXdLSnBgUnQQeb59nT6s4nHeeeZDnJICjz9ud2ts1bxhNFnlBO1yh0P27DFXMePGVXi8oG0YYoWICPOl/p//mNIT0SE4BBSKKpknsZv0yCvjyVIIsW/fahk6FNLT4eBBu1tiK5+GQ9wyT3JGckYWD548H06c4KG7Z4VmmmQoqmSexG4SyCtSUGDGEIcMCblv32oZOtT8/O9/7W2HzXwaDgnBLAV/eSZ55zfqSG6N2ly8aXXo5ryHmpQUEwuKreYMFS6KUBbz1FJwSS+sSEyMGWJZsABuu83u1tjKq+EQT5bC/fcHpU2BVjTJW6MWX5zflQG703nidD7TF+5wztCQXVJSoF+/kNxtS3rkFUlNLVNLwRWUMlcZCxeazAVRucWLXTVPUnwyd3n7Szg/+3+0+ykrNHPeQ8nu3bBtW8i+DySQVyQlxZT9LFZLwek8C2Du+qkZHDvGynfm2d2kkFR8odCC/3ud0w3Ogj597G6WJYpP5i5v1xOAAbvTQzPnPZRUspozFEggL8/338PmzSH7ovmj+AKYVW1iyIuI5Jt/J8rYaCnFnydVWMCl29eysFV3kje7Y3K4+CRv5lnN2NmoNYP2rAvdnPdQkZJi1pOE6GYiEsjL41koEaKXUf4ovgAmp3Zd0lt25opvvw7NeiA2Kv48xRzYSaPc4yxud4lrnqfSk7zpnS/jsh+2MPyC0Bv3DRnHjoX8ehKZ7CxPSor55r3wQrtbYpnSY6CftYvlseVvoffts6lFoan48zRo11fkqwiWt+tJjovGkEtM8n5WBwbOMftPXn+9vQ0LVZ75pBAO5NIjL+3nn00dimuvdcYO6V4qPQa6rP0lAFz/v03OqKUdJMWfp0G7vuLrVr/heFQ9944h9+ljdoBfsMDuloSulBQ455wyVS9DiQTy0pYsgdOnXTWsAmUXwOxq1Iqss5pybdYG59TSDgLP89Ty2EEu+vF7lra/JLTrplRXrVpw9dUmkEsxtbIKCsx+vSG4mrM4CeSlpaaauuN9+9rdEkuVWQBzdh1OXTOYNhvWUJB7ssR9Q7aWdhB4nqfh+zcCsLV7v9Cum2KF+HjIzDSbJVQgbK/a0tPhyBHzHIWw0P2KsYPWpmcSFwc1a9rdGsuVWQCTegI++g+XZG7lizYxJe4bznnFw7u3gILvoF07Zs+41VVDbOUaMsT8XLAALr64zK89mTyeSWDPVRvg7i84MM9JRARcc43dLamU9MiL27jRbIHleWO73YABnKpRk4Hfld01yLVjwt44edLMkwwZ4v4gDtC8uVnxW8E4uaN2QLJaWhr06gWNGtndkkpJIC8uLc38dNtqzorUrUv2Jb0ZuDu9xM2uHhP2xooVpipguHyhgxk6+OKLcjebcNQOSFY6fNgMrTjgfSCBvLi0NLPL+Hnn2d2SoGk2eiRtf8qiV/5PzqilHQxpaVC7ttlRKVzEx5uJvcWLy/zKUTsgWWnhQjPcGuLj4yCB/FfZ2bB6tSO+fS115u+dc3520DeMDVkLFkD//lCnjt0tCZ5LLzXlKMoZXnHcDkhWWbAAmjaFHj3sbkmVJJB7LF5seiThFsg9C58kj9jYvRt27gy/94FnG7u0NCgsLPErx+2AZIWCAtMjj4szk50hTrJWPNLSzDZel11md0uCLz4eXnkFTpwIr15oeTzzJOEWyMG8DxITISMDevYs8StH7YBkha+/hp9+csz7wJKvGqXUYKXUDqXULqXUZCuOGVRam40WrrkmpJP+AyY+Hk6dgs8+s7sl9ktLMxttu2ATCZ/FxZksHbk6M+8DB6QdelQ7kCulIoGXgSFAZ2C0UqpzdY8bVOGWdlhav35Qt67sqh5uaYelNW0Kl1wigRzMc3DppSGfduhhRY+8F7BLa71ba30aSAScVX3H88YNl7TD0mrXhquukmXan38efmmHpcXHw5dfmt2xwtWhQ45JO/SwIpC3AH4o9v+ZZ24rQSl1h1IqXSmVfvjwYQtOayFP2uG559rdEvvEx5s67Nu3290S+3jSDvv3t7sl9omP/3WoMVwtXGh+hlkgL+8atEy3Tmv9utY6Vmsd26RJEwtOa5HsbFizBuLjw7eeBJRcph2u0tJM7ng4T/j27GmGWDyTvuEoLc0xaYceVgTyTKBVsf9vCey34LgBUzxg/+WPM6GggBXtY8O7CmCrVqbOhgsDuVdf0N99F55ph6UkbzzAgpYxHJ2bQt//Wxw+738PT9rh4MGOSDv0sKKlXwMXKKXaKqVqATcBIbsZZPGtvDTQbdNqjkXV48G9tcO3noRHfDysWmV2jneJ0q93hV/Q4Zx2eIbnuZrfsjtnn8yhybaN4dWZAfjqK0elHXpUO5BrrfOBe4CFwHZgjtZ6a3WPGyjFCwApXciVe9axsk13fjxVWO79XV9PorghQ8yO8UuX2t0Sy3hd8Cmc0w7P8DxXK9t2J19FMGB3evh1ZhyWduhhybWD1nqB1vpCrXV7rfVTVhwzUIoH5s6H9tD0l6NFu4mXx/X1JIrr3dvUYnfR8IpXBZ9OnjQ59A7rhVnN85wcj6rHuhadGHCmmFpYdWbS0syiwHPOsbslPnHOIJBFigfmK3evA+Dztj1pGF0zPOtJFFezpumJuCgN0auCT560QwcURwqk4s/J8vaxdDn4HU1zjoRPZ+bgQcelHXqEXSAvXgBowHfpbG7Wnp/PbswT1/0m/OpJlCc+Hvbvh02b7G6JJbwq+LRgAURFhXfaISWfq8/axQJw9b6M8OnMlJN26JRMtrBbj+4JzK8mp9Nj/ze82390iYAddoG7NM+iqAULoFs3e9tiAc/rOX3hDvZn59K8YTST4jqWfJ3T0kwQjw6TnmcFij9XO3QbDjVozN25O2kRLp8JT9ph9+6As3ZGUtqGS+jY2Fidnp5e9R0Dac4cGDXKZGn06WNvW0JNz55Qpw7JLyRWHgDd4LvvTAXI55+H++6zuzWh5Y47TBGtI0dcufVhCQUFJogPGwbvvANAn6nLyCpnfqBFw2i+mDww2C0EQCm1TmsdW/r2sBtaKZKWZuovX3qp3S0JPfHx6NWrefr91e7Pq/ekHYb5+Hi5hg6FnByzc5DbrV1r0g6LvQ+ctDNSeAbywkKfqh06ZZzMMvHxqMJCLvm25FWTK1PRUlNNymGHDna3JPQMGmR64uFQTC0l5dea7Gc4aWek8Azk69bB//5nehxV8HpBiZv06sXRqPpF6WfFhWJvxG85OSbt8Npr7W5JaKpXD6680lXpqBVKTYW+fc2eBGc4aWek8AzkKSkm6d+Ly+mw3EE8MpKvO/biyt3rULrkQqlQ7I34bckSOH1aAnll4uNh2zbYu9fulgTOnj2wdWuZ94GTdkYKu6wVwATy3r29qjXspHEyK51z4/U0/stSLv7fLjaddyEQur0Rv6WkwFlnyWR3ZeLj4cEHzVzCXXfZ3ZrASE01P4cNK/Mrp+yM5JgeuWXj1D/8ABs2eN0Lc9I4mZVi7xyNVorrsjaGfG/EL4WFZux3yBD3Z2RUx4UXQrt27h5eSU2Fjh0dXZ7BET1yS/M5Pd++XgbySXEdS5wbXNgzLU/jxqhLL2V8znbGT616LsFxvv7abCAgwyqVU8r0yt9805QyiIqyu0XWysmB5cv5dtRYbp+6zLGpto7okVs6Tp2SYoojXXSRV3d30jiZ5eLjfw14bpOSApGR4bsrlC+GDjUlDD7/3O6WWG/xYjh9mr/rdo5OaHBEILdsnPqXX8yejNde69OejMO7t+CLyQPZM3UoX0weGB5BHH7dLcazdNlNUlPN2LjDiiPZ4sorzapXNw6vpKZyPKoeq88teYXttIQGRwRyy8apFy82u8XL5bR3uneHZs3c9wHet89suC3vA+9ER8PAge57H5yZJ1netgf5kWVHmZ2U0OCIQG5ZPmdKiinT2revha1zsYgIMxm4cCHk59vdGutUkqUgKhAfD7t2mV2U3OLMsOH6rleU+2snJTQ4IpBbMk7tyVIYPFiyFHwxdCgcPWpq0rhFaqpZydnR5RPWVvJUBHTTKs/UVIiI4JIJox2z8KcijshaAQvyOdPTTb1huZz2zeDBZmf5pCR3lHn1zJPcfbdP8yRhr21b6NIFPv0UHnjA7tZY48w8ydD+Xcg762xHF4hzTCCvNs9qTgcWjbdVvXqmJk1yMsya5fzg55knkWEV3yUkwFNPweHD0KSJ3a2pHs96kmnTAOcs/KmII4ZWLJGSYrIUvFjNKUpJSDAThOvXF93k2EJiqalmNafMk/huxAgzRDkvZPdW955niMglX+jhEcglS6F6rr3WXM0kJQEOLiRWWGgCucyT+KdbN2jTpuh94GipqWbFqpfrSUKdqwO5p9c4ZfxUAJa072VzixyqcWPo16/oA+zYQmLr1pl5Epf0woJOKdMrX7wYjh+3uzX+O3ECli71eT1JKHNtIC/eaxy06yv2nH0e964/Efq9xlCVkGCq4O3Y4dxCYjJPUn0JCaZipGdDDidautSUG3DRF7prA7mn11jndC6X79vI0va9yM0vDP1eY6gaPtz8TEpybiExmSepvssvN4vE5s61uyX+S02F+vXNVaZLuDaQe3qHffdmULsgn6UdLi1xu/BR69YQGwtJSY4quF8kM9NkKbioF2aLyEi4/nqzyvPkSbtb4zutTSCPi4NatexujWVcG8g9vcNBu77ieO26fN2yc4nbhR8SEuCrrxjeFOcVEvOx6qWoxIgR8PPPZmMOp1m/Hvbvd90XumsD+aS4jtSpoRjwXTrL2/UkP7JG6PcaQ11CgvmZnOy8QmLz5vlU9VJUYsAAk8LpsOGV5Iws3n1oBnkRkQzZ1cBV82WuDeTDu7fg1XanaHIim6Udejmj1xjqOnUyy9qdln525IjJtLjhBtdkKdiqVi3To503zzE1eJIzsnj0k00MyFjGyjbd2Z5Xyxkps15ybSAH6LduCURH8/z7jzuj1+gECQmwfDn89JPdLfHexx+bgHPTTXa3xD0SEswX5MqVdrfEK9MX7uCi77fR8vghUjuZxWCOSJn1knsDeV4efPSRGROtX9/u1rhHQgIUFPw65uwEs2ebIZVu3exuiXsMHmx2C3LI1dn+7Fyu3b6CU5E1WXzBZSVudwP3BvKlS01NiDFj7G6Ju8TGQosWjvkAk5UFK1bA6NEyrGKlunVN5kdSkskECXEtG9Ri6I5VfNY+lpzadYtud0vyg3sD+ezZ0LChbOVltYgIk1O+cKGpJBjq5swxgUaGVaw3YgRkZjLunldCvubO002yafbzT6Rc9GvuuJuSH9wZyHNzzYz6yJGmBKuwVkKCeY6dsAXc7NnQo4fZDV5Yan7rHuSrCHqu/yzka+5csW4p+dF1+KZnX+ekzPrAnWVs5883ea6jR9vdEnfq18/sdZmUZHploWrXLrMLzPTpdrfElf5v7SEatO5K3M41TOt3GyhVNIEYUgEyLw8+/pgaw69n6eND7W5NQLizR/7BB3Duue7YCCEU1axpJpFTU82HJFR9+KH5+dvf2tsOl9qfncvCCy+n/U9ZdDjyQ4nbQ8rSpSbDxsXDa+4L5MeOmeXDo0aZ5cQiMBISIDvbfEhC1ezZcMUVpryAsFzzhtEsOpMBErdzTYnbQ0piolnAFBdnd0sCxn2BPCnJ7AAjwyqBFRdnhlfeesvulpRv82bYulXeBwE0Ka4jOec0ZV3zi7hu++egdehNIJ48+esQoIvny9wXyD/4wBSM7yW1xwMqKgpuvtlsAXfkiN2tKSsx0VyR3XCD3S1xLc+m6EsuHULHH/dxTc7e0JtA/O9/Te10Fw+rgNsC+cGD5lJfcoaDY9w4U5v6vffsbklJWptAPmgQNG1qd2tcbXj3FjzyzpNQty6vn94QWkEczPugcWMYONDulgRUtQK5Umq6UuobpdQmpVSSUqqhRe3yz5w5ZjsvWQQUHF27wiWXwL/+FVqLQr76CnbvlmGVYKlf3/R4ExMhJ8fu1vzql19MDfobb4Qa7kzQ86huj3wx0EVr3RXYCTxa/SZVw+zZJrh07mxrM8LK+PGwZYtJ8wsVs2ebwk6ezTBE4I0fbwJnYqLdLflVSorZ1s3lwypQzUCutV6ktfaUP1sLtKx+k/y0Zw+sWSO98WC76SaoU8f0ykNBQYG5MouPNyt7RXBceil06QJvvGF3S36VmAjNm5vMJZezcoz890CFG/kppe5QSqUrpdIPHz5s4WnP8PQEwuDbN6Q0aGDytGfPNouw7LZiBRw4IMMqwaaU6ZV//TVs3Gh3a0xqbFqaSUOOcNdUYHmq/AuVUkuUUlvK+Xd9sfv8GcgH3q/oOFrr17XWsVrr2CZNmljT+uI++MDsx3j++dYfW1Ru3DgTxD/6yO6WmC+UunVdtwOMI9xyi0nxC4Wrs7lzzUR8mHTslK7mJJVS6jZgAjBIa33Cm8fExsbq9PT0ap23hC1b4OKL4aWX4I9/tO64wjtam00nGjeGVavsa8cvv0DLlmZY5f0K+xQikH73O7Mgb/9+iLZpYZDWpr7O6dMmNrgog00ptU5rHVv69upmrQwGHgGu8zaIB8TLL5vJrRtvtK0JYU0p0yv/4gv45htLDpmckUWfqct8q6r35pvmkvqeeyxpg/DD+PHmNfjkE/vasGyZ2Wj7oYdcFcQrU60euVJqF1Ab8KwIWau1nlDV4yztkR88aIZTbr0VXn/dmmMK3x08aHrDEydWu0hVckYWj87dTG5eQdFt0TUjK19skp8PHTqYNth5VRDutDaVJps3h88/t6cNQ4ZARgZ8/73rVnMGpEeute6gtW6ltY4586/KIG655583l1APPxz0U4timjUzhbT+8x/zelTD9IU7SgRx8GJbro8+Mh/cP/2pWucW1eSZ9FyxAnaUfb38utLyxZYtZjXnffe5LohXxtnTucePwz//aeqOS71p+40fD4cOVXsbuIqq51VYVU9rmDbNbOcmk5z2u+02swDnzTdL3Oy50srKzg1c/fJnnzXpsBOC36e0k7MD+WuvwbFjjGs2MOR3KAkLcXFmG7hSH2BfVVQ9r8KqekuWmDHRhx8Oi1SzkHfuuebq7O23S1yd+XWl5Yv9+80k97hxpqBbGHHuu/7UKXKnPcuaNjEsrdc65HcoCQuRkTB2rLm0zcz0+zCT4joSXbNkCeJKq+pNn26Cx803+31OYbHx482eufPmFd3k85WWr1580SwImzjRmuM5iHMD+bvvEv3jQV7uNbLEzZZ+wwvf/f73pt7NK6/4fQhPVb0WDaOr3pYrIwMWLzYf3jAaEw15cXHQqlWJlZ4+X2n5IifHvOdGjjTVT8OMMyvJFBTA9OlsbtaeVW1iyvw65HYoCSPJ2bWo020gfac9y5iCi7ltVF+/KuIN797Cu8dNnw716sGdd/rRWhEwkZFwxx0wZQqsXg29ezMprmO52UiW1C9/802zqcxDD1X/WA7kzB55cjLs3MmcgWPKzRMNuR1KwoRnMuuJ3jej0Nya8lpgh7r27jV1Ve68U+qqhKKJE00a4n33QWFhlVdafme05OfDzJnQt6+p+RKGnNcj1xqeeQbat6fng+P5+NNtgfmGFz7zTGblNmjK65ckcN+aD3mn57VMX1grMHWqZ840X+RhOCbqCPXqmWyim282E5+//32FV1ql1w545ruAqt87H38M+/aZMfIw5bwe+WefmcI8kyYxPLa192OpIuCKD2m9etkNHKx3Do8vfZ39RwOw6PfIEVPTY8wYswhIhKYxY6B3b3j0UZMuXAG/M1q0NsNrF14Y1qmnzuuRP/OMWXxy222AD2OpIuCaN4wm60wwP1Ermun9bmXGglnc+v1awOIP2T//aWpNy0Kw0KaUWbTXqxf8/e8Vrvr1O6Pl889h/XqTihzGqafO+sszMmDRInMpHRVld2tEKaXTBj/pMpCt53bgkc/fglwLJ6BzcsxldHy8KZYmQltsrElLff552Lmz3Lv4ldGSlwd//jM0aWIqL4YxZwXy554z20qF2aotpyg9mdX87Lr89Pep1PnffrPizgpamxTHI0dMRoRwhqeeMp2vBx8s99c+rx0AE8RXrzZzJXZVWgwVWuug/+vZs6f2y48/ar1okX+PFfYZOVLrunW1zsqq/rGmTdMatJ4+vfrHEsE1fbp57RYsKPfXSeszde+nl+o2j6Tq3k8v1UnrMys+VnKyOdaECQFqbGgC0nU5MbXa9cj9YXk9chHadu829crHjIG33vL/OMuWwdVXm0UfH34YNiVKXeP0abMdXEQEbN4MNWv6d5zdu0298Q4dTKXLMBpmDUj1QyG80q6dmdd4+21Yt86/Y+zbZ7btuugis/hDgnhIKzcnvFYtMwyyY4fZBMYfJ0+afQeUYtETL9Jn1mqps4QFOwT5Q3rkYej4cdODOuss+PRT6NzZ+8eePGkWe+zYYVJPO8o6gVBWZT35+HizCcnSpWYi1BcTJsBrr7F25r8Ze+S8cs8BJp1xf3YuzRtGMymuo2sy26RHLuzVoAEkJZmMk169zCIOvFzNd++9kJ5uap1LEA95VeaEv/ii+ULv3dv8dxWdSc975IFrH4bXXmPn7Xfz0Mnzyz3HE/O2Br5UbgiSQC6Cp08fM7Ry8cVw443svP1u/vLxhso/dG+8YRb+PPYYDB9uV8uFD6rMCW/f3pQdHjzYLN+/4QazPVw5PL37Ot9+w1MLX+LLVl0Y0Ty+aL1Cadm5eYEtlRuiJJCL4GrRwiziuOsuLnznFV59/8+cc+JY0a9z8wp4fe5XZhf0e+81+29ecw387W82Nlr4wquc8HPOMUNszz5rSt326GGGzYrLyWHtc2/y+LyZJM5+lF9qRXPPdX/i5wKI9HGOxO2F9Jy3slM4UnJGVslxyz/8mS/21uIfi/5JytsTmXXFaDof2sNl+zbT6fBe86A6dUw51LfeMtX0hCN4XeVQKZNX3ru3mcju08fkm9esCfPnw4oVTD19muO16rCibQ9euewGDtczG0YUaE10zcgy54iqGcHRE3ll2uT2Qnoy2SkCrqLJr6iaEbTYvZ1Xk56i5fHD5NaozboWF7G1Y0/ufPz3cMklJtNBOE6ZL+6qJhx/+gluvx1SUsz/d+oEQ4dyz7Hz+O9Z7cmPLNnnbHHmmKXPAfi+cbeDVDTZKYFcBFyfqcvKHdNsGF2TU/mFRP6cQ7ufMvmmSVsio6Nc86ETPtLaDLu1bl20OUSVGTDl8PlLxEEqCuQytCICrqLxyWO5ecwcFcP0hTvYXLuO6z50wkdKQf/+JW7yvBd8CczhWEhPArkIuOJVEUvfHo4fOuEbeY9UTbJWRMD5VRBJCOE16ZGLgPPn8lgI4T0J5CIo5PJYiMCRoRUhhHA46ZEL13Bz2pkQlZFALlyhWruwC+FwMrQiXMHvXdiFcAEJ5MIV/N6FXQgXkEAuXMGvXdiFcAkJ5MIVZNGRCGcy2SlcQRYdiXAmgVy4hiw6EuFKArlwHMkXF6IkCeTCUSRfXIiyZLJTOIrkiwtRliWBXCn1sFJKK6UaW3E8ISoi+eJClFXtQK6UagVcDeyrfnOEqJzkiwtRlhU98pnAn4Dgb/4pwo7kiwtRVrUmO5VS1wFZWuuNSimLmiRExSRfXIiyqgzkSqklwLnl/OrPwGPANd6cSCl1B3AHQOvWrX1oohAlSb64ECUprf0bEVFKXQwsBU6cuaklsB/opbX+X2WPjY2N1enp6X6dVwghwpVSap3WOrb07X4PrWitNwNNi51gLxCrtf7R32MKIYTwneSRCyGEw1m2slNr3caqYwkhhPCe9MiFEMLh/J7srNZJlToMfB/0E1dfYyCc5gDC7e8F+ZvDhVP/5vO11k1K32hLIHcqpVR6eTPGbhVufy/I3xwu3PY3y9CKEEI4nARyIYRwOAnkvnnd7gYEWbj9vSB/c7hw1d8sY+RCCOFw0iMXQgiHk0AuhBAOJ4HcD+G0I5JSarpS6hul1CalVJJSqqHdbQoUpdRgpdQOpdQupdRku9sTaEqpVkqpz5RS25VSW5VS99vdpmBQSkUqpTKUUql2t8UqEsh9FIY7Ii0GumituwI7gUdtbk9AKKUigZeBIUBnYLRSqrO9rQq4fOAhrXUn4DLgj2HwNwPcD2y3uxFWkkDuu7DaEUlrvUhrnX/mf9diyhW7US9gl9Z6t9b6NJAIXG9zmwJKa31Aa73+zH/nYIKbqwu9K6VaAkOBf9ndFitJIPdB8R2R7G6LTX4PpNndiABpAfxQ7P8zcXlQK04p1QboDnxpc1MCbRamI1ZoczssZVn1Q7ewakckJ6nsb9Zaf3rmPn/GXIq/H8y2BVF5exWGxVWXUqoe8AkwUWt93O72BIpSahhwSGu9TinV3+bmWEoCeSla66vKu/3MjkhtAc/+pC2B9UqpKndECnUV/c0eSqnbgGHAIO3ehQeZQKti/+/Z8crVlFI1MUH8fa31XLvbE2B9gOuUUvFAFNBAKfWe1vpmm9tVbbIgyE/hsiOSUmow8Bxwpdb6sN3tCRSlVA3MZO4gIAv4Ghijtd5qa8MCSJkeyTvAT1rriTY3J6jO9Mgf1loPs7kplpAxclGVl4D6wGKl1Aal1Kt2NygQzkzo3gMsxEz6zXFzED+jD3ALMPDMa7vhTG9VOIz0yIUQwuGkRy6EEA4ngVwIIRxOArkQQjicBHIhhHA4CeRCCOFwEsiFEMLhJJALIYTD/T/p7o8yk/8W8gAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "import numpy as np\n",
    "from scipy import optimize\n",
    "from matplotlib import pyplot as plt\n",
    "\n",
    "x_data = np.linspace(-5, 5, num=50)\n",
    "y_data = 2.9 * np.sin(1.5 * x_data) + np.random.normal(size=50)\n",
    "\n",
    "plt.plot(x_data,y_data,'o',label='Data')\n",
    "\n",
    "def test_func(x, a, b):\n",
    "    return a * np.sin(b * x)\n",
    "\n",
    "params, params_covariance = optimize.curve_fit(test_func, x_data, y_data, p0=[2, 2])\n",
    "print(params)\n",
    "print(params_covariance)\n",
    "\n",
    "plt.plot(x_data,test_func(x_data, *params),'-r',label='Fitted Curve')\n",
    "plt.legend()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "4f3c351e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[3.02434955 1.50937374]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<matplotlib.legend.Legend at 0x167fd60a520>"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXIAAAD4CAYAAADxeG0DAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjQuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/MnkTPAAAACXBIWXMAAAsTAAALEwEAmpwYAAAxL0lEQVR4nO2de3xU1bXHf5vwCI8oCAF5B3kHDAFmxvCKQriIlatERbQ+8F29vbVqtfVWrb21t7c+Wr0+rhSrvVJojVVBJNDwEESSMJmBBAzhYSoIQYQAhUAIBMi+f6wcCTBJ5nHO2WefWd/PJ58xmNl7TU7O7+y91tprCSklGIZhGH1podoAhmEYJjZYyBmGYTSHhZxhGEZzWMgZhmE0h4WcYRhGc1qqmLRLly4yJSVFxdQMwzDasn79+gNSyuTz/12JkKekpCAYDKqYmmEYRluEEF+H+nd2rTAMw2gOCznDMIzmsJAzDMNoDgs5wzCM5rCQMwzDaA4LOcMwjObELORCiEQhRJEQYqMQYrMQ4j/NMIxhGIYJDzNW5CcBTJJSjgCQDmCqECLDhHHjj4ICYO5c4MQJ1ZYwDKMRMR8IklTQ/Fj9t63qv7jIeaSsXg1ccw2J+BNPAD/8IfBv/wZ06aLaMsZKjhwBtm8/+7VtG70ePQr86U/A+PGqLWQ0wBQfuRAiQQhRAmA/gOVSSn+In3lACBEUQgQrKyvNmNY9rFsHTJsGXHYZ8PHHgMcDPPss0Ls38OCDdHMz7uOtt4BOnQCfD7j9duDXvwYCAaBbN6C2FrjlFuDAAdVWMhogzOwQJIToCGABgB9JKUsb+zmPxyP5iH49JSXAxIlA587AmjVAjx7072VlwMsvA3/+M3DyJN3U8+YBCQlKzWVMYs8eYMgQYORI4LHHgEGDgP79gTZt6P8XFwMZGUBWFrB4MdCC8xIYQAixXkrpOf/fTf3rkFIeBrAawFQzx3UtW7YAU6YASUnAypVnRRwAUlNpxfb113Sjv/ceiTrjDh59FDh9Gvi//wOmT6frbYg4QAL/8svA0qXASy+pspIxmx07LBnWjKyV5PqVOIQQbQFMBrA11nFdz1dfAZMn00prxQqgb9/QP9etG93IHg/wy1/S6pzRm7//Hfjb34CnnyZ3WmM89BBw003Az39OgXBGb4qK6HovWGD60GasyLsDWCWE2AQgAPKRLzZhXPdSUUFb5hMnSMQHDWr654UAfvMbWp3PmWOPjYw11NRQIHvwYODxx5v+WSGAP/6RHvK33AIcPGiPjYw1vPgi0LEjLeBMJmYhl1JuklKOlFKmSSmHSyl/ZYZhrkVK4F//FTh0CFi2DBg+PLz3TZ4MXHUVBcSqqy01kbGQ3/yGdmNvvnmuK6UxLr4YyMkBvv0WuPtu+vth9KO8HPjwQ9plJSWZPjxHUOxm1SoKcP7P/wCjR4f/PmNVvn8/vZfRj61bgeefpwyViRPDf5/HQ+61Tz4hvzmjH7/7HdCqFfDww5YMb2rWSrjEddbKzTdTYLOiAmjbNvL3X3cdZbfs2EGpa4weSEnutOJiEvRu3SJ//403kpgXFpK4M3qwfz+5x+64I2bXqC1ZK0wz7N1LgY677opOxAFyrVRVAS+8YKppjMXMn0+7sf/+78hFHKAd2dtvk6uFr71evP46JSn85CeWTcFCbifvvEMpZz/4QfRjpKUBt95K7pVvvzXPNsY6/vlPuomvuAJ44IHox+nUCfj+94FFi2hMxvlUVwNvvAFcfz0FuC2ChdwuzpyhbVVWVvNZKs3xn/8JnDpFq3PG+Tz3HJ3QnD079oM9s2bR6u79982xjbGWd96hxIYnnrB0GhZyu1iyBNi1i6LWsTJgAHDvvfRgsOiAAWMSp0/TQa6bbgLS02Mfb9QoOjz07ruxj8VYy+nTwO9/D4wbB4wda+lULOR2MXs20L07BSvN4Jln6Lj+L39pzniMNXz6Ka3Gb7nFnPGEoFV5YSEV12KcywcfADt3Aj/9qeVTsZDbwY4ddNT6vvsoBckMevYE/v3fabXHN7RzycmhvOFrrjFvzNtvJxcNl2xwLlJSUHrIECqIZzEs5Hbw1lu0krr/fnPHffRR+oNhf6kzqa2lLKXrrwcSE80bt0cPOiA2dy5QV2feuIx5rFxJqaaPP25LwTMWcqupraW0sWnTqCytmfToQRXyFi40d1zGHFasoOySmTPNH3vWLIq5fPaZ+WMzsfPii8Cll9LuyQZYyK1mwQI6EGBGkDMU2dnA+vXA7t3WjM9ET04O1daYMsX8sadPJ5fN3Lnmj83ExsaNVH7jxz8OrwyDCbCQW82bbwL9+llzMwN0QwO8KncaJ07QNcnOBlq3Nn/8du3olPAHH3DtHacxdy5d8wcftG1KFnIr2bKFtr4/+IF1frJBg4ChQ1nInUZeHp3Avflm6+a4807g2DHgo4+sm4OJnNxcqqXTsaNtU7KQW8ns2ZSlcvfd1s6TnU0PDC5z6hxycqjrU1aWdXOMH0+7Pc4pdw7l5dSa8dprbZ2WhdwqamroBrvpJqBrV2vnmj6dTo7m5lo7DxMeNTV0jP6GG8xLNw1Fixa0Kv/0U46ROAXjHmQhdwmffUYd0u+4w/q5PB7KK7eg8wgTBUuWkN/aimyV87njDkpBnTfP+rmY5snNpdzxpjo/WQALuVUsW0YR6yuvtH4uIWhVnpcHHD9u/XxM0+Tk0C7Mjmvfvz+5WN59l5tOqOboUVrA2XAA6HxYyK0iLw/IzKTsAjvIzqYt/bJl9szHhObYMep6f9NNQMuW9sw5axb5ZQMBe+ZjQrNiBZ0bsdmtArCQW8Pu3UBZGXD11fbNmZlJZU45e0UtixfTA9UOt4rBjBl0cpRzytWSm0v14seNs31qFnIrMFbFdgp5q1a0pfvkE6q6xqjh/ffpxO348fbNefHFdE5h6VL75mTORUqKjUyZYm2AuxFYyK1g2TK6mYcNs3fe6dOp9vHnn9s7L0NUVdHNPGOGLfU1ziEri5o679xp77wMUVxMHcAUuFUAFnLzOXMGWL6cnsxC2Dv31VfTFpuzV9SwaBE1fbDTrWIweTK9rlxp/9wMudSEMLfKZQSwkJtNMEiFkux0qxi0b08PkIULOYNBBTk5VBjtiivsn3voUKp3z0KuhtxcwOez/sxII8Qs5EKI3kKIVUKILUKIzUKIH5thmLbk5dGT+V/+Rc382dkUbC0uVjN/vHLyJIlodrb9bhWA/uYmTSIb+CFuL/v3U8aQIrcKYM6K/DSAn0gphwLIAPBDIUSqCePqybJldECnc2c180+bRkLC7hV7KSqibJVJk9TZkJVFolJaqs6GeGTpUnp46izkUsq9UsoN9f99FMAWAD1jHVdLjhwB1q2zrtJhOHTpQqmInIZoL6tX06o4M1OdDUZdF3av2MvixZTcMHKkMhNM3QMKIVIAjATgN3NcbVi5koKdKvzjDZk+nVZl5eVq7YgnVq+m5sqdOqmzoU8faszNQm4fp07RLvx737M/uaEBpgm5EKIDgA8BPCKlrArx/x8QQgSFEMHKykqzpnUWeXlU7D8jQ60dRo3yRYuUmhE3nDwJFBQAV12l2hLKXvnsMxIYxnrWrqW0U4VuFcAkIRdCtAKJ+HwpZcjiyFLKOVJKj5TSk5ycbMa0zkJKEvKsLCUHAs6hb19ama1Zo9aOeKGoiBpJOEHIs7Ko5gcf17eH3FxqImGkfyrCjKwVAeBtAFuklL+P3SRN+fJL4Ouv1btVDDIz6WAQN+e1HsM/PmGCakuooYEQ7F6xi8WLqThahw5KzTBjRT4OwB0AJgkhSuq/vmfCuHqRl0evKgOdDcnMpFOeW7aotsT9OME/btC5M9nCQm49//gHFStTUO3wfMzIWlkrpRRSyjQpZXr91xIzjNOKvDxyZ9hch7hRjNUhu1esxUn+cYOsLKCwkEsaW42iJhKh4JOdZnDyJLBqlXPcKgC1AOvZk4Xcavx+5/jHDbKyqJzq2rWqLXE3y5cDAwdSTXjFsJCbQUEBrX6cJOSGz/bzz/mkn5U4yT9uMGECBdxXrFBtiXupqwPy89WeG2gAC7kZ5OVREwEnrcoA+iPbswfYsUO1Je7FSf5xg/btgTFj2E9uJVu2UE0lBbXHQ8FCbgZ5eXRBk5JUW3IuxiqRy9paw4kT5It22gMcIPdKcTEFvBnzyc+nVzvrzjcBC3ms7NsHlJQ4y61ikJoKXHIJ+8mtwsgfnzhRtSUXkpVFLrVVq1Rb4k7y84HkZEpwcAAs5LFi3ChOSTtsSIsWtGLgFbk1ONE/buDzUW4zu1esYe1aurcUHstvCAt5rBQUkE9yxAjVloQmM5MOK+3dq9oS97F6NRVK6thRtSUX0qoVXXsOeJrP3r3Ujckh/nGAhTx2CgqokYBdHdMjxYiq86rcXJzsHzfIyqKH+O7dqi1xFw7zjwMs5LFRXU3+8TFjVFvSOCNH0o6BhdxcnJg/fj7c/s0a8vOppaLCsrXnw0IeC8Egla0dO1a1JY3TsiXZxwFPc3Gyf9xg+HAKyLGQm8vatbQLb91atSXfwUIeCwUF9Kq6bG1zTJgAfPEF5b0y5uBk/7hBixbUsejTT/lQmFlUV1Nap4P84wALeWwUFABDhlCKn5PJzKQb2fDtMbGhg3/cYNw44JtvgIoK1Za4g6Ii2oWzkLsEKelmdrJbxcDnoywG9pObg99P9XV0EHIjfrNunVo73MLateRSc1hcjIU8Wr78Ejh40HEXNCRt25KYs5/cHFavJreFk/3jBmlpFJhjITeH/Hxg2DBnlWQAC3n0FBbSqw4rcoDcK8EglzY1Ax384watWwOjR5/9e2Wi58wZcqc6KO3QgIU8WgoK6EYeMkS1JeExYQJw+jSvzGJFJ/+4QUYGsGEDuYOY6CktpTZ6DvOPAyzk0VNQQDdIC01+hWPHkq3sXomN4mISRAeuyholI4Ns3rhRtSV6Y9R3d+C110SFHMaRI8Dmzfq4VQDg4oupjAAHPGPD76fXK65Qa0ckGOmxvBuLjfx8oEcPam7uMFjIo8Hvp6wVHQKdDcnMJLdAba1qS/TF7wd69wa6d1dtSfj06kXdoljIY8NhhbIawkIeDYWF5Kbw+VRbEhkTJgA1NeQvZaLD79fvugO0Kmchj57du+nLgf5xgIU8OgoKgMsvBy66SLUlkcENmWOjspK6LenkVjEYM4Zs37dPtSV64sBCWQ1hIY+Uujpa2ejmVgGArl2BwYO5KW+06OgfNzD85MZnYCJj7VoqPpeWptqSkLCQR0pZGVBVpVegsyEZGXTMmGtvRI7fDyQkUF62bowaRQXU2L0SHfn5dO84tFw1C3mkGIWydFyRA4DXS9trrr0ROX4/VRRs3161JZHTti01ieaDQZFTVQVs2uRYtwpgkpALId4RQuwXQpSaMZ6jKSyk0qD9+6u2JDqMQF1RkVo7dKOujn5nOrpVDDIygECADoYx4bNuHV1/hwY6AfNW5P8HYKpJYzmbggJyqzgwBSks0tLo2DYLeWRs307nB3QX8upqOgPBhM/atZSl5uBy1aYIuZRyDYBDZozlaA4coBtaV7cKALRpQ1tsFvLI0DnQacAHg6IjP58O0yUlqbakUWzzkQshHhBCBIUQwcrKSrumNRfjBtA10Gng853tbsSER1ER3ci61NYJxWWXAV26sJBHwpkzdO0dvnizTcillHOklB4ppSc5Odmuac2loICi1h6Paktiw+cDjh0Dtm5VbYk++P0UKE5IUG1J9Bh1tFnIw2fbNrpXHH4IjLNWIqGwkMqXtm2r2pLY4IBnZNTUUMEpnd0qBhkZ9ADntn/hEQjQq9er1o5mYCEPl1OnSPh0d6sAwMCBVESLhTw8iosp08Phq7KwMPzkfO3Dw3CpDR6s2pImMSv98K8ACgEMFkJUCCHuNWNcR7FpEzVlcLivLCxatKAVBt/M4eGGQKeB10suFnavhEcgQAfAHO5SMytr5VYpZXcpZSspZS8p5dtmjOsodOsI1Bw+Hz2campUW+J8dKx42BhJSXSoiQ8GNU9tLbnUHO5WAdi1Ej5+P93IvXurtsQcfD5yF5SUqLbE+fj97liNG2Rk0Geqq1NtibPZtInEnIXcRQQCWlzQsOGAZ3js3w/s3Ok+IT98mM5EMI1j3BsaxEZYyMOhqorSkNwk5N27U8MBFvKmcZN/3IAPBoVHIEDlOPr0UW1Js7CQh8P69fTqJiEHaKXBQt40Olc8bIwhQyhriYW8aYxduAblOFjIw8HIJdX9IND5+HxAeTlwyP3VFaLG76cmIu3aqbbEPFq0oB0GC3njHD1KJas1WbyxkIdDIAD06wd07qzaEnMxfH/Gg4o5FzdUPGyMjAzgiy/o1CJzIRs2UM1+DfzjAAt5eASD2jyZI2L0aNo2snslNNu3U3zEjULu89GDivu3hkaTE50GLOTNUVlJWQuaXNCIuOgiYOhQFvLGcGOg08BwEwaDau1wKoEA0LcvBTs1gIW8OYw/dLf5xw2MgCe3frsQv1+L49lR0a0bnYlgt1poNEs3ZiFvjmCQ3A9uylpoiM9HudK7dqm2xHm4oeJhU3g8vCIPxYEDwI4dLOSuIhCgdC0HF5WPCT4YFJqaGjrZ50a3ioHXS1lLXAnxXIxdiiaBToCFvGmkpIvqVrcKQKl1bdqwkJ/Phg1UwsDNQm78XRvnJBgiENBuF85C3hR79gDffqvVFitiWremGuss5Oei4aosYjjgGRoNd+Es5E1h/IG7WciBs63fuLv6WYJBoGdPd1Q8bIxOnYD+/Tng2RBjF67ZPc9C3hSBALV2GzFCtSXW4vVSrfUtW1Rb4hyCQXe71Ay8Xl6RN2T3bmDfPu12YizkTREIUO1m3Vu7NQcHPM/lyBEqkhYPQu7xUMbS/v2qLXEGmh0EMmAhbwwp3Xui83wGDAA6dmQhNzBOO8bDtTc+I6/KiUAAaNVKu104C3ljfPUVpWXFw81stH4zTjLGO4aoaZS1EDUjR1KGBgs5EQgAaWmUyaURLVUb4FjcWvGwMbxe4PnnKX86BlfSqVOnUFFRgRMnTphonM1ccQWwfDmVZ6isVG1NRCQmJqJXr15o1apVeG9ISqIMDQ54Uu2ZYBC47TbVlkQMC3ljBAJAYiL5yOMBrxc4c4Z6FBqNB6KgoqICSUlJSElJgdCgjnNITp+msrX9+6u2JCKklDh48CAqKirQr1+/8N/o9QLLlpE7UddrZgZGkTQNd+HsWmmMYBBITyd/WTxg7DxiXJmdOHECnTt31lvET54E2rdXbUnECCHQuXPnyHdDHg+dl9izxxrDdEHTQCfAQh6aM2fotFu8uFUAypm+9FJTfKXaijgAVFfTq6aNJKL63XPAkwgE6AE+dKhqSyKGhTwUW7fSDa3hkzlqhKAHV7z7So8fp1dNhTwqRoyg8xLxfu0DAWDUKC2LpLGQh0LjLVZMeL30EDt6VLUlMZGQkID09PTvvnbu3ImxY8cCAHbu3Im//OUv3/1sSUkJlixZcvbN1dUUG2nZdPjoqquuQjDECvbUqVN48sknMXDgQAwfPhw+nw9Lly4154NZRdu2FAuK5xX5qVNAcbF2B4EMTBFyIcRUIcQ2IUS5EOJJM8ZUSjAIdOjgzjrUTeHxUMCruFi1JTHRtm1blJSUfPeVkpKCgoICAGEKeQyr8WeeeQZ79+5FaWkpSktL8cknn+BoBA/G06rKJBglbeO1Lv0XX1BsRFN3asxCLoRIAPAGgGsApAK4VQiRGuu4SgkEKIe4RZxtWEwKeDqRDh06AACefPJJfP7550hPT8fzzz+PX/ziF8jJyUF6ejpy5s9HdVUV7nnqKXi9XowcORIff/wxAKCmpga33HIL0tLSMHPmTNTU1Fwwx/Hjx/HWW2/htddeQ5v6PORu3brh5ptvPscGAPjggw9w1113AQDuuusuPPbYY5g4cSKeeOIJpKSk4PDhw9/97IABA7Bv3z5UVlbixhtvhNfrhdfrRX5+vnm/II+HmnDv2GHemDqheV0lM9IPfQDKpZRfAYAQ4j0A1wMoM2Fs+6mtBUpKgIcfVm2J/XTtCvTpY94W+5FH6HdpJunpwCuvNPkjNTU1SE9PBwD069cPCxYs+O7//fa3v8VLL72ExYsXAyChDQaDeP3114HDh/HzRx/FpEmT8M5f/4rDhw/D5/Nh8uTJ+MMf/oB27dph06ZN2LRpE0aNGnXBvOXl5ejTpw8uuuiiiD/W9u3bsWLFCiQkJKCurg4LFizA3XffDb/fj5SUFHTr1g3f//738eijj2L8+PHYtWsXrr76amwxqz5Ow4DnZZeZM6ZOBALAJZdo+9nNEPKeAHY3+L4CwAVFnIUQDwB4AAD69OljwrQWUVpKYq7pkzlmXBDwNFwrEVNdjWV+PxYFAnjpjTcAUDrlrl27sGbNGjxc/3BPS0tDWlqaiRYDM2bMQEJ9kG3mzJn41a9+hbvvvhvvvfceZs6cCQBYsWIFysrOro+qqqpw9OhRJJlRbnX4cCppHAgA9TuIuMLoO6BpxpUZQh7qk1/gaJNSzgEwBwA8Ho9zHXHxdqLzfLxe4KOPqDxBp06xjdXMytlxHD8OKQQ+/PBDDA4RH2kutW/AgAHYtWtXo+La8P3n53q3b5C3PmbMGJSXl6OyshILFy7E008/DQCoq6tDYWEh2lpRxK11a9rtxGPAs6aGFnDTpqm2JGrMcAJXAOjd4PteAL4xYVw1BAJA585AJCfj3ITLmw0kJSWdE3z87nspgepqXH3llXjttdcg64N+xfWB38zMTMyfPx8AUFpaik2bNl0wdrt27XDvvffi4YcfRm1tLQBg7969mDdvHgBy42zZsuU710ljCCGQnZ2Nxx57DEOHDkXnzp0BAFOmTCEXUD1R7TqawuOh8xN1deaO63RKSujsiMaLNzOEPABgoBCinxCiNYBbACwyYVw1aL7FihmjUJRLhTwtLQ0tW7bEiBEj8PLLL2PixIkoKyujYOeSJXjmZz/DqVOnkJaWhuHDh+OZZ54BADz00EM4duwY0tLS8MILL8DXSJrar3/9ayQnJyM1NRXDhw/H9OnTkZycDID889OmTcOkSZPQvZmGFTNnzsS8efO+c6sAwKuvvopgMIi0tDSkpqZi9uzZJv1W6vF6KfV0+3Zzx3U6bkg3llLG/AXgewC2A/gHgKea+/nRo0dLR1JdLWVCgpRPP63aErUMGCDlDTdE9daysjKTjbGJQ4ekDASkPHZMtSUxE/U1+OILKQEp//xncw1yOrffLmX37qqtCAsAQRlCU03Jr5NSLpFSDpJS9pdS/pcZYyqhuJi2WDo/mc3ABQHPiKmupl2Y25uINMWQIZRDH2/X3gV9B+IsUboZjMYKml/UmPF6z7a8iheOHycRj7ezAw1p2ZKOqLvUrRaSqipXdIOK47/aEAQCQK9e7m64Gw4xBjylbqcD6wOdOlY8PJ+Yf/ceD+1M46UR9/r1dP01X7yxkDdEw+7ZljBqVNRdYxITE3Hw4EG9xPzkSXKpaV4oS9bXI09MTIx+EK+X0vHK9DzPFzEuSTfmxhIGhw4B5eXAPfeotkQ9HTpQKc8ofKW9evVCRUUFKnXqrFNdDRw4QLXnDxxQbU1MGB2CosZYyBQVUcsztxMMUqpxly6qLYkJFnIDY/WpafUz0/F6gb//PeKuMa1atYqsO40TePxx4I03yF8aL41EGsNoxB0IAPfdp9oa6wkEXHHPs2vFwFh9xkPD3XDweCjYWVGh2hLribduUE0hBD3EjcC/m6msBHbu1N6tArCQnyUQAAYNotUI4/oTnt8Rj92gmsPno7KuRpMNt6J5xcOGsJAbcKDzXIyuMW4X8u3bgWPHWMgb4vPRA07zuvTNEgzSDsQFu3AWcoCazn7zjSt8ZaZhdI1x++EQF63KTKNhwNPNBAJ0CMqM6pGKYSEH3FFrwQq8Xvd3jSkqovzxeOsG1RTduwO9e7tbyKU8W1fJBbCQA3RBW7akgBdzFo+Hytl+9ZVqS6zD76cHloYNdy3F53O3kO/ZA3z7rWsWbyzkAAn58OHxXWcjFA27xriREyeohOkVF/RBYXw+eoBrnlffKC5zqbGQG1ss9o9fyLBhQJs27vWTl5RQ93QW8gsx7ge3XntjFz5ihGpLTIGFvLwcOHzYNU9mU2ndmv7Q3boi9/vplYX8QkaPpowOt7pXXLYLZyHnQGfTeL3u7Rrj91ORtB49VFviPJKSgNRUdwq5lK4oXdsQFvKiInoqDxum2hJn4vVSnrVZ3dqdhN/Pq/GmMAKebsta+uorCuKzkLuIQICq/bXksjMhycigV8MN4RYqK+mGZiFvHJ+Pgp07d6q2xFxcuAuPbyE/fZpOr7nogprOwIFAp07AunWqLTEXw2XAQe7GMX43bnOvBAJAYqKrduHxLeSbN1PtZRbyxmnRglathYWqLTEXv58+mwuOZ1vG5ZdT1pLbhNyFRdLiW8h5VRYeGRn00KuqUm2JeRQVUdZChw6qLXEurVqR29FNQm4USXPZ4i2+hTwQILdB//6qLXE2Y8aczbd3A1KSOLF/vHl8PhI+t7R+27qVGomwkLsIo9ZCBI0T4hJjx+IWP/mXX1LWAgt58/h85H7cvFm1JeZgBO1ZyF1CTQ3VXHbZBbWEjh2p9ZtbhJwPAoWP2wKeBQXAJZe4rkha/Ap5cTH5y9g/Hh4ZGSTkbsgp9vvP9iVlmqZ/f3I/ukXICwvJVeiyXXhMQi6EmCGE2CyEqBNC6FUP0oW5pJaSkUE5xW6ohMgVD8NHCPdUQvznP4GyMhJylxHrirwUwA0A1phgi70EAnQ0m49nh4dxMEh398qJE8DGjexWiQSfDygtpSChzhh/u2PHqrXDAmIScinlFinlNrOMsRVu7RYZw4aRO0L3fPLiYq54GCk+H9Xa2bBBtSWxUVhIZwdceN/b5iMXQjwghAgKIYKVlZV2TRuaAweoVyPfzOGTkEA3tO4rcg50Ro5bWr8VFFA1TxeeHWhWyIUQK4QQpSG+ro9kIinlHCmlR0rpSU5Ojt5iMygooNfx49XaoRsZGeSW0Lm7ut9Pbcy6d1dtiT506wb07au3kJ85Q9fehW4VAGi2UpSUcrIdhthKfj6dWnNJvz7byMiggyEbNuj7EOSKh9Ghe8CztJSqeLow0AnEa/rh2rUk4i4pKm8bhgDq6l6prAR27OCU02jw+agK4v79qi2JDmMX7tIVeazph9lCiAoAYwDkCiHyzDHLQk6coKI548aptkQ/unYFLrtMXyFn/3j06N76rbCQXEQpKaotsYRYs1YWSCl7SSnbSCm7SSmvNsswy1i/Hqit1dc1oBrjYJCO+P0UtOWKh5EzahRlfOhal76ggFbjLjsIZBB/rpW1a+nVpVssy8nIAPbsASoqVFsSOX4/VTxs3161JfrRoQNlfBj3j07s3w/84x+u9Y8D8Sjk+flUZ0F15oyuGDeDbvnkdXVc8TBWMjPputfWqrYkMoy/VRcv3uJLyOvqSMjZPx49aWnUXUU398r27cCRIyzksTBhAsWY1q9XbUlkFBRQlpqLXWrxJeTbtgGHDrGQx0Lr1nRD6CbkHOiMnQkT6HWNZhU5CgvJx5+YqNoSy4gvIc/Pp1cOdMZGRsbZoLEuFBUBSUnAkCGqLdGXrl3p9/f556otCZ/aWsq0cbFbBYg3IV+7lnzjAweqtkRvMjKAkyfplKcu5OdTCh1XPIyNCRPoPjpzRrUl4bFxI7mDXBzoBOJNyA3/uEtTkGxDt0qIhw4BmzYBV12l2hL9ycykWMMXX6i2JDyMg0As5C5h3z6gvJz942bQqxfQs6c+Qr5mDTXEYCGPHcNProt7pbCQauv06qXaEkuJHyFn/7i5jBmjj5CvXk3lGFxYvtR2+vYF+vTRJ+BpHARyOfEl5ImJFL1mYicjg7oF6VB7Y/Vq2om1aaPaEncwYQKtyJ3e9q+iAti92/VuFSCehHztWlqRtW6t2hJ3YPjJDR+kUzl4kAJe7FYxj8xMclV++aVqS5omDg4CGcSHkB8/rnfpVSfi9QLt2gGffqrakqYxXAAs5Oahi5+8oIBcaunpqi2xnPgQ8qIiqqPNgU7zaN2abugVK1Rb0jTsHzefIUOALl2c7ycvLKRy1a1aqbbEcuJDyI1AZxxssWwlKwvYsgX45hvVljSO4R9nl5p5CEHuFScLeU0N7cLj5J6PHyEfNgzo1Em1Je4iK4tenepeOXiQ88etYsIEajSxe7dqS0Kzfj012Y6DQCcQD0JeV0e+MvaPm096OnDJJcDKlaotCQ37x60jM5NeneonN3bhLOQuYfNmOonG/nHzadECmDSJhNyJqWjsH7eOESOodo1ThXz5ctqFd+2q2hJbcL+QG4XwWcitISuLttfl5aotuRD2j1tHQgL9bp3oJz9+nB4wVzu/YZlZuF/I8/OB7t2Bfv1UW+JODD+507JX2D9uPRMmAGVlwIEDqi05l88+o6qHLOQuggtlWcuAAVTLwml+8s8+o1cWcusw/OROa/+Wl0enuI189zjA3UJeUUGRdQ50WocQtCpftYoCy05h9Wo6sMT+cevweqnsgdPcK3l59JBp21a1JbbhbiH/+9/p1dj+M9YweTKVii0pUW3JWdg/bj1t2lDHJScFPHftArZujSu3CuB2Ic/NpUptw4aptsTdTJpEr05xrxw4QPWy2a1iPZmZdPDm6FHVlhDLltErC3n4CCFeFEJsFUJsEkIsEEJ0NMmu2Dl5klKQrr2W/eNW0707kJrqHCHn/HH7mDCBXGpGgSrV5OVRrfzUVNWW2EqsK/LlAIZLKdMAbAfwH7GbZBKffQZUVwPTpqm2JD7IyiIBPXlStSVn/eMej2pL3M+YMZSK6AQ/+enTlD119dVxt3iLScillMuklKfrv10HwDltOHJzKdgxcaJqS+KDrCyqb+GEZhPsH7ePpCSq8e+EMg2BAHD4MDBlimpLbMdMH/k9AJaaOF70SAksXky+2ziKXCvlyivppKdq9wr7x+3n2mvpAb5vn1o7li2jlfjkyWrtUECzQi6EWCGEKA3xdX2Dn3kKwGkA85sY5wEhRFAIEaysrDTH+sbYto2611x7rbXzMGfp2JFcGaqFnP3j9pOdTYunRYvU2pGXRymRnTurtUMBzQq5lHKylHJ4iK+PAUAIMQvANAC3Sdl4wQ0p5RwppUdK6UlOTjbvE4QiN5deWcjtZfJkqv2uMoOB/eP2c/nldHJ6wQJ1Nvzzn4DfH3fZKgaxZq1MBfAzANdJKY+bY5IJ5ObSH1efPqotiS+ysijgpDLwtWoV+8ftRghala9cCVRVqbFh5UrKnmEhj4rXASQBWC6EKBFCzDbBptg4coQOKPBq3H7GjqWj0arqrpSXA6WlcRnsUs706VTfZKmiMFleHnDRRYDPp2Z+xbSM5c1SygFmGWIay5bRqpCF3H4SE2k1rMpP/v779HrzzWrmj2fGjgWSk4GFC4GZM+2dW0q677Oy4qKtWyjcd7IzN5eaHRhd3hl7ycqirJH9++2f+/33Ka+ZXWr2k5AAXHcd3X92nyXYto2O5sepWwVwm5DX1QFLlgBTpwItY9psMNGiqv3btm3Axo32rwaZs2RnU6B71Sp7583Lo1cWcpcQCACVlexWUcno0ZT+ZXcGQ04OBd1mzLB3XuYsWVlAhw72X/u8PGDQICAlxd55HYS7hDw3lw6lTJ2q2pL4JSEBuPVW4OOP6ZSdXeTkUN2PHj3sm5M5l8RE4Jpr6NrbVdL4xAlKOY3zALf7hHzsWPKRM+q4807ykxrBR6spLaVONRzkVM/06XTC065SDfn5VBoijt0qgJuE/JtvqJwmu1XU4/EAQ4cCc+faM19ODu3EbrrJnvmYxrn2WsocWbjQnvny8mi+OD/J6x4hX7KEXlnI1SMEMGsWrZasbsosJQn5VVcB3bpZOxfTPBdfTIXqFiyga2MlZ87QtZ84kXzzcYx7hDw3l3pHDh+u2hIGAG6/nQTd6lV5SQnw5ZecreIkpk+nB3hZmbXzLF1KaYf332/tPBrgDiE3mkhMmxZ3dYgdS8+eVHtl7lxrA185ORRgveEG6+ZgIuP6+np6VmevzJ4NXHrp2fniGHcI+erV1ESC3SrOYtYs4OuvrevpKCUFVCdPBrp0sWYOJnJ69KBenlb6yXfuJHfqfffF7WnOhrhDyF9/nXKXjd6RjDPIzqbGA+++a834wSCwYwe7VZxIdjawfj25PqxgzhzafbNbBYAbhLysjJpI/OhH3ETCabRrRwd0/vY32jGZTU4OrcamTzd/bCY2jGvy8cfmj11bC7z9Nu3AuRwDADcI+UsvkYD/8IeqLWFCceedwLFj5m+z6+rIrTJlCtCpk7ljM7EzeDCloFrhJ1+wgGr5PPSQ+WNrit5CvmcPMG8ecO+97CN1KhMm0NFps90r69YBu3ezW8XJzJhB8avSUnPHnT2b/qbi/BBQQ/QW8ldfpVzSxx5TbQnTGC1a0Kp8xQqgosK8cXNygDZtOGPByTz8MMVInnnGvDG3bKGHww9+QH9bDACdhfzIEXoyz5hBbaYY53LnnZRhMr/Rlq6RceYM+d2vuYaaCTDOpHNn4PHHya1WVGTOmH/4A8VF7rnHnPFcgr5CPmcOtZV64gnVljDN0b8/NZx4911zTvvNnQvs3QvcdlvsYzHW8sgj1HDi5z+Pfazjx+lv6MYbga5dYx/PRegp5LW1wCuvULrh6NGqrWHCYdYs2hYHg7GNc/AgPbzHjeNDQDqQlEQivnJl7J2jcnKooiYHOS9ATyH/y1+oSNZPf6raEiZcbr6Zypz+8Y+xjfOzn9HN/Oab7CPVhQcfpPIZTz0V247szTeB1FQKoDPnoN+dUFcHvPgikJYW9zWIteLii8mv+dZb0Tdnzs+n/OHHHgMuv9xc+xjrSEwEnn0W8PuBRYuiG2P9emoc8+CDXIYjBEJaXaEsBB6PRwaj3WLn5lJNlT//mQozMfpw/Dh1Oa+spGJX3buH/95Tp4BRoyguUlYGtG9vmZmMBZw+DQwbBrRuTdc+ISGy999/P+3E9+wBOna0wkItEEKsl1J6zv93/VbkL7xA2zTOH9aPdu3oEM+xYxSoPHMm/Pe+8grlI7/6Kou4jrRsCTz3HF3Dv/41svcuXAj86U+0cItjEW8KvYTc7wfWrKGtNRfK0ZPUVOCNN6hB73PPhfeeXbuAX/6SurRz3ri+3HQTkJ5Obpba2vDek5dHizavl05xMyHRS8hfe42eyPfdp9oSJhbuuotyy3/1K+DTT5v/+YcfptdXX7XULMZiWrQAfvMb4KuvgHfeaf7n16yh4lupqVTpMCnJehs1JSYhF0I8J4TYJIQoEUIsE0JY2/n2f/+XgiVx3g3EFbzxBtXjuO026vHYGIsWUeGlZ58F+va1zz7GGqZOBcaPp4d4U4eEioooFpaSAixbxvV0miHWFfmLUso0KWU6gMUAfhG7SU1w0UWceuQWOnQgf/nhw+T7PN9fXlVFOec/+hEFyR59VImZjMkIAfz+9xT4vuIKup8XLjz3+m/aRIKfnEwNY5KTlZmrCy1jebOUsqrBt+0B2J8Cw+jL5ZeTu+z++8nV0r49sG0bsH078O239DMtW5I/nWMi7sHrpYJnb79NQezsbGDAAHpYjxlDIt6+PR0g6tlTtbVaEHP6oRDivwDcCeAIgIlSyspGfu4BAA8AQJ8+fUZ//fXXMc3LuAQpKebxzjtUwXLwYGDQIPoaPJiCY1xLx72cPg189BHwu9+ddbV07Ur+8cGD1drmQBpLP2xWyIUQKwBcGuJ/PSWl/LjBz/0HgEQp5bPNGRNTHjnjTo4d49hHPCMlHfiaP596C3AT9ZBELeQRTNAXQK6UstkrwELOMAwTOZYcCBJCDGzw7XUAtsYyHsMwDBM5MQU7AfxWCDEYQB2ArwE8GLtJDMMwTCTEmrVyo1mGMAzDMNGh18lOhmEY5gJYyBmGYTSHhZxhGEZzWMgZhmE0h4WcYRhGc5R0CBJCVILSFXWjC4ADqo2wkXj7vAB/5nhB18/cV0p5QRUxJUKuK0KIYKhTVW4l3j4vwJ85XnDbZ2bXCsMwjOawkDMMw2gOC3lkzFFtgM3E2+cF+DPHC676zOwjZxiG0RxekTMMw2gOCznDMIzmsJBHgRDicSGEFEJ0UW2L1QghXhRCbBVCbBJCLBBCdFRtk1UIIaYKIbYJIcqFEE+qtsdqhBC9hRCrhBBbhBCbhRA/Vm2THQghEoQQxUKIxaptMQsW8ggRQvQG8C8Adqm2xSaWAxgupUwDsB3Afyi2xxKEEAkA3gBwDYBUALcKIVLVWmU5pwH8REo5FEAGgB/GwWcGgB8D2KLaCDNhIY+clwH8FEBcRImllMuklKfrv10HoJdKeyzEB6BcSvmVlLIWwHsArldsk6VIKfdKKTfU//dRkLi5um29EKIXgGsB/FG1LWbCQh4BQojrAOyRUm5UbYsi7gGwVLURFtETwO4G31fA5aLWECFECoCRAPyKTbGaV0ALsTrFdphKrK3eXIcQYgWAS0P8r6cA/BzAFHstsp6mPrOU8uP6n3kKtBWfb6dtNiJC/Ftc7LqEEB0AfAjgESlllWp7rEIIMQ3AfinleiHEVYrNMRUW8vOQUk4O9e9CiMsB9AOwUQgBkIthgxDCJ6X81kYTTaexz2wghJgFYBqALOnegwcVAHo3+L4XgG8U2WIbQohWIBGfL6X8SLU9FjMOwHVCiO8BSARwkRBinpTydsV2xQwfCIoSIcROAB4ppY4V1MJGCDEVwO8BXCmlrFRtj1UIIVqCgrlZAPYACAD4vpRys1LDLETQiuRdAIeklI8oNsdW6lfkj0sppyk2xRTYR840x+sAkgAsF0KUCCFmqzbICuoDuv8OIA8U9HvfzSJezzgAdwCYVH9tS+pXq4xm8IqcYRhGc3hFzjAMozks5AzDMJrDQs4wDKM5LOQMwzCaw0LOMAyjOSzkDMMwmsNCzjAMozn/D9zaRbf8jOTMAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "2aa21409",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scipy.constants as sc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d68109de",
   "metadata": {},
   "outputs": [],
   "source": [
    "help(sc)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  },
  "latex_envs": {
   "LaTeX_envs_menu_present": true,
   "autoclose": true,
   "autocomplete": true,
   "bibliofile": "biblio.bib",
   "cite_by": "apalike",
   "current_citInitial": 1,
   "eqLabelWithNumbers": true,
   "eqNumInitial": 1,
   "hotkeys": {
    "equation": "Ctrl-E",
    "itemize": "Ctrl-I"
   },
   "labels_anchors": false,
   "latex_user_defs": false,
   "report_style_numbering": false,
   "user_envs_cfg": false
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
